
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../study_index/">
      
      
        <link rel="next" href="../study_locus_overlap/">
      
      
      <link rel="icon" href="../../../assets/imgs/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.18">
    
    
      
        <title>Study Locus - Open Targets Gentropy</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.66ac8b77.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../css/timeago.css">
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../assets/stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="light-blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#gentropy.dataset.study_locus.StudyLocus" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Open Targets Gentropy" class="md-header__button md-logo" aria-label="Open Targets Gentropy" data-md-component="logo">
      
  <img src="../../../assets/imgs/otlogo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Open Targets Gentropy
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Study Locus
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="light-blue"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="light-blue"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/opentargets/gentropy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    opentargets/gentropy
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../installation/" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../howto/_howto/" class="md-tabs__link">
        
  
    
  
  How-to

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../development/_development/" class="md-tabs__link">
        
  
    
  
  Development

      </a>
    </li>
  

      
        
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="../../_python_api/" class="md-tabs__link">
        
  
    
  
  Python API

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Open Targets Gentropy" class="md-nav__button md-logo" aria-label="Open Targets Gentropy" data-md-component="logo">
      
  <img src="../../../assets/imgs/otlogo.png" alt="logo">

    </a>
    Open Targets Gentropy
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/opentargets/gentropy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    opentargets/gentropy
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../howto/_howto/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    How-to
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3" id="__nav_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            How-to
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../howto/command_line/_command_line/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Command line
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1">
            <span class="md-nav__icon md-icon"></span>
            Command line
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/command_line/run_step_in_cli/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run step in CLI
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/command_line/run_step_using_config/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run step using YAML config
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../howto/python_api/_python_api/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Python api
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2">
            <span class="md-nav__icon md-icon"></span>
            Python api
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/python_api/a_creating_spark_session/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Creating a Spark Session
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/python_api/b_create_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Create a dataset
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/python_api/c_applying_methods/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Applying methods
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/python_api/d_inspect_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Inspect a dataset
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../development/_development/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Development
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4" id="__nav_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Development
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../development/contributing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Contributing guidelines
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../development/troubleshooting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Troubleshooting
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../_python_api/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Python API
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5" id="__nav_5_label" tabindex="">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Python API
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../common/_common/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Common
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            Common
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../common/genomic_region/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Genomic Region
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../common/types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Literal Types
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_2" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../_datasets/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Datasets
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_2" id="__nav_5_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            Datasets
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../biosample_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Biosample index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../intervals/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Intervals
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_feature_matrix/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Feature Matrix
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_gold_standard/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Gold Standard
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_prediction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Prediction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ld_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LD Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Study Locus
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Study Locus
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus" class="md-nav__link">
    <span class="md-ellipsis">
      StudyLocus
    </span>
  </a>
  
    <nav class="md-nav" aria-label="StudyLocus">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_credible_sets" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_credible_sets
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_ld" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_ld
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_locus_statistics" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_locus_statistics
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_locus_statistics_boundaries" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_locus_statistics_boundaries
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_study_type" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_study_type
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.assign_confidence" class="md-nav__link">
    <span class="md-ellipsis">
      assign_confidence
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.assign_study_locus_id" class="md-nav__link">
    <span class="md-ellipsis">
      assign_study_locus_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.build_feature_matrix" class="md-nav__link">
    <span class="md-ellipsis">
      build_feature_matrix
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.calculate_credible_set_log10bf" class="md-nav__link">
    <span class="md-ellipsis">
      calculate_credible_set_log10bf
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.clump" class="md-nav__link">
    <span class="md-ellipsis">
      clump
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.exclude_region" class="md-nav__link">
    <span class="md-ellipsis">
      exclude_region
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.filter_credible_set" class="md-nav__link">
    <span class="md-ellipsis">
      filter_credible_set
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.filter_ld_set" class="md-nav__link">
    <span class="md-ellipsis">
      filter_ld_set
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.find_overlaps" class="md-nav__link">
    <span class="md-ellipsis">
      find_overlaps
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.flag_trans_qtls" class="md-nav__link">
    <span class="md-ellipsis">
      flag_trans_qtls
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.get_QC_column_name" class="md-nav__link">
    <span class="md-ellipsis">
      get_QC_column_name
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.get_QC_mappings" class="md-nav__link">
    <span class="md-ellipsis">
      get_QC_mappings
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.get_schema" class="md-nav__link">
    <span class="md-ellipsis">
      get_schema
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.neglog_pvalue" class="md-nav__link">
    <span class="md-ellipsis">
      neglog_pvalue
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.qc_MHC_region" class="md-nav__link">
    <span class="md-ellipsis">
      qc_MHC_region
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.qc_abnormal_pips" class="md-nav__link">
    <span class="md-ellipsis">
      qc_abnormal_pips
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.qc_explained_by_SuSiE" class="md-nav__link">
    <span class="md-ellipsis">
      qc_explained_by_SuSiE
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.qc_redundant_top_hits_from_PICS" class="md-nav__link">
    <span class="md-ellipsis">
      qc_redundant_top_hits_from_PICS
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.unique_variants_in_locus" class="md-nav__link">
    <span class="md-ellipsis">
      unique_variants_in_locus
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.validate_chromosome_label" class="md-nav__link">
    <span class="md-ellipsis">
      validate_chromosome_label
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.validate_lead_pvalue" class="md-nav__link">
    <span class="md-ellipsis">
      validate_lead_pvalue
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.validate_study" class="md-nav__link">
    <span class="md-ellipsis">
      validate_study
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.validate_unique_study_locus_id" class="md-nav__link">
    <span class="md-ellipsis">
      validate_unique_study_locus_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.validate_variant_identifiers" class="md-nav__link">
    <span class="md-ellipsis">
      validate_variant_identifiers
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.window_based_clumping" class="md-nav__link">
    <span class="md-ellipsis">
      window_based_clumping
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.FinemappingMethod" class="md-nav__link">
    <span class="md-ellipsis">
      FinemappingMethod
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocusQualityCheck" class="md-nav__link">
    <span class="md-ellipsis">
      StudyLocusQualityCheck
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.CredibleInterval" class="md-nav__link">
    <span class="md-ellipsis">
      CredibleInterval
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#schema" class="md-nav__link">
    <span class="md-ellipsis">
      Schema
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../study_locus_overlap/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Locus Overlap
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../summary_statistics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../summary_statistics_qc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../target_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Target Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variant_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_2_15" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../l2g_features/_l2g_feature/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    L2g features
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_2_15" id="__nav_5_2_15_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_2_15_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_2_15">
            <span class="md-nav__icon md-icon"></span>
            L2g features
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_features/colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_features/distance/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From distance
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_features/other/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Other features
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_features/vep/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From VEP
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/_datasources/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Datasources
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3" id="__nav_5_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3">
            <span class="md-nav__icon md-icon"></span>
            Datasources
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/biosample_ontologies/_cell_ontology/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Biosample ontologies
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_1" id="__nav_5_3_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_1">
            <span class="md-nav__icon md-icon"></span>
            Biosample ontologies
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/biosample_ontologies/_uberon/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Uberon
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/ensembl/_ensembl/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Ensembl
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_2" id="__nav_5_3_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_2">
            <span class="md-nav__icon md-icon"></span>
            Ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/ensembl/variant_effect_predictor_parser/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant effector parser
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/eqtl_catalogue/_eqtl_catalogue/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Eqtl catalogue
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_3" id="__nav_5_3_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_3">
            <span class="md-nav__icon md-icon"></span>
            Eqtl catalogue
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/eqtl_catalogue/finemapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Fine mapping results
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/eqtl_catalogue/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/eqtl_catalogue/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Stats
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/finngen/_finngen/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Finngen
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_4" id="__nav_5_3_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_4">
            <span class="md-nav__icon md-icon"></span>
            Finngen
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/finngen/finemapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Finemapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/finngen/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/finngen/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/gnomad/_gnomad/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Gnomad
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_5" id="__nav_5_3_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_5">
            <span class="md-nav__icon md-icon"></span>
            Gnomad
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gnomad/gnomad_ld/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LD Matrix
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gnomad/gnomad_variants/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variants
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_6" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/gwas_catalog/_gwas_catalog/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Gwas catalog
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_6" id="__nav_5_3_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_6">
            <span class="md-nav__icon md-icon"></span>
            Gwas catalog
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/associations/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Associations
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/study_splitter/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Splitter
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/summary_statistics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary statistics
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_7" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/intervals/_intervals/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Intervals
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_7" id="__nav_5_3_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_7">
            <span class="md-nav__icon md-icon"></span>
            Intervals
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/andersson/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Andersson et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/javierre/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Javierre et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/jung/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Jung et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/thurman/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Thurman et al.
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/open_targets/_open_targets/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Open targets
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_8" id="__nav_5_3_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_8">
            <span class="md-nav__icon md-icon"></span>
            Open targets
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/open_targets/l2g_gold_standard/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Gold Standard
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_9" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/ukb_ppp_eur/_ukb_ppp_eur/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Ukb ppp eur
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_9" id="__nav_5_3_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_9">
            <span class="md-nav__icon md-icon"></span>
            Ukb ppp eur
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/ukb_ppp_eur/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/ukb_ppp_eur/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../methods/_methods/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Methods
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4" id="__nav_5_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4">
            <span class="md-nav__icon md-icon"></span>
            Methods
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/carma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CARMA
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/coloc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Coloc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/ecaviar/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    eCAVIAR
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/ld_annotator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LDAnnotator
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/pics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PICS
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/sumstat_imputation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics Imputation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/sumstat_quality_controls/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    QC of GWAS Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/susie_inf/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SuSiE-inf
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4_10" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../methods/l2g/_l2g/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    L2g
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4_10" id="__nav_5_4_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_4_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4_10">
            <span class="md-nav__icon md-icon"></span>
            L2g
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/l2g/feature_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Feature Factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/l2g/model/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/l2g/trainer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Trainer
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../steps/_steps/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Steps
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_5" id="__nav_5_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_5">
            <span class="md-nav__icon md-icon"></span>
            Steps
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/biosample_index_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    biosample_index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/credible_set_qc_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    credible_set_qc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/eqtl_catalogue/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    eQTL Catalogue
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/finngen_studies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    finngen_studies
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/finngen_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    finngen_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_curation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_study_curation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_study_inclusion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_top_hits/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GWAS Catalog Top Hits Ingestion Step
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/l2g/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Locus to Gene (L2G)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/ld_clump/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ld_based_clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/ld_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GnomAD Linkage data ingestion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/locus_breaker_clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    locus_breaker_clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/pics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    pics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/study_locus_validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study-Locus Validation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/study_validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Validation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/summary_statistics_qc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    summary_statistics_qc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/ukb_ppp_eur_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ukb_ppp_eur_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/variant_annotation_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GnomAD variant data ingestion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/variant_index_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    variant_index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/window_based_clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    window_based_clumping
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus" class="md-nav__link">
    <span class="md-ellipsis">
      StudyLocus
    </span>
  </a>
  
    <nav class="md-nav" aria-label="StudyLocus">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_credible_sets" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_credible_sets
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_ld" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_ld
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_locus_statistics" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_locus_statistics
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_locus_statistics_boundaries" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_locus_statistics_boundaries
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_study_type" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_study_type
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.assign_confidence" class="md-nav__link">
    <span class="md-ellipsis">
      assign_confidence
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.assign_study_locus_id" class="md-nav__link">
    <span class="md-ellipsis">
      assign_study_locus_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.build_feature_matrix" class="md-nav__link">
    <span class="md-ellipsis">
      build_feature_matrix
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.calculate_credible_set_log10bf" class="md-nav__link">
    <span class="md-ellipsis">
      calculate_credible_set_log10bf
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.clump" class="md-nav__link">
    <span class="md-ellipsis">
      clump
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.exclude_region" class="md-nav__link">
    <span class="md-ellipsis">
      exclude_region
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.filter_credible_set" class="md-nav__link">
    <span class="md-ellipsis">
      filter_credible_set
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.filter_ld_set" class="md-nav__link">
    <span class="md-ellipsis">
      filter_ld_set
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.find_overlaps" class="md-nav__link">
    <span class="md-ellipsis">
      find_overlaps
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.flag_trans_qtls" class="md-nav__link">
    <span class="md-ellipsis">
      flag_trans_qtls
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.get_QC_column_name" class="md-nav__link">
    <span class="md-ellipsis">
      get_QC_column_name
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.get_QC_mappings" class="md-nav__link">
    <span class="md-ellipsis">
      get_QC_mappings
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.get_schema" class="md-nav__link">
    <span class="md-ellipsis">
      get_schema
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.neglog_pvalue" class="md-nav__link">
    <span class="md-ellipsis">
      neglog_pvalue
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.qc_MHC_region" class="md-nav__link">
    <span class="md-ellipsis">
      qc_MHC_region
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.qc_abnormal_pips" class="md-nav__link">
    <span class="md-ellipsis">
      qc_abnormal_pips
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.qc_explained_by_SuSiE" class="md-nav__link">
    <span class="md-ellipsis">
      qc_explained_by_SuSiE
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.qc_redundant_top_hits_from_PICS" class="md-nav__link">
    <span class="md-ellipsis">
      qc_redundant_top_hits_from_PICS
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.unique_variants_in_locus" class="md-nav__link">
    <span class="md-ellipsis">
      unique_variants_in_locus
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.validate_chromosome_label" class="md-nav__link">
    <span class="md-ellipsis">
      validate_chromosome_label
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.validate_lead_pvalue" class="md-nav__link">
    <span class="md-ellipsis">
      validate_lead_pvalue
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.validate_study" class="md-nav__link">
    <span class="md-ellipsis">
      validate_study
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.validate_unique_study_locus_id" class="md-nav__link">
    <span class="md-ellipsis">
      validate_unique_study_locus_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.validate_variant_identifiers" class="md-nav__link">
    <span class="md-ellipsis">
      validate_variant_identifiers
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.window_based_clumping" class="md-nav__link">
    <span class="md-ellipsis">
      window_based_clumping
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.FinemappingMethod" class="md-nav__link">
    <span class="md-ellipsis">
      FinemappingMethod
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocusQualityCheck" class="md-nav__link">
    <span class="md-ellipsis">
      StudyLocusQualityCheck
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.CredibleInterval" class="md-nav__link">
    <span class="md-ellipsis">
      CredibleInterval
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#schema" class="md-nav__link">
    <span class="md-ellipsis">
      Schema
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>Study Locus</h1>

<div class="doc doc-object doc-class">



<h2 id="gentropy.dataset.study_locus.StudyLocus" class="doc doc-heading">
            <code>gentropy.dataset.study_locus.StudyLocus</code>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-dataclass"><code>dataclass</code></small>
  </span>

<a href="#gentropy.dataset.study_locus.StudyLocus" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="gentropy.dataset.dataset.Dataset" href="../_datasets/#gentropy.dataset.dataset.Dataset">Dataset</a></code></p>


        <p>Study-Locus dataset.</p>
<p>This dataset captures associations between study/traits and a genetic loci as provided by finemapping methods.</p>






              <details class="quote">
                <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 153</span>
<span class="normal"> 154</span>
<span class="normal"> 155</span>
<span class="normal"> 156</span>
<span class="normal"> 157</span>
<span class="normal"> 158</span>
<span class="normal"> 159</span>
<span class="normal"> 160</span>
<span class="normal"> 161</span>
<span class="normal"> 162</span>
<span class="normal"> 163</span>
<span class="normal"> 164</span>
<span class="normal"> 165</span>
<span class="normal"> 166</span>
<span class="normal"> 167</span>
<span class="normal"> 168</span>
<span class="normal"> 169</span>
<span class="normal"> 170</span>
<span class="normal"> 171</span>
<span class="normal"> 172</span>
<span class="normal"> 173</span>
<span class="normal"> 174</span>
<span class="normal"> 175</span>
<span class="normal"> 176</span>
<span class="normal"> 177</span>
<span class="normal"> 178</span>
<span class="normal"> 179</span>
<span class="normal"> 180</span>
<span class="normal"> 181</span>
<span class="normal"> 182</span>
<span class="normal"> 183</span>
<span class="normal"> 184</span>
<span class="normal"> 185</span>
<span class="normal"> 186</span>
<span class="normal"> 187</span>
<span class="normal"> 188</span>
<span class="normal"> 189</span>
<span class="normal"> 190</span>
<span class="normal"> 191</span>
<span class="normal"> 192</span>
<span class="normal"> 193</span>
<span class="normal"> 194</span>
<span class="normal"> 195</span>
<span class="normal"> 196</span>
<span class="normal"> 197</span>
<span class="normal"> 198</span>
<span class="normal"> 199</span>
<span class="normal"> 200</span>
<span class="normal"> 201</span>
<span class="normal"> 202</span>
<span class="normal"> 203</span>
<span class="normal"> 204</span>
<span class="normal"> 205</span>
<span class="normal"> 206</span>
<span class="normal"> 207</span>
<span class="normal"> 208</span>
<span class="normal"> 209</span>
<span class="normal"> 210</span>
<span class="normal"> 211</span>
<span class="normal"> 212</span>
<span class="normal"> 213</span>
<span class="normal"> 214</span>
<span class="normal"> 215</span>
<span class="normal"> 216</span>
<span class="normal"> 217</span>
<span class="normal"> 218</span>
<span class="normal"> 219</span>
<span class="normal"> 220</span>
<span class="normal"> 221</span>
<span class="normal"> 222</span>
<span class="normal"> 223</span>
<span class="normal"> 224</span>
<span class="normal"> 225</span>
<span class="normal"> 226</span>
<span class="normal"> 227</span>
<span class="normal"> 228</span>
<span class="normal"> 229</span>
<span class="normal"> 230</span>
<span class="normal"> 231</span>
<span class="normal"> 232</span>
<span class="normal"> 233</span>
<span class="normal"> 234</span>
<span class="normal"> 235</span>
<span class="normal"> 236</span>
<span class="normal"> 237</span>
<span class="normal"> 238</span>
<span class="normal"> 239</span>
<span class="normal"> 240</span>
<span class="normal"> 241</span>
<span class="normal"> 242</span>
<span class="normal"> 243</span>
<span class="normal"> 244</span>
<span class="normal"> 245</span>
<span class="normal"> 246</span>
<span class="normal"> 247</span>
<span class="normal"> 248</span>
<span class="normal"> 249</span>
<span class="normal"> 250</span>
<span class="normal"> 251</span>
<span class="normal"> 252</span>
<span class="normal"> 253</span>
<span class="normal"> 254</span>
<span class="normal"> 255</span>
<span class="normal"> 256</span>
<span class="normal"> 257</span>
<span class="normal"> 258</span>
<span class="normal"> 259</span>
<span class="normal"> 260</span>
<span class="normal"> 261</span>
<span class="normal"> 262</span>
<span class="normal"> 263</span>
<span class="normal"> 264</span>
<span class="normal"> 265</span>
<span class="normal"> 266</span>
<span class="normal"> 267</span>
<span class="normal"> 268</span>
<span class="normal"> 269</span>
<span class="normal"> 270</span>
<span class="normal"> 271</span>
<span class="normal"> 272</span>
<span class="normal"> 273</span>
<span class="normal"> 274</span>
<span class="normal"> 275</span>
<span class="normal"> 276</span>
<span class="normal"> 277</span>
<span class="normal"> 278</span>
<span class="normal"> 279</span>
<span class="normal"> 280</span>
<span class="normal"> 281</span>
<span class="normal"> 282</span>
<span class="normal"> 283</span>
<span class="normal"> 284</span>
<span class="normal"> 285</span>
<span class="normal"> 286</span>
<span class="normal"> 287</span>
<span class="normal"> 288</span>
<span class="normal"> 289</span>
<span class="normal"> 290</span>
<span class="normal"> 291</span>
<span class="normal"> 292</span>
<span class="normal"> 293</span>
<span class="normal"> 294</span>
<span class="normal"> 295</span>
<span class="normal"> 296</span>
<span class="normal"> 297</span>
<span class="normal"> 298</span>
<span class="normal"> 299</span>
<span class="normal"> 300</span>
<span class="normal"> 301</span>
<span class="normal"> 302</span>
<span class="normal"> 303</span>
<span class="normal"> 304</span>
<span class="normal"> 305</span>
<span class="normal"> 306</span>
<span class="normal"> 307</span>
<span class="normal"> 308</span>
<span class="normal"> 309</span>
<span class="normal"> 310</span>
<span class="normal"> 311</span>
<span class="normal"> 312</span>
<span class="normal"> 313</span>
<span class="normal"> 314</span>
<span class="normal"> 315</span>
<span class="normal"> 316</span>
<span class="normal"> 317</span>
<span class="normal"> 318</span>
<span class="normal"> 319</span>
<span class="normal"> 320</span>
<span class="normal"> 321</span>
<span class="normal"> 322</span>
<span class="normal"> 323</span>
<span class="normal"> 324</span>
<span class="normal"> 325</span>
<span class="normal"> 326</span>
<span class="normal"> 327</span>
<span class="normal"> 328</span>
<span class="normal"> 329</span>
<span class="normal"> 330</span>
<span class="normal"> 331</span>
<span class="normal"> 332</span>
<span class="normal"> 333</span>
<span class="normal"> 334</span>
<span class="normal"> 335</span>
<span class="normal"> 336</span>
<span class="normal"> 337</span>
<span class="normal"> 338</span>
<span class="normal"> 339</span>
<span class="normal"> 340</span>
<span class="normal"> 341</span>
<span class="normal"> 342</span>
<span class="normal"> 343</span>
<span class="normal"> 344</span>
<span class="normal"> 345</span>
<span class="normal"> 346</span>
<span class="normal"> 347</span>
<span class="normal"> 348</span>
<span class="normal"> 349</span>
<span class="normal"> 350</span>
<span class="normal"> 351</span>
<span class="normal"> 352</span>
<span class="normal"> 353</span>
<span class="normal"> 354</span>
<span class="normal"> 355</span>
<span class="normal"> 356</span>
<span class="normal"> 357</span>
<span class="normal"> 358</span>
<span class="normal"> 359</span>
<span class="normal"> 360</span>
<span class="normal"> 361</span>
<span class="normal"> 362</span>
<span class="normal"> 363</span>
<span class="normal"> 364</span>
<span class="normal"> 365</span>
<span class="normal"> 366</span>
<span class="normal"> 367</span>
<span class="normal"> 368</span>
<span class="normal"> 369</span>
<span class="normal"> 370</span>
<span class="normal"> 371</span>
<span class="normal"> 372</span>
<span class="normal"> 373</span>
<span class="normal"> 374</span>
<span class="normal"> 375</span>
<span class="normal"> 376</span>
<span class="normal"> 377</span>
<span class="normal"> 378</span>
<span class="normal"> 379</span>
<span class="normal"> 380</span>
<span class="normal"> 381</span>
<span class="normal"> 382</span>
<span class="normal"> 383</span>
<span class="normal"> 384</span>
<span class="normal"> 385</span>
<span class="normal"> 386</span>
<span class="normal"> 387</span>
<span class="normal"> 388</span>
<span class="normal"> 389</span>
<span class="normal"> 390</span>
<span class="normal"> 391</span>
<span class="normal"> 392</span>
<span class="normal"> 393</span>
<span class="normal"> 394</span>
<span class="normal"> 395</span>
<span class="normal"> 396</span>
<span class="normal"> 397</span>
<span class="normal"> 398</span>
<span class="normal"> 399</span>
<span class="normal"> 400</span>
<span class="normal"> 401</span>
<span class="normal"> 402</span>
<span class="normal"> 403</span>
<span class="normal"> 404</span>
<span class="normal"> 405</span>
<span class="normal"> 406</span>
<span class="normal"> 407</span>
<span class="normal"> 408</span>
<span class="normal"> 409</span>
<span class="normal"> 410</span>
<span class="normal"> 411</span>
<span class="normal"> 412</span>
<span class="normal"> 413</span>
<span class="normal"> 414</span>
<span class="normal"> 415</span>
<span class="normal"> 416</span>
<span class="normal"> 417</span>
<span class="normal"> 418</span>
<span class="normal"> 419</span>
<span class="normal"> 420</span>
<span class="normal"> 421</span>
<span class="normal"> 422</span>
<span class="normal"> 423</span>
<span class="normal"> 424</span>
<span class="normal"> 425</span>
<span class="normal"> 426</span>
<span class="normal"> 427</span>
<span class="normal"> 428</span>
<span class="normal"> 429</span>
<span class="normal"> 430</span>
<span class="normal"> 431</span>
<span class="normal"> 432</span>
<span class="normal"> 433</span>
<span class="normal"> 434</span>
<span class="normal"> 435</span>
<span class="normal"> 436</span>
<span class="normal"> 437</span>
<span class="normal"> 438</span>
<span class="normal"> 439</span>
<span class="normal"> 440</span>
<span class="normal"> 441</span>
<span class="normal"> 442</span>
<span class="normal"> 443</span>
<span class="normal"> 444</span>
<span class="normal"> 445</span>
<span class="normal"> 446</span>
<span class="normal"> 447</span>
<span class="normal"> 448</span>
<span class="normal"> 449</span>
<span class="normal"> 450</span>
<span class="normal"> 451</span>
<span class="normal"> 452</span>
<span class="normal"> 453</span>
<span class="normal"> 454</span>
<span class="normal"> 455</span>
<span class="normal"> 456</span>
<span class="normal"> 457</span>
<span class="normal"> 458</span>
<span class="normal"> 459</span>
<span class="normal"> 460</span>
<span class="normal"> 461</span>
<span class="normal"> 462</span>
<span class="normal"> 463</span>
<span class="normal"> 464</span>
<span class="normal"> 465</span>
<span class="normal"> 466</span>
<span class="normal"> 467</span>
<span class="normal"> 468</span>
<span class="normal"> 469</span>
<span class="normal"> 470</span>
<span class="normal"> 471</span>
<span class="normal"> 472</span>
<span class="normal"> 473</span>
<span class="normal"> 474</span>
<span class="normal"> 475</span>
<span class="normal"> 476</span>
<span class="normal"> 477</span>
<span class="normal"> 478</span>
<span class="normal"> 479</span>
<span class="normal"> 480</span>
<span class="normal"> 481</span>
<span class="normal"> 482</span>
<span class="normal"> 483</span>
<span class="normal"> 484</span>
<span class="normal"> 485</span>
<span class="normal"> 486</span>
<span class="normal"> 487</span>
<span class="normal"> 488</span>
<span class="normal"> 489</span>
<span class="normal"> 490</span>
<span class="normal"> 491</span>
<span class="normal"> 492</span>
<span class="normal"> 493</span>
<span class="normal"> 494</span>
<span class="normal"> 495</span>
<span class="normal"> 496</span>
<span class="normal"> 497</span>
<span class="normal"> 498</span>
<span class="normal"> 499</span>
<span class="normal"> 500</span>
<span class="normal"> 501</span>
<span class="normal"> 502</span>
<span class="normal"> 503</span>
<span class="normal"> 504</span>
<span class="normal"> 505</span>
<span class="normal"> 506</span>
<span class="normal"> 507</span>
<span class="normal"> 508</span>
<span class="normal"> 509</span>
<span class="normal"> 510</span>
<span class="normal"> 511</span>
<span class="normal"> 512</span>
<span class="normal"> 513</span>
<span class="normal"> 514</span>
<span class="normal"> 515</span>
<span class="normal"> 516</span>
<span class="normal"> 517</span>
<span class="normal"> 518</span>
<span class="normal"> 519</span>
<span class="normal"> 520</span>
<span class="normal"> 521</span>
<span class="normal"> 522</span>
<span class="normal"> 523</span>
<span class="normal"> 524</span>
<span class="normal"> 525</span>
<span class="normal"> 526</span>
<span class="normal"> 527</span>
<span class="normal"> 528</span>
<span class="normal"> 529</span>
<span class="normal"> 530</span>
<span class="normal"> 531</span>
<span class="normal"> 532</span>
<span class="normal"> 533</span>
<span class="normal"> 534</span>
<span class="normal"> 535</span>
<span class="normal"> 536</span>
<span class="normal"> 537</span>
<span class="normal"> 538</span>
<span class="normal"> 539</span>
<span class="normal"> 540</span>
<span class="normal"> 541</span>
<span class="normal"> 542</span>
<span class="normal"> 543</span>
<span class="normal"> 544</span>
<span class="normal"> 545</span>
<span class="normal"> 546</span>
<span class="normal"> 547</span>
<span class="normal"> 548</span>
<span class="normal"> 549</span>
<span class="normal"> 550</span>
<span class="normal"> 551</span>
<span class="normal"> 552</span>
<span class="normal"> 553</span>
<span class="normal"> 554</span>
<span class="normal"> 555</span>
<span class="normal"> 556</span>
<span class="normal"> 557</span>
<span class="normal"> 558</span>
<span class="normal"> 559</span>
<span class="normal"> 560</span>
<span class="normal"> 561</span>
<span class="normal"> 562</span>
<span class="normal"> 563</span>
<span class="normal"> 564</span>
<span class="normal"> 565</span>
<span class="normal"> 566</span>
<span class="normal"> 567</span>
<span class="normal"> 568</span>
<span class="normal"> 569</span>
<span class="normal"> 570</span>
<span class="normal"> 571</span>
<span class="normal"> 572</span>
<span class="normal"> 573</span>
<span class="normal"> 574</span>
<span class="normal"> 575</span>
<span class="normal"> 576</span>
<span class="normal"> 577</span>
<span class="normal"> 578</span>
<span class="normal"> 579</span>
<span class="normal"> 580</span>
<span class="normal"> 581</span>
<span class="normal"> 582</span>
<span class="normal"> 583</span>
<span class="normal"> 584</span>
<span class="normal"> 585</span>
<span class="normal"> 586</span>
<span class="normal"> 587</span>
<span class="normal"> 588</span>
<span class="normal"> 589</span>
<span class="normal"> 590</span>
<span class="normal"> 591</span>
<span class="normal"> 592</span>
<span class="normal"> 593</span>
<span class="normal"> 594</span>
<span class="normal"> 595</span>
<span class="normal"> 596</span>
<span class="normal"> 597</span>
<span class="normal"> 598</span>
<span class="normal"> 599</span>
<span class="normal"> 600</span>
<span class="normal"> 601</span>
<span class="normal"> 602</span>
<span class="normal"> 603</span>
<span class="normal"> 604</span>
<span class="normal"> 605</span>
<span class="normal"> 606</span>
<span class="normal"> 607</span>
<span class="normal"> 608</span>
<span class="normal"> 609</span>
<span class="normal"> 610</span>
<span class="normal"> 611</span>
<span class="normal"> 612</span>
<span class="normal"> 613</span>
<span class="normal"> 614</span>
<span class="normal"> 615</span>
<span class="normal"> 616</span>
<span class="normal"> 617</span>
<span class="normal"> 618</span>
<span class="normal"> 619</span>
<span class="normal"> 620</span>
<span class="normal"> 621</span>
<span class="normal"> 622</span>
<span class="normal"> 623</span>
<span class="normal"> 624</span>
<span class="normal"> 625</span>
<span class="normal"> 626</span>
<span class="normal"> 627</span>
<span class="normal"> 628</span>
<span class="normal"> 629</span>
<span class="normal"> 630</span>
<span class="normal"> 631</span>
<span class="normal"> 632</span>
<span class="normal"> 633</span>
<span class="normal"> 634</span>
<span class="normal"> 635</span>
<span class="normal"> 636</span>
<span class="normal"> 637</span>
<span class="normal"> 638</span>
<span class="normal"> 639</span>
<span class="normal"> 640</span>
<span class="normal"> 641</span>
<span class="normal"> 642</span>
<span class="normal"> 643</span>
<span class="normal"> 644</span>
<span class="normal"> 645</span>
<span class="normal"> 646</span>
<span class="normal"> 647</span>
<span class="normal"> 648</span>
<span class="normal"> 649</span>
<span class="normal"> 650</span>
<span class="normal"> 651</span>
<span class="normal"> 652</span>
<span class="normal"> 653</span>
<span class="normal"> 654</span>
<span class="normal"> 655</span>
<span class="normal"> 656</span>
<span class="normal"> 657</span>
<span class="normal"> 658</span>
<span class="normal"> 659</span>
<span class="normal"> 660</span>
<span class="normal"> 661</span>
<span class="normal"> 662</span>
<span class="normal"> 663</span>
<span class="normal"> 664</span>
<span class="normal"> 665</span>
<span class="normal"> 666</span>
<span class="normal"> 667</span>
<span class="normal"> 668</span>
<span class="normal"> 669</span>
<span class="normal"> 670</span>
<span class="normal"> 671</span>
<span class="normal"> 672</span>
<span class="normal"> 673</span>
<span class="normal"> 674</span>
<span class="normal"> 675</span>
<span class="normal"> 676</span>
<span class="normal"> 677</span>
<span class="normal"> 678</span>
<span class="normal"> 679</span>
<span class="normal"> 680</span>
<span class="normal"> 681</span>
<span class="normal"> 682</span>
<span class="normal"> 683</span>
<span class="normal"> 684</span>
<span class="normal"> 685</span>
<span class="normal"> 686</span>
<span class="normal"> 687</span>
<span class="normal"> 688</span>
<span class="normal"> 689</span>
<span class="normal"> 690</span>
<span class="normal"> 691</span>
<span class="normal"> 692</span>
<span class="normal"> 693</span>
<span class="normal"> 694</span>
<span class="normal"> 695</span>
<span class="normal"> 696</span>
<span class="normal"> 697</span>
<span class="normal"> 698</span>
<span class="normal"> 699</span>
<span class="normal"> 700</span>
<span class="normal"> 701</span>
<span class="normal"> 702</span>
<span class="normal"> 703</span>
<span class="normal"> 704</span>
<span class="normal"> 705</span>
<span class="normal"> 706</span>
<span class="normal"> 707</span>
<span class="normal"> 708</span>
<span class="normal"> 709</span>
<span class="normal"> 710</span>
<span class="normal"> 711</span>
<span class="normal"> 712</span>
<span class="normal"> 713</span>
<span class="normal"> 714</span>
<span class="normal"> 715</span>
<span class="normal"> 716</span>
<span class="normal"> 717</span>
<span class="normal"> 718</span>
<span class="normal"> 719</span>
<span class="normal"> 720</span>
<span class="normal"> 721</span>
<span class="normal"> 722</span>
<span class="normal"> 723</span>
<span class="normal"> 724</span>
<span class="normal"> 725</span>
<span class="normal"> 726</span>
<span class="normal"> 727</span>
<span class="normal"> 728</span>
<span class="normal"> 729</span>
<span class="normal"> 730</span>
<span class="normal"> 731</span>
<span class="normal"> 732</span>
<span class="normal"> 733</span>
<span class="normal"> 734</span>
<span class="normal"> 735</span>
<span class="normal"> 736</span>
<span class="normal"> 737</span>
<span class="normal"> 738</span>
<span class="normal"> 739</span>
<span class="normal"> 740</span>
<span class="normal"> 741</span>
<span class="normal"> 742</span>
<span class="normal"> 743</span>
<span class="normal"> 744</span>
<span class="normal"> 745</span>
<span class="normal"> 746</span>
<span class="normal"> 747</span>
<span class="normal"> 748</span>
<span class="normal"> 749</span>
<span class="normal"> 750</span>
<span class="normal"> 751</span>
<span class="normal"> 752</span>
<span class="normal"> 753</span>
<span class="normal"> 754</span>
<span class="normal"> 755</span>
<span class="normal"> 756</span>
<span class="normal"> 757</span>
<span class="normal"> 758</span>
<span class="normal"> 759</span>
<span class="normal"> 760</span>
<span class="normal"> 761</span>
<span class="normal"> 762</span>
<span class="normal"> 763</span>
<span class="normal"> 764</span>
<span class="normal"> 765</span>
<span class="normal"> 766</span>
<span class="normal"> 767</span>
<span class="normal"> 768</span>
<span class="normal"> 769</span>
<span class="normal"> 770</span>
<span class="normal"> 771</span>
<span class="normal"> 772</span>
<span class="normal"> 773</span>
<span class="normal"> 774</span>
<span class="normal"> 775</span>
<span class="normal"> 776</span>
<span class="normal"> 777</span>
<span class="normal"> 778</span>
<span class="normal"> 779</span>
<span class="normal"> 780</span>
<span class="normal"> 781</span>
<span class="normal"> 782</span>
<span class="normal"> 783</span>
<span class="normal"> 784</span>
<span class="normal"> 785</span>
<span class="normal"> 786</span>
<span class="normal"> 787</span>
<span class="normal"> 788</span>
<span class="normal"> 789</span>
<span class="normal"> 790</span>
<span class="normal"> 791</span>
<span class="normal"> 792</span>
<span class="normal"> 793</span>
<span class="normal"> 794</span>
<span class="normal"> 795</span>
<span class="normal"> 796</span>
<span class="normal"> 797</span>
<span class="normal"> 798</span>
<span class="normal"> 799</span>
<span class="normal"> 800</span>
<span class="normal"> 801</span>
<span class="normal"> 802</span>
<span class="normal"> 803</span>
<span class="normal"> 804</span>
<span class="normal"> 805</span>
<span class="normal"> 806</span>
<span class="normal"> 807</span>
<span class="normal"> 808</span>
<span class="normal"> 809</span>
<span class="normal"> 810</span>
<span class="normal"> 811</span>
<span class="normal"> 812</span>
<span class="normal"> 813</span>
<span class="normal"> 814</span>
<span class="normal"> 815</span>
<span class="normal"> 816</span>
<span class="normal"> 817</span>
<span class="normal"> 818</span>
<span class="normal"> 819</span>
<span class="normal"> 820</span>
<span class="normal"> 821</span>
<span class="normal"> 822</span>
<span class="normal"> 823</span>
<span class="normal"> 824</span>
<span class="normal"> 825</span>
<span class="normal"> 826</span>
<span class="normal"> 827</span>
<span class="normal"> 828</span>
<span class="normal"> 829</span>
<span class="normal"> 830</span>
<span class="normal"> 831</span>
<span class="normal"> 832</span>
<span class="normal"> 833</span>
<span class="normal"> 834</span>
<span class="normal"> 835</span>
<span class="normal"> 836</span>
<span class="normal"> 837</span>
<span class="normal"> 838</span>
<span class="normal"> 839</span>
<span class="normal"> 840</span>
<span class="normal"> 841</span>
<span class="normal"> 842</span>
<span class="normal"> 843</span>
<span class="normal"> 844</span>
<span class="normal"> 845</span>
<span class="normal"> 846</span>
<span class="normal"> 847</span>
<span class="normal"> 848</span>
<span class="normal"> 849</span>
<span class="normal"> 850</span>
<span class="normal"> 851</span>
<span class="normal"> 852</span>
<span class="normal"> 853</span>
<span class="normal"> 854</span>
<span class="normal"> 855</span>
<span class="normal"> 856</span>
<span class="normal"> 857</span>
<span class="normal"> 858</span>
<span class="normal"> 859</span>
<span class="normal"> 860</span>
<span class="normal"> 861</span>
<span class="normal"> 862</span>
<span class="normal"> 863</span>
<span class="normal"> 864</span>
<span class="normal"> 865</span>
<span class="normal"> 866</span>
<span class="normal"> 867</span>
<span class="normal"> 868</span>
<span class="normal"> 869</span>
<span class="normal"> 870</span>
<span class="normal"> 871</span>
<span class="normal"> 872</span>
<span class="normal"> 873</span>
<span class="normal"> 874</span>
<span class="normal"> 875</span>
<span class="normal"> 876</span>
<span class="normal"> 877</span>
<span class="normal"> 878</span>
<span class="normal"> 879</span>
<span class="normal"> 880</span>
<span class="normal"> 881</span>
<span class="normal"> 882</span>
<span class="normal"> 883</span>
<span class="normal"> 884</span>
<span class="normal"> 885</span>
<span class="normal"> 886</span>
<span class="normal"> 887</span>
<span class="normal"> 888</span>
<span class="normal"> 889</span>
<span class="normal"> 890</span>
<span class="normal"> 891</span>
<span class="normal"> 892</span>
<span class="normal"> 893</span>
<span class="normal"> 894</span>
<span class="normal"> 895</span>
<span class="normal"> 896</span>
<span class="normal"> 897</span>
<span class="normal"> 898</span>
<span class="normal"> 899</span>
<span class="normal"> 900</span>
<span class="normal"> 901</span>
<span class="normal"> 902</span>
<span class="normal"> 903</span>
<span class="normal"> 904</span>
<span class="normal"> 905</span>
<span class="normal"> 906</span>
<span class="normal"> 907</span>
<span class="normal"> 908</span>
<span class="normal"> 909</span>
<span class="normal"> 910</span>
<span class="normal"> 911</span>
<span class="normal"> 912</span>
<span class="normal"> 913</span>
<span class="normal"> 914</span>
<span class="normal"> 915</span>
<span class="normal"> 916</span>
<span class="normal"> 917</span>
<span class="normal"> 918</span>
<span class="normal"> 919</span>
<span class="normal"> 920</span>
<span class="normal"> 921</span>
<span class="normal"> 922</span>
<span class="normal"> 923</span>
<span class="normal"> 924</span>
<span class="normal"> 925</span>
<span class="normal"> 926</span>
<span class="normal"> 927</span>
<span class="normal"> 928</span>
<span class="normal"> 929</span>
<span class="normal"> 930</span>
<span class="normal"> 931</span>
<span class="normal"> 932</span>
<span class="normal"> 933</span>
<span class="normal"> 934</span>
<span class="normal"> 935</span>
<span class="normal"> 936</span>
<span class="normal"> 937</span>
<span class="normal"> 938</span>
<span class="normal"> 939</span>
<span class="normal"> 940</span>
<span class="normal"> 941</span>
<span class="normal"> 942</span>
<span class="normal"> 943</span>
<span class="normal"> 944</span>
<span class="normal"> 945</span>
<span class="normal"> 946</span>
<span class="normal"> 947</span>
<span class="normal"> 948</span>
<span class="normal"> 949</span>
<span class="normal"> 950</span>
<span class="normal"> 951</span>
<span class="normal"> 952</span>
<span class="normal"> 953</span>
<span class="normal"> 954</span>
<span class="normal"> 955</span>
<span class="normal"> 956</span>
<span class="normal"> 957</span>
<span class="normal"> 958</span>
<span class="normal"> 959</span>
<span class="normal"> 960</span>
<span class="normal"> 961</span>
<span class="normal"> 962</span>
<span class="normal"> 963</span>
<span class="normal"> 964</span>
<span class="normal"> 965</span>
<span class="normal"> 966</span>
<span class="normal"> 967</span>
<span class="normal"> 968</span>
<span class="normal"> 969</span>
<span class="normal"> 970</span>
<span class="normal"> 971</span>
<span class="normal"> 972</span>
<span class="normal"> 973</span>
<span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span>
<span class="normal">1124</span>
<span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span>
<span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span>
<span class="normal">1161</span>
<span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span>
<span class="normal">1174</span>
<span class="normal">1175</span>
<span class="normal">1176</span>
<span class="normal">1177</span>
<span class="normal">1178</span>
<span class="normal">1179</span>
<span class="normal">1180</span>
<span class="normal">1181</span>
<span class="normal">1182</span>
<span class="normal">1183</span>
<span class="normal">1184</span>
<span class="normal">1185</span>
<span class="normal">1186</span>
<span class="normal">1187</span>
<span class="normal">1188</span>
<span class="normal">1189</span>
<span class="normal">1190</span>
<span class="normal">1191</span>
<span class="normal">1192</span>
<span class="normal">1193</span>
<span class="normal">1194</span>
<span class="normal">1195</span>
<span class="normal">1196</span>
<span class="normal">1197</span>
<span class="normal">1198</span>
<span class="normal">1199</span>
<span class="normal">1200</span>
<span class="normal">1201</span>
<span class="normal">1202</span>
<span class="normal">1203</span>
<span class="normal">1204</span>
<span class="normal">1205</span>
<span class="normal">1206</span>
<span class="normal">1207</span>
<span class="normal">1208</span>
<span class="normal">1209</span>
<span class="normal">1210</span>
<span class="normal">1211</span>
<span class="normal">1212</span>
<span class="normal">1213</span>
<span class="normal">1214</span>
<span class="normal">1215</span>
<span class="normal">1216</span>
<span class="normal">1217</span>
<span class="normal">1218</span>
<span class="normal">1219</span>
<span class="normal">1220</span>
<span class="normal">1221</span>
<span class="normal">1222</span>
<span class="normal">1223</span>
<span class="normal">1224</span>
<span class="normal">1225</span>
<span class="normal">1226</span>
<span class="normal">1227</span>
<span class="normal">1228</span>
<span class="normal">1229</span>
<span class="normal">1230</span>
<span class="normal">1231</span>
<span class="normal">1232</span>
<span class="normal">1233</span>
<span class="normal">1234</span>
<span class="normal">1235</span>
<span class="normal">1236</span>
<span class="normal">1237</span>
<span class="normal">1238</span>
<span class="normal">1239</span>
<span class="normal">1240</span>
<span class="normal">1241</span>
<span class="normal">1242</span>
<span class="normal">1243</span>
<span class="normal">1244</span>
<span class="normal">1245</span>
<span class="normal">1246</span>
<span class="normal">1247</span>
<span class="normal">1248</span>
<span class="normal">1249</span>
<span class="normal">1250</span>
<span class="normal">1251</span>
<span class="normal">1252</span>
<span class="normal">1253</span>
<span class="normal">1254</span>
<span class="normal">1255</span>
<span class="normal">1256</span>
<span class="normal">1257</span>
<span class="normal">1258</span>
<span class="normal">1259</span>
<span class="normal">1260</span>
<span class="normal">1261</span>
<span class="normal">1262</span>
<span class="normal">1263</span>
<span class="normal">1264</span>
<span class="normal">1265</span>
<span class="normal">1266</span>
<span class="normal">1267</span>
<span class="normal">1268</span>
<span class="normal">1269</span>
<span class="normal">1270</span>
<span class="normal">1271</span>
<span class="normal">1272</span>
<span class="normal">1273</span>
<span class="normal">1274</span>
<span class="normal">1275</span>
<span class="normal">1276</span>
<span class="normal">1277</span>
<span class="normal">1278</span>
<span class="normal">1279</span>
<span class="normal">1280</span>
<span class="normal">1281</span>
<span class="normal">1282</span>
<span class="normal">1283</span>
<span class="normal">1284</span>
<span class="normal">1285</span>
<span class="normal">1286</span>
<span class="normal">1287</span>
<span class="normal">1288</span>
<span class="normal">1289</span>
<span class="normal">1290</span>
<span class="normal">1291</span>
<span class="normal">1292</span>
<span class="normal">1293</span>
<span class="normal">1294</span>
<span class="normal">1295</span>
<span class="normal">1296</span>
<span class="normal">1297</span>
<span class="normal">1298</span>
<span class="normal">1299</span>
<span class="normal">1300</span>
<span class="normal">1301</span>
<span class="normal">1302</span>
<span class="normal">1303</span>
<span class="normal">1304</span>
<span class="normal">1305</span>
<span class="normal">1306</span>
<span class="normal">1307</span>
<span class="normal">1308</span>
<span class="normal">1309</span>
<span class="normal">1310</span>
<span class="normal">1311</span>
<span class="normal">1312</span>
<span class="normal">1313</span>
<span class="normal">1314</span>
<span class="normal">1315</span>
<span class="normal">1316</span>
<span class="normal">1317</span>
<span class="normal">1318</span>
<span class="normal">1319</span>
<span class="normal">1320</span>
<span class="normal">1321</span>
<span class="normal">1322</span>
<span class="normal">1323</span>
<span class="normal">1324</span>
<span class="normal">1325</span>
<span class="normal">1326</span>
<span class="normal">1327</span>
<span class="normal">1328</span>
<span class="normal">1329</span>
<span class="normal">1330</span>
<span class="normal">1331</span>
<span class="normal">1332</span>
<span class="normal">1333</span>
<span class="normal">1334</span>
<span class="normal">1335</span>
<span class="normal">1336</span>
<span class="normal">1337</span>
<span class="normal">1338</span>
<span class="normal">1339</span>
<span class="normal">1340</span>
<span class="normal">1341</span>
<span class="normal">1342</span>
<span class="normal">1343</span>
<span class="normal">1344</span>
<span class="normal">1345</span>
<span class="normal">1346</span>
<span class="normal">1347</span>
<span class="normal">1348</span>
<span class="normal">1349</span>
<span class="normal">1350</span>
<span class="normal">1351</span>
<span class="normal">1352</span>
<span class="normal">1353</span>
<span class="normal">1354</span>
<span class="normal">1355</span>
<span class="normal">1356</span>
<span class="normal">1357</span>
<span class="normal">1358</span>
<span class="normal">1359</span>
<span class="normal">1360</span>
<span class="normal">1361</span>
<span class="normal">1362</span>
<span class="normal">1363</span>
<span class="normal">1364</span>
<span class="normal">1365</span>
<span class="normal">1366</span>
<span class="normal">1367</span>
<span class="normal">1368</span>
<span class="normal">1369</span>
<span class="normal">1370</span>
<span class="normal">1371</span>
<span class="normal">1372</span>
<span class="normal">1373</span>
<span class="normal">1374</span>
<span class="normal">1375</span>
<span class="normal">1376</span>
<span class="normal">1377</span>
<span class="normal">1378</span>
<span class="normal">1379</span>
<span class="normal">1380</span>
<span class="normal">1381</span>
<span class="normal">1382</span>
<span class="normal">1383</span>
<span class="normal">1384</span>
<span class="normal">1385</span>
<span class="normal">1386</span>
<span class="normal">1387</span>
<span class="normal">1388</span>
<span class="normal">1389</span>
<span class="normal">1390</span>
<span class="normal">1391</span>
<span class="normal">1392</span>
<span class="normal">1393</span>
<span class="normal">1394</span>
<span class="normal">1395</span>
<span class="normal">1396</span>
<span class="normal">1397</span>
<span class="normal">1398</span>
<span class="normal">1399</span>
<span class="normal">1400</span>
<span class="normal">1401</span>
<span class="normal">1402</span>
<span class="normal">1403</span>
<span class="normal">1404</span>
<span class="normal">1405</span>
<span class="normal">1406</span>
<span class="normal">1407</span>
<span class="normal">1408</span>
<span class="normal">1409</span>
<span class="normal">1410</span>
<span class="normal">1411</span>
<span class="normal">1412</span>
<span class="normal">1413</span>
<span class="normal">1414</span>
<span class="normal">1415</span>
<span class="normal">1416</span>
<span class="normal">1417</span>
<span class="normal">1418</span>
<span class="normal">1419</span>
<span class="normal">1420</span>
<span class="normal">1421</span>
<span class="normal">1422</span>
<span class="normal">1423</span>
<span class="normal">1424</span>
<span class="normal">1425</span>
<span class="normal">1426</span>
<span class="normal">1427</span>
<span class="normal">1428</span>
<span class="normal">1429</span>
<span class="normal">1430</span>
<span class="normal">1431</span>
<span class="normal">1432</span>
<span class="normal">1433</span>
<span class="normal">1434</span>
<span class="normal">1435</span>
<span class="normal">1436</span>
<span class="normal">1437</span>
<span class="normal">1438</span>
<span class="normal">1439</span>
<span class="normal">1440</span>
<span class="normal">1441</span>
<span class="normal">1442</span>
<span class="normal">1443</span>
<span class="normal">1444</span>
<span class="normal">1445</span>
<span class="normal">1446</span>
<span class="normal">1447</span>
<span class="normal">1448</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@dataclass</span>
<span class="k">class</span><span class="w"> </span><span class="nc">StudyLocus</span><span class="p">(</span><span class="n">Dataset</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Study-Locus dataset.</span>

<span class="sd">    This dataset captures associations between study/traits and a genetic loci as provided by finemapping methods.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validate_study</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Flagging study loci if the corresponding study has issues.</span>

<span class="sd">        There are two different potential flags:</span>
<span class="sd">        - flagged study: flagging locus if the study has quality control flags.</span>
<span class="sd">        - study with summary statistics for top hit: flagging locus if the study has available summary statistics.</span>
<span class="sd">        - missing study: flagging locus if the study was not found in the reference study index.</span>

<span class="sd">        Args:</span>
<span class="sd">            study_index (StudyIndex): Study index to resolve study types.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Updated study locus with quality control flags.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Quality controls is not a mandatory field in the study index schema, so we have to be ready to handle it:</span>
        <span class="n">qc_select_expression</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s2">&quot;qualityControls&quot;</span> <span class="ow">in</span> <span class="n">study_index</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
            <span class="k">else</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">StringType</span><span class="p">())</span>
        <span class="p">)</span>

        <span class="c1"># The study Id of the study index needs to be kept, because we would not know which study was in the index after the left join:</span>
        <span class="n">study_flags</span> <span class="o">=</span> <span class="n">study_index</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;study_studyId&quot;</span><span class="p">),</span>
            <span class="n">qc_select_expression</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;study_qualityControls&quot;</span><span class="p">),</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">study_flags</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_studyId&quot;</span><span class="p">),</span> <span class="s2">&quot;left&quot;</span>
                <span class="p">)</span>
                <span class="c1"># Flagging loci with flagged studies - without propagating the actual flags:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                    <span class="n">StudyLocus</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_qualityControls&quot;</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span>
                        <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">FLAGGED_STUDY</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="c1"># Flagging top-hits, where the study has available summary statistics:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                    <span class="n">StudyLocus</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                        <span class="c1"># Condition is true, if the study has summary statistics available and the locus is a top hit:</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                            <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">TOP_HIT</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                        <span class="p">)</span>
                        <span class="o">&amp;</span> <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_qualityControls&quot;</span><span class="p">),</span>
                            <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">SUMSTATS_NOT_AVAILABLE</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                        <span class="p">),</span>
                        <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">TOP_HIT_AND_SUMMARY_STATS</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="c1"># Flagging loci where no studies were found:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                    <span class="n">StudyLocus</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_studyId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span>
                        <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">MISSING_STUDY</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;study_studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;study_qualityControls&quot;</span><span class="p">)</span>
            <span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">annotate_study_type</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Gets study type from study index and adds it to study locus.</span>

<span class="sd">        Args:</span>
<span class="sd">            study_index (StudyIndex): Study index to get study type.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Updated study locus with study type.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">study_index</span><span class="o">.</span><span class="n">study_type_lut</span><span class="p">(),</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validate_chromosome_label</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Flagging study loci, where chromosome is coded not as 1:22, X, Y, Xy and MT.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Updated study locus with quality control flags.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># QC column might not be present in the variant index schema, so we have to be ready to handle it:</span>
        <span class="n">qc_select_expression</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s2">&quot;qualityControls&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
            <span class="k">else</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">ArrayType</span><span class="p">(</span><span class="n">StringType</span><span class="p">()))</span>
        <span class="p">)</span>
        <span class="n">valid_chromosomes</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">,</span> <span class="s2">&quot;Y&quot;</span><span class="p">,</span> <span class="s2">&quot;XY&quot;</span><span class="p">,</span> <span class="s2">&quot;MT&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                        <span class="n">qc_select_expression</span><span class="p">,</span>
                        <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">valid_chromosomes</span><span class="p">),</span>
                        <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">INVALID_CHROMOSOME</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validate_variant_identifiers</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">variant_index</span><span class="p">:</span> <span class="n">VariantIndex</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Flagging study loci, where tagging variant identifiers are not found in variant index.</span>

<span class="sd">        Args:</span>
<span class="sd">            variant_index (VariantIndex): Variant index to resolve variant identifiers.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Updated study locus with quality control flags.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># QC column might not be present in the variant index schema, so we have to be ready to handle it:</span>
        <span class="n">qc_select_expression</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s2">&quot;qualityControls&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
            <span class="k">else</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">ArrayType</span><span class="p">(</span><span class="n">StringType</span><span class="p">()))</span>
        <span class="p">)</span>

        <span class="c1"># Find out which study loci have variants not in the variant index:</span>
        <span class="n">flag</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span>
            <span class="c1"># Exploding locus:</span>
            <span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span>
            <span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span> <span class="s2">&quot;locus.variantId&quot;</span><span class="p">)</span>
            <span class="c1"># Join with variant index variants:</span>
            <span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">variant_index</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                    <span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inVariantIndex&quot;</span><span class="p">)</span>
                <span class="p">),</span>
                <span class="n">on</span><span class="o">=</span><span class="s2">&quot;variantId&quot;</span><span class="p">,</span>
                <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="c1"># Flagging variants not in the variant index:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;inVariantIndex&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;inVariantIndex&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">())</span>
            <span class="c1"># Flagging study loci with ANY variants not in the variant index:</span>
            <span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="s2">&quot;inVariantIndex&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inVariantIndex&quot;</span><span class="p">))</span>
            <span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span><span class="s2">&quot;inVariantIndex&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;toFlag&quot;</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">flag</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                        <span class="n">qc_select_expression</span><span class="p">,</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;toFlag&quot;</span><span class="p">),</span>
                        <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">INVALID_VARIANT_IDENTIFIER</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;toFlag&quot;</span><span class="p">)</span>
            <span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validate_lead_pvalue</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">pvalue_cutoff</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Flag associations below significant threshold.</span>

<span class="sd">        Args:</span>
<span class="sd">            pvalue_cutoff (float): association p-value cut-off</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Updated study locus with quality control flags.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span>
        <span class="n">qc_colname</span> <span class="o">=</span> <span class="n">StudyLocus</span><span class="o">.</span><span class="n">get_QC_column_name</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">qc_colname</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="n">qc_colname</span><span class="p">,</span>
                <span class="n">create_empty_column_if_not_exists</span><span class="p">(</span>
                    <span class="n">qc_colname</span><span class="p">,</span>
                    <span class="n">get_struct_field_schema</span><span class="p">(</span><span class="n">StudyLocus</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span> <span class="n">qc_colname</span><span class="p">),</span>
                <span class="p">),</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
                <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="n">qc_colname</span><span class="p">,</span>
                    <span class="c1"># Because this QC might already run on the dataset, the unique set of flags is generated:</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array_distinct</span><span class="p">(</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">_qc_subsignificant_associations</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">),</span>
                            <span class="n">pvalue_cutoff</span><span class="p">,</span>
                        <span class="p">)</span>
                    <span class="p">),</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validate_unique_study_locus_id</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Validating the uniqueness of study-locus identifiers and flagging duplicated studyloci.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: with flagged duplicated studies.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">flag_duplicates</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)),</span>
                    <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">DUPLICATED_STUDYLOCUS_ID</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="n">StudyLocus</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_qc_subsignificant_associations</span><span class="p">(</span>
        <span class="n">quality_controls_column</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span>
        <span class="n">p_value_mantissa</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span>
        <span class="n">p_value_exponent</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span>
        <span class="n">pvalue_cutoff</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Flag associations below significant threshold.</span>

<span class="sd">        Args:</span>
<span class="sd">            quality_controls_column (Column): QC column</span>
<span class="sd">            p_value_mantissa (Column): P-value mantissa column</span>
<span class="sd">            p_value_exponent (Column): P-value exponent column</span>
<span class="sd">            pvalue_cutoff (float): association p-value cut-off</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Updated QC column with flag.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; import pyspark.sql.types as t</span>
<span class="sd">            &gt;&gt;&gt; d = [{&#39;qc&#39;: None, &#39;p_value_mantissa&#39;: 1, &#39;p_value_exponent&#39;: -7}, {&#39;qc&#39;: None, &#39;p_value_mantissa&#39;: 1, &#39;p_value_exponent&#39;: -8}, {&#39;qc&#39;: None, &#39;p_value_mantissa&#39;: 5, &#39;p_value_exponent&#39;: -8}, {&#39;qc&#39;: None, &#39;p_value_mantissa&#39;: 1, &#39;p_value_exponent&#39;: -9}]</span>
<span class="sd">            &gt;&gt;&gt; df = spark.createDataFrame(d, t.StructType([t.StructField(&#39;qc&#39;, t.ArrayType(t.StringType()), True), t.StructField(&#39;p_value_mantissa&#39;, t.IntegerType()), t.StructField(&#39;p_value_exponent&#39;, t.IntegerType())]))</span>
<span class="sd">            &gt;&gt;&gt; df.withColumn(&#39;qc&#39;, StudyLocus._qc_subsignificant_associations(f.col(&quot;qc&quot;), f.col(&quot;p_value_mantissa&quot;), f.col(&quot;p_value_exponent&quot;), 5e-8)).show(truncate = False)</span>
<span class="sd">            +------------------------+----------------+----------------+</span>
<span class="sd">            |qc                      |p_value_mantissa|p_value_exponent|</span>
<span class="sd">            +------------------------+----------------+----------------+</span>
<span class="sd">            |[Subsignificant p-value]|1               |-7              |</span>
<span class="sd">            |[]                      |1               |-8              |</span>
<span class="sd">            |[]                      |5               |-8              |</span>
<span class="sd">            |[]                      |1               |-9              |</span>
<span class="sd">            +------------------------+----------------+----------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">StudyLocus</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
            <span class="n">quality_controls_column</span><span class="p">,</span>
            <span class="n">calculate_neglog_pvalue</span><span class="p">(</span><span class="n">p_value_mantissa</span><span class="p">,</span> <span class="n">p_value_exponent</span><span class="p">)</span>
            <span class="o">&lt;</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">pvalue_cutoff</span><span class="p">)),</span>
            <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">SUBSIGNIFICANT_FLAG</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">qc_abnormal_pips</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
        <span class="n">sum_pips_lower_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.99</span><span class="p">,</span>
        <span class="c1"># Set slightly above 1 to account for floating point errors</span>
        <span class="n">sum_pips_upper_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0001</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Filter study-locus by sum of posterior inclusion probabilities to ensure that the sum of PIPs is within a given range.</span>

<span class="sd">        Args:</span>
<span class="sd">            sum_pips_lower_threshold (float): Lower threshold for the sum of PIPs.</span>
<span class="sd">            sum_pips_upper_threshold (float): Upper threshold for the sum of PIPs.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Filtered study-locus dataset.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># QC column might not be present so we have to be ready to handle it:</span>
        <span class="n">qc_select_expression</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s2">&quot;qualityControls&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
            <span class="k">else</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">ArrayType</span><span class="p">(</span><span class="n">StringType</span><span class="p">()))</span>
        <span class="p">)</span>

        <span class="n">flag</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;sumPosteriorProbability&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">0.0</span><span class="p">),</span>
                <span class="k">lambda</span> <span class="n">acc</span><span class="p">,</span> <span class="n">x</span><span class="p">:</span> <span class="n">acc</span> <span class="o">+</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;posteriorProbability&quot;</span><span class="p">],</span>
            <span class="p">),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;pipOutOfRange&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumPosteriorProbability&quot;</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">sum_pips_lower_threshold</span><span class="p">)</span>
                <span class="o">|</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumPosteriorProbability&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">sum_pips_upper_threshold</span><span class="p">),</span>
                <span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="kc">False</span><span class="p">),</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
                <span class="n">flag</span>
                <span class="c1"># Flagging loci with failed studies:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                        <span class="n">qc_select_expression</span><span class="p">,</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;pipOutOfRange&quot;</span><span class="p">),</span>
                        <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">ABNORMAL_PIPS</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;sumPosteriorProbability&quot;</span><span class="p">,</span> <span class="s2">&quot;pipOutOfRange&quot;</span><span class="p">)</span>
            <span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_overlapping_peaks</span><span class="p">(</span>
        <span class="n">credset_to_overlap</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">intra_study_overlap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate overlapping signals (study-locus) between GWAS-GWAS and GWAS-Molecular trait.</span>

<span class="sd">        Args:</span>
<span class="sd">            credset_to_overlap (DataFrame): DataFrame containing at least `studyLocusId`, `studyType`, `chromosome` and `tagVariantId` columns.</span>
<span class="sd">            intra_study_overlap (bool): When True, finds intra-study overlaps for credible set deduplication. Default is False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            DataFrame: containing `leftStudyLocusId`, `rightStudyLocusId` and `chromosome` columns.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Reduce columns to the minimum to reduce the size of the dataframe</span>
        <span class="n">credset_to_overlap</span> <span class="o">=</span> <span class="n">credset_to_overlap</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;studyType&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
            <span class="s2">&quot;region&quot;</span><span class="p">,</span>
            <span class="s2">&quot;tagVariantId&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># Define join condition - if intra_study_overlap is True, finds overlaps within the same study. Otherwise finds gwas vs everything overlaps for coloc.</span>
        <span class="n">join_condition</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">[</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.studyId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.studyId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.chromosome&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.tagVariantId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.tagVariantId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.studyLocusId&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.studyLocusId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.region&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.region&quot;</span><span class="p">),</span>
            <span class="p">]</span>
            <span class="k">if</span> <span class="n">intra_study_overlap</span>
            <span class="k">else</span> <span class="p">[</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.chromosome&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.tagVariantId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.tagVariantId&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.studyType&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;gwas&quot;</span><span class="p">)</span>
                <span class="o">|</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.studyLocusId&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.studyLocusId&quot;</span><span class="p">)),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.studyType&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;gwas&quot;</span><span class="p">),</span>
            <span class="p">]</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="n">credset_to_overlap</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
            <span class="c1"># Self join with complex condition.</span>
            <span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">credset_to_overlap</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;right&quot;</span><span class="p">),</span>
                <span class="n">on</span><span class="o">=</span><span class="n">join_condition</span><span class="p">,</span>
                <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.studyLocusId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;leftStudyLocusId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.studyLocusId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;rightStudyLocusId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.studyType&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;rightStudyType&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.chromosome&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">distinct</span><span class="p">()</span>
            <span class="o">.</span><span class="n">repartition</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_align_overlapping_tags</span><span class="p">(</span>
        <span class="n">loci_to_overlap</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">peak_overlaps</span><span class="p">:</span> <span class="n">DataFrame</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocusOverlap</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Align overlapping tags in pairs of overlapping study-locus, keeping all tags in both loci.</span>

<span class="sd">        Args:</span>
<span class="sd">            loci_to_overlap (DataFrame): containing `studyLocusId`, `studyType`, `chromosome`, `tagVariantId`, `logBF` and `posteriorProbability` columns.</span>
<span class="sd">            peak_overlaps (DataFrame): containing `leftStudyLocusId`, `rightStudyLocusId` and `chromosome` columns.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocusOverlap: Pairs of overlapping study-locus with aligned tags.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Complete information about all tags in the left study-locus of the overlap</span>
        <span class="n">stats_cols</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;logBF&quot;</span><span class="p">,</span>
            <span class="s2">&quot;posteriorProbability&quot;</span><span class="p">,</span>
            <span class="s2">&quot;beta&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pValueMantissa&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pValueExponent&quot;</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">overlapping_left</span> <span class="o">=</span> <span class="n">loci_to_overlap</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tagVariantId&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;leftStudyLocusId&quot;</span><span class="p">),</span>
            <span class="o">*</span><span class="p">[</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">col</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;left_</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">stats_cols</span><span class="p">],</span>
        <span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">peak_overlaps</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;leftStudyLocusId&quot;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>

        <span class="c1"># Complete information about all tags in the right study-locus of the overlap</span>
        <span class="n">overlapping_right</span> <span class="o">=</span> <span class="n">loci_to_overlap</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tagVariantId&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;rightStudyLocusId&quot;</span><span class="p">),</span>
            <span class="o">*</span><span class="p">[</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">col</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;right_</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">stats_cols</span><span class="p">],</span>
        <span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">peak_overlaps</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;rightStudyLocusId&quot;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>

        <span class="c1"># Include information about all tag variants in both study-locus aligned by tag variant id</span>
        <span class="n">overlaps</span> <span class="o">=</span> <span class="n">overlapping_left</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">overlapping_right</span><span class="p">,</span>
            <span class="n">on</span><span class="o">=</span><span class="p">[</span>
                <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
                <span class="s2">&quot;rightStudyLocusId&quot;</span><span class="p">,</span>
                <span class="s2">&quot;leftStudyLocusId&quot;</span><span class="p">,</span>
                <span class="s2">&quot;tagVariantId&quot;</span><span class="p">,</span>
                <span class="s2">&quot;rightStudyType&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;outer&quot;</span><span class="p">,</span>
        <span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="s2">&quot;leftStudyLocusId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;rightStudyLocusId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;rightStudyType&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
            <span class="s2">&quot;tagVariantId&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="o">*</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;left_</span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">stats_cols</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;right_</span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">stats_cols</span><span class="p">]</span>
            <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;statistics&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">StudyLocusOverlap</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="n">overlaps</span><span class="p">,</span>
            <span class="n">_schema</span><span class="o">=</span><span class="n">StudyLocusOverlap</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">assign_study_locus_id</span><span class="p">(</span><span class="n">uniqueness_defining_columns</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Hashes the provided columns to extract a consistent studyLocusId.</span>

<span class="sd">        Args:</span>
<span class="sd">            uniqueness_defining_columns (list[str]): list of columns defining uniqueness</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: column with a study locus ID</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; df = spark.createDataFrame([(&quot;GCST000001&quot;, &quot;1_1000_A_C&quot;, &quot;SuSiE-inf&quot;), (&quot;GCST000002&quot;, &quot;1_1000_A_C&quot;, &quot;pics&quot;)]).toDF(&quot;studyId&quot;, &quot;variantId&quot;, &quot;finemappingMethod&quot;)</span>
<span class="sd">            &gt;&gt;&gt; df.withColumn(&quot;study_locus_id&quot;, StudyLocus.assign_study_locus_id([&quot;studyId&quot;, &quot;variantId&quot;, &quot;finemappingMethod&quot;])).show(truncate=False)</span>
<span class="sd">            +----------+----------+-----------------+--------------------------------+</span>
<span class="sd">            |studyId   |variantId |finemappingMethod|study_locus_id                  |</span>
<span class="sd">            +----------+----------+-----------------+--------------------------------+</span>
<span class="sd">            |GCST000001|1_1000_A_C|SuSiE-inf        |109804fe1e20c94231a31bafd71b566e|</span>
<span class="sd">            |GCST000002|1_1000_A_C|pics             |de310be4558e0482c9cc359c97d37773|</span>
<span class="sd">            +----------+----------+-----------------+--------------------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">Dataset</span><span class="o">.</span><span class="n">generate_identifier</span><span class="p">(</span><span class="n">uniqueness_defining_columns</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
            <span class="s2">&quot;studyLocusId&quot;</span>
        <span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">calculate_credible_set_log10bf</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyLocus</span><span class="p">],</span> <span class="n">logbfs</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">num_variants_region</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">500</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate Bayes factor for the entire credible set. The Bayes factor is calculated as the logsumexp of the logBF values of the variants in the locus.</span>

<span class="sd">        Args:</span>
<span class="sd">            logbfs (Column): Array column with the logBF values of the variants in the locus.</span>
<span class="sd">            num_variants_region (int): Number of variants in the region for calculation of priors. Default: 500.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: log10 Bayes factor for the entire credible set.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; spark.createDataFrame([([1.0, 0.5, 0.25, 0.0],)]).toDF(&quot;logBF&quot;).select(f.round(StudyLocus.calculate_credible_set_log10bf(f.col(&quot;logBF&quot;), 4), 7).alias(&quot;credibleSetlog10BF&quot;)).show()</span>
<span class="sd">            +------------------+</span>
<span class="sd">            |credibleSetlog10BF|</span>
<span class="sd">            +------------------+</span>
<span class="sd">            |         0.2208288|</span>
<span class="sd">            +------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># log10=log/log(10)=log*0.43429448190325176</span>
        <span class="n">logsumexp_udf</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">udf</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span>
                <span class="n">get_logsum</span><span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="n">num_variants_region</span><span class="p">))</span> <span class="o">*</span> <span class="mf">0.43429448190325176</span>
            <span class="p">),</span>
            <span class="n">FloatType</span><span class="p">(),</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">logsumexp_udf</span><span class="p">(</span><span class="n">logbfs</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="s2">&quot;double&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;credibleSetlog10BF&quot;</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_schema</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyLocus</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">StructType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Provides the schema for the StudyLocus dataset.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StructType: schema for the StudyLocus dataset.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">parse_spark_schema</span><span class="p">(</span><span class="s2">&quot;study_locus.json&quot;</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_QC_column_name</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyLocus</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Quality control column.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: Name of the quality control column.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;qualityControls&quot;</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_QC_mappings</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyLocus</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Quality control flag to QC column category mappings.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict[str, str]: Mapping between flag name and QC column category value.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">member</span><span class="o">.</span><span class="n">name</span><span class="p">:</span> <span class="n">member</span><span class="o">.</span><span class="n">value</span> <span class="k">for</span> <span class="n">member</span> <span class="ow">in</span> <span class="n">StudyLocusQualityCheck</span><span class="p">}</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">flag_trans_qtls</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
        <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span>
        <span class="n">target_index</span><span class="p">:</span> <span class="n">TargetIndex</span><span class="p">,</span>
        <span class="n">trans_threshold</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5_000_000</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Flagging transQTL credible sets based on genomic location of the measured gene.</span>

<span class="sd">        Process:</span>
<span class="sd">        0. Make sure that the `isTransQtl` column does not exist (remove if exists)</span>
<span class="sd">        1. Enrich study-locus dataset with geneId based on study metadata. (only QTL studies are considered)</span>
<span class="sd">        2. Enrich with transcription start site and chromosome of the studied gegne.</span>
<span class="sd">        3. Flagging any tagging variant of QTL credible sets, if chromosome is different from the gene or distance is above the threshold.</span>
<span class="sd">        4. Propagate flags to credible sets where any tags are considered as trans.</span>
<span class="sd">        5. Return study locus object with annotation stored in &#39;isTransQtl` boolean column, where gwas credible sets will be `null`</span>

<span class="sd">        Args:</span>
<span class="sd">            study_index (StudyIndex): study index to extract identifier of the measured gene</span>
<span class="sd">            target_index (TargetIndex): target index bringing TSS and chromosome of the measured gene</span>
<span class="sd">            trans_threshold (int): Distance above which the QTL is considered trans. Default: 5_000_000bp</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: new column added indicating if the QTL credibles sets are trans.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># As the `geneId` column in the study index is optional, we have to test for that:</span>
        <span class="k">if</span> <span class="s2">&quot;geneId&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">study_index</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>

        <span class="c1"># We have to remove the column `isTransQtl` to ensure the column is not duplicated</span>
        <span class="c1"># The duplication can happen when one reads the StudyLocus from parquet with</span>
        <span class="c1"># predefined schema that already contains the `isTransQtl` column.</span>
        <span class="k">if</span> <span class="s2">&quot;isTransQtl&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;isTransQtl&quot;</span><span class="p">)</span>

        <span class="c1"># Process study index:</span>
        <span class="n">processed_studies</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">study_index</span><span class="o">.</span><span class="n">df</span>
            <span class="c1"># Dropping gwas studies. This ensures that only QTLs will have &quot;isTrans&quot; annotation:</span>
            <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;gwas&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;geneId&quot;</span><span class="p">,</span> <span class="s2">&quot;projectId&quot;</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># Process study locus:</span>
        <span class="n">processed_credible_set</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span>
            <span class="c1"># Exploding locus to test all tag variants:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
                <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;locus.variantId&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;locus.variantId&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">LongType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># Process target index:</span>
        <span class="n">processed_targets</span> <span class="o">=</span> <span class="n">target_index</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;geneId&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tss&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;genomicLocation.chromosome&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;geneChromosome&quot;</span><span class="p">),</span>
        <span class="p">)</span>

        <span class="c1"># Pool datasets:</span>
        <span class="n">joined_data</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">processed_credible_set</span>
            <span class="c1"># Join processed studies:</span>
            <span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">processed_studies</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>
            <span class="c1"># Join processed targets:</span>
            <span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">processed_targets</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;geneId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
            <span class="c1"># Assign True/False for QTL studies:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;isTagTrans&quot;</span><span class="p">,</span>
                <span class="c1"># The QTL signal is considered trans if the locus is on a different chromosome than the measured gene.</span>
                <span class="c1"># OR the distance from the gene&#39;s transcription start site is &gt; threshold.</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;geneChromosome&quot;</span><span class="p">))</span>
                    <span class="o">|</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tss&quot;</span><span class="p">)</span> <span class="o">-</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">))</span> <span class="o">&gt;</span> <span class="n">trans_threshold</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">),</span>
                <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">)),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
                <span class="c1"># If all tagging variants of the locus is in trans position, the QTL is considered trans:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="s2">&quot;isTagTrans&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">)),</span> <span class="kc">False</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
                <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;isTransQtl&quot;</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="c1"># Adding new column, where the value is null for gwas loci:</span>
        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">joined_data</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">))</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">filter_credible_set</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
        <span class="n">credible_interval</span><span class="p">:</span> <span class="n">CredibleInterval</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Annotate and filter study-locus tag variants based on given credible interval.</span>

<span class="sd">        Args:</span>
<span class="sd">            credible_interval (CredibleInterval): Credible interval to filter for.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Filtered study-locus dataset.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">annotate_credible_sets</span><span class="p">()</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
                    <span class="k">lambda</span> <span class="n">tag</span><span class="p">:</span> <span class="p">(</span><span class="n">tag</span><span class="p">[</span><span class="n">credible_interval</span><span class="o">.</span><span class="n">value</span><span class="p">]),</span>
                <span class="p">),</span>
            <span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_schema</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">filter_ld_set</span><span class="p">(</span><span class="n">ld_set</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">r2_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Filter the LD set by a given R2 threshold.</span>

<span class="sd">        Args:</span>
<span class="sd">            ld_set (Column): LD set</span>
<span class="sd">            r2_threshold (float): R2 threshold to filter the LD set on</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Filtered LD index</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">ld_set</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">ld_set</span><span class="p">,</span>
                <span class="k">lambda</span> <span class="n">tag</span><span class="p">:</span> <span class="n">tag</span><span class="p">[</span><span class="s2">&quot;r2Overall&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">r2_threshold</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">find_overlaps</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">intra_study_overlap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocusOverlap</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate overlapping study-locus.</span>

<span class="sd">        Find overlapping study-locus that share at least one tagging variant. All GWAS-GWAS and all GWAS-Molecular traits are computed with the Molecular traits always</span>
<span class="sd">        appearing on the right side.</span>

<span class="sd">        Args:</span>
<span class="sd">            intra_study_overlap (bool): If True, finds intra-study overlaps for credible set deduplication. Default is False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocusOverlap: Pairs of overlapping study-locus with aligned tags.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">loci_to_overlap</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">())</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span>
            <span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
                <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
                <span class="s2">&quot;studyType&quot;</span><span class="p">,</span>
                <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
                <span class="s2">&quot;region&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.variantId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;tagVariantId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.logBF&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;logBF&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.posteriorProbability&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;posteriorProbability&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.pValueMantissa&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.pValueExponent&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.beta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="c1"># overlapping study-locus</span>
        <span class="n">peak_overlaps</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_overlapping_peaks</span><span class="p">(</span><span class="n">loci_to_overlap</span><span class="p">,</span> <span class="n">intra_study_overlap</span><span class="p">)</span>

        <span class="c1"># study-locus overlap by aligning overlapping variants</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_align_overlapping_tags</span><span class="p">(</span><span class="n">loci_to_overlap</span><span class="p">,</span> <span class="n">peak_overlaps</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">unique_variants_in_locus</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;All unique variants collected in a `StudyLocus` dataframe.</span>

<span class="sd">        Returns:</span>
<span class="sd">            DataFrame: A dataframe containing `variantId` and `chromosome` columns.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;variantId&quot;</span><span class="p">,</span>
                <span class="c1"># Joint array of variants in that studylocus. Locus can be null</span>
                <span class="n">f</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array_union</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">)),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.variantId&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">()),</span>
                    <span class="p">)</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span> <span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">distinct</span><span class="p">()</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">neglog_pvalue</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the negative log p-value.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Negative log p-value</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">calculate_neglog_pvalue</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueMantissa</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueExponent</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">build_feature_matrix</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
        <span class="n">features_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">features_input_loader</span><span class="p">:</span> <span class="n">L2GFeatureInputLoader</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">L2GFeatureMatrix</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the feature matrix for a StudyLocus.</span>

<span class="sd">        Args:</span>
<span class="sd">            features_list (list[str]): List of features to include in the feature matrix.</span>
<span class="sd">            features_input_loader (L2GFeatureInputLoader): Feature input loader to use.</span>

<span class="sd">        Returns:</span>
<span class="sd">            L2GFeatureMatrix: Feature matrix for this study-locus.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span><span class="w"> </span><span class="nn">gentropy.dataset.l2g_feature_matrix</span><span class="w"> </span><span class="kn">import</span> <span class="n">L2GFeatureMatrix</span>

        <span class="k">return</span> <span class="n">L2GFeatureMatrix</span><span class="o">.</span><span class="n">from_features_list</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">features_list</span><span class="p">,</span>
            <span class="n">features_input_loader</span><span class="p">,</span>
        <span class="p">)</span><span class="o">.</span><span class="n">fill_na</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">annotate_credible_sets</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Annotate study-locus dataset with credible set flags.</span>

<span class="sd">        Sorts the array in the `locus` column elements by their `posteriorProbability` values in descending order and adds</span>
<span class="sd">        `is95CredibleSet` and `is99CredibleSet` fields to the elements, indicating which are the tagging variants whose cumulative sum</span>
<span class="sd">        of their `posteriorProbability` values is below 0.95 and 0.99, respectively.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: including annotation on `is95CredibleSet` and `is99CredibleSet`.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If `locus` column is not available.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="s2">&quot;locus&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Locus column not available.&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="c1"># Sort credible set by posterior probability in descending order</span>
            <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">),</span>
                <span class="n">order_array_of_structs_by_field</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;posteriorProbability&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="c1"># Calculate array of cumulative sums of posterior probabilities to determine which variants are in the 95% and 99% credible sets</span>
            <span class="c1"># and zip the cumulative sums array with the credible set array to add the flags</span>
            <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">zip_with</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">sequence</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))),</span>
                        <span class="k">lambda</span> <span class="n">index</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">slice</span><span class="p">(</span>
                                <span class="c1"># By using `index - 1` we introduce a value of `0.0` in the cumulative sums array. to ensure that the last variant</span>
                                <span class="c1"># that exceeds the 0.95 threshold is included in the cumulative sum, as its probability is necessary to satisfy the threshold.</span>
                                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.posteriorProbability&quot;</span><span class="p">),</span>
                                <span class="mi">1</span><span class="p">,</span>
                                <span class="n">index</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span>
                            <span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">0.0</span><span class="p">),</span>
                            <span class="k">lambda</span> <span class="n">acc</span><span class="p">,</span> <span class="n">el</span><span class="p">:</span> <span class="n">acc</span> <span class="o">+</span> <span class="n">el</span><span class="p">,</span>
                        <span class="p">),</span>
                    <span class="p">),</span>
                    <span class="k">lambda</span> <span class="n">struct_e</span><span class="p">,</span> <span class="n">acc</span><span class="p">:</span> <span class="n">struct_e</span><span class="o">.</span><span class="n">withField</span><span class="p">(</span>
                        <span class="n">CredibleInterval</span><span class="o">.</span><span class="n">IS95</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="p">(</span><span class="n">acc</span> <span class="o">&lt;</span> <span class="mf">0.95</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">acc</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">withField</span><span class="p">(</span>
                        <span class="n">CredibleInterval</span><span class="o">.</span><span class="n">IS99</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="p">(</span><span class="n">acc</span> <span class="o">&lt;</span> <span class="mf">0.99</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">acc</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
                    <span class="p">),</span>
                <span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">annotate_locus_statistics</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
        <span class="n">summary_statistics</span><span class="p">:</span> <span class="n">SummaryStatistics</span><span class="p">,</span>
        <span class="n">collect_locus_distance</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Annotates study locus with summary statistics in the specified distance around the position.</span>

<span class="sd">        Args:</span>
<span class="sd">            summary_statistics (SummaryStatistics): Summary statistics to be used for annotation.</span>
<span class="sd">            collect_locus_distance (int): distance from variant defining window for inclusion of variants in locus.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Study locus annotated with summary statistics in `locus` column. If no statistics are found, the `locus` column will be empty.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># The clumps will be used several times (persisting)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="c1"># Renaming columns:</span>
        <span class="n">sumstats_renamed</span> <span class="o">=</span> <span class="n">summary_statistics</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">selectExpr</span><span class="p">(</span>
            <span class="o">*</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2"> as tag_</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">summary_statistics</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
        <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sumstat&quot;</span><span class="p">)</span>

        <span class="n">locus_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sumstats_renamed</span>
            <span class="c1"># Joining the two datasets together:</span>
            <span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">broadcast</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;clumped&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                        <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;studyLocusId&quot;</span>
                    <span class="p">)</span>
                <span class="p">),</span>
                <span class="n">on</span><span class="o">=</span><span class="p">[</span>
                    <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_studyId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.studyId&quot;</span><span class="p">))</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.chromosome&quot;</span><span class="p">))</span>
                    <span class="o">&amp;</span> <span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_position&quot;</span><span class="p">)</span>
                        <span class="o">&gt;=</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.position&quot;</span><span class="p">)</span> <span class="o">-</span> <span class="n">collect_locus_distance</span><span class="p">)</span>
                    <span class="p">)</span>
                    <span class="o">&amp;</span> <span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_position&quot;</span><span class="p">)</span>
                        <span class="o">&lt;=</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.position&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">collect_locus_distance</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">],</span>
                <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_variantId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_beta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_pValueMantissa&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_pValueExponent&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_standardError&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">),</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">collect_list</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">locus_df</span><span class="p">,</span>
            <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">annotate_ld</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
        <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span>
        <span class="n">ld_index</span><span class="p">:</span> <span class="n">LDIndex</span><span class="p">,</span>
        <span class="n">r2_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Annotate LD information to study-locus.</span>

<span class="sd">        Args:</span>
<span class="sd">            study_index (StudyIndex): Study index to resolve ancestries.</span>
<span class="sd">            ld_index (LDIndex): LD index to resolve LD information.</span>
<span class="sd">            r2_threshold (float): R2 threshold to filter the LD index. Default is 0.0.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Study locus annotated with ld information from LD index.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span><span class="w"> </span><span class="nn">gentropy.method.ld</span><span class="w"> </span><span class="kn">import</span> <span class="n">LDAnnotator</span>

        <span class="k">return</span> <span class="n">LDAnnotator</span><span class="o">.</span><span class="n">ld_annotate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">study_index</span><span class="p">,</span> <span class="n">ld_index</span><span class="p">,</span> <span class="n">r2_threshold</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">clump</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Perform LD clumping of the studyLocus.</span>

<span class="sd">        Evaluates whether a lead variant is linked to a tag (with lowest p-value) in the same studyLocus dataset.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: with empty credible sets for linked variants and QC flag.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">clumped_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;is_lead_linked&quot;</span><span class="p">,</span>
                <span class="n">LDclumping</span><span class="o">.</span><span class="n">_is_lead_linked</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">studyId</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">chromosome</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">variantId</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueExponent</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueMantissa</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">ldSet</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;ldSet&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;is_lead_linked&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">())</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ldSet&quot;</span><span class="p">)),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">StudyLocus</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;is_lead_linked&quot;</span><span class="p">),</span>
                    <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">LD_CLUMPED</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;is_lead_linked&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="n">clumped_df</span><span class="p">,</span>
            <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">exclude_region</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">region</span><span class="p">:</span> <span class="n">GenomicRegion</span><span class="p">,</span> <span class="n">exclude_overlap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Exclude a region from the StudyLocus dataset.</span>

<span class="sd">        Args:</span>
<span class="sd">            region (GenomicRegion): genomic region object.</span>
<span class="sd">            exclude_overlap (bool): If True, excludes StudyLocus windows with any overlap with the region.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: filtered StudyLocus object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">exclude_overlap</span><span class="p">:</span>
            <span class="n">filter_condition</span> <span class="o">=</span> <span class="o">~</span><span class="p">(</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">region</span><span class="o">.</span><span class="n">chromosome</span><span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span>
                    <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locusStart&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">region</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locusEnd&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">region</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">filter_condition</span> <span class="o">=</span> <span class="o">~</span><span class="p">(</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">region</span><span class="o">.</span><span class="n">chromosome</span><span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span>
                    <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">region</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">region</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">filter_condition</span><span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="n">StudyLocus</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">qc_MHC_region</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adds qualityControl flag when lead overlaps with MHC region.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: including qualityControl flag if in MHC region.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">region</span> <span class="o">=</span> <span class="n">GenomicRegion</span><span class="o">.</span><span class="n">from_known_genomic_region</span><span class="p">(</span><span class="n">KnownGenomicRegions</span><span class="o">.</span><span class="n">MHC</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="p">(</span>
                    <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">region</span><span class="o">.</span><span class="n">chromosome</span><span class="p">)</span>
                    <span class="o">&amp;</span> <span class="p">(</span>
                        <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">region</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">region</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">),</span>
                <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">IN_MHC</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">qc_redundant_top_hits_from_PICS</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Flag associations from top hits when the study contains other PICS associations from summary statistics.</span>

<span class="sd">        This flag can be useful to identify top hits that should be explained by other associations in the study derived from the summary statistics.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Updated study locus with redundant top hits flagged.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">studies_with_pics_sumstats</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">PICS</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>
            <span class="c1"># Returns True if the study contains any PICS associations from summary statistics</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;hasPicsSumstats&quot;</span><span class="p">,</span>
                <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                    <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span> <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">TOP_HIT</span><span class="o">.</span><span class="n">value</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasPicsSumstats&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;studiesWithPicsSumstats&quot;</span><span class="p">))</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">studies_with_pics_sumstats</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                        <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span> <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">TOP_HIT</span><span class="o">.</span><span class="n">value</span>
                    <span class="p">)</span>
                    <span class="o">&amp;</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studiesWithPicsSumstats&quot;</span><span class="p">),</span>
                    <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">REDUNDANT_PICS_TOP_HIT</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;studiesWithPicsSumstats&quot;</span><span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="n">StudyLocus</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">qc_explained_by_SuSiE</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Flag associations that are explained by SuSiE associations.</span>

<span class="sd">        Credible sets overlapping in the same region as a SuSiE credible set are flagged as explained by SuSiE.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Updated study locus with SuSiE explained flags.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># unique study-regions covered by SuSie credible sets</span>
        <span class="n">susie_study_regions</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                    <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE_INF</span><span class="o">.</span><span class="n">value</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
                <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
                <span class="s2">&quot;locusStart&quot;</span><span class="p">,</span>
                <span class="s2">&quot;locusEnd&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inSuSiE&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">distinct</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="c1"># non SuSiE credible sets (studyLocusId) overlapping in any variant with SuSiE locus</span>
        <span class="n">redundant_study_locus</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                    <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE_INF</span><span class="o">.</span><span class="n">value</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;l&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span>
            <span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
                <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
                <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;l.variantId&quot;</span><span class="p">),</span> <span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;tag_position&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;study_locus&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">susie_study_regions</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;regions&quot;</span><span class="p">),</span>
                <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span>
                <span class="n">on</span><span class="o">=</span><span class="p">[</span>
                    <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_locus.chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;regions.chromosome&quot;</span><span class="p">))</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_locus.studyId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;regions.studyId&quot;</span><span class="p">))</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_locus.tag_position&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;regions.locusStart&quot;</span><span class="p">))</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_locus.tag_position&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;regions.locusEnd&quot;</span><span class="p">))</span>
                <span class="p">],</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span> <span class="s2">&quot;inSuSiE&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">distinct</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">redundant_study_locus</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                        <span class="c1"># credible set in SuSiE overlapping region</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;inSuSiE&quot;</span><span class="p">)</span>
                        <span class="c1"># credible set not based on SuSiE</span>
                        <span class="o">&amp;</span> <span class="p">(</span>
                            <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                                <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                                <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE_INF</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                            <span class="p">)</span>
                        <span class="p">),</span>
                        <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">EXPLAINED_BY_SUSIE</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;inSuSiE&quot;</span><span class="p">)</span>
            <span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="n">StudyLocus</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_qc_no_population</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Flag associations where the study doesn&#39;t have population information to resolve LD.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Updated study locus.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># If the tested column is not present, return self unchanged:</span>
        <span class="k">if</span> <span class="s2">&quot;ldPopulationStructure&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ldPopulationStructure&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span>
                <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">NO_POPULATION</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">annotate_locus_statistics_boundaries</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
        <span class="n">summary_statistics</span><span class="p">:</span> <span class="n">SummaryStatistics</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Annotates study locus with summary statistics in the specified boundaries - locusStart and locusEnd.</span>

<span class="sd">        Args:</span>
<span class="sd">            summary_statistics (SummaryStatistics): Summary statistics to be used for annotation.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Study locus annotated with summary statistics in `locus` column. If no statistics are found, the `locus` column will be empty.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># The clumps will be used several times (persisting)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="c1"># Renaming columns:</span>
        <span class="n">sumstats_renamed</span> <span class="o">=</span> <span class="n">summary_statistics</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">selectExpr</span><span class="p">(</span>
            <span class="o">*</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2"> as tag_</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">summary_statistics</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
        <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sumstat&quot;</span><span class="p">)</span>

        <span class="n">locus_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sumstats_renamed</span>
            <span class="c1"># Joining the two datasets together:</span>
            <span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">broadcast</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;clumped&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                        <span class="s2">&quot;position&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;locusStart&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;locusEnd&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">),</span>
                <span class="n">on</span><span class="o">=</span><span class="p">[</span>
                    <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_studyId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.studyId&quot;</span><span class="p">))</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.chromosome&quot;</span><span class="p">))</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_position&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.locusStart&quot;</span><span class="p">)))</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_position&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.locusEnd&quot;</span><span class="p">)))</span>
                <span class="p">],</span>
                <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_variantId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_beta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_pValueMantissa&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_pValueExponent&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_standardError&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">),</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">collect_list</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">locus_df</span><span class="p">,</span>
            <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">window_based_clumping</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
        <span class="n">window_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">WindowBasedClumpingStepConfig</span><span class="p">()</span><span class="o">.</span><span class="n">distance</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Clump study locus by window size.</span>

<span class="sd">        Args:</span>
<span class="sd">            window_size (int): Window size for clumping.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Clumped study locus, where clumped associations are flagged.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span><span class="w"> </span><span class="nn">gentropy.method.window_based_clumping</span><span class="w"> </span><span class="kn">import</span> <span class="n">WindowBasedClumping</span>

        <span class="k">return</span> <span class="n">WindowBasedClumping</span><span class="o">.</span><span class="n">clump</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">window_size</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">assign_confidence</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Assign confidence to study locus.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Study locus with confidence assigned.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Return self if the required columns are not in the dataframe:</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
            <span class="ow">or</span> <span class="s2">&quot;finemappingMethod&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
        <span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span>

        <span class="c1"># Assign confidence based on the presence of quality controls</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;confidence&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                        <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                        <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE_INF</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span>
                    <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                        <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">OUT_OF_SAMPLE_LD</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">),</span>
                <span class="n">CredibleSetConfidenceClasses</span><span class="o">.</span><span class="n">FINEMAPPED_IN_SAMPLE_LD</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                        <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                        <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE_INF</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                        <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">OUT_OF_SAMPLE_LD</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">),</span>
                <span class="n">CredibleSetConfidenceClasses</span><span class="o">.</span><span class="n">FINEMAPPED_OUT_OF_SAMPLE_LD</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">PICS</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span>
                    <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span> <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">TOP_HIT</span><span class="o">.</span><span class="n">value</span>
                    <span class="p">)</span>
                <span class="p">),</span>
                <span class="n">CredibleSetConfidenceClasses</span><span class="o">.</span><span class="n">PICSED_SUMMARY_STATS</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">PICS</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span> <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">TOP_HIT</span><span class="o">.</span><span class="n">value</span>
                    <span class="p">)</span>
                <span class="p">),</span>
                <span class="n">CredibleSetConfidenceClasses</span><span class="o">.</span><span class="n">PICSED_TOP_HIT</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">CredibleSetConfidenceClasses</span><span class="o">.</span><span class="n">UNKNOWN</span><span class="o">.</span><span class="n">value</span><span class="p">),</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="n">df</span><span class="p">,</span>
            <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.annotate_credible_sets" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">annotate_credible_sets</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.annotate_credible_sets" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Annotate study-locus dataset with credible set flags.</p>
<p>Sorts the array in the <code>locus</code> column elements by their <code>posteriorProbability</code> values in descending order and adds
<code>is95CredibleSet</code> and <code>is99CredibleSet</code> fields to the elements, indicating which are the tagging variants whose cumulative sum
of their <code>posteriorProbability</code> values is below 0.95 and 0.99, respectively.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>including annotation on <code>is95CredibleSet</code> and <code>is99CredibleSet</code>.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If <code>locus</code> column is not available.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">922</span>
<span class="normal">923</span>
<span class="normal">924</span>
<span class="normal">925</span>
<span class="normal">926</span>
<span class="normal">927</span>
<span class="normal">928</span>
<span class="normal">929</span>
<span class="normal">930</span>
<span class="normal">931</span>
<span class="normal">932</span>
<span class="normal">933</span>
<span class="normal">934</span>
<span class="normal">935</span>
<span class="normal">936</span>
<span class="normal">937</span>
<span class="normal">938</span>
<span class="normal">939</span>
<span class="normal">940</span>
<span class="normal">941</span>
<span class="normal">942</span>
<span class="normal">943</span>
<span class="normal">944</span>
<span class="normal">945</span>
<span class="normal">946</span>
<span class="normal">947</span>
<span class="normal">948</span>
<span class="normal">949</span>
<span class="normal">950</span>
<span class="normal">951</span>
<span class="normal">952</span>
<span class="normal">953</span>
<span class="normal">954</span>
<span class="normal">955</span>
<span class="normal">956</span>
<span class="normal">957</span>
<span class="normal">958</span>
<span class="normal">959</span>
<span class="normal">960</span>
<span class="normal">961</span>
<span class="normal">962</span>
<span class="normal">963</span>
<span class="normal">964</span>
<span class="normal">965</span>
<span class="normal">966</span>
<span class="normal">967</span>
<span class="normal">968</span>
<span class="normal">969</span>
<span class="normal">970</span>
<span class="normal">971</span>
<span class="normal">972</span>
<span class="normal">973</span>
<span class="normal">974</span>
<span class="normal">975</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">annotate_credible_sets</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Annotate study-locus dataset with credible set flags.</span>

<span class="sd">    Sorts the array in the `locus` column elements by their `posteriorProbability` values in descending order and adds</span>
<span class="sd">    `is95CredibleSet` and `is99CredibleSet` fields to the elements, indicating which are the tagging variants whose cumulative sum</span>
<span class="sd">    of their `posteriorProbability` values is below 0.95 and 0.99, respectively.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: including annotation on `is95CredibleSet` and `is99CredibleSet`.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If `locus` column is not available.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s2">&quot;locus&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Locus column not available.&quot;</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
        <span class="c1"># Sort credible set by posterior probability in descending order</span>
        <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">),</span>
            <span class="n">order_array_of_structs_by_field</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;posteriorProbability&quot;</span><span class="p">),</span>
        <span class="p">),</span>
    <span class="p">)</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
        <span class="c1"># Calculate array of cumulative sums of posterior probabilities to determine which variants are in the 95% and 99% credible sets</span>
        <span class="c1"># and zip the cumulative sums array with the credible set array to add the flags</span>
        <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">zip_with</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">sequence</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))),</span>
                    <span class="k">lambda</span> <span class="n">index</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">slice</span><span class="p">(</span>
                            <span class="c1"># By using `index - 1` we introduce a value of `0.0` in the cumulative sums array. to ensure that the last variant</span>
                            <span class="c1"># that exceeds the 0.95 threshold is included in the cumulative sum, as its probability is necessary to satisfy the threshold.</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.posteriorProbability&quot;</span><span class="p">),</span>
                            <span class="mi">1</span><span class="p">,</span>
                            <span class="n">index</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span>
                        <span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">0.0</span><span class="p">),</span>
                        <span class="k">lambda</span> <span class="n">acc</span><span class="p">,</span> <span class="n">el</span><span class="p">:</span> <span class="n">acc</span> <span class="o">+</span> <span class="n">el</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">),</span>
                <span class="k">lambda</span> <span class="n">struct_e</span><span class="p">,</span> <span class="n">acc</span><span class="p">:</span> <span class="n">struct_e</span><span class="o">.</span><span class="n">withField</span><span class="p">(</span>
                    <span class="n">CredibleInterval</span><span class="o">.</span><span class="n">IS95</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="p">(</span><span class="n">acc</span> <span class="o">&lt;</span> <span class="mf">0.95</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">acc</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
                <span class="p">)</span><span class="o">.</span><span class="n">withField</span><span class="p">(</span>
                    <span class="n">CredibleInterval</span><span class="o">.</span><span class="n">IS99</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="p">(</span><span class="n">acc</span> <span class="o">&lt;</span> <span class="mf">0.99</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">acc</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
                <span class="p">),</span>
            <span class="p">),</span>
        <span class="p">),</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.annotate_ld" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">annotate_ld</span><span class="p">(</span><span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span> <span class="n">ld_index</span><span class="p">:</span> <span class="n">LDIndex</span><span class="p">,</span> <span class="n">r2_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.annotate_ld" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Annotate LD information to study-locus.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>study_index</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="../study_index/#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study index to resolve ancestries.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>ld_index</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.ld_index.LDIndex" href="../ld_index/#gentropy.dataset.ld_index.LDIndex">LDIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>LD index to resolve LD information.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>r2_threshold</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>R2 threshold to filter the LD index. Default is 0.0.</p>
              </div>
            </td>
            <td>
                  <code>0.0</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study locus annotated with ld information from LD index.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">annotate_ld</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
    <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span>
    <span class="n">ld_index</span><span class="p">:</span> <span class="n">LDIndex</span><span class="p">,</span>
    <span class="n">r2_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Annotate LD information to study-locus.</span>

<span class="sd">    Args:</span>
<span class="sd">        study_index (StudyIndex): Study index to resolve ancestries.</span>
<span class="sd">        ld_index (LDIndex): LD index to resolve LD information.</span>
<span class="sd">        r2_threshold (float): R2 threshold to filter the LD index. Default is 0.0.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Study locus annotated with ld information from LD index.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">gentropy.method.ld</span><span class="w"> </span><span class="kn">import</span> <span class="n">LDAnnotator</span>

    <span class="k">return</span> <span class="n">LDAnnotator</span><span class="o">.</span><span class="n">ld_annotate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">study_index</span><span class="p">,</span> <span class="n">ld_index</span><span class="p">,</span> <span class="n">r2_threshold</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.annotate_locus_statistics" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">annotate_locus_statistics</span><span class="p">(</span><span class="n">summary_statistics</span><span class="p">:</span> <span class="n">SummaryStatistics</span><span class="p">,</span> <span class="n">collect_locus_distance</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.annotate_locus_statistics" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Annotates study locus with summary statistics in the specified distance around the position.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>summary_statistics</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.summary_statistics.SummaryStatistics" href="../summary_statistics/#gentropy.dataset.summary_statistics.SummaryStatistics">SummaryStatistics</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Summary statistics to be used for annotation.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>collect_locus_distance</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>distance from variant defining window for inclusion of variants in locus.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study locus annotated with summary statistics in <code>locus</code> column. If no statistics are found, the <code>locus</code> column will be empty.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">annotate_locus_statistics</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
    <span class="n">summary_statistics</span><span class="p">:</span> <span class="n">SummaryStatistics</span><span class="p">,</span>
    <span class="n">collect_locus_distance</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Annotates study locus with summary statistics in the specified distance around the position.</span>

<span class="sd">    Args:</span>
<span class="sd">        summary_statistics (SummaryStatistics): Summary statistics to be used for annotation.</span>
<span class="sd">        collect_locus_distance (int): distance from variant defining window for inclusion of variants in locus.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Study locus annotated with summary statistics in `locus` column. If no statistics are found, the `locus` column will be empty.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># The clumps will be used several times (persisting)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>
    <span class="c1"># Renaming columns:</span>
    <span class="n">sumstats_renamed</span> <span class="o">=</span> <span class="n">summary_statistics</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">selectExpr</span><span class="p">(</span>
        <span class="o">*</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2"> as tag_</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">summary_statistics</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sumstat&quot;</span><span class="p">)</span>

    <span class="n">locus_df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">sumstats_renamed</span>
        <span class="c1"># Joining the two datasets together:</span>
        <span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">broadcast</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;clumped&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                    <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;studyLocusId&quot;</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="n">on</span><span class="o">=</span><span class="p">[</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_studyId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.studyId&quot;</span><span class="p">))</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.chromosome&quot;</span><span class="p">))</span>
                <span class="o">&amp;</span> <span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_position&quot;</span><span class="p">)</span>
                    <span class="o">&gt;=</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.position&quot;</span><span class="p">)</span> <span class="o">-</span> <span class="n">collect_locus_distance</span><span class="p">)</span>
                <span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_position&quot;</span><span class="p">)</span>
                    <span class="o">&lt;=</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.position&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">collect_locus_distance</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">],</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_variantId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_beta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_pValueMantissa&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_pValueExponent&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_standardError&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">collect_list</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">locus_df</span><span class="p">,</span>
        <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
        <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.annotate_locus_statistics_boundaries" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">annotate_locus_statistics_boundaries</span><span class="p">(</span><span class="n">summary_statistics</span><span class="p">:</span> <span class="n">SummaryStatistics</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.annotate_locus_statistics_boundaries" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Annotates study locus with summary statistics in the specified boundaries - locusStart and locusEnd.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>summary_statistics</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.summary_statistics.SummaryStatistics" href="../summary_statistics/#gentropy.dataset.summary_statistics.SummaryStatistics">SummaryStatistics</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Summary statistics to be used for annotation.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study locus annotated with summary statistics in <code>locus</code> column. If no statistics are found, the <code>locus</code> column will be empty.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1297</span>
<span class="normal">1298</span>
<span class="normal">1299</span>
<span class="normal">1300</span>
<span class="normal">1301</span>
<span class="normal">1302</span>
<span class="normal">1303</span>
<span class="normal">1304</span>
<span class="normal">1305</span>
<span class="normal">1306</span>
<span class="normal">1307</span>
<span class="normal">1308</span>
<span class="normal">1309</span>
<span class="normal">1310</span>
<span class="normal">1311</span>
<span class="normal">1312</span>
<span class="normal">1313</span>
<span class="normal">1314</span>
<span class="normal">1315</span>
<span class="normal">1316</span>
<span class="normal">1317</span>
<span class="normal">1318</span>
<span class="normal">1319</span>
<span class="normal">1320</span>
<span class="normal">1321</span>
<span class="normal">1322</span>
<span class="normal">1323</span>
<span class="normal">1324</span>
<span class="normal">1325</span>
<span class="normal">1326</span>
<span class="normal">1327</span>
<span class="normal">1328</span>
<span class="normal">1329</span>
<span class="normal">1330</span>
<span class="normal">1331</span>
<span class="normal">1332</span>
<span class="normal">1333</span>
<span class="normal">1334</span>
<span class="normal">1335</span>
<span class="normal">1336</span>
<span class="normal">1337</span>
<span class="normal">1338</span>
<span class="normal">1339</span>
<span class="normal">1340</span>
<span class="normal">1341</span>
<span class="normal">1342</span>
<span class="normal">1343</span>
<span class="normal">1344</span>
<span class="normal">1345</span>
<span class="normal">1346</span>
<span class="normal">1347</span>
<span class="normal">1348</span>
<span class="normal">1349</span>
<span class="normal">1350</span>
<span class="normal">1351</span>
<span class="normal">1352</span>
<span class="normal">1353</span>
<span class="normal">1354</span>
<span class="normal">1355</span>
<span class="normal">1356</span>
<span class="normal">1357</span>
<span class="normal">1358</span>
<span class="normal">1359</span>
<span class="normal">1360</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">annotate_locus_statistics_boundaries</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
    <span class="n">summary_statistics</span><span class="p">:</span> <span class="n">SummaryStatistics</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Annotates study locus with summary statistics in the specified boundaries - locusStart and locusEnd.</span>

<span class="sd">    Args:</span>
<span class="sd">        summary_statistics (SummaryStatistics): Summary statistics to be used for annotation.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Study locus annotated with summary statistics in `locus` column. If no statistics are found, the `locus` column will be empty.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># The clumps will be used several times (persisting)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>
    <span class="c1"># Renaming columns:</span>
    <span class="n">sumstats_renamed</span> <span class="o">=</span> <span class="n">summary_statistics</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">selectExpr</span><span class="p">(</span>
        <span class="o">*</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2"> as tag_</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">summary_statistics</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sumstat&quot;</span><span class="p">)</span>

    <span class="n">locus_df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">sumstats_renamed</span>
        <span class="c1"># Joining the two datasets together:</span>
        <span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">broadcast</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;clumped&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                    <span class="s2">&quot;position&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;locusStart&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;locusEnd&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="n">on</span><span class="o">=</span><span class="p">[</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_studyId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.studyId&quot;</span><span class="p">))</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.chromosome&quot;</span><span class="p">))</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_position&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.locusStart&quot;</span><span class="p">)))</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_position&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.locusEnd&quot;</span><span class="p">)))</span>
            <span class="p">],</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_variantId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_beta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_pValueMantissa&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_pValueExponent&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_standardError&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">collect_list</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">locus_df</span><span class="p">,</span>
        <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
        <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.annotate_study_type" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">annotate_study_type</span><span class="p">(</span><span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.annotate_study_type" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Gets study type from study index and adds it to study locus.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>study_index</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="../study_index/#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study index to get study type.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Updated study locus with study type.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">annotate_study_type</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Gets study type from study index and adds it to study locus.</span>

<span class="sd">    Args:</span>
<span class="sd">        study_index (StudyIndex): Study index to get study type.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Updated study locus with study type.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">study_index</span><span class="o">.</span><span class="n">study_type_lut</span><span class="p">(),</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span>
            <span class="p">)</span>
        <span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.assign_confidence" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">assign_confidence</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.assign_confidence" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Assign confidence to study locus.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study locus with confidence assigned.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1378</span>
<span class="normal">1379</span>
<span class="normal">1380</span>
<span class="normal">1381</span>
<span class="normal">1382</span>
<span class="normal">1383</span>
<span class="normal">1384</span>
<span class="normal">1385</span>
<span class="normal">1386</span>
<span class="normal">1387</span>
<span class="normal">1388</span>
<span class="normal">1389</span>
<span class="normal">1390</span>
<span class="normal">1391</span>
<span class="normal">1392</span>
<span class="normal">1393</span>
<span class="normal">1394</span>
<span class="normal">1395</span>
<span class="normal">1396</span>
<span class="normal">1397</span>
<span class="normal">1398</span>
<span class="normal">1399</span>
<span class="normal">1400</span>
<span class="normal">1401</span>
<span class="normal">1402</span>
<span class="normal">1403</span>
<span class="normal">1404</span>
<span class="normal">1405</span>
<span class="normal">1406</span>
<span class="normal">1407</span>
<span class="normal">1408</span>
<span class="normal">1409</span>
<span class="normal">1410</span>
<span class="normal">1411</span>
<span class="normal">1412</span>
<span class="normal">1413</span>
<span class="normal">1414</span>
<span class="normal">1415</span>
<span class="normal">1416</span>
<span class="normal">1417</span>
<span class="normal">1418</span>
<span class="normal">1419</span>
<span class="normal">1420</span>
<span class="normal">1421</span>
<span class="normal">1422</span>
<span class="normal">1423</span>
<span class="normal">1424</span>
<span class="normal">1425</span>
<span class="normal">1426</span>
<span class="normal">1427</span>
<span class="normal">1428</span>
<span class="normal">1429</span>
<span class="normal">1430</span>
<span class="normal">1431</span>
<span class="normal">1432</span>
<span class="normal">1433</span>
<span class="normal">1434</span>
<span class="normal">1435</span>
<span class="normal">1436</span>
<span class="normal">1437</span>
<span class="normal">1438</span>
<span class="normal">1439</span>
<span class="normal">1440</span>
<span class="normal">1441</span>
<span class="normal">1442</span>
<span class="normal">1443</span>
<span class="normal">1444</span>
<span class="normal">1445</span>
<span class="normal">1446</span>
<span class="normal">1447</span>
<span class="normal">1448</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">assign_confidence</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Assign confidence to study locus.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Study locus with confidence assigned.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Return self if the required columns are not in the dataframe:</span>
    <span class="k">if</span> <span class="p">(</span>
        <span class="s2">&quot;qualityControls&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
        <span class="ow">or</span> <span class="s2">&quot;finemappingMethod&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
    <span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="c1"># Assign confidence based on the presence of quality controls</span>
    <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
        <span class="s2">&quot;confidence&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                    <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                    <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE_INF</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="o">&amp;</span> <span class="p">(</span>
                <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">OUT_OF_SAMPLE_LD</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="n">CredibleSetConfidenceClasses</span><span class="o">.</span><span class="n">FINEMAPPED_IN_SAMPLE_LD</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                    <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                    <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE_INF</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="o">&amp;</span> <span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">OUT_OF_SAMPLE_LD</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="n">CredibleSetConfidenceClasses</span><span class="o">.</span><span class="n">FINEMAPPED_OUT_OF_SAMPLE_LD</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">PICS</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>
            <span class="o">&amp;</span> <span class="p">(</span>
                <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span> <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">TOP_HIT</span><span class="o">.</span><span class="n">value</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="n">CredibleSetConfidenceClasses</span><span class="o">.</span><span class="n">PICSED_SUMMARY_STATS</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">PICS</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>
            <span class="o">&amp;</span> <span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span> <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">TOP_HIT</span><span class="o">.</span><span class="n">value</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="n">CredibleSetConfidenceClasses</span><span class="o">.</span><span class="n">PICSED_TOP_HIT</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">CredibleSetConfidenceClasses</span><span class="o">.</span><span class="n">UNKNOWN</span><span class="o">.</span><span class="n">value</span><span class="p">),</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="n">df</span><span class="p">,</span>
        <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.assign_study_locus_id" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">assign_study_locus_id</span><span class="p">(</span><span class="n">uniqueness_defining_columns</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

<a href="#gentropy.dataset.study_locus.StudyLocus.assign_study_locus_id" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Hashes the provided columns to extract a consistent studyLocusId.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>uniqueness_defining_columns</code>
            </td>
            <td>
                  <code>list[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of columns defining uniqueness</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>column with a study locus ID</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">([(</span><span class="s2">&quot;GCST000001&quot;</span><span class="p">,</span> <span class="s2">&quot;1_1000_A_C&quot;</span><span class="p">,</span> <span class="s2">&quot;SuSiE-inf&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;GCST000002&quot;</span><span class="p">,</span> <span class="s2">&quot;1_1000_A_C&quot;</span><span class="p">,</span> <span class="s2">&quot;pics&quot;</span><span class="p">)])</span><span class="o">.</span><span class="n">toDF</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;study_locus_id&quot;</span><span class="p">,</span> <span class="n">StudyLocus</span><span class="o">.</span><span class="n">assign_study_locus_id</span><span class="p">([</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="s2">&quot;finemappingMethod&quot;</span><span class="p">]))</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+----------+----------+-----------------+--------------------------------+</span>
<span class="go">|studyId   |variantId |finemappingMethod|study_locus_id                  |</span>
<span class="go">+----------+----------+-----------------+--------------------------------+</span>
<span class="go">|GCST000001|1_1000_A_C|SuSiE-inf        |109804fe1e20c94231a31bafd71b566e|</span>
<span class="go">|GCST000002|1_1000_A_C|pics             |de310be4558e0482c9cc359c97d37773|</span>
<span class="go">+----------+----------+-----------------+--------------------------------+</span>
</code></pre></div>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">assign_study_locus_id</span><span class="p">(</span><span class="n">uniqueness_defining_columns</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Hashes the provided columns to extract a consistent studyLocusId.</span>

<span class="sd">    Args:</span>
<span class="sd">        uniqueness_defining_columns (list[str]): list of columns defining uniqueness</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: column with a study locus ID</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame([(&quot;GCST000001&quot;, &quot;1_1000_A_C&quot;, &quot;SuSiE-inf&quot;), (&quot;GCST000002&quot;, &quot;1_1000_A_C&quot;, &quot;pics&quot;)]).toDF(&quot;studyId&quot;, &quot;variantId&quot;, &quot;finemappingMethod&quot;)</span>
<span class="sd">        &gt;&gt;&gt; df.withColumn(&quot;study_locus_id&quot;, StudyLocus.assign_study_locus_id([&quot;studyId&quot;, &quot;variantId&quot;, &quot;finemappingMethod&quot;])).show(truncate=False)</span>
<span class="sd">        +----------+----------+-----------------+--------------------------------+</span>
<span class="sd">        |studyId   |variantId |finemappingMethod|study_locus_id                  |</span>
<span class="sd">        +----------+----------+-----------------+--------------------------------+</span>
<span class="sd">        |GCST000001|1_1000_A_C|SuSiE-inf        |109804fe1e20c94231a31bafd71b566e|</span>
<span class="sd">        |GCST000002|1_1000_A_C|pics             |de310be4558e0482c9cc359c97d37773|</span>
<span class="sd">        +----------+----------+-----------------+--------------------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">Dataset</span><span class="o">.</span><span class="n">generate_identifier</span><span class="p">(</span><span class="n">uniqueness_defining_columns</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
        <span class="s2">&quot;studyLocusId&quot;</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.build_feature_matrix" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">build_feature_matrix</span><span class="p">(</span><span class="n">features_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">features_input_loader</span><span class="p">:</span> <span class="n">L2GFeatureInputLoader</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">L2GFeatureMatrix</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.build_feature_matrix" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Returns the feature matrix for a StudyLocus.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>features_list</code>
            </td>
            <td>
                  <code>list[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of features to include in the feature matrix.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>features_input_loader</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.method.l2g.feature_factory.L2GFeatureInputLoader" href="../../methods/l2g/feature_factory/#gentropy.method.l2g.feature_factory.L2GFeatureInputLoader">L2GFeatureInputLoader</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Feature input loader to use.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>L2GFeatureMatrix</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.l2g_feature_matrix.L2GFeatureMatrix" href="../l2g_feature_matrix/#gentropy.dataset.l2g_feature_matrix.L2GFeatureMatrix">L2GFeatureMatrix</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Feature matrix for this study-locus.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span>
<span class="normal">917</span>
<span class="normal">918</span>
<span class="normal">919</span>
<span class="normal">920</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">build_feature_matrix</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
    <span class="n">features_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">features_input_loader</span><span class="p">:</span> <span class="n">L2GFeatureInputLoader</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">L2GFeatureMatrix</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the feature matrix for a StudyLocus.</span>

<span class="sd">    Args:</span>
<span class="sd">        features_list (list[str]): List of features to include in the feature matrix.</span>
<span class="sd">        features_input_loader (L2GFeatureInputLoader): Feature input loader to use.</span>

<span class="sd">    Returns:</span>
<span class="sd">        L2GFeatureMatrix: Feature matrix for this study-locus.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">gentropy.dataset.l2g_feature_matrix</span><span class="w"> </span><span class="kn">import</span> <span class="n">L2GFeatureMatrix</span>

    <span class="k">return</span> <span class="n">L2GFeatureMatrix</span><span class="o">.</span><span class="n">from_features_list</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">features_list</span><span class="p">,</span>
        <span class="n">features_input_loader</span><span class="p">,</span>
    <span class="p">)</span><span class="o">.</span><span class="n">fill_na</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.calculate_credible_set_log10bf" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">calculate_credible_set_log10bf</span><span class="p">(</span><span class="n">logbfs</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">num_variants_region</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">500</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.dataset.study_locus.StudyLocus.calculate_credible_set_log10bf" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Calculate Bayes factor for the entire credible set. The Bayes factor is calculated as the logsumexp of the logBF values of the variants in the locus.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>logbfs</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array column with the logBF values of the variants in the locus.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>num_variants_region</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of variants in the region for calculation of priors. Default: 500.</p>
              </div>
            </td>
            <td>
                  <code>500</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>log10 Bayes factor for the entire credible set.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">([([</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">],)])</span><span class="o">.</span><span class="n">toDF</span><span class="p">(</span><span class="s2">&quot;logBF&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">StudyLocus</span><span class="o">.</span><span class="n">calculate_credible_set_log10bf</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;logBF&quot;</span><span class="p">),</span> <span class="mi">4</span><span class="p">),</span> <span class="mi">7</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;credibleSetlog10BF&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+------------------+</span>
<span class="go">|credibleSetlog10BF|</span>
<span class="go">+------------------+</span>
<span class="go">|         0.2208288|</span>
<span class="go">+------------------+</span>
</code></pre></div>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">calculate_credible_set_log10bf</span><span class="p">(</span>
    <span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyLocus</span><span class="p">],</span> <span class="n">logbfs</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">num_variants_region</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">500</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate Bayes factor for the entire credible set. The Bayes factor is calculated as the logsumexp of the logBF values of the variants in the locus.</span>

<span class="sd">    Args:</span>
<span class="sd">        logbfs (Column): Array column with the logBF values of the variants in the locus.</span>
<span class="sd">        num_variants_region (int): Number of variants in the region for calculation of priors. Default: 500.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: log10 Bayes factor for the entire credible set.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; spark.createDataFrame([([1.0, 0.5, 0.25, 0.0],)]).toDF(&quot;logBF&quot;).select(f.round(StudyLocus.calculate_credible_set_log10bf(f.col(&quot;logBF&quot;), 4), 7).alias(&quot;credibleSetlog10BF&quot;)).show()</span>
<span class="sd">        +------------------+</span>
<span class="sd">        |credibleSetlog10BF|</span>
<span class="sd">        +------------------+</span>
<span class="sd">        |         0.2208288|</span>
<span class="sd">        +------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># log10=log/log(10)=log*0.43429448190325176</span>
    <span class="n">logsumexp_udf</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">udf</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span>
            <span class="n">get_logsum</span><span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="mi">1</span> <span class="o">/</span> <span class="n">num_variants_region</span><span class="p">))</span> <span class="o">*</span> <span class="mf">0.43429448190325176</span>
        <span class="p">),</span>
        <span class="n">FloatType</span><span class="p">(),</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">logsumexp_udf</span><span class="p">(</span><span class="n">logbfs</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="s2">&quot;double&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;credibleSetlog10BF&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.clump" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">clump</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.clump" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Perform LD clumping of the studyLocus.</p>
<p>Evaluates whether a lead variant is linked to a tag (with lowest p-value) in the same studyLocus dataset.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>with empty credible sets for linked variants and QC flag.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">clump</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Perform LD clumping of the studyLocus.</span>

<span class="sd">    Evaluates whether a lead variant is linked to a tag (with lowest p-value) in the same studyLocus dataset.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: with empty credible sets for linked variants and QC flag.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">clumped_df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;is_lead_linked&quot;</span><span class="p">,</span>
            <span class="n">LDclumping</span><span class="o">.</span><span class="n">_is_lead_linked</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">studyId</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">chromosome</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">variantId</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueExponent</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueMantissa</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">ldSet</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;ldSet&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;is_lead_linked&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">())</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ldSet&quot;</span><span class="p">)),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="n">StudyLocus</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;is_lead_linked&quot;</span><span class="p">),</span>
                <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">LD_CLUMPED</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;is_lead_linked&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="n">clumped_df</span><span class="p">,</span>
        <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.exclude_region" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">exclude_region</span><span class="p">(</span><span class="n">region</span><span class="p">:</span> <span class="n">GenomicRegion</span><span class="p">,</span> <span class="n">exclude_overlap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.exclude_region" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Exclude a region from the StudyLocus dataset.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>region</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.common.genomic_region.GenomicRegion" href="../../common/genomic_region/#gentropy.common.genomic_region.GenomicRegion">GenomicRegion</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>genomic region object.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>exclude_overlap</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, excludes StudyLocus windows with any overlap with the region.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>filtered StudyLocus object.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span>
<span class="normal">1124</span>
<span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">exclude_region</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">region</span><span class="p">:</span> <span class="n">GenomicRegion</span><span class="p">,</span> <span class="n">exclude_overlap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Exclude a region from the StudyLocus dataset.</span>

<span class="sd">    Args:</span>
<span class="sd">        region (GenomicRegion): genomic region object.</span>
<span class="sd">        exclude_overlap (bool): If True, excludes StudyLocus windows with any overlap with the region.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: filtered StudyLocus object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">exclude_overlap</span><span class="p">:</span>
        <span class="n">filter_condition</span> <span class="o">=</span> <span class="o">~</span><span class="p">(</span>
            <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">region</span><span class="o">.</span><span class="n">chromosome</span><span class="p">)</span>
            <span class="o">&amp;</span> <span class="p">(</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locusStart&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">region</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locusEnd&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">region</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">filter_condition</span> <span class="o">=</span> <span class="o">~</span><span class="p">(</span>
            <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">region</span><span class="o">.</span><span class="n">chromosome</span><span class="p">)</span>
            <span class="o">&amp;</span> <span class="p">(</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">region</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">region</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">filter_condition</span><span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="n">StudyLocus</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.filter_credible_set" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">filter_credible_set</span><span class="p">(</span><span class="n">credible_interval</span><span class="p">:</span> <span class="n">CredibleInterval</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.filter_credible_set" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Annotate and filter study-locus tag variants based on given credible interval.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>credible_interval</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.CredibleInterval" href="#gentropy.dataset.study_locus.CredibleInterval">CredibleInterval</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Credible interval to filter for.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Filtered study-locus dataset.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">filter_credible_set</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
    <span class="n">credible_interval</span><span class="p">:</span> <span class="n">CredibleInterval</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Annotate and filter study-locus tag variants based on given credible interval.</span>

<span class="sd">    Args:</span>
<span class="sd">        credible_interval (CredibleInterval): Credible interval to filter for.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Filtered study-locus dataset.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">annotate_credible_sets</span><span class="p">()</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
                <span class="k">lambda</span> <span class="n">tag</span><span class="p">:</span> <span class="p">(</span><span class="n">tag</span><span class="p">[</span><span class="n">credible_interval</span><span class="o">.</span><span class="n">value</span><span class="p">]),</span>
            <span class="p">),</span>
        <span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_schema</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.filter_ld_set" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">filter_ld_set</span><span class="p">(</span><span class="n">ld_set</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">r2_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

<a href="#gentropy.dataset.study_locus.StudyLocus.filter_ld_set" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Filter the LD set by a given R2 threshold.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>ld_set</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>LD set</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>r2_threshold</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>R2 threshold to filter the LD set on</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Filtered LD index</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">filter_ld_set</span><span class="p">(</span><span class="n">ld_set</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">r2_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Filter the LD set by a given R2 threshold.</span>

<span class="sd">    Args:</span>
<span class="sd">        ld_set (Column): LD set</span>
<span class="sd">        r2_threshold (float): R2 threshold to filter the LD set on</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Filtered LD index</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
        <span class="n">ld_set</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">ld_set</span><span class="p">,</span>
            <span class="k">lambda</span> <span class="n">tag</span><span class="p">:</span> <span class="n">tag</span><span class="p">[</span><span class="s2">&quot;r2Overall&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">r2_threshold</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.find_overlaps" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">find_overlaps</span><span class="p">(</span><span class="n">intra_study_overlap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocusOverlap</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.find_overlaps" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Calculate overlapping study-locus.</p>
<p>Find overlapping study-locus that share at least one tagging variant. All GWAS-GWAS and all GWAS-Molecular traits are computed with the Molecular traits always
appearing on the right side.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>intra_study_overlap</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, finds intra-study overlaps for credible set deduplication. Default is False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocusOverlap</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus_overlap.StudyLocusOverlap" href="../study_locus_overlap/#gentropy.dataset.study_locus_overlap.StudyLocusOverlap">StudyLocusOverlap</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Pairs of overlapping study-locus with aligned tags.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">find_overlaps</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">intra_study_overlap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocusOverlap</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate overlapping study-locus.</span>

<span class="sd">    Find overlapping study-locus that share at least one tagging variant. All GWAS-GWAS and all GWAS-Molecular traits are computed with the Molecular traits always</span>
<span class="sd">    appearing on the right side.</span>

<span class="sd">    Args:</span>
<span class="sd">        intra_study_overlap (bool): If True, finds intra-study overlaps for credible set deduplication. Default is False.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocusOverlap: Pairs of overlapping study-locus with aligned tags.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">loci_to_overlap</span> <span class="o">=</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">())</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span>
        <span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;studyType&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
            <span class="s2">&quot;region&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.variantId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;tagVariantId&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.logBF&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;logBF&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.posteriorProbability&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;posteriorProbability&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.pValueMantissa&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.pValueExponent&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.beta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">persist</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="c1"># overlapping study-locus</span>
    <span class="n">peak_overlaps</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_overlapping_peaks</span><span class="p">(</span><span class="n">loci_to_overlap</span><span class="p">,</span> <span class="n">intra_study_overlap</span><span class="p">)</span>

    <span class="c1"># study-locus overlap by aligning overlapping variants</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_align_overlapping_tags</span><span class="p">(</span><span class="n">loci_to_overlap</span><span class="p">,</span> <span class="n">peak_overlaps</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.flag_trans_qtls" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">flag_trans_qtls</span><span class="p">(</span><span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span> <span class="n">target_index</span><span class="p">:</span> <span class="n">TargetIndex</span><span class="p">,</span> <span class="n">trans_threshold</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5000000</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.flag_trans_qtls" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Flagging transQTL credible sets based on genomic location of the measured gene.</p>
<p>Process:
0. Make sure that the <code>isTransQtl</code> column does not exist (remove if exists)
1. Enrich study-locus dataset with geneId based on study metadata. (only QTL studies are considered)
2. Enrich with transcription start site and chromosome of the studied gegne.
3. Flagging any tagging variant of QTL credible sets, if chromosome is different from the gene or distance is above the threshold.
4. Propagate flags to credible sets where any tags are considered as trans.
5. Return study locus object with annotation stored in 'isTransQtl<code>boolean column, where gwas credible sets will be</code>null`</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>study_index</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="../study_index/#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>study index to extract identifier of the measured gene</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>target_index</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.target_index.TargetIndex" href="../target_index/#gentropy.dataset.target_index.TargetIndex">TargetIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>target index bringing TSS and chromosome of the measured gene</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>trans_threshold</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Distance above which the QTL is considered trans. Default: 5_000_000bp</p>
              </div>
            </td>
            <td>
                  <code>5000000</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>new column added indicating if the QTL credibles sets are trans.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">flag_trans_qtls</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
    <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span>
    <span class="n">target_index</span><span class="p">:</span> <span class="n">TargetIndex</span><span class="p">,</span>
    <span class="n">trans_threshold</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5_000_000</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Flagging transQTL credible sets based on genomic location of the measured gene.</span>

<span class="sd">    Process:</span>
<span class="sd">    0. Make sure that the `isTransQtl` column does not exist (remove if exists)</span>
<span class="sd">    1. Enrich study-locus dataset with geneId based on study metadata. (only QTL studies are considered)</span>
<span class="sd">    2. Enrich with transcription start site and chromosome of the studied gegne.</span>
<span class="sd">    3. Flagging any tagging variant of QTL credible sets, if chromosome is different from the gene or distance is above the threshold.</span>
<span class="sd">    4. Propagate flags to credible sets where any tags are considered as trans.</span>
<span class="sd">    5. Return study locus object with annotation stored in &#39;isTransQtl` boolean column, where gwas credible sets will be `null`</span>

<span class="sd">    Args:</span>
<span class="sd">        study_index (StudyIndex): study index to extract identifier of the measured gene</span>
<span class="sd">        target_index (TargetIndex): target index bringing TSS and chromosome of the measured gene</span>
<span class="sd">        trans_threshold (int): Distance above which the QTL is considered trans. Default: 5_000_000bp</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: new column added indicating if the QTL credibles sets are trans.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># As the `geneId` column in the study index is optional, we have to test for that:</span>
    <span class="k">if</span> <span class="s2">&quot;geneId&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">study_index</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="c1"># We have to remove the column `isTransQtl` to ensure the column is not duplicated</span>
    <span class="c1"># The duplication can happen when one reads the StudyLocus from parquet with</span>
    <span class="c1"># predefined schema that already contains the `isTransQtl` column.</span>
    <span class="k">if</span> <span class="s2">&quot;isTransQtl&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;isTransQtl&quot;</span><span class="p">)</span>

    <span class="c1"># Process study index:</span>
    <span class="n">processed_studies</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">study_index</span><span class="o">.</span><span class="n">df</span>
        <span class="c1"># Dropping gwas studies. This ensures that only QTLs will have &quot;isTrans&quot; annotation:</span>
        <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;gwas&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;geneId&quot;</span><span class="p">,</span> <span class="s2">&quot;projectId&quot;</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Process study locus:</span>
    <span class="n">processed_credible_set</span> <span class="o">=</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span>
        <span class="c1"># Exploding locus to test all tag variants:</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;locus.variantId&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;locus.variantId&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">LongType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Process target index:</span>
    <span class="n">processed_targets</span> <span class="o">=</span> <span class="n">target_index</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;geneId&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tss&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;genomicLocation.chromosome&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;geneChromosome&quot;</span><span class="p">),</span>
    <span class="p">)</span>

    <span class="c1"># Pool datasets:</span>
    <span class="n">joined_data</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">processed_credible_set</span>
        <span class="c1"># Join processed studies:</span>
        <span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">processed_studies</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>
        <span class="c1"># Join processed targets:</span>
        <span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">processed_targets</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;geneId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
        <span class="c1"># Assign True/False for QTL studies:</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;isTagTrans&quot;</span><span class="p">,</span>
            <span class="c1"># The QTL signal is considered trans if the locus is on a different chromosome than the measured gene.</span>
            <span class="c1"># OR the distance from the gene&#39;s transcription start site is &gt; threshold.</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;geneChromosome&quot;</span><span class="p">))</span>
                <span class="o">|</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tss&quot;</span><span class="p">)</span> <span class="o">-</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">))</span> <span class="o">&gt;</span> <span class="n">trans_threshold</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">),</span>
            <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">)),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
            <span class="c1"># If all tagging variants of the locus is in trans position, the QTL is considered trans:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="s2">&quot;isTagTrans&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">)),</span> <span class="kc">False</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
            <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;isTransQtl&quot;</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="c1"># Adding new column, where the value is null for gwas loci:</span>
    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">joined_data</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.get_QC_column_name" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_QC_column_name</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="nb">str</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.dataset.study_locus.StudyLocus.get_QC_column_name" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Quality control column.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the quality control column.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">get_QC_column_name</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyLocus</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Quality control column.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Name of the quality control column.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="s2">&quot;qualityControls&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.get_QC_mappings" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_QC_mappings</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.dataset.study_locus.StudyLocus.get_QC_mappings" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Quality control flag to QC column category mappings.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>dict[str, str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dict[str, str]: Mapping between flag name and QC column category value.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">get_QC_mappings</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyLocus</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Quality control flag to QC column category mappings.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict[str, str]: Mapping between flag name and QC column category value.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">{</span><span class="n">member</span><span class="o">.</span><span class="n">name</span><span class="p">:</span> <span class="n">member</span><span class="o">.</span><span class="n">value</span> <span class="k">for</span> <span class="n">member</span> <span class="ow">in</span> <span class="n">StudyLocusQualityCheck</span><span class="p">}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.get_schema" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_schema</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StructType</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.dataset.study_locus.StudyLocus.get_schema" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Provides the schema for the StudyLocus dataset.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StructType</code></td>            <td>
                  <code><span title="pyspark.sql.types.StructType">StructType</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>schema for the StudyLocus dataset.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">get_schema</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyLocus</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">StructType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Provides the schema for the StudyLocus dataset.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StructType: schema for the StudyLocus dataset.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">parse_spark_schema</span><span class="p">(</span><span class="s2">&quot;study_locus.json&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.neglog_pvalue" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">neglog_pvalue</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.neglog_pvalue" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Returns the negative log p-value.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Negative log p-value</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">neglog_pvalue</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the negative log p-value.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Negative log p-value</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">calculate_neglog_pvalue</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueMantissa</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueExponent</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.qc_MHC_region" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">qc_MHC_region</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.qc_MHC_region" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Adds qualityControl flag when lead overlaps with MHC region.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>including qualityControl flag if in MHC region.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span>
<span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">qc_MHC_region</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Adds qualityControl flag when lead overlaps with MHC region.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: including qualityControl flag if in MHC region.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">region</span> <span class="o">=</span> <span class="n">GenomicRegion</span><span class="o">.</span><span class="n">from_known_genomic_region</span><span class="p">(</span><span class="n">KnownGenomicRegions</span><span class="o">.</span><span class="n">MHC</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
        <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
            <span class="p">(</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">region</span><span class="o">.</span><span class="n">chromosome</span><span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span>
                    <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">region</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">region</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">IN_MHC</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.qc_abnormal_pips" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">qc_abnormal_pips</span><span class="p">(</span><span class="n">sum_pips_lower_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.99</span><span class="p">,</span> <span class="n">sum_pips_upper_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0001</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.qc_abnormal_pips" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Filter study-locus by sum of posterior inclusion probabilities to ensure that the sum of PIPs is within a given range.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>sum_pips_lower_threshold</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Lower threshold for the sum of PIPs.</p>
              </div>
            </td>
            <td>
                  <code>0.99</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sum_pips_upper_threshold</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Upper threshold for the sum of PIPs.</p>
              </div>
            </td>
            <td>
                  <code>1.0001</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Filtered study-locus dataset.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">qc_abnormal_pips</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
    <span class="n">sum_pips_lower_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.99</span><span class="p">,</span>
    <span class="c1"># Set slightly above 1 to account for floating point errors</span>
    <span class="n">sum_pips_upper_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0001</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Filter study-locus by sum of posterior inclusion probabilities to ensure that the sum of PIPs is within a given range.</span>

<span class="sd">    Args:</span>
<span class="sd">        sum_pips_lower_threshold (float): Lower threshold for the sum of PIPs.</span>
<span class="sd">        sum_pips_upper_threshold (float): Upper threshold for the sum of PIPs.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Filtered study-locus dataset.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># QC column might not be present so we have to be ready to handle it:</span>
    <span class="n">qc_select_expression</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;qualityControls&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
        <span class="k">else</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">ArrayType</span><span class="p">(</span><span class="n">StringType</span><span class="p">()))</span>
    <span class="p">)</span>

    <span class="n">flag</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
        <span class="s2">&quot;sumPosteriorProbability&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">0.0</span><span class="p">),</span>
            <span class="k">lambda</span> <span class="n">acc</span><span class="p">,</span> <span class="n">x</span><span class="p">:</span> <span class="n">acc</span> <span class="o">+</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;posteriorProbability&quot;</span><span class="p">],</span>
        <span class="p">),</span>
    <span class="p">)</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
        <span class="s2">&quot;pipOutOfRange&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumPosteriorProbability&quot;</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">sum_pips_lower_threshold</span><span class="p">)</span>
            <span class="o">|</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumPosteriorProbability&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">sum_pips_upper_threshold</span><span class="p">),</span>
            <span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="kc">False</span><span class="p">),</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
            <span class="n">flag</span>
            <span class="c1"># Flagging loci with failed studies:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">qc_select_expression</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;pipOutOfRange&quot;</span><span class="p">),</span>
                    <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">ABNORMAL_PIPS</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;sumPosteriorProbability&quot;</span><span class="p">,</span> <span class="s2">&quot;pipOutOfRange&quot;</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.qc_explained_by_SuSiE" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">qc_explained_by_SuSiE</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.qc_explained_by_SuSiE" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Flag associations that are explained by SuSiE associations.</p>
<p>Credible sets overlapping in the same region as a SuSiE credible set are flagged as explained by SuSiE.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Updated study locus with SuSiE explained flags.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1199</span>
<span class="normal">1200</span>
<span class="normal">1201</span>
<span class="normal">1202</span>
<span class="normal">1203</span>
<span class="normal">1204</span>
<span class="normal">1205</span>
<span class="normal">1206</span>
<span class="normal">1207</span>
<span class="normal">1208</span>
<span class="normal">1209</span>
<span class="normal">1210</span>
<span class="normal">1211</span>
<span class="normal">1212</span>
<span class="normal">1213</span>
<span class="normal">1214</span>
<span class="normal">1215</span>
<span class="normal">1216</span>
<span class="normal">1217</span>
<span class="normal">1218</span>
<span class="normal">1219</span>
<span class="normal">1220</span>
<span class="normal">1221</span>
<span class="normal">1222</span>
<span class="normal">1223</span>
<span class="normal">1224</span>
<span class="normal">1225</span>
<span class="normal">1226</span>
<span class="normal">1227</span>
<span class="normal">1228</span>
<span class="normal">1229</span>
<span class="normal">1230</span>
<span class="normal">1231</span>
<span class="normal">1232</span>
<span class="normal">1233</span>
<span class="normal">1234</span>
<span class="normal">1235</span>
<span class="normal">1236</span>
<span class="normal">1237</span>
<span class="normal">1238</span>
<span class="normal">1239</span>
<span class="normal">1240</span>
<span class="normal">1241</span>
<span class="normal">1242</span>
<span class="normal">1243</span>
<span class="normal">1244</span>
<span class="normal">1245</span>
<span class="normal">1246</span>
<span class="normal">1247</span>
<span class="normal">1248</span>
<span class="normal">1249</span>
<span class="normal">1250</span>
<span class="normal">1251</span>
<span class="normal">1252</span>
<span class="normal">1253</span>
<span class="normal">1254</span>
<span class="normal">1255</span>
<span class="normal">1256</span>
<span class="normal">1257</span>
<span class="normal">1258</span>
<span class="normal">1259</span>
<span class="normal">1260</span>
<span class="normal">1261</span>
<span class="normal">1262</span>
<span class="normal">1263</span>
<span class="normal">1264</span>
<span class="normal">1265</span>
<span class="normal">1266</span>
<span class="normal">1267</span>
<span class="normal">1268</span>
<span class="normal">1269</span>
<span class="normal">1270</span>
<span class="normal">1271</span>
<span class="normal">1272</span>
<span class="normal">1273</span>
<span class="normal">1274</span>
<span class="normal">1275</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">qc_explained_by_SuSiE</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Flag associations that are explained by SuSiE associations.</span>

<span class="sd">    Credible sets overlapping in the same region as a SuSiE credible set are flagged as explained by SuSiE.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Updated study locus with SuSiE explained flags.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># unique study-regions covered by SuSie credible sets</span>
    <span class="n">susie_study_regions</span> <span class="o">=</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE_INF</span><span class="o">.</span><span class="n">value</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
            <span class="s2">&quot;locusStart&quot;</span><span class="p">,</span>
            <span class="s2">&quot;locusEnd&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inSuSiE&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">distinct</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="c1"># non SuSiE credible sets (studyLocusId) overlapping in any variant with SuSiE locus</span>
    <span class="n">redundant_study_locus</span> <span class="o">=</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE_INF</span><span class="o">.</span><span class="n">value</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;l&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span>
        <span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;l.variantId&quot;</span><span class="p">),</span> <span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;tag_position&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;study_locus&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">susie_study_regions</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;regions&quot;</span><span class="p">),</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span>
            <span class="n">on</span><span class="o">=</span><span class="p">[</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_locus.chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;regions.chromosome&quot;</span><span class="p">))</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_locus.studyId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;regions.studyId&quot;</span><span class="p">))</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_locus.tag_position&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;regions.locusStart&quot;</span><span class="p">))</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_locus.tag_position&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;regions.locusEnd&quot;</span><span class="p">))</span>
            <span class="p">],</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span> <span class="s2">&quot;inSuSiE&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">distinct</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">redundant_study_locus</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="c1"># credible set in SuSiE overlapping region</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;inSuSiE&quot;</span><span class="p">)</span>
                    <span class="c1"># credible set not based on SuSiE</span>
                    <span class="o">&amp;</span> <span class="p">(</span>
                        <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                            <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                            <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">SUSIE_INF</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                        <span class="p">)</span>
                    <span class="p">),</span>
                    <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">EXPLAINED_BY_SUSIE</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;inSuSiE&quot;</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="n">StudyLocus</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.qc_redundant_top_hits_from_PICS" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">qc_redundant_top_hits_from_PICS</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.qc_redundant_top_hits_from_PICS" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Flag associations from top hits when the study contains other PICS associations from summary statistics.</p>
<p>This flag can be useful to identify top hits that should be explained by other associations in the study derived from the summary statistics.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Updated study locus with redundant top hits flagged.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1161</span>
<span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span>
<span class="normal">1174</span>
<span class="normal">1175</span>
<span class="normal">1176</span>
<span class="normal">1177</span>
<span class="normal">1178</span>
<span class="normal">1179</span>
<span class="normal">1180</span>
<span class="normal">1181</span>
<span class="normal">1182</span>
<span class="normal">1183</span>
<span class="normal">1184</span>
<span class="normal">1185</span>
<span class="normal">1186</span>
<span class="normal">1187</span>
<span class="normal">1188</span>
<span class="normal">1189</span>
<span class="normal">1190</span>
<span class="normal">1191</span>
<span class="normal">1192</span>
<span class="normal">1193</span>
<span class="normal">1194</span>
<span class="normal">1195</span>
<span class="normal">1196</span>
<span class="normal">1197</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">qc_redundant_top_hits_from_PICS</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Flag associations from top hits when the study contains other PICS associations from summary statistics.</span>

<span class="sd">    This flag can be useful to identify top hits that should be explained by other associations in the study derived from the summary statistics.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Updated study locus with redundant top hits flagged.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">studies_with_pics_sumstats</span> <span class="o">=</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;finemappingMethod&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">FinemappingMethod</span><span class="o">.</span><span class="n">PICS</span><span class="o">.</span><span class="n">value</span><span class="p">)</span>
        <span class="c1"># Returns True if the study contains any PICS associations from summary statistics</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;hasPicsSumstats&quot;</span><span class="p">,</span>
            <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span> <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">TOP_HIT</span><span class="o">.</span><span class="n">value</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasPicsSumstats&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;studiesWithPicsSumstats&quot;</span><span class="p">))</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">studies_with_pics_sumstats</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                    <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span> <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">TOP_HIT</span><span class="o">.</span><span class="n">value</span>
                <span class="p">)</span>
                <span class="o">&amp;</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studiesWithPicsSumstats&quot;</span><span class="p">),</span>
                <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">REDUNDANT_PICS_TOP_HIT</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;studiesWithPicsSumstats&quot;</span><span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="n">StudyLocus</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.unique_variants_in_locus" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">unique_variants_in_locus</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.unique_variants_in_locus" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>All unique variants collected in a <code>StudyLocus</code> dataframe.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>DataFrame</code></td>            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dataframe containing <code>variantId</code> and <code>chromosome</code> columns.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">unique_variants_in_locus</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;All unique variants collected in a `StudyLocus` dataframe.</span>

<span class="sd">    Returns:</span>
<span class="sd">        DataFrame: A dataframe containing `variantId` and `chromosome` columns.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;variantId&quot;</span><span class="p">,</span>
            <span class="c1"># Joint array of variants in that studylocus. Locus can be null</span>
            <span class="n">f</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">array_union</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">)),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.variantId&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">()),</span>
                <span class="p">)</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span> <span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">distinct</span><span class="p">()</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.validate_chromosome_label" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">validate_chromosome_label</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.validate_chromosome_label" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Flagging study loci, where chromosome is coded not as 1:22, X, Y, Xy and MT.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Updated study locus with quality control flags.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate_chromosome_label</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Flagging study loci, where chromosome is coded not as 1:22, X, Y, Xy and MT.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Updated study locus with quality control flags.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># QC column might not be present in the variant index schema, so we have to be ready to handle it:</span>
    <span class="n">qc_select_expression</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;qualityControls&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
        <span class="k">else</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">ArrayType</span><span class="p">(</span><span class="n">StringType</span><span class="p">()))</span>
    <span class="p">)</span>
    <span class="n">valid_chromosomes</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">,</span> <span class="s2">&quot;Y&quot;</span><span class="p">,</span> <span class="s2">&quot;XY&quot;</span><span class="p">,</span> <span class="s2">&quot;MT&quot;</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">qc_select_expression</span><span class="p">,</span>
                    <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">valid_chromosomes</span><span class="p">),</span>
                    <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">INVALID_CHROMOSOME</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
        <span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.validate_lead_pvalue" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">validate_lead_pvalue</span><span class="p">(</span><span class="n">pvalue_cutoff</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.validate_lead_pvalue" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Flag associations below significant threshold.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>pvalue_cutoff</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>association p-value cut-off</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Updated study locus with quality control flags.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate_lead_pvalue</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">pvalue_cutoff</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Flag associations below significant threshold.</span>

<span class="sd">    Args:</span>
<span class="sd">        pvalue_cutoff (float): association p-value cut-off</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Updated study locus with quality control flags.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span>
    <span class="n">qc_colname</span> <span class="o">=</span> <span class="n">StudyLocus</span><span class="o">.</span><span class="n">get_QC_column_name</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">qc_colname</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="n">qc_colname</span><span class="p">,</span>
            <span class="n">create_empty_column_if_not_exists</span><span class="p">(</span>
                <span class="n">qc_colname</span><span class="p">,</span>
                <span class="n">get_struct_field_schema</span><span class="p">(</span><span class="n">StudyLocus</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span> <span class="n">qc_colname</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
            <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="n">qc_colname</span><span class="p">,</span>
                <span class="c1"># Because this QC might already run on the dataset, the unique set of flags is generated:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">array_distinct</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_qc_subsignificant_associations</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">),</span>
                        <span class="n">pvalue_cutoff</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">),</span>
            <span class="p">)</span>
        <span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.validate_study" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">validate_study</span><span class="p">(</span><span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.validate_study" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Flagging study loci if the corresponding study has issues.</p>
<p>There are two different potential flags:
- flagged study: flagging locus if the study has quality control flags.
- study with summary statistics for top hit: flagging locus if the study has available summary statistics.
- missing study: flagging locus if the study was not found in the reference study index.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>study_index</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="../study_index/#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study index to resolve study types.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Updated study locus with quality control flags.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate_study</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Flagging study loci if the corresponding study has issues.</span>

<span class="sd">    There are two different potential flags:</span>
<span class="sd">    - flagged study: flagging locus if the study has quality control flags.</span>
<span class="sd">    - study with summary statistics for top hit: flagging locus if the study has available summary statistics.</span>
<span class="sd">    - missing study: flagging locus if the study was not found in the reference study index.</span>

<span class="sd">    Args:</span>
<span class="sd">        study_index (StudyIndex): Study index to resolve study types.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Updated study locus with quality control flags.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Quality controls is not a mandatory field in the study index schema, so we have to be ready to handle it:</span>
    <span class="n">qc_select_expression</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;qualityControls&quot;</span> <span class="ow">in</span> <span class="n">study_index</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
        <span class="k">else</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">StringType</span><span class="p">())</span>
    <span class="p">)</span>

    <span class="c1"># The study Id of the study index needs to be kept, because we would not know which study was in the index after the left join:</span>
    <span class="n">study_flags</span> <span class="o">=</span> <span class="n">study_index</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;study_studyId&quot;</span><span class="p">),</span>
        <span class="n">qc_select_expression</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;study_qualityControls&quot;</span><span class="p">),</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">study_flags</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_studyId&quot;</span><span class="p">),</span> <span class="s2">&quot;left&quot;</span>
            <span class="p">)</span>
            <span class="c1"># Flagging loci with flagged studies - without propagating the actual flags:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">StudyLocus</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_qualityControls&quot;</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span>
                    <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">FLAGGED_STUDY</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Flagging top-hits, where the study has available summary statistics:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">StudyLocus</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="c1"># Condition is true, if the study has summary statistics available and the locus is a top hit:</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                        <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">TOP_HIT</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="o">&amp;</span> <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_qualityControls&quot;</span><span class="p">),</span>
                        <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">SUMSTATS_NOT_AVAILABLE</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                    <span class="p">),</span>
                    <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">TOP_HIT_AND_SUMMARY_STATS</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Flagging loci where no studies were found:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">StudyLocus</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;study_studyId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span>
                    <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">MISSING_STUDY</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;study_studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;study_qualityControls&quot;</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.validate_unique_study_locus_id" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">validate_unique_study_locus_id</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.validate_unique_study_locus_id" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Validating the uniqueness of study-locus identifiers and flagging duplicated studyloci.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>with flagged duplicated studies.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate_unique_study_locus_id</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Validating the uniqueness of study-locus identifiers and flagging duplicated studyloci.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: with flagged duplicated studies.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">flag_duplicates</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)),</span>
                <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">DUPLICATED_STUDYLOCUS_ID</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="n">StudyLocus</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.validate_variant_identifiers" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">validate_variant_identifiers</span><span class="p">(</span><span class="n">variant_index</span><span class="p">:</span> <span class="n">VariantIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.validate_variant_identifiers" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Flagging study loci, where tagging variant identifiers are not found in variant index.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>variant_index</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.variant_index.VariantIndex" href="../variant_index/#gentropy.dataset.variant_index.VariantIndex">VariantIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Variant index to resolve variant identifiers.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Updated study locus with quality control flags.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate_variant_identifiers</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">variant_index</span><span class="p">:</span> <span class="n">VariantIndex</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Flagging study loci, where tagging variant identifiers are not found in variant index.</span>

<span class="sd">    Args:</span>
<span class="sd">        variant_index (VariantIndex): Variant index to resolve variant identifiers.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Updated study locus with quality control flags.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># QC column might not be present in the variant index schema, so we have to be ready to handle it:</span>
    <span class="n">qc_select_expression</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;qualityControls&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
        <span class="k">else</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">ArrayType</span><span class="p">(</span><span class="n">StringType</span><span class="p">()))</span>
    <span class="p">)</span>

    <span class="c1"># Find out which study loci have variants not in the variant index:</span>
    <span class="n">flag</span> <span class="o">=</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span>
        <span class="c1"># Exploding locus:</span>
        <span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span>
        <span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span> <span class="s2">&quot;locus.variantId&quot;</span><span class="p">)</span>
        <span class="c1"># Join with variant index variants:</span>
        <span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">variant_index</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inVariantIndex&quot;</span><span class="p">)</span>
            <span class="p">),</span>
            <span class="n">on</span><span class="o">=</span><span class="s2">&quot;variantId&quot;</span><span class="p">,</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># Flagging variants not in the variant index:</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;inVariantIndex&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;inVariantIndex&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">())</span>
        <span class="c1"># Flagging study loci with ANY variants not in the variant index:</span>
        <span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="s2">&quot;inVariantIndex&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inVariantIndex&quot;</span><span class="p">))</span>
        <span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span><span class="s2">&quot;inVariantIndex&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;toFlag&quot;</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">flag</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">qc_select_expression</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;toFlag&quot;</span><span class="p">),</span>
                    <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">INVALID_VARIANT_IDENTIFIER</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;toFlag&quot;</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.window_based_clumping" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">window_based_clumping</span><span class="p">(</span><span class="n">window_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">WindowBasedClumpingStepConfig</span><span class="p">()</span><span class="o">.</span><span class="n">distance</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.window_based_clumping" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Clump study locus by window size.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>window_size</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Window size for clumping.</p>
              </div>
            </td>
            <td>
                  <code>distance</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Clumped study locus, where clumped associations are flagged.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1362</span>
<span class="normal">1363</span>
<span class="normal">1364</span>
<span class="normal">1365</span>
<span class="normal">1366</span>
<span class="normal">1367</span>
<span class="normal">1368</span>
<span class="normal">1369</span>
<span class="normal">1370</span>
<span class="normal">1371</span>
<span class="normal">1372</span>
<span class="normal">1373</span>
<span class="normal">1374</span>
<span class="normal">1375</span>
<span class="normal">1376</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">window_based_clumping</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
    <span class="n">window_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">WindowBasedClumpingStepConfig</span><span class="p">()</span><span class="o">.</span><span class="n">distance</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Clump study locus by window size.</span>

<span class="sd">    Args:</span>
<span class="sd">        window_size (int): Window size for clumping.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Clumped study locus, where clumped associations are flagged.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">gentropy.method.window_based_clumping</span><span class="w"> </span><span class="kn">import</span> <span class="n">WindowBasedClumping</span>

    <span class="k">return</span> <span class="n">WindowBasedClumping</span><span class="o">.</span><span class="n">clump</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">window_size</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><hr />


<div class="doc doc-object doc-class">



<h2 id="gentropy.dataset.study_locus.FinemappingMethod" class="doc doc-heading">
            <code>gentropy.dataset.study_locus.FinemappingMethod</code>


<a href="#gentropy.dataset.study_locus.FinemappingMethod" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><span title="enum.Enum">Enum</span></code></p>


        <p>Finemapping method enum.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.FinemappingMethod.PICS">PICS</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>PICS</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.FinemappingMethod.SUSIE">SUSIE</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>SuSiE method</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.FinemappingMethod.SUSIE_INF">SUSIE_INF</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>SuSiE-inf method implemented in <code>gentropy</code></p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>






              <details class="quote">
                <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">FinemappingMethod</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Finemapping method enum.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        PICS (str): PICS</span>
<span class="sd">        SUSIE (str): SuSiE method</span>
<span class="sd">        SUSIE_INF (str): SuSiE-inf method implemented in `gentropy`</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">PICS</span> <span class="o">=</span> <span class="s2">&quot;PICS&quot;</span>
    <span class="n">SUSIE</span> <span class="o">=</span> <span class="s2">&quot;SuSie&quot;</span>
    <span class="n">SUSIE_INF</span> <span class="o">=</span> <span class="s2">&quot;SuSiE-inf&quot;</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">











  </div>

    </div>

</div><hr />


<div class="doc doc-object doc-class">



<h2 id="gentropy.dataset.study_locus.StudyLocusQualityCheck" class="doc doc-heading">
            <code>gentropy.dataset.study_locus.StudyLocusQualityCheck</code>


<a href="#gentropy.dataset.study_locus.StudyLocusQualityCheck" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><span title="enum.Enum">Enum</span></code></p>


        <p>Study-Locus quality control options listing concerns on the quality of the association.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.SUBSIGNIFICANT_FLAG">SUBSIGNIFICANT_FLAG</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>p-value below significance threshold</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.NO_GENOMIC_LOCATION_FLAG">NO_GENOMIC_LOCATION_FLAG</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Incomplete genomic mapping</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.COMPOSITE_FLAG">COMPOSITE_FLAG</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Composite association due to variant x variant interactions</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.INCONSISTENCY_FLAG">INCONSISTENCY_FLAG</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Inconsistencies in the reported variants</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.NON_MAPPED_VARIANT_FLAG">NON_MAPPED_VARIANT_FLAG</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Variant not mapped to GnomAd</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.PALINDROMIC_ALLELE_FLAG">PALINDROMIC_ALLELE_FLAG</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Alleles are palindromic - cannot harmonize</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.AMBIGUOUS_STUDY">AMBIGUOUS_STUDY</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Association with ambiguous study</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.UNRESOLVED_LD">UNRESOLVED_LD</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Variant not found in LD reference</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.LD_CLUMPED">LD_CLUMPED</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Explained by a more significant variant in high LD</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.WINDOW_CLUMPED">WINDOW_CLUMPED</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Explained by a more significant variant in the same window</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.NO_POPULATION">NO_POPULATION</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study does not have population annotation to resolve LD</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.FLAGGED_STUDY">FLAGGED_STUDY</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study has quality control flag(s)</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.MISSING_STUDY">MISSING_STUDY</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Flagging study loci if the study is not found in the study index as a reference</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.DUPLICATED_STUDYLOCUS_ID">DUPLICATED_STUDYLOCUS_ID</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study-locus identifier is not unique</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.INVALID_VARIANT_IDENTIFIER">INVALID_VARIANT_IDENTIFIER</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Flagging study loci where identifier of any tagging variant was not found in the variant index</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.TOP_HIT">TOP_HIT</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study locus from curated top hit</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.IN_MHC">IN_MHC</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Flagging study loci in the MHC region</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.REDUNDANT_PICS_TOP_HIT">REDUNDANT_PICS_TOP_HIT</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Flagging study loci in studies with PICS results from summary statistics</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.EXPLAINED_BY_SUSIE">EXPLAINED_BY_SUSIE</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study locus in region explained by a SuSiE credible set</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.ABNORMAL_PIPS">ABNORMAL_PIPS</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Flagging study loci with a sum of PIPs that are not in [0.99,1]</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.OUT_OF_SAMPLE_LD">OUT_OF_SAMPLE_LD</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study locus finemapped without in-sample LD reference</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.INVALID_CHROMOSOME">INVALID_CHROMOSOME</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Chromosome not in 1:22, X, Y, XY or MT</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.TOP_HIT_AND_SUMMARY_STATS">TOP_HIT_AND_SUMMARY_STATS</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Curated top hit is flagged because summary statistics are available for study</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>






              <details class="quote">
                <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">StudyLocusQualityCheck</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Study-Locus quality control options listing concerns on the quality of the association.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        SUBSIGNIFICANT_FLAG (str): p-value below significance threshold</span>
<span class="sd">        NO_GENOMIC_LOCATION_FLAG (str): Incomplete genomic mapping</span>
<span class="sd">        COMPOSITE_FLAG (str): Composite association due to variant x variant interactions</span>
<span class="sd">        INCONSISTENCY_FLAG (str): Inconsistencies in the reported variants</span>
<span class="sd">        NON_MAPPED_VARIANT_FLAG (str): Variant not mapped to GnomAd</span>
<span class="sd">        PALINDROMIC_ALLELE_FLAG (str): Alleles are palindromic - cannot harmonize</span>
<span class="sd">        AMBIGUOUS_STUDY (str): Association with ambiguous study</span>
<span class="sd">        UNRESOLVED_LD (str): Variant not found in LD reference</span>
<span class="sd">        LD_CLUMPED (str): Explained by a more significant variant in high LD</span>
<span class="sd">        WINDOW_CLUMPED (str): Explained by a more significant variant in the same window</span>
<span class="sd">        NO_POPULATION (str): Study does not have population annotation to resolve LD</span>
<span class="sd">        FLAGGED_STUDY (str): Study has quality control flag(s)</span>
<span class="sd">        MISSING_STUDY (str): Flagging study loci if the study is not found in the study index as a reference</span>
<span class="sd">        DUPLICATED_STUDYLOCUS_ID (str): Study-locus identifier is not unique</span>
<span class="sd">        INVALID_VARIANT_IDENTIFIER (str): Flagging study loci where identifier of any tagging variant was not found in the variant index</span>
<span class="sd">        TOP_HIT (str): Study locus from curated top hit</span>
<span class="sd">        IN_MHC (str): Flagging study loci in the MHC region</span>
<span class="sd">        REDUNDANT_PICS_TOP_HIT (str): Flagging study loci in studies with PICS results from summary statistics</span>
<span class="sd">        EXPLAINED_BY_SUSIE (str): Study locus in region explained by a SuSiE credible set</span>
<span class="sd">        ABNORMAL_PIPS (str): Flagging study loci with a sum of PIPs that are not in [0.99,1]</span>
<span class="sd">        OUT_OF_SAMPLE_LD (str): Study locus finemapped without in-sample LD reference</span>
<span class="sd">        INVALID_CHROMOSOME (str): Chromosome not in 1:22, X, Y, XY or MT</span>
<span class="sd">        TOP_HIT_AND_SUMMARY_STATS (str): Curated top hit is flagged because summary statistics are available for study</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">SUBSIGNIFICANT_FLAG</span> <span class="o">=</span> <span class="s2">&quot;Subsignificant p-value&quot;</span>
    <span class="n">NO_GENOMIC_LOCATION_FLAG</span> <span class="o">=</span> <span class="s2">&quot;Incomplete genomic mapping&quot;</span>
    <span class="n">COMPOSITE_FLAG</span> <span class="o">=</span> <span class="s2">&quot;Composite association&quot;</span>
    <span class="n">INCONSISTENCY_FLAG</span> <span class="o">=</span> <span class="s2">&quot;Variant inconsistency&quot;</span>
    <span class="n">NON_MAPPED_VARIANT_FLAG</span> <span class="o">=</span> <span class="s2">&quot;No mapping in GnomAd&quot;</span>
    <span class="n">PALINDROMIC_ALLELE_FLAG</span> <span class="o">=</span> <span class="s2">&quot;Palindrome alleles - cannot harmonize&quot;</span>
    <span class="n">AMBIGUOUS_STUDY</span> <span class="o">=</span> <span class="s2">&quot;Association with ambiguous study&quot;</span>
    <span class="n">UNRESOLVED_LD</span> <span class="o">=</span> <span class="s2">&quot;Variant not found in LD reference&quot;</span>
    <span class="n">LD_CLUMPED</span> <span class="o">=</span> <span class="s2">&quot;Explained by a more significant variant in high LD&quot;</span>
    <span class="n">WINDOW_CLUMPED</span> <span class="o">=</span> <span class="s2">&quot;Explained by a more significant variant in the same window&quot;</span>
    <span class="n">NO_POPULATION</span> <span class="o">=</span> <span class="s2">&quot;Study does not have population annotation to resolve LD&quot;</span>
    <span class="n">FLAGGED_STUDY</span> <span class="o">=</span> <span class="s2">&quot;Study has quality control flag(s)&quot;</span>
    <span class="n">MISSING_STUDY</span> <span class="o">=</span> <span class="s2">&quot;Study not found in the study index&quot;</span>
    <span class="n">DUPLICATED_STUDYLOCUS_ID</span> <span class="o">=</span> <span class="s2">&quot;Non-unique study locus identifier&quot;</span>
    <span class="n">INVALID_VARIANT_IDENTIFIER</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;Some variant identifiers of this locus were not found in variant index&quot;</span>
    <span class="p">)</span>
    <span class="n">IN_MHC</span> <span class="o">=</span> <span class="s2">&quot;MHC region&quot;</span>
    <span class="n">REDUNDANT_PICS_TOP_HIT</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;PICS results from summary statistics available for this same study&quot;</span>
    <span class="p">)</span>
    <span class="n">TOP_HIT</span> <span class="o">=</span> <span class="s2">&quot;Study locus from curated top hit&quot;</span>
    <span class="n">EXPLAINED_BY_SUSIE</span> <span class="o">=</span> <span class="s2">&quot;Study locus in region explained by a SuSiE credible set&quot;</span>
    <span class="n">OUT_OF_SAMPLE_LD</span> <span class="o">=</span> <span class="s2">&quot;Study locus finemapped without in-sample LD reference&quot;</span>
    <span class="n">ABNORMAL_PIPS</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;Study locus with a sum of PIPs that not in the expected range [0.95,1]&quot;</span>
    <span class="p">)</span>
    <span class="n">INVALID_CHROMOSOME</span> <span class="o">=</span> <span class="s2">&quot;Chromosome not in 1:22, X, Y, XY or MT&quot;</span>
    <span class="n">TOP_HIT_AND_SUMMARY_STATS</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;Curated top hit is flagged because summary statistics are available for study&quot;</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">











  </div>

    </div>

</div><hr />


<div class="doc doc-object doc-class">



<h2 id="gentropy.dataset.study_locus.CredibleInterval" class="doc doc-heading">
            <code>gentropy.dataset.study_locus.CredibleInterval</code>


<a href="#gentropy.dataset.study_locus.CredibleInterval" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><span title="enum.Enum">Enum</span></code></p>


        <p>Credible interval enum.</p>
<p>Interval within which an unobserved parameter value falls with a particular probability.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.CredibleInterval.IS95">IS95</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>95% credible interval</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.CredibleInterval.IS99">IS99</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>99% credible interval</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>






              <details class="quote">
                <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">CredibleInterval</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Credible interval enum.</span>

<span class="sd">    Interval within which an unobserved parameter value falls with a particular probability.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        IS95 (str): 95% credible interval</span>
<span class="sd">        IS99 (str): 99% credible interval</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">IS95</span> <span class="o">=</span> <span class="s2">&quot;is95CredibleSet&quot;</span>
    <span class="n">IS99</span> <span class="o">=</span> <span class="s2">&quot;is99CredibleSet&quot;</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">











  </div>

    </div>

</div><h2 id="schema">Schema<a class="headerlink" href="#schema" title="Permanent link">&para;</a></h2>
<div class="highlight"><pre><span></span><code>root
 |-- studyLocusId: string (nullable = false)
 |-- studyType: string (nullable = true)
 |-- variantId: string (nullable = false)
 |-- chromosome: string (nullable = true)
 |-- position: integer (nullable = true)
 |-- region: string (nullable = true)
 |-- studyId: string (nullable = false)
 |-- beta: double (nullable = true)
 |-- zScore: double (nullable = true)
 |-- pValueMantissa: float (nullable = true)
 |-- pValueExponent: integer (nullable = true)
 |-- effectAlleleFrequencyFromSource: float (nullable = true)
 |-- standardError: double (nullable = true)
 |-- subStudyDescription: string (nullable = true)
 |-- qualityControls: array (nullable = true)
 |    |-- element: string (containsNull = false)
 |-- finemappingMethod: string (nullable = true)
 |-- credibleSetIndex: integer (nullable = true)
 |-- credibleSetlog10BF: double (nullable = true)
 |-- purityMeanR2: double (nullable = true)
 |-- purityMinR2: double (nullable = true)
 |-- locusStart: integer (nullable = true)
 |-- locusEnd: integer (nullable = true)
 |-- sampleSize: integer (nullable = true)
 |-- ldSet: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- tagVariantId: string (nullable = true)
 |    |    |-- r2Overall: double (nullable = true)
 |-- locus: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- is95CredibleSet: boolean (nullable = true)
 |    |    |-- is99CredibleSet: boolean (nullable = true)
 |    |    |-- logBF: double (nullable = true)
 |    |    |-- posteriorProbability: double (nullable = true)
 |    |    |-- variantId: string (nullable = true)
 |    |    |-- pValueMantissa: float (nullable = true)
 |    |    |-- pValueExponent: integer (nullable = true)
 |    |    |-- beta: double (nullable = true)
 |    |    |-- standardError: double (nullable = true)
 |    |    |-- r2Overall: double (nullable = true)
 |-- confidence: string (nullable = true)
 |-- isTransQtl: boolean (nullable = true)
</code></pre></div>
<hr />
<div class="md-source-file">
  <small>
    <ul
      class="metadata page-metadata"
      data-bi-name="page info"
      lang="en-us"
      dir="ltr"
    >
      <!-- mkdocs-git-revision-date-plugin -->
      
      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3h-2Z"/></svg>
        </span>
        <span class="label"> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class="timeago" datetime="2023-02-10T22:28:52+00:00" locale="en"></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2023-02-10</span> </span>
      </li>
       
      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg>
        </span>
        <span class="label"> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class="timeago" datetime="2024-11-08T12:41:54+00:00" locale="en"></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2024-11-08</span> </span>
      </li>
      

      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2A10 10 0 0 0 2 12c0 4.42 2.87 8.17 6.84 9.5.5.08.66-.23.66-.5v-1.69c-2.77.6-3.36-1.34-3.36-1.34-.46-1.16-1.11-1.47-1.11-1.47-.91-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.87 1.52 2.34 1.07 2.91.83.09-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.92 0-1.11.38-2 1.03-2.71-.1-.25-.45-1.29.1-2.64 0 0 .84-.27 2.75 1.02.79-.22 1.65-.33 2.5-.33.85 0 1.71.11 2.5.33 1.91-1.29 2.75-1.02 2.75-1.02.55 1.35.2 2.39.1 2.64.65.71 1.03 1.6 1.03 2.71 0 3.82-2.34 4.66-4.57 4.91.36.31.69.92.69 1.85V21c0 .27.16.59.67.5C19.14 20.16 22 16.42 22 12A10 10 0 0 0 12 2Z"/></svg>
        </span>
        <span class="label"> Contributors </span>
      </li>
      <li class="contributors-holder">
        <ul class="contributors" data-bi-name="contributors"><li>
            <a
              href="https://github.com/d0choa"
              title="d0choa"
              data-bi-name="contributorprofile"
              target="_blank"
              ><img src="https://avatars.githubusercontent.com/u/5097586?v=4" alt="d0choa"
            /></a>
          </li><li>
            <a
              href="https://github.com/web-flow"
              title="web-flow"
              data-bi-name="contributorprofile"
              target="_blank"
              ><img src="https://avatars.githubusercontent.com/u/19864447?v=4" alt="web-flow"
            /></a>
          </li></ul>
      </li>
    </ul>
  </small>
</div>




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/gentropy" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.top", "navigation.tabs", "navigation.tabs.sticky", "content.code.annotate", "search.suggest", "search.highlight"], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.3220b9d7.min.js"></script>
      
        <script src="../../../js/timeago.min.js"></script>
      
        <script src="../../../js/timeago_mkdocs_material.js"></script>
      
        <script src="../../../assets/javascripts/extra.js"></script>
      
    
  </body>
</html>