<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><link href=../_gwas_catalog/ rel=prev><link href=../study_index/ rel=next><link rel=icon href=../../../../assets/imgs/favicon.png><meta name=generator content="mkdocs-1.5.3, mkdocs-material-9.4.10"><title>Associations - Open Targets Genetics</title><link rel=stylesheet href=../../../../assets/stylesheets/main.fad675c6.min.css><link rel=stylesheet href=../../../../assets/stylesheets/palette.356b1318.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../../../css/timeago.css><link rel=stylesheet href=../../../../assets/_mkdocstrings.css><link rel=stylesheet href=../../../../assets/stylesheets/extra.css><script>__md_scope=new URL("../../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=blue-grey data-md-color-accent=light-blue> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#otg.datasource.gwas_catalog.associations.GWASCatalogAssociations class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../../../.. title="Open Targets Genetics" class="md-header__button md-logo" aria-label="Open Targets Genetics" data-md-component=logo> <img src=../../../../assets/imgs/otlogo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Open Targets Genetics </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Associations </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=blue-grey data-md-color-accent=light-blue aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=blue-grey data-md-color-accent=light-blue aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/opentargets/genetics_etl_python title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> opentargets/genetics_etl_python </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../../../../installation/ class=md-tabs__link> Installation </a> </li> <li class=md-tabs__item> <a href=../../../../usage/ class=md-tabs__link> How-to </a> </li> <li class=md-tabs__item> <a href=../../../../development/_development/ class=md-tabs__link> Development </a> </li> <li class="md-tabs__item md-tabs__item--active"> <a href=../../../_python_api/ class=md-tabs__link> Python API </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../../../.. title="Open Targets Genetics" class="md-nav__button md-logo" aria-label="Open Targets Genetics" data-md-component=logo> <img src=../../../../assets/imgs/otlogo.png alt=logo> </a> Open Targets Genetics </label> <div class=md-nav__source> <a href=https://github.com/opentargets/genetics_etl_python title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> opentargets/genetics_etl_python </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../../installation/ class=md-nav__link> <span class=md-ellipsis> Installation </span> </a> </li> <li class=md-nav__item> <a href=../../../../usage/ class=md-nav__link> <span class=md-ellipsis> How-to </span> </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <div class="md-nav__link md-nav__container"> <a href=../../../../development/_development/ class="md-nav__link "> <span class=md-ellipsis> Development </span> </a> <label class="md-nav__link " for=__nav_3 id=__nav_3_label tabindex> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Development </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../../development/airflow/ class=md-nav__link> <span class=md-ellipsis> Airflow configuration </span> </a> </li> <li class=md-nav__item> <a href=../../../../development/contributing/ class=md-nav__link> <span class=md-ellipsis> Contributing guidelines </span> </a> </li> <li class=md-nav__item> <a href=../../../../development/troubleshooting/ class=md-nav__link> <span class=md-ellipsis> Troubleshooting </span> </a> </li> <li class=md-nav__item> <a href=../../../../development/workflows/ class=md-nav__link> <span class=md-ellipsis> Pipeline workflows </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4 checked> <div class="md-nav__link md-nav__container"> <a href=../../../_python_api/ class="md-nav__link "> <span class=md-ellipsis> Python API </span> </a> <label class="md-nav__link " for=__nav_4 id=__nav_4_label tabindex> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=true> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Python API </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_1> <div class="md-nav__link md-nav__container"> <a href=../../../dataset/_dataset/ class="md-nav__link "> <span class=md-ellipsis> Dataset </span> </a> <label class="md-nav__link " for=__nav_4_1 id=__nav_4_1_label tabindex=0> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_1_label aria-expanded=false> <label class=md-nav__title for=__nav_4_1> <span class="md-nav__icon md-icon"></span> Dataset </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../dataset/colocalisation/ class=md-nav__link> <span class=md-ellipsis> Colocalisation </span> </a> </li> <li class=md-nav__item> <a href=../../../dataset/gene_index/ class=md-nav__link> <span class=md-ellipsis> Gene Index </span> </a> </li> <li class=md-nav__item> <a href=../../../dataset/intervals/ class=md-nav__link> <span class=md-ellipsis> Intervals </span> </a> </li> <li class=md-nav__item> <a href=../../../dataset/l2g_feature/ class=md-nav__link> <span class=md-ellipsis> L2G Feature </span> </a> </li> <li class=md-nav__item> <a href=../../../dataset/l2g_feature_matrix/ class=md-nav__link> <span class=md-ellipsis> L2G Feature Matrix </span> </a> </li> <li class=md-nav__item> <a href=../../../dataset/l2g_gold_standard/ class=md-nav__link> <span class=md-ellipsis> L2G Gold Standard </span> </a> </li> <li class=md-nav__item> <a href=../../../dataset/l2g_prediction/ class=md-nav__link> <span class=md-ellipsis> L2G Prediction </span> </a> </li> <li class=md-nav__item> <a href=../../../dataset/ld_index/ class=md-nav__link> <span class=md-ellipsis> LD Index </span> </a> </li> <li class=md-nav__item> <a href=../../../dataset/study_index/ class=md-nav__link> <span class=md-ellipsis> Study Index </span> </a> </li> <li class=md-nav__item> <a href=../../../dataset/study_locus/ class=md-nav__link> <span class=md-ellipsis> Study Locus </span> </a> </li> <li class=md-nav__item> <a href=../../../dataset/study_locus_overlap/ class=md-nav__link> <span class=md-ellipsis> Study Locus Overlap </span> </a> </li> <li class=md-nav__item> <a href=../../../dataset/summary_statistics/ class=md-nav__link> <span class=md-ellipsis> Summary Statistics </span> </a> </li> <li class=md-nav__item> <a href=../../../dataset/variant_annotation/ class=md-nav__link> <span class=md-ellipsis> Variant annotation </span> </a> </li> <li class=md-nav__item> <a href=../../../dataset/variant_index/ class=md-nav__link> <span class=md-ellipsis> Variant index </span> </a> </li> <li class=md-nav__item> <a href=../../../dataset/variant_to_gene/ class=md-nav__link> <span class=md-ellipsis> Variant-to-gene </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2 checked> <div class="md-nav__link md-nav__container"> <a href=../../_datasource/ class="md-nav__link "> <span class=md-ellipsis> Datasource </span> </a> <label class="md-nav__link " for=__nav_4_2 id=__nav_4_2_label tabindex=0> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_2_label aria-expanded=true> <label class=md-nav__title for=__nav_4_2> <span class="md-nav__icon md-icon"></span> Datasource </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2_1> <div class="md-nav__link md-nav__container"> <a href=../../eqtl_catalogue/study_index/ class="md-nav__link "> <span class=md-ellipsis> Eqtl catalogue </span> </a> <label class="md-nav__link " for=__nav_4_2_1 id=__nav_4_2_1_label tabindex=0> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_2_1_label aria-expanded=false> <label class=md-nav__title for=__nav_4_2_1> <span class="md-nav__icon md-icon"></span> Eqtl catalogue </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../eqtl_catalogue/summary_stats/ class=md-nav__link> <span class=md-ellipsis> Summary Stats </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2_2> <div class="md-nav__link md-nav__container"> <a href=../../finngen/_finngen/ class="md-nav__link "> <span class=md-ellipsis> Finngen </span> </a> <label class="md-nav__link " for=__nav_4_2_2 id=__nav_4_2_2_label tabindex=0> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_2_2_label aria-expanded=false> <label class=md-nav__title for=__nav_4_2_2> <span class="md-nav__icon md-icon"></span> Finngen </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../finngen/study_index/ class=md-nav__link> <span class=md-ellipsis> Study Index </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2_3> <div class="md-nav__link md-nav__container"> <a href=../../gnomad/_gnomad/ class="md-nav__link "> <span class=md-ellipsis> Gnomad </span> </a> <label class="md-nav__link " for=__nav_4_2_3 id=__nav_4_2_3_label tabindex=0> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_2_3_label aria-expanded=false> <label class=md-nav__title for=__nav_4_2_3> <span class="md-nav__icon md-icon"></span> Gnomad </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../gnomad/gnomad_ld/ class=md-nav__link> <span class=md-ellipsis> LD Matrix </span> </a> </li> <li class=md-nav__item> <a href=../../gnomad/gnomad_variants/ class=md-nav__link> <span class=md-ellipsis> Variants </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2_4 checked> <div class="md-nav__link md-nav__container"> <a href=../_gwas_catalog/ class="md-nav__link "> <span class=md-ellipsis> Gwas catalog </span> </a> <label class="md-nav__link " for=__nav_4_2_4 id=__nav_4_2_4_label tabindex=0> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_2_4_label aria-expanded=true> <label class=md-nav__title for=__nav_4_2_4> <span class="md-nav__icon md-icon"></span> Gwas catalog </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Associations </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Associations </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#otg.datasource.gwas_catalog.associations.GWASCatalogAssociations class=md-nav__link> <span class=md-ellipsis> GWASCatalogAssociations </span> </a> <nav class=md-nav aria-label=GWASCatalogAssociations> <ul class=md-nav__list> <li class=md-nav__item> <a href=#otg.datasource.gwas_catalog.associations.GWASCatalogAssociations.from_source class=md-nav__link> <span class=md-ellipsis> from_source() </span> </a> </li> <li class=md-nav__item> <a href=#otg.datasource.gwas_catalog.associations.GWASCatalogAssociations.update_study_id class=md-nav__link> <span class=md-ellipsis> update_study_id() </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../study_index/ class=md-nav__link> <span class=md-ellipsis> Study Index </span> </a> </li> <li class=md-nav__item> <a href=../study_splitter/ class=md-nav__link> <span class=md-ellipsis> Study Splitter </span> </a> </li> <li class=md-nav__item> <a href=../summary_statistics/ class=md-nav__link> <span class=md-ellipsis> Summary statistics </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2_5> <div class="md-nav__link md-nav__container"> <a href=../../intervals/_intervals/ class="md-nav__link "> <span class=md-ellipsis> Intervals </span> </a> <label class="md-nav__link " for=__nav_4_2_5 id=__nav_4_2_5_label tabindex=0> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_2_5_label aria-expanded=false> <label class=md-nav__title for=__nav_4_2_5> <span class="md-nav__icon md-icon"></span> Intervals </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../intervals/andersson/ class=md-nav__link> <span class=md-ellipsis> Andersson et al. </span> </a> </li> <li class=md-nav__item> <a href=../../intervals/javierre/ class=md-nav__link> <span class=md-ellipsis> Javierre et al. </span> </a> </li> <li class=md-nav__item> <a href=../../intervals/jung/ class=md-nav__link> <span class=md-ellipsis> Jung et al. </span> </a> </li> <li class=md-nav__item> <a href=../../intervals/thurman/ class=md-nav__link> <span class=md-ellipsis> Thurman et al. </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2_6> <div class="md-nav__link md-nav__container"> <a href=../../open_targets/_open_targets/ class="md-nav__link "> <span class=md-ellipsis> Open targets </span> </a> <label class="md-nav__link " for=__nav_4_2_6 id=__nav_4_2_6_label tabindex=0> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_2_6_label aria-expanded=false> <label class=md-nav__title for=__nav_4_2_6> <span class="md-nav__icon md-icon"></span> Open targets </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../open_targets/l2g_gold_standard/ class=md-nav__link> <span class=md-ellipsis> L2G Gold Standard </span> </a> </li> <li class=md-nav__item> <a href=../../open_targets/target/ class=md-nav__link> <span class=md-ellipsis> Target </span> </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2_7> <div class="md-nav__link md-nav__container"> <a href=../../ukbiobank/_ukbiobank/ class="md-nav__link "> <span class=md-ellipsis> Ukbiobank </span> </a> <label class="md-nav__link " for=__nav_4_2_7 id=__nav_4_2_7_label tabindex=0> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_2_7_label aria-expanded=false> <label class=md-nav__title for=__nav_4_2_7> <span class="md-nav__icon md-icon"></span> Ukbiobank </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../ukbiobank/study_index/ class=md-nav__link> <span class=md-ellipsis> Study Index </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_3> <div class="md-nav__link md-nav__container"> <a href=../../../method/_method/ class="md-nav__link "> <span class=md-ellipsis> Method </span> </a> <label class="md-nav__link " for=__nav_4_3 id=__nav_4_3_label tabindex=0> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_3_label aria-expanded=false> <label class=md-nav__title for=__nav_4_3> <span class="md-nav__icon md-icon"></span> Method </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../method/clumping/ class=md-nav__link> <span class=md-ellipsis> Clumping </span> </a> </li> <li class=md-nav__item> <a href=../../../method/coloc/ class=md-nav__link> <span class=md-ellipsis> Coloc </span> </a> </li> <li class=md-nav__item> <a href=../../../method/ecaviar/ class=md-nav__link> <span class=md-ellipsis> eCAVIAR </span> </a> </li> <li class=md-nav__item> <a href=../../../method/ld_annotator/ class=md-nav__link> <span class=md-ellipsis> LDAnnotator </span> </a> </li> <li class=md-nav__item> <a href=../../../method/pics/ class=md-nav__link> <span class=md-ellipsis> PICS </span> </a> </li> <li class=md-nav__item> <a href=../../../method/window_based_clumping/ class=md-nav__link> <span class=md-ellipsis> Window-based clumping </span> </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_3_7> <div class="md-nav__link md-nav__container"> <a href=../../../method/l2g/_l2g/ class="md-nav__link "> <span class=md-ellipsis> L2g </span> </a> <label class="md-nav__link " for=__nav_4_3_7 id=__nav_4_3_7_label tabindex=0> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_3_7_label aria-expanded=false> <label class=md-nav__title for=__nav_4_3_7> <span class="md-nav__icon md-icon"></span> L2g </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../method/l2g/evaluator/ class=md-nav__link> <span class=md-ellipsis> W&B evaluator </span> </a> </li> <li class=md-nav__item> <a href=../../../method/l2g/feature_factory/ class=md-nav__link> <span class=md-ellipsis> L2G Feature Factory </span> </a> </li> <li class=md-nav__item> <a href=../../../method/l2g/model/ class=md-nav__link> <span class=md-ellipsis> L2G Model </span> </a> </li> <li class=md-nav__item> <a href=../../../method/l2g/trainer/ class=md-nav__link> <span class=md-ellipsis> L2G Trainer </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_4> <div class="md-nav__link md-nav__container"> <a href=../../../step/_step/ class="md-nav__link "> <span class=md-ellipsis> Step </span> </a> <label class="md-nav__link " for=__nav_4_4 id=__nav_4_4_label tabindex=0> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4_4> <span class="md-nav__icon md-icon"></span> Step </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../step/colocalisation/ class=md-nav__link> <span class=md-ellipsis> Colocalisation </span> </a> </li> <li class=md-nav__item> <a href=../../../step/eqtl_catalogue/ class=md-nav__link> <span class=md-ellipsis> eQTL Catalogue </span> </a> </li> <li class=md-nav__item> <a href=../../../step/finngen/ class=md-nav__link> <span class=md-ellipsis> FinnGen </span> </a> </li> <li class=md-nav__item> <a href=../../../step/gene_index/ class=md-nav__link> <span class=md-ellipsis> Gene Index </span> </a> </li> <li class=md-nav__item> <a href=../../../step/gwas_catalog/ class=md-nav__link> <span class=md-ellipsis> GWAS Catalog </span> </a> </li> <li class=md-nav__item> <a href=../../../step/gwas_catalog_sumstat_preprocess/ class=md-nav__link> <span class=md-ellipsis> GWAS Catalog sumstat preprocess </span> </a> </li> <li class=md-nav__item> <a href=../../../step/l2g/ class=md-nav__link> <span class=md-ellipsis> Locus-to-gene (L2G) </span> </a> </li> <li class=md-nav__item> <a href=../../../step/ld_index/ class=md-nav__link> <span class=md-ellipsis> LD Index </span> </a> </li> <li class=md-nav__item> <a href=../../../step/pics/ class=md-nav__link> <span class=md-ellipsis> PICS </span> </a> </li> <li class=md-nav__item> <a href=../../../step/ukbiobank/ class=md-nav__link> <span class=md-ellipsis> UK Biobank </span> </a> </li> <li class=md-nav__item> <a href=../../../step/variant_annotation_step/ class=md-nav__link> <span class=md-ellipsis> Variant Annotation </span> </a> </li> <li class=md-nav__item> <a href=../../../step/variant_index_step/ class=md-nav__link> <span class=md-ellipsis> Variant Index </span> </a> </li> <li class=md-nav__item> <a href=../../../step/variant_to_gene_step/ class=md-nav__link> <span class=md-ellipsis> Variant-to-gene </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#otg.datasource.gwas_catalog.associations.GWASCatalogAssociations class=md-nav__link> <span class=md-ellipsis> GWASCatalogAssociations </span> </a> <nav class=md-nav aria-label=GWASCatalogAssociations> <ul class=md-nav__list> <li class=md-nav__item> <a href=#otg.datasource.gwas_catalog.associations.GWASCatalogAssociations.from_source class=md-nav__link> <span class=md-ellipsis> from_source() </span> </a> </li> <li class=md-nav__item> <a href=#otg.datasource.gwas_catalog.associations.GWASCatalogAssociations.update_study_id class=md-nav__link> <span class=md-ellipsis> update_study_id() </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1>Associations</h1> <div class="doc doc-object doc-class"> <h2 id=otg.datasource.gwas_catalog.associations.GWASCatalogAssociations class="doc doc-heading"> <code>otg.datasource.gwas_catalog.associations.GWASCatalogAssociations</code> <span class="doc doc-labels"> <small class="doc doc-label doc-label-dataclass"><code>dataclass</code></small> </span> <a href=#otg.datasource.gwas_catalog.associations.GWASCatalogAssociations class=headerlink title="Permanent link">&para;</a></h2> <div class="doc doc-contents first"> <p class="doc doc-class-bases"> Bases: <code><a class="autorefs autorefs-internal" title="otg.dataset.study_locus.StudyLocus" href="../../../dataset/study_locus/#otg.dataset.study_locus.StudyLocus">StudyLocus</a></code></p> <p>Study-locus dataset derived from GWAS Catalog.</p> <details class=quote> <summary>Source code in <code>src/otg/datasource/gwas_catalog/associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>  30</span>
<span class=normal>  31</span>
<span class=normal>  32</span>
<span class=normal>  33</span>
<span class=normal>  34</span>
<span class=normal>  35</span>
<span class=normal>  36</span>
<span class=normal>  37</span>
<span class=normal>  38</span>
<span class=normal>  39</span>
<span class=normal>  40</span>
<span class=normal>  41</span>
<span class=normal>  42</span>
<span class=normal>  43</span>
<span class=normal>  44</span>
<span class=normal>  45</span>
<span class=normal>  46</span>
<span class=normal>  47</span>
<span class=normal>  48</span>
<span class=normal>  49</span>
<span class=normal>  50</span>
<span class=normal>  51</span>
<span class=normal>  52</span>
<span class=normal>  53</span>
<span class=normal>  54</span>
<span class=normal>  55</span>
<span class=normal>  56</span>
<span class=normal>  57</span>
<span class=normal>  58</span>
<span class=normal>  59</span>
<span class=normal>  60</span>
<span class=normal>  61</span>
<span class=normal>  62</span>
<span class=normal>  63</span>
<span class=normal>  64</span>
<span class=normal>  65</span>
<span class=normal>  66</span>
<span class=normal>  67</span>
<span class=normal>  68</span>
<span class=normal>  69</span>
<span class=normal>  70</span>
<span class=normal>  71</span>
<span class=normal>  72</span>
<span class=normal>  73</span>
<span class=normal>  74</span>
<span class=normal>  75</span>
<span class=normal>  76</span>
<span class=normal>  77</span>
<span class=normal>  78</span>
<span class=normal>  79</span>
<span class=normal>  80</span>
<span class=normal>  81</span>
<span class=normal>  82</span>
<span class=normal>  83</span>
<span class=normal>  84</span>
<span class=normal>  85</span>
<span class=normal>  86</span>
<span class=normal>  87</span>
<span class=normal>  88</span>
<span class=normal>  89</span>
<span class=normal>  90</span>
<span class=normal>  91</span>
<span class=normal>  92</span>
<span class=normal>  93</span>
<span class=normal>  94</span>
<span class=normal>  95</span>
<span class=normal>  96</span>
<span class=normal>  97</span>
<span class=normal>  98</span>
<span class=normal>  99</span>
<span class=normal> 100</span>
<span class=normal> 101</span>
<span class=normal> 102</span>
<span class=normal> 103</span>
<span class=normal> 104</span>
<span class=normal> 105</span>
<span class=normal> 106</span>
<span class=normal> 107</span>
<span class=normal> 108</span>
<span class=normal> 109</span>
<span class=normal> 110</span>
<span class=normal> 111</span>
<span class=normal> 112</span>
<span class=normal> 113</span>
<span class=normal> 114</span>
<span class=normal> 115</span>
<span class=normal> 116</span>
<span class=normal> 117</span>
<span class=normal> 118</span>
<span class=normal> 119</span>
<span class=normal> 120</span>
<span class=normal> 121</span>
<span class=normal> 122</span>
<span class=normal> 123</span>
<span class=normal> 124</span>
<span class=normal> 125</span>
<span class=normal> 126</span>
<span class=normal> 127</span>
<span class=normal> 128</span>
<span class=normal> 129</span>
<span class=normal> 130</span>
<span class=normal> 131</span>
<span class=normal> 132</span>
<span class=normal> 133</span>
<span class=normal> 134</span>
<span class=normal> 135</span>
<span class=normal> 136</span>
<span class=normal> 137</span>
<span class=normal> 138</span>
<span class=normal> 139</span>
<span class=normal> 140</span>
<span class=normal> 141</span>
<span class=normal> 142</span>
<span class=normal> 143</span>
<span class=normal> 144</span>
<span class=normal> 145</span>
<span class=normal> 146</span>
<span class=normal> 147</span>
<span class=normal> 148</span>
<span class=normal> 149</span>
<span class=normal> 150</span>
<span class=normal> 151</span>
<span class=normal> 152</span>
<span class=normal> 153</span>
<span class=normal> 154</span>
<span class=normal> 155</span>
<span class=normal> 156</span>
<span class=normal> 157</span>
<span class=normal> 158</span>
<span class=normal> 159</span>
<span class=normal> 160</span>
<span class=normal> 161</span>
<span class=normal> 162</span>
<span class=normal> 163</span>
<span class=normal> 164</span>
<span class=normal> 165</span>
<span class=normal> 166</span>
<span class=normal> 167</span>
<span class=normal> 168</span>
<span class=normal> 169</span>
<span class=normal> 170</span>
<span class=normal> 171</span>
<span class=normal> 172</span>
<span class=normal> 173</span>
<span class=normal> 174</span>
<span class=normal> 175</span>
<span class=normal> 176</span>
<span class=normal> 177</span>
<span class=normal> 178</span>
<span class=normal> 179</span>
<span class=normal> 180</span>
<span class=normal> 181</span>
<span class=normal> 182</span>
<span class=normal> 183</span>
<span class=normal> 184</span>
<span class=normal> 185</span>
<span class=normal> 186</span>
<span class=normal> 187</span>
<span class=normal> 188</span>
<span class=normal> 189</span>
<span class=normal> 190</span>
<span class=normal> 191</span>
<span class=normal> 192</span>
<span class=normal> 193</span>
<span class=normal> 194</span>
<span class=normal> 195</span>
<span class=normal> 196</span>
<span class=normal> 197</span>
<span class=normal> 198</span>
<span class=normal> 199</span>
<span class=normal> 200</span>
<span class=normal> 201</span>
<span class=normal> 202</span>
<span class=normal> 203</span>
<span class=normal> 204</span>
<span class=normal> 205</span>
<span class=normal> 206</span>
<span class=normal> 207</span>
<span class=normal> 208</span>
<span class=normal> 209</span>
<span class=normal> 210</span>
<span class=normal> 211</span>
<span class=normal> 212</span>
<span class=normal> 213</span>
<span class=normal> 214</span>
<span class=normal> 215</span>
<span class=normal> 216</span>
<span class=normal> 217</span>
<span class=normal> 218</span>
<span class=normal> 219</span>
<span class=normal> 220</span>
<span class=normal> 221</span>
<span class=normal> 222</span>
<span class=normal> 223</span>
<span class=normal> 224</span>
<span class=normal> 225</span>
<span class=normal> 226</span>
<span class=normal> 227</span>
<span class=normal> 228</span>
<span class=normal> 229</span>
<span class=normal> 230</span>
<span class=normal> 231</span>
<span class=normal> 232</span>
<span class=normal> 233</span>
<span class=normal> 234</span>
<span class=normal> 235</span>
<span class=normal> 236</span>
<span class=normal> 237</span>
<span class=normal> 238</span>
<span class=normal> 239</span>
<span class=normal> 240</span>
<span class=normal> 241</span>
<span class=normal> 242</span>
<span class=normal> 243</span>
<span class=normal> 244</span>
<span class=normal> 245</span>
<span class=normal> 246</span>
<span class=normal> 247</span>
<span class=normal> 248</span>
<span class=normal> 249</span>
<span class=normal> 250</span>
<span class=normal> 251</span>
<span class=normal> 252</span>
<span class=normal> 253</span>
<span class=normal> 254</span>
<span class=normal> 255</span>
<span class=normal> 256</span>
<span class=normal> 257</span>
<span class=normal> 258</span>
<span class=normal> 259</span>
<span class=normal> 260</span>
<span class=normal> 261</span>
<span class=normal> 262</span>
<span class=normal> 263</span>
<span class=normal> 264</span>
<span class=normal> 265</span>
<span class=normal> 266</span>
<span class=normal> 267</span>
<span class=normal> 268</span>
<span class=normal> 269</span>
<span class=normal> 270</span>
<span class=normal> 271</span>
<span class=normal> 272</span>
<span class=normal> 273</span>
<span class=normal> 274</span>
<span class=normal> 275</span>
<span class=normal> 276</span>
<span class=normal> 277</span>
<span class=normal> 278</span>
<span class=normal> 279</span>
<span class=normal> 280</span>
<span class=normal> 281</span>
<span class=normal> 282</span>
<span class=normal> 283</span>
<span class=normal> 284</span>
<span class=normal> 285</span>
<span class=normal> 286</span>
<span class=normal> 287</span>
<span class=normal> 288</span>
<span class=normal> 289</span>
<span class=normal> 290</span>
<span class=normal> 291</span>
<span class=normal> 292</span>
<span class=normal> 293</span>
<span class=normal> 294</span>
<span class=normal> 295</span>
<span class=normal> 296</span>
<span class=normal> 297</span>
<span class=normal> 298</span>
<span class=normal> 299</span>
<span class=normal> 300</span>
<span class=normal> 301</span>
<span class=normal> 302</span>
<span class=normal> 303</span>
<span class=normal> 304</span>
<span class=normal> 305</span>
<span class=normal> 306</span>
<span class=normal> 307</span>
<span class=normal> 308</span>
<span class=normal> 309</span>
<span class=normal> 310</span>
<span class=normal> 311</span>
<span class=normal> 312</span>
<span class=normal> 313</span>
<span class=normal> 314</span>
<span class=normal> 315</span>
<span class=normal> 316</span>
<span class=normal> 317</span>
<span class=normal> 318</span>
<span class=normal> 319</span>
<span class=normal> 320</span>
<span class=normal> 321</span>
<span class=normal> 322</span>
<span class=normal> 323</span>
<span class=normal> 324</span>
<span class=normal> 325</span>
<span class=normal> 326</span>
<span class=normal> 327</span>
<span class=normal> 328</span>
<span class=normal> 329</span>
<span class=normal> 330</span>
<span class=normal> 331</span>
<span class=normal> 332</span>
<span class=normal> 333</span>
<span class=normal> 334</span>
<span class=normal> 335</span>
<span class=normal> 336</span>
<span class=normal> 337</span>
<span class=normal> 338</span>
<span class=normal> 339</span>
<span class=normal> 340</span>
<span class=normal> 341</span>
<span class=normal> 342</span>
<span class=normal> 343</span>
<span class=normal> 344</span>
<span class=normal> 345</span>
<span class=normal> 346</span>
<span class=normal> 347</span>
<span class=normal> 348</span>
<span class=normal> 349</span>
<span class=normal> 350</span>
<span class=normal> 351</span>
<span class=normal> 352</span>
<span class=normal> 353</span>
<span class=normal> 354</span>
<span class=normal> 355</span>
<span class=normal> 356</span>
<span class=normal> 357</span>
<span class=normal> 358</span>
<span class=normal> 359</span>
<span class=normal> 360</span>
<span class=normal> 361</span>
<span class=normal> 362</span>
<span class=normal> 363</span>
<span class=normal> 364</span>
<span class=normal> 365</span>
<span class=normal> 366</span>
<span class=normal> 367</span>
<span class=normal> 368</span>
<span class=normal> 369</span>
<span class=normal> 370</span>
<span class=normal> 371</span>
<span class=normal> 372</span>
<span class=normal> 373</span>
<span class=normal> 374</span>
<span class=normal> 375</span>
<span class=normal> 376</span>
<span class=normal> 377</span>
<span class=normal> 378</span>
<span class=normal> 379</span>
<span class=normal> 380</span>
<span class=normal> 381</span>
<span class=normal> 382</span>
<span class=normal> 383</span>
<span class=normal> 384</span>
<span class=normal> 385</span>
<span class=normal> 386</span>
<span class=normal> 387</span>
<span class=normal> 388</span>
<span class=normal> 389</span>
<span class=normal> 390</span>
<span class=normal> 391</span>
<span class=normal> 392</span>
<span class=normal> 393</span>
<span class=normal> 394</span>
<span class=normal> 395</span>
<span class=normal> 396</span>
<span class=normal> 397</span>
<span class=normal> 398</span>
<span class=normal> 399</span>
<span class=normal> 400</span>
<span class=normal> 401</span>
<span class=normal> 402</span>
<span class=normal> 403</span>
<span class=normal> 404</span>
<span class=normal> 405</span>
<span class=normal> 406</span>
<span class=normal> 407</span>
<span class=normal> 408</span>
<span class=normal> 409</span>
<span class=normal> 410</span>
<span class=normal> 411</span>
<span class=normal> 412</span>
<span class=normal> 413</span>
<span class=normal> 414</span>
<span class=normal> 415</span>
<span class=normal> 416</span>
<span class=normal> 417</span>
<span class=normal> 418</span>
<span class=normal> 419</span>
<span class=normal> 420</span>
<span class=normal> 421</span>
<span class=normal> 422</span>
<span class=normal> 423</span>
<span class=normal> 424</span>
<span class=normal> 425</span>
<span class=normal> 426</span>
<span class=normal> 427</span>
<span class=normal> 428</span>
<span class=normal> 429</span>
<span class=normal> 430</span>
<span class=normal> 431</span>
<span class=normal> 432</span>
<span class=normal> 433</span>
<span class=normal> 434</span>
<span class=normal> 435</span>
<span class=normal> 436</span>
<span class=normal> 437</span>
<span class=normal> 438</span>
<span class=normal> 439</span>
<span class=normal> 440</span>
<span class=normal> 441</span>
<span class=normal> 442</span>
<span class=normal> 443</span>
<span class=normal> 444</span>
<span class=normal> 445</span>
<span class=normal> 446</span>
<span class=normal> 447</span>
<span class=normal> 448</span>
<span class=normal> 449</span>
<span class=normal> 450</span>
<span class=normal> 451</span>
<span class=normal> 452</span>
<span class=normal> 453</span>
<span class=normal> 454</span>
<span class=normal> 455</span>
<span class=normal> 456</span>
<span class=normal> 457</span>
<span class=normal> 458</span>
<span class=normal> 459</span>
<span class=normal> 460</span>
<span class=normal> 461</span>
<span class=normal> 462</span>
<span class=normal> 463</span>
<span class=normal> 464</span>
<span class=normal> 465</span>
<span class=normal> 466</span>
<span class=normal> 467</span>
<span class=normal> 468</span>
<span class=normal> 469</span>
<span class=normal> 470</span>
<span class=normal> 471</span>
<span class=normal> 472</span>
<span class=normal> 473</span>
<span class=normal> 474</span>
<span class=normal> 475</span>
<span class=normal> 476</span>
<span class=normal> 477</span>
<span class=normal> 478</span>
<span class=normal> 479</span>
<span class=normal> 480</span>
<span class=normal> 481</span>
<span class=normal> 482</span>
<span class=normal> 483</span>
<span class=normal> 484</span>
<span class=normal> 485</span>
<span class=normal> 486</span>
<span class=normal> 487</span>
<span class=normal> 488</span>
<span class=normal> 489</span>
<span class=normal> 490</span>
<span class=normal> 491</span>
<span class=normal> 492</span>
<span class=normal> 493</span>
<span class=normal> 494</span>
<span class=normal> 495</span>
<span class=normal> 496</span>
<span class=normal> 497</span>
<span class=normal> 498</span>
<span class=normal> 499</span>
<span class=normal> 500</span>
<span class=normal> 501</span>
<span class=normal> 502</span>
<span class=normal> 503</span>
<span class=normal> 504</span>
<span class=normal> 505</span>
<span class=normal> 506</span>
<span class=normal> 507</span>
<span class=normal> 508</span>
<span class=normal> 509</span>
<span class=normal> 510</span>
<span class=normal> 511</span>
<span class=normal> 512</span>
<span class=normal> 513</span>
<span class=normal> 514</span>
<span class=normal> 515</span>
<span class=normal> 516</span>
<span class=normal> 517</span>
<span class=normal> 518</span>
<span class=normal> 519</span>
<span class=normal> 520</span>
<span class=normal> 521</span>
<span class=normal> 522</span>
<span class=normal> 523</span>
<span class=normal> 524</span>
<span class=normal> 525</span>
<span class=normal> 526</span>
<span class=normal> 527</span>
<span class=normal> 528</span>
<span class=normal> 529</span>
<span class=normal> 530</span>
<span class=normal> 531</span>
<span class=normal> 532</span>
<span class=normal> 533</span>
<span class=normal> 534</span>
<span class=normal> 535</span>
<span class=normal> 536</span>
<span class=normal> 537</span>
<span class=normal> 538</span>
<span class=normal> 539</span>
<span class=normal> 540</span>
<span class=normal> 541</span>
<span class=normal> 542</span>
<span class=normal> 543</span>
<span class=normal> 544</span>
<span class=normal> 545</span>
<span class=normal> 546</span>
<span class=normal> 547</span>
<span class=normal> 548</span>
<span class=normal> 549</span>
<span class=normal> 550</span>
<span class=normal> 551</span>
<span class=normal> 552</span>
<span class=normal> 553</span>
<span class=normal> 554</span>
<span class=normal> 555</span>
<span class=normal> 556</span>
<span class=normal> 557</span>
<span class=normal> 558</span>
<span class=normal> 559</span>
<span class=normal> 560</span>
<span class=normal> 561</span>
<span class=normal> 562</span>
<span class=normal> 563</span>
<span class=normal> 564</span>
<span class=normal> 565</span>
<span class=normal> 566</span>
<span class=normal> 567</span>
<span class=normal> 568</span>
<span class=normal> 569</span>
<span class=normal> 570</span>
<span class=normal> 571</span>
<span class=normal> 572</span>
<span class=normal> 573</span>
<span class=normal> 574</span>
<span class=normal> 575</span>
<span class=normal> 576</span>
<span class=normal> 577</span>
<span class=normal> 578</span>
<span class=normal> 579</span>
<span class=normal> 580</span>
<span class=normal> 581</span>
<span class=normal> 582</span>
<span class=normal> 583</span>
<span class=normal> 584</span>
<span class=normal> 585</span>
<span class=normal> 586</span>
<span class=normal> 587</span>
<span class=normal> 588</span>
<span class=normal> 589</span>
<span class=normal> 590</span>
<span class=normal> 591</span>
<span class=normal> 592</span>
<span class=normal> 593</span>
<span class=normal> 594</span>
<span class=normal> 595</span>
<span class=normal> 596</span>
<span class=normal> 597</span>
<span class=normal> 598</span>
<span class=normal> 599</span>
<span class=normal> 600</span>
<span class=normal> 601</span>
<span class=normal> 602</span>
<span class=normal> 603</span>
<span class=normal> 604</span>
<span class=normal> 605</span>
<span class=normal> 606</span>
<span class=normal> 607</span>
<span class=normal> 608</span>
<span class=normal> 609</span>
<span class=normal> 610</span>
<span class=normal> 611</span>
<span class=normal> 612</span>
<span class=normal> 613</span>
<span class=normal> 614</span>
<span class=normal> 615</span>
<span class=normal> 616</span>
<span class=normal> 617</span>
<span class=normal> 618</span>
<span class=normal> 619</span>
<span class=normal> 620</span>
<span class=normal> 621</span>
<span class=normal> 622</span>
<span class=normal> 623</span>
<span class=normal> 624</span>
<span class=normal> 625</span>
<span class=normal> 626</span>
<span class=normal> 627</span>
<span class=normal> 628</span>
<span class=normal> 629</span>
<span class=normal> 630</span>
<span class=normal> 631</span>
<span class=normal> 632</span>
<span class=normal> 633</span>
<span class=normal> 634</span>
<span class=normal> 635</span>
<span class=normal> 636</span>
<span class=normal> 637</span>
<span class=normal> 638</span>
<span class=normal> 639</span>
<span class=normal> 640</span>
<span class=normal> 641</span>
<span class=normal> 642</span>
<span class=normal> 643</span>
<span class=normal> 644</span>
<span class=normal> 645</span>
<span class=normal> 646</span>
<span class=normal> 647</span>
<span class=normal> 648</span>
<span class=normal> 649</span>
<span class=normal> 650</span>
<span class=normal> 651</span>
<span class=normal> 652</span>
<span class=normal> 653</span>
<span class=normal> 654</span>
<span class=normal> 655</span>
<span class=normal> 656</span>
<span class=normal> 657</span>
<span class=normal> 658</span>
<span class=normal> 659</span>
<span class=normal> 660</span>
<span class=normal> 661</span>
<span class=normal> 662</span>
<span class=normal> 663</span>
<span class=normal> 664</span>
<span class=normal> 665</span>
<span class=normal> 666</span>
<span class=normal> 667</span>
<span class=normal> 668</span>
<span class=normal> 669</span>
<span class=normal> 670</span>
<span class=normal> 671</span>
<span class=normal> 672</span>
<span class=normal> 673</span>
<span class=normal> 674</span>
<span class=normal> 675</span>
<span class=normal> 676</span>
<span class=normal> 677</span>
<span class=normal> 678</span>
<span class=normal> 679</span>
<span class=normal> 680</span>
<span class=normal> 681</span>
<span class=normal> 682</span>
<span class=normal> 683</span>
<span class=normal> 684</span>
<span class=normal> 685</span>
<span class=normal> 686</span>
<span class=normal> 687</span>
<span class=normal> 688</span>
<span class=normal> 689</span>
<span class=normal> 690</span>
<span class=normal> 691</span>
<span class=normal> 692</span>
<span class=normal> 693</span>
<span class=normal> 694</span>
<span class=normal> 695</span>
<span class=normal> 696</span>
<span class=normal> 697</span>
<span class=normal> 698</span>
<span class=normal> 699</span>
<span class=normal> 700</span>
<span class=normal> 701</span>
<span class=normal> 702</span>
<span class=normal> 703</span>
<span class=normal> 704</span>
<span class=normal> 705</span>
<span class=normal> 706</span>
<span class=normal> 707</span>
<span class=normal> 708</span>
<span class=normal> 709</span>
<span class=normal> 710</span>
<span class=normal> 711</span>
<span class=normal> 712</span>
<span class=normal> 713</span>
<span class=normal> 714</span>
<span class=normal> 715</span>
<span class=normal> 716</span>
<span class=normal> 717</span>
<span class=normal> 718</span>
<span class=normal> 719</span>
<span class=normal> 720</span>
<span class=normal> 721</span>
<span class=normal> 722</span>
<span class=normal> 723</span>
<span class=normal> 724</span>
<span class=normal> 725</span>
<span class=normal> 726</span>
<span class=normal> 727</span>
<span class=normal> 728</span>
<span class=normal> 729</span>
<span class=normal> 730</span>
<span class=normal> 731</span>
<span class=normal> 732</span>
<span class=normal> 733</span>
<span class=normal> 734</span>
<span class=normal> 735</span>
<span class=normal> 736</span>
<span class=normal> 737</span>
<span class=normal> 738</span>
<span class=normal> 739</span>
<span class=normal> 740</span>
<span class=normal> 741</span>
<span class=normal> 742</span>
<span class=normal> 743</span>
<span class=normal> 744</span>
<span class=normal> 745</span>
<span class=normal> 746</span>
<span class=normal> 747</span>
<span class=normal> 748</span>
<span class=normal> 749</span>
<span class=normal> 750</span>
<span class=normal> 751</span>
<span class=normal> 752</span>
<span class=normal> 753</span>
<span class=normal> 754</span>
<span class=normal> 755</span>
<span class=normal> 756</span>
<span class=normal> 757</span>
<span class=normal> 758</span>
<span class=normal> 759</span>
<span class=normal> 760</span>
<span class=normal> 761</span>
<span class=normal> 762</span>
<span class=normal> 763</span>
<span class=normal> 764</span>
<span class=normal> 765</span>
<span class=normal> 766</span>
<span class=normal> 767</span>
<span class=normal> 768</span>
<span class=normal> 769</span>
<span class=normal> 770</span>
<span class=normal> 771</span>
<span class=normal> 772</span>
<span class=normal> 773</span>
<span class=normal> 774</span>
<span class=normal> 775</span>
<span class=normal> 776</span>
<span class=normal> 777</span>
<span class=normal> 778</span>
<span class=normal> 779</span>
<span class=normal> 780</span>
<span class=normal> 781</span>
<span class=normal> 782</span>
<span class=normal> 783</span>
<span class=normal> 784</span>
<span class=normal> 785</span>
<span class=normal> 786</span>
<span class=normal> 787</span>
<span class=normal> 788</span>
<span class=normal> 789</span>
<span class=normal> 790</span>
<span class=normal> 791</span>
<span class=normal> 792</span>
<span class=normal> 793</span>
<span class=normal> 794</span>
<span class=normal> 795</span>
<span class=normal> 796</span>
<span class=normal> 797</span>
<span class=normal> 798</span>
<span class=normal> 799</span>
<span class=normal> 800</span>
<span class=normal> 801</span>
<span class=normal> 802</span>
<span class=normal> 803</span>
<span class=normal> 804</span>
<span class=normal> 805</span>
<span class=normal> 806</span>
<span class=normal> 807</span>
<span class=normal> 808</span>
<span class=normal> 809</span>
<span class=normal> 810</span>
<span class=normal> 811</span>
<span class=normal> 812</span>
<span class=normal> 813</span>
<span class=normal> 814</span>
<span class=normal> 815</span>
<span class=normal> 816</span>
<span class=normal> 817</span>
<span class=normal> 818</span>
<span class=normal> 819</span>
<span class=normal> 820</span>
<span class=normal> 821</span>
<span class=normal> 822</span>
<span class=normal> 823</span>
<span class=normal> 824</span>
<span class=normal> 825</span>
<span class=normal> 826</span>
<span class=normal> 827</span>
<span class=normal> 828</span>
<span class=normal> 829</span>
<span class=normal> 830</span>
<span class=normal> 831</span>
<span class=normal> 832</span>
<span class=normal> 833</span>
<span class=normal> 834</span>
<span class=normal> 835</span>
<span class=normal> 836</span>
<span class=normal> 837</span>
<span class=normal> 838</span>
<span class=normal> 839</span>
<span class=normal> 840</span>
<span class=normal> 841</span>
<span class=normal> 842</span>
<span class=normal> 843</span>
<span class=normal> 844</span>
<span class=normal> 845</span>
<span class=normal> 846</span>
<span class=normal> 847</span>
<span class=normal> 848</span>
<span class=normal> 849</span>
<span class=normal> 850</span>
<span class=normal> 851</span>
<span class=normal> 852</span>
<span class=normal> 853</span>
<span class=normal> 854</span>
<span class=normal> 855</span>
<span class=normal> 856</span>
<span class=normal> 857</span>
<span class=normal> 858</span>
<span class=normal> 859</span>
<span class=normal> 860</span>
<span class=normal> 861</span>
<span class=normal> 862</span>
<span class=normal> 863</span>
<span class=normal> 864</span>
<span class=normal> 865</span>
<span class=normal> 866</span>
<span class=normal> 867</span>
<span class=normal> 868</span>
<span class=normal> 869</span>
<span class=normal> 870</span>
<span class=normal> 871</span>
<span class=normal> 872</span>
<span class=normal> 873</span>
<span class=normal> 874</span>
<span class=normal> 875</span>
<span class=normal> 876</span>
<span class=normal> 877</span>
<span class=normal> 878</span>
<span class=normal> 879</span>
<span class=normal> 880</span>
<span class=normal> 881</span>
<span class=normal> 882</span>
<span class=normal> 883</span>
<span class=normal> 884</span>
<span class=normal> 885</span>
<span class=normal> 886</span>
<span class=normal> 887</span>
<span class=normal> 888</span>
<span class=normal> 889</span>
<span class=normal> 890</span>
<span class=normal> 891</span>
<span class=normal> 892</span>
<span class=normal> 893</span>
<span class=normal> 894</span>
<span class=normal> 895</span>
<span class=normal> 896</span>
<span class=normal> 897</span>
<span class=normal> 898</span>
<span class=normal> 899</span>
<span class=normal> 900</span>
<span class=normal> 901</span>
<span class=normal> 902</span>
<span class=normal> 903</span>
<span class=normal> 904</span>
<span class=normal> 905</span>
<span class=normal> 906</span>
<span class=normal> 907</span>
<span class=normal> 908</span>
<span class=normal> 909</span>
<span class=normal> 910</span>
<span class=normal> 911</span>
<span class=normal> 912</span>
<span class=normal> 913</span>
<span class=normal> 914</span>
<span class=normal> 915</span>
<span class=normal> 916</span>
<span class=normal> 917</span>
<span class=normal> 918</span>
<span class=normal> 919</span>
<span class=normal> 920</span>
<span class=normal> 921</span>
<span class=normal> 922</span>
<span class=normal> 923</span>
<span class=normal> 924</span>
<span class=normal> 925</span>
<span class=normal> 926</span>
<span class=normal> 927</span>
<span class=normal> 928</span>
<span class=normal> 929</span>
<span class=normal> 930</span>
<span class=normal> 931</span>
<span class=normal> 932</span>
<span class=normal> 933</span>
<span class=normal> 934</span>
<span class=normal> 935</span>
<span class=normal> 936</span>
<span class=normal> 937</span>
<span class=normal> 938</span>
<span class=normal> 939</span>
<span class=normal> 940</span>
<span class=normal> 941</span>
<span class=normal> 942</span>
<span class=normal> 943</span>
<span class=normal> 944</span>
<span class=normal> 945</span>
<span class=normal> 946</span>
<span class=normal> 947</span>
<span class=normal> 948</span>
<span class=normal> 949</span>
<span class=normal> 950</span>
<span class=normal> 951</span>
<span class=normal> 952</span>
<span class=normal> 953</span>
<span class=normal> 954</span>
<span class=normal> 955</span>
<span class=normal> 956</span>
<span class=normal> 957</span>
<span class=normal> 958</span>
<span class=normal> 959</span>
<span class=normal> 960</span>
<span class=normal> 961</span>
<span class=normal> 962</span>
<span class=normal> 963</span>
<span class=normal> 964</span>
<span class=normal> 965</span>
<span class=normal> 966</span>
<span class=normal> 967</span>
<span class=normal> 968</span>
<span class=normal> 969</span>
<span class=normal> 970</span>
<span class=normal> 971</span>
<span class=normal> 972</span>
<span class=normal> 973</span>
<span class=normal> 974</span>
<span class=normal> 975</span>
<span class=normal> 976</span>
<span class=normal> 977</span>
<span class=normal> 978</span>
<span class=normal> 979</span>
<span class=normal> 980</span>
<span class=normal> 981</span>
<span class=normal> 982</span>
<span class=normal> 983</span>
<span class=normal> 984</span>
<span class=normal> 985</span>
<span class=normal> 986</span>
<span class=normal> 987</span>
<span class=normal> 988</span>
<span class=normal> 989</span>
<span class=normal> 990</span>
<span class=normal> 991</span>
<span class=normal> 992</span>
<span class=normal> 993</span>
<span class=normal> 994</span>
<span class=normal> 995</span>
<span class=normal> 996</span>
<span class=normal> 997</span>
<span class=normal> 998</span>
<span class=normal> 999</span>
<span class=normal>1000</span>
<span class=normal>1001</span>
<span class=normal>1002</span>
<span class=normal>1003</span>
<span class=normal>1004</span>
<span class=normal>1005</span>
<span class=normal>1006</span>
<span class=normal>1007</span>
<span class=normal>1008</span>
<span class=normal>1009</span>
<span class=normal>1010</span>
<span class=normal>1011</span>
<span class=normal>1012</span>
<span class=normal>1013</span>
<span class=normal>1014</span>
<span class=normal>1015</span>
<span class=normal>1016</span>
<span class=normal>1017</span>
<span class=normal>1018</span>
<span class=normal>1019</span>
<span class=normal>1020</span>
<span class=normal>1021</span>
<span class=normal>1022</span>
<span class=normal>1023</span>
<span class=normal>1024</span>
<span class=normal>1025</span>
<span class=normal>1026</span>
<span class=normal>1027</span>
<span class=normal>1028</span>
<span class=normal>1029</span>
<span class=normal>1030</span>
<span class=normal>1031</span>
<span class=normal>1032</span>
<span class=normal>1033</span>
<span class=normal>1034</span>
<span class=normal>1035</span>
<span class=normal>1036</span>
<span class=normal>1037</span>
<span class=normal>1038</span>
<span class=normal>1039</span>
<span class=normal>1040</span>
<span class=normal>1041</span>
<span class=normal>1042</span>
<span class=normal>1043</span>
<span class=normal>1044</span>
<span class=normal>1045</span>
<span class=normal>1046</span>
<span class=normal>1047</span>
<span class=normal>1048</span>
<span class=normal>1049</span>
<span class=normal>1050</span>
<span class=normal>1051</span>
<span class=normal>1052</span>
<span class=normal>1053</span>
<span class=normal>1054</span>
<span class=normal>1055</span>
<span class=normal>1056</span>
<span class=normal>1057</span>
<span class=normal>1058</span>
<span class=normal>1059</span>
<span class=normal>1060</span>
<span class=normal>1061</span>
<span class=normal>1062</span>
<span class=normal>1063</span>
<span class=normal>1064</span>
<span class=normal>1065</span>
<span class=normal>1066</span>
<span class=normal>1067</span>
<span class=normal>1068</span>
<span class=normal>1069</span>
<span class=normal>1070</span>
<span class=normal>1071</span>
<span class=normal>1072</span>
<span class=normal>1073</span>
<span class=normal>1074</span>
<span class=normal>1075</span>
<span class=normal>1076</span>
<span class=normal>1077</span>
<span class=normal>1078</span>
<span class=normal>1079</span>
<span class=normal>1080</span>
<span class=normal>1081</span>
<span class=normal>1082</span>
<span class=normal>1083</span>
<span class=normal>1084</span>
<span class=normal>1085</span>
<span class=normal>1086</span>
<span class=normal>1087</span>
<span class=normal>1088</span>
<span class=normal>1089</span>
<span class=normal>1090</span>
<span class=normal>1091</span>
<span class=normal>1092</span>
<span class=normal>1093</span>
<span class=normal>1094</span>
<span class=normal>1095</span>
<span class=normal>1096</span>
<span class=normal>1097</span>
<span class=normal>1098</span>
<span class=normal>1099</span>
<span class=normal>1100</span>
<span class=normal>1101</span>
<span class=normal>1102</span>
<span class=normal>1103</span>
<span class=normal>1104</span>
<span class=normal>1105</span>
<span class=normal>1106</span>
<span class=normal>1107</span>
<span class=normal>1108</span>
<span class=normal>1109</span>
<span class=normal>1110</span>
<span class=normal>1111</span>
<span class=normal>1112</span>
<span class=normal>1113</span>
<span class=normal>1114</span>
<span class=normal>1115</span>
<span class=normal>1116</span>
<span class=normal>1117</span>
<span class=normal>1118</span>
<span class=normal>1119</span>
<span class=normal>1120</span>
<span class=normal>1121</span>
<span class=normal>1122</span>
<span class=normal>1123</span>
<span class=normal>1124</span>
<span class=normal>1125</span>
<span class=normal>1126</span>
<span class=normal>1127</span>
<span class=normal>1128</span>
<span class=normal>1129</span>
<span class=normal>1130</span>
<span class=normal>1131</span>
<span class=normal>1132</span>
<span class=normal>1133</span>
<span class=normal>1134</span>
<span class=normal>1135</span>
<span class=normal>1136</span>
<span class=normal>1137</span>
<span class=normal>1138</span>
<span class=normal>1139</span>
<span class=normal>1140</span>
<span class=normal>1141</span>
<span class=normal>1142</span>
<span class=normal>1143</span>
<span class=normal>1144</span>
<span class=normal>1145</span>
<span class=normal>1146</span>
<span class=normal>1147</span>
<span class=normal>1148</span>
<span class=normal>1149</span>
<span class=normal>1150</span>
<span class=normal>1151</span>
<span class=normal>1152</span>
<span class=normal>1153</span>
<span class=normal>1154</span>
<span class=normal>1155</span>
<span class=normal>1156</span>
<span class=normal>1157</span>
<span class=normal>1158</span></pre></div></td><td class=code><div><pre><span></span><code><span class=nd>@dataclass</span>
<span class=k>class</span> <span class=nc>GWASCatalogAssociations</span><span class=p>(</span><span class=n>StudyLocus</span><span class=p>):</span>
<span class=w>    </span><span class=sd>&quot;&quot;&quot;Study-locus dataset derived from GWAS Catalog.&quot;&quot;&quot;</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_parse_pvalue</span><span class=p>(</span><span class=n>pvalue</span><span class=p>:</span> <span class=n>Column</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=nb>tuple</span><span class=p>[</span><span class=n>Column</span><span class=p>,</span> <span class=n>Column</span><span class=p>]:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Parse p-value column.</span>

<span class=sd>        Args:</span>
<span class=sd>            pvalue (Column): p-value [string]</span>

<span class=sd>        Returns:</span>
<span class=sd>            tuple[Column, Column]: p-value mantissa and exponent</span>

<span class=sd>        Example:</span>
<span class=sd>            &gt;&gt;&gt; import pyspark.sql.types as t</span>
<span class=sd>            &gt;&gt;&gt; d = [(&quot;1.0&quot;), (&quot;0.5&quot;), (&quot;1E-20&quot;), (&quot;3E-3&quot;), (&quot;1E-1000&quot;)]</span>
<span class=sd>            &gt;&gt;&gt; df = spark.createDataFrame(d, t.StringType())</span>
<span class=sd>            &gt;&gt;&gt; df.select(&#39;value&#39;,*GWASCatalogAssociations._parse_pvalue(f.col(&#39;value&#39;))).show()</span>
<span class=sd>            +-------+--------------+--------------+</span>
<span class=sd>            |  value|pValueMantissa|pValueExponent|</span>
<span class=sd>            +-------+--------------+--------------+</span>
<span class=sd>            |    1.0|           1.0|             1|</span>
<span class=sd>            |    0.5|           0.5|             1|</span>
<span class=sd>            |  1E-20|           1.0|           -20|</span>
<span class=sd>            |   3E-3|           3.0|            -3|</span>
<span class=sd>            |1E-1000|           1.0|         -1000|</span>
<span class=sd>            +-------+--------------+--------------+</span>
<span class=sd>            &lt;BLANKLINE&gt;</span>

<span class=sd>        &quot;&quot;&quot;</span>
        <span class=n>split</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>pvalue</span><span class=p>,</span> <span class=s2>&quot;E&quot;</span><span class=p>)</span>
        <span class=k>return</span> <span class=n>split</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>0</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=s2>&quot;float&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;pValueMantissa&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>coalesce</span><span class=p>(</span>
            <span class=n>split</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>1</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=s2>&quot;integer&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mi>1</span><span class=p>)</span>
        <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;pValueExponent&quot;</span><span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_normalise_pvaluetext</span><span class=p>(</span><span class=n>p_value_text</span><span class=p>:</span> <span class=n>Column</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Normalised p-value text column to a standardised format.</span>

<span class=sd>        For cases where there is no mapping, the value is set to null.</span>

<span class=sd>        Args:</span>
<span class=sd>            p_value_text (Column): `pValueText` column from GWASCatalog</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: Array column after using GWAS Catalog mappings. There might be multiple mappings for a single p-value text.</span>

<span class=sd>        Example:</span>
<span class=sd>            &gt;&gt;&gt; import pyspark.sql.types as t</span>
<span class=sd>            &gt;&gt;&gt; d = [(&quot;European Ancestry&quot;), (&quot;African ancestry&quot;), (&quot;Alzheimers Disease&quot;), (&quot;(progression)&quot;), (&quot;&quot;), (None)]</span>
<span class=sd>            &gt;&gt;&gt; df = spark.createDataFrame(d, t.StringType())</span>
<span class=sd>            &gt;&gt;&gt; df.withColumn(&#39;normalised&#39;, GWASCatalogAssociations._normalise_pvaluetext(f.col(&#39;value&#39;))).show()</span>
<span class=sd>            +-------------------+----------+</span>
<span class=sd>            |              value|normalised|</span>
<span class=sd>            +-------------------+----------+</span>
<span class=sd>            |  European Ancestry|      [EA]|</span>
<span class=sd>            |   African ancestry|      [AA]|</span>
<span class=sd>            |Alzheimers Disease|      [AD]|</span>
<span class=sd>            |      (progression)|      null|</span>
<span class=sd>            |                   |      null|</span>
<span class=sd>            |               null|      null|</span>
<span class=sd>            +-------------------+----------+</span>
<span class=sd>            &lt;BLANKLINE&gt;</span>

<span class=sd>        &quot;&quot;&quot;</span>
        <span class=c1># GWAS Catalog to p-value mapping</span>
        <span class=n>json_dict</span> <span class=o>=</span> <span class=n>json</span><span class=o>.</span><span class=n>loads</span><span class=p>(</span>
            <span class=n>pkg_resources</span><span class=o>.</span><span class=n>read_text</span><span class=p>(</span><span class=n>data</span><span class=p>,</span> <span class=s2>&quot;gwas_pValueText_map.json&quot;</span><span class=p>,</span> <span class=n>encoding</span><span class=o>=</span><span class=s2>&quot;utf-8&quot;</span><span class=p>)</span>
        <span class=p>)</span>
        <span class=n>map_expr</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>create_map</span><span class=p>(</span><span class=o>*</span><span class=p>[</span><span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>x</span><span class=p>)</span> <span class=k>for</span> <span class=n>x</span> <span class=ow>in</span> <span class=n>chain</span><span class=p>(</span><span class=o>*</span><span class=n>json_dict</span><span class=o>.</span><span class=n>items</span><span class=p>())])</span>

        <span class=n>splitted_col</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>p_value_text</span><span class=p>,</span> <span class=sa>r</span><span class=s2>&quot;[\(\)]&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>),</span> <span class=s2>&quot;,&quot;</span><span class=p>)</span>
        <span class=n>mapped_col</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=n>splitted_col</span><span class=p>,</span> <span class=k>lambda</span> <span class=n>x</span><span class=p>:</span> <span class=n>map_expr</span><span class=p>[</span><span class=n>x</span><span class=p>])</span>
        <span class=k>return</span> <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>forall</span><span class=p>(</span><span class=n>mapped_col</span><span class=p>,</span> <span class=k>lambda</span> <span class=n>x</span><span class=p>:</span> <span class=n>x</span><span class=o>.</span><span class=n>isNull</span><span class=p>()),</span> <span class=kc>None</span><span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span>
            <span class=n>mapped_col</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_normalise_risk_allele</span><span class=p>(</span><span class=n>risk_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Normalised risk allele column to a standardised format.</span>

<span class=sd>        If multiple risk alleles are present, the first one is returned.</span>

<span class=sd>        Args:</span>
<span class=sd>            risk_allele (Column): `riskAllele` column from GWASCatalog</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: mapped using GWAS Catalog mapping</span>

<span class=sd>        Example:</span>
<span class=sd>            &gt;&gt;&gt; import pyspark.sql.types as t</span>
<span class=sd>            &gt;&gt;&gt; d = [(&quot;rs1234-A-G&quot;), (&quot;rs1234-A&quot;), (&quot;rs1234-A; rs1235-G&quot;)]</span>
<span class=sd>            &gt;&gt;&gt; df = spark.createDataFrame(d, t.StringType())</span>
<span class=sd>            &gt;&gt;&gt; df.withColumn(&#39;normalised&#39;, GWASCatalogAssociations._normalise_risk_allele(f.col(&#39;value&#39;))).show()</span>
<span class=sd>            +------------------+----------+</span>
<span class=sd>            |             value|normalised|</span>
<span class=sd>            +------------------+----------+</span>
<span class=sd>            |        rs1234-A-G|         A|</span>
<span class=sd>            |          rs1234-A|         A|</span>
<span class=sd>            |rs1234-A; rs1235-G|         A|</span>
<span class=sd>            +------------------+----------+</span>
<span class=sd>            &lt;BLANKLINE&gt;</span>

<span class=sd>        &quot;&quot;&quot;</span>
        <span class=c1># GWAS Catalog to risk allele mapping</span>
        <span class=k>return</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>risk_allele</span><span class=p>,</span> <span class=s2>&quot;; &quot;</span><span class=p>)</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>0</span><span class=p>),</span> <span class=s2>&quot;-&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>1</span><span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_collect_rsids</span><span class=p>(</span>
        <span class=n>snp_id</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>snp_id_current</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>risk_allele</span><span class=p>:</span> <span class=n>Column</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;It takes three columns, and returns an array of distinct values from those columns.</span>

<span class=sd>        Args:</span>
<span class=sd>            snp_id (Column): The original snp id from the GWAS catalog.</span>
<span class=sd>            snp_id_current (Column): The current snp id field is just a number at the moment (stored as a string). Adding &#39;rs&#39; prefix if looks good.</span>
<span class=sd>            risk_allele (Column): The risk allele for the SNP.</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: An array of distinct values.</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=c1># The current snp id field is just a number at the moment (stored as a string). Adding &#39;rs&#39; prefix if looks good.</span>
        <span class=n>snp_id_current</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
            <span class=n>snp_id_current</span><span class=o>.</span><span class=n>rlike</span><span class=p>(</span><span class=s2>&quot;^[0-9]*$&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>format_string</span><span class=p>(</span><span class=s2>&quot;rs</span><span class=si>%s</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>snp_id_current</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Cleaning risk allele:</span>
        <span class=n>risk_allele</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>risk_allele</span><span class=p>,</span> <span class=s2>&quot;-&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>0</span><span class=p>)</span>

        <span class=c1># Collecting all values:</span>
        <span class=k>return</span> <span class=n>f</span><span class=o>.</span><span class=n>array_distinct</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>array</span><span class=p>(</span><span class=n>snp_id</span><span class=p>,</span> <span class=n>snp_id_current</span><span class=p>,</span> <span class=n>risk_allele</span><span class=p>))</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_map_to_variant_annotation_variants</span><span class=p>(</span>
        <span class=n>gwas_associations</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span> <span class=n>variant_annotation</span><span class=p>:</span> <span class=n>VariantAnnotation</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Add variant metadata in associations.</span>

<span class=sd>        Args:</span>
<span class=sd>            gwas_associations (DataFrame): raw GWAS Catalog associations</span>
<span class=sd>            variant_annotation (VariantAnnotation): variant annotation dataset</span>

<span class=sd>        Returns:</span>
<span class=sd>            DataFrame: GWAS Catalog associations data including `variantId`, `referenceAllele`,</span>
<span class=sd>            `alternateAllele`, `chromosome`, `position` with variant metadata</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=c1># Subset of GWAS Catalog associations required for resolving variant IDs:</span>
        <span class=n>gwas_associations_subset</span> <span class=o>=</span> <span class=n>gwas_associations</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;studyLocusId&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;CHR_ID&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;CHR_POS&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>IntegerType</span><span class=p>())</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;position&quot;</span><span class=p>),</span>
            <span class=c1># List of all SNPs associated with the variant</span>
            <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_collect_rsids</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;SNPS&quot;</span><span class=p>),</span> <span class=s2>&quot;; &quot;</span><span class=p>)</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>0</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;SNP_ID_CURRENT&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>),</span> <span class=s2>&quot;; &quot;</span><span class=p>)</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>0</span><span class=p>),</span>
            <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;rsIdsGwasCatalog&quot;</span><span class=p>),</span>
            <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_normalise_risk_allele</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>)</span>
            <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;riskAllele&quot;</span><span class=p>),</span>
        <span class=p>)</span>

        <span class=c1># Subset of variant annotation required for GWAS Catalog annotations:</span>
        <span class=n>va_subset</span> <span class=o>=</span> <span class=n>variant_annotation</span><span class=o>.</span><span class=n>df</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;variantId&quot;</span><span class=p>,</span>
            <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
            <span class=s2>&quot;position&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;rsIds&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;rsIdsGnomad&quot;</span><span class=p>),</span>
            <span class=s2>&quot;referenceAllele&quot;</span><span class=p>,</span>
            <span class=s2>&quot;alternateAllele&quot;</span><span class=p>,</span>
            <span class=s2>&quot;alleleFrequencies&quot;</span><span class=p>,</span>
            <span class=n>variant_annotation</span><span class=o>.</span><span class=n>max_maf</span><span class=p>()</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;maxMaf&quot;</span><span class=p>),</span>
        <span class=p>)</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>broadcast</span><span class=p>(</span>
                <span class=n>gwas_associations_subset</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>,</span> <span class=s2>&quot;position&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span>
            <span class=p>),</span>
            <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=s2>&quot;chromosome&quot;</span><span class=p>,</span> <span class=s2>&quot;position&quot;</span><span class=p>],</span>
            <span class=n>how</span><span class=o>=</span><span class=s2>&quot;inner&quot;</span><span class=p>,</span>
        <span class=p>)</span>

        <span class=c1># Semi-resolved ids (still contains duplicates when conclusion was not possible to make</span>
        <span class=c1># based on rsIds or allele concordance)</span>
        <span class=n>filtered_associations</span> <span class=o>=</span> <span class=p>(</span>
            <span class=n>gwas_associations_subset</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>broadcast</span><span class=p>(</span><span class=n>va_subset</span><span class=p>),</span>
                <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=s2>&quot;chromosome&quot;</span><span class=p>,</span> <span class=s2>&quot;position&quot;</span><span class=p>],</span>
                <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
                <span class=s2>&quot;rsIdFilter&quot;</span><span class=p>,</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_flag_mappings_to_retain</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;studyLocusId&quot;</span><span class=p>),</span>
                    <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_compare_rsids</span><span class=p>(</span>
                        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;rsIdsGnomad&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;rsIdsGwasCatalog&quot;</span><span class=p>)</span>
                    <span class=p>),</span>
                <span class=p>),</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
                <span class=s2>&quot;concordanceFilter&quot;</span><span class=p>,</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_flag_mappings_to_retain</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;studyLocusId&quot;</span><span class=p>),</span>
                    <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_check_concordance</span><span class=p>(</span>
                        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;riskAllele&quot;</span><span class=p>),</span>
                        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                    <span class=p>),</span>
                <span class=p>),</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
                <span class=c1># Filter out rows where GWAS Catalog rsId does not match with GnomAD rsId,</span>
                <span class=c1># but there is corresponding variant for the same association</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;rsIdFilter&quot;</span><span class=p>)</span>
                <span class=c1># or filter out rows where GWAS Catalog alleles are not concordant with GnomAD alleles,</span>
                <span class=c1># but there is corresponding variant for the same association</span>
                <span class=o>|</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;concordanceFilter&quot;</span><span class=p>)</span>
            <span class=p>)</span>
        <span class=p>)</span>

        <span class=c1># Keep only highest maxMaf variant per studyLocusId</span>
        <span class=n>fully_mapped_associations</span> <span class=o>=</span> <span class=n>get_record_with_maximum_value</span><span class=p>(</span>
            <span class=n>filtered_associations</span><span class=p>,</span> <span class=n>grouping_col</span><span class=o>=</span><span class=s2>&quot;studyLocusId&quot;</span><span class=p>,</span> <span class=n>sorting_col</span><span class=o>=</span><span class=s2>&quot;maxMaf&quot;</span>
        <span class=p>)</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;studyLocusId&quot;</span><span class=p>,</span>
            <span class=s2>&quot;variantId&quot;</span><span class=p>,</span>
            <span class=s2>&quot;referenceAllele&quot;</span><span class=p>,</span>
            <span class=s2>&quot;alternateAllele&quot;</span><span class=p>,</span>
            <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
            <span class=s2>&quot;position&quot;</span><span class=p>,</span>
        <span class=p>)</span>

        <span class=k>return</span> <span class=n>gwas_associations</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>fully_mapped_associations</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=s2>&quot;studyLocusId&quot;</span><span class=p>,</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_compare_rsids</span><span class=p>(</span><span class=n>gnomad</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>gwas</span><span class=p>:</span> <span class=n>Column</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;If the intersection of the two arrays is greater than 0, return True, otherwise return False.</span>

<span class=sd>        Args:</span>
<span class=sd>            gnomad (Column): rsids from gnomad</span>
<span class=sd>            gwas (Column): rsids from the GWAS Catalog</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: A boolean column that is true if the GnomAD rsIDs can be found in the GWAS rsIDs.</span>

<span class=sd>        Examples:</span>
<span class=sd>            &gt;&gt;&gt; d = [</span>
<span class=sd>            ...    (1, [&quot;rs123&quot;, &quot;rs523&quot;], [&quot;rs123&quot;]),</span>
<span class=sd>            ...    (2, [], [&quot;rs123&quot;]),</span>
<span class=sd>            ...    (3, [&quot;rs123&quot;, &quot;rs523&quot;], []),</span>
<span class=sd>            ...    (4, [], []),</span>
<span class=sd>            ... ]</span>
<span class=sd>            &gt;&gt;&gt; df = spark.createDataFrame(d, [&#39;associationId&#39;, &#39;gnomad&#39;, &#39;gwas&#39;])</span>
<span class=sd>            &gt;&gt;&gt; df.withColumn(&quot;rsid_matches&quot;, GWASCatalogAssociations._compare_rsids(f.col(&quot;gnomad&quot;),f.col(&#39;gwas&#39;))).show()</span>
<span class=sd>            +-------------+--------------+-------+------------+</span>
<span class=sd>            |associationId|        gnomad|   gwas|rsid_matches|</span>
<span class=sd>            +-------------+--------------+-------+------------+</span>
<span class=sd>            |            1|[rs123, rs523]|[rs123]|        true|</span>
<span class=sd>            |            2|            []|[rs123]|       false|</span>
<span class=sd>            |            3|[rs123, rs523]|     []|       false|</span>
<span class=sd>            |            4|            []|     []|       false|</span>
<span class=sd>            +-------------+--------------+-------+------------+</span>
<span class=sd>            &lt;BLANKLINE&gt;</span>

<span class=sd>        &quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>size</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>array_intersect</span><span class=p>(</span><span class=n>gnomad</span><span class=p>,</span> <span class=n>gwas</span><span class=p>))</span> <span class=o>&gt;</span> <span class=mi>0</span><span class=p>,</span> <span class=kc>True</span><span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span>
            <span class=kc>False</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_flag_mappings_to_retain</span><span class=p>(</span>
        <span class=n>association_id</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>filter_column</span><span class=p>:</span> <span class=n>Column</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Flagging mappings to drop for each association.</span>

<span class=sd>        Some associations have multiple mappings. Some has matching rsId others don&#39;t. We only</span>
<span class=sd>        want to drop the non-matching mappings, when a matching is available for the given association.</span>
<span class=sd>        This logic can be generalised for other measures eg. allele concordance.</span>

<span class=sd>        Args:</span>
<span class=sd>            association_id (Column): association identifier column</span>
<span class=sd>            filter_column (Column): boolean col indicating to keep a mapping</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: A column with a boolean value.</span>

<span class=sd>        Examples:</span>
<span class=sd>        &gt;&gt;&gt; d = [</span>
<span class=sd>        ...    (1, False),</span>
<span class=sd>        ...    (1, False),</span>
<span class=sd>        ...    (2, False),</span>
<span class=sd>        ...    (2, True),</span>
<span class=sd>        ...    (3, True),</span>
<span class=sd>        ...    (3, True),</span>
<span class=sd>        ... ]</span>
<span class=sd>        &gt;&gt;&gt; df = spark.createDataFrame(d, [&#39;associationId&#39;, &#39;filter&#39;])</span>
<span class=sd>        &gt;&gt;&gt; df.withColumn(&quot;isConcordant&quot;, GWASCatalogAssociations._flag_mappings_to_retain(f.col(&quot;associationId&quot;),f.col(&#39;filter&#39;))).show()</span>
<span class=sd>        +-------------+------+------------+</span>
<span class=sd>        |associationId|filter|isConcordant|</span>
<span class=sd>        +-------------+------+------------+</span>
<span class=sd>        |            1| false|        true|</span>
<span class=sd>        |            1| false|        true|</span>
<span class=sd>        |            2| false|       false|</span>
<span class=sd>        |            2|  true|        true|</span>
<span class=sd>        |            3|  true|        true|</span>
<span class=sd>        |            3|  true|        true|</span>
<span class=sd>        +-------------+------+------------+</span>
<span class=sd>        &lt;BLANKLINE&gt;</span>

<span class=sd>        &quot;&quot;&quot;</span>
        <span class=n>w</span> <span class=o>=</span> <span class=n>Window</span><span class=o>.</span><span class=n>partitionBy</span><span class=p>(</span><span class=n>association_id</span><span class=p>)</span>

        <span class=c1># Generating a boolean column informing if the filter column contains true anywhere for the association:</span>
        <span class=n>aggregated_filter</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>array_contains</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>collect_set</span><span class=p>(</span><span class=n>filter_column</span><span class=p>)</span><span class=o>.</span><span class=n>over</span><span class=p>(</span><span class=n>w</span><span class=p>),</span> <span class=kc>True</span><span class=p>),</span> <span class=kc>True</span>
        <span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=kc>False</span><span class=p>)</span>

        <span class=c1># Generate a filter column:</span>
        <span class=k>return</span> <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>aggregated_filter</span> <span class=o>&amp;</span> <span class=p>(</span><span class=o>~</span><span class=n>filter_column</span><span class=p>),</span> <span class=kc>False</span><span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=kc>True</span><span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_check_concordance</span><span class=p>(</span>
        <span class=n>risk_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>reference_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>alternate_allele</span><span class=p>:</span> <span class=n>Column</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;A function to check if the risk allele is concordant with the alt or ref allele.</span>

<span class=sd>        If the risk allele is the same as the reference or alternate allele, or if the reverse complement of</span>
<span class=sd>        the risk allele is the same as the reference or alternate allele, then the allele is concordant.</span>
<span class=sd>        If no mapping is available (ref/alt is null), the function returns True.</span>

<span class=sd>        Args:</span>
<span class=sd>            risk_allele (Column): The allele that is associated with the risk of the disease.</span>
<span class=sd>            reference_allele (Column): The reference allele from the GWAS catalog</span>
<span class=sd>            alternate_allele (Column): The alternate allele of the variant.</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: A boolean column that is True if the risk allele is the same as the reference or alternate allele,</span>
<span class=sd>            or if the reverse complement of the risk allele is the same as the reference or alternate allele.</span>

<span class=sd>        Examples:</span>
<span class=sd>            &gt;&gt;&gt; d = [</span>
<span class=sd>            ...     (&#39;A&#39;, &#39;A&#39;, &#39;G&#39;),</span>
<span class=sd>            ...     (&#39;A&#39;, &#39;T&#39;, &#39;G&#39;),</span>
<span class=sd>            ...     (&#39;A&#39;, &#39;C&#39;, &#39;G&#39;),</span>
<span class=sd>            ...     (&#39;A&#39;, &#39;A&#39;, &#39;?&#39;),</span>
<span class=sd>            ...     (None, None, &#39;A&#39;),</span>
<span class=sd>            ... ]</span>
<span class=sd>            &gt;&gt;&gt; df = spark.createDataFrame(d, [&#39;riskAllele&#39;, &#39;referenceAllele&#39;, &#39;alternateAllele&#39;])</span>
<span class=sd>            &gt;&gt;&gt; df.withColumn(&quot;isConcordant&quot;, GWASCatalogAssociations._check_concordance(f.col(&quot;riskAllele&quot;),f.col(&#39;referenceAllele&#39;), f.col(&#39;alternateAllele&#39;))).show()</span>
<span class=sd>            +----------+---------------+---------------+------------+</span>
<span class=sd>            |riskAllele|referenceAllele|alternateAllele|isConcordant|</span>
<span class=sd>            +----------+---------------+---------------+------------+</span>
<span class=sd>            |         A|              A|              G|        true|</span>
<span class=sd>            |         A|              T|              G|        true|</span>
<span class=sd>            |         A|              C|              G|       false|</span>
<span class=sd>            |         A|              A|              ?|        true|</span>
<span class=sd>            |      null|           null|              A|        true|</span>
<span class=sd>            +----------+---------------+---------------+------------+</span>
<span class=sd>            &lt;BLANKLINE&gt;</span>

<span class=sd>        &quot;&quot;&quot;</span>
        <span class=c1># Calculating the reverse complement of the risk allele:</span>
        <span class=n>risk_allele_reverse_complement</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
            <span class=n>risk_allele</span><span class=o>.</span><span class=n>rlike</span><span class=p>(</span><span class=sa>r</span><span class=s2>&quot;^[ACTG]+$&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>reverse</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>translate</span><span class=p>(</span><span class=n>risk_allele</span><span class=p>,</span> <span class=s2>&quot;ACTG&quot;</span><span class=p>,</span> <span class=s2>&quot;TGAC&quot;</span><span class=p>)),</span>
        <span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>risk_allele</span><span class=p>)</span>

        <span class=c1># OK, is the risk allele or the reverse complent is the same as the mapped alleles:</span>
        <span class=k>return</span> <span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=p>(</span><span class=n>risk_allele</span> <span class=o>==</span> <span class=n>reference_allele</span><span class=p>)</span> <span class=o>|</span> <span class=p>(</span><span class=n>risk_allele</span> <span class=o>==</span> <span class=n>alternate_allele</span><span class=p>),</span>
                <span class=kc>True</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=c1># If risk allele is found on the negative strand:</span>
            <span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=p>(</span><span class=n>risk_allele_reverse_complement</span> <span class=o>==</span> <span class=n>reference_allele</span><span class=p>)</span>
                <span class=o>|</span> <span class=p>(</span><span class=n>risk_allele_reverse_complement</span> <span class=o>==</span> <span class=n>alternate_allele</span><span class=p>),</span>
                <span class=kc>True</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=c1># If risk allele is ambiguous, still accepted: &lt; This condition could be reconsidered</span>
            <span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>risk_allele</span> <span class=o>==</span> <span class=s2>&quot;?&quot;</span><span class=p>,</span> <span class=kc>True</span><span class=p>)</span>
            <span class=c1># If the association could not be mapped we keep it:</span>
            <span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>reference_allele</span><span class=o>.</span><span class=n>isNull</span><span class=p>(),</span> <span class=kc>True</span><span class=p>)</span>
            <span class=c1># Allele is discordant:</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=kc>False</span><span class=p>)</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_get_reverse_complement</span><span class=p>(</span><span class=n>allele_col</span><span class=p>:</span> <span class=n>Column</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;A function to return the reverse complement of an allele column.</span>

<span class=sd>        It takes a string and returns the reverse complement of that string if it&#39;s a DNA sequence,</span>
<span class=sd>        otherwise it returns the original string. Assumes alleles in upper case.</span>

<span class=sd>        Args:</span>
<span class=sd>            allele_col (Column): The column containing the allele to reverse complement.</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: A column that is the reverse complement of the allele column.</span>

<span class=sd>        Examples:</span>
<span class=sd>            &gt;&gt;&gt; d = [{&quot;allele&quot;: &#39;A&#39;}, {&quot;allele&quot;: &#39;T&#39;},{&quot;allele&quot;: &#39;G&#39;}, {&quot;allele&quot;: &#39;C&#39;},{&quot;allele&quot;: &#39;AC&#39;}, {&quot;allele&quot;: &#39;GTaatc&#39;},{&quot;allele&quot;: &#39;?&#39;}, {&quot;allele&quot;: None}]</span>
<span class=sd>            &gt;&gt;&gt; df = spark.createDataFrame(d)</span>
<span class=sd>            &gt;&gt;&gt; df.withColumn(&quot;revcom_allele&quot;, GWASCatalogAssociations._get_reverse_complement(f.col(&quot;allele&quot;))).show()</span>
<span class=sd>            +------+-------------+</span>
<span class=sd>            |allele|revcom_allele|</span>
<span class=sd>            +------+-------------+</span>
<span class=sd>            |     A|            T|</span>
<span class=sd>            |     T|            A|</span>
<span class=sd>            |     G|            C|</span>
<span class=sd>            |     C|            G|</span>
<span class=sd>            |    AC|           GT|</span>
<span class=sd>            |GTaatc|       GATTAC|</span>
<span class=sd>            |     ?|            ?|</span>
<span class=sd>            |  null|         null|</span>
<span class=sd>            +------+-------------+</span>
<span class=sd>            &lt;BLANKLINE&gt;</span>

<span class=sd>        &quot;&quot;&quot;</span>
        <span class=n>allele_col</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>upper</span><span class=p>(</span><span class=n>allele_col</span><span class=p>)</span>
        <span class=k>return</span> <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
            <span class=n>allele_col</span><span class=o>.</span><span class=n>rlike</span><span class=p>(</span><span class=s2>&quot;[ACTG]+&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>reverse</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>translate</span><span class=p>(</span><span class=n>allele_col</span><span class=p>,</span> <span class=s2>&quot;ACTG&quot;</span><span class=p>,</span> <span class=s2>&quot;TGAC&quot;</span><span class=p>)),</span>
        <span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>allele_col</span><span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_effect_needs_harmonisation</span><span class=p>(</span>
        <span class=n>risk_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>reference_allele</span><span class=p>:</span> <span class=n>Column</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;A function to check if the effect allele needs to be harmonised.</span>

<span class=sd>        Args:</span>
<span class=sd>            risk_allele (Column): Risk allele column</span>
<span class=sd>            reference_allele (Column): Effect allele column</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: A boolean column indicating if the effect allele needs to be harmonised.</span>

<span class=sd>        Examples:</span>
<span class=sd>            &gt;&gt;&gt; d = [{&quot;risk&quot;: &#39;A&#39;, &quot;reference&quot;: &#39;A&#39;}, {&quot;risk&quot;: &#39;A&#39;, &quot;reference&quot;: &#39;T&#39;}, {&quot;risk&quot;: &#39;AT&#39;, &quot;reference&quot;: &#39;TA&#39;}, {&quot;risk&quot;: &#39;AT&#39;, &quot;reference&quot;: &#39;AT&#39;}]</span>
<span class=sd>            &gt;&gt;&gt; df = spark.createDataFrame(d)</span>
<span class=sd>            &gt;&gt;&gt; df.withColumn(&quot;needs_harmonisation&quot;, GWASCatalogAssociations._effect_needs_harmonisation(f.col(&quot;risk&quot;), f.col(&quot;reference&quot;))).show()</span>
<span class=sd>            +---------+----+-------------------+</span>
<span class=sd>            |reference|risk|needs_harmonisation|</span>
<span class=sd>            +---------+----+-------------------+</span>
<span class=sd>            |        A|   A|               true|</span>
<span class=sd>            |        T|   A|               true|</span>
<span class=sd>            |       TA|  AT|              false|</span>
<span class=sd>            |       AT|  AT|               true|</span>
<span class=sd>            +---------+----+-------------------+</span>
<span class=sd>            &lt;BLANKLINE&gt;</span>

<span class=sd>        &quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=p>(</span><span class=n>risk_allele</span> <span class=o>==</span> <span class=n>reference_allele</span><span class=p>)</span> <span class=o>|</span> <span class=p>(</span>
            <span class=n>risk_allele</span>
            <span class=o>==</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_get_reverse_complement</span><span class=p>(</span><span class=n>reference_allele</span><span class=p>)</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_are_alleles_palindromic</span><span class=p>(</span>
        <span class=n>reference_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>alternate_allele</span><span class=p>:</span> <span class=n>Column</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;A function to check if the alleles are palindromic.</span>

<span class=sd>        Args:</span>
<span class=sd>            reference_allele (Column): Reference allele column</span>
<span class=sd>            alternate_allele (Column): Alternate allele column</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: A boolean column indicating if the alleles are palindromic.</span>

<span class=sd>        Examples:</span>
<span class=sd>            &gt;&gt;&gt; d = [{&quot;reference&quot;: &#39;A&#39;, &quot;alternate&quot;: &#39;T&#39;}, {&quot;reference&quot;: &#39;AT&#39;, &quot;alternate&quot;: &#39;AG&#39;}, {&quot;reference&quot;: &#39;AT&#39;, &quot;alternate&quot;: &#39;AT&#39;}, {&quot;reference&quot;: &#39;CATATG&#39;, &quot;alternate&quot;: &#39;CATATG&#39;}, {&quot;reference&quot;: &#39;-&#39;, &quot;alternate&quot;: None}]</span>
<span class=sd>            &gt;&gt;&gt; df = spark.createDataFrame(d)</span>
<span class=sd>            &gt;&gt;&gt; df.withColumn(&quot;is_palindromic&quot;, GWASCatalogAssociations._are_alleles_palindromic(f.col(&quot;reference&quot;), f.col(&quot;alternate&quot;))).show()</span>
<span class=sd>            +---------+---------+--------------+</span>
<span class=sd>            |alternate|reference|is_palindromic|</span>
<span class=sd>            +---------+---------+--------------+</span>
<span class=sd>            |        T|        A|          true|</span>
<span class=sd>            |       AG|       AT|         false|</span>
<span class=sd>            |       AT|       AT|          true|</span>
<span class=sd>            |   CATATG|   CATATG|          true|</span>
<span class=sd>            |     null|        -|         false|</span>
<span class=sd>            +---------+---------+--------------+</span>
<span class=sd>            &lt;BLANKLINE&gt;</span>

<span class=sd>        &quot;&quot;&quot;</span>
        <span class=n>revcomp</span> <span class=o>=</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_get_reverse_complement</span><span class=p>(</span><span class=n>alternate_allele</span><span class=p>)</span>
        <span class=k>return</span> <span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>reference_allele</span> <span class=o>==</span> <span class=n>revcomp</span><span class=p>,</span> <span class=kc>True</span><span class=p>)</span>
            <span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>revcomp</span><span class=o>.</span><span class=n>isNull</span><span class=p>(),</span> <span class=kc>False</span><span class=p>)</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=kc>False</span><span class=p>)</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_harmonise_beta</span><span class=p>(</span>
        <span class=n>risk_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>reference_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>alternate_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>effect_size</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>confidence_interval</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;A function to extract the beta value from the effect size and confidence interval.</span>

<span class=sd>        If the confidence interval contains the word &quot;increase&quot; or &quot;decrease&quot; it indicates, we are dealing with betas.</span>
<span class=sd>        If it&#39;s &quot;increase&quot; and the effect size needs to be harmonized, then multiply the effect size by -1</span>

<span class=sd>        Args:</span>
<span class=sd>            risk_allele (Column): Risk allele column</span>
<span class=sd>            reference_allele (Column): Reference allele column</span>
<span class=sd>            alternate_allele (Column): Alternate allele column</span>
<span class=sd>            effect_size (Column): GWAS Catalog effect size column</span>
<span class=sd>            confidence_interval (Column): GWAS Catalog confidence interval column</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: A column containing the beta value.</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_are_alleles_palindromic</span><span class=p>(</span>
                    <span class=n>reference_allele</span><span class=p>,</span> <span class=n>alternate_allele</span>
                <span class=p>),</span>
                <span class=kc>None</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=p>(</span>
                    <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_effect_needs_harmonisation</span><span class=p>(</span>
                        <span class=n>risk_allele</span><span class=p>,</span> <span class=n>reference_allele</span>
                    <span class=p>)</span>
                    <span class=o>&amp;</span> <span class=n>confidence_interval</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot;increase&quot;</span><span class=p>)</span>
                <span class=p>)</span>
                <span class=o>|</span> <span class=p>(</span>
                    <span class=o>~</span><span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_effect_needs_harmonisation</span><span class=p>(</span>
                        <span class=n>risk_allele</span><span class=p>,</span> <span class=n>reference_allele</span>
                    <span class=p>)</span>
                    <span class=o>&amp;</span> <span class=n>confidence_interval</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot;decrease&quot;</span><span class=p>)</span>
                <span class=p>),</span>
                <span class=o>-</span><span class=n>effect_size</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>effect_size</span><span class=p>)</span>
            <span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>DoubleType</span><span class=p>())</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_harmonise_beta_ci</span><span class=p>(</span>
        <span class=n>risk_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>reference_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>alternate_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>effect_size</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>confidence_interval</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>p_value</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>direction</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Calculating confidence intervals for beta values.</span>

<span class=sd>        Args:</span>
<span class=sd>            risk_allele (Column): Risk allele column</span>
<span class=sd>            reference_allele (Column): Reference allele column</span>
<span class=sd>            alternate_allele (Column): Alternate allele column</span>
<span class=sd>            effect_size (Column): GWAS Catalog effect size column</span>
<span class=sd>            confidence_interval (Column): GWAS Catalog confidence interval column</span>
<span class=sd>            p_value (Column): GWAS Catalog p-value column</span>
<span class=sd>            direction (str): This is the direction of the confidence interval. It can be either &quot;upper&quot; or &quot;lower&quot;.</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: The upper and lower bounds of the confidence interval for the beta coefficient.</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=n>zscore_95</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mf>1.96</span><span class=p>)</span>
        <span class=n>beta</span> <span class=o>=</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_harmonise_beta</span><span class=p>(</span>
            <span class=n>risk_allele</span><span class=p>,</span>
            <span class=n>reference_allele</span><span class=p>,</span>
            <span class=n>alternate_allele</span><span class=p>,</span>
            <span class=n>effect_size</span><span class=p>,</span>
            <span class=n>confidence_interval</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=n>zscore</span> <span class=o>=</span> <span class=n>pvalue_to_zscore</span><span class=p>(</span><span class=n>p_value</span><span class=p>)</span>
        <span class=k>return</span> <span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>direction</span><span class=p>)</span> <span class=o>==</span> <span class=s2>&quot;upper&quot;</span><span class=p>,</span> <span class=n>beta</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>abs</span><span class=p>(</span><span class=n>zscore_95</span> <span class=o>*</span> <span class=n>beta</span><span class=p>)</span> <span class=o>/</span> <span class=n>zscore</span><span class=p>)</span>
            <span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>direction</span><span class=p>)</span> <span class=o>==</span> <span class=s2>&quot;lower&quot;</span><span class=p>,</span> <span class=n>beta</span> <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>abs</span><span class=p>(</span><span class=n>zscore_95</span> <span class=o>*</span> <span class=n>beta</span><span class=p>)</span> <span class=o>/</span> <span class=n>zscore</span><span class=p>)</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=kc>None</span><span class=p>)</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_harmonise_odds_ratio</span><span class=p>(</span>
        <span class=n>risk_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>reference_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>alternate_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>effect_size</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>confidence_interval</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Harmonizing odds ratio.</span>

<span class=sd>        Args:</span>
<span class=sd>            risk_allele (Column): Risk allele column</span>
<span class=sd>            reference_allele (Column): Reference allele column</span>
<span class=sd>            alternate_allele (Column): Alternate allele column</span>
<span class=sd>            effect_size (Column): GWAS Catalog effect size column</span>
<span class=sd>            confidence_interval (Column): GWAS Catalog confidence interval column</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: A column with the odds ratio, or 1/odds_ratio if harmonization required.</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_are_alleles_palindromic</span><span class=p>(</span>
                    <span class=n>reference_allele</span><span class=p>,</span> <span class=n>alternate_allele</span>
                <span class=p>),</span>
                <span class=kc>None</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=p>(</span>
                    <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_effect_needs_harmonisation</span><span class=p>(</span>
                        <span class=n>risk_allele</span><span class=p>,</span> <span class=n>reference_allele</span>
                    <span class=p>)</span>
                    <span class=o>&amp;</span> <span class=o>~</span><span class=n>confidence_interval</span><span class=o>.</span><span class=n>rlike</span><span class=p>(</span><span class=s2>&quot;|&quot;</span><span class=o>.</span><span class=n>join</span><span class=p>([</span><span class=s2>&quot;decrease&quot;</span><span class=p>,</span> <span class=s2>&quot;increase&quot;</span><span class=p>]))</span>
                <span class=p>),</span>
                <span class=mi>1</span> <span class=o>/</span> <span class=n>effect_size</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>effect_size</span><span class=p>)</span>
            <span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>DoubleType</span><span class=p>())</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_harmonise_odds_ratio_ci</span><span class=p>(</span>
        <span class=n>risk_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>reference_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>alternate_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>effect_size</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>confidence_interval</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>p_value</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>direction</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Calculating confidence intervals for beta values.</span>

<span class=sd>        Args:</span>
<span class=sd>            risk_allele (Column): Risk allele column</span>
<span class=sd>            reference_allele (Column): Reference allele column</span>
<span class=sd>            alternate_allele (Column): Alternate allele column</span>
<span class=sd>            effect_size (Column): GWAS Catalog effect size column</span>
<span class=sd>            confidence_interval (Column): GWAS Catalog confidence interval column</span>
<span class=sd>            p_value (Column): GWAS Catalog p-value column</span>
<span class=sd>            direction (str): This is the direction of the confidence interval. It can be either &quot;upper&quot; or &quot;lower&quot;.</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: The upper and lower bounds of the 95% confidence interval for the odds ratio.</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=n>zscore_95</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mf>1.96</span><span class=p>)</span>
        <span class=n>odds_ratio</span> <span class=o>=</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_harmonise_odds_ratio</span><span class=p>(</span>
            <span class=n>risk_allele</span><span class=p>,</span>
            <span class=n>reference_allele</span><span class=p>,</span>
            <span class=n>alternate_allele</span><span class=p>,</span>
            <span class=n>effect_size</span><span class=p>,</span>
            <span class=n>confidence_interval</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=n>odds_ratio_estimate</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>log</span><span class=p>(</span><span class=n>odds_ratio</span><span class=p>)</span>
        <span class=n>zscore</span> <span class=o>=</span> <span class=n>pvalue_to_zscore</span><span class=p>(</span><span class=n>p_value</span><span class=p>)</span>
        <span class=n>odds_ratio_se</span> <span class=o>=</span> <span class=n>odds_ratio_estimate</span> <span class=o>/</span> <span class=n>zscore</span>
        <span class=k>return</span> <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>direction</span><span class=p>)</span> <span class=o>==</span> <span class=s2>&quot;upper&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>exp</span><span class=p>(</span><span class=n>odds_ratio_estimate</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>abs</span><span class=p>(</span><span class=n>zscore_95</span> <span class=o>*</span> <span class=n>odds_ratio_se</span><span class=p>)),</span>
        <span class=p>)</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>direction</span><span class=p>)</span> <span class=o>==</span> <span class=s2>&quot;lower&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>exp</span><span class=p>(</span><span class=n>odds_ratio_estimate</span> <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>abs</span><span class=p>(</span><span class=n>zscore_95</span> <span class=o>*</span> <span class=n>odds_ratio_se</span><span class=p>)),</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_concatenate_substudy_description</span><span class=p>(</span>
        <span class=n>association_trait</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>pvalue_text</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>mapped_trait_uri</span><span class=p>:</span> <span class=n>Column</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Substudy description parsing. Complex string containing metadata about the substudy (e.g. QTL, specific EFO, etc.).</span>

<span class=sd>        Args:</span>
<span class=sd>            association_trait (Column): GWAS Catalog association trait column</span>
<span class=sd>            pvalue_text (Column): GWAS Catalog p-value text column</span>
<span class=sd>            mapped_trait_uri (Column): GWAS Catalog mapped trait URI column</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: A column with the substudy description in the shape trait|pvaluetext1_pvaluetext2|EFO1_EFO2.</span>

<span class=sd>        Examples:</span>
<span class=sd>        &gt;&gt;&gt; df = spark.createDataFrame([</span>
<span class=sd>        ...    (&quot;Height&quot;, &quot;http://www.ebi.ac.uk/efo/EFO_0000001,http://www.ebi.ac.uk/efo/EFO_0000002&quot;, &quot;European Ancestry&quot;),</span>
<span class=sd>        ...    (&quot;Schizophrenia&quot;, &quot;http://www.ebi.ac.uk/efo/MONDO_0005090&quot;, None)],</span>
<span class=sd>        ...    [&quot;association_trait&quot;, &quot;mapped_trait_uri&quot;, &quot;pvalue_text&quot;]</span>
<span class=sd>        ... )</span>
<span class=sd>        &gt;&gt;&gt; df.withColumn(&#39;substudy_description&#39;, GWASCatalogAssociations._concatenate_substudy_description(df.association_trait, df.pvalue_text, df.mapped_trait_uri)).show(truncate=False)</span>
<span class=sd>        +-----------------+-------------------------------------------------------------------------+-----------------+------------------------------------------+</span>
<span class=sd>        |association_trait|mapped_trait_uri                                                         |pvalue_text      |substudy_description                      |</span>
<span class=sd>        +-----------------+-------------------------------------------------------------------------+-----------------+------------------------------------------+</span>
<span class=sd>        |Height           |http://www.ebi.ac.uk/efo/EFO_0000001,http://www.ebi.ac.uk/efo/EFO_0000002|European Ancestry|Height|EA|EFO_0000001/EFO_0000002         |</span>
<span class=sd>        |Schizophrenia    |http://www.ebi.ac.uk/efo/MONDO_0005090                                   |null             |Schizophrenia|no_pvalue_text|MONDO_0005090|</span>
<span class=sd>        +-----------------+-------------------------------------------------------------------------+-----------------+------------------------------------------+</span>
<span class=sd>        &lt;BLANKLINE&gt;</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=n>p_value_text</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>coalesce</span><span class=p>(</span>
            <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_normalise_pvaluetext</span><span class=p>(</span><span class=n>pvalue_text</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>array</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;no_pvalue_text&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=k>return</span> <span class=n>f</span><span class=o>.</span><span class=n>concat_ws</span><span class=p>(</span>
            <span class=s2>&quot;|&quot;</span><span class=p>,</span>
            <span class=n>association_trait</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>concat_ws</span><span class=p>(</span>
                <span class=s2>&quot;/&quot;</span><span class=p>,</span>
                <span class=n>p_value_text</span><span class=p>,</span>
            <span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>concat_ws</span><span class=p>(</span>
                <span class=s2>&quot;/&quot;</span><span class=p>,</span>
                <span class=n>parse_efos</span><span class=p>(</span><span class=n>mapped_trait_uri</span><span class=p>),</span>
            <span class=p>),</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_qc_all</span><span class=p>(</span>
        <span class=n>qc</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>chromosome</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>position</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>reference_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>alternate_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>strongest_snp_risk_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>p_value_mantissa</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>p_value_exponent</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>p_value_cutoff</span><span class=p>:</span> <span class=nb>float</span><span class=p>,</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Flag associations that fail any QC.</span>

<span class=sd>        Args:</span>
<span class=sd>            qc (Column): QC column</span>
<span class=sd>            chromosome (Column): Chromosome column</span>
<span class=sd>            position (Column): Position column</span>
<span class=sd>            reference_allele (Column): Reference allele column</span>
<span class=sd>            alternate_allele (Column): Alternate allele column</span>
<span class=sd>            strongest_snp_risk_allele (Column): Strongest SNP risk allele column</span>
<span class=sd>            p_value_mantissa (Column): P-value mantissa column</span>
<span class=sd>            p_value_exponent (Column): P-value exponent column</span>
<span class=sd>            p_value_cutoff (float): P-value cutoff</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: Updated QC column with flag.</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=n>qc</span> <span class=o>=</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_qc_variant_interactions</span><span class=p>(</span>
            <span class=n>qc</span><span class=p>,</span> <span class=n>strongest_snp_risk_allele</span>
        <span class=p>)</span>
        <span class=n>qc</span> <span class=o>=</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_qc_subsignificant_associations</span><span class=p>(</span>
            <span class=n>qc</span><span class=p>,</span> <span class=n>p_value_mantissa</span><span class=p>,</span> <span class=n>p_value_exponent</span><span class=p>,</span> <span class=n>p_value_cutoff</span>
        <span class=p>)</span>
        <span class=n>qc</span> <span class=o>=</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_qc_genomic_location</span><span class=p>(</span><span class=n>qc</span><span class=p>,</span> <span class=n>chromosome</span><span class=p>,</span> <span class=n>position</span><span class=p>)</span>
        <span class=n>qc</span> <span class=o>=</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_qc_variant_inconsistencies</span><span class=p>(</span>
            <span class=n>qc</span><span class=p>,</span> <span class=n>chromosome</span><span class=p>,</span> <span class=n>position</span><span class=p>,</span> <span class=n>strongest_snp_risk_allele</span>
        <span class=p>)</span>
        <span class=n>qc</span> <span class=o>=</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_qc_unmapped_variants</span><span class=p>(</span><span class=n>qc</span><span class=p>,</span> <span class=n>alternate_allele</span><span class=p>)</span>
        <span class=n>qc</span> <span class=o>=</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_qc_palindromic_alleles</span><span class=p>(</span>
            <span class=n>qc</span><span class=p>,</span> <span class=n>reference_allele</span><span class=p>,</span> <span class=n>alternate_allele</span>
        <span class=p>)</span>
        <span class=k>return</span> <span class=n>qc</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_qc_variant_interactions</span><span class=p>(</span>
        <span class=n>qc</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>strongest_snp_risk_allele</span><span class=p>:</span> <span class=n>Column</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Flag associations based on variant x variant interactions.</span>

<span class=sd>        Args:</span>
<span class=sd>            qc (Column): QC column</span>
<span class=sd>            strongest_snp_risk_allele (Column): Column with the strongest SNP risk allele</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: Updated QC column with flag.</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_update_quality_flag</span><span class=p>(</span>
            <span class=n>qc</span><span class=p>,</span>
            <span class=n>strongest_snp_risk_allele</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot;;&quot;</span><span class=p>),</span>
            <span class=n>StudyLocusQualityCheck</span><span class=o>.</span><span class=n>COMPOSITE_FLAG</span><span class=p>,</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_qc_subsignificant_associations</span><span class=p>(</span>
        <span class=n>qc</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>p_value_mantissa</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>p_value_exponent</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>pvalue_cutoff</span><span class=p>:</span> <span class=nb>float</span><span class=p>,</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Flag associations below significant threshold.</span>

<span class=sd>        Args:</span>
<span class=sd>            qc (Column): QC column</span>
<span class=sd>            p_value_mantissa (Column): P-value mantissa column</span>
<span class=sd>            p_value_exponent (Column): P-value exponent column</span>
<span class=sd>            pvalue_cutoff (float): association p-value cut-off</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: Updated QC column with flag.</span>

<span class=sd>        Examples:</span>
<span class=sd>            &gt;&gt;&gt; import pyspark.sql.types as t</span>
<span class=sd>            &gt;&gt;&gt; d = [{&#39;qc&#39;: None, &#39;p_value_mantissa&#39;: 1, &#39;p_value_exponent&#39;: -7}, {&#39;qc&#39;: None, &#39;p_value_mantissa&#39;: 1, &#39;p_value_exponent&#39;: -8}, {&#39;qc&#39;: None, &#39;p_value_mantissa&#39;: 5, &#39;p_value_exponent&#39;: -8}, {&#39;qc&#39;: None, &#39;p_value_mantissa&#39;: 1, &#39;p_value_exponent&#39;: -9}]</span>
<span class=sd>            &gt;&gt;&gt; df = spark.createDataFrame(d, t.StructType([t.StructField(&#39;qc&#39;, t.ArrayType(t.StringType()), True), t.StructField(&#39;p_value_mantissa&#39;, t.IntegerType()), t.StructField(&#39;p_value_exponent&#39;, t.IntegerType())]))</span>
<span class=sd>            &gt;&gt;&gt; df.withColumn(&#39;qc&#39;, GWASCatalogAssociations._qc_subsignificant_associations(f.col(&quot;qc&quot;), f.col(&quot;p_value_mantissa&quot;), f.col(&quot;p_value_exponent&quot;), 5e-8)).show(truncate = False)</span>
<span class=sd>            +------------------------+----------------+----------------+</span>
<span class=sd>            |qc                      |p_value_mantissa|p_value_exponent|</span>
<span class=sd>            +------------------------+----------------+----------------+</span>
<span class=sd>            |[Subsignificant p-value]|1               |-7              |</span>
<span class=sd>            |[]                      |1               |-8              |</span>
<span class=sd>            |[]                      |5               |-8              |</span>
<span class=sd>            |[]                      |1               |-9              |</span>
<span class=sd>            +------------------------+----------------+----------------+</span>
<span class=sd>            &lt;BLANKLINE&gt;</span>

<span class=sd>        &quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=n>StudyLocus</span><span class=o>.</span><span class=n>_update_quality_flag</span><span class=p>(</span>
            <span class=n>qc</span><span class=p>,</span>
            <span class=n>calculate_neglog_pvalue</span><span class=p>(</span><span class=n>p_value_mantissa</span><span class=p>,</span> <span class=n>p_value_exponent</span><span class=p>)</span>
            <span class=o>&lt;</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=o>-</span><span class=n>np</span><span class=o>.</span><span class=n>log10</span><span class=p>(</span><span class=n>pvalue_cutoff</span><span class=p>)),</span>
            <span class=n>StudyLocusQualityCheck</span><span class=o>.</span><span class=n>SUBSIGNIFICANT_FLAG</span><span class=p>,</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_qc_genomic_location</span><span class=p>(</span>
        <span class=n>qc</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>chromosome</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>position</span><span class=p>:</span> <span class=n>Column</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Flag associations without genomic location in GWAS Catalog.</span>

<span class=sd>        Args:</span>
<span class=sd>            qc (Column): QC column</span>
<span class=sd>            chromosome (Column): Chromosome column in GWAS Catalog</span>
<span class=sd>            position (Column): Position column in GWAS Catalog</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: Updated QC column with flag.</span>

<span class=sd>        Examples:</span>
<span class=sd>            &gt;&gt;&gt; import pyspark.sql.types as t</span>
<span class=sd>            &gt;&gt;&gt; d = [{&#39;qc&#39;: None, &#39;chromosome&#39;: None, &#39;position&#39;: None}, {&#39;qc&#39;: None, &#39;chromosome&#39;: &#39;1&#39;, &#39;position&#39;: None}, {&#39;qc&#39;: None, &#39;chromosome&#39;: None, &#39;position&#39;: 1}, {&#39;qc&#39;: None, &#39;chromosome&#39;: &#39;1&#39;, &#39;position&#39;: 1}]</span>
<span class=sd>            &gt;&gt;&gt; df = spark.createDataFrame(d, schema=t.StructType([t.StructField(&#39;qc&#39;, t.ArrayType(t.StringType()), True), t.StructField(&#39;chromosome&#39;, t.StringType()), t.StructField(&#39;position&#39;, t.IntegerType())]))</span>
<span class=sd>            &gt;&gt;&gt; df.withColumn(&#39;qc&#39;, GWASCatalogAssociations._qc_genomic_location(df.qc, df.chromosome, df.position)).show(truncate=False)</span>
<span class=sd>            +----------------------------+----------+--------+</span>
<span class=sd>            |qc                          |chromosome|position|</span>
<span class=sd>            +----------------------------+----------+--------+</span>
<span class=sd>            |[Incomplete genomic mapping]|null      |null    |</span>
<span class=sd>            |[Incomplete genomic mapping]|1         |null    |</span>
<span class=sd>            |[Incomplete genomic mapping]|null      |1       |</span>
<span class=sd>            |[]                          |1         |1       |</span>
<span class=sd>            +----------------------------+----------+--------+</span>
<span class=sd>            &lt;BLANKLINE&gt;</span>

<span class=sd>        &quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=n>StudyLocus</span><span class=o>.</span><span class=n>_update_quality_flag</span><span class=p>(</span>
            <span class=n>qc</span><span class=p>,</span>
            <span class=n>position</span><span class=o>.</span><span class=n>isNull</span><span class=p>()</span> <span class=o>|</span> <span class=n>chromosome</span><span class=o>.</span><span class=n>isNull</span><span class=p>(),</span>
            <span class=n>StudyLocusQualityCheck</span><span class=o>.</span><span class=n>NO_GENOMIC_LOCATION_FLAG</span><span class=p>,</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_qc_variant_inconsistencies</span><span class=p>(</span>
        <span class=n>qc</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>chromosome</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>position</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>strongest_snp_risk_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Flag associations with inconsistencies in the variant annotation.</span>

<span class=sd>        Args:</span>
<span class=sd>            qc (Column): QC column</span>
<span class=sd>            chromosome (Column): Chromosome column in GWAS Catalog</span>
<span class=sd>            position (Column): Position column in GWAS Catalog</span>
<span class=sd>            strongest_snp_risk_allele (Column): Strongest SNP risk allele column in GWAS Catalog</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: Updated QC column with flag.</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_update_quality_flag</span><span class=p>(</span>
            <span class=n>qc</span><span class=p>,</span>
            <span class=c1># Number of chromosomes does not correspond to the number of positions:</span>
            <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>size</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>chromosome</span><span class=p>,</span> <span class=s2>&quot;;&quot;</span><span class=p>))</span> <span class=o>!=</span> <span class=n>f</span><span class=o>.</span><span class=n>size</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>position</span><span class=p>,</span> <span class=s2>&quot;;&quot;</span><span class=p>)))</span>
            <span class=c1># Number of chromosome values different from riskAllele values:</span>
            <span class=o>|</span> <span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>size</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>chromosome</span><span class=p>,</span> <span class=s2>&quot;;&quot;</span><span class=p>))</span>
                <span class=o>!=</span> <span class=n>f</span><span class=o>.</span><span class=n>size</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>strongest_snp_risk_allele</span><span class=p>,</span> <span class=s2>&quot;;&quot;</span><span class=p>))</span>
            <span class=p>),</span>
            <span class=n>StudyLocusQualityCheck</span><span class=o>.</span><span class=n>INCONSISTENCY_FLAG</span><span class=p>,</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_qc_unmapped_variants</span><span class=p>(</span><span class=n>qc</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>alternate_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Flag associations with variants not mapped to variantAnnotation.</span>

<span class=sd>        Args:</span>
<span class=sd>            qc (Column): QC column</span>
<span class=sd>            alternate_allele (Column): alternate allele</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: Updated QC column with flag.</span>

<span class=sd>        Example:</span>
<span class=sd>            &gt;&gt;&gt; import pyspark.sql.types as t</span>
<span class=sd>            &gt;&gt;&gt; d = [{&#39;alternate_allele&#39;: &#39;A&#39;, &#39;qc&#39;: None}, {&#39;alternate_allele&#39;: None, &#39;qc&#39;: None}]</span>
<span class=sd>            &gt;&gt;&gt; schema = t.StructType([t.StructField(&#39;alternate_allele&#39;, t.StringType(), True), t.StructField(&#39;qc&#39;, t.ArrayType(t.StringType()), True)])</span>
<span class=sd>            &gt;&gt;&gt; df = spark.createDataFrame(data=d, schema=schema)</span>
<span class=sd>            &gt;&gt;&gt; df.withColumn(&quot;new_qc&quot;, GWASCatalogAssociations._qc_unmapped_variants(f.col(&quot;qc&quot;), f.col(&quot;alternate_allele&quot;))).show()</span>
<span class=sd>            +----------------+----+--------------------+</span>
<span class=sd>            |alternate_allele|  qc|              new_qc|</span>
<span class=sd>            +----------------+----+--------------------+</span>
<span class=sd>            |               A|null|                  []|</span>
<span class=sd>            |            null|null|[No mapping in Gn...|</span>
<span class=sd>            +----------------+----+--------------------+</span>
<span class=sd>            &lt;BLANKLINE&gt;</span>

<span class=sd>        &quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_update_quality_flag</span><span class=p>(</span>
            <span class=n>qc</span><span class=p>,</span>
            <span class=n>alternate_allele</span><span class=o>.</span><span class=n>isNull</span><span class=p>(),</span>
            <span class=n>StudyLocusQualityCheck</span><span class=o>.</span><span class=n>NON_MAPPED_VARIANT_FLAG</span><span class=p>,</span>
        <span class=p>)</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_qc_palindromic_alleles</span><span class=p>(</span>
        <span class=n>qc</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>reference_allele</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span> <span class=n>alternate_allele</span><span class=p>:</span> <span class=n>Column</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Flag associations with palindromic variants which effects can not be harmonised.</span>

<span class=sd>        Args:</span>
<span class=sd>            qc (Column): QC column</span>
<span class=sd>            reference_allele (Column): reference allele</span>
<span class=sd>            alternate_allele (Column): alternate allele</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: Updated QC column with flag.</span>

<span class=sd>        Example:</span>
<span class=sd>            &gt;&gt;&gt; import pyspark.sql.types as t</span>
<span class=sd>            &gt;&gt;&gt; schema = t.StructType([t.StructField(&#39;reference_allele&#39;, t.StringType(), True), t.StructField(&#39;alternate_allele&#39;, t.StringType(), True), t.StructField(&#39;qc&#39;, t.ArrayType(t.StringType()), True)])</span>
<span class=sd>            &gt;&gt;&gt; d = [{&#39;reference_allele&#39;: &#39;A&#39;, &#39;alternate_allele&#39;: &#39;T&#39;, &#39;qc&#39;: None}, {&#39;reference_allele&#39;: &#39;AT&#39;, &#39;alternate_allele&#39;: &#39;TA&#39;, &#39;qc&#39;: None}, {&#39;reference_allele&#39;: &#39;AT&#39;, &#39;alternate_allele&#39;: &#39;AT&#39;, &#39;qc&#39;: None}]</span>
<span class=sd>            &gt;&gt;&gt; df = spark.createDataFrame(data=d, schema=schema)</span>
<span class=sd>            &gt;&gt;&gt; df.withColumn(&quot;qc&quot;, GWASCatalogAssociations._qc_palindromic_alleles(f.col(&quot;qc&quot;), f.col(&quot;reference_allele&quot;), f.col(&quot;alternate_allele&quot;))).show(truncate=False)</span>
<span class=sd>            +----------------+----------------+---------------------------------------+</span>
<span class=sd>            |reference_allele|alternate_allele|qc                                     |</span>
<span class=sd>            +----------------+----------------+---------------------------------------+</span>
<span class=sd>            |A               |T               |[Palindrome alleles - cannot harmonize]|</span>
<span class=sd>            |AT              |TA              |[]                                     |</span>
<span class=sd>            |AT              |AT              |[Palindrome alleles - cannot harmonize]|</span>
<span class=sd>            +----------------+----------------+---------------------------------------+</span>
<span class=sd>            &lt;BLANKLINE&gt;</span>

<span class=sd>        &quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=n>StudyLocus</span><span class=o>.</span><span class=n>_update_quality_flag</span><span class=p>(</span>
            <span class=n>qc</span><span class=p>,</span>
            <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_are_alleles_palindromic</span><span class=p>(</span>
                <span class=n>reference_allele</span><span class=p>,</span> <span class=n>alternate_allele</span>
            <span class=p>),</span>
            <span class=n>StudyLocusQualityCheck</span><span class=o>.</span><span class=n>PALINDROMIC_ALLELE_FLAG</span><span class=p>,</span>
        <span class=p>)</span>

    <span class=nd>@classmethod</span>
    <span class=k>def</span> <span class=nf>from_source</span><span class=p>(</span>
        <span class=bp>cls</span><span class=p>:</span> <span class=nb>type</span><span class=p>[</span><span class=n>GWASCatalogAssociations</span><span class=p>],</span>
        <span class=n>gwas_associations</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
        <span class=n>variant_annotation</span><span class=p>:</span> <span class=n>VariantAnnotation</span><span class=p>,</span>
        <span class=n>pvalue_threshold</span><span class=p>:</span> <span class=nb>float</span> <span class=o>=</span> <span class=mf>5e-8</span><span class=p>,</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>GWASCatalogAssociations</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Read GWASCatalog associations.</span>

<span class=sd>        It reads the GWAS Catalog association dataset, selects and renames columns, casts columns, and</span>
<span class=sd>        applies some pre-defined filters on the data:</span>

<span class=sd>        Args:</span>
<span class=sd>            gwas_associations (DataFrame): GWAS Catalog raw associations dataset</span>
<span class=sd>            variant_annotation (VariantAnnotation): Variant annotation dataset</span>
<span class=sd>            pvalue_threshold (float): P-value threshold for flagging associations</span>

<span class=sd>        Returns:</span>
<span class=sd>            GWASCatalogAssociations: GWASCatalogAssociations dataset</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=n>GWASCatalogAssociations</span><span class=p>(</span>
            <span class=n>_df</span><span class=o>=</span><span class=n>gwas_associations</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
                <span class=s2>&quot;studyLocusId&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>monotonically_increasing_id</span><span class=p>()</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>LongType</span><span class=p>())</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>transform</span><span class=p>(</span>
                <span class=c1># Map/harmonise variants to variant annotation dataset:</span>
                <span class=c1># This function adds columns: variantId, referenceAllele, alternateAllele, chromosome, position</span>
                <span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_map_to_variant_annotation_variants</span><span class=p>(</span>
                    <span class=n>df</span><span class=p>,</span> <span class=n>variant_annotation</span>
                <span class=p>)</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
                <span class=c1># Perform all quality control checks:</span>
                <span class=s2>&quot;qualityControls&quot;</span><span class=p>,</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_qc_all</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>array</span><span class=p>()</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;qualityControls&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;CHR_ID&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;CHR_POS&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>IntegerType</span><span class=p>()),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>),</span>
                    <span class=o>*</span><span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_parse_pvalue</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;P-VALUE&quot;</span><span class=p>)),</span>
                    <span class=n>pvalue_threshold</span><span class=p>,</span>
                <span class=p>),</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>select</span><span class=p>(</span>
                <span class=c1># INSIDE STUDY-LOCUS SCHEMA:</span>
                <span class=s2>&quot;studyLocusId&quot;</span><span class=p>,</span>
                <span class=s2>&quot;variantId&quot;</span><span class=p>,</span>
                <span class=c1># Mapped genomic location of the variant (; separated list)</span>
                <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
                <span class=s2>&quot;position&quot;</span><span class=p>,</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STUDY ACCESSION&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;studyId&quot;</span><span class=p>),</span>
                <span class=c1># beta value of the association</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_harmonise_beta</span><span class=p>(</span>
                    <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_normalise_risk_allele</span><span class=p>(</span>
                        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>)</span>
                    <span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;OR or BETA&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;95% CI (TEXT)&quot;</span><span class=p>),</span>
                <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;beta&quot;</span><span class=p>),</span>
                <span class=c1># odds ratio of the association</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_harmonise_odds_ratio</span><span class=p>(</span>
                    <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_normalise_risk_allele</span><span class=p>(</span>
                        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>)</span>
                    <span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;OR or BETA&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;95% CI (TEXT)&quot;</span><span class=p>),</span>
                <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;oddsRatio&quot;</span><span class=p>),</span>
                <span class=c1># CI lower of the beta value</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_harmonise_beta_ci</span><span class=p>(</span>
                    <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_normalise_risk_allele</span><span class=p>(</span>
                        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>)</span>
                    <span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;OR or BETA&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;95% CI (TEXT)&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;P-VALUE&quot;</span><span class=p>),</span>
                    <span class=s2>&quot;lower&quot;</span><span class=p>,</span>
                <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;betaConfidenceIntervalLower&quot;</span><span class=p>),</span>
                <span class=c1># CI upper for the beta value</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_harmonise_beta_ci</span><span class=p>(</span>
                    <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_normalise_risk_allele</span><span class=p>(</span>
                        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>)</span>
                    <span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;OR or BETA&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;95% CI (TEXT)&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;P-VALUE&quot;</span><span class=p>),</span>
                    <span class=s2>&quot;upper&quot;</span><span class=p>,</span>
                <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;betaConfidenceIntervalUpper&quot;</span><span class=p>),</span>
                <span class=c1># CI lower of the odds ratio value</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_harmonise_odds_ratio_ci</span><span class=p>(</span>
                    <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_normalise_risk_allele</span><span class=p>(</span>
                        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>)</span>
                    <span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;OR or BETA&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;95% CI (TEXT)&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;P-VALUE&quot;</span><span class=p>),</span>
                    <span class=s2>&quot;lower&quot;</span><span class=p>,</span>
                <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;oddsRatioConfidenceIntervalLower&quot;</span><span class=p>),</span>
                <span class=c1># CI upper of the odds ratio value</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_harmonise_odds_ratio_ci</span><span class=p>(</span>
                    <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_normalise_risk_allele</span><span class=p>(</span>
                        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>)</span>
                    <span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;OR or BETA&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;95% CI (TEXT)&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;P-VALUE&quot;</span><span class=p>),</span>
                    <span class=s2>&quot;upper&quot;</span><span class=p>,</span>
                <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;oddsRatioConfidenceIntervalUpper&quot;</span><span class=p>),</span>
                <span class=c1># p-value of the association, string: split into exponent and mantissa.</span>
                <span class=o>*</span><span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_parse_pvalue</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;P-VALUE&quot;</span><span class=p>)),</span>
                <span class=c1># Capturing phenotype granularity at the association level</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_concatenate_substudy_description</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;DISEASE/TRAIT&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;P-VALUE (TEXT)&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;MAPPED_TRAIT_URI&quot;</span><span class=p>),</span>
                <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;subStudyDescription&quot;</span><span class=p>),</span>
                <span class=c1># Quality controls (array of strings)</span>
                <span class=s2>&quot;qualityControls&quot;</span><span class=p>,</span>
            <span class=p>),</span>
            <span class=n>_schema</span><span class=o>=</span><span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>get_schema</span><span class=p>(),</span>
        <span class=p>)</span>

    <span class=k>def</span> <span class=nf>update_study_id</span><span class=p>(</span>
        <span class=bp>self</span><span class=p>:</span> <span class=n>GWASCatalogAssociations</span><span class=p>,</span> <span class=n>study_annotation</span><span class=p>:</span> <span class=n>DataFrame</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>GWASCatalogAssociations</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Update final studyId and studyLocusId with a dataframe containing study annotation.</span>

<span class=sd>        Args:</span>
<span class=sd>            study_annotation (DataFrame): Dataframe containing `updatedStudyId` and key columns `studyId` and `subStudyDescription`.</span>

<span class=sd>        Returns:</span>
<span class=sd>            GWASCatalogAssociations: Updated study locus with new `studyId` and `studyLocusId`.</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>df</span> <span class=o>=</span> <span class=p>(</span>
            <span class=bp>self</span><span class=o>.</span><span class=n>_df</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
                <span class=n>study_annotation</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=s2>&quot;studyId&quot;</span><span class=p>,</span> <span class=s2>&quot;subStudyDescription&quot;</span><span class=p>],</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;studyId&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>coalesce</span><span class=p>(</span><span class=s2>&quot;updatedStudyId&quot;</span><span class=p>,</span> <span class=s2>&quot;studyId&quot;</span><span class=p>))</span>
            <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;subStudyDescription&quot;</span><span class=p>,</span> <span class=s2>&quot;updatedStudyId&quot;</span><span class=p>)</span>
        <span class=p>)</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;studyLocusId&quot;</span><span class=p>,</span>
            <span class=n>StudyLocus</span><span class=o>.</span><span class=n>assign_study_locus_id</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;studyId&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;variantId&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=k>return</span> <span class=bp>self</span>

    <span class=k>def</span> <span class=nf>_qc_ambiguous_study</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>GWASCatalogAssociations</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>GWASCatalogAssociations</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Flag associations with variants that can not be unambiguously associated with one study.</span>

<span class=sd>        Returns:</span>
<span class=sd>            GWASCatalogAssociations: Updated study locus.</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=n>assoc_ambiguity_window</span> <span class=o>=</span> <span class=n>Window</span><span class=o>.</span><span class=n>partitionBy</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;studyId&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;variantId&quot;</span><span class=p>)</span>
        <span class=p>)</span>

        <span class=bp>self</span><span class=o>.</span><span class=n>_df</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;qualityControls&quot;</span><span class=p>,</span>
            <span class=n>StudyLocus</span><span class=o>.</span><span class=n>_update_quality_flag</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;qualityControls&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>count</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;variantId&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>over</span><span class=p>(</span><span class=n>assoc_ambiguity_window</span><span class=p>)</span> <span class=o>&gt;</span> <span class=mi>1</span><span class=p>,</span>
                <span class=n>StudyLocusQualityCheck</span><span class=o>.</span><span class=n>AMBIGUOUS_STUDY</span><span class=p>,</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=k>return</span> <span class=bp>self</span>
</code></pre></div></td></tr></table></div> </details> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h3 id=otg.datasource.gwas_catalog.associations.GWASCatalogAssociations.from_source class="doc doc-heading"> <code class="highlight language-python"><span class=n>from_source</span><span class=p>(</span><span class=n>gwas_associations</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span> <span class=n>variant_annotation</span><span class=p>:</span> <span class=n>VariantAnnotation</span><span class=p>,</span> <span class=n>pvalue_threshold</span><span class=p>:</span> <span class=nb>float</span> <span class=o>=</span> <span class=mf>5e-08</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>GWASCatalogAssociations</span></code> <span class="doc doc-labels"> <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small> </span> <a href=#otg.datasource.gwas_catalog.associations.GWASCatalogAssociations.from_source class=headerlink title="Permanent link">&para;</a></h3> <div class="doc doc-contents "> <p>Read GWASCatalog associations.</p> <p>It reads the GWAS Catalog association dataset, selects and renames columns, casts columns, and applies some pre-defined filters on the data:</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>gwas_associations</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td> <div class=doc-md-description> <p>GWAS Catalog raw associations dataset</p> </div> </td> <td> <em>required</em> </td> </tr> <tr> <td><code>variant_annotation</code></td> <td> <code><a class="autorefs autorefs-internal" title="otg.dataset.variant_annotation.VariantAnnotation" href="../../../dataset/variant_annotation/#otg.dataset.variant_annotation.VariantAnnotation">VariantAnnotation</a></code> </td> <td> <div class=doc-md-description> <p>Variant annotation dataset</p> </div> </td> <td> <em>required</em> </td> </tr> <tr> <td><code>pvalue_threshold</code></td> <td> <code>float</code> </td> <td> <div class=doc-md-description> <p>P-value threshold for flagging associations</p> </div> </td> <td> <code>5e-08</code> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>GWASCatalogAssociations</code></td> <td> <code><a class="autorefs autorefs-internal" title="otg.datasource.gwas_catalog.associations.GWASCatalogAssociations" href="#otg.datasource.gwas_catalog.associations.GWASCatalogAssociations">GWASCatalogAssociations</a></code> </td> <td> <div class=doc-md-description> <p>GWASCatalogAssociations dataset</p> </div> </td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>src/otg/datasource/gwas_catalog/associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 982</span>
<span class=normal> 983</span>
<span class=normal> 984</span>
<span class=normal> 985</span>
<span class=normal> 986</span>
<span class=normal> 987</span>
<span class=normal> 988</span>
<span class=normal> 989</span>
<span class=normal> 990</span>
<span class=normal> 991</span>
<span class=normal> 992</span>
<span class=normal> 993</span>
<span class=normal> 994</span>
<span class=normal> 995</span>
<span class=normal> 996</span>
<span class=normal> 997</span>
<span class=normal> 998</span>
<span class=normal> 999</span>
<span class=normal>1000</span>
<span class=normal>1001</span>
<span class=normal>1002</span>
<span class=normal>1003</span>
<span class=normal>1004</span>
<span class=normal>1005</span>
<span class=normal>1006</span>
<span class=normal>1007</span>
<span class=normal>1008</span>
<span class=normal>1009</span>
<span class=normal>1010</span>
<span class=normal>1011</span>
<span class=normal>1012</span>
<span class=normal>1013</span>
<span class=normal>1014</span>
<span class=normal>1015</span>
<span class=normal>1016</span>
<span class=normal>1017</span>
<span class=normal>1018</span>
<span class=normal>1019</span>
<span class=normal>1020</span>
<span class=normal>1021</span>
<span class=normal>1022</span>
<span class=normal>1023</span>
<span class=normal>1024</span>
<span class=normal>1025</span>
<span class=normal>1026</span>
<span class=normal>1027</span>
<span class=normal>1028</span>
<span class=normal>1029</span>
<span class=normal>1030</span>
<span class=normal>1031</span>
<span class=normal>1032</span>
<span class=normal>1033</span>
<span class=normal>1034</span>
<span class=normal>1035</span>
<span class=normal>1036</span>
<span class=normal>1037</span>
<span class=normal>1038</span>
<span class=normal>1039</span>
<span class=normal>1040</span>
<span class=normal>1041</span>
<span class=normal>1042</span>
<span class=normal>1043</span>
<span class=normal>1044</span>
<span class=normal>1045</span>
<span class=normal>1046</span>
<span class=normal>1047</span>
<span class=normal>1048</span>
<span class=normal>1049</span>
<span class=normal>1050</span>
<span class=normal>1051</span>
<span class=normal>1052</span>
<span class=normal>1053</span>
<span class=normal>1054</span>
<span class=normal>1055</span>
<span class=normal>1056</span>
<span class=normal>1057</span>
<span class=normal>1058</span>
<span class=normal>1059</span>
<span class=normal>1060</span>
<span class=normal>1061</span>
<span class=normal>1062</span>
<span class=normal>1063</span>
<span class=normal>1064</span>
<span class=normal>1065</span>
<span class=normal>1066</span>
<span class=normal>1067</span>
<span class=normal>1068</span>
<span class=normal>1069</span>
<span class=normal>1070</span>
<span class=normal>1071</span>
<span class=normal>1072</span>
<span class=normal>1073</span>
<span class=normal>1074</span>
<span class=normal>1075</span>
<span class=normal>1076</span>
<span class=normal>1077</span>
<span class=normal>1078</span>
<span class=normal>1079</span>
<span class=normal>1080</span>
<span class=normal>1081</span>
<span class=normal>1082</span>
<span class=normal>1083</span>
<span class=normal>1084</span>
<span class=normal>1085</span>
<span class=normal>1086</span>
<span class=normal>1087</span>
<span class=normal>1088</span>
<span class=normal>1089</span>
<span class=normal>1090</span>
<span class=normal>1091</span>
<span class=normal>1092</span>
<span class=normal>1093</span>
<span class=normal>1094</span>
<span class=normal>1095</span>
<span class=normal>1096</span>
<span class=normal>1097</span>
<span class=normal>1098</span>
<span class=normal>1099</span>
<span class=normal>1100</span>
<span class=normal>1101</span>
<span class=normal>1102</span>
<span class=normal>1103</span>
<span class=normal>1104</span>
<span class=normal>1105</span>
<span class=normal>1106</span>
<span class=normal>1107</span>
<span class=normal>1108</span>
<span class=normal>1109</span>
<span class=normal>1110</span>
<span class=normal>1111</span>
<span class=normal>1112</span>
<span class=normal>1113</span>
<span class=normal>1114</span>
<span class=normal>1115</span></pre></div></td><td class=code><div><pre><span></span><code><span class=nd>@classmethod</span>
<span class=k>def</span> <span class=nf>from_source</span><span class=p>(</span>
    <span class=bp>cls</span><span class=p>:</span> <span class=nb>type</span><span class=p>[</span><span class=n>GWASCatalogAssociations</span><span class=p>],</span>
    <span class=n>gwas_associations</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
    <span class=n>variant_annotation</span><span class=p>:</span> <span class=n>VariantAnnotation</span><span class=p>,</span>
    <span class=n>pvalue_threshold</span><span class=p>:</span> <span class=nb>float</span> <span class=o>=</span> <span class=mf>5e-8</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>GWASCatalogAssociations</span><span class=p>:</span>
<span class=w>    </span><span class=sd>&quot;&quot;&quot;Read GWASCatalog associations.</span>

<span class=sd>    It reads the GWAS Catalog association dataset, selects and renames columns, casts columns, and</span>
<span class=sd>    applies some pre-defined filters on the data:</span>

<span class=sd>    Args:</span>
<span class=sd>        gwas_associations (DataFrame): GWAS Catalog raw associations dataset</span>
<span class=sd>        variant_annotation (VariantAnnotation): Variant annotation dataset</span>
<span class=sd>        pvalue_threshold (float): P-value threshold for flagging associations</span>

<span class=sd>    Returns:</span>
<span class=sd>        GWASCatalogAssociations: GWASCatalogAssociations dataset</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=n>GWASCatalogAssociations</span><span class=p>(</span>
        <span class=n>_df</span><span class=o>=</span><span class=n>gwas_associations</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;studyLocusId&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>monotonically_increasing_id</span><span class=p>()</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>LongType</span><span class=p>())</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span>
            <span class=c1># Map/harmonise variants to variant annotation dataset:</span>
            <span class=c1># This function adds columns: variantId, referenceAllele, alternateAllele, chromosome, position</span>
            <span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_map_to_variant_annotation_variants</span><span class=p>(</span>
                <span class=n>df</span><span class=p>,</span> <span class=n>variant_annotation</span>
            <span class=p>)</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=c1># Perform all quality control checks:</span>
            <span class=s2>&quot;qualityControls&quot;</span><span class=p>,</span>
            <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_qc_all</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>array</span><span class=p>()</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;qualityControls&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;CHR_ID&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;CHR_POS&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>IntegerType</span><span class=p>()),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>),</span>
                <span class=o>*</span><span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_parse_pvalue</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;P-VALUE&quot;</span><span class=p>)),</span>
                <span class=n>pvalue_threshold</span><span class=p>,</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=c1># INSIDE STUDY-LOCUS SCHEMA:</span>
            <span class=s2>&quot;studyLocusId&quot;</span><span class=p>,</span>
            <span class=s2>&quot;variantId&quot;</span><span class=p>,</span>
            <span class=c1># Mapped genomic location of the variant (; separated list)</span>
            <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
            <span class=s2>&quot;position&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STUDY ACCESSION&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;studyId&quot;</span><span class=p>),</span>
            <span class=c1># beta value of the association</span>
            <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_harmonise_beta</span><span class=p>(</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_normalise_risk_allele</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>)</span>
                <span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;OR or BETA&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;95% CI (TEXT)&quot;</span><span class=p>),</span>
            <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;beta&quot;</span><span class=p>),</span>
            <span class=c1># odds ratio of the association</span>
            <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_harmonise_odds_ratio</span><span class=p>(</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_normalise_risk_allele</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>)</span>
                <span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;OR or BETA&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;95% CI (TEXT)&quot;</span><span class=p>),</span>
            <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;oddsRatio&quot;</span><span class=p>),</span>
            <span class=c1># CI lower of the beta value</span>
            <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_harmonise_beta_ci</span><span class=p>(</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_normalise_risk_allele</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>)</span>
                <span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;OR or BETA&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;95% CI (TEXT)&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;P-VALUE&quot;</span><span class=p>),</span>
                <span class=s2>&quot;lower&quot;</span><span class=p>,</span>
            <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;betaConfidenceIntervalLower&quot;</span><span class=p>),</span>
            <span class=c1># CI upper for the beta value</span>
            <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_harmonise_beta_ci</span><span class=p>(</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_normalise_risk_allele</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>)</span>
                <span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;OR or BETA&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;95% CI (TEXT)&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;P-VALUE&quot;</span><span class=p>),</span>
                <span class=s2>&quot;upper&quot;</span><span class=p>,</span>
            <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;betaConfidenceIntervalUpper&quot;</span><span class=p>),</span>
            <span class=c1># CI lower of the odds ratio value</span>
            <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_harmonise_odds_ratio_ci</span><span class=p>(</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_normalise_risk_allele</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>)</span>
                <span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;OR or BETA&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;95% CI (TEXT)&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;P-VALUE&quot;</span><span class=p>),</span>
                <span class=s2>&quot;lower&quot;</span><span class=p>,</span>
            <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;oddsRatioConfidenceIntervalLower&quot;</span><span class=p>),</span>
            <span class=c1># CI upper of the odds ratio value</span>
            <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_harmonise_odds_ratio_ci</span><span class=p>(</span>
                <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_normalise_risk_allele</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;STRONGEST SNP-RISK ALLELE&quot;</span><span class=p>)</span>
                <span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;OR or BETA&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;95% CI (TEXT)&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;P-VALUE&quot;</span><span class=p>),</span>
                <span class=s2>&quot;upper&quot;</span><span class=p>,</span>
            <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;oddsRatioConfidenceIntervalUpper&quot;</span><span class=p>),</span>
            <span class=c1># p-value of the association, string: split into exponent and mantissa.</span>
            <span class=o>*</span><span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_parse_pvalue</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;P-VALUE&quot;</span><span class=p>)),</span>
            <span class=c1># Capturing phenotype granularity at the association level</span>
            <span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>_concatenate_substudy_description</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;DISEASE/TRAIT&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;P-VALUE (TEXT)&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;MAPPED_TRAIT_URI&quot;</span><span class=p>),</span>
            <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;subStudyDescription&quot;</span><span class=p>),</span>
            <span class=c1># Quality controls (array of strings)</span>
            <span class=s2>&quot;qualityControls&quot;</span><span class=p>,</span>
        <span class=p>),</span>
        <span class=n>_schema</span><span class=o>=</span><span class=n>GWASCatalogAssociations</span><span class=o>.</span><span class=n>get_schema</span><span class=p>(),</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h3 id=otg.datasource.gwas_catalog.associations.GWASCatalogAssociations.update_study_id class="doc doc-heading"> <code class="highlight language-python"><span class=n>update_study_id</span><span class=p>(</span><span class=n>study_annotation</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>GWASCatalogAssociations</span></code> <a href=#otg.datasource.gwas_catalog.associations.GWASCatalogAssociations.update_study_id class=headerlink title="Permanent link">&para;</a></h3> <div class="doc doc-contents "> <p>Update final studyId and studyLocusId with a dataframe containing study annotation.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>study_annotation</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td> <div class=doc-md-description> <p>Dataframe containing <code>updatedStudyId</code> and key columns <code>studyId</code> and <code>subStudyDescription</code>.</p> </div> </td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>GWASCatalogAssociations</code></td> <td> <code><a class="autorefs autorefs-internal" title="otg.datasource.gwas_catalog.associations.GWASCatalogAssociations" href="#otg.datasource.gwas_catalog.associations.GWASCatalogAssociations">GWASCatalogAssociations</a></code> </td> <td> <div class=doc-md-description> <p>Updated study locus with new <code>studyId</code> and <code>studyLocusId</code>.</p> </div> </td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>src/otg/datasource/gwas_catalog/associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>1117</span>
<span class=normal>1118</span>
<span class=normal>1119</span>
<span class=normal>1120</span>
<span class=normal>1121</span>
<span class=normal>1122</span>
<span class=normal>1123</span>
<span class=normal>1124</span>
<span class=normal>1125</span>
<span class=normal>1126</span>
<span class=normal>1127</span>
<span class=normal>1128</span>
<span class=normal>1129</span>
<span class=normal>1130</span>
<span class=normal>1131</span>
<span class=normal>1132</span>
<span class=normal>1133</span>
<span class=normal>1134</span>
<span class=normal>1135</span>
<span class=normal>1136</span>
<span class=normal>1137</span>
<span class=normal>1138</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>update_study_id</span><span class=p>(</span>
    <span class=bp>self</span><span class=p>:</span> <span class=n>GWASCatalogAssociations</span><span class=p>,</span> <span class=n>study_annotation</span><span class=p>:</span> <span class=n>DataFrame</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>GWASCatalogAssociations</span><span class=p>:</span>
<span class=w>    </span><span class=sd>&quot;&quot;&quot;Update final studyId and studyLocusId with a dataframe containing study annotation.</span>

<span class=sd>    Args:</span>
<span class=sd>        study_annotation (DataFrame): Dataframe containing `updatedStudyId` and key columns `studyId` and `subStudyDescription`.</span>

<span class=sd>    Returns:</span>
<span class=sd>        GWASCatalogAssociations: Updated study locus with new `studyId` and `studyLocusId`.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>df</span> <span class=o>=</span> <span class=p>(</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>_df</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>study_annotation</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=s2>&quot;studyId&quot;</span><span class=p>,</span> <span class=s2>&quot;subStudyDescription&quot;</span><span class=p>],</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;studyId&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>coalesce</span><span class=p>(</span><span class=s2>&quot;updatedStudyId&quot;</span><span class=p>,</span> <span class=s2>&quot;studyId&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;subStudyDescription&quot;</span><span class=p>,</span> <span class=s2>&quot;updatedStudyId&quot;</span><span class=p>)</span>
    <span class=p>)</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
        <span class=s2>&quot;studyLocusId&quot;</span><span class=p>,</span>
        <span class=n>StudyLocus</span><span class=o>.</span><span class=n>assign_study_locus_id</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;studyId&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;variantId&quot;</span><span class=p>)),</span>
    <span class=p>)</span>
    <span class=k>return</span> <span class=bp>self</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> <hr> <div class=md-source-file> <small> <ul class="metadata page-metadata" data-bi-name="page info" lang=en-us dir=ltr> <!-- mkdocs-git-revision-date-plugin --> <li> <span class="icon twemoji"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3h-2Z"/></svg> </span> <span class=label> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class=timeago datetime=2023-09-25T14:33:18+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2023-09-25</span> </span> </li> <li> <span class="icon twemoji"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg> </span> <span class=label> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class=timeago datetime=2023-10-31T12:33:26+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2023-10-31</span> </span> </li> <li> <span class="icon twemoji"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 2A10 10 0 0 0 2 12c0 4.42 2.87 8.17 6.84 9.5.5.08.66-.23.66-.5v-1.69c-2.77.6-3.36-1.34-3.36-1.34-.46-1.16-1.11-1.47-1.11-1.47-.91-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.87 1.52 2.34 1.07 2.91.83.09-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.92 0-1.11.38-2 1.03-2.71-.1-.25-.45-1.29.1-2.64 0 0 .84-.27 2.75 1.02.79-.22 1.65-.33 2.5-.33.85 0 1.71.11 2.5.33 1.91-1.29 2.75-1.02 2.75-1.02.55 1.35.2 2.39.1 2.64.65.71 1.03 1.6 1.03 2.71 0 3.82-2.34 4.66-4.57 4.91.36.31.69.92.69 1.85V21c0 .27.16.59.67.5C19.14 20.16 22 16.42 22 12A10 10 0 0 0 12 2Z"/></svg> </span> <span class=label> Contributors </span> </li> <li class=contributors-holder> <ul class=contributors data-bi-name=contributors><li><a href=https://github.com/d0choa title=d0choa data-bi-name=contributorprofile target=_blank><img src=https://avatars.githubusercontent.com/u/5097586 alt=d0choa></a></li></ul> </li> </ul> </small> </div> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/genetics_etl_python target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../../../..", "features": ["navigation.instant", "navigation.top", "navigation.tabs", "navigation.tabs.sticky", "content.code.annotate", "search.suggest", "search.highlight"], "search": "../../../../assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=../../../../assets/javascripts/bundle.6c14ae12.min.js></script> <script src=../../../../js/timeago.min.js></script> <script src=../../../../js/timeago_mkdocs_material.js></script> <script src=../../../../assets/javascripts/extra.js></script> </body> </html>