
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../sumstat_imputation/">
      
      
        <link rel="next" href="../susie_inf/">
      
      
      <link rel="icon" href="../../../assets/imgs/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.18">
    
    
      
        <title>QC of GWAS Summary Statistics - Open Targets Gentropy</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.66ac8b77.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../css/timeago.css">
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../assets/stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="light-blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#summary-statistics-qc-checks" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Open Targets Gentropy" class="md-header__button md-logo" aria-label="Open Targets Gentropy" data-md-component="logo">
      
  <img src="../../../assets/imgs/otlogo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Open Targets Gentropy
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              QC of GWAS Summary Statistics
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="light-blue"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="light-blue"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/opentargets/gentropy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    opentargets/gentropy
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../installation/" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../howto/_howto/" class="md-tabs__link">
        
  
    
  
  How-to

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../development/_development/" class="md-tabs__link">
        
  
    
  
  Development

      </a>
    </li>
  

      
        
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="../../_python_api/" class="md-tabs__link">
        
  
    
  
  Python API

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Open Targets Gentropy" class="md-nav__button md-logo" aria-label="Open Targets Gentropy" data-md-component="logo">
      
  <img src="../../../assets/imgs/otlogo.png" alt="logo">

    </a>
    Open Targets Gentropy
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/opentargets/gentropy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    opentargets/gentropy
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../howto/_howto/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    How-to
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3" id="__nav_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            How-to
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../howto/command_line/_command_line/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Command line
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1">
            <span class="md-nav__icon md-icon"></span>
            Command line
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/command_line/run_step_in_cli/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run step in cli
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/command_line/run_step_using_config/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run step using config
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../howto/tutorials/_tutorials/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Tutorials
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/tutorials/a_creating_spark_session/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Creating a Spark Session
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/tutorials/b_create_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Create a dataset
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/tutorials/c_applying_methods/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Applying methods
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/tutorials/d_inspect_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Inspect a dataset
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../development/_development/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Development
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4" id="__nav_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Development
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../development/contributing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Contributing guidelines
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../development/troubleshooting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Troubleshooting
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../_python_api/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Python API
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5" id="__nav_5_label" tabindex="">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Python API
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../common/_common/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Common
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            Common
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../common/genomic_region/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genomic_region
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../common/processing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    processing
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../common/session/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    session
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../common/spark/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    spark
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../common/stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../common/types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    types
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../common/udf/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    udf
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasets/_datasets/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Datasets
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_2" id="__nav_5_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            Datasets
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/biosample_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Biosample index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/intervals/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Intervals
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/l2g_feature_matrix/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Feature Matrix
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/l2g_gold_standard/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Gold Standard
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/l2g_prediction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Prediction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/ld_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LD Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/study_locus/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Locus
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/study_locus_overlap/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Locus Overlap
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/summary_statistics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/summary_statistics_qc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/target_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Target Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/variant_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_2_15" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasets/l2g_features/_l2g_feature/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    L2g features
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_2_15" id="__nav_5_2_15_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_2_15_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_2_15">
            <span class="md-nav__icon md-icon"></span>
            L2g features
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/l2g_features/colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/l2g_features/distance/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From distance
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/l2g_features/other/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Other features
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasets/l2g_features/vep/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From VEP
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/_datasources/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Datasources
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3" id="__nav_5_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3">
            <span class="md-nav__icon md-icon"></span>
            Datasources
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/biosample_ontologies/_cell_ontology/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Biosample ontologies
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_1" id="__nav_5_3_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_1">
            <span class="md-nav__icon md-icon"></span>
            Biosample ontologies
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/biosample_ontologies/_uberon/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Uberon
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/ensembl/_ensembl/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Ensembl
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_2" id="__nav_5_3_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_2">
            <span class="md-nav__icon md-icon"></span>
            Ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/ensembl/variant_effect_predictor_parser/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant effector parser
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/eqtl_catalogue/_eqtl_catalogue/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Eqtl catalogue
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_3" id="__nav_5_3_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_3">
            <span class="md-nav__icon md-icon"></span>
            Eqtl catalogue
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/eqtl_catalogue/finemapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Fine mapping results
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/eqtl_catalogue/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/eqtl_catalogue/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Stats
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/finngen/_finngen/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Finngen
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_4" id="__nav_5_3_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_4">
            <span class="md-nav__icon md-icon"></span>
            Finngen
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/finngen/finemapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Finemapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/finngen/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/finngen/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/gnomad/_gnomad/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Gnomad
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_5" id="__nav_5_3_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_5">
            <span class="md-nav__icon md-icon"></span>
            Gnomad
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gnomad/gnomad_ld/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LD Matrix
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gnomad/gnomad_variants/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variants
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_6" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/gwas_catalog/_gwas_catalog/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Gwas catalog
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_6" id="__nav_5_3_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_6">
            <span class="md-nav__icon md-icon"></span>
            Gwas catalog
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/associations/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Associations
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/study_splitter/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Splitter
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/summary_statistics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary statistics
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_7" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/intervals/_intervals/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Intervals
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_7" id="__nav_5_3_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_7">
            <span class="md-nav__icon md-icon"></span>
            Intervals
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/andersson/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Andersson et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/javierre/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Javierre et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/jung/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Jung et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/thurman/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Thurman et al.
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/open_targets/_open_targets/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Open targets
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_8" id="__nav_5_3_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_8">
            <span class="md-nav__icon md-icon"></span>
            Open targets
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/open_targets/l2g_gold_standard/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Gold Standard
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_9" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/ukb_ppp_eur/_ukb_ppp_eur/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Ukb ppp eur
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_9" id="__nav_5_3_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_9">
            <span class="md-nav__icon md-icon"></span>
            Ukb ppp eur
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/ukb_ppp_eur/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/ukb_ppp_eur/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../_methods/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Methods
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4" id="__nav_5_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_4">
            <span class="md-nav__icon md-icon"></span>
            Methods
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../carma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CARMA
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../coloc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Coloc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ecaviar/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    eCAVIAR
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ld_annotator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LDAnnotator
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PICS
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../sumstat_imputation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics Imputation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    QC of GWAS Summary Statistics
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    QC of GWAS Summary Statistics
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#summary-statistics-qc-checks" class="md-nav__link">
    <span class="md-ellipsis">
      Summary Statistics QC checks
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.method.sumstat_quality_controls.gc_lambda_check" class="md-nav__link">
    <span class="md-ellipsis">
      gc_lambda_check
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.method.sumstat_quality_controls.p_z_test" class="md-nav__link">
    <span class="md-ellipsis">
      p_z_test
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.method.sumstat_quality_controls.mean_beta_check" class="md-nav__link">
    <span class="md-ellipsis">
      mean_beta_check
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.method.sumstat_quality_controls.sumstat_n_eff_check" class="md-nav__link">
    <span class="md-ellipsis">
      sumstat_n_eff_check
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.method.sumstat_quality_controls.number_of_variants" class="md-nav__link">
    <span class="md-ellipsis">
      number_of_variants
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../susie_inf/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SuSiE-inf
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4_10" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../l2g/_l2g/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    L2g
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4_10" id="__nav_5_4_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_4_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4_10">
            <span class="md-nav__icon md-icon"></span>
            L2g
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g/feature_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Feature Factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g/model/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g/trainer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Trainer
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../steps/_steps/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Steps
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_5" id="__nav_5_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_5">
            <span class="md-nav__icon md-icon"></span>
            Steps
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/biosample_index_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    biosample_index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/credible_set_qc_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    credible_set_qc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/eqtl_catalogue/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    eQTL Catalogue
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/finngen_studies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    finngen_studies
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/finngen_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    finngen_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_curation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_study_curation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_study_inclusion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_top_hits/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GWAS Catalog Top Hits Ingestion Step
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/l2g/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Locus to Gene (L2G)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/ld_clump/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ld_based_clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/ld_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GnomAD Linkage data ingestion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/locus_breaker_clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    locus_breaker_clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/pics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    pics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/study_locus_validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study-Locus Validation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/study_validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Validation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/summary_statistics_qc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    summary_statistics_qc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/ukb_ppp_eur_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ukb_ppp_eur_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/variant_annotation_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GnomAD variant data ingestion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/variant_index_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    variant_index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/window_based_clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    window_based_clumping
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#summary-statistics-qc-checks" class="md-nav__link">
    <span class="md-ellipsis">
      Summary Statistics QC checks
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.method.sumstat_quality_controls.gc_lambda_check" class="md-nav__link">
    <span class="md-ellipsis">
      gc_lambda_check
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.method.sumstat_quality_controls.p_z_test" class="md-nav__link">
    <span class="md-ellipsis">
      p_z_test
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.method.sumstat_quality_controls.mean_beta_check" class="md-nav__link">
    <span class="md-ellipsis">
      mean_beta_check
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.method.sumstat_quality_controls.sumstat_n_eff_check" class="md-nav__link">
    <span class="md-ellipsis">
      sumstat_n_eff_check
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.method.sumstat_quality_controls.number_of_variants" class="md-nav__link">
    <span class="md-ellipsis">
      number_of_variants
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>QC of GWAS Summary Statistics</h1>

<p>This class consists of several general quality control checks for GWAS with full summary statistics.
There are several checks included:</p>
<ol>
<li>
<p>Genomic control lambda (median of the distribution of Chi2 statistics divided by expected for Chi2 with df=1). Lambda should be reasonably close to 1. Ideally not bigger than 2.</p>
</li>
<li>
<p>P-Z check: the linear regression between log10 of reported p-values and log10 of p-values inferred from betas and standard errors. Intercept of the regression should be close to 0, slope close to 1.</p>
</li>
<li>
<p>Mean beta check: mean of beta. Should be close to 0.</p>
</li>
<li>
<p>The N_eff check: It estimates the ratio between effective sample size and the expected one and checks its distribution. It is possible to conduct only if the effective allele frequency is provided in the study. The median ratio is always close to 1, standard error should be close to 0.</p>
</li>
<li>
<p>Number of SNPs and number of significant SNPs.</p>
</li>
</ol>
<h2 id="summary-statistics-qc-checks">Summary Statistics QC checks<a class="headerlink" href="#summary-statistics-qc-checks" title="Permanent link">&para;</a></h2>


<div class="doc doc-object doc-function">


<h2 id="gentropy.method.sumstat_quality_controls.gc_lambda_check" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">gentropy</span><span class="o">.</span><span class="n">method</span><span class="o">.</span><span class="n">sumstat_quality_controls</span><span class="o">.</span><span class="n">gc_lambda_check</span><span class="p">(</span><span class="n">gwas_for_qc</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span></code>

<a href="#gentropy.method.sumstat_quality_controls.gc_lambda_check" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents first">

        <p>The genomic control lambda check for QC of GWAS summary statistics.</p>
<p>The genomic control lambda is a measure of the inflation of test statistics in a GWAS.
It is calculated as the ratio of the median of the squared test statistics to the expected median under the null hypothesis.
The expected median under the null hypothesis is calculated using the chi-squared distribution with 1 degree of freedom.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gwas_for_qc</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dataframe with <code>studyId</code>, <code>beta</code>, and <code>standardError</code> columns.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>DataFrame</code></td>            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>PySpark DataFrame with the genomic control lambda for each study.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="warning" open>
  <summary>Warning</summary>
  <p>High lambda values indicate inflation of test statistics, which may be due to the population stratification or other confounding factors.</p>
</details>

<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="s1">&#39;studyId STRING, beta DOUBLE, standardError DOUBLE&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d1</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;S1&quot;</span><span class="p">,</span> <span class="mf">1.81</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;S1&quot;</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d2</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;S2&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;S2&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">d1</span> <span class="o">+</span> <span class="n">d2</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+-------+----+-------------+</span>
<span class="go">|studyId|beta|standardError|</span>
<span class="go">+-------+----+-------------+</span>
<span class="go">|     S1|1.81|          0.2|</span>
<span class="go">|     S1|-0.1|          0.2|</span>
<span class="go">|     S2| 1.0|          0.1|</span>
<span class="go">|     S2| 1.0|          0.1|</span>
<span class="go">+-------+----+-------------+</span>
</code></pre></div>
    <p><strong>This method outputs one value per study</strong></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">gc_lambda</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="s2">&quot;gc_lambda&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;gc_lambda&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gc_lambda_check</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">gc_lambda</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+-------+---------+</span>
<span class="go">|studyId|gc_lambda|</span>
<span class="go">+-------+---------+</span>
<span class="go">|     S1|     0.55|</span>
<span class="go">|     S2|   219.81|</span>
<span class="go">+-------+---------+</span>
</code></pre></div>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/method/sumstat_quality_controls.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">gc_lambda_check</span><span class="p">(</span><span class="n">gwas_for_qc</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;The genomic control lambda check for QC of GWAS summary statistics.</span>

<span class="sd">    The genomic control lambda is a measure of the inflation of test statistics in a GWAS.</span>
<span class="sd">    It is calculated as the ratio of the median of the squared test statistics to the expected median under the null hypothesis.</span>
<span class="sd">    The expected median under the null hypothesis is calculated using the chi-squared distribution with 1 degree of freedom.</span>

<span class="sd">    Args:</span>
<span class="sd">        gwas_for_qc (DataFrame): Dataframe with `studyId`, `beta`, and `standardError` columns.</span>

<span class="sd">    Returns:</span>
<span class="sd">        DataFrame: PySpark DataFrame with the genomic control lambda for each study.</span>

<span class="sd">    Warning:</span>
<span class="sd">        High lambda values indicate inflation of test statistics, which may be due to the population stratification or other confounding factors.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; s = &#39;studyId STRING, beta DOUBLE, standardError DOUBLE&#39;</span>
<span class="sd">        &gt;&gt;&gt; d1 = [(&quot;S1&quot;, 1.81, 0.2), (&quot;S1&quot;, -0.1, 0.2)]</span>
<span class="sd">        &gt;&gt;&gt; d2 = [(&quot;S2&quot;, 1.0, 0.1), (&quot;S2&quot;, 1.0, 0.1)]</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame(d1 + d2, s)</span>
<span class="sd">        &gt;&gt;&gt; df.show()</span>
<span class="sd">        +-------+----+-------------+</span>
<span class="sd">        |studyId|beta|standardError|</span>
<span class="sd">        +-------+----+-------------+</span>
<span class="sd">        |     S1|1.81|          0.2|</span>
<span class="sd">        |     S1|-0.1|          0.2|</span>
<span class="sd">        |     S2| 1.0|          0.1|</span>
<span class="sd">        |     S2| 1.0|          0.1|</span>
<span class="sd">        +-------+----+-------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        **This method outputs one value per study**</span>

<span class="sd">        &gt;&gt;&gt; gc_lambda = f.round(&quot;gc_lambda&quot;, 2).alias(&quot;gc_lambda&quot;)</span>
<span class="sd">        &gt;&gt;&gt; gc_lambda_check(df).select(&quot;studyId&quot;, gc_lambda).show()</span>
<span class="sd">        +-------+---------+</span>
<span class="sd">        |studyId|gc_lambda|</span>
<span class="sd">        +-------+---------+</span>
<span class="sd">        |     S1|     0.55|</span>
<span class="sd">        |     S2|   219.81|</span>
<span class="sd">        +-------+---------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">z_score</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">)</span> <span class="o">/</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">)</span>
    <span class="c1"># NOTE! The statistic can be calculated using scipy.stats.chi2.isf(0.5, df=1)</span>
    <span class="c1"># as well, since both functions are equal at the 0.5 quantile.</span>
    <span class="n">stat</span> <span class="o">=</span> <span class="n">chi2</span><span class="o">.</span><span class="n">ppf</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">df</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">qc_c</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">gwas_for_qc</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;beta&quot;</span><span class="p">,</span> <span class="s2">&quot;standardError&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;Z2&quot;</span><span class="p">,</span> <span class="n">z_score</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span>
        <span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">percentile_approx</span><span class="p">(</span><span class="s2">&quot;Z2&quot;</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;z2_median&quot;</span><span class="p">))</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;gc_lambda&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;z2_median&quot;</span><span class="p">)</span> <span class="o">/</span> <span class="n">stat</span><span class="p">)</span>
        <span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;gc_lambda&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">qc_c</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gentropy.method.sumstat_quality_controls.p_z_test" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">gentropy</span><span class="o">.</span><span class="n">method</span><span class="o">.</span><span class="n">sumstat_quality_controls</span><span class="o">.</span><span class="n">p_z_test</span><span class="p">(</span><span class="n">gwas_for_qc</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span></code>

<a href="#gentropy.method.sumstat_quality_controls.p_z_test" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents first">

        <p>The P-Z test for QC of GWAS summary statistics.</p>
<p>This function expects to have a dataframe with <code>studyId</code>, <code>beta</code>, <code>standardError</code>, <code>pValueMantissa</code> and <code>pValueExponent</code> columns</p>
<p>It runs linear regression between reported p-values and p-values inferred from z-scores.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gwas_for_qc</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dataframe with <code>studyId</code>, <code>beta</code>, <code>standardError</code>, <code>pValueMantissa</code> and <code>pValueExponent</code> columns.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>DataFrame</code></td>            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>PySpark DataFrame with the results of the linear regression for each study.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="note" open>
  <summary>Note</summary>
  <p>The algorithm does following things:
1. recalculates the negative logarithm of p-value from the square z-score
2. calculates the mean and se difference between the sum of logarithms derived form reported p-value mantissa and exponent and value recalculated from z-score.</p>
</details>

<details class="warning" open>
  <summary>Warning</summary>
  <p>The function requires the calculation of the <strong>chi-squared survival function</strong> to obtain the p-value from the z-score.
which is not available in PySpark. The function uses scipy instead, thus <strong>it is not optimized for large datasets</strong>.</p>
</details>

<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="s1">&#39;studyId STRING, beta DOUBLE, standardError DOUBLE, pValueMantissa FLOAT, pValueExponent INTEGER&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Example where the variant reaches upper and lower boundaries for mantissa and upper bound for exponent</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d1</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;S1&quot;</span><span class="p">,</span> <span class="mf">1.81</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="mf">9.9</span><span class="p">,</span> <span class="o">-</span><span class="mi">20</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;S1&quot;</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Example where z-score^2 (beta / se) &gt; 100</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d2</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;S2&quot;</span><span class="p">,</span> <span class="mf">101.0</span><span class="p">,</span> <span class="mf">10.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;S2&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;S2&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">,</span> <span class="o">-</span><span class="mi">2</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">d1</span> <span class="o">+</span> <span class="n">d2</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+-------+-----+-------------+--------------+--------------+</span>
<span class="go">|studyId| beta|standardError|pValueMantissa|pValueExponent|</span>
<span class="go">+-------+-----+-------------+--------------+--------------+</span>
<span class="go">|     S1| 1.81|          0.2|           9.9|           -20|</span>
<span class="go">|     S1| -0.1|          0.2|           1.0|            -1|</span>
<span class="go">|     S2|101.0|         10.0|           1.0|            -1|</span>
<span class="go">|     S2|  1.0|          0.1|           1.0|            -1|</span>
<span class="go">|     S2|  1.0|          0.1|           2.0|            -2|</span>
<span class="go">+-------+-----+-------------+--------------+--------------+</span>
</code></pre></div>
    <p><strong>This method outputs two values per study, mean and standard deviation of the difference between log p-value(s)</strong></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">mean_diff_pz</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="s2">&quot;mean_diff_pz&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;mean_diff_pz&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">se_diff_pz</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="s2">&quot;se_diff_pz&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;se_diff_pz&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p_z_test</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s1">&#39;studyId&#39;</span><span class="p">,</span> <span class="n">mean_diff_pz</span><span class="p">,</span> <span class="n">se_diff_pz</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+-------+------------+----------+</span>
<span class="go">|studyId|mean_diff_pz|se_diff_pz|</span>
<span class="go">+-------+------------+----------+</span>
<span class="go">|     S1|        0.47|      0.45|</span>
<span class="go">|     S2|      -21.47|      0.49|</span>
<span class="go">+-------+------------+----------+</span>
</code></pre></div>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/method/sumstat_quality_controls.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">p_z_test</span><span class="p">(</span><span class="n">gwas_for_qc</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;The P-Z test for QC of GWAS summary statistics.</span>

<span class="sd">    This function expects to have a dataframe with `studyId`, `beta`, `standardError`, `pValueMantissa` and `pValueExponent` columns</span>

<span class="sd">    It runs linear regression between reported p-values and p-values inferred from z-scores.</span>

<span class="sd">    Args:</span>
<span class="sd">        gwas_for_qc (DataFrame): Dataframe with `studyId`, `beta`, `standardError`, `pValueMantissa` and `pValueExponent` columns.</span>

<span class="sd">    Returns:</span>
<span class="sd">        DataFrame: PySpark DataFrame with the results of the linear regression for each study.</span>

<span class="sd">    Note:</span>
<span class="sd">        The algorithm does following things:</span>
<span class="sd">        1. recalculates the negative logarithm of p-value from the square z-score</span>
<span class="sd">        2. calculates the mean and se difference between the sum of logarithms derived form reported p-value mantissa and exponent and value recalculated from z-score.</span>

<span class="sd">    Warning:</span>
<span class="sd">         The function requires the calculation of the **chi-squared survival function** to obtain the p-value from the z-score.</span>
<span class="sd">         which is not available in PySpark. The function uses scipy instead, thus **it is not optimized for large datasets**.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; s = &#39;studyId STRING, beta DOUBLE, standardError DOUBLE, pValueMantissa FLOAT, pValueExponent INTEGER&#39;</span>
<span class="sd">        &gt;&gt;&gt; # Example where the variant reaches upper and lower boundaries for mantissa and upper bound for exponent</span>
<span class="sd">        &gt;&gt;&gt; d1 = [(&quot;S1&quot;, 1.81, 0.2, 9.9, -20), (&quot;S1&quot;, -0.1, 0.2, 1.0, -1)]</span>
<span class="sd">        &gt;&gt;&gt; # Example where z-score^2 (beta / se) &gt; 100</span>
<span class="sd">        &gt;&gt;&gt; d2 = [(&quot;S2&quot;, 101.0, 10.0, 1.0, -1), (&quot;S2&quot;, 1.0, 0.1, 1.0, -1), (&quot;S2&quot;, 1.0, 0.1, 2.0, -2)]</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame(d1 + d2, s)</span>
<span class="sd">        &gt;&gt;&gt; df.show()</span>
<span class="sd">        +-------+-----+-------------+--------------+--------------+</span>
<span class="sd">        |studyId| beta|standardError|pValueMantissa|pValueExponent|</span>
<span class="sd">        +-------+-----+-------------+--------------+--------------+</span>
<span class="sd">        |     S1| 1.81|          0.2|           9.9|           -20|</span>
<span class="sd">        |     S1| -0.1|          0.2|           1.0|            -1|</span>
<span class="sd">        |     S2|101.0|         10.0|           1.0|            -1|</span>
<span class="sd">        |     S2|  1.0|          0.1|           1.0|            -1|</span>
<span class="sd">        |     S2|  1.0|          0.1|           2.0|            -2|</span>
<span class="sd">        +-------+-----+-------------+--------------+--------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        **This method outputs two values per study, mean and standard deviation of the difference between log p-value(s)**</span>

<span class="sd">        &gt;&gt;&gt; mean_diff_pz = f.round(&quot;mean_diff_pz&quot;, 2).alias(&quot;mean_diff_pz&quot;)</span>
<span class="sd">        &gt;&gt;&gt; se_diff_pz = f.round(&quot;se_diff_pz&quot;, 2).alias(&quot;se_diff_pz&quot;)</span>
<span class="sd">        &gt;&gt;&gt; p_z_test(df).select(&#39;studyId&#39;, mean_diff_pz, se_diff_pz).show()</span>
<span class="sd">        +-------+------------+----------+</span>
<span class="sd">        |studyId|mean_diff_pz|se_diff_pz|</span>
<span class="sd">        +-------+------------+----------+</span>
<span class="sd">        |     S1|        0.47|      0.45|</span>
<span class="sd">        |     S2|      -21.47|      0.49|</span>
<span class="sd">        +-------+------------+----------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">qc_c</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">gwas_for_qc</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;Z2&quot;</span><span class="p">,</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">)</span> <span class="o">/</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">))</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
        <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;Z2&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">100</span><span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;neglogpValFromZScore&quot;</span><span class="p">,</span> <span class="n">neglogpval_from_z2</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;Z2&quot;</span><span class="p">)))</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;neglogpVal&quot;</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">))</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;diffpval&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;neglogpVal&quot;</span><span class="p">)</span> <span class="o">-</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;neglogpValfromZScore&quot;</span><span class="p">))</span>
        <span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="s2">&quot;diffpval&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;mean_diff_pz&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">stddev</span><span class="p">(</span><span class="s2">&quot;diffpval&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;se_diff_pz&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;mean_diff_pz&quot;</span><span class="p">,</span> <span class="s2">&quot;se_diff_pz&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">qc_c</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gentropy.method.sumstat_quality_controls.mean_beta_check" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">gentropy</span><span class="o">.</span><span class="n">method</span><span class="o">.</span><span class="n">sumstat_quality_controls</span><span class="o">.</span><span class="n">mean_beta_check</span><span class="p">(</span><span class="n">gwas_for_qc</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span></code>

<a href="#gentropy.method.sumstat_quality_controls.mean_beta_check" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents first">

        <p>The mean beta check for QC of GWAS summary statistics.</p>
<p>This function expects to have a dataframe with <code>studyId</code> and <code>beta</code> columns and
outputs the dataframe with mean beta aggregated over the studyId.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gwas_for_qc</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dataframe with <code>studyId</code> and <code>beta</code> columns.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>DataFrame</code></td>            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>PySpark DataFrame with the mean beta for each study.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="s2">&quot;studyId STRING, variantId STRING, beta DOUBLE&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">([(</span><span class="s1">&#39;S1&#39;</span><span class="p">,</span> <span class="s1">&#39;1_10000_A_T&#39;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;S1&#39;</span><span class="p">,</span> <span class="s1">&#39;1_10001_C_T&#39;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;S2&#39;</span><span class="p">,</span> <span class="s1">&#39;1_10001_C_T&#39;</span><span class="p">,</span> <span class="mf">0.028</span><span class="p">)],</span> <span class="n">schema</span><span class="o">=</span><span class="n">s</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+-------+-----------+-----+</span>
<span class="go">|studyId|  variantId| beta|</span>
<span class="go">+-------+-----------+-----+</span>
<span class="go">|     S1|1_10000_A_T|  1.0|</span>
<span class="go">|     S1|1_10001_C_T|  1.0|</span>
<span class="go">|     S2|1_10001_C_T|0.028|</span>
<span class="go">+-------+-----------+-----+</span>
</code></pre></div>
    <p><strong>This method outputs one value per study</strong></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">mean_beta</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="s2">&quot;mean_beta&quot;</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;mean_beta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mean_beta_check</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s1">&#39;studyId&#39;</span><span class="p">,</span> <span class="n">mean_beta</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+-------+---------+</span>
<span class="go">|studyId|mean_beta|</span>
<span class="go">+-------+---------+</span>
<span class="go">|     S1|      1.0|</span>
<span class="go">|     S2|    0.028|</span>
<span class="go">+-------+---------+</span>
</code></pre></div>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/method/sumstat_quality_controls.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">mean_beta_check</span><span class="p">(</span>
    <span class="n">gwas_for_qc</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;The mean beta check for QC of GWAS summary statistics.</span>

<span class="sd">    This function expects to have a dataframe with `studyId` and `beta` columns and</span>
<span class="sd">    outputs the dataframe with mean beta aggregated over the studyId.</span>

<span class="sd">    Args:</span>
<span class="sd">        gwas_for_qc (DataFrame): Dataframe with `studyId` and `beta` columns.</span>

<span class="sd">    Returns:</span>
<span class="sd">        DataFrame: PySpark DataFrame with the mean beta for each study.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; s = &quot;studyId STRING, variantId STRING, beta DOUBLE&quot;</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame([(&#39;S1&#39;, &#39;1_10000_A_T&#39;, 1.0), (&#39;S1&#39;, &#39;1_10001_C_T&#39;, 1.0), (&#39;S2&#39;, &#39;1_10001_C_T&#39;, 0.028)], schema=s)</span>
<span class="sd">        &gt;&gt;&gt; df.show()</span>
<span class="sd">        +-------+-----------+-----+</span>
<span class="sd">        |studyId|  variantId| beta|</span>
<span class="sd">        +-------+-----------+-----+</span>
<span class="sd">        |     S1|1_10000_A_T|  1.0|</span>
<span class="sd">        |     S1|1_10001_C_T|  1.0|</span>
<span class="sd">        |     S2|1_10001_C_T|0.028|</span>
<span class="sd">        +-------+-----------+-----+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        **This method outputs one value per study**</span>

<span class="sd">        &gt;&gt;&gt; mean_beta = f.round(&quot;mean_beta&quot;, 3).alias(&quot;mean_beta&quot;)</span>
<span class="sd">        &gt;&gt;&gt; mean_beta_check(df).select(&#39;studyId&#39;, mean_beta).show()</span>
<span class="sd">        +-------+---------+</span>
<span class="sd">        |studyId|mean_beta|</span>
<span class="sd">        +-------+---------+</span>
<span class="sd">        |     S1|      1.0|</span>
<span class="sd">        |     S2|    0.028|</span>
<span class="sd">        +-------+---------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">study_id</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span>
    <span class="n">beta</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">)</span>
    <span class="n">qc_c</span> <span class="o">=</span> <span class="n">gwas_for_qc</span><span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="n">study_id</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">beta</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;mean_beta&quot;</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">qc_c</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gentropy.method.sumstat_quality_controls.sumstat_n_eff_check" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">gentropy</span><span class="o">.</span><span class="n">method</span><span class="o">.</span><span class="n">sumstat_quality_controls</span><span class="o">.</span><span class="n">sumstat_n_eff_check</span><span class="p">(</span><span class="n">gwas_for_qc</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">n_total</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100000</span><span class="p">,</span> <span class="n">limit</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10000000</span><span class="p">,</span> <span class="n">min_count</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span></code>

<a href="#gentropy.method.sumstat_quality_controls.sumstat_n_eff_check" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents first">

        <p>The effective sample size check for QC of GWAS summary statistics.</p>
<p>It estimates the ratio between effective sample size and the expected one and checks it's distribution.
It is possible to conduct only if the effective allele frequency is provided in the study.
The median ratio is always close to 1, but standard error could be inflated.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gwas_for_qc</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dataframe with <code>studyId</code>, <code>beta</code>, <code>standardError</code>, and <code>effectAlleleFrequencyFromSource</code> columns.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_total</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The reported sample size of the study. The QC metrics is robust toward the sample size.</p>
              </div>
            </td>
            <td>
                  <code>100000</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>limit</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The limit for the number of variants to be used for the estimation.</p>
              </div>
            </td>
            <td>
                  <code>10000000</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_count</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The minimum number of variants to be used for the estimation.</p>
              </div>
            </td>
            <td>
                  <code>100</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>DataFrame</code></td>            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>PySpark DataFrame with the effective sample size ratio for each study.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="s1">&#39;studyId STRING, beta DOUBLE, standardError DOUBLE, effectAlleleFrequencyFromSource FLOAT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Example where we have a very common and very rare variant</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d1</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;S1&quot;</span><span class="p">,</span> <span class="mf">1.81</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.999</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;S1&quot;</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.001</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;S1&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Example where z-score^2 (beta / se) &gt; 100</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d2</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;S2&quot;</span><span class="p">,</span> <span class="mf">1.81</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="kc">None</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;S2&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;S2&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">d1</span> <span class="o">+</span> <span class="n">d2</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+-------+----+-------------+-------------------------------+</span>
<span class="go">|studyId|beta|standardError|effectAlleleFrequencyFromSource|</span>
<span class="go">+-------+----+-------------+-------------------------------+</span>
<span class="go">|     S1|1.81|          0.2|                          0.999|</span>
<span class="go">|     S1|-0.1|          0.2|                          0.001|</span>
<span class="go">|     S1| 1.0|          0.1|                            0.5|</span>
<span class="go">|     S2|1.81|          0.2|                           NULL|</span>
<span class="go">|     S2| 1.0|          0.1|                            0.5|</span>
<span class="go">|     S2| 1.0|          0.1|                            0.5|</span>
<span class="go">+-------+----+-------------+-------------------------------+</span>
</code></pre></div>
    <p>This method outputs one value per study</p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">se_n</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="s2">&quot;se_N&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sumstat_n_eff_check</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">min_count</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">limit</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">se_n</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+-------+--------------+</span>
<span class="go">|studyId|round(se_N, 2)|</span>
<span class="go">+-------+--------------+</span>
<span class="go">|     S1|           0.0|</span>
<span class="go">|     S2|           0.0|</span>
<span class="go">+-------+--------------+</span>
</code></pre></div>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/method/sumstat_quality_controls.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">sumstat_n_eff_check</span><span class="p">(</span>
    <span class="n">gwas_for_qc</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">n_total</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100_000</span><span class="p">,</span>
    <span class="n">limit</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10_000_000</span><span class="p">,</span>
    <span class="n">min_count</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;The effective sample size check for QC of GWAS summary statistics.</span>

<span class="sd">    It estimates the ratio between effective sample size and the expected one and checks it&#39;s distribution.</span>
<span class="sd">    It is possible to conduct only if the effective allele frequency is provided in the study.</span>
<span class="sd">    The median ratio is always close to 1, but standard error could be inflated.</span>

<span class="sd">    Args:</span>
<span class="sd">        gwas_for_qc (DataFrame): Dataframe with `studyId`, `beta`, `standardError`, and `effectAlleleFrequencyFromSource` columns.</span>
<span class="sd">        n_total (int): The reported sample size of the study. The QC metrics is robust toward the sample size.</span>
<span class="sd">        limit (int): The limit for the number of variants to be used for the estimation.</span>
<span class="sd">        min_count (int): The minimum number of variants to be used for the estimation.</span>

<span class="sd">    Returns:</span>
<span class="sd">        DataFrame: PySpark DataFrame with the effective sample size ratio for each study.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; s = &#39;studyId STRING, beta DOUBLE, standardError DOUBLE, effectAlleleFrequencyFromSource FLOAT&#39;</span>
<span class="sd">        &gt;&gt;&gt; # Example where we have a very common and very rare variant</span>
<span class="sd">        &gt;&gt;&gt; d1 = [(&quot;S1&quot;, 1.81, 0.2, 0.999), (&quot;S1&quot;, -0.1, 0.2, 0.001), (&quot;S1&quot;, 1.0, 0.1, 0.5)]</span>
<span class="sd">        &gt;&gt;&gt; # Example where z-score^2 (beta / se) &gt; 100</span>
<span class="sd">        &gt;&gt;&gt; d2 = [(&quot;S2&quot;, 1.81, 0.2, None), (&quot;S2&quot;, 1.0, 0.1, 0.5), (&quot;S2&quot;, 1.0, 0.1, 0.5)]</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame(d1 + d2, s)</span>
<span class="sd">        &gt;&gt;&gt; df.show()</span>
<span class="sd">        +-------+----+-------------+-------------------------------+</span>
<span class="sd">        |studyId|beta|standardError|effectAlleleFrequencyFromSource|</span>
<span class="sd">        +-------+----+-------------+-------------------------------+</span>
<span class="sd">        |     S1|1.81|          0.2|                          0.999|</span>
<span class="sd">        |     S1|-0.1|          0.2|                          0.001|</span>
<span class="sd">        |     S1| 1.0|          0.1|                            0.5|</span>
<span class="sd">        |     S2|1.81|          0.2|                           NULL|</span>
<span class="sd">        |     S2| 1.0|          0.1|                            0.5|</span>
<span class="sd">        |     S2| 1.0|          0.1|                            0.5|</span>
<span class="sd">        +-------+----+-------------+-------------------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        This method outputs one value per study</span>

<span class="sd">        &gt;&gt;&gt; se_n = f.round(&quot;se_N&quot;, 2)</span>
<span class="sd">        &gt;&gt;&gt; sumstat_n_eff_check(df, min_count=2, limit=2).select(&quot;studyId&quot;, se_n).show()</span>
<span class="sd">        +-------+--------------+</span>
<span class="sd">        |studyId|round(se_N, 2)|</span>
<span class="sd">        +-------+--------------+</span>
<span class="sd">        |     S1|           0.0|</span>
<span class="sd">        |     S2|           0.0|</span>
<span class="sd">        +-------+--------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">af</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;effectAlleleFrequencyFromSource&quot;</span><span class="p">)</span>
    <span class="n">se</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">)</span>
    <span class="n">beta</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">)</span>
    <span class="n">varG</span> <span class="o">=</span> <span class="n">genotypic_variance</span><span class="p">(</span><span class="n">af</span><span class="p">)</span>
    <span class="n">pheno_var</span> <span class="o">=</span> <span class="p">(</span><span class="n">se</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span> <span class="o">*</span> <span class="n">n_total</span> <span class="o">*</span> <span class="n">varG</span> <span class="o">+</span> <span class="p">((</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span> <span class="o">*</span> <span class="n">varG</span><span class="p">)</span>

    <span class="n">window</span> <span class="o">=</span> <span class="n">Window</span><span class="o">.</span><span class="n">partitionBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">orderBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span>
    <span class="n">pheno_median</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">percentile_approx</span><span class="p">(</span><span class="n">pheno_var</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">window</span><span class="p">)</span>
    <span class="n">N_hat_ratio</span> <span class="o">=</span> <span class="p">(</span><span class="n">pheno_median</span> <span class="o">-</span> <span class="p">(</span><span class="n">beta</span><span class="o">**</span><span class="mi">2</span> <span class="o">*</span> <span class="n">varG</span><span class="p">))</span> <span class="o">/</span> <span class="p">(</span><span class="n">se</span><span class="o">**</span><span class="mi">2</span> <span class="o">*</span> <span class="n">varG</span> <span class="o">*</span> <span class="n">n_total</span><span class="p">)</span>

    <span class="n">df_with_counts</span> <span class="o">=</span> <span class="n">gwas_for_qc</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span>
        <span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;effectAlleleFrequencyFromSource&quot;</span><span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;count&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">window</span><span class="p">))</span>

    <span class="c1"># Filter the DataFrame to keep only the groups with count greater than or equal to min_count</span>
    <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">df_with_counts</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;count&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">min_count</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;count&quot;</span><span class="p">)</span>

    <span class="c1"># Keep the number of variants up to the limit</span>
    <span class="n">gwas_df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">filtered_df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;row_num&quot;</span><span class="p">,</span> <span class="n">row_number</span><span class="p">()</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">window</span><span class="p">))</span>
        <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;row_num&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">limit</span><span class="p">)</span>
        <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;row_num&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="c1"># Calculate the genotypic variance following the formula 2 * AlleleFrequency * (1 - AlleleFrequency)</span>
    <span class="n">qc_c</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">gwas_df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;pheno_var&quot;</span><span class="p">,</span> <span class="n">pheno_var</span><span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;pheno_median&quot;</span><span class="p">,</span> <span class="n">pheno_median</span><span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;N_hat_ratio&quot;</span><span class="p">,</span> <span class="n">N_hat_ratio</span><span class="p">)</span>
        <span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">stddev</span><span class="p">(</span><span class="s2">&quot;N_hat_ratio&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;se_N&quot;</span><span class="p">))</span>
        <span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;se_N&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">qc_c</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gentropy.method.sumstat_quality_controls.number_of_variants" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">gentropy</span><span class="o">.</span><span class="n">method</span><span class="o">.</span><span class="n">sumstat_quality_controls</span><span class="o">.</span><span class="n">number_of_variants</span><span class="p">(</span><span class="n">gwas_for_qc</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">pval_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5e-08</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span></code>

<a href="#gentropy.method.sumstat_quality_controls.number_of_variants" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents first">

        <p>The function calculates number of SNPs and number of SNPs with p-value less than the threshold (default to 5e-8).</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gwas_for_qc</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dataframe with <code>studyId</code>, <code>variantId</code>, <code>pValueMantissa</code> and <code>pValueExponent</code> columns.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>pval_threshold</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The threshold for the p-value.</p>
              </div>
            </td>
            <td>
                  <code>5e-08</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>DataFrame</code></td>            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>PySpark DataFrame with the number of SNPs and number of SNPs with p-value less than threshold.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="s1">&#39;studyId STRING, variantId STRING, pValueMantissa FLOAT, pValueExponent INTEGER&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d1</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;S1&quot;</span><span class="p">,</span> <span class="s2">&quot;1_10000_A_T&quot;</span><span class="p">,</span> <span class="mf">9.9</span><span class="p">,</span> <span class="o">-</span><span class="mi">20</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;S1&quot;</span><span class="p">,</span> <span class="s2">&quot;1_10001_C_T&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;S1&quot;</span><span class="p">,</span> <span class="s2">&quot;1_10002_G_C&quot;</span><span class="p">,</span> <span class="mf">5.0</span><span class="p">,</span> <span class="o">-</span><span class="mi">8</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d2</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;S2&quot;</span><span class="p">,</span> <span class="s2">&quot;1_10001_C_T&quot;</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;S2&quot;</span><span class="p">,</span> <span class="s2">&quot;1_10002_G_C&quot;</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">,</span> <span class="o">-</span><span class="mi">2</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">d1</span> <span class="o">+</span> <span class="n">d2</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+-------+-----------+--------------+--------------+</span>
<span class="go">|studyId|  variantId|pValueMantissa|pValueExponent|</span>
<span class="go">+-------+-----------+--------------+--------------+</span>
<span class="go">|     S1|1_10000_A_T|           9.9|           -20|</span>
<span class="go">|     S1|1_10001_C_T|           1.0|            -1|</span>
<span class="go">|     S1|1_10002_G_C|           5.0|            -8|</span>
<span class="go">|     S2|1_10001_C_T|           1.0|            -1|</span>
<span class="go">|     S2|1_10002_G_C|           2.0|            -2|</span>
<span class="go">+-------+-----------+--------------+--------------+</span>
</code></pre></div>
    <p><strong>This method outputs two values per study <code>n_variants_sig</code> and <code>n_variants</code></strong></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">number_of_variants</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+-------+----------+--------------+</span>
<span class="go">|studyId|n_variants|n_variants_sig|</span>
<span class="go">+-------+----------+--------------+</span>
<span class="go">|     S1|         3|             2|</span>
<span class="go">|     S2|         2|             0|</span>
<span class="go">+-------+----------+--------------+</span>
</code></pre></div>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/method/sumstat_quality_controls.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">number_of_variants</span><span class="p">(</span>
    <span class="n">gwas_for_qc</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">pval_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5e-8</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;The function calculates number of SNPs and number of SNPs with p-value less than the threshold (default to 5e-8).</span>

<span class="sd">    Args:</span>
<span class="sd">        gwas_for_qc (DataFrame): Dataframe with `studyId`, `variantId`, `pValueMantissa` and `pValueExponent` columns.</span>
<span class="sd">        pval_threshold (float): The threshold for the p-value.</span>

<span class="sd">    Returns:</span>
<span class="sd">        DataFrame: PySpark DataFrame with the number of SNPs and number of SNPs with p-value less than threshold.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; s = &#39;studyId STRING, variantId STRING, pValueMantissa FLOAT, pValueExponent INTEGER&#39;</span>
<span class="sd">        &gt;&gt;&gt; d1 = [(&quot;S1&quot;, &quot;1_10000_A_T&quot;, 9.9, -20), (&quot;S1&quot;, &quot;1_10001_C_T&quot;, 1.0, -1), (&quot;S1&quot;, &quot;1_10002_G_C&quot;, 5.0, -8)]</span>
<span class="sd">        &gt;&gt;&gt; d2 = [(&quot;S2&quot;, &quot;1_10001_C_T&quot;, 1.0, -1), (&quot;S2&quot;, &quot;1_10002_G_C&quot;, 2.0, -2)]</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame(d1 + d2, s)</span>
<span class="sd">        &gt;&gt;&gt; df.show()</span>
<span class="sd">        +-------+-----------+--------------+--------------+</span>
<span class="sd">        |studyId|  variantId|pValueMantissa|pValueExponent|</span>
<span class="sd">        +-------+-----------+--------------+--------------+</span>
<span class="sd">        |     S1|1_10000_A_T|           9.9|           -20|</span>
<span class="sd">        |     S1|1_10001_C_T|           1.0|            -1|</span>
<span class="sd">        |     S1|1_10002_G_C|           5.0|            -8|</span>
<span class="sd">        |     S2|1_10001_C_T|           1.0|            -1|</span>
<span class="sd">        |     S2|1_10002_G_C|           2.0|            -2|</span>
<span class="sd">        +-------+-----------+--------------+--------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        **This method outputs two values per study `n_variants_sig` and `n_variants`**</span>

<span class="sd">        &gt;&gt;&gt; number_of_variants(df).show()</span>
<span class="sd">        +-------+----------+--------------+</span>
<span class="sd">        |studyId|n_variants|n_variants_sig|</span>
<span class="sd">        +-------+----------+--------------+</span>
<span class="sd">        |     S1|         3|             2|</span>
<span class="sd">        |     S2|         2|             0|</span>
<span class="sd">        +-------+----------+--------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">log10pval</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">))</span> <span class="o">+</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">)</span>
    <span class="n">threshold</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="n">pval_threshold</span><span class="p">))</span>
    <span class="n">snp_counts</span> <span class="o">=</span> <span class="n">gwas_for_qc</span><span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;*&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;n_variants&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">log10pval</span> <span class="o">&lt;=</span> <span class="n">threshold</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">IntegerType</span><span class="p">()))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;n_variants_sig&quot;</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">snp_counts</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>
<hr />
<div class="md-source-file">
  <small>
    <ul
      class="metadata page-metadata"
      data-bi-name="page info"
      lang="en-us"
      dir="ltr"
    >
      <!-- mkdocs-git-revision-date-plugin -->
      
      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3h-2Z"/></svg>
        </span>
        <span class="label"> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class="timeago" datetime="2024-04-24T08:55:51+00:00" locale="en"></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2024-04-24</span> </span>
      </li>
       
      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg>
        </span>
        <span class="label"> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class="timeago" datetime="2025-06-12T20:08:08+00:00" locale="en"></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2025-06-12</span> </span>
      </li>
      

      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2A10 10 0 0 0 2 12c0 4.42 2.87 8.17 6.84 9.5.5.08.66-.23.66-.5v-1.69c-2.77.6-3.36-1.34-3.36-1.34-.46-1.16-1.11-1.47-1.11-1.47-.91-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.87 1.52 2.34 1.07 2.91.83.09-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.92 0-1.11.38-2 1.03-2.71-.1-.25-.45-1.29.1-2.64 0 0 .84-.27 2.75 1.02.79-.22 1.65-.33 2.5-.33.85 0 1.71.11 2.5.33 1.91-1.29 2.75-1.02 2.75-1.02.55 1.35.2 2.39.1 2.64.65.71 1.03 1.6 1.03 2.71 0 3.82-2.34 4.66-4.57 4.91.36.31.69.92.69 1.85V21c0 .27.16.59.67.5C19.14 20.16 22 16.42 22 12A10 10 0 0 0 12 2Z"/></svg>
        </span>
        <span class="label"> Contributors </span>
      </li>
      <li class="contributors-holder">
        <ul class="contributors" data-bi-name="contributors"></ul>
      </li>
    </ul>
  </small>
</div>




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/gentropy" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.top", "navigation.tabs", "navigation.tabs.sticky", "content.code.annotate", "search.suggest", "search.highlight"], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.3220b9d7.min.js"></script>
      
        <script src="../../../js/timeago.min.js"></script>
      
        <script src="../../../js/timeago_mkdocs_material.js"></script>
      
        <script src="../../../assets/javascripts/extra.js"></script>
      
    
  </body>
</html>