{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9306d86d",
   "metadata": {},
   "source": [
    "# UKBB MVP Finngen summary statistics preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4855a12",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4c2013dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "    <style>\n",
       "        .bk-notebook-logo {\n",
       "            display: block;\n",
       "            width: 20px;\n",
       "            height: 20px;\n",
       "            background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAABx0RVh0U29mdHdhcmUAQWRvYmUgRmlyZXdvcmtzIENTNui8sowAAAOkSURBVDiNjZRtaJVlGMd/1/08zzln5zjP1LWcU9N0NkN8m2CYjpgQYQXqSs0I84OLIC0hkEKoPtiH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);\n",
       "        }\n",
       "    </style>\n",
       "    <div>\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-notebook-logo\"></a>\n",
       "        <span id=\"f1e11fba-ab36-467d-ae17-e856a94a5627\">Loading BokehJS ...</span>\n",
       "    </div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\nconst JS_MIME_TYPE = 'application/javascript';\n  const HTML_MIME_TYPE = 'text/html';\n  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n  const CLASS_NAME = 'output_bokeh rendered_html';\n\n  /**\n   * Render data to the DOM node\n   */\n  function render(props, node) {\n    const script = document.createElement(\"script\");\n    node.appendChild(script);\n  }\n\n  /**\n   * Handle when an output is cleared or removed\n   */\n  function handleClearOutput(event, handle) {\n    function drop(id) {\n      const view = Bokeh.index.get_by_id(id)\n      if (view != null) {\n        view.model.document.clear()\n        Bokeh.index.delete(view)\n      }\n    }\n\n    const cell = handle.cell;\n\n    const id = cell.output_area._bokeh_element_id;\n    const server_id = cell.output_area._bokeh_server_id;\n\n    // Clean up Bokeh references\n    if (id != null) {\n      drop(id)\n    }\n\n    if (server_id !== undefined) {\n      // Clean up Bokeh references\n      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n      cell.notebook.kernel.execute(cmd_clean, {\n        iopub: {\n          output: function(msg) {\n            const id = msg.content.text.trim()\n            drop(id)\n          }\n        }\n      });\n      // Destroy server and session\n      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n      cell.notebook.kernel.execute(cmd_destroy);\n    }\n  }\n\n  /**\n   * Handle when a new output is added\n   */\n  function handleAddOutput(event, handle) {\n    const output_area = handle.output_area;\n    const output = handle.output;\n\n    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n      return\n    }\n\n    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n\n    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n      // store reference to embed id on output_area\n      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n    }\n    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n      const bk_div = document.createElement(\"div\");\n      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n      const script_attrs = bk_div.children[0].attributes;\n      for (let i = 0; i < script_attrs.length; i++) {\n        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n      }\n      // store reference to server id on output_area\n      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n    }\n  }\n\n  function register_renderer(events, OutputArea) {\n\n    function append_mime(data, metadata, element) {\n      // create a DOM node to render to\n      const toinsert = this.create_output_subarea(\n        metadata,\n        CLASS_NAME,\n        EXEC_MIME_TYPE\n      );\n      this.keyboard_manager.register_events(toinsert);\n      // Render to node\n      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n      render(props, toinsert[toinsert.length - 1]);\n      element.append(toinsert);\n      return toinsert\n    }\n\n    /* Handle when an output is cleared or removed */\n    events.on('clear_output.CodeCell', handleClearOutput);\n    events.on('delete.Cell', handleClearOutput);\n\n    /* Handle when a new output is added */\n    events.on('output_added.OutputArea', handleAddOutput);\n\n    /**\n     * Register the mime type and append_mime function with output_area\n     */\n    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n      /* Is output safe? */\n      safe: true,\n      /* Index of renderer in `output_area.display_order` */\n      index: 0\n    });\n  }\n\n  // register the mime type if in Jupyter Notebook environment and previously unregistered\n  if (root.Jupyter !== undefined) {\n    const events = require('base/js/events');\n    const OutputArea = require('notebook/js/outputarea').OutputArea;\n\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  }\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(\"f1e11fba-ab36-467d-ae17-e856a94a5627\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.3.4.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.3.4.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\nif (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"f1e11fba-ab36-467d-ae17-e856a94a5627\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));",
      "application/vnd.bokehjs_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/mindos/gentropy/.venv/lib/python3.11/site-packages/pyspark/sql/pandas/functions.py:407: UserWarning:\n",
      "\n",
      "In Python 3.6+ and Spark 3.0+, it is preferred to specify type hints for pandas UDF instead of specifying pandas UDF type which will be deprecated in the future releases. See SPARK-28264 for more details.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from gentropy import Session\n",
    "from pyspark.sql import types as t\n",
    "from pyspark.sql import functions as f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "24a3f82d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/10/06 16:55:30 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.\n",
      "25/10/06 16:55:30 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.\n",
      "25/10/06 16:55:30 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.\n",
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n",
      "25/10/06 16:55:31 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.\n",
      "25/10/06 16:55:31 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.\n",
      "25/10/06 16:55:32 INFO SparkEnv: Registering MapOutputTracker\n",
      "25/10/06 16:55:32 INFO SparkEnv: Registering BlockManagerMaster\n",
      "25/10/06 16:55:32 INFO SparkEnv: Registering BlockManagerMasterHeartbeat\n",
      "25/10/06 16:55:32 INFO SparkEnv: Registering OutputCommitCoordinator\n",
      "25/10/06 16:55:36 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.\n"
     ]
    }
   ],
   "source": [
    "session = Session(\n",
    "    spark_uri=\"yarn\",\n",
    "    extended_spark_conf={\"spark.master.memory\": \"16G\"},\n",
    "    write_mode=\"overwrite\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2dfacfb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Inputs\n",
    "source_manifest_path = \"gs://finngen-public-data-r12/meta_analysis/mvp_ukbb/FinnGen_R12_MVP_UKBB_manifest.tsv\"\n",
    "source_summary_statistics_glob = (\n",
    "    \"gs://finngen-public-data-r12/meta_analysis/mvp_ukbb/summary_stats/*.gz\"\n",
    ")\n",
    "efo_curation_path = \"https://raw.githubusercontent.com/opentargets/curation/refs/tags/25.09/mappings/disease/manual_string.tsv\"\n",
    "gnomad_variant_index_path = \"gs://gnomad_data_2/v4.1/variant_index/\"\n",
    "\n",
    "## Outputs\n",
    "study_index_output_path = \"gs://finngen_ukb_mvp_meta_data/study_index/\"\n",
    "raw_summary_statistics_output_path = (\n",
    "    \"gs://finngen_ukb_mvp_meta_data/raw_summary_statistics/\"\n",
    ")\n",
    "harmonised_summary_statistics_output_path = (\n",
    "    \"gs://finngen_ukb_mvp_meta_data/harmonised_summary_statistics/\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a556fa1e",
   "metadata": {},
   "source": [
    "### Building StudyIndex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7e5c707b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gentropy.finngen_ukb_mvp_meta import FinngenUkbbMvpMetaIngestionStep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3d5f92fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/10/06 13:54:10 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.\n",
      "25/10/06 13:54:10 WARN SparkConf: The configuration key 'spark.yarn.executor.failuresValidityInterval' has been deprecated as of Spark 3.5 and may be removed in the future. Please use the new key 'spark.executor.failuresValidityInterval' instead.\n",
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<gentropy.finngen_ukb_mvp_meta.FinngenUkbbMvpMetaIngestionStep at 0x7f3745af12d0>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FinngenUkbbMvpMetaIngestionStep(\n",
    "    session=session,\n",
    "    source_manifest_path=source_manifest_path,\n",
    "    source_summary_statistics_glob=source_summary_statistics_glob,\n",
    "    efo_curation_path=efo_curation_path,\n",
    "    gnomad_variant_index_path=gnomad_variant_index_path,\n",
    "    study_index_output_path=study_index_output_path,\n",
    "    raw_summary_statistics_output_path=raw_summary_statistics_output_path,\n",
    "    harmonised_summary_statistics_output_path=harmonised_summary_statistics_output_path,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c4180e0",
   "metadata": {},
   "source": [
    "#### Reading studyIndex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "08a9ab31",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------+--------------------+---------+--------------------+------------------------+----------+------+---------------------+-----------+--------+----------------+----------------------+---------------+------------------+----------------------------------+--------------------+--------------------+------+---------+--------+--------------------+---------------------+--------------------+------------------+---------------+-------------+--------------------+-----------+---------+---------------+\n",
      "|             studyId|           projectId|studyType|     traitFromSource|traitFromSourceMappedIds|diseaseIds|geneId|biosampleFromSourceId|biosampleId|pubmedId|publicationTitle|publicationFirstAuthor|publicationDate|publicationJournal|backgroundTraitFromSourceMappedIds|backgroundDiseaseIds|   initialSampleSize|nCases|nControls|nSamples|             cohorts|ldPopulationStructure|    discoverySamples|replicationSamples|qualityControls|analysisFlags|summarystatsLocation|hasSumstats|condition|sumstatQCValues|\n",
      "+--------------------+--------------------+---------+--------------------+------------------------+----------+------+---------------------+-----------+--------+----------------+----------------------+---------------+------------------+----------------------------------+--------------------+--------------------+------+---------+--------+--------------------+---------------------+--------------------+------------------+---------------+-------------+--------------------+-----------+---------+---------------+\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|            AV-block|         [MONDO_0000465]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...| 29658|  1356907| 1386565|[MVP, FinnGen, pa...| [{fin, 0.27636136...|[{383193, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|Abdominal aortic ...|           [EFO_0004214]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...| 24557|  1474384| 1498941|[MVP, FinnGen, pa...| [{fin, 0.31191687...|[{467545, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|                Acne|           [EFO_0003894]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...| 21289|  1466258| 1487547|[MVP, FinnGen, pa...| [{fin, 0.32336524...|[{481021, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|Acquired haemolyt...|           [EFO_0005558]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...|  1693|  1352577| 1354270|[MVP, FinnGen, pa...| [{fin, 0.36922696...|[{500033, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|     Acute gastritis|           [EFO_0000217]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...|  6945|  1369643| 1376588|[MVP, FinnGen, pa...| [{fin, 0.31008914...|[{426865, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|Acute laryngitis ...|    [MONDO_0004777, E...|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...|  6890|  1274083| 1280973|[MVP, FinnGen, pa...| [{fin, 0.31904653...|[{408690, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|  Acute pancreatitis|           [EFO_1000652]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...| 22643|  1468789| 1491432|[MVP, FinnGen, pa...| [{fin, 0.29895027...|[{445864, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|   Acute peritonitis|           [EFO_0008588]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...|  7969|  1485835| 1493804|[MVP, FinnGen, pa...| [{fin, 0.33386374...|[{498727, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|   Acute pharyngitis|           [EFO_0009657]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...| 30951|  1374204| 1405155|[MVP, FinnGen, pa...| [{fin, 0.29150164...|[{409605, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|Acute posthaemorr...|         [MONDO_0002280]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...| 10198|  1480255| 1490453|[MVP, FinnGen, pa...| [{fin, 0.32263748...|[{480876, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas| Acute renal failure|            [HP_0001919]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...| 90152|  1394880| 1485032|[MVP, FinnGen, pa...| [{fin, 0.32917203...|[{488831, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|     Acute sinusitis|           [EFO_0007486]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...| 86773|  1283244| 1370017|[MVP, FinnGen, pa...| [{fin, 0.31479025...|[{431268, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|Adrenocortical in...|           [EFO_0009491]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...|  3262|  1350537| 1353799|[MVP, FinnGen, pa...| [{fin, 0.35480599...|[{480336, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|     Agranulocytosis|            [HP_0012234]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...| 14543|  1470638| 1485181|[MVP, FinnGen, pa...| [{fin, 0.33360378...|[{495462, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|Alcoholic cardiom...|           [EFO_1000801]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...|   758|   949979|  950737|[MVP, FinnGen, pa...| [{fin, 0.51840203...|[{492864, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|       All influenza|           [EFO_0007328]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...| 19666|  1419437| 1439103|[MVP, FinnGen, pa...| [{fin, 0.29677931...|[{427096, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|      All pneumoniae|           [EFO_0003106]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...|151723|  1360172| 1511895|[MVP, FinnGen, pa...| [{fin, 0.33094097...|[{500348, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|     Alopecia areata|           [EFO_0004192]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...|  2940|  1460632| 1463572|[MVP, FinnGen, pa...| [{fin, 0.32616092...|[{477360, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|Amblyopia ex anopsia|         [MONDO_0001020]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...|  9829|  1460072| 1469901|[MVP, FinnGen, pa...| [{fin, 0.32485249...|[{477501, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "|FINNGEN_R12_UKB_M...|FINNGEN_R12_UKB_M...|     gwas|     Angina pectoris|           [EFO_0003913]|      NULL|  NULL|                 NULL|       NULL|    NULL|            NULL|                  NULL|           NULL|              NULL|                              NULL|                NULL|1,550,147 (MVP: n...|114520|  1341912| 1456432|[MVP, FinnGen, pa...| [{fin, 0.31636149...|[{460759, Finnish...|              NULL|           NULL|         NULL|gs://finngen-publ...|       true|     NULL|           NULL|\n",
      "+--------------------+--------------------+---------+--------------------+------------------------+----------+------+---------------------+-----------+--------+----------------+----------------------+---------------+------------------+----------------------------------+--------------------+--------------------+------+---------+--------+--------------------+---------------------+--------------------+------------------+---------------+-------------+--------------------+-----------+---------+---------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from gentropy import StudyIndex\n",
    "\n",
    "si = StudyIndex.from_parquet(session, study_index_output_path)\n",
    "si.df.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d75e677b",
   "metadata": {},
   "source": [
    "#### nSamples validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "87659226",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------+---------------+------------+-------------+--------+----------+---------------+------------------+---------------+------------------+---------------+------------------+--------------------+\n",
      "|ukbb_n_cases|ukbb_n_controls|fg_phenotype|fg_n_controls|    name|fg_n_cases|MVP_AMR_n_cases|MVP_AFR_n_controls|MVP_EUR_n_cases|MVP_EUR_n_controls|MVP_AFR_n_cases|MVP_AMR_n_controls|         path_bucket|\n",
      "+------------+---------------+------------+-------------+--------+----------+---------------+------------------+---------------+------------------+---------------+------------------+--------------------+\n",
      "|        3088|         384657|  I9_AVBLOCK|       375343|AV-block|      7850|            817|            115972|          15010|            431314|           2893|             49621|gs://finngen-publ...|\n",
      "+------------+---------------+------------+-------------+--------+----------+---------------+------------------+---------------+------------------+---------------+------------------+--------------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from gentropy.datasource.finngen_meta import FinnGenMetaManifest\n",
    "\n",
    "finngen_manifest = FinnGenMetaManifest.from_path(session, source_manifest_path)\n",
    "finngen_manifest._df.filter(f.col(\"fg_phenotype\") == \"I9_AVBLOCK\").show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29890590",
   "metadata": {},
   "source": [
    "#### Disease validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4c3d02d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_index_path = (\n",
    "    \"gs://open-targets-pipeline-runs/il/25.09-testrun-1/output/disease/\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3572a96c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "disease_index = (\n",
    "    session.spark.read.parquet(disease_index_path)\n",
    "    .select(\n",
    "        f.col(\"id\").alias(\"diseaseId\"),\n",
    "        f.explode_outer(\n",
    "            f.when(\n",
    "                f.col(\"obsoleteTerms\").isNotNull(),\n",
    "                f.array_union(f.array(\"id\"), f.col(\"obsoleteTerms\")),\n",
    "            )\n",
    "        ).alias(\"efo\"),\n",
    "    )\n",
    "    .withColumn(\"efo\", f.coalesce(f.col(\"efo\"), f.col(\"diseaseId\")))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "98380c3d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/10/06 12:16:27 WARN CacheManager: Asked to cache already cached data.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-------------------------------------------------------------+------------------------------------------------------+------------------------+----------+-----------------------------------+\n",
      "|studyId                                                      |traitFromSource                                       |traitFromSourceMappedIds|diseaseIds|qualityControls                    |\n",
      "+-------------------------------------------------------------+------------------------------------------------------+------------------------+----------+-----------------------------------+\n",
      "|FINNGEN_R12_UKB_MVP_META_L12_OTHPIGMENTATION                 |Other disorders of pigmentation                       |[EFO_1000755]           |[]        |[No valid disease identifier found]|\n",
      "|FINNGEN_R12_UKB_MVP_META_M13_FATIGFRACT                      |Fatigue fracture of vertebra                          |[HP_0041166]            |[]        |[No valid disease identifier found]|\n",
      "|FINNGEN_R12_UKB_MVP_META_M13_FINGERDEFORM                    |Deformity of finger(s)                                |[NCIT_C36287]           |[]        |[No valid disease identifier found]|\n",
      "|FINNGEN_R12_UKB_MVP_META_D3_ANAEMIA_FOLATEDEF                |Folate deficiency anaemia                             |[DOID_14026]            |[]        |[No valid disease identifier found]|\n",
      "|FINNGEN_R12_UKB_MVP_META_K11_PARODON_OPER                    |Parodontitis or operation codes                       |[HP_0000704]            |[]        |[No valid disease identifier found]|\n",
      "|FINNGEN_R12_UKB_MVP_META_Q17_OTHER_CONGEN_MALFO_DIGES_SYSTEM1|Other congenital malformations of the digestive system|[MONDO_0004335]         |[]        |[No valid disease identifier found]|\n",
      "|FINNGEN_R12_UKB_MVP_META_N14_HYPERTROPHYBREAST_BOTH          |Hypertrophy of breast in women and men                |[MONDO_0001100]         |[]        |[No valid disease identifier found]|\n",
      "|FINNGEN_R12_UKB_MVP_META_H8_EXTOTITIS                        |Otitis externa                                        |[HP_0000389]            |[]        |[No valid disease identifier found]|\n",
      "|FINNGEN_R12_UKB_MVP_META_H7_PERIPHRETINADEGEN                |Peripheral retinal degeneration                       |[HP_0007769]            |[]        |[No valid disease identifier found]|\n",
      "+-------------------------------------------------------------+------------------------------------------------------+------------------------+----------+-----------------------------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "si.validate_disease(disease_index).df.filter(f.size(\"qualityControls\") > 0).limit(\n",
    "    10\n",
    ").select(\n",
    "    \"studyId\",\n",
    "    \"traitFromSource\",\n",
    "    \"traitFromSourceMappedIds\",\n",
    "    \"diseaseIds\",\n",
    "    \"qualityControls\",\n",
    ").show(truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f52ab29d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 89:>                                                         (0 + 1) / 1]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------+---+\n",
      "|diseaseId|efo|\n",
      "+---------+---+\n",
      "+---------+---+\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "disease_index.filter(f.col(\"efo\") == \"EFO_1000755\").show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54789707",
   "metadata": {},
   "source": [
    "### Preparing variant annotations for flipping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c2e48675",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gentropy.dataset.variant_direction import VariantDirection\n",
    "from gentropy.dataset.variant_index import VariantIndex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "245efa77",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "vi = VariantIndex.from_parquet(session, gnomad_variant_index_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6a9bd8b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "vd = VariantDirection.from_variant_index(variant_index=vi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "57dd8540",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----------+-------------------+----+-------------------+---------+------+-------------+\n",
      "|chromosome|  originalVariantId|type|          variantId|direction|strand|isPalindromic|\n",
      "+----------+-------------------+----+-------------------+---------+------+-------------+\n",
      "|         4|     4_64274529_A_G|   1|     4_64274529_A_G|        1|     1|        false|\n",
      "|         4|     4_64274529_A_G|   1|     4_64274529_G_A|       -1|     1|        false|\n",
      "|         4|     4_64274529_A_G|   1|     4_64274529_T_C|        1|    -1|        false|\n",
      "|         4|     4_64274529_A_G|   1|     4_64274529_C_T|       -1|    -1|        false|\n",
      "|         4|     4_64274530_G_C|   1|     4_64274530_G_C|        1|     1|         true|\n",
      "|         4|     4_64274530_G_C|   1|     4_64274530_C_G|       -1|     1|         true|\n",
      "|         4|     4_64274530_G_C|   1|     4_64274530_C_G|        1|    -1|         true|\n",
      "|         4|     4_64274530_G_C|   1|     4_64274530_G_C|       -1|    -1|         true|\n",
      "|         4|     4_64274537_A_G|   1|     4_64274537_A_G|        1|     1|        false|\n",
      "|         4|     4_64274537_A_G|   1|     4_64274537_G_A|       -1|     1|        false|\n",
      "|         4|     4_64274537_A_G|   1|     4_64274537_T_C|        1|    -1|        false|\n",
      "|         4|     4_64274537_A_G|   1|     4_64274537_C_T|       -1|    -1|        false|\n",
      "|         4|4_64274537_A_AAATCG|   2|4_64274537_A_AAATCG|        1|     1|        false|\n",
      "|         4|4_64274537_A_AAATCG|   2|4_64274537_AAATCG_A|       -1|     1|        false|\n",
      "|         4|4_64274537_A_AAATCG|   2|4_64274537_T_CGATTT|        1|    -1|        false|\n",
      "|         4|4_64274537_A_AAATCG|   2|4_64274537_CGATTT_T|       -1|    -1|        false|\n",
      "|         4|     4_64274550_A_G|   1|     4_64274550_A_G|        1|     1|        false|\n",
      "|         4|     4_64274550_A_G|   1|     4_64274550_G_A|       -1|     1|        false|\n",
      "|         4|     4_64274550_A_G|   1|     4_64274550_T_C|        1|    -1|        false|\n",
      "|         4|     4_64274550_A_G|   1|     4_64274550_C_T|       -1|    -1|        false|\n",
      "+----------+-------------------+----+-------------------+---------+------+-------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "vd.df.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bd89142",
   "metadata": {},
   "source": [
    "## Munge summary statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7e38994b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gentropy.common.stats import pvalue_from_neglogpval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "68c6d365",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "sst = session.spark.read.parquet(raw_summary_statistics_output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e4e8f19b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['#CHR', 'POS', 'REF', 'ALT', 'SNP', 'fg_beta', 'fg_sebeta', 'fg_pval', 'fg_af_alt', 'fg_af_alt_cases', 'fg_af_alt_controls', 'MVP_EUR_beta', 'MVP_EUR_sebeta', 'MVP_EUR_pval', 'MVP_EUR_af_alt', 'MVP_EUR_r2', 'MVP_AFR_beta', 'MVP_AFR_sebeta', 'MVP_AFR_pval', 'MVP_AFR_af_alt', 'MVP_AFR_r2', 'MVP_HIS_beta', 'MVP_HIS_sebeta', 'MVP_HIS_pval', 'MVP_HIS_af_alt', 'MVP_HIS_r2', 'all_meta_N', 'all_inv_var_meta_beta', 'all_inv_var_meta_sebeta', 'all_inv_var_meta_p', 'all_inv_var_meta_mlogp', 'all_inv_var_het_p', 'leave_fg_N', 'leave_fg_inv_var_meta_beta', 'leave_fg_inv_var_meta_sebeta', 'leave_fg_inv_var_meta_p', 'leave_fg_inv_var_meta_mlogp', 'leave_fg_inv_var_meta_het_p', 'leave_MVP_EUR_N', 'leave_MVP_EUR_inv_var_meta_beta', 'leave_MVP_EUR_inv_var_meta_sebeta', 'leave_MVP_EUR_inv_var_meta_p', 'leave_MVP_EUR_inv_var_meta_mlogp', 'leave_MVP_EUR_inv_var_meta_het_p', 'leave_MVP_AFR_N', 'leave_MVP_AFR_inv_var_meta_beta', 'leave_MVP_AFR_inv_var_meta_sebeta', 'leave_MVP_AFR_inv_var_meta_p', 'leave_MVP_AFR_inv_var_meta_mlogp', 'leave_MVP_AFR_inv_var_meta_het_p', 'leave_MVP_HIS_N', 'leave_MVP_HIS_inv_var_meta_beta', 'leave_MVP_HIS_inv_var_meta_sebeta', 'leave_MVP_HIS_inv_var_meta_p', 'leave_MVP_HIS_inv_var_meta_mlogp', 'leave_MVP_HIS_inv_var_meta_het_p', 'rsid', 'studyId']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 10:>                                                         (0 + 1) / 1]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------------------------------+----------+--------+---------------+---------------+--------+-------------+------------------+----------------------+--------+------------+------------+------------+\n",
      "|studyId                               |chromosome|position|referenceAllele|alternateAllele|beta    |standardError|all_inv_var_meta_p|all_inv_var_meta_mlogp|fg_pval |MVP_EUR_pval|MVP_AFR_pval|MVP_HIS_pval|\n",
      "+--------------------------------------+----------+--------+---------------+---------------+--------+-------------+------------------+----------------------+--------+------------+------------+------------+\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |11063   |T              |G              |1.54e+00|7.25e-01     |4.747e-05         |1                     |NA      |NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |13259   |G              |A              |1.92e-01|7.06e-01     |0.0002696         |1                     |NA      |NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |17641   |G              |A              |1.14e-01|8.05e-01     |0.0008418         |1                     |NA      |NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |57222   |T              |C              |1.27e-01|5.62e-01     |0.0006442         |1                     |NA      |NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |58396   |T              |C              |1.93e-01|9.83e-01     |0.0002396         |1                     |NA      |NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |63668   |G              |A              |6.44e-01|5.81e-01     |2.6e-05           |1                     |NA      |NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |69569   |T              |C              |2.34e-01|9.78e-01     |0.0001882         |1                     |NA      |NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |79192   |T              |G              |3.33e-01|9.68e-01     |7.96e-05          |1                     |NA      |NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |91588   |G              |A              |2.37e-01|1.04e-01     |0.0001461         |1                     |NA      |NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |96882   |A              |G              |2.74e-01|9.46e-01     |0.0002096         |1                     |NA      |NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |97269   |T              |TAC            |1.09e-02|5.14e-01     |0.09473           |1                     |NA      |NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |108787  |G              |T              |NA      |NA           |NA                |1                     |2.51e-01|NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |126470  |A              |G              |NA      |NA           |NA                |1                     |3.62e-01|NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |135173  |G              |A              |NA      |NA           |NA                |1                     |5.23e-01|NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |138619  |G              |A              |NA      |NA           |NA                |1                     |6.34e-02|NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |181963  |T              |G              |NA      |NA           |NA                |1                     |4.07e-01|NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |185296  |C              |T              |NA      |NA           |NA                |1                     |8.98e-01|NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |186404  |G              |A              |NA      |NA           |NA                |1                     |3.86e-01|NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |205992  |T              |G              |NA      |NA           |NA                |1                     |1.15e-01|NA          |NA          |NA          |\n",
      "|FINNGEN_R12_UKB_MVP_META_I9_HYPTENSESS|1         |267915  |A              |T              |NA      |NA           |NA                |1                     |7.47e-01|NA          |NA          |NA          |\n",
      "+--------------------------------------+----------+--------+---------------+---------------+--------+-------------+------------------+----------------------+--------+------------+------------+------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "cols = sst.columns\n",
    "print(cols)\n",
    "pval_cols = [col for col in cols if \"_pval\" in col]\n",
    "sst.select(\n",
    "    f.col(\"studyId\"),\n",
    "    f.col(\"#CHR\").alias(\"chromosome\"),\n",
    "    f.col(\"POS\").alias(\"position\"),\n",
    "    f.col(\"REF\").alias(\"referenceAllele\"),\n",
    "    f.col(\"ALT\").alias(\"alternateAllele\"),\n",
    "    f.col(\"all_inv_var_meta_beta\").alias(\"beta\"),\n",
    "    f.col(\"all_inv_var_meta_sebeta\").alias(\"standardError\"),\n",
    "    f.col(\"all_inv_var_meta_p\"),\n",
    "    f.col(\"all_inv_var_meta_mlogp\"),\n",
    "    *pval_cols,\n",
    ").show(truncate=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gentropy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
