
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../ld_index/">
      
      
        <link rel="next" href="../study_locus/">
      
      
      <link rel="icon" href="../../../assets/imgs/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.18">
    
    
      
        <title>Study Index - Open Targets Gentropy</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.66ac8b77.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../css/timeago.css">
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../assets/stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="light-blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#gentropy.dataset.study_index.StudyIndex" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Open Targets Gentropy" class="md-header__button md-logo" aria-label="Open Targets Gentropy" data-md-component="logo">
      
  <img src="../../../assets/imgs/otlogo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Open Targets Gentropy
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Study Index
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="light-blue"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="light-blue"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/opentargets/gentropy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    opentargets/gentropy
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../installation/" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../howto/_howto/" class="md-tabs__link">
        
  
    
  
  How-to

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../development/_development/" class="md-tabs__link">
        
  
    
  
  Development

      </a>
    </li>
  

      
        
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="../../_python_api/" class="md-tabs__link">
        
  
    
  
  Python API

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Open Targets Gentropy" class="md-nav__button md-logo" aria-label="Open Targets Gentropy" data-md-component="logo">
      
  <img src="../../../assets/imgs/otlogo.png" alt="logo">

    </a>
    Open Targets Gentropy
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/opentargets/gentropy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    opentargets/gentropy
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../howto/_howto/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    How-to
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3" id="__nav_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            How-to
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../howto/command_line/_command_line/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Command line
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1">
            <span class="md-nav__icon md-icon"></span>
            Command line
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/command_line/run_step_in_cli/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run step in CLI
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/command_line/run_step_using_config/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run step using YAML config
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../howto/python_api/_python_api/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Python api
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2">
            <span class="md-nav__icon md-icon"></span>
            Python api
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/python_api/a_creating_spark_session/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Creating a Spark Session
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/python_api/b_create_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Create a dataset
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/python_api/c_applying_methods/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Applying methods
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/python_api/d_inspect_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Inspect a dataset
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../development/_development/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Development
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4" id="__nav_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Development
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../development/contributing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Contributing guidelines
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../development/troubleshooting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Troubleshooting
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../_python_api/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Python API
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5" id="__nav_5_label" tabindex="">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Python API
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../common/_common/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Common
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            Common
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../common/genomic_region/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Genomic Region
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../common/types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Literal Types
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_2" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../_datasets/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Datasets
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_2" id="__nav_5_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            Datasets
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../biosample_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Biosample index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../intervals/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Intervals
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_feature_matrix/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Feature Matrix
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_gold_standard/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Gold Standard
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_prediction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Prediction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ld_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LD Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex" class="md-nav__link">
    <span class="md-ellipsis">
      StudyIndex
    </span>
  </a>
  
    <nav class="md-nav" aria-label="StudyIndex">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.aggregate_and_map_ancestries" class="md-nav__link">
    <span class="md-ellipsis">
      aggregate_and_map_ancestries
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.annotate_sumstats_qc" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_sumstats_qc
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.deconvolute_studies" class="md-nav__link">
    <span class="md-ellipsis">
      deconvolute_studies
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.get_QC_column_name" class="md-nav__link">
    <span class="md-ellipsis">
      get_QC_column_name
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.get_QC_mappings" class="md-nav__link">
    <span class="md-ellipsis">
      get_QC_mappings
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.get_schema" class="md-nav__link">
    <span class="md-ellipsis">
      get_schema
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.has_mapped_trait" class="md-nav__link">
    <span class="md-ellipsis">
      has_mapped_trait
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.has_summarystats" class="md-nav__link">
    <span class="md-ellipsis">
      has_summarystats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.is_gwas" class="md-nav__link">
    <span class="md-ellipsis">
      is_gwas
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.is_qtl" class="md-nav__link">
    <span class="md-ellipsis">
      is_qtl
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.is_quality_flagged" class="md-nav__link">
    <span class="md-ellipsis">
      is_quality_flagged
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.study_type_lut" class="md-nav__link">
    <span class="md-ellipsis">
      study_type_lut
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.validate_analysis_flags" class="md-nav__link">
    <span class="md-ellipsis">
      validate_analysis_flags
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.validate_biosample" class="md-nav__link">
    <span class="md-ellipsis">
      validate_biosample
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.validate_disease" class="md-nav__link">
    <span class="md-ellipsis">
      validate_disease
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.validate_study_type" class="md-nav__link">
    <span class="md-ellipsis">
      validate_study_type
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.validate_target" class="md-nav__link">
    <span class="md-ellipsis">
      validate_target
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.validate_unique_study_id" class="md-nav__link">
    <span class="md-ellipsis">
      validate_unique_study_id
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#schema" class="md-nav__link">
    <span class="md-ellipsis">
      Schema
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../study_locus/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Locus
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../study_locus_overlap/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Locus Overlap
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../summary_statistics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../summary_statistics_qc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../target_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Target Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variant_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_2_15" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../l2g_features/_l2g_feature/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    L2g features
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_2_15" id="__nav_5_2_15_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_2_15_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_2_15">
            <span class="md-nav__icon md-icon"></span>
            L2g features
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_features/colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_features/distance/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From distance
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_features/other/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Other features
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_features/vep/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From VEP
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/_datasources/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Datasources
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3" id="__nav_5_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3">
            <span class="md-nav__icon md-icon"></span>
            Datasources
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/biosample_ontologies/_cell_ontology/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Biosample ontologies
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_1" id="__nav_5_3_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_1">
            <span class="md-nav__icon md-icon"></span>
            Biosample ontologies
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/biosample_ontologies/_uberon/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Uberon
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/ensembl/_ensembl/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Ensembl
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_2" id="__nav_5_3_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_2">
            <span class="md-nav__icon md-icon"></span>
            Ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/ensembl/variant_effect_predictor_parser/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant effector parser
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/eqtl_catalogue/_eqtl_catalogue/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Eqtl catalogue
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_3" id="__nav_5_3_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_3">
            <span class="md-nav__icon md-icon"></span>
            Eqtl catalogue
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/eqtl_catalogue/finemapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Fine mapping results
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/eqtl_catalogue/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/eqtl_catalogue/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Stats
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/finngen/_finngen/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Finngen
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_4" id="__nav_5_3_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_4">
            <span class="md-nav__icon md-icon"></span>
            Finngen
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/finngen/finemapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Finemapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/finngen/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/finngen/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/gnomad/_gnomad/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Gnomad
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_5" id="__nav_5_3_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_5">
            <span class="md-nav__icon md-icon"></span>
            Gnomad
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gnomad/gnomad_ld/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LD Matrix
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gnomad/gnomad_variants/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variants
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_6" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/gwas_catalog/_gwas_catalog/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Gwas catalog
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_6" id="__nav_5_3_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_6">
            <span class="md-nav__icon md-icon"></span>
            Gwas catalog
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/associations/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Associations
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/study_splitter/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Splitter
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/summary_statistics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary statistics
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_7" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/intervals/_intervals/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Intervals
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_7" id="__nav_5_3_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_7">
            <span class="md-nav__icon md-icon"></span>
            Intervals
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/andersson/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Andersson et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/javierre/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Javierre et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/jung/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Jung et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/thurman/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Thurman et al.
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/open_targets/_open_targets/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Open targets
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_8" id="__nav_5_3_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_8">
            <span class="md-nav__icon md-icon"></span>
            Open targets
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/open_targets/l2g_gold_standard/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Gold Standard
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_9" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/ukb_ppp_eur/_ukb_ppp_eur/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Ukb ppp eur
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_9" id="__nav_5_3_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_9">
            <span class="md-nav__icon md-icon"></span>
            Ukb ppp eur
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/ukb_ppp_eur/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/ukb_ppp_eur/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../methods/_methods/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Methods
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4" id="__nav_5_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4">
            <span class="md-nav__icon md-icon"></span>
            Methods
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/carma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CARMA
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/coloc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Coloc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/ecaviar/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    eCAVIAR
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/ld_annotator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LDAnnotator
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/pics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PICS
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/sumstat_imputation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics Imputation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/sumstat_quality_controls/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    QC of GWAS Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/susie_inf/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SuSiE-inf
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4_10" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../methods/l2g/_l2g/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    L2g
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4_10" id="__nav_5_4_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_4_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4_10">
            <span class="md-nav__icon md-icon"></span>
            L2g
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/l2g/feature_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Feature Factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/l2g/model/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/l2g/trainer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Trainer
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../steps/_steps/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Steps
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_5" id="__nav_5_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_5">
            <span class="md-nav__icon md-icon"></span>
            Steps
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/biosample_index_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    biosample_index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/credible_set_qc_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    credible_set_qc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/eqtl_catalogue/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    eQTL Catalogue
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/finngen_studies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    finngen_studies
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/finngen_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    finngen_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_curation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_study_curation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_study_inclusion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_top_hits/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GWAS Catalog Top Hits Ingestion Step
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/l2g/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Locus to Gene (L2G)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/ld_clump/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ld_based_clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/ld_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GnomAD Linkage data ingestion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/locus_breaker_clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    locus_breaker_clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/pics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    pics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/study_locus_validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study-Locus Validation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/study_validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Validation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/summary_statistics_qc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    summary_statistics_qc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/ukb_ppp_eur_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ukb_ppp_eur_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/variant_annotation_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GnomAD variant data ingestion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/variant_index_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    variant_index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/window_based_clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    window_based_clumping
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex" class="md-nav__link">
    <span class="md-ellipsis">
      StudyIndex
    </span>
  </a>
  
    <nav class="md-nav" aria-label="StudyIndex">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.aggregate_and_map_ancestries" class="md-nav__link">
    <span class="md-ellipsis">
      aggregate_and_map_ancestries
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.annotate_sumstats_qc" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_sumstats_qc
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.deconvolute_studies" class="md-nav__link">
    <span class="md-ellipsis">
      deconvolute_studies
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.get_QC_column_name" class="md-nav__link">
    <span class="md-ellipsis">
      get_QC_column_name
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.get_QC_mappings" class="md-nav__link">
    <span class="md-ellipsis">
      get_QC_mappings
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.get_schema" class="md-nav__link">
    <span class="md-ellipsis">
      get_schema
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.has_mapped_trait" class="md-nav__link">
    <span class="md-ellipsis">
      has_mapped_trait
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.has_summarystats" class="md-nav__link">
    <span class="md-ellipsis">
      has_summarystats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.is_gwas" class="md-nav__link">
    <span class="md-ellipsis">
      is_gwas
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.is_qtl" class="md-nav__link">
    <span class="md-ellipsis">
      is_qtl
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.is_quality_flagged" class="md-nav__link">
    <span class="md-ellipsis">
      is_quality_flagged
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.study_type_lut" class="md-nav__link">
    <span class="md-ellipsis">
      study_type_lut
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.validate_analysis_flags" class="md-nav__link">
    <span class="md-ellipsis">
      validate_analysis_flags
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.validate_biosample" class="md-nav__link">
    <span class="md-ellipsis">
      validate_biosample
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.validate_disease" class="md-nav__link">
    <span class="md-ellipsis">
      validate_disease
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.validate_study_type" class="md-nav__link">
    <span class="md-ellipsis">
      validate_study_type
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.validate_target" class="md-nav__link">
    <span class="md-ellipsis">
      validate_target
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_index.StudyIndex.validate_unique_study_id" class="md-nav__link">
    <span class="md-ellipsis">
      validate_unique_study_id
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#schema" class="md-nav__link">
    <span class="md-ellipsis">
      Schema
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>Study Index</h1>

<div class="doc doc-object doc-class">



<h2 id="gentropy.dataset.study_index.StudyIndex" class="doc doc-heading">
            <code>gentropy.dataset.study_index.StudyIndex</code>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-dataclass"><code>dataclass</code></small>
  </span>

<a href="#gentropy.dataset.study_index.StudyIndex" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="gentropy.dataset.dataset.Dataset" href="../_datasets/#gentropy.dataset.dataset.Dataset">Dataset</a></code></p>


        <p>Study index dataset.</p>
<p>A study index dataset captures all the metadata for all studies including GWAS and Molecular QTL.</p>






              <details class="quote">
                <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@dataclass</span>
<span class="k">class</span><span class="w"> </span><span class="nc">StudyIndex</span><span class="p">(</span><span class="n">Dataset</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Study index dataset.</span>

<span class="sd">    A study index dataset captures all the metadata for all studies including GWAS and Molecular QTL.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">VALID_TYPES</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;gwas&quot;</span><span class="p">,</span>
        <span class="s2">&quot;eqtl&quot;</span><span class="p">,</span>
        <span class="s2">&quot;pqtl&quot;</span><span class="p">,</span>
        <span class="s2">&quot;sqtl&quot;</span><span class="p">,</span>
        <span class="s2">&quot;tuqtl&quot;</span><span class="p">,</span>
        <span class="s2">&quot;sceqtl&quot;</span><span class="p">,</span>
        <span class="s2">&quot;scpqtl&quot;</span><span class="p">,</span>
        <span class="s2">&quot;scsqtl&quot;</span><span class="p">,</span>
        <span class="s2">&quot;sctuqtl&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_aggregate_samples_by_ancestry</span><span class="p">(</span><span class="n">merged</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">ancestry</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Aggregate sample counts by ancestry in a list of struct colmns.</span>

<span class="sd">        Args:</span>
<span class="sd">            merged (Column): A column representing merged data (list of structs).</span>
<span class="sd">            ancestry (Column): The `ancestry` parameter is a column that represents the ancestry of each</span>
<span class="sd">                sample. (a struct)</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: the modified &quot;merged&quot; column after aggregating the samples by ancestry.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Iterating over the list of ancestries and adding the sample size if label matches:</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
            <span class="n">merged</span><span class="p">,</span>
            <span class="k">lambda</span> <span class="n">a</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="n">a</span><span class="o">.</span><span class="n">ancestry</span> <span class="o">==</span> <span class="n">ancestry</span><span class="o">.</span><span class="n">ancestry</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">a</span><span class="o">.</span><span class="n">ancestry</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;ancestry&quot;</span><span class="p">),</span>
                    <span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">sampleSize</span> <span class="o">+</span> <span class="n">ancestry</span><span class="o">.</span><span class="n">sampleSize</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sampleSize&quot;</span><span class="p">),</span>
                <span class="p">),</span>
            <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">a</span><span class="p">),</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_map_ancestries_to_ld_population</span><span class="p">(</span><span class="n">gwas_ancestry_label</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Normalise ancestry column from GWAS studies into reference LD panel based on a pre-defined map.</span>

<span class="sd">        This function assumes all possible ancestry categories have a corresponding</span>
<span class="sd">        LD panel in the LD index. It is very important to have the ancestry labels</span>
<span class="sd">        moved to the LD panel map.</span>

<span class="sd">        Args:</span>
<span class="sd">            gwas_ancestry_label (Column): A struct column with ancestry label like Finnish,</span>
<span class="sd">                European, African etc. and the corresponding sample size.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Struct column with the mapped LD population label and the sample size.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Loading ancestry label to LD population label:</span>
        <span class="n">json_dict</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span>
            <span class="n">pkg_resources</span><span class="o">.</span><span class="n">read_text</span><span class="p">(</span>
                <span class="n">data</span><span class="p">,</span> <span class="s2">&quot;gwas_population_2_LD_panel_map.json&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">map_expr</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">create_map</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">(</span><span class="o">*</span><span class="n">json_dict</span><span class="o">.</span><span class="n">items</span><span class="p">())])</span>

        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">map_expr</span><span class="p">[</span><span class="n">gwas_ancestry_label</span><span class="o">.</span><span class="n">ancestry</span><span class="p">]</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;ancestry&quot;</span><span class="p">),</span>
            <span class="n">gwas_ancestry_label</span><span class="o">.</span><span class="n">sampleSize</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sampleSize&quot;</span><span class="p">),</span>
        <span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_schema</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyIndex</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">StructType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Provide the schema for the StudyIndex dataset.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StructType: The schema of the StudyIndex dataset.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">parse_spark_schema</span><span class="p">(</span><span class="s2">&quot;study_index.json&quot;</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_QC_column_name</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyIndex</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the name of the quality control column.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: The name of the quality control column.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;qualityControls&quot;</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_QC_mappings</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyIndex</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Quality control flag to QC column category mappings.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict[str, str]: Mapping between flag name and QC column category value.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">member</span><span class="o">.</span><span class="n">name</span><span class="p">:</span> <span class="n">member</span><span class="o">.</span><span class="n">value</span> <span class="k">for</span> <span class="n">member</span> <span class="ow">in</span> <span class="n">StudyQualityCheck</span><span class="p">}</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">aggregate_and_map_ancestries</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyIndex</span><span class="p">],</span> <span class="n">discovery_samples</span><span class="p">:</span> <span class="n">Column</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Map ancestries to populations in the LD reference and calculate relative sample size.</span>

<span class="sd">        Args:</span>
<span class="sd">            discovery_samples (Column): A list of struct column. Has an `ancestry` column and a `sampleSize` columns</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: A list of struct with mapped LD population and their relative sample size.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Map ancestry categories to population labels of the LD index:</span>
        <span class="n">mapped_ancestries</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
            <span class="n">discovery_samples</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_map_ancestries_to_ld_population</span>
        <span class="p">)</span>

        <span class="c1"># Aggregate sample sizes belonging to the same LD population:</span>
        <span class="n">aggregated_counts</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
            <span class="n">mapped_ancestries</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">array_distinct</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                    <span class="n">mapped_ancestries</span><span class="p">,</span>
                    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                        <span class="n">x</span><span class="o">.</span><span class="n">ancestry</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;ancestry&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sampleSize&quot;</span><span class="p">)</span>
                    <span class="p">),</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="bp">cls</span><span class="o">.</span><span class="n">_aggregate_samples_by_ancestry</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># Getting total sample count:</span>
        <span class="n">total_sample_count</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
            <span class="n">aggregated_counts</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">0.0</span><span class="p">),</span> <span class="k">lambda</span> <span class="n">total</span><span class="p">,</span> <span class="n">pop</span><span class="p">:</span> <span class="n">total</span> <span class="o">+</span> <span class="n">pop</span><span class="o">.</span><span class="n">sampleSize</span>
        <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sampleSize&quot;</span><span class="p">)</span>
        <span class="c1"># Calculating relative sample size for each LD population:</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
            <span class="n">aggregated_counts</span><span class="p">,</span>
            <span class="k">lambda</span> <span class="n">ld_population</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">ld_population</span><span class="o">.</span><span class="n">ancestry</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;ldPopulation&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="n">ld_population</span><span class="o">.</span><span class="n">sampleSize</span> <span class="o">/</span> <span class="n">total_sample_count</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                    <span class="s2">&quot;relativeSampleSize&quot;</span>
                <span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">study_type_lut</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return a lookup table of study type.</span>

<span class="sd">        Returns:</span>
<span class="sd">            DataFrame: A dataframe containing `studyId` and `studyType` columns.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;studyType&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">is_qtl</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return a boolean column with true values for QTL studies.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: True if the study is a QTL study.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">studyType</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;qtl&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">is_gwas</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return a boolean column with true values for GWAS studies.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: True if the study is a GWAS study.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">studyType</span> <span class="o">==</span> <span class="s2">&quot;gwas&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">has_mapped_trait</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return a boolean column indicating if a study has mapped disease.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: True if the study has mapped disease.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">traitFromSourceMappedIds</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">is_quality_flagged</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return a boolean column indicating if a study is flagged due to quality issues.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: True if the study is flagged.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Testing for the presence of the qualityControls column:</span>
        <span class="k">if</span> <span class="s2">&quot;qualityControls&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">])</span> <span class="o">!=</span> <span class="mi">0</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">has_summarystats</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return a boolean column indicating if a study has harmonized summary statistics.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: True if the study has harmonized summary statistics.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">hasSumstats</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validate_unique_study_id</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Validating the uniqueness of study identifiers and flagging duplicated studies.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyIndex: with flagged duplicated studies.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">flag_duplicates</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)),</span>
                    <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">DUPLICATED_STUDY</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_normalise_disease</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span>
        <span class="n">source_disease_column_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">disease_column_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">disease_map</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Normalising diseases in the study index.</span>

<span class="sd">        Given a reference disease map (containing all potential EFO ids with the corresponding reference disease ids),</span>
<span class="sd">        this function maps all EFO ids in the study index to the reference disease ids.</span>

<span class="sd">        Args:</span>
<span class="sd">            source_disease_column_name (str): The column name of the disease column to validate.</span>
<span class="sd">            disease_column_name (str): The resulting disease column name that contains the validated ids.</span>
<span class="sd">            disease_map (DataFrame): Reference dataframe with diseases</span>

<span class="sd">        Returns:</span>
<span class="sd">            DataFrame: where the newly added diseaseIds column will contain the validated EFO identifiers.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span>
            <span class="c1"># Only validating studies with diseases:</span>
            <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">source_disease_column_name</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
            <span class="c1"># Explode disease column:</span>
            <span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
                <span class="s2">&quot;studyType&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">explode_outer</span><span class="p">(</span><span class="n">source_disease_column_name</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;efo&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Join disease map:</span>
            <span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">disease_map</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;efo&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;diseaseId&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="n">disease_column_name</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validate_disease</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span> <span class="n">disease_map</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Validate diseases in the study index dataset.</span>

<span class="sd">        Args:</span>
<span class="sd">            disease_map (DataFrame): a dataframe with two columns (efo, diseaseId).</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyIndex: where gwas studies are flagged where no valid disease id could be found.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Because the disease ids are not mandatory fields of the schema, we skip vaildation if these columns are not present:</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;traitFromSourceMappedIds&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span>
            <span class="s2">&quot;backgroundTraitFromSourceMappedIds&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
        <span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span>

        <span class="c1"># Disease Column names:</span>
        <span class="n">foreground_disease_column</span> <span class="o">=</span> <span class="s2">&quot;diseaseIds&quot;</span>
        <span class="n">background_disease_column</span> <span class="o">=</span> <span class="s2">&quot;backgroundDiseaseIds&quot;</span>

        <span class="c1"># If diseaseId in schema, we need to drop it:</span>
        <span class="n">drop_columns</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">column</span>
            <span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
            <span class="k">if</span> <span class="n">column</span> <span class="ow">in</span> <span class="p">[</span><span class="n">foreground_disease_column</span><span class="p">,</span> <span class="n">background_disease_column</span><span class="p">]</span>
        <span class="p">]</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">drop_columns</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="o">*</span><span class="n">drop_columns</span><span class="p">)</span>

        <span class="c1"># Normalise disease:</span>
        <span class="n">normalised_disease</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalise_disease</span><span class="p">(</span>
            <span class="s2">&quot;traitFromSourceMappedIds&quot;</span><span class="p">,</span> <span class="n">foreground_disease_column</span><span class="p">,</span> <span class="n">disease_map</span>
        <span class="p">)</span>
        <span class="n">normalised_background_disease</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalise_disease</span><span class="p">(</span>
            <span class="s2">&quot;backgroundTraitFromSourceMappedIds&quot;</span><span class="p">,</span> <span class="n">background_disease_column</span><span class="p">,</span> <span class="n">disease_map</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">normalised_disease</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
                <span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">normalised_background_disease</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
                <span class="c1"># Updating disease columns:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="n">foreground_disease_column</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">foreground_disease_column</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">()</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">foreground_disease_column</span><span class="p">)),</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="n">background_disease_column</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">background_disease_column</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">()</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">background_disease_column</span><span class="p">)),</span>
                <span class="p">)</span>
                <span class="c1"># Flagging gwas studies where no valid disease is avilable:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                    <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                        <span class="c1"># Flagging all gwas studies with no normalised disease:</span>
                        <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">foreground_disease_column</span><span class="p">))</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;gwas&quot;</span><span class="p">),</span>
                        <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">UNRESOLVED_DISEASE</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="c1"># Added to avoid Spark optimisation (see: https://github.com/opentargets/issues/issues/3906#issuecomment-2949299965)</span>
                <span class="o">.</span><span class="n">persist</span><span class="p">()</span>
            <span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validate_study_type</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Validating study type and flag unsupported types.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyIndex: with flagged studies with unsupported type.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">validated_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span>
            <span class="c1"># Flagging unsupported study types:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                        <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;gwas&quot;</span><span class="p">)</span>
                        <span class="o">|</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;qtl&quot;</span><span class="p">),</span>
                        <span class="kc">False</span><span class="p">,</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="kc">True</span><span class="p">),</span>
                    <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">UNKNOWN_STUDY_TYPE</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span><span class="n">_df</span><span class="o">=</span><span class="n">validated_df</span><span class="p">,</span> <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">())</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validate_target</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span> <span class="n">target_index</span><span class="p">:</span> <span class="n">TargetIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Validating gene identifiers in the study index against the provided target index.</span>

<span class="sd">        Args:</span>
<span class="sd">            target_index (TargetIndex): target index containing the reference gene identifiers (Ensembl gene identifiers).</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyIndex: with flagged studies if geneId could not be validated.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">gene_set</span> <span class="o">=</span> <span class="n">target_index</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;geneId&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># As the geneId is not a mandatory field of study index, we return if the column is not there:</span>
        <span class="k">if</span> <span class="s2">&quot;geneId&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>

        <span class="n">validated_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_set</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;geneId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;isIdFound&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;gwas&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">),</span>
                <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">)),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">),</span>
                    <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">UNRESOLVED_TARGET</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span><span class="n">_df</span><span class="o">=</span><span class="n">validated_df</span><span class="p">,</span> <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">())</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validate_biosample</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span> <span class="n">biosample_index</span><span class="p">:</span> <span class="n">BiosampleIndex</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Validating biosample identifiers in the study index against the provided biosample index.</span>

<span class="sd">        Args:</span>
<span class="sd">            biosample_index (BiosampleIndex): Biosample index containing a reference of biosample identifiers e.g. cell types, tissues, cell lines, etc.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyIndex: where non-gwas studies are flagged if biosampleIndex could not be validated.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">biosample_set</span> <span class="o">=</span> <span class="n">biosample_index</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="s2">&quot;biosampleId&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># If biosampleId in df, we need to drop it:</span>
        <span class="k">if</span> <span class="s2">&quot;biosampleId&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;biosampleId&quot;</span><span class="p">)</span>

        <span class="c1"># As the biosampleFromSourceId is not a mandatory field of study index, we return if the column is not there:</span>
        <span class="k">if</span> <span class="s2">&quot;biosampleFromSourceId&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>

        <span class="n">validated_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">biosample_set</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">biosampleFromSourceId</span> <span class="o">==</span> <span class="n">biosample_set</span><span class="o">.</span><span class="n">biosampleId</span><span class="p">,</span>
                <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;isIdFound&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;gwas&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">()),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">),</span>
                <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">)),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">),</span>
                    <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">UNKNOWN_BIOSAMPLE</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span><span class="n">_df</span><span class="o">=</span><span class="n">validated_df</span><span class="p">,</span> <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">())</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">annotate_sumstats_qc</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span>
        <span class="n">sumstats_qc</span><span class="p">:</span> <span class="n">SummaryStatisticsQC</span><span class="p">,</span>
        <span class="n">threshold_mean_beta</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
        <span class="n">threshold_mean_diff_pz</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
        <span class="n">threshold_se_diff_pz</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
        <span class="n">threshold_min_gc_lambda</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.7</span><span class="p">,</span>
        <span class="n">threshold_max_gc_lambda</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">2.5</span><span class="p">,</span>
        <span class="n">threshold_min_n_variants</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2_000_000</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Annotate summary stats QC information.</span>

<span class="sd">        Args:</span>
<span class="sd">            sumstats_qc (SummaryStatisticsQC): Dataset containing summary statistics-based quality controls.</span>
<span class="sd">            threshold_mean_beta (float): Threshold for mean beta check. Defaults to 0.05.</span>
<span class="sd">            threshold_mean_diff_pz (float): Threshold for mean diff PZ check. Defaults to 0.05.</span>
<span class="sd">            threshold_se_diff_pz (float): Threshold for SE diff PZ check. Defaults to 0.05.</span>
<span class="sd">            threshold_min_gc_lambda (float): Minimum threshold for GC lambda check. Defaults to 0.7.</span>
<span class="sd">            threshold_max_gc_lambda (float): Maximum threshold for GC lambda check. Defaults to 2.5.</span>
<span class="sd">            threshold_min_n_variants (int): Minimum number of variants for SuSiE check. Defaults to 2_000_000.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyIndex: Updated study index with QC information</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># convert all columns in sumstats_qc dataframe in array of structs grouped by studyId</span>
        <span class="n">qc_check_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">sumstats_qc</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="s2">&quot;studyId&quot;</span><span class="p">]</span>

        <span class="n">studies</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span>

        <span class="c1"># Prepare the QC flags format for the study index:</span>
        <span class="n">melted_qc</span> <span class="o">=</span> <span class="n">convert_from_wide_to_long</span><span class="p">(</span>
            <span class="n">sumstats_qc</span><span class="o">.</span><span class="n">df</span><span class="p">,</span>
            <span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;studyId&quot;</span><span class="p">],</span>
            <span class="n">value_vars</span><span class="o">=</span><span class="n">qc_check_cols</span><span class="p">,</span>
            <span class="n">var_name</span><span class="o">=</span><span class="s2">&quot;QCCheckName&quot;</span><span class="p">,</span>
            <span class="n">value_name</span><span class="o">=</span><span class="s2">&quot;QCCheckValue&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">qc_struct</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;QCCheckName&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;QCCheckValue&quot;</span><span class="p">))</span>
        <span class="n">qc_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">melted_qc</span><span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">collect_list</span><span class="p">(</span><span class="n">qc_struct</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sumstatQCValues&quot;</span><span class="p">))</span>
            <span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;sumstatQCValues&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
        <span class="p">)</span>
        <span class="n">extract_qc_value</span><span class="p">:</span> <span class="n">Callable</span><span class="p">[[</span><span class="nb">str</span><span class="p">],</span> <span class="n">Column</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">filter_array_struct</span><span class="p">(</span>
            <span class="s2">&quot;sumstatQCValues&quot;</span><span class="p">,</span> <span class="s2">&quot;QCCheckName&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="s2">&quot;QCCheckValue&quot;</span>
        <span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">studies</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;sumstatQCValues&quot;</span><span class="p">,</span> <span class="s2">&quot;hasSumstats&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">qc_df</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">)))</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">),</span>
                    <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">SUMSTATS_NOT_AVAILABLE</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="o">~</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">extract_qc_value</span><span class="p">(</span><span class="s2">&quot;mean_beta&quot;</span><span class="p">))</span> <span class="o">&lt;=</span> <span class="n">threshold_mean_beta</span><span class="p">),</span>
                    <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">FAILED_MEAN_BETA_CHECK</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="o">~</span><span class="p">(</span>
                        <span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">extract_qc_value</span><span class="p">(</span><span class="s2">&quot;mean_diff_pz&quot;</span><span class="p">))</span>
                            <span class="o">&lt;=</span> <span class="n">threshold_mean_diff_pz</span>
                        <span class="p">)</span>
                        <span class="o">&amp;</span> <span class="p">(</span><span class="n">extract_qc_value</span><span class="p">(</span><span class="s2">&quot;se_diff_pz&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">threshold_se_diff_pz</span><span class="p">)</span>
                    <span class="p">),</span>
                    <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">FAILED_PZ_CHECK</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="o">~</span><span class="p">(</span>
                        <span class="p">(</span><span class="n">extract_qc_value</span><span class="p">(</span><span class="s2">&quot;gc_lambda&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">threshold_max_gc_lambda</span><span class="p">)</span>
                        <span class="o">&amp;</span> <span class="p">(</span><span class="n">extract_qc_value</span><span class="p">(</span><span class="s2">&quot;gc_lambda&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">threshold_min_gc_lambda</span><span class="p">)</span>
                    <span class="p">),</span>
                    <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">FAILED_GC_LAMBDA_CHECK</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="n">extract_qc_value</span><span class="p">(</span><span class="s2">&quot;n_variants&quot;</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">threshold_min_n_variants</span><span class="p">,</span>
                    <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">SMALL_NUMBER_OF_SNPS</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># Annotate study index with QC information:</span>
        <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="n">df</span><span class="p">,</span>
            <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validate_analysis_flags</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Validating analysis flags in the study index dataset.</span>

<span class="sd">        This method checks the analysis flags and reports the the outcome of predicates to the quality control column.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyIndex: with qualityControls column flagged based on analysisFlags.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">predicate</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
            <span class="s2">&quot;analysisFlags&quot;</span><span class="p">,</span> <span class="n">StudyAnalysisFlag</span><span class="o">.</span><span class="n">CASE_CASE_STUDY</span><span class="o">.</span><span class="n">value</span>
        <span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="n">predicate</span><span class="p">,</span>
                <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">CASE_CASE_STUDY_DESIGN</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span><span class="n">_df</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">())</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">deconvolute_studies</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Deconvolute the study index dataset.</span>

<span class="sd">        When ingesting the study index dataset, the same studyId might be ingested from more than one source.</span>
<span class="sd">        In such cases, the data needs to be merged and the quality control flags need to be combined.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyIndex: Deconvoluted study index dataset.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Windowing by study ID assume random order, but this is OK, because we are not selecting rows by a specific order.</span>
        <span class="n">study_id_window</span> <span class="o">=</span> <span class="n">Window</span><span class="o">.</span><span class="n">partitionBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">orderBy</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">rand</span><span class="p">())</span>

        <span class="c1"># For certain aggregation, the full window is needed to be considered:</span>
        <span class="n">full_study_id_window</span> <span class="o">=</span> <span class="n">study_id_window</span><span class="o">.</span><span class="n">orderBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rangeBetween</span><span class="p">(</span>
            <span class="n">Window</span><span class="o">.</span><span class="n">unboundedPreceding</span><span class="p">,</span> <span class="n">Window</span><span class="o">.</span><span class="n">unboundedFollowing</span>
        <span class="p">)</span>

        <span class="c1"># Temporary columns to drop at the end:</span>
        <span class="n">columns_to_drop</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;keepTopHit&quot;</span><span class="p">,</span> <span class="s2">&quot;mostGranular&quot;</span><span class="p">,</span> <span class="s2">&quot;rank&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span>
                <span class="c1"># Initialising quality controls column, if not present:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">()</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">ArrayType</span><span class="p">(</span><span class="n">StringType</span><span class="p">())),</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)),</span>
                <span class="p">)</span>
                <span class="c1"># Keeping top hit studies unless the same study is available from a summmary statistics source:</span>
                <span class="c1"># This value will be set for all rows for the same `studyId`:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;keepTopHit&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">over</span><span class="p">(</span>
                                <span class="n">full_study_id_window</span>
                            <span class="p">),</span>
                            <span class="kc">True</span><span class="p">,</span>
                        <span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">),</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="kc">True</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="c1"># For studies without summary statistics, we remove the &quot;Not curated by Open Targets&quot; flag:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                        <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">array_remove</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                            <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">NO_OT_CURATION</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                        <span class="p">),</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)),</span>
                <span class="p">)</span>
                <span class="c1"># If top hits are not kept, we remove the &quot;sumstats not available&quot; flag from all QC lists:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                        <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;keepTopHit&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">array_remove</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                            <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">SUMSTATS_NOT_AVAILABLE</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                        <span class="p">),</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)),</span>
                <span class="p">)</span>
                <span class="c1"># Then propagate quality checks for all sources of the same study:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array_distinct</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">full_study_id_window</span><span class="p">)</span>
                        <span class="p">)</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="c1"># Propagating sumstatQCValues -&gt; map, cannot be flatten:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;sumstatQCValues&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">first</span><span class="p">(</span><span class="s2">&quot;sumstatQCValues&quot;</span><span class="p">,</span> <span class="n">ignorenulls</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span>
                        <span class="n">full_study_id_window</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="c1"># Propagating analysisFlags:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;analysisFlags&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">collect_list</span><span class="p">(</span><span class="s2">&quot;analysisFlags&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">full_study_id_window</span><span class="p">)</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="c1"># Propagating hasSumstatsFlag - if no flag, leave null:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;hasSumstats&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                        <span class="c1"># There&#39;s a true:</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">full_study_id_window</span><span class="p">),</span>
                            <span class="kc">True</span><span class="p">,</span>
                        <span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">),</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                        <span class="c1"># There&#39;s a false:</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">full_study_id_window</span><span class="p">),</span>
                            <span class="kc">False</span><span class="p">,</span>
                        <span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">),</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="c1"># Propagating disease: when different sets of diseases available for the same study,</span>
                <span class="c1"># we pick the shortest list, becasuse we assume, that is the most accurate disease assignment:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;mostGranular&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;traitFromSourceMappedIds&quot;</span><span class="p">))</span>
                    <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;traitFromSourceMappedIds&quot;</span><span class="p">)))</span><span class="o">.</span><span class="n">over</span><span class="p">(</span>
                        <span class="n">full_study_id_window</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="c1"># Remove less granular disease mappings:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;traitFromSourceMappedIds&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;mostGranular&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;traitFromSourceMappedIds&quot;</span><span class="p">)),</span>
                <span class="p">)</span>
                <span class="c1"># Propagate mapped disease:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;traitFromSourceMappedIds&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">last</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;traitFromSourceMappedIds&quot;</span><span class="p">),</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span>
                        <span class="n">full_study_id_window</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="c1"># Repeating these steps for the `traitFromSource` column:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;traitFromSource&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;mostGranular&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;traitFromSource&quot;</span><span class="p">)),</span>
                <span class="p">)</span>
                <span class="c1"># Propagate disease:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;traitFromSource&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">last</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;traitFromSource&quot;</span><span class="p">),</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">full_study_id_window</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="c1"># Distinct study types are joined together into a string. So, if there&#39;s ambiguite, the study will be flagged when the study type is validated:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;studyType&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">concat_ws</span><span class="p">(</span>
                        <span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">full_study_id_window</span><span class="p">)</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="c1"># At this point, all studies in one window is expected to be identical. Let&#39;s just pick one:</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;rank&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">row_number</span><span class="p">()</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">study_id_window</span><span class="p">))</span>
                <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;rank&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
                <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="o">*</span><span class="n">columns_to_drop</span><span class="p">)</span>
                <span class="c1"># Added to avoid Spark optimisation (see: https://github.com/opentargets/issues/issues/3906#issuecomment-2949299965)</span>
                <span class="o">.</span><span class="n">persist</span><span class="p">()</span>
            <span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.aggregate_and_map_ancestries" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">aggregate_and_map_ancestries</span><span class="p">(</span><span class="n">discovery_samples</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.dataset.study_index.StudyIndex.aggregate_and_map_ancestries" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Map ancestries to populations in the LD reference and calculate relative sample size.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>discovery_samples</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of struct column. Has an <code>ancestry</code> column and a <code>sampleSize</code> columns</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of struct with mapped LD population and their relative sample size.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">aggregate_and_map_ancestries</span><span class="p">(</span>
    <span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyIndex</span><span class="p">],</span> <span class="n">discovery_samples</span><span class="p">:</span> <span class="n">Column</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Map ancestries to populations in the LD reference and calculate relative sample size.</span>

<span class="sd">    Args:</span>
<span class="sd">        discovery_samples (Column): A list of struct column. Has an `ancestry` column and a `sampleSize` columns</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: A list of struct with mapped LD population and their relative sample size.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Map ancestry categories to population labels of the LD index:</span>
    <span class="n">mapped_ancestries</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
        <span class="n">discovery_samples</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_map_ancestries_to_ld_population</span>
    <span class="p">)</span>

    <span class="c1"># Aggregate sample sizes belonging to the same LD population:</span>
    <span class="n">aggregated_counts</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="n">mapped_ancestries</span><span class="p">,</span>
        <span class="n">f</span><span class="o">.</span><span class="n">array_distinct</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                <span class="n">mapped_ancestries</span><span class="p">,</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">x</span><span class="o">.</span><span class="n">ancestry</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;ancestry&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sampleSize&quot;</span><span class="p">)</span>
                <span class="p">),</span>
            <span class="p">)</span>
        <span class="p">),</span>
        <span class="bp">cls</span><span class="o">.</span><span class="n">_aggregate_samples_by_ancestry</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># Getting total sample count:</span>
    <span class="n">total_sample_count</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="n">aggregated_counts</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">0.0</span><span class="p">),</span> <span class="k">lambda</span> <span class="n">total</span><span class="p">,</span> <span class="n">pop</span><span class="p">:</span> <span class="n">total</span> <span class="o">+</span> <span class="n">pop</span><span class="o">.</span><span class="n">sampleSize</span>
    <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sampleSize&quot;</span><span class="p">)</span>
    <span class="c1"># Calculating relative sample size for each LD population:</span>
    <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
        <span class="n">aggregated_counts</span><span class="p">,</span>
        <span class="k">lambda</span> <span class="n">ld_population</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">ld_population</span><span class="o">.</span><span class="n">ancestry</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;ldPopulation&quot;</span><span class="p">),</span>
            <span class="p">(</span><span class="n">ld_population</span><span class="o">.</span><span class="n">sampleSize</span> <span class="o">/</span> <span class="n">total_sample_count</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                <span class="s2">&quot;relativeSampleSize&quot;</span>
            <span class="p">),</span>
        <span class="p">),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.annotate_sumstats_qc" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">annotate_sumstats_qc</span><span class="p">(</span><span class="n">sumstats_qc</span><span class="p">:</span> <span class="n">SummaryStatisticsQC</span><span class="p">,</span> <span class="n">threshold_mean_beta</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">threshold_mean_diff_pz</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">threshold_se_diff_pz</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">threshold_min_gc_lambda</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.7</span><span class="p">,</span> <span class="n">threshold_max_gc_lambda</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">2.5</span><span class="p">,</span> <span class="n">threshold_min_n_variants</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2000000</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span></code>

<a href="#gentropy.dataset.study_index.StudyIndex.annotate_sumstats_qc" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Annotate summary stats QC information.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>sumstats_qc</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.summary_statistics_qc.SummaryStatisticsQC" href="../summary_statistics_qc/#gentropy.dataset.summary_statistics_qc.SummaryStatisticsQC">SummaryStatisticsQC</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dataset containing summary statistics-based quality controls.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>threshold_mean_beta</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Threshold for mean beta check. Defaults to 0.05.</p>
              </div>
            </td>
            <td>
                  <code>0.05</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>threshold_mean_diff_pz</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Threshold for mean diff PZ check. Defaults to 0.05.</p>
              </div>
            </td>
            <td>
                  <code>0.05</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>threshold_se_diff_pz</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Threshold for SE diff PZ check. Defaults to 0.05.</p>
              </div>
            </td>
            <td>
                  <code>0.05</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>threshold_min_gc_lambda</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum threshold for GC lambda check. Defaults to 0.7.</p>
              </div>
            </td>
            <td>
                  <code>0.7</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>threshold_max_gc_lambda</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum threshold for GC lambda check. Defaults to 2.5.</p>
              </div>
            </td>
            <td>
                  <code>2.5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>threshold_min_n_variants</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum number of variants for SuSiE check. Defaults to 2_000_000.</p>
              </div>
            </td>
            <td>
                  <code>2000000</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyIndex</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Updated study index with QC information</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">annotate_sumstats_qc</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span>
    <span class="n">sumstats_qc</span><span class="p">:</span> <span class="n">SummaryStatisticsQC</span><span class="p">,</span>
    <span class="n">threshold_mean_beta</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
    <span class="n">threshold_mean_diff_pz</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
    <span class="n">threshold_se_diff_pz</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
    <span class="n">threshold_min_gc_lambda</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.7</span><span class="p">,</span>
    <span class="n">threshold_max_gc_lambda</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">2.5</span><span class="p">,</span>
    <span class="n">threshold_min_n_variants</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2_000_000</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Annotate summary stats QC information.</span>

<span class="sd">    Args:</span>
<span class="sd">        sumstats_qc (SummaryStatisticsQC): Dataset containing summary statistics-based quality controls.</span>
<span class="sd">        threshold_mean_beta (float): Threshold for mean beta check. Defaults to 0.05.</span>
<span class="sd">        threshold_mean_diff_pz (float): Threshold for mean diff PZ check. Defaults to 0.05.</span>
<span class="sd">        threshold_se_diff_pz (float): Threshold for SE diff PZ check. Defaults to 0.05.</span>
<span class="sd">        threshold_min_gc_lambda (float): Minimum threshold for GC lambda check. Defaults to 0.7.</span>
<span class="sd">        threshold_max_gc_lambda (float): Maximum threshold for GC lambda check. Defaults to 2.5.</span>
<span class="sd">        threshold_min_n_variants (int): Minimum number of variants for SuSiE check. Defaults to 2_000_000.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyIndex: Updated study index with QC information</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># convert all columns in sumstats_qc dataframe in array of structs grouped by studyId</span>
    <span class="n">qc_check_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">sumstats_qc</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="s2">&quot;studyId&quot;</span><span class="p">]</span>

    <span class="n">studies</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span>

    <span class="c1"># Prepare the QC flags format for the study index:</span>
    <span class="n">melted_qc</span> <span class="o">=</span> <span class="n">convert_from_wide_to_long</span><span class="p">(</span>
        <span class="n">sumstats_qc</span><span class="o">.</span><span class="n">df</span><span class="p">,</span>
        <span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;studyId&quot;</span><span class="p">],</span>
        <span class="n">value_vars</span><span class="o">=</span><span class="n">qc_check_cols</span><span class="p">,</span>
        <span class="n">var_name</span><span class="o">=</span><span class="s2">&quot;QCCheckName&quot;</span><span class="p">,</span>
        <span class="n">value_name</span><span class="o">=</span><span class="s2">&quot;QCCheckValue&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">qc_struct</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;QCCheckName&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;QCCheckValue&quot;</span><span class="p">))</span>
    <span class="n">qc_df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">melted_qc</span><span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">collect_list</span><span class="p">(</span><span class="n">qc_struct</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sumstatQCValues&quot;</span><span class="p">))</span>
        <span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;sumstatQCValues&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
    <span class="p">)</span>
    <span class="n">extract_qc_value</span><span class="p">:</span> <span class="n">Callable</span><span class="p">[[</span><span class="nb">str</span><span class="p">],</span> <span class="n">Column</span><span class="p">]</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">filter_array_struct</span><span class="p">(</span>
        <span class="s2">&quot;sumstatQCValues&quot;</span><span class="p">,</span> <span class="s2">&quot;QCCheckName&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="s2">&quot;QCCheckValue&quot;</span>
    <span class="p">)</span>

    <span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">studies</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;sumstatQCValues&quot;</span><span class="p">,</span> <span class="s2">&quot;hasSumstats&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">qc_df</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">)))</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">),</span>
                <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">SUMSTATS_NOT_AVAILABLE</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="o">~</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">extract_qc_value</span><span class="p">(</span><span class="s2">&quot;mean_beta&quot;</span><span class="p">))</span> <span class="o">&lt;=</span> <span class="n">threshold_mean_beta</span><span class="p">),</span>
                <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">FAILED_MEAN_BETA_CHECK</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="o">~</span><span class="p">(</span>
                    <span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">extract_qc_value</span><span class="p">(</span><span class="s2">&quot;mean_diff_pz&quot;</span><span class="p">))</span>
                        <span class="o">&lt;=</span> <span class="n">threshold_mean_diff_pz</span>
                    <span class="p">)</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">extract_qc_value</span><span class="p">(</span><span class="s2">&quot;se_diff_pz&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">threshold_se_diff_pz</span><span class="p">)</span>
                <span class="p">),</span>
                <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">FAILED_PZ_CHECK</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="o">~</span><span class="p">(</span>
                    <span class="p">(</span><span class="n">extract_qc_value</span><span class="p">(</span><span class="s2">&quot;gc_lambda&quot;</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">threshold_max_gc_lambda</span><span class="p">)</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">extract_qc_value</span><span class="p">(</span><span class="s2">&quot;gc_lambda&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">threshold_min_gc_lambda</span><span class="p">)</span>
                <span class="p">),</span>
                <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">FAILED_GC_LAMBDA_CHECK</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="n">extract_qc_value</span><span class="p">(</span><span class="s2">&quot;n_variants&quot;</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">threshold_min_n_variants</span><span class="p">,</span>
                <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">SMALL_NUMBER_OF_SNPS</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Annotate study index with QC information:</span>
    <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="n">df</span><span class="p">,</span>
        <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.deconvolute_studies" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">deconvolute_studies</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span></code>

<a href="#gentropy.dataset.study_index.StudyIndex.deconvolute_studies" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Deconvolute the study index dataset.</p>
<p>When ingesting the study index dataset, the same studyId might be ingested from more than one source.
In such cases, the data needs to be merged and the quality control flags need to be combined.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyIndex</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Deconvoluted study index dataset.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">deconvolute_studies</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Deconvolute the study index dataset.</span>

<span class="sd">    When ingesting the study index dataset, the same studyId might be ingested from more than one source.</span>
<span class="sd">    In such cases, the data needs to be merged and the quality control flags need to be combined.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyIndex: Deconvoluted study index dataset.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Windowing by study ID assume random order, but this is OK, because we are not selecting rows by a specific order.</span>
    <span class="n">study_id_window</span> <span class="o">=</span> <span class="n">Window</span><span class="o">.</span><span class="n">partitionBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">orderBy</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">rand</span><span class="p">())</span>

    <span class="c1"># For certain aggregation, the full window is needed to be considered:</span>
    <span class="n">full_study_id_window</span> <span class="o">=</span> <span class="n">study_id_window</span><span class="o">.</span><span class="n">orderBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rangeBetween</span><span class="p">(</span>
        <span class="n">Window</span><span class="o">.</span><span class="n">unboundedPreceding</span><span class="p">,</span> <span class="n">Window</span><span class="o">.</span><span class="n">unboundedFollowing</span>
    <span class="p">)</span>

    <span class="c1"># Temporary columns to drop at the end:</span>
    <span class="n">columns_to_drop</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;keepTopHit&quot;</span><span class="p">,</span> <span class="s2">&quot;mostGranular&quot;</span><span class="p">,</span> <span class="s2">&quot;rank&quot;</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span>
            <span class="c1"># Initialising quality controls column, if not present:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">()</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">ArrayType</span><span class="p">(</span><span class="n">StringType</span><span class="p">())),</span>
                <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)),</span>
            <span class="p">)</span>
            <span class="c1"># Keeping top hit studies unless the same study is available from a summmary statistics source:</span>
            <span class="c1"># This value will be set for all rows for the same `studyId`:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;keepTopHit&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">over</span><span class="p">(</span>
                            <span class="n">full_study_id_window</span>
                        <span class="p">),</span>
                        <span class="kc">True</span><span class="p">,</span>
                    <span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">),</span>
                <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="kc">True</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># For studies without summary statistics, we remove the &quot;Not curated by Open Targets&quot; flag:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array_remove</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                        <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">NO_OT_CURATION</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)),</span>
            <span class="p">)</span>
            <span class="c1"># If top hits are not kept, we remove the &quot;sumstats not available&quot; flag from all QC lists:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;keepTopHit&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array_remove</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                        <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">SUMSTATS_NOT_AVAILABLE</span><span class="o">.</span><span class="n">value</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)),</span>
            <span class="p">)</span>
            <span class="c1"># Then propagate quality checks for all sources of the same study:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">array_distinct</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">full_study_id_window</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Propagating sumstatQCValues -&gt; map, cannot be flatten:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;sumstatQCValues&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">first</span><span class="p">(</span><span class="s2">&quot;sumstatQCValues&quot;</span><span class="p">,</span> <span class="n">ignorenulls</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span>
                    <span class="n">full_study_id_window</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Propagating analysisFlags:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;analysisFlags&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">collect_list</span><span class="p">(</span><span class="s2">&quot;analysisFlags&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">full_study_id_window</span><span class="p">)</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Propagating hasSumstatsFlag - if no flag, leave null:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;hasSumstats&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="c1"># There&#39;s a true:</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">full_study_id_window</span><span class="p">),</span>
                        <span class="kc">True</span><span class="p">,</span>
                    <span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">),</span>
                <span class="p">)</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="c1"># There&#39;s a false:</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="s2">&quot;hasSumstats&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">full_study_id_window</span><span class="p">),</span>
                        <span class="kc">False</span><span class="p">,</span>
                    <span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">),</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Propagating disease: when different sets of diseases available for the same study,</span>
            <span class="c1"># we pick the shortest list, becasuse we assume, that is the most accurate disease assignment:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;mostGranular&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;traitFromSourceMappedIds&quot;</span><span class="p">))</span>
                <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;traitFromSourceMappedIds&quot;</span><span class="p">)))</span><span class="o">.</span><span class="n">over</span><span class="p">(</span>
                    <span class="n">full_study_id_window</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Remove less granular disease mappings:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;traitFromSourceMappedIds&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;mostGranular&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;traitFromSourceMappedIds&quot;</span><span class="p">)),</span>
            <span class="p">)</span>
            <span class="c1"># Propagate mapped disease:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;traitFromSourceMappedIds&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">last</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;traitFromSourceMappedIds&quot;</span><span class="p">),</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span>
                    <span class="n">full_study_id_window</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Repeating these steps for the `traitFromSource` column:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;traitFromSource&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;mostGranular&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;traitFromSource&quot;</span><span class="p">)),</span>
            <span class="p">)</span>
            <span class="c1"># Propagate disease:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;traitFromSource&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">last</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;traitFromSource&quot;</span><span class="p">),</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">full_study_id_window</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Distinct study types are joined together into a string. So, if there&#39;s ambiguite, the study will be flagged when the study type is validated:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;studyType&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">concat_ws</span><span class="p">(</span>
                    <span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">collect_set</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">full_study_id_window</span><span class="p">)</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># At this point, all studies in one window is expected to be identical. Let&#39;s just pick one:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;rank&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">row_number</span><span class="p">()</span><span class="o">.</span><span class="n">over</span><span class="p">(</span><span class="n">study_id_window</span><span class="p">))</span>
            <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;rank&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="o">*</span><span class="n">columns_to_drop</span><span class="p">)</span>
            <span class="c1"># Added to avoid Spark optimisation (see: https://github.com/opentargets/issues/issues/3906#issuecomment-2949299965)</span>
            <span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.get_QC_column_name" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_QC_column_name</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="nb">str</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.dataset.study_index.StudyIndex.get_QC_column_name" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Return the name of the quality control column.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the quality control column.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">get_QC_column_name</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyIndex</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return the name of the quality control column.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: The name of the quality control column.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="s2">&quot;qualityControls&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.get_QC_mappings" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_QC_mappings</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.dataset.study_index.StudyIndex.get_QC_mappings" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Quality control flag to QC column category mappings.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>dict[str, str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>dict[str, str]: Mapping between flag name and QC column category value.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">get_QC_mappings</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyIndex</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Quality control flag to QC column category mappings.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict[str, str]: Mapping between flag name and QC column category value.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">{</span><span class="n">member</span><span class="o">.</span><span class="n">name</span><span class="p">:</span> <span class="n">member</span><span class="o">.</span><span class="n">value</span> <span class="k">for</span> <span class="n">member</span> <span class="ow">in</span> <span class="n">StudyQualityCheck</span><span class="p">}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.get_schema" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_schema</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StructType</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.dataset.study_index.StudyIndex.get_schema" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Provide the schema for the StudyIndex dataset.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StructType</code></td>            <td>
                  <code><span title="pyspark.sql.types.StructType">StructType</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The schema of the StudyIndex dataset.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">get_schema</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyIndex</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">StructType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Provide the schema for the StudyIndex dataset.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StructType: The schema of the StudyIndex dataset.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">parse_spark_schema</span><span class="p">(</span><span class="s2">&quot;study_index.json&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.has_mapped_trait" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">has_mapped_trait</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

<a href="#gentropy.dataset.study_index.StudyIndex.has_mapped_trait" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Return a boolean column indicating if a study has mapped disease.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if the study has mapped disease.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">has_mapped_trait</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return a boolean column indicating if a study has mapped disease.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: True if the study has mapped disease.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">traitFromSourceMappedIds</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.has_summarystats" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">has_summarystats</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

<a href="#gentropy.dataset.study_index.StudyIndex.has_summarystats" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Return a boolean column indicating if a study has harmonized summary statistics.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if the study has harmonized summary statistics.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">has_summarystats</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return a boolean column indicating if a study has harmonized summary statistics.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: True if the study has harmonized summary statistics.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">hasSumstats</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.is_gwas" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">is_gwas</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

<a href="#gentropy.dataset.study_index.StudyIndex.is_gwas" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Return a boolean column with true values for GWAS studies.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if the study is a GWAS study.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">is_gwas</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return a boolean column with true values for GWAS studies.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: True if the study is a GWAS study.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">studyType</span> <span class="o">==</span> <span class="s2">&quot;gwas&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.is_qtl" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">is_qtl</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

<a href="#gentropy.dataset.study_index.StudyIndex.is_qtl" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Return a boolean column with true values for QTL studies.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if the study is a QTL study.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">is_qtl</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return a boolean column with true values for QTL studies.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: True if the study is a QTL study.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">studyType</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;qtl&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.is_quality_flagged" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">is_quality_flagged</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

<a href="#gentropy.dataset.study_index.StudyIndex.is_quality_flagged" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Return a boolean column indicating if a study is flagged due to quality issues.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if the study is flagged.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">is_quality_flagged</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return a boolean column indicating if a study is flagged due to quality issues.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: True if the study is flagged.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Testing for the presence of the qualityControls column:</span>
    <span class="k">if</span> <span class="s2">&quot;qualityControls&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">])</span> <span class="o">!=</span> <span class="mi">0</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.study_type_lut" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">study_type_lut</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span></code>

<a href="#gentropy.dataset.study_index.StudyIndex.study_type_lut" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Return a lookup table of study type.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>DataFrame</code></td>            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dataframe containing <code>studyId</code> and <code>studyType</code> columns.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">study_type_lut</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return a lookup table of study type.</span>

<span class="sd">    Returns:</span>
<span class="sd">        DataFrame: A dataframe containing `studyId` and `studyType` columns.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;studyType&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.validate_analysis_flags" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">validate_analysis_flags</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span></code>

<a href="#gentropy.dataset.study_index.StudyIndex.validate_analysis_flags" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Validating analysis flags in the study index dataset.</p>
<p>This method checks the analysis flags and reports the the outcome of predicates to the quality control column.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyIndex</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>with qualityControls column flagged based on analysisFlags.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate_analysis_flags</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Validating analysis flags in the study index dataset.</span>

<span class="sd">    This method checks the analysis flags and reports the the outcome of predicates to the quality control column.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyIndex: with qualityControls column flagged based on analysisFlags.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">predicate</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
        <span class="s2">&quot;analysisFlags&quot;</span><span class="p">,</span> <span class="n">StudyAnalysisFlag</span><span class="o">.</span><span class="n">CASE_CASE_STUDY</span><span class="o">.</span><span class="n">value</span>
    <span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
        <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
        <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
            <span class="n">predicate</span><span class="p">,</span>
            <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">CASE_CASE_STUDY_DESIGN</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span><span class="n">_df</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">())</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.validate_biosample" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">validate_biosample</span><span class="p">(</span><span class="n">biosample_index</span><span class="p">:</span> <span class="n">BiosampleIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span></code>

<a href="#gentropy.dataset.study_index.StudyIndex.validate_biosample" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Validating biosample identifiers in the study index against the provided biosample index.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>biosample_index</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.biosample_index.BiosampleIndex" href="../biosample_index/#gentropy.dataset.biosample_index.BiosampleIndex">BiosampleIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Biosample index containing a reference of biosample identifiers e.g. cell types, tissues, cell lines, etc.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyIndex</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>where non-gwas studies are flagged if biosampleIndex could not be validated.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate_biosample</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span> <span class="n">biosample_index</span><span class="p">:</span> <span class="n">BiosampleIndex</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Validating biosample identifiers in the study index against the provided biosample index.</span>

<span class="sd">    Args:</span>
<span class="sd">        biosample_index (BiosampleIndex): Biosample index containing a reference of biosample identifiers e.g. cell types, tissues, cell lines, etc.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyIndex: where non-gwas studies are flagged if biosampleIndex could not be validated.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">biosample_set</span> <span class="o">=</span> <span class="n">biosample_index</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="s2">&quot;biosampleId&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># If biosampleId in df, we need to drop it:</span>
    <span class="k">if</span> <span class="s2">&quot;biosampleId&quot;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;biosampleId&quot;</span><span class="p">)</span>

    <span class="c1"># As the biosampleFromSourceId is not a mandatory field of study index, we return if the column is not there:</span>
    <span class="k">if</span> <span class="s2">&quot;biosampleFromSourceId&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="n">validated_df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">biosample_set</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">biosampleFromSourceId</span> <span class="o">==</span> <span class="n">biosample_set</span><span class="o">.</span><span class="n">biosampleId</span><span class="p">,</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;isIdFound&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;gwas&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">()),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">),</span>
            <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">)),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">),</span>
                <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">UNKNOWN_BIOSAMPLE</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span><span class="n">_df</span><span class="o">=</span><span class="n">validated_df</span><span class="p">,</span> <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">())</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.validate_disease" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">validate_disease</span><span class="p">(</span><span class="n">disease_map</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span></code>

<a href="#gentropy.dataset.study_index.StudyIndex.validate_disease" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Validate diseases in the study index dataset.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>disease_map</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>a dataframe with two columns (efo, diseaseId).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyIndex</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>where gwas studies are flagged where no valid disease id could be found.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate_disease</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span> <span class="n">disease_map</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Validate diseases in the study index dataset.</span>

<span class="sd">    Args:</span>
<span class="sd">        disease_map (DataFrame): a dataframe with two columns (efo, diseaseId).</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyIndex: where gwas studies are flagged where no valid disease id could be found.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Because the disease ids are not mandatory fields of the schema, we skip vaildation if these columns are not present:</span>
    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;traitFromSourceMappedIds&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span>
        <span class="s2">&quot;backgroundTraitFromSourceMappedIds&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
    <span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="c1"># Disease Column names:</span>
    <span class="n">foreground_disease_column</span> <span class="o">=</span> <span class="s2">&quot;diseaseIds&quot;</span>
    <span class="n">background_disease_column</span> <span class="o">=</span> <span class="s2">&quot;backgroundDiseaseIds&quot;</span>

    <span class="c1"># If diseaseId in schema, we need to drop it:</span>
    <span class="n">drop_columns</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">column</span>
        <span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span>
        <span class="k">if</span> <span class="n">column</span> <span class="ow">in</span> <span class="p">[</span><span class="n">foreground_disease_column</span><span class="p">,</span> <span class="n">background_disease_column</span><span class="p">]</span>
    <span class="p">]</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">drop_columns</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="o">*</span><span class="n">drop_columns</span><span class="p">)</span>

    <span class="c1"># Normalise disease:</span>
    <span class="n">normalised_disease</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalise_disease</span><span class="p">(</span>
        <span class="s2">&quot;traitFromSourceMappedIds&quot;</span><span class="p">,</span> <span class="n">foreground_disease_column</span><span class="p">,</span> <span class="n">disease_map</span>
    <span class="p">)</span>
    <span class="n">normalised_background_disease</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_normalise_disease</span><span class="p">(</span>
        <span class="s2">&quot;backgroundTraitFromSourceMappedIds&quot;</span><span class="p">,</span> <span class="n">background_disease_column</span><span class="p">,</span> <span class="n">disease_map</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">normalised_disease</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">normalised_background_disease</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
            <span class="c1"># Updating disease columns:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="n">foreground_disease_column</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">foreground_disease_column</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">()</span>
                <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">foreground_disease_column</span><span class="p">)),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="n">background_disease_column</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">background_disease_column</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">()</span>
                <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">background_disease_column</span><span class="p">)),</span>
            <span class="p">)</span>
            <span class="c1"># Flagging gwas studies where no valid disease is avilable:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="c1"># Flagging all gwas studies with no normalised disease:</span>
                    <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">foreground_disease_column</span><span class="p">))</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;gwas&quot;</span><span class="p">),</span>
                    <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">UNRESOLVED_DISEASE</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Added to avoid Spark optimisation (see: https://github.com/opentargets/issues/issues/3906#issuecomment-2949299965)</span>
            <span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.validate_study_type" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">validate_study_type</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span></code>

<a href="#gentropy.dataset.study_index.StudyIndex.validate_study_type" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Validating study type and flag unsupported types.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyIndex</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>with flagged studies with unsupported type.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate_study_type</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Validating study type and flag unsupported types.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyIndex: with flagged studies with unsupported type.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">validated_df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span>
        <span class="c1"># Flagging unsupported study types:</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;gwas&quot;</span><span class="p">)</span>
                    <span class="o">|</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;qtl&quot;</span><span class="p">),</span>
                    <span class="kc">False</span><span class="p">,</span>
                <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="kc">True</span><span class="p">),</span>
                <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">UNKNOWN_STUDY_TYPE</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span><span class="n">_df</span><span class="o">=</span><span class="n">validated_df</span><span class="p">,</span> <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">())</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.validate_target" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">validate_target</span><span class="p">(</span><span class="n">target_index</span><span class="p">:</span> <span class="n">TargetIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span></code>

<a href="#gentropy.dataset.study_index.StudyIndex.validate_target" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Validating gene identifiers in the study index against the provided target index.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>target_index</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.target_index.TargetIndex" href="../target_index/#gentropy.dataset.target_index.TargetIndex">TargetIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>target index containing the reference gene identifiers (Ensembl gene identifiers).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyIndex</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>with flagged studies if geneId could not be validated.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate_target</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span> <span class="n">target_index</span><span class="p">:</span> <span class="n">TargetIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Validating gene identifiers in the study index against the provided target index.</span>

<span class="sd">    Args:</span>
<span class="sd">        target_index (TargetIndex): target index containing the reference gene identifiers (Ensembl gene identifiers).</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyIndex: with flagged studies if geneId could not be validated.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gene_set</span> <span class="o">=</span> <span class="n">target_index</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;geneId&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># As the geneId is not a mandatory field of study index, we return if the column is not there:</span>
    <span class="k">if</span> <span class="s2">&quot;geneId&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="n">validated_df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_set</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;geneId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;isIdFound&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;gwas&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">),</span>
            <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">)),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="n">StudyIndex</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">),</span>
                <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">UNRESOLVED_TARGET</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;isIdFound&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span><span class="n">_df</span><span class="o">=</span><span class="n">validated_df</span><span class="p">,</span> <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">())</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_index.StudyIndex.validate_unique_study_id" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">validate_unique_study_id</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span></code>

<a href="#gentropy.dataset.study_index.StudyIndex.validate_unique_study_id" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Validating the uniqueness of study identifiers and flagging duplicated studies.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyIndex</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>with flagged duplicated studies.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_index.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate_unique_study_id</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyIndex</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Validating the uniqueness of study identifiers and flagging duplicated studies.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyIndex: with flagged duplicated studies.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">StudyIndex</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">flag_duplicates</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)),</span>
                <span class="n">StudyQualityCheck</span><span class="o">.</span><span class="n">DUPLICATED_STUDY</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="n">StudyIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><h2 id="schema">Schema<a class="headerlink" href="#schema" title="Permanent link">&para;</a></h2>
<div class="highlight"><pre><span></span><code>root
 |-- studyId: string (nullable = false)
 |-- projectId: string (nullable = false)
 |-- studyType: string (nullable = false)
 |-- traitFromSource: string (nullable = true)
 |-- traitFromSourceMappedIds: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- diseaseIds: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- geneId: string (nullable = true)
 |-- biosampleFromSourceId: string (nullable = true)
 |-- biosampleId: string (nullable = true)
 |-- pubmedId: string (nullable = true)
 |-- publicationTitle: string (nullable = true)
 |-- publicationFirstAuthor: string (nullable = true)
 |-- publicationDate: string (nullable = true)
 |-- publicationJournal: string (nullable = true)
 |-- backgroundTraitFromSourceMappedIds: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- backgroundDiseaseIds: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- initialSampleSize: string (nullable = true)
 |-- nCases: integer (nullable = true)
 |-- nControls: integer (nullable = true)
 |-- nSamples: integer (nullable = true)
 |-- cohorts: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- ldPopulationStructure: array (nullable = true)
 |    |-- element: struct (containsNull = false)
 |    |    |-- ldPopulation: string (nullable = true)
 |    |    |-- relativeSampleSize: double (nullable = true)
 |-- discoverySamples: array (nullable = true)
 |    |-- element: struct (containsNull = false)
 |    |    |-- sampleSize: integer (nullable = true)
 |    |    |-- ancestry: string (nullable = true)
 |-- replicationSamples: array (nullable = true)
 |    |-- element: struct (containsNull = false)
 |    |    |-- sampleSize: integer (nullable = true)
 |    |    |-- ancestry: string (nullable = true)
 |-- qualityControls: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- analysisFlags: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- summarystatsLocation: string (nullable = true)
 |-- hasSumstats: boolean (nullable = true)
 |-- condition: string (nullable = true)
 |-- sumstatQCValues: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- QCCheckName: string (nullable = false)
 |    |    |-- QCCheckValue: float (nullable = false)
</code></pre></div>
<hr />
<div class="md-source-file">
  <small>
    <ul
      class="metadata page-metadata"
      data-bi-name="page info"
      lang="en-us"
      dir="ltr"
    >
      <!-- mkdocs-git-revision-date-plugin -->
      
      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3h-2Z"/></svg>
        </span>
        <span class="label"> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class="timeago" datetime="2023-01-20T17:37:40+00:00" locale="en"></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2023-01-20</span> </span>
      </li>
       
      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg>
        </span>
        <span class="label"> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class="timeago" datetime="2024-01-17T10:30:44+00:00" locale="en"></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2024-01-17</span> </span>
      </li>
      

      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2A10 10 0 0 0 2 12c0 4.42 2.87 8.17 6.84 9.5.5.08.66-.23.66-.5v-1.69c-2.77.6-3.36-1.34-3.36-1.34-.46-1.16-1.11-1.47-1.11-1.47-.91-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.87 1.52 2.34 1.07 2.91.83.09-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.92 0-1.11.38-2 1.03-2.71-.1-.25-.45-1.29.1-2.64 0 0 .84-.27 2.75 1.02.79-.22 1.65-.33 2.5-.33.85 0 1.71.11 2.5.33 1.91-1.29 2.75-1.02 2.75-1.02.55 1.35.2 2.39.1 2.64.65.71 1.03 1.6 1.03 2.71 0 3.82-2.34 4.66-4.57 4.91.36.31.69.92.69 1.85V21c0 .27.16.59.67.5C19.14 20.16 22 16.42 22 12A10 10 0 0 0 12 2Z"/></svg>
        </span>
        <span class="label"> Contributors </span>
      </li>
      <li class="contributors-holder">
        <ul class="contributors" data-bi-name="contributors"><li>
            <a
              href="https://github.com/d0choa"
              title="d0choa"
              data-bi-name="contributorprofile"
              target="_blank"
              ><img src="https://avatars.githubusercontent.com/u/5097586?v=4" alt="d0choa"
            /></a>
          </li><li>
            <a
              href="https://github.com/web-flow"
              title="web-flow"
              data-bi-name="contributorprofile"
              target="_blank"
              ><img src="https://avatars.githubusercontent.com/u/19864447?v=4" alt="web-flow"
            /></a>
          </li></ul>
      </li>
    </ul>
  </small>
</div>




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/gentropy" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.top", "navigation.tabs", "navigation.tabs.sticky", "content.code.annotate", "search.suggest", "search.highlight"], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.3220b9d7.min.js"></script>
      
        <script src="../../../js/timeago.min.js"></script>
      
        <script src="../../../js/timeago_mkdocs_material.js"></script>
      
        <script src="../../../assets/javascripts/extra.js"></script>
      
    
  </body>
</html>