
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../_ensembl/">
      
      
        <link rel="next" href="../../eqtl_catalogue/_eqtl_catalogue/">
      
      
      <link rel="icon" href="../../../../assets/imgs/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.18">
    
    
      
        <title>Variant effector parser - Open Targets Gentropy</title>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/main.66ac8b77.min.css">
      
        
        <link rel="stylesheet" href="../../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../css/timeago.css">
    
      <link rel="stylesheet" href="../../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="light-blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../.." title="Open Targets Gentropy" class="md-header__button md-logo" aria-label="Open Targets Gentropy" data-md-component="logo">
      
  <img src="../../../../assets/imgs/otlogo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Open Targets Gentropy
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Variant effector parser
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="light-blue"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="light-blue"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/opentargets/gentropy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    opentargets/gentropy
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../.." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../../installation/" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../../howto/_howto/" class="md-tabs__link">
        
  
    
  
  How-to

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../../roadmap/" class="md-tabs__link">
        
  
    
  
  Roadmap

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../../development/_development/" class="md-tabs__link">
        
  
    
  
  Development

      </a>
    </li>
  

      
        
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="../../../_python_api/" class="md-tabs__link">
        
  
    
  
  Python API

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../.." title="Open Targets Gentropy" class="md-nav__button md-logo" aria-label="Open Targets Gentropy" data-md-component="logo">
      
  <img src="../../../../assets/imgs/otlogo.png" alt="logo">

    </a>
    Open Targets Gentropy
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/opentargets/gentropy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    opentargets/gentropy
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../howto/_howto/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    How-to
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3" id="__nav_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            How-to
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../howto/command_line/_command_line/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Command line
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1">
            <span class="md-nav__icon md-icon"></span>
            Command line
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../howto/command_line/run_step_in_cli/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run step in CLI
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../howto/command_line/run_step_using_config/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run step using YAML config
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../howto/python_api/_python_api/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Python api
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2">
            <span class="md-nav__icon md-icon"></span>
            Python api
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../howto/python_api/a_creating_spark_session/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Creating a Spark Session
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../howto/python_api/b_create_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Create a dataset
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../howto/python_api/c_applying_methods/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Applying methods
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../howto/python_api/d_inspect_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Inspect a dataset
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../roadmap/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Roadmap
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../development/_development/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Development
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5" id="__nav_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Development
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../development/contributing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Contributing guidelines
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../development/troubleshooting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Troubleshooting
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../_python_api/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Python API
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6" id="__nav_6_label" tabindex="">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Python API
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../common/_common/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Common
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_1" id="__nav_6_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_1">
            <span class="md-nav__icon md-icon"></span>
            Common
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../common/genomic_region/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Genomic Region
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../common/types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Literal Types
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../datasets/_datasets/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Datasets
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_2" id="__nav_6_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_2">
            <span class="md-nav__icon md-icon"></span>
            Datasets
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/biosample_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Biosample index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/intervals/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Intervals
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/l2g_feature_matrix/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Feature Matrix
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/l2g_gold_standard/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Gold Standard
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/l2g_prediction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Prediction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/ld_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LD Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/study_locus/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Locus
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/study_locus_overlap/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Locus Overlap
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/summary_statistics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/target_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Target Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/variant_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_2_14" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../datasets/l2g_features/_l2g_feature/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    L2g features
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_2_14" id="__nav_6_2_14_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_2_14_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_2_14">
            <span class="md-nav__icon md-icon"></span>
            L2g features
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/l2g_features/colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/l2g_features/distance/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From distance
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/l2g_features/other/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Other features
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/l2g_features/vep/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From VEP
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../_datasources/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Datasources
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3" id="__nav_6_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_6_3">
            <span class="md-nav__icon md-icon"></span>
            Datasources
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../biosample_ontologies/_cell_ontology/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Biosample ontologies
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_1" id="__nav_6_3_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3_1">
            <span class="md-nav__icon md-icon"></span>
            Biosample ontologies
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../biosample_ontologies/_uberon/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Uberon
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_2" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../_ensembl/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Ensembl
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_2" id="__nav_6_3_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_6_3_2">
            <span class="md-nav__icon md-icon"></span>
            Ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Variant effector parser
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Variant effector parser
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser" class="md-nav__link">
    <span class="md-ellipsis">
      VariantEffectPredictorParser
    </span>
  </a>
  
    <nav class="md-nav" aria-label="VariantEffectPredictorParser">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser.extract_variant_index_from_vep" class="md-nav__link">
    <span class="md-ellipsis">
      extract_variant_index_from_vep
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser.get_schema" class="md-nav__link">
    <span class="md-ellipsis">
      get_schema
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser.process_vep_output" class="md-nav__link">
    <span class="md-ellipsis">
      process_vep_output
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../eqtl_catalogue/_eqtl_catalogue/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Eqtl catalogue
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_3" id="__nav_6_3_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3_3">
            <span class="md-nav__icon md-icon"></span>
            Eqtl catalogue
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../eqtl_catalogue/finemapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Fine mapping results
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../eqtl_catalogue/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../eqtl_catalogue/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Stats
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../finngen/_finngen/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Finngen
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_4" id="__nav_6_3_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3_4">
            <span class="md-nav__icon md-icon"></span>
            Finngen
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../finngen/finemapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Finemapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../finngen/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../finngen/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../gnomad/_gnomad/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Gnomad
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_5" id="__nav_6_3_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3_5">
            <span class="md-nav__icon md-icon"></span>
            Gnomad
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gnomad/gnomad_ld/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LD Matrix
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gnomad/gnomad_variants/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variants
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_6" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../gwas_catalog/_gwas_catalog/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Gwas catalog
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_6" id="__nav_6_3_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3_6">
            <span class="md-nav__icon md-icon"></span>
            Gwas catalog
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gwas_catalog/associations/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Associations
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gwas_catalog/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gwas_catalog/study_splitter/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Splitter
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gwas_catalog/summary_statistics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary statistics
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_7" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../intervals/_intervals/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Intervals
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_7" id="__nav_6_3_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3_7">
            <span class="md-nav__icon md-icon"></span>
            Intervals
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../intervals/andersson/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Andersson et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../intervals/javierre/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Javierre et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../intervals/jung/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Jung et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../intervals/thurman/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Thurman et al.
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../open_targets/_open_targets/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Open targets
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_8" id="__nav_6_3_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3_8">
            <span class="md-nav__icon md-icon"></span>
            Open targets
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../open_targets/l2g_gold_standard/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Gold Standard
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_9" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../ukb_ppp_eur/_ukb_ppp_eur/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Ukb ppp eur
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_9" id="__nav_6_3_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3_9">
            <span class="md-nav__icon md-icon"></span>
            Ukb ppp eur
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ukb_ppp_eur/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ukb_ppp_eur/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../methods/_methods/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Methods
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_4" id="__nav_6_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_4">
            <span class="md-nav__icon md-icon"></span>
            Methods
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/carma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CARMA
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/coloc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Coloc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/ecaviar/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    eCAVIAR
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/ld_annotator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LDAnnotator
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/pics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PICS
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/sumstat_imputation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics Imputation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/sumstat_quality_controls/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    QC of GWAS Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/susie_inf/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SuSiE-inf
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_4_10" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../methods/l2g/_l2g/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    L2g
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_4_10" id="__nav_6_4_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_4_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_4_10">
            <span class="md-nav__icon md-icon"></span>
            L2g
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/l2g/feature_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Feature Factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/l2g/model/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/l2g/trainer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Trainer
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../steps/_steps/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Steps
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_5" id="__nav_6_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_5">
            <span class="md-nav__icon md-icon"></span>
            Steps
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/biosample_index_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    biosample_index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/credible_set_qc_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    credible_set_qc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/eqtl_catalogue/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    eQTL Catalogue
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/finngen_studies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    finngen_studies
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/finngen_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    finngen_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/gwas_catalog_curation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_study_curation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/gwas_catalog_study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_study_inclusion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/gwas_catalog_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/gwas_catalog_top_hits/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GWAS Catalog Top Hits Ingestion Step
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/l2g/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Locus to Gene (L2G)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/ld_clump/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ld_based_clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/ld_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GnomAD Linkage data ingestion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/locus_breaker_clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    locus_breaker_clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/pics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    pics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/study_locus_validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study-Locus Validation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/study_validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Validation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/summary_statistics_qc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    summary_statistics_qc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/ukb_ppp_eur_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ukb_ppp_eur_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/variant_annotation_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GnomAD variant data ingestion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/variant_index_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    variant_index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/window_based_clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    window_based_clumping
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser" class="md-nav__link">
    <span class="md-ellipsis">
      VariantEffectPredictorParser
    </span>
  </a>
  
    <nav class="md-nav" aria-label="VariantEffectPredictorParser">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser.extract_variant_index_from_vep" class="md-nav__link">
    <span class="md-ellipsis">
      extract_variant_index_from_vep
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser.get_schema" class="md-nav__link">
    <span class="md-ellipsis">
      get_schema
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser.process_vep_output" class="md-nav__link">
    <span class="md-ellipsis">
      process_vep_output
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>Variant effector parser</h1>

<div class="doc doc-object doc-class">



<h2 id="gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser" class="doc doc-heading">
            <code>gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser</code>


<a href="#gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents first">


        <p>Collection of methods to parse VEP output in json format.</p>






              <details class="quote">
                <summary>Source code in <code>src/gentropy/datasource/ensembl/vep_parser.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  27</span>
<span class="normal">  28</span>
<span class="normal">  29</span>
<span class="normal">  30</span>
<span class="normal">  31</span>
<span class="normal">  32</span>
<span class="normal">  33</span>
<span class="normal">  34</span>
<span class="normal">  35</span>
<span class="normal">  36</span>
<span class="normal">  37</span>
<span class="normal">  38</span>
<span class="normal">  39</span>
<span class="normal">  40</span>
<span class="normal">  41</span>
<span class="normal">  42</span>
<span class="normal">  43</span>
<span class="normal">  44</span>
<span class="normal">  45</span>
<span class="normal">  46</span>
<span class="normal">  47</span>
<span class="normal">  48</span>
<span class="normal">  49</span>
<span class="normal">  50</span>
<span class="normal">  51</span>
<span class="normal">  52</span>
<span class="normal">  53</span>
<span class="normal">  54</span>
<span class="normal">  55</span>
<span class="normal">  56</span>
<span class="normal">  57</span>
<span class="normal">  58</span>
<span class="normal">  59</span>
<span class="normal">  60</span>
<span class="normal">  61</span>
<span class="normal">  62</span>
<span class="normal">  63</span>
<span class="normal">  64</span>
<span class="normal">  65</span>
<span class="normal">  66</span>
<span class="normal">  67</span>
<span class="normal">  68</span>
<span class="normal">  69</span>
<span class="normal">  70</span>
<span class="normal">  71</span>
<span class="normal">  72</span>
<span class="normal">  73</span>
<span class="normal">  74</span>
<span class="normal">  75</span>
<span class="normal">  76</span>
<span class="normal">  77</span>
<span class="normal">  78</span>
<span class="normal">  79</span>
<span class="normal">  80</span>
<span class="normal">  81</span>
<span class="normal">  82</span>
<span class="normal">  83</span>
<span class="normal">  84</span>
<span class="normal">  85</span>
<span class="normal">  86</span>
<span class="normal">  87</span>
<span class="normal">  88</span>
<span class="normal">  89</span>
<span class="normal">  90</span>
<span class="normal">  91</span>
<span class="normal">  92</span>
<span class="normal">  93</span>
<span class="normal">  94</span>
<span class="normal">  95</span>
<span class="normal">  96</span>
<span class="normal">  97</span>
<span class="normal">  98</span>
<span class="normal">  99</span>
<span class="normal"> 100</span>
<span class="normal"> 101</span>
<span class="normal"> 102</span>
<span class="normal"> 103</span>
<span class="normal"> 104</span>
<span class="normal"> 105</span>
<span class="normal"> 106</span>
<span class="normal"> 107</span>
<span class="normal"> 108</span>
<span class="normal"> 109</span>
<span class="normal"> 110</span>
<span class="normal"> 111</span>
<span class="normal"> 112</span>
<span class="normal"> 113</span>
<span class="normal"> 114</span>
<span class="normal"> 115</span>
<span class="normal"> 116</span>
<span class="normal"> 117</span>
<span class="normal"> 118</span>
<span class="normal"> 119</span>
<span class="normal"> 120</span>
<span class="normal"> 121</span>
<span class="normal"> 122</span>
<span class="normal"> 123</span>
<span class="normal"> 124</span>
<span class="normal"> 125</span>
<span class="normal"> 126</span>
<span class="normal"> 127</span>
<span class="normal"> 128</span>
<span class="normal"> 129</span>
<span class="normal"> 130</span>
<span class="normal"> 131</span>
<span class="normal"> 132</span>
<span class="normal"> 133</span>
<span class="normal"> 134</span>
<span class="normal"> 135</span>
<span class="normal"> 136</span>
<span class="normal"> 137</span>
<span class="normal"> 138</span>
<span class="normal"> 139</span>
<span class="normal"> 140</span>
<span class="normal"> 141</span>
<span class="normal"> 142</span>
<span class="normal"> 143</span>
<span class="normal"> 144</span>
<span class="normal"> 145</span>
<span class="normal"> 146</span>
<span class="normal"> 147</span>
<span class="normal"> 148</span>
<span class="normal"> 149</span>
<span class="normal"> 150</span>
<span class="normal"> 151</span>
<span class="normal"> 152</span>
<span class="normal"> 153</span>
<span class="normal"> 154</span>
<span class="normal"> 155</span>
<span class="normal"> 156</span>
<span class="normal"> 157</span>
<span class="normal"> 158</span>
<span class="normal"> 159</span>
<span class="normal"> 160</span>
<span class="normal"> 161</span>
<span class="normal"> 162</span>
<span class="normal"> 163</span>
<span class="normal"> 164</span>
<span class="normal"> 165</span>
<span class="normal"> 166</span>
<span class="normal"> 167</span>
<span class="normal"> 168</span>
<span class="normal"> 169</span>
<span class="normal"> 170</span>
<span class="normal"> 171</span>
<span class="normal"> 172</span>
<span class="normal"> 173</span>
<span class="normal"> 174</span>
<span class="normal"> 175</span>
<span class="normal"> 176</span>
<span class="normal"> 177</span>
<span class="normal"> 178</span>
<span class="normal"> 179</span>
<span class="normal"> 180</span>
<span class="normal"> 181</span>
<span class="normal"> 182</span>
<span class="normal"> 183</span>
<span class="normal"> 184</span>
<span class="normal"> 185</span>
<span class="normal"> 186</span>
<span class="normal"> 187</span>
<span class="normal"> 188</span>
<span class="normal"> 189</span>
<span class="normal"> 190</span>
<span class="normal"> 191</span>
<span class="normal"> 192</span>
<span class="normal"> 193</span>
<span class="normal"> 194</span>
<span class="normal"> 195</span>
<span class="normal"> 196</span>
<span class="normal"> 197</span>
<span class="normal"> 198</span>
<span class="normal"> 199</span>
<span class="normal"> 200</span>
<span class="normal"> 201</span>
<span class="normal"> 202</span>
<span class="normal"> 203</span>
<span class="normal"> 204</span>
<span class="normal"> 205</span>
<span class="normal"> 206</span>
<span class="normal"> 207</span>
<span class="normal"> 208</span>
<span class="normal"> 209</span>
<span class="normal"> 210</span>
<span class="normal"> 211</span>
<span class="normal"> 212</span>
<span class="normal"> 213</span>
<span class="normal"> 214</span>
<span class="normal"> 215</span>
<span class="normal"> 216</span>
<span class="normal"> 217</span>
<span class="normal"> 218</span>
<span class="normal"> 219</span>
<span class="normal"> 220</span>
<span class="normal"> 221</span>
<span class="normal"> 222</span>
<span class="normal"> 223</span>
<span class="normal"> 224</span>
<span class="normal"> 225</span>
<span class="normal"> 226</span>
<span class="normal"> 227</span>
<span class="normal"> 228</span>
<span class="normal"> 229</span>
<span class="normal"> 230</span>
<span class="normal"> 231</span>
<span class="normal"> 232</span>
<span class="normal"> 233</span>
<span class="normal"> 234</span>
<span class="normal"> 235</span>
<span class="normal"> 236</span>
<span class="normal"> 237</span>
<span class="normal"> 238</span>
<span class="normal"> 239</span>
<span class="normal"> 240</span>
<span class="normal"> 241</span>
<span class="normal"> 242</span>
<span class="normal"> 243</span>
<span class="normal"> 244</span>
<span class="normal"> 245</span>
<span class="normal"> 246</span>
<span class="normal"> 247</span>
<span class="normal"> 248</span>
<span class="normal"> 249</span>
<span class="normal"> 250</span>
<span class="normal"> 251</span>
<span class="normal"> 252</span>
<span class="normal"> 253</span>
<span class="normal"> 254</span>
<span class="normal"> 255</span>
<span class="normal"> 256</span>
<span class="normal"> 257</span>
<span class="normal"> 258</span>
<span class="normal"> 259</span>
<span class="normal"> 260</span>
<span class="normal"> 261</span>
<span class="normal"> 262</span>
<span class="normal"> 263</span>
<span class="normal"> 264</span>
<span class="normal"> 265</span>
<span class="normal"> 266</span>
<span class="normal"> 267</span>
<span class="normal"> 268</span>
<span class="normal"> 269</span>
<span class="normal"> 270</span>
<span class="normal"> 271</span>
<span class="normal"> 272</span>
<span class="normal"> 273</span>
<span class="normal"> 274</span>
<span class="normal"> 275</span>
<span class="normal"> 276</span>
<span class="normal"> 277</span>
<span class="normal"> 278</span>
<span class="normal"> 279</span>
<span class="normal"> 280</span>
<span class="normal"> 281</span>
<span class="normal"> 282</span>
<span class="normal"> 283</span>
<span class="normal"> 284</span>
<span class="normal"> 285</span>
<span class="normal"> 286</span>
<span class="normal"> 287</span>
<span class="normal"> 288</span>
<span class="normal"> 289</span>
<span class="normal"> 290</span>
<span class="normal"> 291</span>
<span class="normal"> 292</span>
<span class="normal"> 293</span>
<span class="normal"> 294</span>
<span class="normal"> 295</span>
<span class="normal"> 296</span>
<span class="normal"> 297</span>
<span class="normal"> 298</span>
<span class="normal"> 299</span>
<span class="normal"> 300</span>
<span class="normal"> 301</span>
<span class="normal"> 302</span>
<span class="normal"> 303</span>
<span class="normal"> 304</span>
<span class="normal"> 305</span>
<span class="normal"> 306</span>
<span class="normal"> 307</span>
<span class="normal"> 308</span>
<span class="normal"> 309</span>
<span class="normal"> 310</span>
<span class="normal"> 311</span>
<span class="normal"> 312</span>
<span class="normal"> 313</span>
<span class="normal"> 314</span>
<span class="normal"> 315</span>
<span class="normal"> 316</span>
<span class="normal"> 317</span>
<span class="normal"> 318</span>
<span class="normal"> 319</span>
<span class="normal"> 320</span>
<span class="normal"> 321</span>
<span class="normal"> 322</span>
<span class="normal"> 323</span>
<span class="normal"> 324</span>
<span class="normal"> 325</span>
<span class="normal"> 326</span>
<span class="normal"> 327</span>
<span class="normal"> 328</span>
<span class="normal"> 329</span>
<span class="normal"> 330</span>
<span class="normal"> 331</span>
<span class="normal"> 332</span>
<span class="normal"> 333</span>
<span class="normal"> 334</span>
<span class="normal"> 335</span>
<span class="normal"> 336</span>
<span class="normal"> 337</span>
<span class="normal"> 338</span>
<span class="normal"> 339</span>
<span class="normal"> 340</span>
<span class="normal"> 341</span>
<span class="normal"> 342</span>
<span class="normal"> 343</span>
<span class="normal"> 344</span>
<span class="normal"> 345</span>
<span class="normal"> 346</span>
<span class="normal"> 347</span>
<span class="normal"> 348</span>
<span class="normal"> 349</span>
<span class="normal"> 350</span>
<span class="normal"> 351</span>
<span class="normal"> 352</span>
<span class="normal"> 353</span>
<span class="normal"> 354</span>
<span class="normal"> 355</span>
<span class="normal"> 356</span>
<span class="normal"> 357</span>
<span class="normal"> 358</span>
<span class="normal"> 359</span>
<span class="normal"> 360</span>
<span class="normal"> 361</span>
<span class="normal"> 362</span>
<span class="normal"> 363</span>
<span class="normal"> 364</span>
<span class="normal"> 365</span>
<span class="normal"> 366</span>
<span class="normal"> 367</span>
<span class="normal"> 368</span>
<span class="normal"> 369</span>
<span class="normal"> 370</span>
<span class="normal"> 371</span>
<span class="normal"> 372</span>
<span class="normal"> 373</span>
<span class="normal"> 374</span>
<span class="normal"> 375</span>
<span class="normal"> 376</span>
<span class="normal"> 377</span>
<span class="normal"> 378</span>
<span class="normal"> 379</span>
<span class="normal"> 380</span>
<span class="normal"> 381</span>
<span class="normal"> 382</span>
<span class="normal"> 383</span>
<span class="normal"> 384</span>
<span class="normal"> 385</span>
<span class="normal"> 386</span>
<span class="normal"> 387</span>
<span class="normal"> 388</span>
<span class="normal"> 389</span>
<span class="normal"> 390</span>
<span class="normal"> 391</span>
<span class="normal"> 392</span>
<span class="normal"> 393</span>
<span class="normal"> 394</span>
<span class="normal"> 395</span>
<span class="normal"> 396</span>
<span class="normal"> 397</span>
<span class="normal"> 398</span>
<span class="normal"> 399</span>
<span class="normal"> 400</span>
<span class="normal"> 401</span>
<span class="normal"> 402</span>
<span class="normal"> 403</span>
<span class="normal"> 404</span>
<span class="normal"> 405</span>
<span class="normal"> 406</span>
<span class="normal"> 407</span>
<span class="normal"> 408</span>
<span class="normal"> 409</span>
<span class="normal"> 410</span>
<span class="normal"> 411</span>
<span class="normal"> 412</span>
<span class="normal"> 413</span>
<span class="normal"> 414</span>
<span class="normal"> 415</span>
<span class="normal"> 416</span>
<span class="normal"> 417</span>
<span class="normal"> 418</span>
<span class="normal"> 419</span>
<span class="normal"> 420</span>
<span class="normal"> 421</span>
<span class="normal"> 422</span>
<span class="normal"> 423</span>
<span class="normal"> 424</span>
<span class="normal"> 425</span>
<span class="normal"> 426</span>
<span class="normal"> 427</span>
<span class="normal"> 428</span>
<span class="normal"> 429</span>
<span class="normal"> 430</span>
<span class="normal"> 431</span>
<span class="normal"> 432</span>
<span class="normal"> 433</span>
<span class="normal"> 434</span>
<span class="normal"> 435</span>
<span class="normal"> 436</span>
<span class="normal"> 437</span>
<span class="normal"> 438</span>
<span class="normal"> 439</span>
<span class="normal"> 440</span>
<span class="normal"> 441</span>
<span class="normal"> 442</span>
<span class="normal"> 443</span>
<span class="normal"> 444</span>
<span class="normal"> 445</span>
<span class="normal"> 446</span>
<span class="normal"> 447</span>
<span class="normal"> 448</span>
<span class="normal"> 449</span>
<span class="normal"> 450</span>
<span class="normal"> 451</span>
<span class="normal"> 452</span>
<span class="normal"> 453</span>
<span class="normal"> 454</span>
<span class="normal"> 455</span>
<span class="normal"> 456</span>
<span class="normal"> 457</span>
<span class="normal"> 458</span>
<span class="normal"> 459</span>
<span class="normal"> 460</span>
<span class="normal"> 461</span>
<span class="normal"> 462</span>
<span class="normal"> 463</span>
<span class="normal"> 464</span>
<span class="normal"> 465</span>
<span class="normal"> 466</span>
<span class="normal"> 467</span>
<span class="normal"> 468</span>
<span class="normal"> 469</span>
<span class="normal"> 470</span>
<span class="normal"> 471</span>
<span class="normal"> 472</span>
<span class="normal"> 473</span>
<span class="normal"> 474</span>
<span class="normal"> 475</span>
<span class="normal"> 476</span>
<span class="normal"> 477</span>
<span class="normal"> 478</span>
<span class="normal"> 479</span>
<span class="normal"> 480</span>
<span class="normal"> 481</span>
<span class="normal"> 482</span>
<span class="normal"> 483</span>
<span class="normal"> 484</span>
<span class="normal"> 485</span>
<span class="normal"> 486</span>
<span class="normal"> 487</span>
<span class="normal"> 488</span>
<span class="normal"> 489</span>
<span class="normal"> 490</span>
<span class="normal"> 491</span>
<span class="normal"> 492</span>
<span class="normal"> 493</span>
<span class="normal"> 494</span>
<span class="normal"> 495</span>
<span class="normal"> 496</span>
<span class="normal"> 497</span>
<span class="normal"> 498</span>
<span class="normal"> 499</span>
<span class="normal"> 500</span>
<span class="normal"> 501</span>
<span class="normal"> 502</span>
<span class="normal"> 503</span>
<span class="normal"> 504</span>
<span class="normal"> 505</span>
<span class="normal"> 506</span>
<span class="normal"> 507</span>
<span class="normal"> 508</span>
<span class="normal"> 509</span>
<span class="normal"> 510</span>
<span class="normal"> 511</span>
<span class="normal"> 512</span>
<span class="normal"> 513</span>
<span class="normal"> 514</span>
<span class="normal"> 515</span>
<span class="normal"> 516</span>
<span class="normal"> 517</span>
<span class="normal"> 518</span>
<span class="normal"> 519</span>
<span class="normal"> 520</span>
<span class="normal"> 521</span>
<span class="normal"> 522</span>
<span class="normal"> 523</span>
<span class="normal"> 524</span>
<span class="normal"> 525</span>
<span class="normal"> 526</span>
<span class="normal"> 527</span>
<span class="normal"> 528</span>
<span class="normal"> 529</span>
<span class="normal"> 530</span>
<span class="normal"> 531</span>
<span class="normal"> 532</span>
<span class="normal"> 533</span>
<span class="normal"> 534</span>
<span class="normal"> 535</span>
<span class="normal"> 536</span>
<span class="normal"> 537</span>
<span class="normal"> 538</span>
<span class="normal"> 539</span>
<span class="normal"> 540</span>
<span class="normal"> 541</span>
<span class="normal"> 542</span>
<span class="normal"> 543</span>
<span class="normal"> 544</span>
<span class="normal"> 545</span>
<span class="normal"> 546</span>
<span class="normal"> 547</span>
<span class="normal"> 548</span>
<span class="normal"> 549</span>
<span class="normal"> 550</span>
<span class="normal"> 551</span>
<span class="normal"> 552</span>
<span class="normal"> 553</span>
<span class="normal"> 554</span>
<span class="normal"> 555</span>
<span class="normal"> 556</span>
<span class="normal"> 557</span>
<span class="normal"> 558</span>
<span class="normal"> 559</span>
<span class="normal"> 560</span>
<span class="normal"> 561</span>
<span class="normal"> 562</span>
<span class="normal"> 563</span>
<span class="normal"> 564</span>
<span class="normal"> 565</span>
<span class="normal"> 566</span>
<span class="normal"> 567</span>
<span class="normal"> 568</span>
<span class="normal"> 569</span>
<span class="normal"> 570</span>
<span class="normal"> 571</span>
<span class="normal"> 572</span>
<span class="normal"> 573</span>
<span class="normal"> 574</span>
<span class="normal"> 575</span>
<span class="normal"> 576</span>
<span class="normal"> 577</span>
<span class="normal"> 578</span>
<span class="normal"> 579</span>
<span class="normal"> 580</span>
<span class="normal"> 581</span>
<span class="normal"> 582</span>
<span class="normal"> 583</span>
<span class="normal"> 584</span>
<span class="normal"> 585</span>
<span class="normal"> 586</span>
<span class="normal"> 587</span>
<span class="normal"> 588</span>
<span class="normal"> 589</span>
<span class="normal"> 590</span>
<span class="normal"> 591</span>
<span class="normal"> 592</span>
<span class="normal"> 593</span>
<span class="normal"> 594</span>
<span class="normal"> 595</span>
<span class="normal"> 596</span>
<span class="normal"> 597</span>
<span class="normal"> 598</span>
<span class="normal"> 599</span>
<span class="normal"> 600</span>
<span class="normal"> 601</span>
<span class="normal"> 602</span>
<span class="normal"> 603</span>
<span class="normal"> 604</span>
<span class="normal"> 605</span>
<span class="normal"> 606</span>
<span class="normal"> 607</span>
<span class="normal"> 608</span>
<span class="normal"> 609</span>
<span class="normal"> 610</span>
<span class="normal"> 611</span>
<span class="normal"> 612</span>
<span class="normal"> 613</span>
<span class="normal"> 614</span>
<span class="normal"> 615</span>
<span class="normal"> 616</span>
<span class="normal"> 617</span>
<span class="normal"> 618</span>
<span class="normal"> 619</span>
<span class="normal"> 620</span>
<span class="normal"> 621</span>
<span class="normal"> 622</span>
<span class="normal"> 623</span>
<span class="normal"> 624</span>
<span class="normal"> 625</span>
<span class="normal"> 626</span>
<span class="normal"> 627</span>
<span class="normal"> 628</span>
<span class="normal"> 629</span>
<span class="normal"> 630</span>
<span class="normal"> 631</span>
<span class="normal"> 632</span>
<span class="normal"> 633</span>
<span class="normal"> 634</span>
<span class="normal"> 635</span>
<span class="normal"> 636</span>
<span class="normal"> 637</span>
<span class="normal"> 638</span>
<span class="normal"> 639</span>
<span class="normal"> 640</span>
<span class="normal"> 641</span>
<span class="normal"> 642</span>
<span class="normal"> 643</span>
<span class="normal"> 644</span>
<span class="normal"> 645</span>
<span class="normal"> 646</span>
<span class="normal"> 647</span>
<span class="normal"> 648</span>
<span class="normal"> 649</span>
<span class="normal"> 650</span>
<span class="normal"> 651</span>
<span class="normal"> 652</span>
<span class="normal"> 653</span>
<span class="normal"> 654</span>
<span class="normal"> 655</span>
<span class="normal"> 656</span>
<span class="normal"> 657</span>
<span class="normal"> 658</span>
<span class="normal"> 659</span>
<span class="normal"> 660</span>
<span class="normal"> 661</span>
<span class="normal"> 662</span>
<span class="normal"> 663</span>
<span class="normal"> 664</span>
<span class="normal"> 665</span>
<span class="normal"> 666</span>
<span class="normal"> 667</span>
<span class="normal"> 668</span>
<span class="normal"> 669</span>
<span class="normal"> 670</span>
<span class="normal"> 671</span>
<span class="normal"> 672</span>
<span class="normal"> 673</span>
<span class="normal"> 674</span>
<span class="normal"> 675</span>
<span class="normal"> 676</span>
<span class="normal"> 677</span>
<span class="normal"> 678</span>
<span class="normal"> 679</span>
<span class="normal"> 680</span>
<span class="normal"> 681</span>
<span class="normal"> 682</span>
<span class="normal"> 683</span>
<span class="normal"> 684</span>
<span class="normal"> 685</span>
<span class="normal"> 686</span>
<span class="normal"> 687</span>
<span class="normal"> 688</span>
<span class="normal"> 689</span>
<span class="normal"> 690</span>
<span class="normal"> 691</span>
<span class="normal"> 692</span>
<span class="normal"> 693</span>
<span class="normal"> 694</span>
<span class="normal"> 695</span>
<span class="normal"> 696</span>
<span class="normal"> 697</span>
<span class="normal"> 698</span>
<span class="normal"> 699</span>
<span class="normal"> 700</span>
<span class="normal"> 701</span>
<span class="normal"> 702</span>
<span class="normal"> 703</span>
<span class="normal"> 704</span>
<span class="normal"> 705</span>
<span class="normal"> 706</span>
<span class="normal"> 707</span>
<span class="normal"> 708</span>
<span class="normal"> 709</span>
<span class="normal"> 710</span>
<span class="normal"> 711</span>
<span class="normal"> 712</span>
<span class="normal"> 713</span>
<span class="normal"> 714</span>
<span class="normal"> 715</span>
<span class="normal"> 716</span>
<span class="normal"> 717</span>
<span class="normal"> 718</span>
<span class="normal"> 719</span>
<span class="normal"> 720</span>
<span class="normal"> 721</span>
<span class="normal"> 722</span>
<span class="normal"> 723</span>
<span class="normal"> 724</span>
<span class="normal"> 725</span>
<span class="normal"> 726</span>
<span class="normal"> 727</span>
<span class="normal"> 728</span>
<span class="normal"> 729</span>
<span class="normal"> 730</span>
<span class="normal"> 731</span>
<span class="normal"> 732</span>
<span class="normal"> 733</span>
<span class="normal"> 734</span>
<span class="normal"> 735</span>
<span class="normal"> 736</span>
<span class="normal"> 737</span>
<span class="normal"> 738</span>
<span class="normal"> 739</span>
<span class="normal"> 740</span>
<span class="normal"> 741</span>
<span class="normal"> 742</span>
<span class="normal"> 743</span>
<span class="normal"> 744</span>
<span class="normal"> 745</span>
<span class="normal"> 746</span>
<span class="normal"> 747</span>
<span class="normal"> 748</span>
<span class="normal"> 749</span>
<span class="normal"> 750</span>
<span class="normal"> 751</span>
<span class="normal"> 752</span>
<span class="normal"> 753</span>
<span class="normal"> 754</span>
<span class="normal"> 755</span>
<span class="normal"> 756</span>
<span class="normal"> 757</span>
<span class="normal"> 758</span>
<span class="normal"> 759</span>
<span class="normal"> 760</span>
<span class="normal"> 761</span>
<span class="normal"> 762</span>
<span class="normal"> 763</span>
<span class="normal"> 764</span>
<span class="normal"> 765</span>
<span class="normal"> 766</span>
<span class="normal"> 767</span>
<span class="normal"> 768</span>
<span class="normal"> 769</span>
<span class="normal"> 770</span>
<span class="normal"> 771</span>
<span class="normal"> 772</span>
<span class="normal"> 773</span>
<span class="normal"> 774</span>
<span class="normal"> 775</span>
<span class="normal"> 776</span>
<span class="normal"> 777</span>
<span class="normal"> 778</span>
<span class="normal"> 779</span>
<span class="normal"> 780</span>
<span class="normal"> 781</span>
<span class="normal"> 782</span>
<span class="normal"> 783</span>
<span class="normal"> 784</span>
<span class="normal"> 785</span>
<span class="normal"> 786</span>
<span class="normal"> 787</span>
<span class="normal"> 788</span>
<span class="normal"> 789</span>
<span class="normal"> 790</span>
<span class="normal"> 791</span>
<span class="normal"> 792</span>
<span class="normal"> 793</span>
<span class="normal"> 794</span>
<span class="normal"> 795</span>
<span class="normal"> 796</span>
<span class="normal"> 797</span>
<span class="normal"> 798</span>
<span class="normal"> 799</span>
<span class="normal"> 800</span>
<span class="normal"> 801</span>
<span class="normal"> 802</span>
<span class="normal"> 803</span>
<span class="normal"> 804</span>
<span class="normal"> 805</span>
<span class="normal"> 806</span>
<span class="normal"> 807</span>
<span class="normal"> 808</span>
<span class="normal"> 809</span>
<span class="normal"> 810</span>
<span class="normal"> 811</span>
<span class="normal"> 812</span>
<span class="normal"> 813</span>
<span class="normal"> 814</span>
<span class="normal"> 815</span>
<span class="normal"> 816</span>
<span class="normal"> 817</span>
<span class="normal"> 818</span>
<span class="normal"> 819</span>
<span class="normal"> 820</span>
<span class="normal"> 821</span>
<span class="normal"> 822</span>
<span class="normal"> 823</span>
<span class="normal"> 824</span>
<span class="normal"> 825</span>
<span class="normal"> 826</span>
<span class="normal"> 827</span>
<span class="normal"> 828</span>
<span class="normal"> 829</span>
<span class="normal"> 830</span>
<span class="normal"> 831</span>
<span class="normal"> 832</span>
<span class="normal"> 833</span>
<span class="normal"> 834</span>
<span class="normal"> 835</span>
<span class="normal"> 836</span>
<span class="normal"> 837</span>
<span class="normal"> 838</span>
<span class="normal"> 839</span>
<span class="normal"> 840</span>
<span class="normal"> 841</span>
<span class="normal"> 842</span>
<span class="normal"> 843</span>
<span class="normal"> 844</span>
<span class="normal"> 845</span>
<span class="normal"> 846</span>
<span class="normal"> 847</span>
<span class="normal"> 848</span>
<span class="normal"> 849</span>
<span class="normal"> 850</span>
<span class="normal"> 851</span>
<span class="normal"> 852</span>
<span class="normal"> 853</span>
<span class="normal"> 854</span>
<span class="normal"> 855</span>
<span class="normal"> 856</span>
<span class="normal"> 857</span>
<span class="normal"> 858</span>
<span class="normal"> 859</span>
<span class="normal"> 860</span>
<span class="normal"> 861</span>
<span class="normal"> 862</span>
<span class="normal"> 863</span>
<span class="normal"> 864</span>
<span class="normal"> 865</span>
<span class="normal"> 866</span>
<span class="normal"> 867</span>
<span class="normal"> 868</span>
<span class="normal"> 869</span>
<span class="normal"> 870</span>
<span class="normal"> 871</span>
<span class="normal"> 872</span>
<span class="normal"> 873</span>
<span class="normal"> 874</span>
<span class="normal"> 875</span>
<span class="normal"> 876</span>
<span class="normal"> 877</span>
<span class="normal"> 878</span>
<span class="normal"> 879</span>
<span class="normal"> 880</span>
<span class="normal"> 881</span>
<span class="normal"> 882</span>
<span class="normal"> 883</span>
<span class="normal"> 884</span>
<span class="normal"> 885</span>
<span class="normal"> 886</span>
<span class="normal"> 887</span>
<span class="normal"> 888</span>
<span class="normal"> 889</span>
<span class="normal"> 890</span>
<span class="normal"> 891</span>
<span class="normal"> 892</span>
<span class="normal"> 893</span>
<span class="normal"> 894</span>
<span class="normal"> 895</span>
<span class="normal"> 896</span>
<span class="normal"> 897</span>
<span class="normal"> 898</span>
<span class="normal"> 899</span>
<span class="normal"> 900</span>
<span class="normal"> 901</span>
<span class="normal"> 902</span>
<span class="normal"> 903</span>
<span class="normal"> 904</span>
<span class="normal"> 905</span>
<span class="normal"> 906</span>
<span class="normal"> 907</span>
<span class="normal"> 908</span>
<span class="normal"> 909</span>
<span class="normal"> 910</span>
<span class="normal"> 911</span>
<span class="normal"> 912</span>
<span class="normal"> 913</span>
<span class="normal"> 914</span>
<span class="normal"> 915</span>
<span class="normal"> 916</span>
<span class="normal"> 917</span>
<span class="normal"> 918</span>
<span class="normal"> 919</span>
<span class="normal"> 920</span>
<span class="normal"> 921</span>
<span class="normal"> 922</span>
<span class="normal"> 923</span>
<span class="normal"> 924</span>
<span class="normal"> 925</span>
<span class="normal"> 926</span>
<span class="normal"> 927</span>
<span class="normal"> 928</span>
<span class="normal"> 929</span>
<span class="normal"> 930</span>
<span class="normal"> 931</span>
<span class="normal"> 932</span>
<span class="normal"> 933</span>
<span class="normal"> 934</span>
<span class="normal"> 935</span>
<span class="normal"> 936</span>
<span class="normal"> 937</span>
<span class="normal"> 938</span>
<span class="normal"> 939</span>
<span class="normal"> 940</span>
<span class="normal"> 941</span>
<span class="normal"> 942</span>
<span class="normal"> 943</span>
<span class="normal"> 944</span>
<span class="normal"> 945</span>
<span class="normal"> 946</span>
<span class="normal"> 947</span>
<span class="normal"> 948</span>
<span class="normal"> 949</span>
<span class="normal"> 950</span>
<span class="normal"> 951</span>
<span class="normal"> 952</span>
<span class="normal"> 953</span>
<span class="normal"> 954</span>
<span class="normal"> 955</span>
<span class="normal"> 956</span>
<span class="normal"> 957</span>
<span class="normal"> 958</span>
<span class="normal"> 959</span>
<span class="normal"> 960</span>
<span class="normal"> 961</span>
<span class="normal"> 962</span>
<span class="normal"> 963</span>
<span class="normal"> 964</span>
<span class="normal"> 965</span>
<span class="normal"> 966</span>
<span class="normal"> 967</span>
<span class="normal"> 968</span>
<span class="normal"> 969</span>
<span class="normal"> 970</span>
<span class="normal"> 971</span>
<span class="normal"> 972</span>
<span class="normal"> 973</span>
<span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span>
<span class="normal">1124</span>
<span class="normal">1125</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">VariantEffectPredictorParser</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Collection of methods to parse VEP output in json format.&quot;&quot;&quot;</span>

    <span class="c1"># NOTE: Due to the fact that the comparison of the xrefs is done om the base of rsids</span>
    <span class="c1"># if the field `colocalised_variants` have multiple rsids, this extracting xrefs will result in</span>
    <span class="c1"># an array of xref structs, rather then the struct itself.</span>

    <span class="n">DBXREF_SCHEMA</span> <span class="o">=</span> <span class="n">VariantIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">()[</span><span class="s2">&quot;dbXrefs&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">dataType</span>

    <span class="c1"># Schema description of the variant effect object:</span>
    <span class="n">VARIANT_EFFECT_SCHEMA</span> <span class="o">=</span> <span class="n">get_nested_struct_schema</span><span class="p">(</span>
        <span class="n">VariantIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">()[</span><span class="s2">&quot;variantEffect&quot;</span><span class="p">]</span>
    <span class="p">)</span>

    <span class="c1"># Schema for the allele frequency column:</span>
    <span class="n">ALLELE_FREQUENCY_SCHEMA</span> <span class="o">=</span> <span class="n">VariantIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">()[</span><span class="s2">&quot;alleleFrequencies&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">dataType</span>

    <span class="c1"># Consequence to sequence ontology map:</span>
    <span class="n">SEQUENCE_ONTOLOGY_MAP</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">item</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]:</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">VariantIndexConfig</span><span class="o">.</span><span class="n">consequence_to_pathogenicity_score</span>
    <span class="p">}</span>

    <span class="c1"># Sequence ontology to score map:</span>
    <span class="n">LABEL_TO_SCORE_MAP</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">item</span><span class="p">[</span><span class="s2">&quot;label&quot;</span><span class="p">]:</span> <span class="n">item</span><span class="p">[</span><span class="s2">&quot;score&quot;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">VariantIndexConfig</span><span class="o">.</span><span class="n">consequence_to_pathogenicity_score</span>
    <span class="p">}</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_schema</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">t</span><span class="o">.</span><span class="n">StructType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the schema of the VEP output.</span>

<span class="sd">        Returns:</span>
<span class="sd">            t.StructType: VEP output schema.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; type(VariantEffectPredictorParser.get_schema())</span>
<span class="sd">            &lt;class &#39;pyspark.sql.types.StructType&#39;&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">parse_spark_schema</span><span class="p">(</span><span class="s2">&quot;vep_json_output.json&quot;</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">extract_variant_index_from_vep</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">VariantEffectPredictorParser</span><span class="p">],</span>
        <span class="n">spark</span><span class="p">:</span> <span class="n">SparkSession</span><span class="p">,</span>
        <span class="n">vep_output_path</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">hash_threshold</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">|</span> <span class="nb">float</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">VariantIndex</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract variant index from VEP output.</span>

<span class="sd">        Args:</span>
<span class="sd">            spark (SparkSession): Spark session.</span>
<span class="sd">            vep_output_path (str | list[str]): Path to the VEP output.</span>
<span class="sd">            hash_threshold (int): Threshold above which variant identifiers will be hashed.</span>
<span class="sd">            **kwargs (bool | float | int | str | None): Additional arguments to pass to spark.read.json.</span>

<span class="sd">        Returns:</span>
<span class="sd">            VariantIndex: Variant index dataset.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: Failed reading file or if the dataset is empty.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># To speed things up and simplify the json structure, read data following an expected schema:</span>
        <span class="n">vep_schema</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">get_schema</span><span class="p">()</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">vep_data</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">json</span><span class="p">(</span><span class="n">vep_output_path</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="n">vep_schema</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed reading file: </span><span class="si">{</span><span class="n">vep_output_path</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span> <span class="kn">from</span><span class="w"> </span><span class="nn">e</span>

        <span class="k">if</span> <span class="n">vep_data</span><span class="o">.</span><span class="n">isEmpty</span><span class="p">():</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The dataset is empty: </span><span class="si">{</span><span class="n">vep_output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Convert to VariantAnnotation dataset:</span>
        <span class="k">return</span> <span class="n">VariantIndex</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="n">VariantEffectPredictorParser</span><span class="o">.</span><span class="n">process_vep_output</span><span class="p">(</span>
                <span class="n">vep_data</span><span class="p">,</span> <span class="n">hash_threshold</span>
            <span class="p">),</span>
            <span class="n">_schema</span><span class="o">=</span><span class="n">VariantIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
            <span class="n">id_threshold</span><span class="o">=</span><span class="n">hash_threshold</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_extract_ensembl_xrefs</span><span class="p">(</span><span class="n">colocated_variants</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract rs identifiers and build cross reference to Ensembl&#39;s variant page.</span>

<span class="sd">        Args:</span>
<span class="sd">            colocated_variants (Column): Colocated variants field from VEP output.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: List of dbXrefs for rs identifiers.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">VariantEffectPredictorParser</span><span class="o">.</span><span class="n">_generate_dbxrefs</span><span class="p">(</span>
            <span class="n">VariantEffectPredictorParser</span><span class="o">.</span><span class="n">_colocated_variants_to_rsids</span><span class="p">(</span>
                <span class="n">colocated_variants</span>
            <span class="p">),</span>
            <span class="s2">&quot;ensembl_variation&quot;</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="nd">@enforce_schema</span><span class="p">(</span><span class="n">DBXREF_SCHEMA</span><span class="p">)</span>
    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_generate_dbxrefs</span><span class="p">(</span><span class="n">ids</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">source</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Convert a list of variant identifiers to dbXrefs given the source label.</span>

<span class="sd">        Identifiers are cast to strings, then Null values are filtered out of the id list.</span>

<span class="sd">        Args:</span>
<span class="sd">            ids (Column): List of variant identifiers.</span>
<span class="sd">            source (str): Source label for the dbXrefs.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: List of dbXrefs.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; (</span>
<span class="sd">            ...     spark.createDataFrame([(&#39;rs12&#39;,),(None,)], [&#39;test_id&#39;])</span>
<span class="sd">            ...     .select(VariantEffectPredictorParser._generate_dbxrefs(f.array(f.col(&#39;test_id&#39;)), &quot;ensemblVariation&quot;).alias(&#39;col&#39;))</span>
<span class="sd">            ...     .show(truncate=False)</span>
<span class="sd">            ... )</span>
<span class="sd">            +--------------------------+</span>
<span class="sd">            |col                       |</span>
<span class="sd">            +--------------------------+</span>
<span class="sd">            |[{rs12, ensemblVariation}]|</span>
<span class="sd">            |[]                        |</span>
<span class="sd">            +--------------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">            &gt;&gt;&gt; (</span>
<span class="sd">            ...     spark.createDataFrame([(&#39;rs12&#39;,),(None,)], [&#39;test_id&#39;])</span>
<span class="sd">            ...     .select(VariantEffectPredictorParser._generate_dbxrefs(f.array(f.col(&#39;test_id&#39;)), &quot;ensemblVariation&quot;).alias(&#39;col&#39;))</span>
<span class="sd">            ...     .first().col[0].asDict()</span>
<span class="sd">            ... )</span>
<span class="sd">            {&#39;id&#39;: &#39;rs12&#39;, &#39;source&#39;: &#39;ensemblVariation&#39;}</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">ids</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span> <span class="k">lambda</span> <span class="nb">id</span><span class="p">:</span> <span class="nb">id</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">())</span>
        <span class="n">xref_column</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
            <span class="n">ids</span><span class="p">,</span>
            <span class="k">lambda</span> <span class="nb">id</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="nb">id</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;id&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="n">source</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;source&quot;</span><span class="p">)</span>
            <span class="p">),</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">xref_column</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">())</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">xref_column</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_colocated_variants_to_rsids</span><span class="p">(</span><span class="n">colocated_variants</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract rs identifiers from the colocated variants VEP field.</span>

<span class="sd">        Args:</span>
<span class="sd">            colocated_variants (Column): Colocated variants field from VEP output.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: List of rs identifiers.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; data = [(&#39;s1&#39;, &#39;rs1&#39;),(&#39;s1&#39;, &#39;rs2&#39;),(&#39;s1&#39;, &#39;rs3&#39;),(&#39;s2&#39;, None),]</span>
<span class="sd">            &gt;&gt;&gt; (</span>
<span class="sd">            ...    spark.createDataFrame(data, [&#39;s&#39;,&#39;v&#39;])</span>
<span class="sd">            ...    .groupBy(&#39;s&#39;)</span>
<span class="sd">            ...    .agg(f.collect_list(f.struct(f.col(&#39;v&#39;).alias(&#39;id&#39;))).alias(&#39;cv&#39;))</span>
<span class="sd">            ...    .select(VariantEffectPredictorParser._colocated_variants_to_rsids(f.col(&#39;cv&#39;)).alias(&#39;rsIds&#39;),)</span>
<span class="sd">            ...    .show(truncate=False)</span>
<span class="sd">            ... )</span>
<span class="sd">            +---------------+</span>
<span class="sd">            |rsIds          |</span>
<span class="sd">            +---------------+</span>
<span class="sd">            |[rs1, rs2, rs3]|</span>
<span class="sd">            |[NULL]         |</span>
<span class="sd">            +---------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">colocated_variants</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                <span class="n">colocated_variants</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">variant</span><span class="p">:</span> <span class="n">variant</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;id&quot;</span><span class="p">)</span>
            <span class="p">),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;rsIds&quot;</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_extract_omim_xrefs</span><span class="p">(</span><span class="n">colocated_variants</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Build xdbRefs for OMIM identifiers.</span>

<span class="sd">        OMIM identifiers are extracted from the colocated variants field, casted to strings and formatted as dbXrefs.</span>

<span class="sd">        Args:</span>
<span class="sd">            colocated_variants (Column): Colocated variants field from VEP output.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: List of dbXrefs for OMIM identifiers.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; data = [(&#39;234234.32&#39;, &#39;rs1&#39;, &#39;s1&#39;,),(None, &#39;rs1&#39;, &#39;s1&#39;,),]</span>
<span class="sd">            &gt;&gt;&gt; (</span>
<span class="sd">            ...    spark.createDataFrame(data, [&#39;id&#39;, &#39;rsid&#39;, &#39;s&#39;])</span>
<span class="sd">            ...    .groupBy(&#39;s&#39;)</span>
<span class="sd">            ...    .agg(f.collect_list(f.struct(f.struct(f.array(f.col(&#39;id&#39;)).alias(&#39;OMIM&#39;)).alias(&#39;var_synonyms&#39;),f.col(&#39;rsid&#39;).alias(&#39;id&#39;),),).alias(&#39;cv&#39;),).select(VariantEffectPredictorParser._extract_omim_xrefs(f.col(&#39;cv&#39;)).alias(&#39;dbXrefs&#39;))</span>
<span class="sd">            ...    .show(truncate=False)</span>
<span class="sd">            ... )</span>
<span class="sd">            +-------------------+</span>
<span class="sd">            |dbXrefs            |</span>
<span class="sd">            +-------------------+</span>
<span class="sd">            |[{234234#32, omim}]|</span>
<span class="sd">            +-------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">variants_w_omim_ref</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">colocated_variants</span><span class="p">,</span>
            <span class="k">lambda</span> <span class="n">variant</span><span class="p">:</span> <span class="n">variant</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;var_synonyms&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;OMIM&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
        <span class="p">)</span>

        <span class="n">omim_var_ids</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
            <span class="n">variants_w_omim_ref</span><span class="p">,</span>
            <span class="k">lambda</span> <span class="n">var</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                <span class="n">var</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;var_synonyms&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;OMIM&quot;</span><span class="p">),</span>
                <span class="k">lambda</span> <span class="n">var</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">regexp_replace</span><span class="p">(</span><span class="n">var</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">()),</span> <span class="sa">r</span><span class="s2">&quot;\.&quot;</span><span class="p">,</span> <span class="sa">r</span><span class="s2">&quot;#&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">VariantEffectPredictorParser</span><span class="o">.</span><span class="n">_generate_dbxrefs</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span><span class="n">omim_var_ids</span><span class="p">),</span> <span class="s2">&quot;omim&quot;</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_extract_clinvar_xrefs</span><span class="p">(</span><span class="n">colocated_variants</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Build xdbRefs for VCV ClinVar identifiers.</span>

<span class="sd">        ClinVar identifiers are extracted from the colocated variants field.</span>
<span class="sd">        VCV-identifiers are filtered out to generate cross-references.</span>

<span class="sd">        Args:</span>
<span class="sd">            colocated_variants (Column): Colocated variants field from VEP output.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: List of dbXrefs for VCV ClinVar identifiers.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; data = [(&#39;VCV2323,RCV3423&#39;, &#39;rs1&#39;, &#39;s1&#39;,),(None, &#39;rs1&#39;, &#39;s1&#39;,),]</span>
<span class="sd">            &gt;&gt;&gt; (</span>
<span class="sd">            ...    spark.createDataFrame(data, [&#39;id&#39;, &#39;rsid&#39;, &#39;s&#39;])</span>
<span class="sd">            ...    .groupBy(&#39;s&#39;)</span>
<span class="sd">            ...    .agg(f.collect_list(f.struct(f.struct(f.split(f.col(&#39;id&#39;),&#39;,&#39;).alias(&#39;ClinVar&#39;)).alias(&#39;var_synonyms&#39;),f.col(&#39;rsid&#39;).alias(&#39;id&#39;),),).alias(&#39;cv&#39;),).select(VariantEffectPredictorParser._extract_clinvar_xrefs(f.col(&#39;cv&#39;)).alias(&#39;dbXrefs&#39;))</span>
<span class="sd">            ...    .show(truncate=False)</span>
<span class="sd">            ... )</span>
<span class="sd">            +--------------------+</span>
<span class="sd">            |dbXrefs             |</span>
<span class="sd">            +--------------------+</span>
<span class="sd">            |[{VCV2323, clinvar}]|</span>
<span class="sd">            +--------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">variants_w_clinvar_ref</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">colocated_variants</span><span class="p">,</span>
            <span class="k">lambda</span> <span class="n">variant</span><span class="p">:</span> <span class="n">variant</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;var_synonyms&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;ClinVar&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
        <span class="p">)</span>

        <span class="n">clin_var_ids</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
            <span class="n">variants_w_clinvar_ref</span><span class="p">,</span>
            <span class="k">lambda</span> <span class="n">var</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">var</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;var_synonyms&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;ClinVar&quot;</span><span class="p">),</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;VCV&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">VariantEffectPredictorParser</span><span class="o">.</span><span class="n">_generate_dbxrefs</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span><span class="n">clin_var_ids</span><span class="p">),</span> <span class="s2">&quot;clinvar&quot;</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_get_most_severe_transcript</span><span class="p">(</span>
        <span class="n">transcript_column_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">score_field_name</span><span class="p">:</span> <span class="nb">str</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return transcript with the highest in silico predicted score.</span>

<span class="sd">        This method assumes the higher the score, the more severe the consequence of the variant is.</span>
<span class="sd">        Hence, by selecting the transcript with the highest score, we are selecting the most severe consequence.</span>

<span class="sd">        Args:</span>
<span class="sd">            transcript_column_name (str): Name of the column containing the list of transcripts.</span>
<span class="sd">            score_field_name (str): Name of the field containing the severity score.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Most severe transcript.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; data = [(&quot;v1&quot;, 0.2, &#39;transcript1&#39;),(&quot;v1&quot;, None, &#39;transcript2&#39;),(&quot;v1&quot;, 0.6, &#39;transcript3&#39;),(&quot;v1&quot;, 0.4, &#39;transcript4&#39;),]</span>
<span class="sd">            &gt;&gt;&gt; (</span>
<span class="sd">            ...    spark.createDataFrame(data, [&#39;v&#39;, &#39;score&#39;, &#39;transcriptId&#39;])</span>
<span class="sd">            ...    .groupBy(&#39;v&#39;)</span>
<span class="sd">            ...    .agg(f.collect_list(f.struct(f.col(&#39;score&#39;),f.col(&#39;transcriptId&#39;))).alias(&#39;transcripts&#39;))</span>
<span class="sd">            ...    .select(VariantEffectPredictorParser._get_most_severe_transcript(&#39;transcripts&#39;, &#39;score&#39;).alias(&#39;most_severe_transcript&#39;))</span>
<span class="sd">            ...    .show(truncate=False)</span>
<span class="sd">            ... )</span>
<span class="sd">            +----------------------+</span>
<span class="sd">            |most_severe_transcript|</span>
<span class="sd">            +----------------------+</span>
<span class="sd">            |{0.6, transcript3}    |</span>
<span class="sd">            +----------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>

<span class="sd">        Raises:</span>
<span class="sd">            AssertionError: When `transcript_column_name` is not a string.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span>
            <span class="n">transcript_column_name</span><span class="p">,</span> <span class="nb">str</span>
        <span class="p">),</span> <span class="s2">&quot;transcript_column_name must be a string and not a column.&quot;</span>
        <span class="c1"># Order transcripts by severity score:</span>
        <span class="n">ordered_transcripts</span> <span class="o">=</span> <span class="n">order_array_of_structs_by_field</span><span class="p">(</span>
            <span class="n">transcript_column_name</span><span class="p">,</span> <span class="n">score_field_name</span>
        <span class="p">)</span>

        <span class="c1"># Drop transcripts with no severity score and return the most severe one:</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">ordered_transcripts</span><span class="p">,</span>
            <span class="k">lambda</span> <span class="n">transcript</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="n">score_field_name</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
        <span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

    <span class="nd">@classmethod</span>
    <span class="nd">@enforce_schema</span><span class="p">(</span><span class="n">VARIANT_EFFECT_SCHEMA</span><span class="p">)</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_get_vep_prediction</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">most_severe_consequence</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;VEP&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;method&quot;</span><span class="p">),</span>
            <span class="n">most_severe_consequence</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;assessment&quot;</span><span class="p">),</span>
            <span class="n">map_column_by_dictionary</span><span class="p">(</span>
                <span class="n">most_severe_consequence</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">LABEL_TO_SCORE_MAP</span>
            <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;score&quot;</span><span class="p">),</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="nd">@enforce_schema</span><span class="p">(</span><span class="n">VARIANT_EFFECT_SCHEMA</span><span class="p">)</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_get_max_alpha_missense</span><span class="p">(</span><span class="n">transcripts</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return the most severe alpha missense prediction from all transcripts.</span>

<span class="sd">        This function assumes one variant can fall onto only one gene with alpha-sense prediction available on the canonical transcript.</span>

<span class="sd">        Args:</span>
<span class="sd">            transcripts (Column): List of transcripts.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Most severe alpha missense prediction.</span>

<span class="sd">        Examples:</span>
<span class="sd">        &gt;&gt;&gt; data = [(&#39;v1&#39;, 0.4, &#39;assessment 1&#39;), (&#39;v1&#39;, None, None), (&#39;v1&#39;, None, None),(&#39;v2&#39;, None, None),]</span>
<span class="sd">        &gt;&gt;&gt; (</span>
<span class="sd">        ...     spark.createDataFrame(data, [&#39;v&#39;, &#39;a&#39;, &#39;b&#39;])</span>
<span class="sd">        ...    .groupBy(&#39;v&#39;)</span>
<span class="sd">        ...    .agg(f.collect_list(f.struct(f.struct(</span>
<span class="sd">        ...        f.col(&#39;a&#39;).alias(&#39;am_pathogenicity&#39;),</span>
<span class="sd">        ...        f.col(&#39;b&#39;).alias(&#39;am_class&#39;)).alias(&#39;alphamissense&#39;),</span>
<span class="sd">        ...        f.lit(&#39;gene1&#39;).alias(&#39;gene_id&#39;))).alias(&#39;transcripts&#39;)</span>
<span class="sd">        ...    )</span>
<span class="sd">        ...    .select(VariantEffectPredictorParser._get_max_alpha_missense(f.col(&#39;transcripts&#39;)).alias(&#39;am&#39;))</span>
<span class="sd">        ...    .show(truncate=False)</span>
<span class="sd">        ... )</span>
<span class="sd">        +-----------------------------------------------------+</span>
<span class="sd">        |am                                                   |</span>
<span class="sd">        +-----------------------------------------------------+</span>
<span class="sd">        |{AlphaMissense, assessment 1, 0.4, NULL, gene1, NULL}|</span>
<span class="sd">        |{AlphaMissense, NULL, NULL, NULL, gene1, NULL}       |</span>
<span class="sd">        +-----------------------------------------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Extracting transcript with alpha missense values:</span>
        <span class="n">transcript</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">transcripts</span><span class="p">,</span>
            <span class="k">lambda</span> <span class="n">transcript</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;alphamissense&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
        <span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="c1"># Adding method:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;AlphaMissense&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;method&quot;</span><span class="p">),</span>
                <span class="c1"># Extracting assessment:</span>
                <span class="n">transcript</span><span class="o">.</span><span class="n">alphamissense</span><span class="o">.</span><span class="n">am_class</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;assessment&quot;</span><span class="p">),</span>
                <span class="c1"># Extracting score:</span>
                <span class="n">transcript</span><span class="o">.</span><span class="n">alphamissense</span><span class="o">.</span><span class="n">am_pathogenicity</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">FloatType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                    <span class="s2">&quot;score&quot;</span>
                <span class="p">),</span>
                <span class="c1"># Adding assessment flag:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;assessmentFlag&quot;</span><span class="p">),</span>
                <span class="c1"># Extracting target id:</span>
                <span class="n">transcript</span><span class="o">.</span><span class="n">gene_id</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;targetId&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="nd">@enforce_schema</span><span class="p">(</span><span class="n">VARIANT_EFFECT_SCHEMA</span><span class="p">)</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_vep_variant_effect_extractor</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">VariantEffectPredictorParser</span><span class="p">],</span>
        <span class="n">transcript_column_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">method_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">score_column_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">assessment_column_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">assessment_flag_column_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract variant effect from VEP output.</span>

<span class="sd">        Args:</span>
<span class="sd">            transcript_column_name (str): Name of the column containing the list of transcripts.</span>
<span class="sd">            method_name (str): Name of the variant effect.</span>
<span class="sd">            score_column_name (str | None): Name of the column containing the score.</span>
<span class="sd">            assessment_column_name (str | None): Name of the column containing the assessment.</span>
<span class="sd">            assessment_flag_column_name (str | None): Name of the column containing the assessment flag.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Variant effect.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Get transcript with the highest score:</span>
        <span class="n">most_severe_transcript</span><span class="p">:</span> <span class="n">Column</span> <span class="o">=</span> <span class="p">(</span>
            <span class="c1"># Getting the most severe transcript:</span>
            <span class="n">VariantEffectPredictorParser</span><span class="o">.</span><span class="n">_get_most_severe_transcript</span><span class="p">(</span>
                <span class="n">transcript_column_name</span><span class="p">,</span> <span class="n">score_column_name</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">score_column_name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="c1"># If we don&#39;t have score, just pick one of the transcript where assessment is available:</span>
            <span class="k">else</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">transcript_column_name</span><span class="p">),</span>
                <span class="k">lambda</span> <span class="n">transcript</span><span class="p">:</span> <span class="n">transcript</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span>
                    <span class="n">assessment_column_name</span>
                <span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># Get assessment:</span>
        <span class="n">assessment</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;assessment&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">assessment_column_name</span> <span class="ow">is</span> <span class="kc">None</span>
            <span class="k">else</span> <span class="n">most_severe_transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="n">assessment_column_name</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                <span class="s2">&quot;assessment&quot;</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># Get score:</span>
        <span class="n">score</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">FloatType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;score&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">score_column_name</span> <span class="ow">is</span> <span class="kc">None</span>
            <span class="k">else</span> <span class="n">most_severe_transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="n">score_column_name</span><span class="p">)</span>
            <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">FloatType</span><span class="p">())</span>
            <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;score&quot;</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># Get assessment flag:</span>
        <span class="n">assessment_flag</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;assessmentFlag&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">assessment_flag_column_name</span> <span class="ow">is</span> <span class="kc">None</span>
            <span class="k">else</span> <span class="n">most_severe_transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="n">assessment_flag_column_name</span><span class="p">)</span>
            <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">())</span>
            <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;assessmentFlag&quot;</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="c1"># Extract gene id:</span>
        <span class="n">gene_id</span> <span class="o">=</span> <span class="n">most_severe_transcript</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;gene_id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;targetId&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">most_severe_transcript</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="n">method_name</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;method&quot;</span><span class="p">),</span>
                <span class="n">assessment</span><span class="p">,</span>
                <span class="n">score</span><span class="p">,</span>
                <span class="n">assessment_flag</span><span class="p">,</span>
                <span class="n">gene_id</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_parser_amino_acid_change</span><span class="p">(</span><span class="n">amino_acids</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">protein_end</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Convert VEP amino acid change information to one letter code aa substitution code.</span>

<span class="sd">        The logic assumes that the amino acid change is given in the format &quot;from/to&quot; and the protein end is given also.</span>
<span class="sd">        If any of the information is missing, the amino acid change will be set to None.</span>

<span class="sd">        Args:</span>
<span class="sd">            amino_acids (Column): Amino acid change information.</span>
<span class="sd">            protein_end (Column): Protein end information.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Amino acid change in one letter code.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; data = [(&#39;A/B&#39;, 1),(&#39;A/B&#39;, None),(None, 1),]</span>
<span class="sd">            &gt;&gt;&gt; (</span>
<span class="sd">            ...    spark.createDataFrame(data, [&#39;amino_acids&#39;, &#39;protein_end&#39;])</span>
<span class="sd">            ...    .select(VariantEffectPredictorParser._parser_amino_acid_change(f.col(&#39;amino_acids&#39;), f.col(&#39;protein_end&#39;)).alias(&#39;amino_acid_change&#39;))</span>
<span class="sd">            ...    .show()</span>
<span class="sd">            ... )</span>
<span class="sd">            +-----------------+</span>
<span class="sd">            |amino_acid_change|</span>
<span class="sd">            +-----------------+</span>
<span class="sd">            |              A1B|</span>
<span class="sd">            |             NULL|</span>
<span class="sd">            |             NULL|</span>
<span class="sd">            +-----------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">amino_acids</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span> <span class="o">&amp;</span> <span class="n">protein_end</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">amino_acids</span><span class="p">,</span> <span class="sa">r</span><span class="s2">&quot;\/&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span>
                <span class="n">protein_end</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">amino_acids</span><span class="p">,</span> <span class="sa">r</span><span class="s2">&quot;\/&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span>
            <span class="p">),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">))</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_collect_uniprot_accessions</span><span class="p">(</span><span class="n">trembl</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">swissprot</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Flatten arrays containing Uniprot accessions.</span>

<span class="sd">        Args:</span>
<span class="sd">            trembl (Column): List of TrEMBL protein accessions.</span>
<span class="sd">            swissprot (Column): List of SwissProt protein accessions.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: List of unique Uniprot accessions extracted from swissprot and trembl arrays, splitted from version numbers.</span>

<span class="sd">        Examples:</span>
<span class="sd">        &gt;&gt;&gt; data = [(&#39;v1&#39;, [&quot;sp_1&quot;], [&quot;tr_1&quot;]), (&#39;v1&#39;, None, None), (&#39;v1&#39;, None, [&quot;tr_2&quot;]),]</span>
<span class="sd">        &gt;&gt;&gt; (</span>
<span class="sd">        ...    spark.createDataFrame(data, [&#39;v&#39;, &#39;sp&#39;, &#39;tr&#39;])</span>
<span class="sd">        ...    .select(VariantEffectPredictorParser._collect_uniprot_accessions(f.col(&#39;sp&#39;), f.col(&#39;tr&#39;)).alias(&#39;proteinIds&#39;))</span>
<span class="sd">        ...    .show()</span>
<span class="sd">        ... )</span>
<span class="sd">        +------------+</span>
<span class="sd">        |  proteinIds|</span>
<span class="sd">        +------------+</span>
<span class="sd">        |[sp_1, tr_1]|</span>
<span class="sd">        |          []|</span>
<span class="sd">        |      [tr_2]|</span>
<span class="sd">        +------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Dropping Null values and flattening the arrays:</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">array_distinct</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">trembl</span><span class="p">,</span> <span class="n">swissprot</span><span class="p">),</span>
                            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
                        <span class="p">)</span>
                    <span class="p">),</span>
                    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="sa">r</span><span class="s2">&quot;\.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_parse_variant_location_id</span><span class="p">(</span><span class="n">vep_input_field</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">Column</span><span class="p">]:</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;Parse variant identifier, chromosome, position, reference allele and alternate allele from VEP input field.</span>

<span class="sd">        Args:</span>
<span class="sd">            vep_input_field (Column): Column containing variant vcf string used as VEP input.</span>

<span class="sd">        Returns:</span>
<span class="sd">            list[Column]: List of columns containing chromosome, position, reference allele and alternate allele.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">variant_fields</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">vep_input_field</span><span class="p">,</span> <span class="sa">r</span><span class="s2">&quot;\t&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">[</span>
            <span class="n">f</span><span class="o">.</span><span class="n">concat_ws</span><span class="p">(</span>
                <span class="s2">&quot;_&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                    <span class="n">variant_fields</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                    <span class="n">variant_fields</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                    <span class="n">variant_fields</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span>
                    <span class="n">variant_fields</span><span class="p">[</span><span class="mi">4</span><span class="p">],</span>
                <span class="p">),</span>
            <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
            <span class="n">variant_fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
            <span class="n">variant_fields</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">IntegerType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">),</span>
            <span class="n">variant_fields</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;referenceAllele&quot;</span><span class="p">),</span>
            <span class="n">variant_fields</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alternateAllele&quot;</span><span class="p">),</span>
        <span class="p">]</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">process_vep_output</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">,</span> <span class="n">vep_output</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">hash_threshold</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Process and format a VEP output in JSON format.</span>

<span class="sd">        Args:</span>
<span class="sd">            vep_output (DataFrame): raw VEP output, read as spark DataFrame.</span>
<span class="sd">            hash_threshold (int): threshold above which variant identifiers will be hashed.</span>

<span class="sd">        Returns:</span>
<span class="sd">           DataFrame: processed data in the right shape.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Processing VEP output:</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="n">vep_output</span>
            <span class="c1"># Dropping non-canonical transcript consequences:  # TODO: parametrize this.</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;transcript_consequences&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">),</span>
                    <span class="k">lambda</span> <span class="n">consequence</span><span class="p">:</span> <span class="n">consequence</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;canonical&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="c1"># Parse id and chr:pos:alt:ref:</span>
                <span class="o">*</span><span class="bp">cls</span><span class="o">.</span><span class="n">_parse_variant_location_id</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;input&quot;</span><span class="p">)),</span>
                <span class="c1"># Extracting corss_references from colocated variants:</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">_extract_ensembl_xrefs</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;colocated_variants&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                    <span class="s2">&quot;ensembl_xrefs&quot;</span>
                <span class="p">),</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">_extract_omim_xrefs</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;colocated_variants&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                    <span class="s2">&quot;omim_xrefs&quot;</span>
                <span class="p">),</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">_extract_clinvar_xrefs</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;colocated_variants&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                    <span class="s2">&quot;clinvar_xrefs&quot;</span>
                <span class="p">),</span>
                <span class="c1"># Extracting variant effect assessments</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="c1"># The following variant effect assessments are only available for variants with transcript consequences:</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                            <span class="c1"># Extract CADD scores:</span>
                            <span class="bp">cls</span><span class="o">.</span><span class="n">_vep_variant_effect_extractor</span><span class="p">(</span>
                                <span class="n">transcript_column_name</span><span class="o">=</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">,</span>
                                <span class="n">method_name</span><span class="o">=</span><span class="s2">&quot;CADD&quot;</span><span class="p">,</span>
                                <span class="n">score_column_name</span><span class="o">=</span><span class="s2">&quot;cadd_phred&quot;</span><span class="p">,</span>
                            <span class="p">),</span>
                            <span class="c1"># Extract polyphen scores:</span>
                            <span class="bp">cls</span><span class="o">.</span><span class="n">_vep_variant_effect_extractor</span><span class="p">(</span>
                                <span class="n">transcript_column_name</span><span class="o">=</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">,</span>
                                <span class="n">method_name</span><span class="o">=</span><span class="s2">&quot;PolyPhen&quot;</span><span class="p">,</span>
                                <span class="n">score_column_name</span><span class="o">=</span><span class="s2">&quot;polyphen_score&quot;</span><span class="p">,</span>
                                <span class="n">assessment_column_name</span><span class="o">=</span><span class="s2">&quot;polyphen_prediction&quot;</span><span class="p">,</span>
                            <span class="p">),</span>
                            <span class="c1"># Extract sift scores:</span>
                            <span class="bp">cls</span><span class="o">.</span><span class="n">_vep_variant_effect_extractor</span><span class="p">(</span>
                                <span class="n">transcript_column_name</span><span class="o">=</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">,</span>
                                <span class="n">method_name</span><span class="o">=</span><span class="s2">&quot;SIFT&quot;</span><span class="p">,</span>
                                <span class="n">score_column_name</span><span class="o">=</span><span class="s2">&quot;sift_score&quot;</span><span class="p">,</span>
                                <span class="n">assessment_column_name</span><span class="o">=</span><span class="s2">&quot;sift_prediction&quot;</span><span class="p">,</span>
                            <span class="p">),</span>
                            <span class="c1"># Extract loftee scores:</span>
                            <span class="bp">cls</span><span class="o">.</span><span class="n">_vep_variant_effect_extractor</span><span class="p">(</span>
                                <span class="n">method_name</span><span class="o">=</span><span class="s2">&quot;LOFTEE&quot;</span><span class="p">,</span>
                                <span class="n">transcript_column_name</span><span class="o">=</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">,</span>
                                <span class="n">score_column_name</span><span class="o">=</span><span class="s2">&quot;lof&quot;</span><span class="p">,</span>
                                <span class="n">assessment_column_name</span><span class="o">=</span><span class="s2">&quot;lof&quot;</span><span class="p">,</span>
                                <span class="n">assessment_flag_column_name</span><span class="o">=</span><span class="s2">&quot;lof_filter&quot;</span><span class="p">,</span>
                            <span class="p">),</span>
                            <span class="c1"># Extract GERP conservation score:</span>
                            <span class="bp">cls</span><span class="o">.</span><span class="n">_vep_variant_effect_extractor</span><span class="p">(</span>
                                <span class="n">method_name</span><span class="o">=</span><span class="s2">&quot;GERP&quot;</span><span class="p">,</span>
                                <span class="n">transcript_column_name</span><span class="o">=</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">,</span>
                                <span class="n">score_column_name</span><span class="o">=</span><span class="s2">&quot;conservation&quot;</span><span class="p">,</span>
                            <span class="p">),</span>
                            <span class="c1"># Extract max alpha missense:</span>
                            <span class="bp">cls</span><span class="o">.</span><span class="n">_get_max_alpha_missense</span><span class="p">(</span>
                                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">)</span>
                            <span class="p">),</span>
                            <span class="c1"># Extract VEP prediction:</span>
                            <span class="bp">cls</span><span class="o">.</span><span class="n">_get_vep_prediction</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;most_severe_consequence&quot;</span><span class="p">)),</span>
                        <span class="p">),</span>
                        <span class="k">lambda</span> <span class="n">predictor</span><span class="p">:</span> <span class="n">predictor</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                        <span class="c1"># Extract VEP prediction:</span>
                        <span class="bp">cls</span><span class="o">.</span><span class="n">_get_vep_prediction</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;most_severe_consequence&quot;</span><span class="p">)),</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;variantEffect&quot;</span><span class="p">),</span>
                <span class="c1"># Convert consequence to SO:</span>
                <span class="n">map_column_by_dictionary</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;most_severe_consequence&quot;</span><span class="p">),</span> <span class="bp">cls</span><span class="o">.</span><span class="n">SEQUENCE_ONTOLOGY_MAP</span>
                <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;mostSevereConsequenceId&quot;</span><span class="p">),</span>
                <span class="c1"># Propagate most severe consequence:</span>
                <span class="s2">&quot;most_severe_consequence&quot;</span><span class="p">,</span>
                <span class="c1"># Extract HGVS identifier:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;hgvsg&quot;</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="s2">&quot;intergenic_consequences&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;intergenic_consequences&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;hgvsg&quot;</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">))</span>
                <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;hgvsId&quot;</span><span class="p">),</span>
                <span class="c1"># Collect transcript consequence:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">),</span>
                        <span class="k">lambda</span> <span class="n">transcript</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                            <span class="c1"># Convert consequence terms to SO identifier:</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                                <span class="n">transcript</span><span class="o">.</span><span class="n">consequence_terms</span><span class="p">,</span>
                                <span class="k">lambda</span> <span class="n">y</span><span class="p">:</span> <span class="n">map_column_by_dictionary</span><span class="p">(</span>
                                    <span class="n">y</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">SEQUENCE_ONTOLOGY_MAP</span>
                                <span class="p">),</span>
                            <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;variantFunctionalConsequenceIds&quot;</span><span class="p">),</span>
                            <span class="c1"># Convert consequence terms to consequence score:</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">array_max</span><span class="p">(</span>
                                <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                                    <span class="n">transcript</span><span class="o">.</span><span class="n">consequence_terms</span><span class="p">,</span>
                                    <span class="k">lambda</span> <span class="n">term</span><span class="p">:</span> <span class="n">map_column_by_dictionary</span><span class="p">(</span>
                                        <span class="n">term</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">LABEL_TO_SCORE_MAP</span>
                                    <span class="p">),</span>
                                <span class="p">)</span>
                            <span class="p">)</span>
                            <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">FloatType</span><span class="p">())</span>
                            <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;consequenceScore&quot;</span><span class="p">),</span>
                            <span class="c1"># Format amino acid change:</span>
                            <span class="bp">cls</span><span class="o">.</span><span class="n">_parser_amino_acid_change</span><span class="p">(</span>
                                <span class="n">transcript</span><span class="o">.</span><span class="n">amino_acids</span><span class="p">,</span> <span class="n">transcript</span><span class="o">.</span><span class="n">protein_end</span>
                            <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;aminoAcidChange&quot;</span><span class="p">),</span>
                            <span class="c1"># Extract and clean uniprot identifiers:</span>
                            <span class="bp">cls</span><span class="o">.</span><span class="n">_collect_uniprot_accessions</span><span class="p">(</span>
                                <span class="n">transcript</span><span class="o">.</span><span class="n">swissprot</span><span class="p">,</span>
                                <span class="n">transcript</span><span class="o">.</span><span class="n">trembl</span><span class="p">,</span>
                            <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;uniprotAccessions&quot;</span><span class="p">),</span>
                            <span class="c1"># Add canonical flag:</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">canonical</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
                            <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                            <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;isEnsemblCanonical&quot;</span><span class="p">),</span>
                            <span class="c1"># Extract footprint distance:</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">codons</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;codons&quot;</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span> <span class="n">transcript</span><span class="o">.</span><span class="n">distance</span><span class="p">)</span>
                            <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mi">0</span><span class="p">))</span>
                            <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">LongType</span><span class="p">())</span>
                            <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;distanceFromFootprint&quot;</span><span class="p">),</span>
                            <span class="c1"># Extract distance from the transcription start site:</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">tssdistance</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">LongType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                                <span class="s2">&quot;distanceFromTss&quot;</span>
                            <span class="p">),</span>
                            <span class="c1"># Extracting APPRIS isoform annotation for this transcript:</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">appris</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;appris&quot;</span><span class="p">),</span>
                            <span class="c1"># Extracting MANE select transcript:</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">mane_select</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;maneSelect&quot;</span><span class="p">),</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">gene_id</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;targetId&quot;</span><span class="p">),</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">impact</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;impact&quot;</span><span class="p">),</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">lof</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                                <span class="s2">&quot;lofteePrediction&quot;</span>
                            <span class="p">),</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">lof</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">FloatType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;siftPrediction&quot;</span><span class="p">),</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">lof</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">FloatType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                                <span class="s2">&quot;polyphenPrediction&quot;</span>
                            <span class="p">),</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">transcript_id</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;transcriptId&quot;</span><span class="p">),</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">biotype</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;biotype&quot;</span><span class="p">),</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">gene_symbol</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;approvedSymbol&quot;</span><span class="p">),</span>
                        <span class="p">),</span>
                    <span class="p">),</span>
                <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">),</span>
                <span class="c1"># Extracting rsids:</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">_colocated_variants_to_rsids</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;colocated_variants&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                    <span class="s2">&quot;rsIds&quot;</span>
                <span class="p">),</span>
                <span class="c1"># Adding empty array for allele frequencies - now this piece of data is not coming form the VEP data:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">()</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="bp">cls</span><span class="o">.</span><span class="n">ALLELE_FREQUENCY_SCHEMA</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alleleFrequencies&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Dropping transcripts where the consequence score or the distance is null:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">),</span>
                    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;consequenceScore&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
                    <span class="o">&amp;</span> <span class="n">x</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;distanceFromFootprint&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Sort transcript consequences by consequence score and distance from footprint and add index:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                        <span class="n">order_array_of_structs_by_two_fields</span><span class="p">(</span>
                            <span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;consequenceScore&quot;</span><span class="p">,</span>
                            <span class="s2">&quot;distanceFromFootprint&quot;</span><span class="p">,</span>
                        <span class="p">),</span>
                        <span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">i</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">withField</span><span class="p">(</span><span class="s2">&quot;transcriptIndex&quot;</span><span class="p">,</span> <span class="n">i</span> <span class="o">+</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mi">1</span><span class="p">)),</span>
                    <span class="p">),</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Adding protvar xref for missense variants:  # TODO: making and extendable list of consequences</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;protvar_xrefs&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">),</span>
                            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                                <span class="n">x</span><span class="o">.</span><span class="n">variantFunctionalConsequenceIds</span><span class="p">,</span> <span class="s2">&quot;SO_0001583&quot;</span>
                            <span class="p">),</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                    <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">_generate_dbxrefs</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">)),</span> <span class="s2">&quot;protvar&quot;</span><span class="p">),</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;dbXrefs&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ensembl_xrefs&quot;</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;omim_xrefs&quot;</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clinvar_xrefs&quot;</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;protvar_xrefs&quot;</span><span class="p">),</span>
                        <span class="p">),</span>
                        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
                    <span class="p">)</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># If the variantId is too long, hash it:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;variantId&quot;</span><span class="p">,</span>
                <span class="n">VariantIndex</span><span class="o">.</span><span class="n">hash_long_variant_ids</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">),</span>
                    <span class="n">hash_threshold</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Generating a temporary column with only protein coding transcripts:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;proteinCodingTranscripts&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">),</span>
                    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;biotype&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;protein_coding&quot;</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Generate variant descrioption:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;variantDescription&quot;</span><span class="p">,</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">_compose_variant_description</span><span class="p">(</span>
                    <span class="c1"># Passing the most severe consequence:</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;most_severe_consequence&quot;</span><span class="p">),</span>
                    <span class="c1"># The first transcript:</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">),</span>
                        <span class="k">lambda</span> <span class="n">vep</span><span class="p">:</span> <span class="n">vep</span><span class="o">.</span><span class="n">transcriptIndex</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
                    <span class="c1"># The first protein coding transcript:</span>
                    <span class="n">order_array_of_structs_by_field</span><span class="p">(</span>
                        <span class="s2">&quot;proteinCodingTranscripts&quot;</span><span class="p">,</span> <span class="s2">&quot;transcriptIndex&quot;</span>
                    <span class="p">)[</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="s2">&quot;proteinCodingTranscripts&quot;</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">],</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Normalising variant effect assessments:</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;variantEffect&quot;</span><span class="p">,</span>
                <span class="n">VariantEffectNormaliser</span><span class="o">.</span><span class="n">normalise_variant_effect</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantEffect&quot;</span><span class="p">)</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># Dropping intermediate xref columns:</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span>
                <span class="o">*</span><span class="p">[</span>
                    <span class="s2">&quot;ensembl_xrefs&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;omim_xrefs&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;clinvar_xrefs&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;protvar_xrefs&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;most_severe_consequence&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;proteinCodingTranscripts&quot;</span><span class="p">,</span>
                <span class="p">]</span>
            <span class="p">)</span>
            <span class="c1"># Drooping rows with null position:</span>
            <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">())</span>
        <span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_compose_variant_description</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">VariantEffectPredictorParser</span><span class="p">],</span>
        <span class="n">most_severe_consequence</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span>
        <span class="n">first_transcript</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span>
        <span class="n">first_protein_coding</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Compose variant description based on the most severe consequence.</span>

<span class="sd">        Args:</span>
<span class="sd">            most_severe_consequence (Column): Most severe consequence</span>
<span class="sd">            first_transcript (Column): First transcript</span>
<span class="sd">            first_protein_coding (Column): First protein coding transcript</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Variant description</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="c1"># When there&#39;s no transcript whatsoever:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="n">first_transcript</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;Intergenic variant no gene in window&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># When the biotype of the first gene is protein coding:</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="n">first_transcript</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;biotype&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;protein_coding&quot;</span><span class="p">,</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">_process_protein_coding_transcript</span><span class="p">(</span>
                    <span class="n">first_transcript</span><span class="p">,</span> <span class="n">most_severe_consequence</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># When the first gene is not protein coding, we also pass the first protein coding gene:</span>
            <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">_process_non_protein_coding_transcript</span><span class="p">(</span>
                    <span class="n">most_severe_consequence</span><span class="p">,</span> <span class="n">first_transcript</span><span class="p">,</span> <span class="n">first_protein_coding</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_process_consequence_term</span><span class="p">(</span><span class="n">consequence_term</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Cleaning up consequence term: capitalizing and replacing underscores.</span>

<span class="sd">        Args:</span>
<span class="sd">            consequence_term (Column): Consequence term.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Cleaned up consequence term.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">last</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">consequence_term</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;variant&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span>
            <span class="s2">&quot; variant&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">regexp_replace</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">initcap</span><span class="p">(</span><span class="n">consequence_term</span><span class="p">),</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">),</span> <span class="n">last</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_process_overlap</span><span class="p">(</span><span class="n">transcript</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Process overlap with gene: if the variant overlaps with the gene, return the gene name or distance.</span>

<span class="sd">        Args:</span>
<span class="sd">            transcript (Column): Transcript.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: string column with overlap description.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">gene_label</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;approvedSymbol&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;approvedSymbol&quot;</span><span class="p">),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;targetId&quot;</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;distanceFromFootprint&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span>
            <span class="c1"># &quot;overlapping with CCDC8&quot;</span>
            <span class="n">f</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot; overlapping with &quot;</span><span class="p">),</span> <span class="n">gene_label</span><span class="p">),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span>
            <span class="c1"># &quot; 123 basepair away from CCDC8&quot;</span>
            <span class="n">f</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">format_number</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;distanceFromFootprint&quot;</span><span class="p">),</span> <span class="mi">0</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot; basepair away from &quot;</span><span class="p">),</span>
                <span class="n">gene_label</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_process_aa_change</span><span class="p">(</span><span class="n">transcript</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract amino acid change information from transcript when available.</span>

<span class="sd">        Args:</span>
<span class="sd">            transcript (Column): Transcript.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Amino acid change information.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;aminoAcidChange&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;, causing amio-acid change: &quot;</span><span class="p">),</span>
                <span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;aminoAcidChange&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot; with &quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lower</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;impact&quot;</span><span class="p">)),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot; impact.&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">))</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_process_lof</span><span class="p">(</span><span class="n">transcript</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Process loss of function annotation from LOFTEE prediction.</span>

<span class="sd">        Args:</span>
<span class="sd">            transcript (Column): Transcript.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Loss of function annotation.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;lofteePrediction&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;lofteePrediction&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;HC&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot; A high-confidence loss-of-function variant by loftee.&quot;</span><span class="p">),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">))</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_process_protein_coding_transcript</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">VariantEffectPredictorParser</span><span class="p">],</span>
        <span class="n">transcript</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span>
        <span class="n">most_severe_consequence</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract information from the first, protein coding transcript.</span>

<span class="sd">        Args:</span>
<span class="sd">            transcript (Column): Transcript.</span>
<span class="sd">            most_severe_consequence (Column): Most severe consequence.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Variant description.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Process consequence term:</span>
        <span class="n">consequence_text</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_process_consequence_term</span><span class="p">(</span><span class="n">most_severe_consequence</span><span class="p">)</span>

        <span class="c1"># Does it overlap with the gene:</span>
        <span class="n">overlap</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_process_overlap</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span>

        <span class="c1"># Does it cause amino acid change:</span>
        <span class="n">amino_acid_change</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_process_aa_change</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span>

        <span class="c1"># Processing lof annotation:</span>
        <span class="n">lof_assessment</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_process_lof</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span>

        <span class="c1"># Concat all together:</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">consequence_text</span><span class="p">,</span> <span class="n">overlap</span><span class="p">,</span> <span class="n">amino_acid_change</span><span class="p">,</span> <span class="n">lof_assessment</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_adding_biotype</span><span class="p">(</span><span class="n">transcript</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Adding biotype information to the variant description.</span>

<span class="sd">        Args:</span>
<span class="sd">            transcript (Column): Transcript.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Biotype information.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">biotype</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;biotype&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">regexp_replace</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;biotype&quot;</span><span class="p">),</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">regexp_replace</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;biotype&quot;</span><span class="p">),</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot; gene.&quot;</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;, a &quot;</span><span class="p">),</span> <span class="n">biotype</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_parse_protein_coding_transcript</span><span class="p">(</span><span class="n">transcript</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Parse the closest, not first protein coding transcript: extract gene symbol and distance.</span>

<span class="sd">        Args:</span>
<span class="sd">            transcript (Column): Transcript.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Protein coding transcript information.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">gene_label</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;approvedSymbol&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;approvedSymbol&quot;</span><span class="p">),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;targetId&quot;</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">transcript</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot; The closest protein-coding gene is &quot;</span><span class="p">),</span>
                <span class="n">gene_label</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot; (&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">format_number</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;distanceFromFootprint&quot;</span><span class="p">),</span> <span class="mi">0</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot; basepair away).&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">))</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">_process_non_protein_coding_transcript</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">VariantEffectPredictorParser</span><span class="p">],</span>
        <span class="n">most_severe_consequence</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span>
        <span class="n">first_transcript</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span>
        <span class="n">first_protein_coding</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract information from the first, non-protein coding transcript.</span>

<span class="sd">        Args:</span>
<span class="sd">            most_severe_consequence (Column): Most severe consequence.</span>
<span class="sd">            first_transcript (Column): First transcript.</span>
<span class="sd">            first_protein_coding (Column): First protein coding transcript.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Variant description.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Process consequence term:</span>
        <span class="n">consequence_text</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_process_consequence_term</span><span class="p">(</span><span class="n">most_severe_consequence</span><span class="p">)</span>

        <span class="c1"># Does it overlap with the gene:</span>
        <span class="n">overlap</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_process_overlap</span><span class="p">(</span><span class="n">first_transcript</span><span class="p">)</span>

        <span class="c1"># Adding biotype:</span>
        <span class="n">biotype</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_adding_biotype</span><span class="p">(</span><span class="n">first_transcript</span><span class="p">)</span>

        <span class="c1"># Adding protein coding gene:</span>
        <span class="n">protein_transcript</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_parse_protein_coding_transcript</span><span class="p">(</span><span class="n">first_protein_coding</span><span class="p">)</span>

        <span class="c1"># Concat all together:</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">consequence_text</span><span class="p">,</span> <span class="n">overlap</span><span class="p">,</span> <span class="n">biotype</span><span class="p">,</span> <span class="n">protein_transcript</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser.extract_variant_index_from_vep" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">extract_variant_index_from_vep</span><span class="p">(</span><span class="n">spark</span><span class="p">:</span> <span class="n">SparkSession</span><span class="p">,</span> <span class="n">vep_output_path</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">hash_threshold</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">|</span> <span class="nb">float</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">VariantIndex</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser.extract_variant_index_from_vep" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Extract variant index from VEP output.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>spark</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.SparkSession">SparkSession</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Spark session.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>vep_output_path</code>
            </td>
            <td>
                  <code>str | list[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the VEP output.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>hash_threshold</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Threshold above which variant identifiers will be hashed.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>**kwargs</code>
            </td>
            <td>
                  <code>bool | float | int | str | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Additional arguments to pass to spark.read.json.</p>
              </div>
            </td>
            <td>
                  <code>{}</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>VariantIndex</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.variant_index.VariantIndex" href="../../../datasets/variant_index/#gentropy.dataset.variant_index.VariantIndex">VariantIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Variant index dataset.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Failed reading file or if the dataset is empty.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/ensembl/vep_parser.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">extract_variant_index_from_vep</span><span class="p">(</span>
    <span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">VariantEffectPredictorParser</span><span class="p">],</span>
    <span class="n">spark</span><span class="p">:</span> <span class="n">SparkSession</span><span class="p">,</span>
    <span class="n">vep_output_path</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">hash_threshold</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">|</span> <span class="nb">float</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">VariantIndex</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract variant index from VEP output.</span>

<span class="sd">    Args:</span>
<span class="sd">        spark (SparkSession): Spark session.</span>
<span class="sd">        vep_output_path (str | list[str]): Path to the VEP output.</span>
<span class="sd">        hash_threshold (int): Threshold above which variant identifiers will be hashed.</span>
<span class="sd">        **kwargs (bool | float | int | str | None): Additional arguments to pass to spark.read.json.</span>

<span class="sd">    Returns:</span>
<span class="sd">        VariantIndex: Variant index dataset.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: Failed reading file or if the dataset is empty.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># To speed things up and simplify the json structure, read data following an expected schema:</span>
    <span class="n">vep_schema</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">get_schema</span><span class="p">()</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">vep_data</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">json</span><span class="p">(</span><span class="n">vep_output_path</span><span class="p">,</span> <span class="n">schema</span><span class="o">=</span><span class="n">vep_schema</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed reading file: </span><span class="si">{</span><span class="n">vep_output_path</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span> <span class="kn">from</span><span class="w"> </span><span class="nn">e</span>

    <span class="k">if</span> <span class="n">vep_data</span><span class="o">.</span><span class="n">isEmpty</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The dataset is empty: </span><span class="si">{</span><span class="n">vep_output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Convert to VariantAnnotation dataset:</span>
    <span class="k">return</span> <span class="n">VariantIndex</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="n">VariantEffectPredictorParser</span><span class="o">.</span><span class="n">process_vep_output</span><span class="p">(</span>
            <span class="n">vep_data</span><span class="p">,</span> <span class="n">hash_threshold</span>
        <span class="p">),</span>
        <span class="n">_schema</span><span class="o">=</span><span class="n">VariantIndex</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="n">id_threshold</span><span class="o">=</span><span class="n">hash_threshold</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser.get_schema" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_schema</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">t</span><span class="o">.</span><span class="n">StructType</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

<a href="#gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser.get_schema" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Return the schema of the VEP output.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pyspark.sql.types.StructType">StructType</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>t.StructType: VEP output schema.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">VariantEffectPredictorParser</span><span class="o">.</span><span class="n">get_schema</span><span class="p">())</span>
<span class="go">&lt;class &#39;pyspark.sql.types.StructType&#39;&gt;</span>
</code></pre></div>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/ensembl/vep_parser.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">get_schema</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">t</span><span class="o">.</span><span class="n">StructType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Return the schema of the VEP output.</span>

<span class="sd">    Returns:</span>
<span class="sd">        t.StructType: VEP output schema.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; type(VariantEffectPredictorParser.get_schema())</span>
<span class="sd">        &lt;class &#39;pyspark.sql.types.StructType&#39;&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">parse_spark_schema</span><span class="p">(</span><span class="s2">&quot;vep_json_output.json&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser.process_vep_output" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_vep_output</span><span class="p">(</span><span class="n">vep_output</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">hash_threshold</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.datasource.ensembl.vep_parser.VariantEffectPredictorParser.process_vep_output" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

        <p>Process and format a VEP output in JSON format.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>vep_output</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>raw VEP output, read as spark DataFrame.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>hash_threshold</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>threshold above which variant identifiers will be hashed.</p>
              </div>
            </td>
            <td>
                  <code>100</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>DataFrame</code></td>            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>processed data in the right shape.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/ensembl/vep_parser.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">process_vep_output</span><span class="p">(</span>
    <span class="bp">cls</span><span class="p">,</span> <span class="n">vep_output</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">hash_threshold</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Process and format a VEP output in JSON format.</span>

<span class="sd">    Args:</span>
<span class="sd">        vep_output (DataFrame): raw VEP output, read as spark DataFrame.</span>
<span class="sd">        hash_threshold (int): threshold above which variant identifiers will be hashed.</span>

<span class="sd">    Returns:</span>
<span class="sd">       DataFrame: processed data in the right shape.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Processing VEP output:</span>
    <span class="k">return</span> <span class="p">(</span>
        <span class="n">vep_output</span>
        <span class="c1"># Dropping non-canonical transcript consequences:  # TODO: parametrize this.</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;transcript_consequences&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">),</span>
                <span class="k">lambda</span> <span class="n">consequence</span><span class="p">:</span> <span class="n">consequence</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;canonical&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="c1"># Parse id and chr:pos:alt:ref:</span>
            <span class="o">*</span><span class="bp">cls</span><span class="o">.</span><span class="n">_parse_variant_location_id</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;input&quot;</span><span class="p">)),</span>
            <span class="c1"># Extracting corss_references from colocated variants:</span>
            <span class="bp">cls</span><span class="o">.</span><span class="n">_extract_ensembl_xrefs</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;colocated_variants&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                <span class="s2">&quot;ensembl_xrefs&quot;</span>
            <span class="p">),</span>
            <span class="bp">cls</span><span class="o">.</span><span class="n">_extract_omim_xrefs</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;colocated_variants&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                <span class="s2">&quot;omim_xrefs&quot;</span>
            <span class="p">),</span>
            <span class="bp">cls</span><span class="o">.</span><span class="n">_extract_clinvar_xrefs</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;colocated_variants&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                <span class="s2">&quot;clinvar_xrefs&quot;</span>
            <span class="p">),</span>
            <span class="c1"># Extracting variant effect assessments</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="c1"># The following variant effect assessments are only available for variants with transcript consequences:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                        <span class="c1"># Extract CADD scores:</span>
                        <span class="bp">cls</span><span class="o">.</span><span class="n">_vep_variant_effect_extractor</span><span class="p">(</span>
                            <span class="n">transcript_column_name</span><span class="o">=</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">,</span>
                            <span class="n">method_name</span><span class="o">=</span><span class="s2">&quot;CADD&quot;</span><span class="p">,</span>
                            <span class="n">score_column_name</span><span class="o">=</span><span class="s2">&quot;cadd_phred&quot;</span><span class="p">,</span>
                        <span class="p">),</span>
                        <span class="c1"># Extract polyphen scores:</span>
                        <span class="bp">cls</span><span class="o">.</span><span class="n">_vep_variant_effect_extractor</span><span class="p">(</span>
                            <span class="n">transcript_column_name</span><span class="o">=</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">,</span>
                            <span class="n">method_name</span><span class="o">=</span><span class="s2">&quot;PolyPhen&quot;</span><span class="p">,</span>
                            <span class="n">score_column_name</span><span class="o">=</span><span class="s2">&quot;polyphen_score&quot;</span><span class="p">,</span>
                            <span class="n">assessment_column_name</span><span class="o">=</span><span class="s2">&quot;polyphen_prediction&quot;</span><span class="p">,</span>
                        <span class="p">),</span>
                        <span class="c1"># Extract sift scores:</span>
                        <span class="bp">cls</span><span class="o">.</span><span class="n">_vep_variant_effect_extractor</span><span class="p">(</span>
                            <span class="n">transcript_column_name</span><span class="o">=</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">,</span>
                            <span class="n">method_name</span><span class="o">=</span><span class="s2">&quot;SIFT&quot;</span><span class="p">,</span>
                            <span class="n">score_column_name</span><span class="o">=</span><span class="s2">&quot;sift_score&quot;</span><span class="p">,</span>
                            <span class="n">assessment_column_name</span><span class="o">=</span><span class="s2">&quot;sift_prediction&quot;</span><span class="p">,</span>
                        <span class="p">),</span>
                        <span class="c1"># Extract loftee scores:</span>
                        <span class="bp">cls</span><span class="o">.</span><span class="n">_vep_variant_effect_extractor</span><span class="p">(</span>
                            <span class="n">method_name</span><span class="o">=</span><span class="s2">&quot;LOFTEE&quot;</span><span class="p">,</span>
                            <span class="n">transcript_column_name</span><span class="o">=</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">,</span>
                            <span class="n">score_column_name</span><span class="o">=</span><span class="s2">&quot;lof&quot;</span><span class="p">,</span>
                            <span class="n">assessment_column_name</span><span class="o">=</span><span class="s2">&quot;lof&quot;</span><span class="p">,</span>
                            <span class="n">assessment_flag_column_name</span><span class="o">=</span><span class="s2">&quot;lof_filter&quot;</span><span class="p">,</span>
                        <span class="p">),</span>
                        <span class="c1"># Extract GERP conservation score:</span>
                        <span class="bp">cls</span><span class="o">.</span><span class="n">_vep_variant_effect_extractor</span><span class="p">(</span>
                            <span class="n">method_name</span><span class="o">=</span><span class="s2">&quot;GERP&quot;</span><span class="p">,</span>
                            <span class="n">transcript_column_name</span><span class="o">=</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">,</span>
                            <span class="n">score_column_name</span><span class="o">=</span><span class="s2">&quot;conservation&quot;</span><span class="p">,</span>
                        <span class="p">),</span>
                        <span class="c1"># Extract max alpha missense:</span>
                        <span class="bp">cls</span><span class="o">.</span><span class="n">_get_max_alpha_missense</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">)</span>
                        <span class="p">),</span>
                        <span class="c1"># Extract VEP prediction:</span>
                        <span class="bp">cls</span><span class="o">.</span><span class="n">_get_vep_prediction</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;most_severe_consequence&quot;</span><span class="p">)),</span>
                    <span class="p">),</span>
                    <span class="k">lambda</span> <span class="n">predictor</span><span class="p">:</span> <span class="n">predictor</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                    <span class="c1"># Extract VEP prediction:</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">_get_vep_prediction</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;most_severe_consequence&quot;</span><span class="p">)),</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;variantEffect&quot;</span><span class="p">),</span>
            <span class="c1"># Convert consequence to SO:</span>
            <span class="n">map_column_by_dictionary</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;most_severe_consequence&quot;</span><span class="p">),</span> <span class="bp">cls</span><span class="o">.</span><span class="n">SEQUENCE_ONTOLOGY_MAP</span>
            <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;mostSevereConsequenceId&quot;</span><span class="p">),</span>
            <span class="c1"># Propagate most severe consequence:</span>
            <span class="s2">&quot;most_severe_consequence&quot;</span><span class="p">,</span>
            <span class="c1"># Extract HGVS identifier:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;hgvsg&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="s2">&quot;intergenic_consequences&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;intergenic_consequences&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;hgvsg&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">))</span>
            <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;hgvsId&quot;</span><span class="p">),</span>
            <span class="c1"># Collect transcript consequence:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcript_consequences&quot;</span><span class="p">),</span>
                    <span class="k">lambda</span> <span class="n">transcript</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                        <span class="c1"># Convert consequence terms to SO identifier:</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">consequence_terms</span><span class="p">,</span>
                            <span class="k">lambda</span> <span class="n">y</span><span class="p">:</span> <span class="n">map_column_by_dictionary</span><span class="p">(</span>
                                <span class="n">y</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">SEQUENCE_ONTOLOGY_MAP</span>
                            <span class="p">),</span>
                        <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;variantFunctionalConsequenceIds&quot;</span><span class="p">),</span>
                        <span class="c1"># Convert consequence terms to consequence score:</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">array_max</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                                <span class="n">transcript</span><span class="o">.</span><span class="n">consequence_terms</span><span class="p">,</span>
                                <span class="k">lambda</span> <span class="n">term</span><span class="p">:</span> <span class="n">map_column_by_dictionary</span><span class="p">(</span>
                                    <span class="n">term</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">LABEL_TO_SCORE_MAP</span>
                                <span class="p">),</span>
                            <span class="p">)</span>
                        <span class="p">)</span>
                        <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">FloatType</span><span class="p">())</span>
                        <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;consequenceScore&quot;</span><span class="p">),</span>
                        <span class="c1"># Format amino acid change:</span>
                        <span class="bp">cls</span><span class="o">.</span><span class="n">_parser_amino_acid_change</span><span class="p">(</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">amino_acids</span><span class="p">,</span> <span class="n">transcript</span><span class="o">.</span><span class="n">protein_end</span>
                        <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;aminoAcidChange&quot;</span><span class="p">),</span>
                        <span class="c1"># Extract and clean uniprot identifiers:</span>
                        <span class="bp">cls</span><span class="o">.</span><span class="n">_collect_uniprot_accessions</span><span class="p">(</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">swissprot</span><span class="p">,</span>
                            <span class="n">transcript</span><span class="o">.</span><span class="n">trembl</span><span class="p">,</span>
                        <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;uniprotAccessions&quot;</span><span class="p">),</span>
                        <span class="c1"># Add canonical flag:</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">canonical</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;isEnsemblCanonical&quot;</span><span class="p">),</span>
                        <span class="c1"># Extract footprint distance:</span>
                        <span class="n">transcript</span><span class="o">.</span><span class="n">codons</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;codons&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">transcript</span><span class="o">.</span><span class="n">distance</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span> <span class="n">transcript</span><span class="o">.</span><span class="n">distance</span><span class="p">)</span>
                        <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mi">0</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">LongType</span><span class="p">())</span>
                        <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;distanceFromFootprint&quot;</span><span class="p">),</span>
                        <span class="c1"># Extract distance from the transcription start site:</span>
                        <span class="n">transcript</span><span class="o">.</span><span class="n">tssdistance</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">LongType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                            <span class="s2">&quot;distanceFromTss&quot;</span>
                        <span class="p">),</span>
                        <span class="c1"># Extracting APPRIS isoform annotation for this transcript:</span>
                        <span class="n">transcript</span><span class="o">.</span><span class="n">appris</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;appris&quot;</span><span class="p">),</span>
                        <span class="c1"># Extracting MANE select transcript:</span>
                        <span class="n">transcript</span><span class="o">.</span><span class="n">mane_select</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;maneSelect&quot;</span><span class="p">),</span>
                        <span class="n">transcript</span><span class="o">.</span><span class="n">gene_id</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;targetId&quot;</span><span class="p">),</span>
                        <span class="n">transcript</span><span class="o">.</span><span class="n">impact</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;impact&quot;</span><span class="p">),</span>
                        <span class="n">transcript</span><span class="o">.</span><span class="n">lof</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                            <span class="s2">&quot;lofteePrediction&quot;</span>
                        <span class="p">),</span>
                        <span class="n">transcript</span><span class="o">.</span><span class="n">lof</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">FloatType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;siftPrediction&quot;</span><span class="p">),</span>
                        <span class="n">transcript</span><span class="o">.</span><span class="n">lof</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">FloatType</span><span class="p">())</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                            <span class="s2">&quot;polyphenPrediction&quot;</span>
                        <span class="p">),</span>
                        <span class="n">transcript</span><span class="o">.</span><span class="n">transcript_id</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;transcriptId&quot;</span><span class="p">),</span>
                        <span class="n">transcript</span><span class="o">.</span><span class="n">biotype</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;biotype&quot;</span><span class="p">),</span>
                        <span class="n">transcript</span><span class="o">.</span><span class="n">gene_symbol</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;approvedSymbol&quot;</span><span class="p">),</span>
                    <span class="p">),</span>
                <span class="p">),</span>
            <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">),</span>
            <span class="c1"># Extracting rsids:</span>
            <span class="bp">cls</span><span class="o">.</span><span class="n">_colocated_variants_to_rsids</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;colocated_variants&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span>
                <span class="s2">&quot;rsIds&quot;</span>
            <span class="p">),</span>
            <span class="c1"># Adding empty array for allele frequencies - now this piece of data is not coming form the VEP data:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">()</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="bp">cls</span><span class="o">.</span><span class="n">ALLELE_FREQUENCY_SCHEMA</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alleleFrequencies&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="c1"># Dropping transcripts where the consequence score or the distance is null:</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">),</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;consequenceScore&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
                <span class="o">&amp;</span> <span class="n">x</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;distanceFromFootprint&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="c1"># Sort transcript consequences by consequence score and distance from footprint and add index:</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                    <span class="n">order_array_of_structs_by_two_fields</span><span class="p">(</span>
                        <span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;consequenceScore&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;distanceFromFootprint&quot;</span><span class="p">,</span>
                    <span class="p">),</span>
                    <span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">i</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">withField</span><span class="p">(</span><span class="s2">&quot;transcriptIndex&quot;</span><span class="p">,</span> <span class="n">i</span> <span class="o">+</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mi">1</span><span class="p">)),</span>
                <span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="c1"># Adding protvar xref for missense variants:  # TODO: making and extendable list of consequences</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;protvar_xrefs&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">),</span>
                        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span>
                            <span class="n">x</span><span class="o">.</span><span class="n">variantFunctionalConsequenceIds</span><span class="p">,</span> <span class="s2">&quot;SO_0001583&quot;</span>
                        <span class="p">),</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">_generate_dbxrefs</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">)),</span> <span class="s2">&quot;protvar&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;dbXrefs&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ensembl_xrefs&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;omim_xrefs&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clinvar_xrefs&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;protvar_xrefs&quot;</span><span class="p">),</span>
                    <span class="p">),</span>
                    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
                <span class="p">)</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="c1"># If the variantId is too long, hash it:</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;variantId&quot;</span><span class="p">,</span>
            <span class="n">VariantIndex</span><span class="o">.</span><span class="n">hash_long_variant_ids</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">),</span>
                <span class="n">hash_threshold</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="c1"># Generating a temporary column with only protein coding transcripts:</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;proteinCodingTranscripts&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">),</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="s2">&quot;biotype&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;protein_coding&quot;</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="c1"># Generate variant descrioption:</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;variantDescription&quot;</span><span class="p">,</span>
            <span class="bp">cls</span><span class="o">.</span><span class="n">_compose_variant_description</span><span class="p">(</span>
                <span class="c1"># Passing the most severe consequence:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;most_severe_consequence&quot;</span><span class="p">),</span>
                <span class="c1"># The first transcript:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;transcriptConsequences&quot;</span><span class="p">),</span>
                    <span class="k">lambda</span> <span class="n">vep</span><span class="p">:</span> <span class="n">vep</span><span class="o">.</span><span class="n">transcriptIndex</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span>
                <span class="p">)</span><span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
                <span class="c1"># The first protein coding transcript:</span>
                <span class="n">order_array_of_structs_by_field</span><span class="p">(</span>
                    <span class="s2">&quot;proteinCodingTranscripts&quot;</span><span class="p">,</span> <span class="s2">&quot;transcriptIndex&quot;</span>
                <span class="p">)[</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="s2">&quot;proteinCodingTranscripts&quot;</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">],</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="c1"># Normalising variant effect assessments:</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;variantEffect&quot;</span><span class="p">,</span>
            <span class="n">VariantEffectNormaliser</span><span class="o">.</span><span class="n">normalise_variant_effect</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantEffect&quot;</span><span class="p">)</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="c1"># Dropping intermediate xref columns:</span>
        <span class="o">.</span><span class="n">drop</span><span class="p">(</span>
            <span class="o">*</span><span class="p">[</span>
                <span class="s2">&quot;ensembl_xrefs&quot;</span><span class="p">,</span>
                <span class="s2">&quot;omim_xrefs&quot;</span><span class="p">,</span>
                <span class="s2">&quot;clinvar_xrefs&quot;</span><span class="p">,</span>
                <span class="s2">&quot;protvar_xrefs&quot;</span><span class="p">,</span>
                <span class="s2">&quot;most_severe_consequence&quot;</span><span class="p">,</span>
                <span class="s2">&quot;proteinCodingTranscripts&quot;</span><span class="p">,</span>
            <span class="p">]</span>
        <span class="p">)</span>
        <span class="c1"># Drooping rows with null position:</span>
        <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">())</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>
<hr />
<div class="md-source-file">
  <small>
    <ul
      class="metadata page-metadata"
      data-bi-name="page info"
      lang="en-us"
      dir="ltr"
    >
      <!-- mkdocs-git-revision-date-plugin -->
      
      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3h-2Z"/></svg>
        </span>
        <span class="label"> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class="timeago" datetime="2024-06-30T20:13:04+00:00" locale="en"></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2024-06-30</span> </span>
      </li>
       
      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg>
        </span>
        <span class="label"> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class="timeago" datetime="2024-06-30T20:13:04+00:00" locale="en"></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2024-06-30</span> </span>
      </li>
      

      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2A10 10 0 0 0 2 12c0 4.42 2.87 8.17 6.84 9.5.5.08.66-.23.66-.5v-1.69c-2.77.6-3.36-1.34-3.36-1.34-.46-1.16-1.11-1.47-1.11-1.47-.91-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.87 1.52 2.34 1.07 2.91.83.09-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.92 0-1.11.38-2 1.03-2.71-.1-.25-.45-1.29.1-2.64 0 0 .84-.27 2.75 1.02.79-.22 1.65-.33 2.5-.33.85 0 1.71.11 2.5.33 1.91-1.29 2.75-1.02 2.75-1.02.55 1.35.2 2.39.1 2.64.65.71 1.03 1.6 1.03 2.71 0 3.82-2.34 4.66-4.57 4.91.36.31.69.92.69 1.85V21c0 .27.16.59.67.5C19.14 20.16 22 16.42 22 12A10 10 0 0 0 12 2Z"/></svg>
        </span>
        <span class="label"> Contributors </span>
      </li>
      <li class="contributors-holder">
        <ul class="contributors" data-bi-name="contributors"><li>
            <a
              href="https://github.com/DSuveges"
              title="DSuveges"
              data-bi-name="contributorprofile"
              target="_blank"
              ><img src="https://avatars.githubusercontent.com/u/7122521?v=4" alt="DSuveges"
            /></a>
          </li><li>
            <a
              href="https://github.com/web-flow"
              title="web-flow"
              data-bi-name="contributorprofile"
              target="_blank"
              ><img src="https://avatars.githubusercontent.com/u/19864447?v=4" alt="web-flow"
            /></a>
          </li></ul>
      </li>
    </ul>
  </small>
</div>




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/gentropy" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../../..", "features": ["navigation.instant", "navigation.top", "navigation.tabs", "navigation.tabs.sticky", "content.code.annotate", "search.suggest", "search.highlight"], "search": "../../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../../assets/javascripts/bundle.3220b9d7.min.js"></script>
      
        <script src="../../../../js/timeago.min.js"></script>
      
        <script src="../../../../js/timeago_mkdocs_material.js"></script>
      
        <script src="../../../../assets/javascripts/extra.js"></script>
      
    
  </body>
</html>