<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><link rel=icon href=../../assets/favicon.png><meta name=generator content="mkdocs-1.4.0, mkdocs-material-8.4.4"><title>Variant to gene - Genetics Portal Pipeline</title><link rel=stylesheet href=../../assets/stylesheets/main.6b80c2a2.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.cbb835fc.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../assets/_mkdocstrings.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=blue-grey data-md-color-accent=light-blue> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#summary-of-the-logic class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="Genetics Portal Pipeline" class="md-header__button md-logo" aria-label="Genetics Portal Pipeline" data-md-component=logo> <img src=../../assets/otlogo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Genetics Portal Pipeline </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Variant to gene </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=blue-grey data-md-color-accent=light-blue aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=blue-grey data-md-color-accent=light-blue aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/opentargets/genetics_etl_python title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> opentargets/genetics_etl_python </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="Genetics Portal Pipeline" class="md-nav__button md-logo" aria-label="Genetics Portal Pipeline" data-md-component=logo> <img src=../../assets/otlogo.png alt=logo> </a> Genetics Portal Pipeline </label> <div class=md-nav__source> <a href=https://github.com/opentargets/genetics_etl_python title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> opentargets/genetics_etl_python </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> Home </a> </li> <li class=md-nav__item> <a href=../../roadmap/ class=md-nav__link> Roadmap </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_3 type=checkbox id=__nav_3 checked> <div class="md-nav__link md-nav__link--index "> <a href=../colocalisation/ >Modules</a> <label for=__nav_3> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav aria-label=Modules data-md-level=1> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Modules </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../gwas_ingest/ class=md-nav__link> Gwas ingest </a> </li> <li class=md-nav__item> <a href=../schemas/ class=md-nav__link> Schemas </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Variant to gene <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> Variant to gene </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#summary-of-the-logic class=md-nav__link> Summary of the logic </a> <nav class=md-nav aria-label="Summary of the logic"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#chromatin-interaction-experiments class=md-nav__link> Chromatin interaction experiments </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals class=md-nav__link> etl.v2g.intervals </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.jung2019 class=md-nav__link> etl.v2g.intervals.jung2019 </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.jung2019.ParseJung class=md-nav__link> ParseJung </a> <nav class=md-nav aria-label=ParseJung> <ul class=md-nav__list> <li class=md-nav__item> <a href=#etl.v2g.intervals.jung2019.ParseJung.__init__ class=md-nav__link> __init__() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.jung2019.ParseJung.get_intervals class=md-nav__link> get_intervals() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.jung2019.ParseJung.qc_intervals class=md-nav__link> qc_intervals() </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.javierre2016 class=md-nav__link> etl.v2g.intervals.javierre2016 </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.javierre2016.ParseJavierre class=md-nav__link> ParseJavierre </a> <nav class=md-nav aria-label=ParseJavierre> <ul class=md-nav__list> <li class=md-nav__item> <a href=#etl.v2g.intervals.javierre2016.ParseJavierre.__init__ class=md-nav__link> __init__() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.javierre2016.ParseJavierre.get_intervals class=md-nav__link> get_intervals() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.javierre2016.ParseJavierre.qc_intervals class=md-nav__link> qc_intervals() </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.andersson2014 class=md-nav__link> etl.v2g.intervals.andersson2014 </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.andersson2014.ParseAndersson class=md-nav__link> ParseAndersson </a> <nav class=md-nav aria-label=ParseAndersson> <ul class=md-nav__list> <li class=md-nav__item> <a href=#etl.v2g.intervals.andersson2014.ParseAndersson.__init__ class=md-nav__link> __init__() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.andersson2014.ParseAndersson.get_intervals class=md-nav__link> get_intervals() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.andersson2014.ParseAndersson.qc_intervals class=md-nav__link> qc_intervals() </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.thurman2012 class=md-nav__link> etl.v2g.intervals.thurman2012 </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.thurman2012.ParseThurman class=md-nav__link> ParseThurman </a> <nav class=md-nav aria-label=ParseThurman> <ul class=md-nav__list> <li class=md-nav__item> <a href=#etl.v2g.intervals.thurman2012.ParseThurman.__init__ class=md-nav__link> __init__() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.thurman2012.ParseThurman.get_intervals class=md-nav__link> get_intervals() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.thurman2012.ParseThurman.qc_intervals class=md-nav__link> qc_intervals() </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.Liftover class=md-nav__link> etl.v2g.intervals.Liftover </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.Liftover.LiftOverSpark class=md-nav__link> LiftOverSpark </a> <nav class=md-nav aria-label=LiftOverSpark> <ul class=md-nav__list> <li class=md-nav__item> <a href=#etl.v2g.intervals.Liftover.LiftOverSpark.__init__ class=md-nav__link> __init__() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.Liftover.LiftOverSpark.convert_coordinates class=md-nav__link> convert_coordinates() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.Liftover.LiftOverSpark.convert_intervals class=md-nav__link> convert_intervals() </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.helpers class=md-nav__link> etl.v2g.intervals.helpers </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.helpers.get_variants_in_interval class=md-nav__link> get_variants_in_interval() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.helpers.prepare_gene_interval_lut class=md-nav__link> prepare_gene_interval_lut() </a> <nav class=md-nav aria-label=prepare_gene_interval_lut()> <ul class=md-nav__list> <li class=md-nav__item> <a href=#functional-predictions class=md-nav__link> Functional predictions </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.v2g.functional_predictions.vep class=md-nav__link> etl.v2g.functional_predictions.vep </a> </li> <li class=md-nav__item> <a href=#etl.v2g.functional_predictions.vep.get_plof_flag class=md-nav__link> get_plof_flag() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.functional_predictions.vep.get_polyphen_score class=md-nav__link> get_polyphen_score() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.functional_predictions.vep.get_sift_score class=md-nav__link> get_sift_score() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.functional_predictions.vep.get_variant_consequences class=md-nav__link> get_variant_consequences() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.functional_predictions.vep.main class=md-nav__link> main() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.functional_predictions.vep.read_consequence_lut class=md-nav__link> read_consequence_lut() </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../variants/ class=md-nav__link> Variants </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#summary-of-the-logic class=md-nav__link> Summary of the logic </a> <nav class=md-nav aria-label="Summary of the logic"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#chromatin-interaction-experiments class=md-nav__link> Chromatin interaction experiments </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals class=md-nav__link> etl.v2g.intervals </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.jung2019 class=md-nav__link> etl.v2g.intervals.jung2019 </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.jung2019.ParseJung class=md-nav__link> ParseJung </a> <nav class=md-nav aria-label=ParseJung> <ul class=md-nav__list> <li class=md-nav__item> <a href=#etl.v2g.intervals.jung2019.ParseJung.__init__ class=md-nav__link> __init__() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.jung2019.ParseJung.get_intervals class=md-nav__link> get_intervals() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.jung2019.ParseJung.qc_intervals class=md-nav__link> qc_intervals() </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.javierre2016 class=md-nav__link> etl.v2g.intervals.javierre2016 </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.javierre2016.ParseJavierre class=md-nav__link> ParseJavierre </a> <nav class=md-nav aria-label=ParseJavierre> <ul class=md-nav__list> <li class=md-nav__item> <a href=#etl.v2g.intervals.javierre2016.ParseJavierre.__init__ class=md-nav__link> __init__() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.javierre2016.ParseJavierre.get_intervals class=md-nav__link> get_intervals() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.javierre2016.ParseJavierre.qc_intervals class=md-nav__link> qc_intervals() </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.andersson2014 class=md-nav__link> etl.v2g.intervals.andersson2014 </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.andersson2014.ParseAndersson class=md-nav__link> ParseAndersson </a> <nav class=md-nav aria-label=ParseAndersson> <ul class=md-nav__list> <li class=md-nav__item> <a href=#etl.v2g.intervals.andersson2014.ParseAndersson.__init__ class=md-nav__link> __init__() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.andersson2014.ParseAndersson.get_intervals class=md-nav__link> get_intervals() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.andersson2014.ParseAndersson.qc_intervals class=md-nav__link> qc_intervals() </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.thurman2012 class=md-nav__link> etl.v2g.intervals.thurman2012 </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.thurman2012.ParseThurman class=md-nav__link> ParseThurman </a> <nav class=md-nav aria-label=ParseThurman> <ul class=md-nav__list> <li class=md-nav__item> <a href=#etl.v2g.intervals.thurman2012.ParseThurman.__init__ class=md-nav__link> __init__() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.thurman2012.ParseThurman.get_intervals class=md-nav__link> get_intervals() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.thurman2012.ParseThurman.qc_intervals class=md-nav__link> qc_intervals() </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.Liftover class=md-nav__link> etl.v2g.intervals.Liftover </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.Liftover.LiftOverSpark class=md-nav__link> LiftOverSpark </a> <nav class=md-nav aria-label=LiftOverSpark> <ul class=md-nav__list> <li class=md-nav__item> <a href=#etl.v2g.intervals.Liftover.LiftOverSpark.__init__ class=md-nav__link> __init__() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.Liftover.LiftOverSpark.convert_coordinates class=md-nav__link> convert_coordinates() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.Liftover.LiftOverSpark.convert_intervals class=md-nav__link> convert_intervals() </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.helpers class=md-nav__link> etl.v2g.intervals.helpers </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.helpers.get_variants_in_interval class=md-nav__link> get_variants_in_interval() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.intervals.helpers.prepare_gene_interval_lut class=md-nav__link> prepare_gene_interval_lut() </a> <nav class=md-nav aria-label=prepare_gene_interval_lut()> <ul class=md-nav__list> <li class=md-nav__item> <a href=#functional-predictions class=md-nav__link> Functional predictions </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.v2g.functional_predictions.vep class=md-nav__link> etl.v2g.functional_predictions.vep </a> </li> <li class=md-nav__item> <a href=#etl.v2g.functional_predictions.vep.get_plof_flag class=md-nav__link> get_plof_flag() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.functional_predictions.vep.get_polyphen_score class=md-nav__link> get_polyphen_score() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.functional_predictions.vep.get_sift_score class=md-nav__link> get_sift_score() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.functional_predictions.vep.get_variant_consequences class=md-nav__link> get_variant_consequences() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.functional_predictions.vep.main class=md-nav__link> main() </a> </li> <li class=md-nav__item> <a href=#etl.v2g.functional_predictions.vep.read_consequence_lut class=md-nav__link> read_consequence_lut() </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/opentargets/genetics_etl_python/edit/master/docs/modules/variant_to_gene.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1>Variant to gene</h1> <p>All variants in the variant index are annotated using our Variant-to-Gene (V2G) pipeline. The pipeline integrates V2G evidence that fall into four main data types:</p> <ol> <li>Chromatin interaction experiments, e.g. Promoter Capture Hi-C (PCHi-C).</li> <li>In silico functional predictions, e.g. Variant Effect Predictor (VEP) from Ensembl.</li> <li>Distance between the variant and each gene's canonical transcription start site (TSS).</li> </ol> <p>Within each data type there are multiple sources of information produced by different experimental methods. Some of these sources can further be broken down into separate tissues or cell types (features).</p> <h2 id=summary-of-the-logic>Summary of the logic</h2> <ol> <li>Process each data type separately.</li> <li>Filter out V2G evidence that links to genes that are not of interest (mainly of non protein coding type).</li> <li>Group V2G evidence by variant and gene to compute an aggregated score.</li> </ol> <h3 id=chromatin-interaction-experiments>Chromatin interaction experiments</h3> <div class="doc doc-object doc-module"> <a id=etl.v2g.intervals></a> <div class="doc doc-contents first"> <p>Interval data parsers.</p> <p>This workflow produce intervals dataset that links genes to genomic regions based on genome interaction studies.</p> <div class="doc doc-children"> </div> </div> </div> <div class="doc doc-object doc-module"> <a id=etl.v2g.intervals.jung2019></a> <div class="doc doc-contents first"> <p>PCHI-C (Jung, 2019) intervals.</p> <p>Promoter capture Hi-C was used to map long-range chromatin interactions for 18,943 well-annotated promoters for protein-coding genes in 27 human tissue types. (<a href=https://www.nature.com/articles/s41588-019-0494-8>Link</a> to the publication)</p> <p>This dataset provides tissue level annotation, but no cell type or biofeature is given. All interactions are significant so scores are set to 1.</p> <div class="doc doc-children"> <div class="doc doc-object doc-class"> <h2 id=etl.v2g.intervals.jung2019.ParseJung class="doc doc-heading"> <code>ParseJung</code> </h2> <div class="doc doc-contents "> <p>Parser Jung 2019 dataset.</p> <p><strong>Summary of the logic:</strong></p> <ul> <li>Reading dataset into a PySpark dataframe.</li> <li>Select relevant columns, parsing: start, end.</li> <li>Lifting over the intervals.</li> <li>Split gene names (separated by ;)</li> <li>Look up gene names to gene identifiers.</li> </ul> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/jung2019.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 22</span>
<span class=normal> 23</span>
<span class=normal> 24</span>
<span class=normal> 25</span>
<span class=normal> 26</span>
<span class=normal> 27</span>
<span class=normal> 28</span>
<span class=normal> 29</span>
<span class=normal> 30</span>
<span class=normal> 31</span>
<span class=normal> 32</span>
<span class=normal> 33</span>
<span class=normal> 34</span>
<span class=normal> 35</span>
<span class=normal> 36</span>
<span class=normal> 37</span>
<span class=normal> 38</span>
<span class=normal> 39</span>
<span class=normal> 40</span>
<span class=normal> 41</span>
<span class=normal> 42</span>
<span class=normal> 43</span>
<span class=normal> 44</span>
<span class=normal> 45</span>
<span class=normal> 46</span>
<span class=normal> 47</span>
<span class=normal> 48</span>
<span class=normal> 49</span>
<span class=normal> 50</span>
<span class=normal> 51</span>
<span class=normal> 52</span>
<span class=normal> 53</span>
<span class=normal> 54</span>
<span class=normal> 55</span>
<span class=normal> 56</span>
<span class=normal> 57</span>
<span class=normal> 58</span>
<span class=normal> 59</span>
<span class=normal> 60</span>
<span class=normal> 61</span>
<span class=normal> 62</span>
<span class=normal> 63</span>
<span class=normal> 64</span>
<span class=normal> 65</span>
<span class=normal> 66</span>
<span class=normal> 67</span>
<span class=normal> 68</span>
<span class=normal> 69</span>
<span class=normal> 70</span>
<span class=normal> 71</span>
<span class=normal> 72</span>
<span class=normal> 73</span>
<span class=normal> 74</span>
<span class=normal> 75</span>
<span class=normal> 76</span>
<span class=normal> 77</span>
<span class=normal> 78</span>
<span class=normal> 79</span>
<span class=normal> 80</span>
<span class=normal> 81</span>
<span class=normal> 82</span>
<span class=normal> 83</span>
<span class=normal> 84</span>
<span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span>
<span class=normal>125</span>
<span class=normal>126</span>
<span class=normal>127</span>
<span class=normal>128</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>class</span> <span class=nc>ParseJung</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Parser Jung 2019 dataset.</span>

<span class=sd>    **Summary of the logic:**</span>

<span class=sd>    - Reading dataset into a PySpark dataframe.</span>
<span class=sd>    - Select relevant columns, parsing: start, end.</span>
<span class=sd>    - Lifting over the intervals.</span>
<span class=sd>    - Split gene names (separated by ;)</span>
<span class=sd>    - Look up gene names to gene identifiers.</span>
<span class=sd>    &quot;&quot;&quot;</span>

    <span class=c1># Constants:</span>
    <span class=n>DATASET_NAME</span> <span class=o>=</span> <span class=s2>&quot;jung2019&quot;</span>
    <span class=n>DATA_TYPE</span> <span class=o>=</span> <span class=s2>&quot;interval&quot;</span>
    <span class=n>EXPERIMENT_TYPE</span> <span class=o>=</span> <span class=s2>&quot;pchic&quot;</span>
    <span class=n>PMID</span> <span class=o>=</span> <span class=s2>&quot;31501517&quot;</span>

    <span class=k>def</span> <span class=fm>__init__</span><span class=p>(</span>
        <span class=bp>self</span><span class=p>:</span> <span class=n>ParseJung</span><span class=p>,</span>
        <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span>
        <span class=n>jung_data</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
        <span class=n>gene_index</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
        <span class=n>lift</span><span class=p>:</span> <span class=n>LiftOverSpark</span><span class=p>,</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Initialise Jung parser.</span>

<span class=sd>        Args:</span>
<span class=sd>            etl (ETLSession): current ETL session</span>
<span class=sd>            jung_data (str): path to the csv file containing the Jung 2019 data</span>
<span class=sd>            gene_index (DataFrame): DataFrame containing the gene index</span>
<span class=sd>            lift (LiftOverSpark): LiftOverSpark object</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Parsing Jung 2019 data...&quot;</span><span class=p>)</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Reading data from </span><span class=si>{</span><span class=n>jung_data</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>

        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span> <span class=o>=</span> <span class=n>etl</span>

        <span class=c1># Read Jung data:</span>
        <span class=n>jung_raw</span> <span class=o>=</span> <span class=p>(</span>
            <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>csv</span><span class=p>(</span><span class=n>jung_data</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;,&quot;</span><span class=p>,</span> <span class=n>header</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;interval&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;Interacting_fragment&quot;</span><span class=p>),</span> <span class=sa>r</span><span class=s2>&quot;\.&quot;</span><span class=p>))</span>
            <span class=o>.</span><span class=n>select</span><span class=p>(</span>
                <span class=c1># Parsing intervals:</span>
                <span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;interval&quot;</span><span class=p>)[</span><span class=mi>0</span><span class=p>],</span> <span class=s2>&quot;chr&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;interval&quot;</span><span class=p>)[</span><span class=mi>1</span><span class=p>]</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>())</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;interval&quot;</span><span class=p>)[</span><span class=mi>2</span><span class=p>]</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>())</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>),</span>
                <span class=c1># Extract other columns:</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;Promoter&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;gene_name&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;Tissue_type&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;tissue&quot;</span><span class=p>),</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
        <span class=p>)</span>

        <span class=c1># Lifting over the coordinates:</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>jung_intervals</span> <span class=o>=</span> <span class=p>(</span>
            <span class=n>jung_raw</span>
            <span class=c1># Lifting over to GRCh38 interval 1:</span>
            <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>lift</span><span class=o>.</span><span class=n>convert_intervals</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>))</span>
            <span class=o>.</span><span class=n>select</span><span class=p>(</span>
                <span class=s2>&quot;chrom&quot;</span><span class=p>,</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped_start&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped_end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>explode</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;gene_name&quot;</span><span class=p>),</span> <span class=s2>&quot;;&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;gene_name&quot;</span><span class=p>),</span>
                <span class=s2>&quot;tissue&quot;</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;intervals&quot;</span><span class=p>)</span>
            <span class=c1># Joining with genes:</span>
            <span class=o>.</span><span class=n>join</span><span class=p>(</span>
                <span class=n>gene_index</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;genes&quot;</span><span class=p>),</span>
                <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;intervals.gene_name&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genes.geneSymbol&quot;</span><span class=p>)],</span>
                <span class=n>how</span><span class=o>=</span><span class=s2>&quot;inner&quot;</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=c1># Finalize dataset:</span>
            <span class=o>.</span><span class=n>select</span><span class=p>(</span>
                <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
                <span class=s2>&quot;start&quot;</span><span class=p>,</span>
                <span class=s2>&quot;end&quot;</span><span class=p>,</span>
                <span class=s2>&quot;geneId&quot;</span><span class=p>,</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;tissue&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;biofeature&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mf>1.0</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>DATASET_NAME</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datasourceId&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>EXPERIMENT_TYPE</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datatypeId&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>PMID</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;pmid&quot;</span><span class=p>),</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>drop_duplicates</span><span class=p>()</span>
            <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
        <span class=p>)</span>

    <span class=k>def</span> <span class=nf>get_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseJung</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Get preformatted intervals from Jung.</span>

<span class=sd>        Returns:</span>
<span class=sd>            DataFrame: Jung intervals</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=bp>self</span><span class=o>.</span><span class=n>jung_intervals</span>

    <span class=k>def</span> <span class=nf>qc_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseJung</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Perform QC on the Jung intervals.&quot;&quot;&quot;</span>
        <span class=c1># Get numbers:</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Size of Jung data: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>jung_intervals</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
            <span class=sa>f</span><span class=s1>&#39;Number of unique intervals: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>jung_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
        <span class=p>)</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
            <span class=sa>f</span><span class=s1>&#39;Number genes in the Jung dataset: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>jung_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
        <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.jung2019.ParseJung.__init__ class="doc doc-heading"> <code class="highlight language-python"><span class=fm>__init__</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>jung_data</span><span class=p>,</span> <span class=n>gene_index</span><span class=p>,</span> <span class=n>lift</span><span class=p>)</span></code> </h3> <div class="doc doc-contents "> <p>Initialise Jung parser.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>current ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>jung_data</code></td> <td> <code>str</code> </td> <td><p>path to the csv file containing the Jung 2019 data</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>gene_index</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>DataFrame containing the gene index</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>lift</code></td> <td> <code><a class="autorefs autorefs-internal" title="etl.v2g.intervals.Liftover.LiftOverSpark" href="#etl.v2g.intervals.Liftover.LiftOverSpark">LiftOverSpark</a></code> </td> <td><p>LiftOverSpark object</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/jung2019.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 40</span>
<span class=normal> 41</span>
<span class=normal> 42</span>
<span class=normal> 43</span>
<span class=normal> 44</span>
<span class=normal> 45</span>
<span class=normal> 46</span>
<span class=normal> 47</span>
<span class=normal> 48</span>
<span class=normal> 49</span>
<span class=normal> 50</span>
<span class=normal> 51</span>
<span class=normal> 52</span>
<span class=normal> 53</span>
<span class=normal> 54</span>
<span class=normal> 55</span>
<span class=normal> 56</span>
<span class=normal> 57</span>
<span class=normal> 58</span>
<span class=normal> 59</span>
<span class=normal> 60</span>
<span class=normal> 61</span>
<span class=normal> 62</span>
<span class=normal> 63</span>
<span class=normal> 64</span>
<span class=normal> 65</span>
<span class=normal> 66</span>
<span class=normal> 67</span>
<span class=normal> 68</span>
<span class=normal> 69</span>
<span class=normal> 70</span>
<span class=normal> 71</span>
<span class=normal> 72</span>
<span class=normal> 73</span>
<span class=normal> 74</span>
<span class=normal> 75</span>
<span class=normal> 76</span>
<span class=normal> 77</span>
<span class=normal> 78</span>
<span class=normal> 79</span>
<span class=normal> 80</span>
<span class=normal> 81</span>
<span class=normal> 82</span>
<span class=normal> 83</span>
<span class=normal> 84</span>
<span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=fm>__init__</span><span class=p>(</span>
    <span class=bp>self</span><span class=p>:</span> <span class=n>ParseJung</span><span class=p>,</span>
    <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span>
    <span class=n>jung_data</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
    <span class=n>gene_index</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
    <span class=n>lift</span><span class=p>:</span> <span class=n>LiftOverSpark</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Initialise Jung parser.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): current ETL session</span>
<span class=sd>        jung_data (str): path to the csv file containing the Jung 2019 data</span>
<span class=sd>        gene_index (DataFrame): DataFrame containing the gene index</span>
<span class=sd>        lift (LiftOverSpark): LiftOverSpark object</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Parsing Jung 2019 data...&quot;</span><span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Reading data from </span><span class=si>{</span><span class=n>jung_data</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>

    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span> <span class=o>=</span> <span class=n>etl</span>

    <span class=c1># Read Jung data:</span>
    <span class=n>jung_raw</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>csv</span><span class=p>(</span><span class=n>jung_data</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;,&quot;</span><span class=p>,</span> <span class=n>header</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;interval&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;Interacting_fragment&quot;</span><span class=p>),</span> <span class=sa>r</span><span class=s2>&quot;\.&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=c1># Parsing intervals:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;interval&quot;</span><span class=p>)[</span><span class=mi>0</span><span class=p>],</span> <span class=s2>&quot;chr&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;interval&quot;</span><span class=p>)[</span><span class=mi>1</span><span class=p>]</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>())</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;interval&quot;</span><span class=p>)[</span><span class=mi>2</span><span class=p>]</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>())</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>),</span>
            <span class=c1># Extract other columns:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;Promoter&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;gene_name&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;Tissue_type&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;tissue&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>

    <span class=c1># Lifting over the coordinates:</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>jung_intervals</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>jung_raw</span>
        <span class=c1># Lifting over to GRCh38 interval 1:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>lift</span><span class=o>.</span><span class=n>convert_intervals</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;chrom&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped_start&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped_end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>explode</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;gene_name&quot;</span><span class=p>),</span> <span class=s2>&quot;;&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;gene_name&quot;</span><span class=p>),</span>
            <span class=s2>&quot;tissue&quot;</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;intervals&quot;</span><span class=p>)</span>
        <span class=c1># Joining with genes:</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>gene_index</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;genes&quot;</span><span class=p>),</span>
            <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;intervals.gene_name&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genes.geneSymbol&quot;</span><span class=p>)],</span>
            <span class=n>how</span><span class=o>=</span><span class=s2>&quot;inner&quot;</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=c1># Finalize dataset:</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
            <span class=s2>&quot;start&quot;</span><span class=p>,</span>
            <span class=s2>&quot;end&quot;</span><span class=p>,</span>
            <span class=s2>&quot;geneId&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;tissue&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;biofeature&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mf>1.0</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>DATASET_NAME</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datasourceId&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>EXPERIMENT_TYPE</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datatypeId&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>PMID</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;pmid&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>drop_duplicates</span><span class=p>()</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.jung2019.ParseJung.get_intervals class="doc doc-heading"> <code class="highlight language-python"><span class=n>get_intervals</span><span class=p>()</span></code> </h3> <div class="doc doc-contents "> <p>Get preformatted intervals from Jung.</p> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Jung intervals</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/jung2019.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>get_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseJung</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Get preformatted intervals from Jung.</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Jung intervals</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=bp>self</span><span class=o>.</span><span class=n>jung_intervals</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.jung2019.ParseJung.qc_intervals class="doc doc-heading"> <code class="highlight language-python"><span class=n>qc_intervals</span><span class=p>()</span></code> </h3> <div class="doc doc-contents "> <p>Perform QC on the Jung intervals.</p> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/jung2019.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span>
<span class=normal>125</span>
<span class=normal>126</span>
<span class=normal>127</span>
<span class=normal>128</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>qc_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseJung</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Perform QC on the Jung intervals.&quot;&quot;&quot;</span>
    <span class=c1># Get numbers:</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Size of Jung data: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>jung_intervals</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number of unique intervals: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>jung_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number genes in the Jung dataset: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>jung_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> </div> </div> </div> <div class="doc doc-object doc-module"> <a id=etl.v2g.intervals.javierre2016></a> <div class="doc doc-contents first"> <p>PCHi-C intervals (Javierre et al. 2016).</p> <p>Javierre and colleagues uses promoter capture Hi-C to identify interacting regions of 31,253 promoters in 17 human primary hematopoietic cell types. (<a href=https://www.sciencedirect.com/science/article/pii/S0092867416313228>Link</a> to the publication)</p> <p>The dataset provides cell type resolution, however these cell types are not resolved to tissues. Scores are also preserved.</p> <div class="doc doc-children"> <div class="doc doc-object doc-class"> <h2 id=etl.v2g.intervals.javierre2016.ParseJavierre class="doc doc-heading"> <code>ParseJavierre</code> </h2> <div class="doc doc-contents "> <p>Javierre 2016 dataset parsers.</p> <p><strong>Summary of the logic:</strong></p> <ul> <li>Reading parquet file containing the pre-processed Javierre dataset.</li> <li>Splitting name column into chromosome, start, end, and score.</li> <li>Lifting over the intervals.</li> <li>Mapping intervals to genes by overlapping regions.</li> <li>For each gene/interval pair, keep only the highest scoring interval.</li> <li>Filter gene/interval pairs by the distance between the TSS and the start of the interval.</li> </ul> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/javierre2016.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 22</span>
<span class=normal> 23</span>
<span class=normal> 24</span>
<span class=normal> 25</span>
<span class=normal> 26</span>
<span class=normal> 27</span>
<span class=normal> 28</span>
<span class=normal> 29</span>
<span class=normal> 30</span>
<span class=normal> 31</span>
<span class=normal> 32</span>
<span class=normal> 33</span>
<span class=normal> 34</span>
<span class=normal> 35</span>
<span class=normal> 36</span>
<span class=normal> 37</span>
<span class=normal> 38</span>
<span class=normal> 39</span>
<span class=normal> 40</span>
<span class=normal> 41</span>
<span class=normal> 42</span>
<span class=normal> 43</span>
<span class=normal> 44</span>
<span class=normal> 45</span>
<span class=normal> 46</span>
<span class=normal> 47</span>
<span class=normal> 48</span>
<span class=normal> 49</span>
<span class=normal> 50</span>
<span class=normal> 51</span>
<span class=normal> 52</span>
<span class=normal> 53</span>
<span class=normal> 54</span>
<span class=normal> 55</span>
<span class=normal> 56</span>
<span class=normal> 57</span>
<span class=normal> 58</span>
<span class=normal> 59</span>
<span class=normal> 60</span>
<span class=normal> 61</span>
<span class=normal> 62</span>
<span class=normal> 63</span>
<span class=normal> 64</span>
<span class=normal> 65</span>
<span class=normal> 66</span>
<span class=normal> 67</span>
<span class=normal> 68</span>
<span class=normal> 69</span>
<span class=normal> 70</span>
<span class=normal> 71</span>
<span class=normal> 72</span>
<span class=normal> 73</span>
<span class=normal> 74</span>
<span class=normal> 75</span>
<span class=normal> 76</span>
<span class=normal> 77</span>
<span class=normal> 78</span>
<span class=normal> 79</span>
<span class=normal> 80</span>
<span class=normal> 81</span>
<span class=normal> 82</span>
<span class=normal> 83</span>
<span class=normal> 84</span>
<span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span>
<span class=normal>125</span>
<span class=normal>126</span>
<span class=normal>127</span>
<span class=normal>128</span>
<span class=normal>129</span>
<span class=normal>130</span>
<span class=normal>131</span>
<span class=normal>132</span>
<span class=normal>133</span>
<span class=normal>134</span>
<span class=normal>135</span>
<span class=normal>136</span>
<span class=normal>137</span>
<span class=normal>138</span>
<span class=normal>139</span>
<span class=normal>140</span>
<span class=normal>141</span>
<span class=normal>142</span>
<span class=normal>143</span>
<span class=normal>144</span>
<span class=normal>145</span>
<span class=normal>146</span>
<span class=normal>147</span>
<span class=normal>148</span>
<span class=normal>149</span>
<span class=normal>150</span>
<span class=normal>151</span>
<span class=normal>152</span>
<span class=normal>153</span>
<span class=normal>154</span>
<span class=normal>155</span>
<span class=normal>156</span>
<span class=normal>157</span>
<span class=normal>158</span>
<span class=normal>159</span>
<span class=normal>160</span>
<span class=normal>161</span>
<span class=normal>162</span>
<span class=normal>163</span>
<span class=normal>164</span>
<span class=normal>165</span>
<span class=normal>166</span>
<span class=normal>167</span>
<span class=normal>168</span>
<span class=normal>169</span>
<span class=normal>170</span>
<span class=normal>171</span>
<span class=normal>172</span>
<span class=normal>173</span>
<span class=normal>174</span>
<span class=normal>175</span>
<span class=normal>176</span>
<span class=normal>177</span>
<span class=normal>178</span>
<span class=normal>179</span>
<span class=normal>180</span>
<span class=normal>181</span>
<span class=normal>182</span>
<span class=normal>183</span>
<span class=normal>184</span>
<span class=normal>185</span>
<span class=normal>186</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>class</span> <span class=nc>ParseJavierre</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Javierre 2016 dataset parsers.</span>

<span class=sd>    **Summary of the logic:**</span>

<span class=sd>    - Reading parquet file containing the pre-processed Javierre dataset.</span>
<span class=sd>    - Splitting name column into chromosome, start, end, and score.</span>
<span class=sd>    - Lifting over the intervals.</span>
<span class=sd>    - Mapping intervals to genes by overlapping regions.</span>
<span class=sd>    - For each gene/interval pair, keep only the highest scoring interval.</span>
<span class=sd>    - Filter gene/interval pairs by the distance between the TSS and the start of the interval.</span>
<span class=sd>    &quot;&quot;&quot;</span>

    <span class=c1># Constants:</span>
    <span class=n>DATASET_NAME</span> <span class=o>=</span> <span class=s2>&quot;javierre2016&quot;</span>
    <span class=n>DATA_TYPE</span> <span class=o>=</span> <span class=s2>&quot;interval&quot;</span>
    <span class=n>EXPERIMENT_TYPE</span> <span class=o>=</span> <span class=s2>&quot;pchic&quot;</span>
    <span class=n>PMID</span> <span class=o>=</span> <span class=s2>&quot;27863249&quot;</span>
    <span class=n>TWOSIDED_THRESHOLD</span> <span class=o>=</span> <span class=mf>2.45e6</span>  <span class=c1># TODO  this needs to phased out. Filter by percentile instead of absolute value.</span>

    <span class=k>def</span> <span class=fm>__init__</span><span class=p>(</span>
        <span class=bp>self</span><span class=p>:</span> <span class=n>ParseJavierre</span><span class=p>,</span>
        <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span>
        <span class=n>javierre_parquet</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
        <span class=n>gene_index</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
        <span class=n>lift</span><span class=p>:</span> <span class=n>LiftOverSpark</span><span class=p>,</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Initialise Javierre parser.</span>

<span class=sd>        Args:</span>
<span class=sd>            etl (ETLSession): ETL session</span>
<span class=sd>            javierre_parquet (str): path to the parquet file containing the Javierre 2016 data after processing it (see notebooks/Javierre_data_pre-process.ipynb)</span>
<span class=sd>            gene_index (DataFrame): Pyspark dataframe containing the gene index</span>
<span class=sd>            lift (LiftOverSpark): LiftOverSpark object</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span> <span class=o>=</span> <span class=n>etl</span>

        <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Parsing Javierre 2016 data...&quot;</span><span class=p>)</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Reading data from </span><span class=si>{</span><span class=n>javierre_parquet</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>

        <span class=c1># Read Javierre data:</span>
        <span class=n>javierre_raw</span> <span class=o>=</span> <span class=p>(</span>
            <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>parquet</span><span class=p>(</span><span class=n>javierre_parquet</span><span class=p>)</span>
            <span class=c1># Splitting name column into chromosome, start, end, and score:</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;name_split&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name&quot;</span><span class=p>),</span> <span class=sa>r</span><span class=s2>&quot;:|-|,&quot;</span><span class=p>))</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
                <span class=s2>&quot;name_chr&quot;</span><span class=p>,</span>
                <span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_split&quot;</span><span class=p>)[</span><span class=mi>0</span><span class=p>],</span> <span class=s2>&quot;chr&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span>
                    <span class=n>t</span><span class=o>.</span><span class=n>StringType</span><span class=p>()</span>
                <span class=p>),</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;name_start&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_split&quot;</span><span class=p>)[</span><span class=mi>1</span><span class=p>]</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>()))</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;name_end&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_split&quot;</span><span class=p>)[</span><span class=mi>2</span><span class=p>]</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>()))</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;name_score&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_split&quot;</span><span class=p>)[</span><span class=mi>3</span><span class=p>]</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>FloatType</span><span class=p>()))</span>
            <span class=c1># Cleaning up chromosome:</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
                <span class=s2>&quot;chrom&quot;</span><span class=p>,</span>
                <span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>),</span> <span class=s2>&quot;chr&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>StringType</span><span class=p>()),</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;name_split&quot;</span><span class=p>,</span> <span class=s2>&quot;name&quot;</span><span class=p>,</span> <span class=s2>&quot;annotation&quot;</span><span class=p>)</span>
            <span class=c1># Keep canonical chromosomes and consistent chromosomes with scores:</span>
            <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
                <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_score&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>())</span>
                <span class=o>&amp;</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_chr&quot;</span><span class=p>))</span>
                <span class=o>&amp;</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_chr&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isin</span><span class=p>(</span>
                    <span class=p>[</span><span class=sa>f</span><span class=s2>&quot;</span><span class=si>{</span><span class=n>x</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>x</span> <span class=ow>in</span> <span class=nb>range</span><span class=p>(</span><span class=mi>1</span><span class=p>,</span> <span class=mi>23</span><span class=p>)]</span> <span class=o>+</span> <span class=p>[</span><span class=s2>&quot;X&quot;</span><span class=p>,</span> <span class=s2>&quot;Y&quot;</span><span class=p>,</span> <span class=s2>&quot;MT&quot;</span><span class=p>]</span>
                <span class=p>)</span>
            <span class=p>)</span>
        <span class=p>)</span>

        <span class=c1># Lifting over intervals:</span>
        <span class=n>javierre_remapped</span> <span class=o>=</span> <span class=p>(</span>
            <span class=n>javierre_raw</span>
            <span class=c1># Lifting over to GRCh38 interval 1:</span>
            <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>lift</span><span class=o>.</span><span class=n>convert_intervals</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>))</span>
            <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;chrom&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_start&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_end&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span>
            <span class=c1># Lifting over interval 2 to GRCh38:</span>
            <span class=o>.</span><span class=n>transform</span><span class=p>(</span>
                <span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>lift</span><span class=o>.</span><span class=n>convert_intervals</span><span class=p>(</span>
                    <span class=n>df</span><span class=p>,</span> <span class=s2>&quot;name_chr&quot;</span><span class=p>,</span> <span class=s2>&quot;name_start&quot;</span><span class=p>,</span> <span class=s2>&quot;name_end&quot;</span>
                <span class=p>)</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;name_start&quot;</span><span class=p>,</span> <span class=s2>&quot;name_end&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_name_chr&quot;</span><span class=p>,</span> <span class=s2>&quot;name_chr&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_name_start&quot;</span><span class=p>,</span> <span class=s2>&quot;name_start&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_name_end&quot;</span><span class=p>,</span> <span class=s2>&quot;name_end&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
        <span class=p>)</span>

        <span class=c1># Once the intervals are lifted, extracting the unique intervals:</span>
        <span class=n>unique_intervals_with_genes</span> <span class=o>=</span> <span class=p>(</span>
            <span class=n>javierre_remapped</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;intervals&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>select</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>()),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>()),</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>distinct</span><span class=p>()</span>
            <span class=o>.</span><span class=n>join</span><span class=p>(</span>
                <span class=n>gene_index</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;genes&quot;</span><span class=p>),</span>
                <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;intervals.chrom&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genes.chromosome&quot;</span><span class=p>)],</span>
                <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
                <span class=p>(</span>
                    <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>)</span> <span class=o>&gt;=</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genomicLocation.start&quot;</span><span class=p>))</span>
                    <span class=o>&amp;</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>)</span> <span class=o>&lt;=</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genomicLocation.end&quot;</span><span class=p>))</span>
                <span class=p>)</span>
                <span class=o>|</span> <span class=p>(</span>
                    <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>)</span> <span class=o>&gt;=</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genomicLocation.start&quot;</span><span class=p>))</span>
                    <span class=o>&amp;</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>)</span> <span class=o>&lt;=</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genomicLocation.end&quot;</span><span class=p>))</span>
                <span class=p>)</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>,</span> <span class=s2>&quot;geneId&quot;</span><span class=p>,</span> <span class=s2>&quot;tss&quot;</span><span class=p>)</span>
        <span class=p>)</span>

        <span class=c1># Joining back the data:</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>javierre_intervals</span> <span class=o>=</span> <span class=p>(</span>
            <span class=n>javierre_remapped</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
                <span class=n>unique_intervals_with_genes</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>],</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
                <span class=c1># Drop rows where the TSS is far from the start of the region</span>
                <span class=n>f</span><span class=o>.</span><span class=n>abs</span><span class=p>((</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>)</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>))</span> <span class=o>/</span> <span class=mi>2</span> <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;tss&quot;</span><span class=p>))</span>
                <span class=o>&lt;=</span> <span class=bp>self</span><span class=o>.</span><span class=n>TWOSIDED_THRESHOLD</span>
            <span class=p>)</span>
            <span class=c1># For each gene, keep only the highest scoring interval:</span>
            <span class=o>.</span><span class=n>groupBy</span><span class=p>(</span>
                <span class=s2>&quot;name_chr&quot;</span><span class=p>,</span> <span class=s2>&quot;name_start&quot;</span><span class=p>,</span> <span class=s2>&quot;name_end&quot;</span><span class=p>,</span> <span class=s2>&quot;genes.geneId&quot;</span><span class=p>,</span> <span class=s2>&quot;bio_feature&quot;</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>agg</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>max</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_score&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;resourceScore&quot;</span><span class=p>))</span>
            <span class=c1># Create the output:</span>
            <span class=o>.</span><span class=n>select</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_chr&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_start&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;resourceScore&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genes.geneId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;bio_feature&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;biofeature&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>DATASET_NAME</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datasourceId&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>EXPERIMENT_TYPE</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datatypeId&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>PMID</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;pmid&quot;</span><span class=p>),</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
        <span class=p>)</span>

    <span class=k>def</span> <span class=nf>get_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseJavierre</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Get preformatted Javierre intervals.&quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=bp>self</span><span class=o>.</span><span class=n>javierre_intervals</span>

    <span class=k>def</span> <span class=nf>qc_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseJavierre</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Perform QC on the anderson intervals.&quot;&quot;&quot;</span>
        <span class=c1># Get numbers:</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
            <span class=sa>f</span><span class=s2>&quot;Size of Javierre data: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>javierre_intervals</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span>
        <span class=p>)</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
            <span class=sa>f</span><span class=s1>&#39;Number of unique intervals: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>javierre_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
        <span class=p>)</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
            <span class=sa>f</span><span class=s1>&#39;Number genes in the Javierre dataset: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>javierre_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
        <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.javierre2016.ParseJavierre.__init__ class="doc doc-heading"> <code class="highlight language-python"><span class=fm>__init__</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>javierre_parquet</span><span class=p>,</span> <span class=n>gene_index</span><span class=p>,</span> <span class=n>lift</span><span class=p>)</span></code> </h3> <div class="doc doc-contents "> <p>Initialise Javierre parser.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>javierre_parquet</code></td> <td> <code>str</code> </td> <td><p>path to the parquet file containing the Javierre 2016 data after processing it (see notebooks/Javierre_data_pre-process.ipynb)</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>gene_index</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Pyspark dataframe containing the gene index</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>lift</code></td> <td> <code><a class="autorefs autorefs-internal" title="etl.v2g.intervals.Liftover.LiftOverSpark" href="#etl.v2g.intervals.Liftover.LiftOverSpark">LiftOverSpark</a></code> </td> <td><p>LiftOverSpark object</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/javierre2016.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 42</span>
<span class=normal> 43</span>
<span class=normal> 44</span>
<span class=normal> 45</span>
<span class=normal> 46</span>
<span class=normal> 47</span>
<span class=normal> 48</span>
<span class=normal> 49</span>
<span class=normal> 50</span>
<span class=normal> 51</span>
<span class=normal> 52</span>
<span class=normal> 53</span>
<span class=normal> 54</span>
<span class=normal> 55</span>
<span class=normal> 56</span>
<span class=normal> 57</span>
<span class=normal> 58</span>
<span class=normal> 59</span>
<span class=normal> 60</span>
<span class=normal> 61</span>
<span class=normal> 62</span>
<span class=normal> 63</span>
<span class=normal> 64</span>
<span class=normal> 65</span>
<span class=normal> 66</span>
<span class=normal> 67</span>
<span class=normal> 68</span>
<span class=normal> 69</span>
<span class=normal> 70</span>
<span class=normal> 71</span>
<span class=normal> 72</span>
<span class=normal> 73</span>
<span class=normal> 74</span>
<span class=normal> 75</span>
<span class=normal> 76</span>
<span class=normal> 77</span>
<span class=normal> 78</span>
<span class=normal> 79</span>
<span class=normal> 80</span>
<span class=normal> 81</span>
<span class=normal> 82</span>
<span class=normal> 83</span>
<span class=normal> 84</span>
<span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span>
<span class=normal>125</span>
<span class=normal>126</span>
<span class=normal>127</span>
<span class=normal>128</span>
<span class=normal>129</span>
<span class=normal>130</span>
<span class=normal>131</span>
<span class=normal>132</span>
<span class=normal>133</span>
<span class=normal>134</span>
<span class=normal>135</span>
<span class=normal>136</span>
<span class=normal>137</span>
<span class=normal>138</span>
<span class=normal>139</span>
<span class=normal>140</span>
<span class=normal>141</span>
<span class=normal>142</span>
<span class=normal>143</span>
<span class=normal>144</span>
<span class=normal>145</span>
<span class=normal>146</span>
<span class=normal>147</span>
<span class=normal>148</span>
<span class=normal>149</span>
<span class=normal>150</span>
<span class=normal>151</span>
<span class=normal>152</span>
<span class=normal>153</span>
<span class=normal>154</span>
<span class=normal>155</span>
<span class=normal>156</span>
<span class=normal>157</span>
<span class=normal>158</span>
<span class=normal>159</span>
<span class=normal>160</span>
<span class=normal>161</span>
<span class=normal>162</span>
<span class=normal>163</span>
<span class=normal>164</span>
<span class=normal>165</span>
<span class=normal>166</span>
<span class=normal>167</span>
<span class=normal>168</span>
<span class=normal>169</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=fm>__init__</span><span class=p>(</span>
    <span class=bp>self</span><span class=p>:</span> <span class=n>ParseJavierre</span><span class=p>,</span>
    <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span>
    <span class=n>javierre_parquet</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
    <span class=n>gene_index</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
    <span class=n>lift</span><span class=p>:</span> <span class=n>LiftOverSpark</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Initialise Javierre parser.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): ETL session</span>
<span class=sd>        javierre_parquet (str): path to the parquet file containing the Javierre 2016 data after processing it (see notebooks/Javierre_data_pre-process.ipynb)</span>
<span class=sd>        gene_index (DataFrame): Pyspark dataframe containing the gene index</span>
<span class=sd>        lift (LiftOverSpark): LiftOverSpark object</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span> <span class=o>=</span> <span class=n>etl</span>

    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Parsing Javierre 2016 data...&quot;</span><span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Reading data from </span><span class=si>{</span><span class=n>javierre_parquet</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>

    <span class=c1># Read Javierre data:</span>
    <span class=n>javierre_raw</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>parquet</span><span class=p>(</span><span class=n>javierre_parquet</span><span class=p>)</span>
        <span class=c1># Splitting name column into chromosome, start, end, and score:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;name_split&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name&quot;</span><span class=p>),</span> <span class=sa>r</span><span class=s2>&quot;:|-|,&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;name_chr&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_split&quot;</span><span class=p>)[</span><span class=mi>0</span><span class=p>],</span> <span class=s2>&quot;chr&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span>
                <span class=n>t</span><span class=o>.</span><span class=n>StringType</span><span class=p>()</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;name_start&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_split&quot;</span><span class=p>)[</span><span class=mi>1</span><span class=p>]</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>()))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;name_end&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_split&quot;</span><span class=p>)[</span><span class=mi>2</span><span class=p>]</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>()))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;name_score&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_split&quot;</span><span class=p>)[</span><span class=mi>3</span><span class=p>]</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>FloatType</span><span class=p>()))</span>
        <span class=c1># Cleaning up chromosome:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;chrom&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>),</span> <span class=s2>&quot;chr&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>StringType</span><span class=p>()),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;name_split&quot;</span><span class=p>,</span> <span class=s2>&quot;name&quot;</span><span class=p>,</span> <span class=s2>&quot;annotation&quot;</span><span class=p>)</span>
        <span class=c1># Keep canonical chromosomes and consistent chromosomes with scores:</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
            <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_score&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>())</span>
            <span class=o>&amp;</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_chr&quot;</span><span class=p>))</span>
            <span class=o>&amp;</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_chr&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isin</span><span class=p>(</span>
                <span class=p>[</span><span class=sa>f</span><span class=s2>&quot;</span><span class=si>{</span><span class=n>x</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>x</span> <span class=ow>in</span> <span class=nb>range</span><span class=p>(</span><span class=mi>1</span><span class=p>,</span> <span class=mi>23</span><span class=p>)]</span> <span class=o>+</span> <span class=p>[</span><span class=s2>&quot;X&quot;</span><span class=p>,</span> <span class=s2>&quot;Y&quot;</span><span class=p>,</span> <span class=s2>&quot;MT&quot;</span><span class=p>]</span>
            <span class=p>)</span>
        <span class=p>)</span>
    <span class=p>)</span>

    <span class=c1># Lifting over intervals:</span>
    <span class=n>javierre_remapped</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>javierre_raw</span>
        <span class=c1># Lifting over to GRCh38 interval 1:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>lift</span><span class=o>.</span><span class=n>convert_intervals</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;chrom&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_start&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_end&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span>
        <span class=c1># Lifting over interval 2 to GRCh38:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span>
            <span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>lift</span><span class=o>.</span><span class=n>convert_intervals</span><span class=p>(</span>
                <span class=n>df</span><span class=p>,</span> <span class=s2>&quot;name_chr&quot;</span><span class=p>,</span> <span class=s2>&quot;name_start&quot;</span><span class=p>,</span> <span class=s2>&quot;name_end&quot;</span>
            <span class=p>)</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;name_start&quot;</span><span class=p>,</span> <span class=s2>&quot;name_end&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_name_chr&quot;</span><span class=p>,</span> <span class=s2>&quot;name_chr&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_name_start&quot;</span><span class=p>,</span> <span class=s2>&quot;name_start&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_name_end&quot;</span><span class=p>,</span> <span class=s2>&quot;name_end&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>

    <span class=c1># Once the intervals are lifted, extracting the unique intervals:</span>
    <span class=n>unique_intervals_with_genes</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>javierre_remapped</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;intervals&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>()),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>()),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>distinct</span><span class=p>()</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>gene_index</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;genes&quot;</span><span class=p>),</span>
            <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;intervals.chrom&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genes.chromosome&quot;</span><span class=p>)],</span>
            <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
            <span class=p>(</span>
                <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>)</span> <span class=o>&gt;=</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genomicLocation.start&quot;</span><span class=p>))</span>
                <span class=o>&amp;</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>)</span> <span class=o>&lt;=</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genomicLocation.end&quot;</span><span class=p>))</span>
            <span class=p>)</span>
            <span class=o>|</span> <span class=p>(</span>
                <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>)</span> <span class=o>&gt;=</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genomicLocation.start&quot;</span><span class=p>))</span>
                <span class=o>&amp;</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>)</span> <span class=o>&lt;=</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genomicLocation.end&quot;</span><span class=p>))</span>
            <span class=p>)</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>,</span> <span class=s2>&quot;geneId&quot;</span><span class=p>,</span> <span class=s2>&quot;tss&quot;</span><span class=p>)</span>
    <span class=p>)</span>

    <span class=c1># Joining back the data:</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>javierre_intervals</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>javierre_remapped</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>unique_intervals_with_genes</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>],</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
            <span class=c1># Drop rows where the TSS is far from the start of the region</span>
            <span class=n>f</span><span class=o>.</span><span class=n>abs</span><span class=p>((</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>)</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>))</span> <span class=o>/</span> <span class=mi>2</span> <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;tss&quot;</span><span class=p>))</span>
            <span class=o>&lt;=</span> <span class=bp>self</span><span class=o>.</span><span class=n>TWOSIDED_THRESHOLD</span>
        <span class=p>)</span>
        <span class=c1># For each gene, keep only the highest scoring interval:</span>
        <span class=o>.</span><span class=n>groupBy</span><span class=p>(</span>
            <span class=s2>&quot;name_chr&quot;</span><span class=p>,</span> <span class=s2>&quot;name_start&quot;</span><span class=p>,</span> <span class=s2>&quot;name_end&quot;</span><span class=p>,</span> <span class=s2>&quot;genes.geneId&quot;</span><span class=p>,</span> <span class=s2>&quot;bio_feature&quot;</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>agg</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>max</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_score&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;resourceScore&quot;</span><span class=p>))</span>
        <span class=c1># Create the output:</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_chr&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_start&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name_end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;resourceScore&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genes.geneId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;bio_feature&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;biofeature&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>DATASET_NAME</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datasourceId&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>EXPERIMENT_TYPE</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datatypeId&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>PMID</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;pmid&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.javierre2016.ParseJavierre.get_intervals class="doc doc-heading"> <code class="highlight language-python"><span class=n>get_intervals</span><span class=p>()</span></code> </h3> <div class="doc doc-contents "> <p>Get preformatted Javierre intervals.</p> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/javierre2016.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>171</span>
<span class=normal>172</span>
<span class=normal>173</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>get_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseJavierre</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Get preformatted Javierre intervals.&quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=bp>self</span><span class=o>.</span><span class=n>javierre_intervals</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.javierre2016.ParseJavierre.qc_intervals class="doc doc-heading"> <code class="highlight language-python"><span class=n>qc_intervals</span><span class=p>()</span></code> </h3> <div class="doc doc-contents "> <p>Perform QC on the anderson intervals.</p> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/javierre2016.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>175</span>
<span class=normal>176</span>
<span class=normal>177</span>
<span class=normal>178</span>
<span class=normal>179</span>
<span class=normal>180</span>
<span class=normal>181</span>
<span class=normal>182</span>
<span class=normal>183</span>
<span class=normal>184</span>
<span class=normal>185</span>
<span class=normal>186</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>qc_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseJavierre</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Perform QC on the anderson intervals.&quot;&quot;&quot;</span>
    <span class=c1># Get numbers:</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s2>&quot;Size of Javierre data: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>javierre_intervals</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span>
    <span class=p>)</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number of unique intervals: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>javierre_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number genes in the Javierre dataset: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>javierre_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> </div> </div> </div> <div class="doc doc-object doc-module"> <a id=etl.v2g.intervals.andersson2014></a> <div class="doc doc-contents first"> <p>Enhancer-TSS correlation (Andersson et al. 2014) intervals.</p> <p>As part of the <a href=https://fantom.gsc.riken.jp/5/ >FANTOM5</a> genome mapping effort, this publication aims to report on actively transcribed enhancers from the majority of human tissues. (<a href=https://www.nature.com/articles/nature12787>Link</a> to the publication)</p> <p>The dataset is not allows to resolve individual tissues, the biotype is <code>aggregate</code>. However the aggregation provides a score quantifying the association of the genomic region and the gene.</p> <div class="doc doc-children"> <div class="doc doc-object doc-class"> <h2 id=etl.v2g.intervals.andersson2014.ParseAndersson class="doc doc-heading"> <code>ParseAndersson</code> </h2> <div class="doc doc-contents "> <p>Parse the anderson file and return a dataframe with the intervals.</p> <p><strong>Summary of the logic</strong></p> <ul> <li>Reading .bed file (input)</li> <li>Parsing the names column -&gt; chr, start, end, gene, score</li> <li>Mapping the coordinates to the new build -&gt; liftover</li> <li>Joining with target index by gene symbol (some loss as input uses obsoleted terms)</li> <li>Dropping rows where the gene is on other chromosomes</li> <li>Dropping rows where the gene TSS is too far from the midpoint of the intervals</li> <li>Adding constant columns for this dataset</li> </ul> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/andersson2014.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 26</span>
<span class=normal> 27</span>
<span class=normal> 28</span>
<span class=normal> 29</span>
<span class=normal> 30</span>
<span class=normal> 31</span>
<span class=normal> 32</span>
<span class=normal> 33</span>
<span class=normal> 34</span>
<span class=normal> 35</span>
<span class=normal> 36</span>
<span class=normal> 37</span>
<span class=normal> 38</span>
<span class=normal> 39</span>
<span class=normal> 40</span>
<span class=normal> 41</span>
<span class=normal> 42</span>
<span class=normal> 43</span>
<span class=normal> 44</span>
<span class=normal> 45</span>
<span class=normal> 46</span>
<span class=normal> 47</span>
<span class=normal> 48</span>
<span class=normal> 49</span>
<span class=normal> 50</span>
<span class=normal> 51</span>
<span class=normal> 52</span>
<span class=normal> 53</span>
<span class=normal> 54</span>
<span class=normal> 55</span>
<span class=normal> 56</span>
<span class=normal> 57</span>
<span class=normal> 58</span>
<span class=normal> 59</span>
<span class=normal> 60</span>
<span class=normal> 61</span>
<span class=normal> 62</span>
<span class=normal> 63</span>
<span class=normal> 64</span>
<span class=normal> 65</span>
<span class=normal> 66</span>
<span class=normal> 67</span>
<span class=normal> 68</span>
<span class=normal> 69</span>
<span class=normal> 70</span>
<span class=normal> 71</span>
<span class=normal> 72</span>
<span class=normal> 73</span>
<span class=normal> 74</span>
<span class=normal> 75</span>
<span class=normal> 76</span>
<span class=normal> 77</span>
<span class=normal> 78</span>
<span class=normal> 79</span>
<span class=normal> 80</span>
<span class=normal> 81</span>
<span class=normal> 82</span>
<span class=normal> 83</span>
<span class=normal> 84</span>
<span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span>
<span class=normal>125</span>
<span class=normal>126</span>
<span class=normal>127</span>
<span class=normal>128</span>
<span class=normal>129</span>
<span class=normal>130</span>
<span class=normal>131</span>
<span class=normal>132</span>
<span class=normal>133</span>
<span class=normal>134</span>
<span class=normal>135</span>
<span class=normal>136</span>
<span class=normal>137</span>
<span class=normal>138</span>
<span class=normal>139</span>
<span class=normal>140</span>
<span class=normal>141</span>
<span class=normal>142</span>
<span class=normal>143</span>
<span class=normal>144</span>
<span class=normal>145</span>
<span class=normal>146</span>
<span class=normal>147</span>
<span class=normal>148</span>
<span class=normal>149</span>
<span class=normal>150</span>
<span class=normal>151</span>
<span class=normal>152</span>
<span class=normal>153</span>
<span class=normal>154</span>
<span class=normal>155</span>
<span class=normal>156</span>
<span class=normal>157</span>
<span class=normal>158</span>
<span class=normal>159</span>
<span class=normal>160</span>
<span class=normal>161</span>
<span class=normal>162</span>
<span class=normal>163</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>class</span> <span class=nc>ParseAndersson</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Parse the anderson file and return a dataframe with the intervals.</span>

<span class=sd>    **Summary of the logic**</span>

<span class=sd>    - Reading .bed file (input)</span>
<span class=sd>    - Parsing the names column -&gt; chr, start, end, gene, score</span>
<span class=sd>    - Mapping the coordinates to the new build -&gt; liftover</span>
<span class=sd>    - Joining with target index by gene symbol (some loss as input uses obsoleted terms)</span>
<span class=sd>    - Dropping rows where the gene is on other chromosomes</span>
<span class=sd>    - Dropping rows where the gene TSS is too far from the midpoint of the intervals</span>
<span class=sd>    - Adding constant columns for this dataset</span>
<span class=sd>    &quot;&quot;&quot;</span>

    <span class=c1># Constant values:</span>
    <span class=n>DATASET_NAME</span> <span class=o>=</span> <span class=s2>&quot;andersson2014&quot;</span>
    <span class=n>DATA_TYPE</span> <span class=o>=</span> <span class=s2>&quot;interval&quot;</span>
    <span class=n>EXPERIMENT_TYPE</span> <span class=o>=</span> <span class=s2>&quot;fantom5&quot;</span>
    <span class=n>PMID</span> <span class=o>=</span> <span class=s2>&quot;24670763&quot;</span>
    <span class=n>BIO_FEATURE</span> <span class=o>=</span> <span class=s2>&quot;aggregate&quot;</span>
    <span class=n>TWOSIDED_THRESHOLD</span> <span class=o>=</span> <span class=mf>2.45e6</span>  <span class=c1># &lt;-  this needs to phased out. Filter by percentile instead of absolute value.</span>

    <span class=k>def</span> <span class=fm>__init__</span><span class=p>(</span>
        <span class=bp>self</span><span class=p>:</span> <span class=n>ParseAndersson</span><span class=p>,</span>
        <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span>
        <span class=n>anderson_data_file</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
        <span class=n>gene_index</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
        <span class=n>lift</span><span class=p>:</span> <span class=n>LiftOverSpark</span><span class=p>,</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Intialise Andersson parser.</span>

<span class=sd>        Args:</span>
<span class=sd>            etl (ETLSession): Spark session</span>
<span class=sd>            anderson_data_file (str): Anderson et al. filepath</span>
<span class=sd>            gene_index (DataFrame): gene index information</span>
<span class=sd>            lift (LiftOverSpark): LiftOverSpark object</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span> <span class=o>=</span> <span class=n>etl</span>

        <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Parsing Andersson 2014 data...&quot;</span><span class=p>)</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Reading data from </span><span class=si>{</span><span class=n>anderson_data_file</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>

        <span class=c1># Expected andersson et al. schema:</span>
        <span class=n>schema</span> <span class=o>=</span> <span class=n>t</span><span class=o>.</span><span class=n>StructType</span><span class=o>.</span><span class=n>fromJson</span><span class=p>(</span>
            <span class=n>json</span><span class=o>.</span><span class=n>loads</span><span class=p>(</span>
                <span class=n>pkg_resources</span><span class=o>.</span><span class=n>read_text</span><span class=p>(</span><span class=n>schemas</span><span class=p>,</span> <span class=s2>&quot;andersson2014.json&quot;</span><span class=p>,</span> <span class=n>encoding</span><span class=o>=</span><span class=s2>&quot;utf-8&quot;</span><span class=p>)</span>
            <span class=p>)</span>
        <span class=p>)</span>

        <span class=c1># Read the anderson file:</span>
        <span class=n>parsed_anderson_df</span> <span class=o>=</span> <span class=p>(</span>
            <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>option</span><span class=p>(</span><span class=s2>&quot;delimiter&quot;</span><span class=p>,</span> <span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>option</span><span class=p>(</span><span class=s2>&quot;header&quot;</span><span class=p>,</span> <span class=s2>&quot;true&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>schema</span><span class=p>(</span><span class=n>schema</span><span class=p>)</span>
            <span class=o>.</span><span class=n>csv</span><span class=p>(</span><span class=n>anderson_data_file</span><span class=p>)</span>
            <span class=c1># Parsing score column and casting as float:</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=s2>&quot;float&quot;</span><span class=p>)</span> <span class=o>/</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mi>1000</span><span class=p>))</span>
            <span class=c1># Parsing the &#39;name&#39; column:</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;parsedName&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name&quot;</span><span class=p>),</span> <span class=s2>&quot;;&quot;</span><span class=p>))</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;gene_symbol&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;parsedName&quot;</span><span class=p>)[</span><span class=mi>2</span><span class=p>])</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;location&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;parsedName&quot;</span><span class=p>)[</span><span class=mi>0</span><span class=p>])</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
                <span class=s2>&quot;chrom&quot;</span><span class=p>,</span>
                <span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;location&quot;</span><span class=p>),</span> <span class=s2>&quot;:|-&quot;</span><span class=p>)[</span><span class=mi>0</span><span class=p>],</span> <span class=s2>&quot;chr&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>),</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
                <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;location&quot;</span><span class=p>),</span> <span class=s2>&quot;:|-&quot;</span><span class=p>)[</span><span class=mi>1</span><span class=p>]</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>())</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
                <span class=s2>&quot;end&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;location&quot;</span><span class=p>),</span> <span class=s2>&quot;:|-&quot;</span><span class=p>)[</span><span class=mi>2</span><span class=p>]</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>())</span>
            <span class=p>)</span>
            <span class=c1># Select relevant columns:</span>
            <span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>,</span> <span class=s2>&quot;gene_symbol&quot;</span><span class=p>,</span> <span class=s2>&quot;score&quot;</span><span class=p>)</span>
            <span class=c1># Drop rows with non-canonical chromosomes:</span>
            <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isin</span><span class=p>([</span><span class=nb>str</span><span class=p>(</span><span class=n>x</span><span class=p>)</span> <span class=k>for</span> <span class=n>x</span> <span class=ow>in</span> <span class=nb>range</span><span class=p>(</span><span class=mi>1</span><span class=p>,</span> <span class=mi>23</span><span class=p>)]</span> <span class=o>+</span> <span class=p>[</span><span class=s2>&quot;X&quot;</span><span class=p>,</span> <span class=s2>&quot;Y&quot;</span><span class=p>,</span> <span class=s2>&quot;MT&quot;</span><span class=p>])</span>
            <span class=p>)</span>
            <span class=c1># For each region/gene, keep only one row with the highest score:</span>
            <span class=o>.</span><span class=n>groupBy</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>,</span> <span class=s2>&quot;gene_symbol&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>agg</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>max</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;resourceScore&quot;</span><span class=p>))</span>
            <span class=o>.</span><span class=n>orderBy</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
        <span class=p>)</span>

        <span class=bp>self</span><span class=o>.</span><span class=n>anderson_intervals</span> <span class=o>=</span> <span class=p>(</span>
            <span class=c1># Lift over the intervals:</span>
            <span class=n>lift</span><span class=o>.</span><span class=n>convert_intervals</span><span class=p>(</span><span class=n>parsed_anderson_df</span><span class=p>,</span> <span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_start&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_end&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>distinct</span><span class=p>()</span>
            <span class=c1># Joining with the gene index</span>
            <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;intervals&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>join</span><span class=p>(</span>
                <span class=n>gene_index</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;genes&quot;</span><span class=p>),</span>
                <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;intervals.gene_symbol&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genes.geneSymbol&quot;</span><span class=p>)],</span>
                <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
                <span class=c1># Drop rows where the gene is not on the same chromosome</span>
                <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>))</span>
                <span class=c1># Drop rows where the TSS is far from the start of the region</span>
                <span class=o>&amp;</span> <span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>abs</span><span class=p>((</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>)</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>))</span> <span class=o>/</span> <span class=mi>2</span> <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;tss&quot;</span><span class=p>))</span>
                    <span class=o>&lt;=</span> <span class=bp>self</span><span class=o>.</span><span class=n>TWOSIDED_THRESHOLD</span>
                <span class=p>)</span>
            <span class=p>)</span>
            <span class=c1># Select relevant columns:</span>
            <span class=o>.</span><span class=n>select</span><span class=p>(</span>
                <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
                <span class=s2>&quot;start&quot;</span><span class=p>,</span>
                <span class=s2>&quot;end&quot;</span><span class=p>,</span>
                <span class=s2>&quot;geneId&quot;</span><span class=p>,</span>
                <span class=s2>&quot;resourceScore&quot;</span><span class=p>,</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>DATASET_NAME</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datasourceId&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>EXPERIMENT_TYPE</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datatypeId&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>PMID</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;pmid&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>BIO_FEATURE</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;biofeature&quot;</span><span class=p>),</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
        <span class=p>)</span>

    <span class=k>def</span> <span class=nf>get_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseAndersson</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Get formatted interval data.&quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=bp>self</span><span class=o>.</span><span class=n>anderson_intervals</span>

    <span class=k>def</span> <span class=nf>qc_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseAndersson</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Perform QC on the anderson intervals.&quot;&quot;&quot;</span>
        <span class=c1># Get numbers:</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
            <span class=sa>f</span><span class=s2>&quot;Size of Andersson data: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>anderson_intervals</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span>
        <span class=p>)</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
            <span class=sa>f</span><span class=s1>&#39;Number of unique intervals: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>anderson_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
        <span class=p>)</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
            <span class=sa>f</span><span class=s1>&#39;Number genes in the Andersson dataset: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>anderson_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
        <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.andersson2014.ParseAndersson.__init__ class="doc doc-heading"> <code class="highlight language-python"><span class=fm>__init__</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>anderson_data_file</span><span class=p>,</span> <span class=n>gene_index</span><span class=p>,</span> <span class=n>lift</span><span class=p>)</span></code> </h3> <div class="doc doc-contents "> <p>Intialise Andersson parser.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>Spark session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>anderson_data_file</code></td> <td> <code>str</code> </td> <td><p>Anderson et al. filepath</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>gene_index</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>gene index information</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>lift</code></td> <td> <code><a class="autorefs autorefs-internal" title="etl.v2g.intervals.Liftover.LiftOverSpark" href="#etl.v2g.intervals.Liftover.LiftOverSpark">LiftOverSpark</a></code> </td> <td><p>LiftOverSpark object</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/andersson2014.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 48</span>
<span class=normal> 49</span>
<span class=normal> 50</span>
<span class=normal> 51</span>
<span class=normal> 52</span>
<span class=normal> 53</span>
<span class=normal> 54</span>
<span class=normal> 55</span>
<span class=normal> 56</span>
<span class=normal> 57</span>
<span class=normal> 58</span>
<span class=normal> 59</span>
<span class=normal> 60</span>
<span class=normal> 61</span>
<span class=normal> 62</span>
<span class=normal> 63</span>
<span class=normal> 64</span>
<span class=normal> 65</span>
<span class=normal> 66</span>
<span class=normal> 67</span>
<span class=normal> 68</span>
<span class=normal> 69</span>
<span class=normal> 70</span>
<span class=normal> 71</span>
<span class=normal> 72</span>
<span class=normal> 73</span>
<span class=normal> 74</span>
<span class=normal> 75</span>
<span class=normal> 76</span>
<span class=normal> 77</span>
<span class=normal> 78</span>
<span class=normal> 79</span>
<span class=normal> 80</span>
<span class=normal> 81</span>
<span class=normal> 82</span>
<span class=normal> 83</span>
<span class=normal> 84</span>
<span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span>
<span class=normal>125</span>
<span class=normal>126</span>
<span class=normal>127</span>
<span class=normal>128</span>
<span class=normal>129</span>
<span class=normal>130</span>
<span class=normal>131</span>
<span class=normal>132</span>
<span class=normal>133</span>
<span class=normal>134</span>
<span class=normal>135</span>
<span class=normal>136</span>
<span class=normal>137</span>
<span class=normal>138</span>
<span class=normal>139</span>
<span class=normal>140</span>
<span class=normal>141</span>
<span class=normal>142</span>
<span class=normal>143</span>
<span class=normal>144</span>
<span class=normal>145</span>
<span class=normal>146</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=fm>__init__</span><span class=p>(</span>
    <span class=bp>self</span><span class=p>:</span> <span class=n>ParseAndersson</span><span class=p>,</span>
    <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span>
    <span class=n>anderson_data_file</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
    <span class=n>gene_index</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
    <span class=n>lift</span><span class=p>:</span> <span class=n>LiftOverSpark</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Intialise Andersson parser.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): Spark session</span>
<span class=sd>        anderson_data_file (str): Anderson et al. filepath</span>
<span class=sd>        gene_index (DataFrame): gene index information</span>
<span class=sd>        lift (LiftOverSpark): LiftOverSpark object</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span> <span class=o>=</span> <span class=n>etl</span>

    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Parsing Andersson 2014 data...&quot;</span><span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Reading data from </span><span class=si>{</span><span class=n>anderson_data_file</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>

    <span class=c1># Expected andersson et al. schema:</span>
    <span class=n>schema</span> <span class=o>=</span> <span class=n>t</span><span class=o>.</span><span class=n>StructType</span><span class=o>.</span><span class=n>fromJson</span><span class=p>(</span>
        <span class=n>json</span><span class=o>.</span><span class=n>loads</span><span class=p>(</span>
            <span class=n>pkg_resources</span><span class=o>.</span><span class=n>read_text</span><span class=p>(</span><span class=n>schemas</span><span class=p>,</span> <span class=s2>&quot;andersson2014.json&quot;</span><span class=p>,</span> <span class=n>encoding</span><span class=o>=</span><span class=s2>&quot;utf-8&quot;</span><span class=p>)</span>
        <span class=p>)</span>
    <span class=p>)</span>

    <span class=c1># Read the anderson file:</span>
    <span class=n>parsed_anderson_df</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>option</span><span class=p>(</span><span class=s2>&quot;delimiter&quot;</span><span class=p>,</span> <span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>option</span><span class=p>(</span><span class=s2>&quot;header&quot;</span><span class=p>,</span> <span class=s2>&quot;true&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>schema</span><span class=p>(</span><span class=n>schema</span><span class=p>)</span>
        <span class=o>.</span><span class=n>csv</span><span class=p>(</span><span class=n>anderson_data_file</span><span class=p>)</span>
        <span class=c1># Parsing score column and casting as float:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=s2>&quot;float&quot;</span><span class=p>)</span> <span class=o>/</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mi>1000</span><span class=p>))</span>
        <span class=c1># Parsing the &#39;name&#39; column:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;parsedName&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;name&quot;</span><span class=p>),</span> <span class=s2>&quot;;&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;gene_symbol&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;parsedName&quot;</span><span class=p>)[</span><span class=mi>2</span><span class=p>])</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;location&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;parsedName&quot;</span><span class=p>)[</span><span class=mi>0</span><span class=p>])</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;chrom&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;location&quot;</span><span class=p>),</span> <span class=s2>&quot;:|-&quot;</span><span class=p>)[</span><span class=mi>0</span><span class=p>],</span> <span class=s2>&quot;chr&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;location&quot;</span><span class=p>),</span> <span class=s2>&quot;:|-&quot;</span><span class=p>)[</span><span class=mi>1</span><span class=p>]</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>())</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;end&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;location&quot;</span><span class=p>),</span> <span class=s2>&quot;:|-&quot;</span><span class=p>)[</span><span class=mi>2</span><span class=p>]</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>())</span>
        <span class=p>)</span>
        <span class=c1># Select relevant columns:</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>,</span> <span class=s2>&quot;gene_symbol&quot;</span><span class=p>,</span> <span class=s2>&quot;score&quot;</span><span class=p>)</span>
        <span class=c1># Drop rows with non-canonical chromosomes:</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isin</span><span class=p>([</span><span class=nb>str</span><span class=p>(</span><span class=n>x</span><span class=p>)</span> <span class=k>for</span> <span class=n>x</span> <span class=ow>in</span> <span class=nb>range</span><span class=p>(</span><span class=mi>1</span><span class=p>,</span> <span class=mi>23</span><span class=p>)]</span> <span class=o>+</span> <span class=p>[</span><span class=s2>&quot;X&quot;</span><span class=p>,</span> <span class=s2>&quot;Y&quot;</span><span class=p>,</span> <span class=s2>&quot;MT&quot;</span><span class=p>])</span>
        <span class=p>)</span>
        <span class=c1># For each region/gene, keep only one row with the highest score:</span>
        <span class=o>.</span><span class=n>groupBy</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>,</span> <span class=s2>&quot;gene_symbol&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>agg</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>max</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;resourceScore&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>orderBy</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>

    <span class=bp>self</span><span class=o>.</span><span class=n>anderson_intervals</span> <span class=o>=</span> <span class=p>(</span>
        <span class=c1># Lift over the intervals:</span>
        <span class=n>lift</span><span class=o>.</span><span class=n>convert_intervals</span><span class=p>(</span><span class=n>parsed_anderson_df</span><span class=p>,</span> <span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_start&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_end&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>distinct</span><span class=p>()</span>
        <span class=c1># Joining with the gene index</span>
        <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;intervals&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>gene_index</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;genes&quot;</span><span class=p>),</span>
            <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;intervals.gene_symbol&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genes.geneSymbol&quot;</span><span class=p>)],</span>
            <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
            <span class=c1># Drop rows where the gene is not on the same chromosome</span>
            <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>))</span>
            <span class=c1># Drop rows where the TSS is far from the start of the region</span>
            <span class=o>&amp;</span> <span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>abs</span><span class=p>((</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>)</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>))</span> <span class=o>/</span> <span class=mi>2</span> <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;tss&quot;</span><span class=p>))</span>
                <span class=o>&lt;=</span> <span class=bp>self</span><span class=o>.</span><span class=n>TWOSIDED_THRESHOLD</span>
            <span class=p>)</span>
        <span class=p>)</span>
        <span class=c1># Select relevant columns:</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
            <span class=s2>&quot;start&quot;</span><span class=p>,</span>
            <span class=s2>&quot;end&quot;</span><span class=p>,</span>
            <span class=s2>&quot;geneId&quot;</span><span class=p>,</span>
            <span class=s2>&quot;resourceScore&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>DATASET_NAME</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datasourceId&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>EXPERIMENT_TYPE</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datatypeId&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>PMID</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;pmid&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>BIO_FEATURE</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;biofeature&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.andersson2014.ParseAndersson.get_intervals class="doc doc-heading"> <code class="highlight language-python"><span class=n>get_intervals</span><span class=p>()</span></code> </h3> <div class="doc doc-contents "> <p>Get formatted interval data.</p> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/andersson2014.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>148</span>
<span class=normal>149</span>
<span class=normal>150</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>get_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseAndersson</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Get formatted interval data.&quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=bp>self</span><span class=o>.</span><span class=n>anderson_intervals</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.andersson2014.ParseAndersson.qc_intervals class="doc doc-heading"> <code class="highlight language-python"><span class=n>qc_intervals</span><span class=p>()</span></code> </h3> <div class="doc doc-contents "> <p>Perform QC on the anderson intervals.</p> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/andersson2014.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>152</span>
<span class=normal>153</span>
<span class=normal>154</span>
<span class=normal>155</span>
<span class=normal>156</span>
<span class=normal>157</span>
<span class=normal>158</span>
<span class=normal>159</span>
<span class=normal>160</span>
<span class=normal>161</span>
<span class=normal>162</span>
<span class=normal>163</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>qc_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseAndersson</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Perform QC on the anderson intervals.&quot;&quot;&quot;</span>
    <span class=c1># Get numbers:</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s2>&quot;Size of Andersson data: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>anderson_intervals</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span>
    <span class=p>)</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number of unique intervals: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>anderson_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number genes in the Andersson dataset: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>anderson_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> </div> </div> </div> <div class="doc doc-object doc-module"> <a id=etl.v2g.intervals.thurman2012></a> <div class="doc doc-contents first"> <p>DHS-promoter correlation (Thurnman, 2012) intervals.</p> <p>In this projet cis regulatory elements were mapped using DNaseI hypersensitive site (DHSs) mapping. (<a href=https://www.nature.com/articles/nature11232>Link</a> to the publication)</p> <p>This is also an aggregated dataset, so no cellular or tissue annotation is preserved, however scores are given.</p> <div class="doc doc-children"> <div class="doc doc-object doc-class"> <h2 id=etl.v2g.intervals.thurman2012.ParseThurman class="doc doc-heading"> <code>ParseThurman</code> </h2> <div class="doc doc-contents "> <p>Parser Thurman 2012 dataset.</p> <p>:param Thurman_parquet: path to the parquet file containing the Thurman 2012 data :param gene_index: Pyspark dataframe containing the gene index :param lift: LiftOverSpark object</p> <p><strong>Summary of the logic:</strong></p> <ul> <li>Lifting over coordinates to GRCh38</li> <li>Mapping genes names to gene IDs -&gt; we might need to measure the loss of genes if there are obsoleted names.</li> </ul> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/thurman2012.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 21</span>
<span class=normal> 22</span>
<span class=normal> 23</span>
<span class=normal> 24</span>
<span class=normal> 25</span>
<span class=normal> 26</span>
<span class=normal> 27</span>
<span class=normal> 28</span>
<span class=normal> 29</span>
<span class=normal> 30</span>
<span class=normal> 31</span>
<span class=normal> 32</span>
<span class=normal> 33</span>
<span class=normal> 34</span>
<span class=normal> 35</span>
<span class=normal> 36</span>
<span class=normal> 37</span>
<span class=normal> 38</span>
<span class=normal> 39</span>
<span class=normal> 40</span>
<span class=normal> 41</span>
<span class=normal> 42</span>
<span class=normal> 43</span>
<span class=normal> 44</span>
<span class=normal> 45</span>
<span class=normal> 46</span>
<span class=normal> 47</span>
<span class=normal> 48</span>
<span class=normal> 49</span>
<span class=normal> 50</span>
<span class=normal> 51</span>
<span class=normal> 52</span>
<span class=normal> 53</span>
<span class=normal> 54</span>
<span class=normal> 55</span>
<span class=normal> 56</span>
<span class=normal> 57</span>
<span class=normal> 58</span>
<span class=normal> 59</span>
<span class=normal> 60</span>
<span class=normal> 61</span>
<span class=normal> 62</span>
<span class=normal> 63</span>
<span class=normal> 64</span>
<span class=normal> 65</span>
<span class=normal> 66</span>
<span class=normal> 67</span>
<span class=normal> 68</span>
<span class=normal> 69</span>
<span class=normal> 70</span>
<span class=normal> 71</span>
<span class=normal> 72</span>
<span class=normal> 73</span>
<span class=normal> 74</span>
<span class=normal> 75</span>
<span class=normal> 76</span>
<span class=normal> 77</span>
<span class=normal> 78</span>
<span class=normal> 79</span>
<span class=normal> 80</span>
<span class=normal> 81</span>
<span class=normal> 82</span>
<span class=normal> 83</span>
<span class=normal> 84</span>
<span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>class</span> <span class=nc>ParseThurman</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Parser Thurman 2012 dataset.</span>

<span class=sd>    :param Thurman_parquet: path to the parquet file containing the Thurman 2012 data</span>
<span class=sd>    :param gene_index: Pyspark dataframe containing the gene index</span>
<span class=sd>    :param lift: LiftOverSpark object</span>

<span class=sd>    **Summary of the logic:**</span>

<span class=sd>    - Lifting over coordinates to GRCh38</span>
<span class=sd>    - Mapping genes names to gene IDs -&gt; we might need to measure the loss of genes if there are obsoleted names.</span>
<span class=sd>    &quot;&quot;&quot;</span>

    <span class=c1># Constants:</span>
    <span class=n>DATASET_NAME</span> <span class=o>=</span> <span class=s2>&quot;thurman2012&quot;</span>
    <span class=n>DATA_TYPE</span> <span class=o>=</span> <span class=s2>&quot;interval&quot;</span>
    <span class=n>EXPERIMENT_TYPE</span> <span class=o>=</span> <span class=s2>&quot;dhscor&quot;</span>
    <span class=n>PMID</span> <span class=o>=</span> <span class=s2>&quot;22955617&quot;</span>
    <span class=n>BIO_FEATURE</span> <span class=o>=</span> <span class=s2>&quot;aggregate&quot;</span>

    <span class=k>def</span> <span class=fm>__init__</span><span class=p>(</span>
        <span class=bp>self</span><span class=p>:</span> <span class=n>ParseThurman</span><span class=p>,</span>
        <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span>
        <span class=n>thurman_datafile</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
        <span class=n>gene_index</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
        <span class=n>lift</span><span class=p>:</span> <span class=n>LiftOverSpark</span><span class=p>,</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Initialise Thurnman parser.</span>

<span class=sd>        Args:</span>
<span class=sd>            etl (ETLSession): current ETL session</span>
<span class=sd>            thurman_datafile (str): filepath to thurnman dataset</span>
<span class=sd>            gene_index (DataFrame): gene/target index</span>
<span class=sd>            lift (LiftOverSpark): LiftOverSpark object</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span> <span class=o>=</span> <span class=n>etl</span>

        <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Parsing Thurman 2012 data...&quot;</span><span class=p>)</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Reading data from </span><span class=si>{</span><span class=n>thurman_datafile</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>

        <span class=n>thurman_schema</span> <span class=o>=</span> <span class=n>t</span><span class=o>.</span><span class=n>StructType</span><span class=p>(</span>
            <span class=p>[</span>
                <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;gene_chr&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>StringType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
                <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;gene_start&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
                <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;gene_end&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
                <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;gene_name&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>StringType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
                <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>StringType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
                <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
                <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
                <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>FloatType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
            <span class=p>]</span>
        <span class=p>)</span>

        <span class=c1># Process Thurman data in a single step:</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>Thurman_intervals</span> <span class=o>=</span> <span class=p>(</span>
            <span class=n>etl</span><span class=o>.</span><span class=n>spark</span>
            <span class=c1># Read table according to the schema, then do some modifications:</span>
            <span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>csv</span><span class=p>(</span><span class=n>thurman_datafile</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>header</span><span class=o>=</span><span class=kc>False</span><span class=p>,</span> <span class=n>schema</span><span class=o>=</span><span class=n>thurman_schema</span><span class=p>)</span>
            <span class=o>.</span><span class=n>select</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>),</span> <span class=s2>&quot;chr&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>),</span>
                <span class=s2>&quot;start&quot;</span><span class=p>,</span>
                <span class=s2>&quot;end&quot;</span><span class=p>,</span>
                <span class=s2>&quot;gene_name&quot;</span><span class=p>,</span>
                <span class=s2>&quot;score&quot;</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=c1># Lift over to the GRCh38 build:</span>
            <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>lift</span><span class=o>.</span><span class=n>convert_intervals</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>))</span>
            <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;intervals&quot;</span><span class=p>)</span>
            <span class=c1># Map gene names to gene IDs:</span>
            <span class=o>.</span><span class=n>join</span><span class=p>(</span>
                <span class=n>gene_index</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;genes&quot;</span><span class=p>),</span>
                <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;intervals.gene_name&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genes.geneSymbol&quot;</span><span class=p>)],</span>
                <span class=n>how</span><span class=o>=</span><span class=s2>&quot;inner&quot;</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=c1># Select relevant columns and add constant columns:</span>
            <span class=o>.</span><span class=n>select</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped_start&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped_end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>),</span>
                <span class=s2>&quot;geneId&quot;</span><span class=p>,</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;resourceScore&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>DATASET_NAME</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datasourceId&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>EXPERIMENT_TYPE</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datatypeId&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>PMID</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;pmid&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>BIO_FEATURE</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;biofeature&quot;</span><span class=p>),</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>distinct</span><span class=p>()</span>
            <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
        <span class=p>)</span>

    <span class=k>def</span> <span class=nf>get_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseThurman</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Get Thurman intervals.&quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=bp>self</span><span class=o>.</span><span class=n>Thurman_intervals</span>

    <span class=k>def</span> <span class=nf>qc_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseThurman</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Perform QC on the anderson intervals.&quot;&quot;&quot;</span>
        <span class=c1># Get numbers:</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Size of Thurman data: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>Thurman_intervals</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
            <span class=sa>f</span><span class=s1>&#39;Number of unique intervals: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>Thurman_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
        <span class=p>)</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
            <span class=sa>f</span><span class=s1>&#39;Number genes in the Thurman dataset: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>Thurman_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
        <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.thurman2012.ParseThurman.__init__ class="doc doc-heading"> <code class="highlight language-python"><span class=fm>__init__</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>thurman_datafile</span><span class=p>,</span> <span class=n>gene_index</span><span class=p>,</span> <span class=n>lift</span><span class=p>)</span></code> </h3> <div class="doc doc-contents "> <p>Initialise Thurnman parser.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>current ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>thurman_datafile</code></td> <td> <code>str</code> </td> <td><p>filepath to thurnman dataset</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>gene_index</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>gene/target index</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>lift</code></td> <td> <code><a class="autorefs autorefs-internal" title="etl.v2g.intervals.Liftover.LiftOverSpark" href="#etl.v2g.intervals.Liftover.LiftOverSpark">LiftOverSpark</a></code> </td> <td><p>LiftOverSpark object</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/thurman2012.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 41</span>
<span class=normal> 42</span>
<span class=normal> 43</span>
<span class=normal> 44</span>
<span class=normal> 45</span>
<span class=normal> 46</span>
<span class=normal> 47</span>
<span class=normal> 48</span>
<span class=normal> 49</span>
<span class=normal> 50</span>
<span class=normal> 51</span>
<span class=normal> 52</span>
<span class=normal> 53</span>
<span class=normal> 54</span>
<span class=normal> 55</span>
<span class=normal> 56</span>
<span class=normal> 57</span>
<span class=normal> 58</span>
<span class=normal> 59</span>
<span class=normal> 60</span>
<span class=normal> 61</span>
<span class=normal> 62</span>
<span class=normal> 63</span>
<span class=normal> 64</span>
<span class=normal> 65</span>
<span class=normal> 66</span>
<span class=normal> 67</span>
<span class=normal> 68</span>
<span class=normal> 69</span>
<span class=normal> 70</span>
<span class=normal> 71</span>
<span class=normal> 72</span>
<span class=normal> 73</span>
<span class=normal> 74</span>
<span class=normal> 75</span>
<span class=normal> 76</span>
<span class=normal> 77</span>
<span class=normal> 78</span>
<span class=normal> 79</span>
<span class=normal> 80</span>
<span class=normal> 81</span>
<span class=normal> 82</span>
<span class=normal> 83</span>
<span class=normal> 84</span>
<span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=fm>__init__</span><span class=p>(</span>
    <span class=bp>self</span><span class=p>:</span> <span class=n>ParseThurman</span><span class=p>,</span>
    <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span>
    <span class=n>thurman_datafile</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
    <span class=n>gene_index</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
    <span class=n>lift</span><span class=p>:</span> <span class=n>LiftOverSpark</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Initialise Thurnman parser.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): current ETL session</span>
<span class=sd>        thurman_datafile (str): filepath to thurnman dataset</span>
<span class=sd>        gene_index (DataFrame): gene/target index</span>
<span class=sd>        lift (LiftOverSpark): LiftOverSpark object</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span> <span class=o>=</span> <span class=n>etl</span>

    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Parsing Thurman 2012 data...&quot;</span><span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Reading data from </span><span class=si>{</span><span class=n>thurman_datafile</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>

    <span class=n>thurman_schema</span> <span class=o>=</span> <span class=n>t</span><span class=o>.</span><span class=n>StructType</span><span class=p>(</span>
        <span class=p>[</span>
            <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;gene_chr&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>StringType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
            <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;gene_start&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
            <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;gene_end&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
            <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;gene_name&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>StringType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
            <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>StringType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
            <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
            <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
            <span class=n>t</span><span class=o>.</span><span class=n>StructField</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>FloatType</span><span class=p>(),</span> <span class=kc>False</span><span class=p>),</span>
        <span class=p>]</span>
    <span class=p>)</span>

    <span class=c1># Process Thurman data in a single step:</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>Thurman_intervals</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>spark</span>
        <span class=c1># Read table according to the schema, then do some modifications:</span>
        <span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>csv</span><span class=p>(</span><span class=n>thurman_datafile</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>header</span><span class=o>=</span><span class=kc>False</span><span class=p>,</span> <span class=n>schema</span><span class=o>=</span><span class=n>thurman_schema</span><span class=p>)</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>),</span> <span class=s2>&quot;chr&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>),</span>
            <span class=s2>&quot;start&quot;</span><span class=p>,</span>
            <span class=s2>&quot;end&quot;</span><span class=p>,</span>
            <span class=s2>&quot;gene_name&quot;</span><span class=p>,</span>
            <span class=s2>&quot;score&quot;</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=c1># Lift over to the GRCh38 build:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>lift</span><span class=o>.</span><span class=n>convert_intervals</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=s2>&quot;chrom&quot;</span><span class=p>,</span> <span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;intervals&quot;</span><span class=p>)</span>
        <span class=c1># Map gene names to gene IDs:</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>gene_index</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;genes&quot;</span><span class=p>),</span>
            <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;intervals.gene_name&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genes.geneSymbol&quot;</span><span class=p>)],</span>
            <span class=n>how</span><span class=o>=</span><span class=s2>&quot;inner&quot;</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=c1># Select relevant columns and add constant columns:</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrom&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped_start&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped_end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>),</span>
            <span class=s2>&quot;geneId&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;resourceScore&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>DATASET_NAME</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datasourceId&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>EXPERIMENT_TYPE</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datatypeId&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>PMID</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;pmid&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=bp>self</span><span class=o>.</span><span class=n>BIO_FEATURE</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;biofeature&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>distinct</span><span class=p>()</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.thurman2012.ParseThurman.get_intervals class="doc doc-heading"> <code class="highlight language-python"><span class=n>get_intervals</span><span class=p>()</span></code> </h3> <div class="doc doc-contents "> <p>Get Thurman intervals.</p> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/thurman2012.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>get_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseThurman</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Get Thurman intervals.&quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=bp>self</span><span class=o>.</span><span class=n>Thurman_intervals</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.thurman2012.ParseThurman.qc_intervals class="doc doc-heading"> <code class="highlight language-python"><span class=n>qc_intervals</span><span class=p>()</span></code> </h3> <div class="doc doc-contents "> <p>Perform QC on the anderson intervals.</p> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/thurman2012.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>qc_intervals</span><span class=p>(</span><span class=bp>self</span><span class=p>:</span> <span class=n>ParseThurman</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Perform QC on the anderson intervals.&quot;&quot;&quot;</span>
    <span class=c1># Get numbers:</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Size of Thurman data: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>Thurman_intervals</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number of unique intervals: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>Thurman_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number genes in the Thurman dataset: </span><span class=si>{</span><span class=bp>self</span><span class=o>.</span><span class=n>Thurman_intervals</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> </div> </div> </div> <div class="doc doc-object doc-module"> <a id=etl.v2g.intervals.Liftover></a> <div class="doc doc-contents first"> <p>LiftOver support.</p> <div class="doc doc-children"> <div class="doc doc-object doc-class"> <h2 id=etl.v2g.intervals.Liftover.LiftOverSpark class="doc doc-heading"> <code>LiftOverSpark</code> </h2> <div class="doc doc-contents "> <p>LiftOver class for mapping genomic coordinates to an other genome build.</p> <p>The input is a Spark DataFrame with a chromosome and position column. This classs can also map regions, if a start and end positions are provided.</p> <p><strong>Logic</strong>:</p> <ul> <li>The mapping is dropped if the mapped chromosome is not on the same as the source.</li> <li>The mapping is dropped if the mapping is ambiguous (more than one mapping is available).</li> <li>If regions are provided, the mapping is dropped if the new region is reversed (mapped_start &gt; mapped_end).</li> <li>If regions are provided, the mapping is dropped if the difference of the lenght of the mapped region and original is larger than a threshold.</li> <li>When lifting over intervals, only unique coordinates are lifted, they joined back to the original dataframe.</li> </ul> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/Liftover.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 15</span>
<span class=normal> 16</span>
<span class=normal> 17</span>
<span class=normal> 18</span>
<span class=normal> 19</span>
<span class=normal> 20</span>
<span class=normal> 21</span>
<span class=normal> 22</span>
<span class=normal> 23</span>
<span class=normal> 24</span>
<span class=normal> 25</span>
<span class=normal> 26</span>
<span class=normal> 27</span>
<span class=normal> 28</span>
<span class=normal> 29</span>
<span class=normal> 30</span>
<span class=normal> 31</span>
<span class=normal> 32</span>
<span class=normal> 33</span>
<span class=normal> 34</span>
<span class=normal> 35</span>
<span class=normal> 36</span>
<span class=normal> 37</span>
<span class=normal> 38</span>
<span class=normal> 39</span>
<span class=normal> 40</span>
<span class=normal> 41</span>
<span class=normal> 42</span>
<span class=normal> 43</span>
<span class=normal> 44</span>
<span class=normal> 45</span>
<span class=normal> 46</span>
<span class=normal> 47</span>
<span class=normal> 48</span>
<span class=normal> 49</span>
<span class=normal> 50</span>
<span class=normal> 51</span>
<span class=normal> 52</span>
<span class=normal> 53</span>
<span class=normal> 54</span>
<span class=normal> 55</span>
<span class=normal> 56</span>
<span class=normal> 57</span>
<span class=normal> 58</span>
<span class=normal> 59</span>
<span class=normal> 60</span>
<span class=normal> 61</span>
<span class=normal> 62</span>
<span class=normal> 63</span>
<span class=normal> 64</span>
<span class=normal> 65</span>
<span class=normal> 66</span>
<span class=normal> 67</span>
<span class=normal> 68</span>
<span class=normal> 69</span>
<span class=normal> 70</span>
<span class=normal> 71</span>
<span class=normal> 72</span>
<span class=normal> 73</span>
<span class=normal> 74</span>
<span class=normal> 75</span>
<span class=normal> 76</span>
<span class=normal> 77</span>
<span class=normal> 78</span>
<span class=normal> 79</span>
<span class=normal> 80</span>
<span class=normal> 81</span>
<span class=normal> 82</span>
<span class=normal> 83</span>
<span class=normal> 84</span>
<span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span>
<span class=normal>125</span>
<span class=normal>126</span>
<span class=normal>127</span>
<span class=normal>128</span>
<span class=normal>129</span>
<span class=normal>130</span>
<span class=normal>131</span>
<span class=normal>132</span>
<span class=normal>133</span>
<span class=normal>134</span>
<span class=normal>135</span>
<span class=normal>136</span>
<span class=normal>137</span>
<span class=normal>138</span>
<span class=normal>139</span>
<span class=normal>140</span>
<span class=normal>141</span>
<span class=normal>142</span>
<span class=normal>143</span>
<span class=normal>144</span>
<span class=normal>145</span>
<span class=normal>146</span>
<span class=normal>147</span>
<span class=normal>148</span>
<span class=normal>149</span>
<span class=normal>150</span>
<span class=normal>151</span>
<span class=normal>152</span>
<span class=normal>153</span>
<span class=normal>154</span>
<span class=normal>155</span>
<span class=normal>156</span>
<span class=normal>157</span>
<span class=normal>158</span>
<span class=normal>159</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>class</span> <span class=nc>LiftOverSpark</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;LiftOver class for mapping genomic coordinates to an other genome build.</span>

<span class=sd>    The input is a Spark DataFrame with a chromosome and position column. This classs can</span>
<span class=sd>    also map regions, if a start and end positions are provided.</span>

<span class=sd>    **Logic**:</span>

<span class=sd>    - The mapping is dropped if the mapped chromosome is not on the same as the source.</span>
<span class=sd>    - The mapping is dropped if the mapping is ambiguous (more than one mapping is available).</span>
<span class=sd>    - If regions are provided, the mapping is dropped if the new region is reversed (mapped_start &gt; mapped_end).</span>
<span class=sd>    - If regions are provided, the mapping is dropped if the difference of the lenght of the mapped region and original is larger than a threshold.</span>
<span class=sd>    - When lifting over intervals, only unique coordinates are lifted, they joined back to the original dataframe.</span>
<span class=sd>    &quot;&quot;&quot;</span>

    <span class=k>def</span> <span class=fm>__init__</span><span class=p>(</span>
        <span class=bp>self</span><span class=p>:</span> <span class=n>LiftOverSpark</span><span class=p>,</span> <span class=n>chain_file</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span> <span class=n>max_difference</span><span class=p>:</span> <span class=nb>int</span> <span class=o>=</span> <span class=mi>100</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Intialise LiftOverSpark object.</span>

<span class=sd>        Args:</span>
<span class=sd>            chain_file (str): Path to the chain file</span>
<span class=sd>            max_difference (int): Maximum difference between the length of the mapped region and the original region. Defaults to 0.</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>chain_file</span> <span class=o>=</span> <span class=n>chain_file</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>max_difference</span> <span class=o>=</span> <span class=n>max_difference</span>

        <span class=c1># Initializing liftover object by opening the chain file:</span>
        <span class=k>if</span> <span class=n>chain_file</span><span class=o>.</span><span class=n>startswith</span><span class=p>(</span><span class=s2>&quot;gs://&quot;</span><span class=p>):</span>
            <span class=k>with</span> <span class=n>gcsfs</span><span class=o>.</span><span class=n>GCSFileSystem</span><span class=p>()</span><span class=o>.</span><span class=n>open</span><span class=p>(</span><span class=n>chain_file</span><span class=p>)</span> <span class=k>as</span> <span class=n>chain_file_object</span><span class=p>:</span>
                <span class=bp>self</span><span class=o>.</span><span class=n>lo</span> <span class=o>=</span> <span class=n>LiftOver</span><span class=p>(</span><span class=n>chain_file_object</span><span class=p>)</span>
        <span class=k>else</span><span class=p>:</span>
            <span class=bp>self</span><span class=o>.</span><span class=n>lo</span> <span class=o>=</span> <span class=n>LiftOver</span><span class=p>(</span><span class=n>chain_file</span><span class=p>)</span>

        <span class=c1># If no maximum difference is provided, set it to 100:</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>max_difference</span> <span class=o>=</span> <span class=n>max_difference</span>

        <span class=c1># UDF to do map genomic coordinates to liftover coordinates:</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>liftover_udf</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>udf</span><span class=p>(</span>
            <span class=k>lambda</span> <span class=n>chrom</span><span class=p>,</span> <span class=n>pos</span><span class=p>:</span> <span class=bp>self</span><span class=o>.</span><span class=n>lo</span><span class=o>.</span><span class=n>convert_coordinate</span><span class=p>(</span><span class=n>chrom</span><span class=p>,</span> <span class=n>pos</span><span class=p>),</span>
            <span class=n>t</span><span class=o>.</span><span class=n>ArrayType</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>ArrayType</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>StringType</span><span class=p>())),</span>
        <span class=p>)</span>

    <span class=k>def</span> <span class=nf>convert_intervals</span><span class=p>(</span>
        <span class=bp>self</span><span class=p>:</span> <span class=n>LiftOverSpark</span><span class=p>,</span>
        <span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
        <span class=n>chrom_col</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
        <span class=n>start_col</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
        <span class=n>end_col</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
        <span class=nb>filter</span><span class=p>:</span> <span class=nb>bool</span> <span class=o>=</span> <span class=kc>True</span><span class=p>,</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Convert genomic intervals to liftover coordinates.</span>

<span class=sd>        Args:</span>
<span class=sd>            df (DataFrame): spark Dataframe with chromosome, start and end columns.</span>
<span class=sd>            chrom_col (str): Name of the chromosome column.</span>
<span class=sd>            start_col (str): Name of the start column.</span>
<span class=sd>            end_col (str): Name of the end column.</span>
<span class=sd>            filter (bool): If True, filter is applied on the mapped data, otherwise return everything. Defaults to True.</span>

<span class=sd>        Returns:</span>
<span class=sd>            DataFrame: Liftovered intervals</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=c1># Lift over start coordinates, changing to 1-based coordinates:</span>
        <span class=n>start_df</span> <span class=o>=</span> <span class=p>(</span>
            <span class=n>df</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=n>start_col</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>start_col</span><span class=p>)</span> <span class=o>+</span> <span class=mi>1</span><span class=p>)</span>
            <span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=n>chrom_col</span><span class=p>,</span> <span class=n>start_col</span><span class=p>)</span>
            <span class=o>.</span><span class=n>distinct</span><span class=p>()</span>
        <span class=p>)</span>
        <span class=n>start_df</span> <span class=o>=</span> <span class=bp>self</span><span class=o>.</span><span class=n>convert_coordinates</span><span class=p>(</span>
            <span class=n>start_df</span><span class=p>,</span> <span class=n>chrom_col</span><span class=p>,</span> <span class=n>start_col</span>
        <span class=p>)</span><span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_pos&quot;</span><span class=p>,</span> <span class=s2>&quot;mapped_&quot;</span> <span class=o>+</span> <span class=n>start_col</span><span class=p>)</span>

        <span class=c1># Lift over end coordinates:</span>
        <span class=n>end_df</span> <span class=o>=</span> <span class=n>df</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=n>chrom_col</span><span class=p>,</span> <span class=n>end_col</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span>
        <span class=n>end_df</span> <span class=o>=</span> <span class=bp>self</span><span class=o>.</span><span class=n>convert_coordinates</span><span class=p>(</span><span class=n>end_df</span><span class=p>,</span> <span class=n>chrom_col</span><span class=p>,</span> <span class=n>end_col</span><span class=p>)</span><span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span>
            <span class=s2>&quot;mapped_pos&quot;</span><span class=p>,</span> <span class=s2>&quot;mapped_&quot;</span> <span class=o>+</span> <span class=n>end_col</span>
        <span class=p>)</span>

        <span class=c1># Join dataframe with mappings (we have to account for the +1 position shift of the start coordinates):</span>
        <span class=n>mapped_df</span> <span class=o>=</span> <span class=n>df</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>start_df</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=n>start_col</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>start_col</span><span class=p>)</span> <span class=o>-</span> <span class=mi>1</span><span class=p>),</span>
            <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=n>chrom_col</span><span class=p>,</span> <span class=n>start_col</span><span class=p>],</span>
            <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>,</span>
        <span class=p>)</span><span class=o>.</span><span class=n>join</span><span class=p>(</span><span class=n>end_df</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=n>chrom_col</span><span class=p>,</span> <span class=n>end_col</span><span class=p>],</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>)</span>

        <span class=c1># The filter option allows to get all the data and filter it afterwards.</span>
        <span class=k>if</span> <span class=nb>filter</span><span class=p>:</span>
            <span class=k>return</span> <span class=p>(</span>
                <span class=n>mapped_df</span>
                <span class=c1># Select only rows where the start is smaller than the end:</span>
                <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
                    <span class=c1># Drop rows with no mappings:</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;mapped_</span><span class=si>{</span><span class=n>start_col</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>()</span>
                    <span class=o>&amp;</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;mapped_</span><span class=si>{</span><span class=n>end_col</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>()</span>
                    <span class=c1># Drop rows where the start is larger than the end:</span>
                    <span class=o>&amp;</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;mapped_</span><span class=si>{</span><span class=n>end_col</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span> <span class=o>&gt;=</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;mapped_</span><span class=si>{</span><span class=n>start_col</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>))</span>
                    <span class=c1># Drop rows where the difference of the length of the regions are larger than the threshold:</span>
                    <span class=o>&amp;</span> <span class=p>(</span>
                        <span class=n>f</span><span class=o>.</span><span class=n>abs</span><span class=p>(</span>
                            <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>end_col</span><span class=p>)</span> <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>start_col</span><span class=p>))</span>
                            <span class=o>-</span> <span class=p>(</span>
                                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;mapped_</span><span class=si>{</span><span class=n>end_col</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
                                <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;mapped_</span><span class=si>{</span><span class=n>start_col</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
                            <span class=p>)</span>
                        <span class=p>)</span>
                        <span class=o>&lt;=</span> <span class=bp>self</span><span class=o>.</span><span class=n>max_difference</span>
                    <span class=p>)</span>
                <span class=p>)</span><span class=o>.</span><span class=n>persist</span><span class=p>()</span>
            <span class=p>)</span>
        <span class=k>else</span><span class=p>:</span>
            <span class=k>return</span> <span class=n>mapped_df</span><span class=o>.</span><span class=n>persist</span><span class=p>()</span>

    <span class=k>def</span> <span class=nf>convert_coordinates</span><span class=p>(</span>
        <span class=bp>self</span><span class=p>:</span> <span class=n>LiftOverSpark</span><span class=p>,</span> <span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span> <span class=n>chrom_name</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span> <span class=n>pos_name</span><span class=p>:</span> <span class=nb>str</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Converts genomic coordinates to coordinates on an other build.</span>

<span class=sd>        Args:</span>
<span class=sd>            df (DataFrame): Spark Dataframe with chromosome and position columns.</span>
<span class=sd>            chrom_name (str): Name of the chromosome column.</span>
<span class=sd>            pos_name (str): Name of the position column.</span>

<span class=sd>        Returns:</span>
<span class=sd>            DataFrame: Spark Dataframe with the mapped position column.</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=n>mapped</span> <span class=o>=</span> <span class=p>(</span>
            <span class=n>df</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
                <span class=s2>&quot;mapped&quot;</span><span class=p>,</span> <span class=bp>self</span><span class=o>.</span><span class=n>liftover_udf</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>chrom_name</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>pos_name</span><span class=p>))</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>filter</span><span class=p>((</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>())</span> <span class=o>&amp;</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>size</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped&quot;</span><span class=p>))</span> <span class=o>==</span> <span class=mi>1</span><span class=p>))</span>
            <span class=c1># Extracting mapped corrdinates:</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;mapped_&quot;</span> <span class=o>+</span> <span class=n>chrom_name</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped&quot;</span><span class=p>)[</span><span class=mi>0</span><span class=p>][</span><span class=mi>0</span><span class=p>])</span>
            <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;mapped_&quot;</span> <span class=o>+</span> <span class=n>pos_name</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped&quot;</span><span class=p>)[</span><span class=mi>0</span><span class=p>][</span><span class=mi>1</span><span class=p>])</span>
            <span class=c1># Drop rows that mapped to the other chromosomes:</span>
            <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped_&quot;</span> <span class=o>+</span> <span class=n>chrom_name</span><span class=p>)</span>
                <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>concat</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;chr&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>chrom_name</span><span class=p>))</span>
            <span class=p>)</span>
            <span class=c1># Dropping unused columns:</span>
            <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;mapped&quot;</span><span class=p>,</span> <span class=s2>&quot;mapped_&quot;</span> <span class=o>+</span> <span class=n>chrom_name</span><span class=p>)</span>
            <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
        <span class=p>)</span>

        <span class=k>return</span> <span class=n>mapped</span>
</code></pre></div></td></tr></table></div> </details> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.Liftover.LiftOverSpark.__init__ class="doc doc-heading"> <code class="highlight language-python"><span class=fm>__init__</span><span class=p>(</span><span class=n>chain_file</span><span class=p>,</span> <span class=n>max_difference</span><span class=o>=</span><span class=mi>100</span><span class=p>)</span></code> </h3> <div class="doc doc-contents "> <p>Intialise LiftOverSpark object.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>chain_file</code></td> <td> <code>str</code> </td> <td><p>Path to the chain file</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>max_difference</code></td> <td> <code>int</code> </td> <td><p>Maximum difference between the length of the mapped region and the original region. Defaults to 0.</p></td> <td> <code>100</code> </td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/Liftover.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>30</span>
<span class=normal>31</span>
<span class=normal>32</span>
<span class=normal>33</span>
<span class=normal>34</span>
<span class=normal>35</span>
<span class=normal>36</span>
<span class=normal>37</span>
<span class=normal>38</span>
<span class=normal>39</span>
<span class=normal>40</span>
<span class=normal>41</span>
<span class=normal>42</span>
<span class=normal>43</span>
<span class=normal>44</span>
<span class=normal>45</span>
<span class=normal>46</span>
<span class=normal>47</span>
<span class=normal>48</span>
<span class=normal>49</span>
<span class=normal>50</span>
<span class=normal>51</span>
<span class=normal>52</span>
<span class=normal>53</span>
<span class=normal>54</span>
<span class=normal>55</span>
<span class=normal>56</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=fm>__init__</span><span class=p>(</span>
    <span class=bp>self</span><span class=p>:</span> <span class=n>LiftOverSpark</span><span class=p>,</span> <span class=n>chain_file</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span> <span class=n>max_difference</span><span class=p>:</span> <span class=nb>int</span> <span class=o>=</span> <span class=mi>100</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=kc>None</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Intialise LiftOverSpark object.</span>

<span class=sd>    Args:</span>
<span class=sd>        chain_file (str): Path to the chain file</span>
<span class=sd>        max_difference (int): Maximum difference between the length of the mapped region and the original region. Defaults to 0.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>chain_file</span> <span class=o>=</span> <span class=n>chain_file</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>max_difference</span> <span class=o>=</span> <span class=n>max_difference</span>

    <span class=c1># Initializing liftover object by opening the chain file:</span>
    <span class=k>if</span> <span class=n>chain_file</span><span class=o>.</span><span class=n>startswith</span><span class=p>(</span><span class=s2>&quot;gs://&quot;</span><span class=p>):</span>
        <span class=k>with</span> <span class=n>gcsfs</span><span class=o>.</span><span class=n>GCSFileSystem</span><span class=p>()</span><span class=o>.</span><span class=n>open</span><span class=p>(</span><span class=n>chain_file</span><span class=p>)</span> <span class=k>as</span> <span class=n>chain_file_object</span><span class=p>:</span>
            <span class=bp>self</span><span class=o>.</span><span class=n>lo</span> <span class=o>=</span> <span class=n>LiftOver</span><span class=p>(</span><span class=n>chain_file_object</span><span class=p>)</span>
    <span class=k>else</span><span class=p>:</span>
        <span class=bp>self</span><span class=o>.</span><span class=n>lo</span> <span class=o>=</span> <span class=n>LiftOver</span><span class=p>(</span><span class=n>chain_file</span><span class=p>)</span>

    <span class=c1># If no maximum difference is provided, set it to 100:</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>max_difference</span> <span class=o>=</span> <span class=n>max_difference</span>

    <span class=c1># UDF to do map genomic coordinates to liftover coordinates:</span>
    <span class=bp>self</span><span class=o>.</span><span class=n>liftover_udf</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>udf</span><span class=p>(</span>
        <span class=k>lambda</span> <span class=n>chrom</span><span class=p>,</span> <span class=n>pos</span><span class=p>:</span> <span class=bp>self</span><span class=o>.</span><span class=n>lo</span><span class=o>.</span><span class=n>convert_coordinate</span><span class=p>(</span><span class=n>chrom</span><span class=p>,</span> <span class=n>pos</span><span class=p>),</span>
        <span class=n>t</span><span class=o>.</span><span class=n>ArrayType</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>ArrayType</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>StringType</span><span class=p>())),</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.Liftover.LiftOverSpark.convert_coordinates class="doc doc-heading"> <code class="highlight language-python"><span class=n>convert_coordinates</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=n>chrom_name</span><span class=p>,</span> <span class=n>pos_name</span><span class=p>)</span></code> </h3> <div class="doc doc-contents "> <p>Converts genomic coordinates to coordinates on an other build.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Spark Dataframe with chromosome and position columns.</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>chrom_name</code></td> <td> <code>str</code> </td> <td><p>Name of the chromosome column.</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>pos_name</code></td> <td> <code>str</code> </td> <td><p>Name of the position column.</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Spark Dataframe with the mapped position column.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/Liftover.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>128</span>
<span class=normal>129</span>
<span class=normal>130</span>
<span class=normal>131</span>
<span class=normal>132</span>
<span class=normal>133</span>
<span class=normal>134</span>
<span class=normal>135</span>
<span class=normal>136</span>
<span class=normal>137</span>
<span class=normal>138</span>
<span class=normal>139</span>
<span class=normal>140</span>
<span class=normal>141</span>
<span class=normal>142</span>
<span class=normal>143</span>
<span class=normal>144</span>
<span class=normal>145</span>
<span class=normal>146</span>
<span class=normal>147</span>
<span class=normal>148</span>
<span class=normal>149</span>
<span class=normal>150</span>
<span class=normal>151</span>
<span class=normal>152</span>
<span class=normal>153</span>
<span class=normal>154</span>
<span class=normal>155</span>
<span class=normal>156</span>
<span class=normal>157</span>
<span class=normal>158</span>
<span class=normal>159</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>convert_coordinates</span><span class=p>(</span>
    <span class=bp>self</span><span class=p>:</span> <span class=n>LiftOverSpark</span><span class=p>,</span> <span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span> <span class=n>chrom_name</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span> <span class=n>pos_name</span><span class=p>:</span> <span class=nb>str</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Converts genomic coordinates to coordinates on an other build.</span>

<span class=sd>    Args:</span>
<span class=sd>        df (DataFrame): Spark Dataframe with chromosome and position columns.</span>
<span class=sd>        chrom_name (str): Name of the chromosome column.</span>
<span class=sd>        pos_name (str): Name of the position column.</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Spark Dataframe with the mapped position column.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>mapped</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>df</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;mapped&quot;</span><span class=p>,</span> <span class=bp>self</span><span class=o>.</span><span class=n>liftover_udf</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>chrom_name</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>pos_name</span><span class=p>))</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>((</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>())</span> <span class=o>&amp;</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>size</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped&quot;</span><span class=p>))</span> <span class=o>==</span> <span class=mi>1</span><span class=p>))</span>
        <span class=c1># Extracting mapped corrdinates:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;mapped_&quot;</span> <span class=o>+</span> <span class=n>chrom_name</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped&quot;</span><span class=p>)[</span><span class=mi>0</span><span class=p>][</span><span class=mi>0</span><span class=p>])</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;mapped_&quot;</span> <span class=o>+</span> <span class=n>pos_name</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped&quot;</span><span class=p>)[</span><span class=mi>0</span><span class=p>][</span><span class=mi>1</span><span class=p>])</span>
        <span class=c1># Drop rows that mapped to the other chromosomes:</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;mapped_&quot;</span> <span class=o>+</span> <span class=n>chrom_name</span><span class=p>)</span>
            <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>concat</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;chr&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>chrom_name</span><span class=p>))</span>
        <span class=p>)</span>
        <span class=c1># Dropping unused columns:</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;mapped&quot;</span><span class=p>,</span> <span class=s2>&quot;mapped_&quot;</span> <span class=o>+</span> <span class=n>chrom_name</span><span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>

    <span class=k>return</span> <span class=n>mapped</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h3 id=etl.v2g.intervals.Liftover.LiftOverSpark.convert_intervals class="doc doc-heading"> <code class="highlight language-python"><span class=n>convert_intervals</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=n>chrom_col</span><span class=p>,</span> <span class=n>start_col</span><span class=p>,</span> <span class=n>end_col</span><span class=p>,</span> <span class=nb>filter</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span></code> </h3> <div class="doc doc-contents "> <p>Convert genomic intervals to liftover coordinates.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>spark Dataframe with chromosome, start and end columns.</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>chrom_col</code></td> <td> <code>str</code> </td> <td><p>Name of the chromosome column.</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>start_col</code></td> <td> <code>str</code> </td> <td><p>Name of the start column.</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>end_col</code></td> <td> <code>str</code> </td> <td><p>Name of the end column.</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>filter</code></td> <td> <code>bool</code> </td> <td><p>If True, filter is applied on the mapped data, otherwise return everything. Defaults to True.</p></td> <td> <code>True</code> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Liftovered intervals</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/Liftover.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 58</span>
<span class=normal> 59</span>
<span class=normal> 60</span>
<span class=normal> 61</span>
<span class=normal> 62</span>
<span class=normal> 63</span>
<span class=normal> 64</span>
<span class=normal> 65</span>
<span class=normal> 66</span>
<span class=normal> 67</span>
<span class=normal> 68</span>
<span class=normal> 69</span>
<span class=normal> 70</span>
<span class=normal> 71</span>
<span class=normal> 72</span>
<span class=normal> 73</span>
<span class=normal> 74</span>
<span class=normal> 75</span>
<span class=normal> 76</span>
<span class=normal> 77</span>
<span class=normal> 78</span>
<span class=normal> 79</span>
<span class=normal> 80</span>
<span class=normal> 81</span>
<span class=normal> 82</span>
<span class=normal> 83</span>
<span class=normal> 84</span>
<span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span>
<span class=normal>125</span>
<span class=normal>126</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>convert_intervals</span><span class=p>(</span>
    <span class=bp>self</span><span class=p>:</span> <span class=n>LiftOverSpark</span><span class=p>,</span>
    <span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
    <span class=n>chrom_col</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
    <span class=n>start_col</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
    <span class=n>end_col</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
    <span class=nb>filter</span><span class=p>:</span> <span class=nb>bool</span> <span class=o>=</span> <span class=kc>True</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Convert genomic intervals to liftover coordinates.</span>

<span class=sd>    Args:</span>
<span class=sd>        df (DataFrame): spark Dataframe with chromosome, start and end columns.</span>
<span class=sd>        chrom_col (str): Name of the chromosome column.</span>
<span class=sd>        start_col (str): Name of the start column.</span>
<span class=sd>        end_col (str): Name of the end column.</span>
<span class=sd>        filter (bool): If True, filter is applied on the mapped data, otherwise return everything. Defaults to True.</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Liftovered intervals</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=c1># Lift over start coordinates, changing to 1-based coordinates:</span>
    <span class=n>start_df</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>df</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=n>start_col</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>start_col</span><span class=p>)</span> <span class=o>+</span> <span class=mi>1</span><span class=p>)</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=n>chrom_col</span><span class=p>,</span> <span class=n>start_col</span><span class=p>)</span>
        <span class=o>.</span><span class=n>distinct</span><span class=p>()</span>
    <span class=p>)</span>
    <span class=n>start_df</span> <span class=o>=</span> <span class=bp>self</span><span class=o>.</span><span class=n>convert_coordinates</span><span class=p>(</span>
        <span class=n>start_df</span><span class=p>,</span> <span class=n>chrom_col</span><span class=p>,</span> <span class=n>start_col</span>
    <span class=p>)</span><span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;mapped_pos&quot;</span><span class=p>,</span> <span class=s2>&quot;mapped_&quot;</span> <span class=o>+</span> <span class=n>start_col</span><span class=p>)</span>

    <span class=c1># Lift over end coordinates:</span>
    <span class=n>end_df</span> <span class=o>=</span> <span class=n>df</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=n>chrom_col</span><span class=p>,</span> <span class=n>end_col</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span>
    <span class=n>end_df</span> <span class=o>=</span> <span class=bp>self</span><span class=o>.</span><span class=n>convert_coordinates</span><span class=p>(</span><span class=n>end_df</span><span class=p>,</span> <span class=n>chrom_col</span><span class=p>,</span> <span class=n>end_col</span><span class=p>)</span><span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span>
        <span class=s2>&quot;mapped_pos&quot;</span><span class=p>,</span> <span class=s2>&quot;mapped_&quot;</span> <span class=o>+</span> <span class=n>end_col</span>
    <span class=p>)</span>

    <span class=c1># Join dataframe with mappings (we have to account for the +1 position shift of the start coordinates):</span>
    <span class=n>mapped_df</span> <span class=o>=</span> <span class=n>df</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
        <span class=n>start_df</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=n>start_col</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>start_col</span><span class=p>)</span> <span class=o>-</span> <span class=mi>1</span><span class=p>),</span>
        <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=n>chrom_col</span><span class=p>,</span> <span class=n>start_col</span><span class=p>],</span>
        <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>,</span>
    <span class=p>)</span><span class=o>.</span><span class=n>join</span><span class=p>(</span><span class=n>end_df</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=n>chrom_col</span><span class=p>,</span> <span class=n>end_col</span><span class=p>],</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>)</span>

    <span class=c1># The filter option allows to get all the data and filter it afterwards.</span>
    <span class=k>if</span> <span class=nb>filter</span><span class=p>:</span>
        <span class=k>return</span> <span class=p>(</span>
            <span class=n>mapped_df</span>
            <span class=c1># Select only rows where the start is smaller than the end:</span>
            <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
                <span class=c1># Drop rows with no mappings:</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;mapped_</span><span class=si>{</span><span class=n>start_col</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>()</span>
                <span class=o>&amp;</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;mapped_</span><span class=si>{</span><span class=n>end_col</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>()</span>
                <span class=c1># Drop rows where the start is larger than the end:</span>
                <span class=o>&amp;</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;mapped_</span><span class=si>{</span><span class=n>end_col</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span> <span class=o>&gt;=</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;mapped_</span><span class=si>{</span><span class=n>start_col</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>))</span>
                <span class=c1># Drop rows where the difference of the length of the regions are larger than the threshold:</span>
                <span class=o>&amp;</span> <span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>abs</span><span class=p>(</span>
                        <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>end_col</span><span class=p>)</span> <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>start_col</span><span class=p>))</span>
                        <span class=o>-</span> <span class=p>(</span>
                            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;mapped_</span><span class=si>{</span><span class=n>end_col</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
                            <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;mapped_</span><span class=si>{</span><span class=n>start_col</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
                        <span class=p>)</span>
                    <span class=p>)</span>
                    <span class=o>&lt;=</span> <span class=bp>self</span><span class=o>.</span><span class=n>max_difference</span>
                <span class=p>)</span>
            <span class=p>)</span><span class=o>.</span><span class=n>persist</span><span class=p>()</span>
        <span class=p>)</span>
    <span class=k>else</span><span class=p>:</span>
        <span class=k>return</span> <span class=n>mapped_df</span><span class=o>.</span><span class=n>persist</span><span class=p>()</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> </div> </div> </div> <div class="doc doc-object doc-module"> <a id=etl.v2g.intervals.helpers></a> <div class="doc doc-contents first"> <p>Interval helper functions.</p> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h2 id=etl.v2g.intervals.helpers.get_variants_in_interval class="doc doc-heading"> <code class="highlight language-python"><span class=n>get_variants_in_interval</span><span class=p>(</span><span class=n>interval_df</span><span class=p>,</span> <span class=n>variants_df</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Explodes the interval dataset to find all variants in the region.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>interval_df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Interval dataset</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>variants_df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>DataFrame with a set of variants of interest with the columns "variantId", "chromosome" and "position"</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>V2G evidence based on all the variants found in the intervals</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/helpers.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>42</span>
<span class=normal>43</span>
<span class=normal>44</span>
<span class=normal>45</span>
<span class=normal>46</span>
<span class=normal>47</span>
<span class=normal>48</span>
<span class=normal>49</span>
<span class=normal>50</span>
<span class=normal>51</span>
<span class=normal>52</span>
<span class=normal>53</span>
<span class=normal>54</span>
<span class=normal>55</span>
<span class=normal>56</span>
<span class=normal>57</span>
<span class=normal>58</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>get_variants_in_interval</span><span class=p>(</span>
    <span class=n>interval_df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span> <span class=n>variants_df</span><span class=p>:</span> <span class=n>DataFrame</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Explodes the interval dataset to find all variants in the region.</span>

<span class=sd>    Args:</span>
<span class=sd>        interval_df (DataFrame): Interval dataset</span>
<span class=sd>        variants_df (DataFrame): DataFrame with a set of variants of interest with the columns &quot;variantId&quot;, &quot;chromosome&quot; and &quot;position&quot;</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: V2G evidence based on all the variants found in the intervals</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=n>interval_df</span><span class=o>.</span><span class=n>join</span><span class=p>(</span><span class=n>variants_df</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=s2>&quot;chromosome&quot;</span><span class=p>,</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;inner&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;position&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>between</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;end&quot;</span><span class=p>)))</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;start&quot;</span><span class=p>,</span> <span class=s2>&quot;end&quot;</span><span class=p>)</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.v2g.intervals.helpers.prepare_gene_interval_lut class="doc doc-heading"> <code class="highlight language-python"><span class=n>prepare_gene_interval_lut</span><span class=p>(</span><span class=n>gene_index</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Gene symbol lookup table.</p> <p>Pre-processess gene/target dataset to create lookup table of gene symbols, including obsoleted gene symbols.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>gene_index</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>gene/target DataFrame</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Gene LUT for symbol mapping</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/intervals/helpers.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>14</span>
<span class=normal>15</span>
<span class=normal>16</span>
<span class=normal>17</span>
<span class=normal>18</span>
<span class=normal>19</span>
<span class=normal>20</span>
<span class=normal>21</span>
<span class=normal>22</span>
<span class=normal>23</span>
<span class=normal>24</span>
<span class=normal>25</span>
<span class=normal>26</span>
<span class=normal>27</span>
<span class=normal>28</span>
<span class=normal>29</span>
<span class=normal>30</span>
<span class=normal>31</span>
<span class=normal>32</span>
<span class=normal>33</span>
<span class=normal>34</span>
<span class=normal>35</span>
<span class=normal>36</span>
<span class=normal>37</span>
<span class=normal>38</span>
<span class=normal>39</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>prepare_gene_interval_lut</span><span class=p>(</span><span class=n>gene_index</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Gene symbol lookup table.</span>

<span class=sd>    Pre-processess gene/target dataset to create lookup table of gene symbols, including</span>
<span class=sd>    obsoleted gene symbols.</span>

<span class=sd>    Args:</span>
<span class=sd>        gene_index (DataFrame): gene/target DataFrame</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Gene LUT for symbol mapping</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=n>gene_index</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;id&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>),</span>
        <span class=s2>&quot;biotype&quot;</span><span class=p>,</span>
        <span class=n>f</span><span class=o>.</span><span class=n>explode</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>array_union</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>array</span><span class=p>(</span><span class=s2>&quot;approvedSymbol&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;obsoleteSymbols.label&quot;</span><span class=p>))</span>
        <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;geneSymbol&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genomicLocation.chromosome&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>),</span>
        <span class=n>get_gene_tss</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genomicLocation.strand&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genomicLocation.start&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;genomicLocation.end&quot;</span><span class=p>),</span>
        <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;tss&quot;</span><span class=p>),</span>
        <span class=s2>&quot;genomicLocation&quot;</span><span class=p>,</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div><h3 id=functional-predictions>Functional predictions</h3> <div class="doc doc-object doc-module"> <a id=etl.v2g.functional_predictions.vep></a> <div class="doc doc-contents first"> <p>The variant annotation dataset contains information about the impact of a variant on a transcript or protein. These can be mapped to genes allowing us to establish significant relationships between variants and genes.</p> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h2 id=etl.v2g.functional_predictions.vep.get_plof_flag class="doc doc-heading"> <code class="highlight language-python"><span class=n>get_plof_flag</span><span class=p>(</span><span class=n>variants_df</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Creates a dataset with variant to gene assignments with a flag indicating if the variant is predicted to be a loss-of-function variant by the LOFTEE algorithm.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>variants_df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Dataframe with two columns: "id" and "transcriptConsequence"</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>variant to gene assignments from the LOFTEE algorithm</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/functional_predictions/vep.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>176</span>
<span class=normal>177</span>
<span class=normal>178</span>
<span class=normal>179</span>
<span class=normal>180</span>
<span class=normal>181</span>
<span class=normal>182</span>
<span class=normal>183</span>
<span class=normal>184</span>
<span class=normal>185</span>
<span class=normal>186</span>
<span class=normal>187</span>
<span class=normal>188</span>
<span class=normal>189</span>
<span class=normal>190</span>
<span class=normal>191</span>
<span class=normal>192</span>
<span class=normal>193</span>
<span class=normal>194</span>
<span class=normal>195</span>
<span class=normal>196</span>
<span class=normal>197</span>
<span class=normal>198</span>
<span class=normal>199</span>
<span class=normal>200</span>
<span class=normal>201</span>
<span class=normal>202</span>
<span class=normal>203</span>
<span class=normal>204</span>
<span class=normal>205</span>
<span class=normal>206</span>
<span class=normal>207</span>
<span class=normal>208</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>get_plof_flag</span><span class=p>(</span><span class=n>variants_df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Creates a dataset with variant to gene assignments with a flag indicating if the variant is predicted to be a loss-of-function variant by the LOFTEE algorithm.</span>

<span class=sd>    Args:</span>
<span class=sd>        variants_df (DataFrame): Dataframe with two columns: &quot;id&quot; and &quot;transcriptConsequence&quot;</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: variant to gene assignments from the LOFTEE algorithm</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=n>variants_df</span><span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;transcriptConsequence.lof&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>())</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;isHighQualityPlof&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;transcriptConsequence.lof&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=s2>&quot;HC&quot;</span><span class=p>,</span> <span class=kc>True</span><span class=p>)</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;transcriptConsequence.lof&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=s2>&quot;LC&quot;</span><span class=p>,</span> <span class=kc>False</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;score&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;isHighQualityPlof&quot;</span><span class=p>),</span> <span class=mf>1.0</span><span class=p>)</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=o>~</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;isHighQualityPlof&quot;</span><span class=p>),</span> <span class=mi>0</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;variantId&quot;</span><span class=p>,</span>
            <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;transcriptConsequence.gene_id&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>),</span>
            <span class=s2>&quot;isHighQualityPlof&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;vep&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datatypeId&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;loftee&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datasourceId&quot;</span><span class=p>),</span>
        <span class=p>)</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.v2g.functional_predictions.vep.get_polyphen_score class="doc doc-heading"> <code class="highlight language-python"><span class=n>get_polyphen_score</span><span class=p>(</span><span class=n>variants_df</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Creates a dataset with variant to gene assignments with a PolyPhen's predicted score on the transcript.</p> <p>Polyphen informs about the probability that a substitution is damaging.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>variants_df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Dataframe with two columns: "id" and "transcriptConsequence"</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>variant to gene assignments with their polyphen scores</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/functional_predictions/vep.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>124</span>
<span class=normal>125</span>
<span class=normal>126</span>
<span class=normal>127</span>
<span class=normal>128</span>
<span class=normal>129</span>
<span class=normal>130</span>
<span class=normal>131</span>
<span class=normal>132</span>
<span class=normal>133</span>
<span class=normal>134</span>
<span class=normal>135</span>
<span class=normal>136</span>
<span class=normal>137</span>
<span class=normal>138</span>
<span class=normal>139</span>
<span class=normal>140</span>
<span class=normal>141</span>
<span class=normal>142</span>
<span class=normal>143</span>
<span class=normal>144</span>
<span class=normal>145</span>
<span class=normal>146</span>
<span class=normal>147</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>get_polyphen_score</span><span class=p>(</span>
    <span class=n>variants_df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Creates a dataset with variant to gene assignments with a PolyPhen&#39;s predicted score on the transcript.</span>

<span class=sd>    Polyphen informs about the probability that a substitution is damaging.</span>

<span class=sd>    Args:</span>
<span class=sd>        variants_df (DataFrame): Dataframe with two columns: &quot;id&quot; and &quot;transcriptConsequence&quot;</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: variant to gene assignments with their polyphen scores</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=n>variants_df</span><span class=o>.</span><span class=n>filter</span><span class=p>(</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;transcriptConsequence.polyphen_score&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>()</span>
    <span class=p>)</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
        <span class=s2>&quot;variantId&quot;</span><span class=p>,</span>
        <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;transcriptConsequence.gene_id&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;transcriptConsequence.polyphen_score&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;transcriptConsequence.polyphen_prediction&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;label&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;vep&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datatypeId&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;polyphen&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datasourceId&quot;</span><span class=p>),</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.v2g.functional_predictions.vep.get_sift_score class="doc doc-heading"> <code class="highlight language-python"><span class=n>get_sift_score</span><span class=p>(</span><span class=n>variants_df</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Creates a dataset with variant to gene assignments with a SIFT's predicted score on the transcript.</p> <p>SIFT informs about the probability that a substitution is tolerated so scores nearer zero are more likely to be deleterious.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>variants_df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Dataframe with two columns: "id" and "transcriptConsequence"</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>variant to gene assignments with their SIFT scores</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/functional_predictions/vep.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>150</span>
<span class=normal>151</span>
<span class=normal>152</span>
<span class=normal>153</span>
<span class=normal>154</span>
<span class=normal>155</span>
<span class=normal>156</span>
<span class=normal>157</span>
<span class=normal>158</span>
<span class=normal>159</span>
<span class=normal>160</span>
<span class=normal>161</span>
<span class=normal>162</span>
<span class=normal>163</span>
<span class=normal>164</span>
<span class=normal>165</span>
<span class=normal>166</span>
<span class=normal>167</span>
<span class=normal>168</span>
<span class=normal>169</span>
<span class=normal>170</span>
<span class=normal>171</span>
<span class=normal>172</span>
<span class=normal>173</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>get_sift_score</span><span class=p>(</span>
    <span class=n>variants_df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Creates a dataset with variant to gene assignments with a SIFT&#39;s predicted score on the transcript.</span>

<span class=sd>    SIFT informs about the probability that a substitution is tolerated so scores nearer zero are more likely to be deleterious.</span>

<span class=sd>    Args:</span>
<span class=sd>        variants_df (DataFrame): Dataframe with two columns: &quot;id&quot; and &quot;transcriptConsequence&quot;</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: variant to gene assignments with their SIFT scores</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=n>variants_df</span><span class=o>.</span><span class=n>filter</span><span class=p>(</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;transcriptConsequence.sift_score&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>()</span>
    <span class=p>)</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
        <span class=s2>&quot;variantId&quot;</span><span class=p>,</span>
        <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;transcriptConsequence.gene_id&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>expr</span><span class=p>(</span><span class=s2>&quot;1 - transcriptConsequence.sift_score&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;transcriptConsequence.sift_prediction&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;label&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;vep&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datatypeId&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;sift&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datasourceId&quot;</span><span class=p>),</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.v2g.functional_predictions.vep.get_variant_consequences class="doc doc-heading"> <code class="highlight language-python"><span class=n>get_variant_consequences</span><span class=p>(</span><span class=n>variants_df</span><span class=p>,</span> <span class=n>variant_consequence_lut</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Creates a dataset with variant to gene assignments based on VEP's predicted consequence on the transcript.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>variants_df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Dataframe with two columns: "id" and "transcriptConsequence"</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>variant_consequence_lut</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Dataframe with the variant consequences sorted by severity</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>High and medium severity variant to gene assignments</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/functional_predictions/vep.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>get_variant_consequences</span><span class=p>(</span>
    <span class=n>variants_df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
    <span class=n>variant_consequence_lut</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Creates a dataset with variant to gene assignments based on VEP&#39;s predicted consequence on the transcript.</span>

<span class=sd>    Args:</span>
<span class=sd>        variants_df (DataFrame): Dataframe with two columns: &quot;id&quot; and &quot;transcriptConsequence&quot;</span>
<span class=sd>        variant_consequence_lut (DataFrame): Dataframe with the variant consequences sorted by severity</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: High and medium severity variant to gene assignments</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=n>variants_df</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;variantId&quot;</span><span class=p>,</span>
            <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;transcriptConsequence.gene_id&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;geneId&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>explode</span><span class=p>(</span><span class=s2>&quot;transcriptConsequence.consequence_terms&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;label&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;vep&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datatypeId&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;variantConsequence&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;datasourceId&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># A variant can have multiple predicted consequences on a transcript, the most severe one is selected</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>broadcast</span><span class=p>(</span><span class=n>variant_consequence_lut</span><span class=p>),</span>
            <span class=n>on</span><span class=o>=</span><span class=s2>&quot;label&quot;</span><span class=p>,</span>
            <span class=n>how</span><span class=o>=</span><span class=s2>&quot;inner&quot;</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>)</span> <span class=o>!=</span> <span class=mi>0</span><span class=p>)</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span>
            <span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>get_record_with_maximum_value</span><span class=p>(</span>
                <span class=n>df</span><span class=p>,</span> <span class=p>[</span><span class=s2>&quot;variantId&quot;</span><span class=p>,</span> <span class=s2>&quot;geneId&quot;</span><span class=p>],</span> <span class=s2>&quot;score&quot;</span>
            <span class=p>)</span>
        <span class=p>)</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.v2g.functional_predictions.vep.main class="doc doc-heading"> <code class="highlight language-python"><span class=n>main</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>variant_index</span><span class=p>,</span> <span class=n>variant_annotation</span><span class=p>,</span> <span class=n>variant_consequence_lut_path</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Extracts variant to gene assignments for the variants included in the index and the features predicted by VEP.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>ETL session,</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>variant_index</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>DataFrame with the OTG variant index</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>variant_annotation</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Dataframe with the annotated variants</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>variant_consequence_lut_path</code></td> <td> <code>str</code> </td> <td><p>The path to the LUT between the functional consequences and their assigned V2G score</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code>tuple[<span title="pyspark.sql.DataFrame">DataFrame</span>, ...]</code> </td> <td><p>variant to gene assignments from VEP</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/functional_predictions/vep.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>16</span>
<span class=normal>17</span>
<span class=normal>18</span>
<span class=normal>19</span>
<span class=normal>20</span>
<span class=normal>21</span>
<span class=normal>22</span>
<span class=normal>23</span>
<span class=normal>24</span>
<span class=normal>25</span>
<span class=normal>26</span>
<span class=normal>27</span>
<span class=normal>28</span>
<span class=normal>29</span>
<span class=normal>30</span>
<span class=normal>31</span>
<span class=normal>32</span>
<span class=normal>33</span>
<span class=normal>34</span>
<span class=normal>35</span>
<span class=normal>36</span>
<span class=normal>37</span>
<span class=normal>38</span>
<span class=normal>39</span>
<span class=normal>40</span>
<span class=normal>41</span>
<span class=normal>42</span>
<span class=normal>43</span>
<span class=normal>44</span>
<span class=normal>45</span>
<span class=normal>46</span>
<span class=normal>47</span>
<span class=normal>48</span>
<span class=normal>49</span>
<span class=normal>50</span>
<span class=normal>51</span>
<span class=normal>52</span>
<span class=normal>53</span>
<span class=normal>54</span>
<span class=normal>55</span>
<span class=normal>56</span>
<span class=normal>57</span>
<span class=normal>58</span>
<span class=normal>59</span>
<span class=normal>60</span>
<span class=normal>61</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>main</span><span class=p>(</span>
    <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span>
    <span class=n>variant_index</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
    <span class=n>variant_annotation</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
    <span class=n>variant_consequence_lut_path</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=nb>tuple</span><span class=p>[</span><span class=n>DataFrame</span><span class=p>,</span> <span class=o>...</span><span class=p>]:</span>
    <span class=sd>&quot;&quot;&quot;Extracts variant to gene assignments for the variants included in the index and the features predicted by VEP.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): ETL session,</span>
<span class=sd>        variant_index (DataFrame): DataFrame with the OTG variant index</span>
<span class=sd>        variant_annotation (DataFrame): Dataframe with the annotated variants</span>
<span class=sd>        variant_consequence_lut_path (str): The path to the LUT between the functional consequences and their assigned V2G score</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: variant to gene assignments from VEP</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Parsing functional predictions...&quot;</span><span class=p>)</span>
    <span class=n>annotated_variants</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>variant_annotation</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;variantId&quot;</span><span class=p>,</span>
            <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
            <span class=c1># exploding the array already removes record without VEP annotation</span>
            <span class=n>f</span><span class=o>.</span><span class=n>explode</span><span class=p>(</span><span class=s2>&quot;vep.transcriptConsequences&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;transcriptConsequence&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>variant_index</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;variantId&quot;</span><span class=p>,</span> <span class=s2>&quot;chromosome&quot;</span><span class=p>),</span>
            <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=s2>&quot;variantId&quot;</span><span class=p>,</span> <span class=s2>&quot;chromosome&quot;</span><span class=p>],</span>
            <span class=n>how</span><span class=o>=</span><span class=s2>&quot;inner&quot;</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>

    <span class=n>variant_consequence_lut</span> <span class=o>=</span> <span class=n>read_consequence_lut</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>variant_consequence_lut_path</span><span class=p>)</span>
    <span class=n>vep_consequences</span> <span class=o>=</span> <span class=n>get_variant_consequences</span><span class=p>(</span>
        <span class=n>annotated_variants</span><span class=p>,</span> <span class=n>variant_consequence_lut</span>
    <span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Extracted functional consequence from VEP.&quot;</span><span class=p>)</span>
    <span class=n>vep_polyphen</span> <span class=o>=</span> <span class=n>get_polyphen_score</span><span class=p>(</span><span class=n>annotated_variants</span><span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Extracted polyphen scores from VEP.&quot;</span><span class=p>)</span>
    <span class=n>vep_sift</span> <span class=o>=</span> <span class=n>get_sift_score</span><span class=p>(</span><span class=n>annotated_variants</span><span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Extracted sift scores from VEP.&quot;</span><span class=p>)</span>
    <span class=n>vep_plof</span> <span class=o>=</span> <span class=n>get_plof_flag</span><span class=p>(</span><span class=n>annotated_variants</span><span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Extracted pLOF assesments from LOFTEE.&quot;</span><span class=p>)</span>

    <span class=k>return</span> <span class=n>vep_consequences</span><span class=p>,</span> <span class=n>vep_polyphen</span><span class=p>,</span> <span class=n>vep_sift</span><span class=p>,</span> <span class=n>vep_plof</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.v2g.functional_predictions.vep.read_consequence_lut class="doc doc-heading"> <code class="highlight language-python"><span class=n>read_consequence_lut</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>variant_consequence_lut_path</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Reads the variant consequence LUT from the given path.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>variant_consequence_lut_path</code></td> <td> <code>str</code> </td> <td><p>Path to the table with the variant consequences sorted by severity</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>variant consequence LUT</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/v2g/functional_predictions/vep.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>64</span>
<span class=normal>65</span>
<span class=normal>66</span>
<span class=normal>67</span>
<span class=normal>68</span>
<span class=normal>69</span>
<span class=normal>70</span>
<span class=normal>71</span>
<span class=normal>72</span>
<span class=normal>73</span>
<span class=normal>74</span>
<span class=normal>75</span>
<span class=normal>76</span>
<span class=normal>77</span>
<span class=normal>78</span>
<span class=normal>79</span>
<span class=normal>80</span>
<span class=normal>81</span>
<span class=normal>82</span>
<span class=normal>83</span>
<span class=normal>84</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>read_consequence_lut</span><span class=p>(</span>
    <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span> <span class=n>variant_consequence_lut_path</span><span class=p>:</span> <span class=nb>str</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Reads the variant consequence LUT from the given path.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): ETL session</span>
<span class=sd>        variant_consequence_lut_path (str): Path to the table with the variant consequences sorted by severity</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: variant consequence LUT</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>csv</span><span class=p>(</span>
        <span class=n>variant_consequence_lut_path</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>header</span><span class=o>=</span><span class=kc>True</span>
    <span class=p>)</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
        <span class=n>f</span><span class=o>.</span><span class=n>element_at</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=s2>&quot;Accession&quot;</span><span class=p>,</span> <span class=sa>r</span><span class=s2>&quot;/&quot;</span><span class=p>),</span> <span class=o>-</span><span class=mi>1</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span>
            <span class=s2>&quot;variantFunctionalConsequenceId&quot;</span>
        <span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;Term&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;label&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;v2g_score&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=s2>&quot;double&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;score&quot;</span><span class=p>),</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> <hr> <div class=md-source-file> <small> Last update: <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">November 21, 2022</span> </small> </div> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../schemas/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Schemas" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Schemas </div> </div> </a> <a href=../variants/ class="md-footer__link md-footer__link--next" aria-label="Next: Variants" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Variants </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/genetics_etl_python target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../..", "features": ["navigation.instant", "navigation.top", "content.code.annotate", "search.suggest", "search.highlight"], "search": "../../assets/javascripts/workers/search.ecf98df9.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script> <script src=../../assets/javascripts/bundle.2ab50757.min.js></script> </body> </html>