
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../study_index/">
      
      
        <link rel="next" href="../../gnomad/_gnomad/">
      
      
      <link rel="icon" href="../../../../assets/imgs/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.18">
    
    
      
        <title>FinnGen Meta Analysis Summary Statistics - Open Targets Gentropy</title>
      
    
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/main.66ac8b77.min.css">
      
        
        <link rel="stylesheet" href="../../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../../css/timeago.css">
    
      <link rel="stylesheet" href="../../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../../assets/stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="light-blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../../.." title="Open Targets Gentropy" class="md-header__button md-logo" aria-label="Open Targets Gentropy" data-md-component="logo">
      
  <img src="../../../../assets/imgs/otlogo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Open Targets Gentropy
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              FinnGen Meta Analysis Summary Statistics
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="light-blue"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="light-blue"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/opentargets/gentropy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    opentargets/gentropy
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../.." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../../installation/" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../../howto/_howto/" class="md-tabs__link">
        
  
    
  
  How-to

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../../development/_development/" class="md-tabs__link">
        
  
    
  
  Development

      </a>
    </li>
  

      
        
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="../../../_python_api/" class="md-tabs__link">
        
  
    
  
  Python API

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../../.." title="Open Targets Gentropy" class="md-nav__button md-logo" aria-label="Open Targets Gentropy" data-md-component="logo">
      
  <img src="../../../../assets/imgs/otlogo.png" alt="logo">

    </a>
    Open Targets Gentropy
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/opentargets/gentropy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    opentargets/gentropy
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../howto/_howto/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    How-to
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3" id="__nav_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            How-to
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../howto/command_line/_command_line/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Command line
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1">
            <span class="md-nav__icon md-icon"></span>
            Command line
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../howto/command_line/run_step_in_cli/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run step in cli
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../howto/command_line/run_step_using_config/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run step using config
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../howto/tutorials/_tutorials/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Tutorials
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../howto/tutorials/a_creating_spark_session/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Creating a Spark Session
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../howto/tutorials/b_create_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Create a dataset
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../howto/tutorials/c_applying_methods/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Applying methods
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../howto/tutorials/d_inspect_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Inspect a dataset
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../../development/_development/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Development
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4" id="__nav_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Development
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../development/contributing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Contributing guidelines
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../../development/troubleshooting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Troubleshooting
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../_python_api/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Python API
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5" id="__nav_5_label" tabindex="">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Python API
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../assets/_assets/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Assets
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            Assets
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../assets/variant_consequences/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant Consequences
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../common/_common/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Common
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_2" id="__nav_5_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            Common
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../common/genomic_region/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    genomic_region
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../common/processing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    processing
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../common/session/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    session
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../common/spark/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    spark
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../common/stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    stats
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../common/types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    types
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../common/udf/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    udf
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../datasets/_datasets/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Datasets
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3" id="__nav_5_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3">
            <span class="md-nav__icon md-icon"></span>
            Datasets
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/biosample_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Biosample index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/intervals/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Intervals
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/l2g_feature_matrix/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Feature Matrix
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/l2g_gold_standard/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Gold Standard
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/l2g_prediction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Prediction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/ld_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LD Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/study_locus/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Locus
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/study_locus_overlap/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Locus Overlap
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/summary_statistics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/summary_statistics_qc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/target_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Target Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/variant_direction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant Direction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/variant_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3_16" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../datasets/l2g_features/_l2g_feature/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    L2g features
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_3_16" id="__nav_5_3_16_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_3_16_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3_16">
            <span class="md-nav__icon md-icon"></span>
            L2g features
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/l2g_features/colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/l2g_features/distance/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From distance
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/l2g_features/other/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Other features
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../datasets/l2g_features/vep/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From VEP
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../_datasources/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Datasources
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4" id="__nav_5_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_4">
            <span class="md-nav__icon md-icon"></span>
            Datasources
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../biosample_ontologies/_biosample_ontologies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Biosample Ontologies
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../ensembl/_ensembl/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Ensembl
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4_2" id="__nav_5_4_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4_2">
            <span class="md-nav__icon md-icon"></span>
            Ensembl
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ensembl/variant_effect_predictor_parser/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant effector parser
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../eqtl_catalogue/_eqtl_catalogue/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Eqtl catalogue
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4_3" id="__nav_5_4_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4_3">
            <span class="md-nav__icon md-icon"></span>
            Eqtl catalogue
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../eqtl_catalogue/finemapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Fine mapping results
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../eqtl_catalogue/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../eqtl_catalogue/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Stats
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../finngen/_finngen/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Finngen
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4_4" id="__nav_5_4_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_4_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4_4">
            <span class="md-nav__icon md-icon"></span>
            Finngen
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../finngen/efo_mapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    FinnGen EFO Phenotype Mapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../finngen/finemapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Finemapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../finngen/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../finngen/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4_5" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../_finngen_meta/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Finngen meta
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4_5" id="__nav_5_4_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_4_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_4_5">
            <span class="md-nav__icon md-icon"></span>
            Finngen meta
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../finngen_manifest/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    FinnGen Meta Analysis Manifest
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    FinnGen Meta Analysis Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    FinnGen Meta Analysis Summary Statistics
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    FinnGen Meta Analysis Summary Statistics
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics" class="md-nav__link">
    <span class="md-ellipsis">
      FinnGenUkbMvpMetaSummaryStatistics
    </span>
  </a>
  
    <nav class="md-nav" aria-label="FinnGenUkbMvpMetaSummaryStatistics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.allele_frequencies" class="md-nav__link">
    <span class="md-ellipsis">
      allele_frequencies
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.bgzip_to_parquet" class="md-nav__link">
    <span class="md-ellipsis">
      bgzip_to_parquet
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.cohorts" class="md-nav__link">
    <span class="md-ellipsis">
      cohorts
    </span>
  </a>
  
    <nav class="md-nav" aria-label="cohorts">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.cohorts--test-case-1-all-cohorts-have-data" class="md-nav__link">
    <span class="md-ellipsis">
      Test case 1: All cohorts have data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.cohorts--test-case-2-only-some-cohorts-have-data" class="md-nav__link">
    <span class="md-ellipsis">
      Test case 2: Only some cohorts have data
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.combined_allele_frequency" class="md-nav__link">
    <span class="md-ellipsis">
      combined_allele_frequency
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.extract_study_phenotype_from_path" class="md-nav__link">
    <span class="md-ellipsis">
      extract_study_phenotype_from_path
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.from_source" class="md-nav__link">
    <span class="md-ellipsis">
      from_source
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.has_low_imputation_score" class="md-nav__link">
    <span class="md-ellipsis">
      has_low_imputation_score
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.has_low_min_allele_count" class="md-nav__link">
    <span class="md-ellipsis">
      has_low_min_allele_count
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.has_low_min_allele_frequency" class="md-nav__link">
    <span class="md-ellipsis">
      has_low_min_allele_frequency
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.is_meta_analyzed_variant" class="md-nav__link">
    <span class="md-ellipsis">
      is_meta_analyzed_variant
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.min_allele_count" class="md-nav__link">
    <span class="md-ellipsis">
      min_allele_count
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.min_allele_frequency" class="md-nav__link">
    <span class="md-ellipsis">
      min_allele_frequency
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.normalize_af" class="md-nav__link">
    <span class="md-ellipsis">
      normalize_af
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4_6" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../gnomad/_gnomad/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Gnomad
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4_6" id="__nav_5_4_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_4_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4_6">
            <span class="md-nav__icon md-icon"></span>
            Gnomad
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gnomad/gnomad_ld/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LD Matrix
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gnomad/gnomad_variants/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variants
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4_7" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../gwas_catalog/_gwas_catalog/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Gwas catalog
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4_7" id="__nav_5_4_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_4_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4_7">
            <span class="md-nav__icon md-icon"></span>
            Gwas catalog
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gwas_catalog/associations/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Associations
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gwas_catalog/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gwas_catalog/study_splitter/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Splitter
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../gwas_catalog/summary_statistics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary statistics
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../intervals/_intervals/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Intervals
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4_8" id="__nav_5_4_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_4_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4_8">
            <span class="md-nav__icon md-icon"></span>
            Intervals
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../intervals/e2g/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ENCODE-rE2G
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../intervals/epiraction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    EPIraction
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4_9" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../open_targets/_open_targets/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Open targets
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4_9" id="__nav_5_4_9_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_4_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4_9">
            <span class="md-nav__icon md-icon"></span>
            Open targets
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../open_targets/l2g_gold_standard/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Gold Standard
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_4_10" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../ukb_ppp_eur/_ukb_ppp_eur/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Ukb ppp eur
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_4_10" id="__nav_5_4_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_4_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_4_10">
            <span class="md-nav__icon md-icon"></span>
            Ukb ppp eur
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ukb_ppp_eur/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ukb_ppp_eur/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../methods/_methods/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Methods
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_5" id="__nav_5_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_5">
            <span class="md-nav__icon md-icon"></span>
            Methods
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/carma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CARMA
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/coloc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Coloc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/coloc_pip/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Coloc using Posterior Inclusion Probabilities (PIP)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/ecaviar/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    eCAVIAR
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/ld_annotator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LDAnnotator
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/pics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PICS
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/sumstat_imputation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics Imputation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/sumstat_quality_controls/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    QC of GWAS Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/susie_inf/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SuSiE-inf
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_5_11" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../methods/l2g/_l2g/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    L2g
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_5_11" id="__nav_5_5_11_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_5_5_11_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_5_11">
            <span class="md-nav__icon md-icon"></span>
            L2g
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/l2g/feature_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Feature Factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/l2g/model/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../methods/l2g/trainer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Trainer
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_6" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../steps/_steps/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Steps
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5_6" id="__nav_5_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_6">
            <span class="md-nav__icon md-icon"></span>
            Steps
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/biosample_index_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    biosample_index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/credible_set_qc_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    credible_set_qc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/eqtl_catalogue/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    eQTL Catalogue
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/finngen_studies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    finngen_studies
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/finngen_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    finngen_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/finngen_ukbb_mvp_meta_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    FinnGen UKBB MVP Meta Analysis Step
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/gwas_catalog_curation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_study_curation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/gwas_catalog_study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_study_inclusion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/gwas_catalog_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/gwas_catalog_top_hits/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GWAS Catalog Top Hits Ingestion Step
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/l2g/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Locus to Gene (L2G)
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/ld_clump/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ld_based_clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/ld_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GnomAD Linkage data ingestion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/locus_breaker_clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    locus_breaker_clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/pics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    pics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/study_locus_validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study-Locus Validation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/study_validation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Validation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/summary_statistics_qc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    summary_statistics_qc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/ukb_ppp_eur_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ukb_ppp_eur_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/variant_annotation_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    GnomAD variant data ingestion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/variant_index_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    variant_index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../steps/window_based_clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    window_based_clumping
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics" class="md-nav__link">
    <span class="md-ellipsis">
      FinnGenUkbMvpMetaSummaryStatistics
    </span>
  </a>
  
    <nav class="md-nav" aria-label="FinnGenUkbMvpMetaSummaryStatistics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.allele_frequencies" class="md-nav__link">
    <span class="md-ellipsis">
      allele_frequencies
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.bgzip_to_parquet" class="md-nav__link">
    <span class="md-ellipsis">
      bgzip_to_parquet
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.cohorts" class="md-nav__link">
    <span class="md-ellipsis">
      cohorts
    </span>
  </a>
  
    <nav class="md-nav" aria-label="cohorts">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.cohorts--test-case-1-all-cohorts-have-data" class="md-nav__link">
    <span class="md-ellipsis">
      Test case 1: All cohorts have data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.cohorts--test-case-2-only-some-cohorts-have-data" class="md-nav__link">
    <span class="md-ellipsis">
      Test case 2: Only some cohorts have data
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.combined_allele_frequency" class="md-nav__link">
    <span class="md-ellipsis">
      combined_allele_frequency
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.extract_study_phenotype_from_path" class="md-nav__link">
    <span class="md-ellipsis">
      extract_study_phenotype_from_path
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.from_source" class="md-nav__link">
    <span class="md-ellipsis">
      from_source
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.has_low_imputation_score" class="md-nav__link">
    <span class="md-ellipsis">
      has_low_imputation_score
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.has_low_min_allele_count" class="md-nav__link">
    <span class="md-ellipsis">
      has_low_min_allele_count
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.has_low_min_allele_frequency" class="md-nav__link">
    <span class="md-ellipsis">
      has_low_min_allele_frequency
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.is_meta_analyzed_variant" class="md-nav__link">
    <span class="md-ellipsis">
      is_meta_analyzed_variant
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.min_allele_count" class="md-nav__link">
    <span class="md-ellipsis">
      min_allele_count
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.min_allele_frequency" class="md-nav__link">
    <span class="md-ellipsis">
      min_allele_frequency
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.normalize_af" class="md-nav__link">
    <span class="md-ellipsis">
      normalize_af
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>FinnGen Meta Analysis Summary Statistics</h1>

<div class="doc doc-object doc-class">



<h3 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics" class="doc doc-heading">
            <code>gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics</code>


<a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents first">


        <p>FinnGen meta summary statistics ingestion and harmonisation.</p>







              <details class="quote">
                <summary>Source code in <code>src/gentropy/datasource/finngen_meta/summary_statistics.py</code></summary>
                <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  25</span>
<span class="normal">  26</span>
<span class="normal">  27</span>
<span class="normal">  28</span>
<span class="normal">  29</span>
<span class="normal">  30</span>
<span class="normal">  31</span>
<span class="normal">  32</span>
<span class="normal">  33</span>
<span class="normal">  34</span>
<span class="normal">  35</span>
<span class="normal">  36</span>
<span class="normal">  37</span>
<span class="normal">  38</span>
<span class="normal">  39</span>
<span class="normal">  40</span>
<span class="normal">  41</span>
<span class="normal">  42</span>
<span class="normal">  43</span>
<span class="normal">  44</span>
<span class="normal">  45</span>
<span class="normal">  46</span>
<span class="normal">  47</span>
<span class="normal">  48</span>
<span class="normal">  49</span>
<span class="normal">  50</span>
<span class="normal">  51</span>
<span class="normal">  52</span>
<span class="normal">  53</span>
<span class="normal">  54</span>
<span class="normal">  55</span>
<span class="normal">  56</span>
<span class="normal">  57</span>
<span class="normal">  58</span>
<span class="normal">  59</span>
<span class="normal">  60</span>
<span class="normal">  61</span>
<span class="normal">  62</span>
<span class="normal">  63</span>
<span class="normal">  64</span>
<span class="normal">  65</span>
<span class="normal">  66</span>
<span class="normal">  67</span>
<span class="normal">  68</span>
<span class="normal">  69</span>
<span class="normal">  70</span>
<span class="normal">  71</span>
<span class="normal">  72</span>
<span class="normal">  73</span>
<span class="normal">  74</span>
<span class="normal">  75</span>
<span class="normal">  76</span>
<span class="normal">  77</span>
<span class="normal">  78</span>
<span class="normal">  79</span>
<span class="normal">  80</span>
<span class="normal">  81</span>
<span class="normal">  82</span>
<span class="normal">  83</span>
<span class="normal">  84</span>
<span class="normal">  85</span>
<span class="normal">  86</span>
<span class="normal">  87</span>
<span class="normal">  88</span>
<span class="normal">  89</span>
<span class="normal">  90</span>
<span class="normal">  91</span>
<span class="normal">  92</span>
<span class="normal">  93</span>
<span class="normal">  94</span>
<span class="normal">  95</span>
<span class="normal">  96</span>
<span class="normal">  97</span>
<span class="normal">  98</span>
<span class="normal">  99</span>
<span class="normal"> 100</span>
<span class="normal"> 101</span>
<span class="normal"> 102</span>
<span class="normal"> 103</span>
<span class="normal"> 104</span>
<span class="normal"> 105</span>
<span class="normal"> 106</span>
<span class="normal"> 107</span>
<span class="normal"> 108</span>
<span class="normal"> 109</span>
<span class="normal"> 110</span>
<span class="normal"> 111</span>
<span class="normal"> 112</span>
<span class="normal"> 113</span>
<span class="normal"> 114</span>
<span class="normal"> 115</span>
<span class="normal"> 116</span>
<span class="normal"> 117</span>
<span class="normal"> 118</span>
<span class="normal"> 119</span>
<span class="normal"> 120</span>
<span class="normal"> 121</span>
<span class="normal"> 122</span>
<span class="normal"> 123</span>
<span class="normal"> 124</span>
<span class="normal"> 125</span>
<span class="normal"> 126</span>
<span class="normal"> 127</span>
<span class="normal"> 128</span>
<span class="normal"> 129</span>
<span class="normal"> 130</span>
<span class="normal"> 131</span>
<span class="normal"> 132</span>
<span class="normal"> 133</span>
<span class="normal"> 134</span>
<span class="normal"> 135</span>
<span class="normal"> 136</span>
<span class="normal"> 137</span>
<span class="normal"> 138</span>
<span class="normal"> 139</span>
<span class="normal"> 140</span>
<span class="normal"> 141</span>
<span class="normal"> 142</span>
<span class="normal"> 143</span>
<span class="normal"> 144</span>
<span class="normal"> 145</span>
<span class="normal"> 146</span>
<span class="normal"> 147</span>
<span class="normal"> 148</span>
<span class="normal"> 149</span>
<span class="normal"> 150</span>
<span class="normal"> 151</span>
<span class="normal"> 152</span>
<span class="normal"> 153</span>
<span class="normal"> 154</span>
<span class="normal"> 155</span>
<span class="normal"> 156</span>
<span class="normal"> 157</span>
<span class="normal"> 158</span>
<span class="normal"> 159</span>
<span class="normal"> 160</span>
<span class="normal"> 161</span>
<span class="normal"> 162</span>
<span class="normal"> 163</span>
<span class="normal"> 164</span>
<span class="normal"> 165</span>
<span class="normal"> 166</span>
<span class="normal"> 167</span>
<span class="normal"> 168</span>
<span class="normal"> 169</span>
<span class="normal"> 170</span>
<span class="normal"> 171</span>
<span class="normal"> 172</span>
<span class="normal"> 173</span>
<span class="normal"> 174</span>
<span class="normal"> 175</span>
<span class="normal"> 176</span>
<span class="normal"> 177</span>
<span class="normal"> 178</span>
<span class="normal"> 179</span>
<span class="normal"> 180</span>
<span class="normal"> 181</span>
<span class="normal"> 182</span>
<span class="normal"> 183</span>
<span class="normal"> 184</span>
<span class="normal"> 185</span>
<span class="normal"> 186</span>
<span class="normal"> 187</span>
<span class="normal"> 188</span>
<span class="normal"> 189</span>
<span class="normal"> 190</span>
<span class="normal"> 191</span>
<span class="normal"> 192</span>
<span class="normal"> 193</span>
<span class="normal"> 194</span>
<span class="normal"> 195</span>
<span class="normal"> 196</span>
<span class="normal"> 197</span>
<span class="normal"> 198</span>
<span class="normal"> 199</span>
<span class="normal"> 200</span>
<span class="normal"> 201</span>
<span class="normal"> 202</span>
<span class="normal"> 203</span>
<span class="normal"> 204</span>
<span class="normal"> 205</span>
<span class="normal"> 206</span>
<span class="normal"> 207</span>
<span class="normal"> 208</span>
<span class="normal"> 209</span>
<span class="normal"> 210</span>
<span class="normal"> 211</span>
<span class="normal"> 212</span>
<span class="normal"> 213</span>
<span class="normal"> 214</span>
<span class="normal"> 215</span>
<span class="normal"> 216</span>
<span class="normal"> 217</span>
<span class="normal"> 218</span>
<span class="normal"> 219</span>
<span class="normal"> 220</span>
<span class="normal"> 221</span>
<span class="normal"> 222</span>
<span class="normal"> 223</span>
<span class="normal"> 224</span>
<span class="normal"> 225</span>
<span class="normal"> 226</span>
<span class="normal"> 227</span>
<span class="normal"> 228</span>
<span class="normal"> 229</span>
<span class="normal"> 230</span>
<span class="normal"> 231</span>
<span class="normal"> 232</span>
<span class="normal"> 233</span>
<span class="normal"> 234</span>
<span class="normal"> 235</span>
<span class="normal"> 236</span>
<span class="normal"> 237</span>
<span class="normal"> 238</span>
<span class="normal"> 239</span>
<span class="normal"> 240</span>
<span class="normal"> 241</span>
<span class="normal"> 242</span>
<span class="normal"> 243</span>
<span class="normal"> 244</span>
<span class="normal"> 245</span>
<span class="normal"> 246</span>
<span class="normal"> 247</span>
<span class="normal"> 248</span>
<span class="normal"> 249</span>
<span class="normal"> 250</span>
<span class="normal"> 251</span>
<span class="normal"> 252</span>
<span class="normal"> 253</span>
<span class="normal"> 254</span>
<span class="normal"> 255</span>
<span class="normal"> 256</span>
<span class="normal"> 257</span>
<span class="normal"> 258</span>
<span class="normal"> 259</span>
<span class="normal"> 260</span>
<span class="normal"> 261</span>
<span class="normal"> 262</span>
<span class="normal"> 263</span>
<span class="normal"> 264</span>
<span class="normal"> 265</span>
<span class="normal"> 266</span>
<span class="normal"> 267</span>
<span class="normal"> 268</span>
<span class="normal"> 269</span>
<span class="normal"> 270</span>
<span class="normal"> 271</span>
<span class="normal"> 272</span>
<span class="normal"> 273</span>
<span class="normal"> 274</span>
<span class="normal"> 275</span>
<span class="normal"> 276</span>
<span class="normal"> 277</span>
<span class="normal"> 278</span>
<span class="normal"> 279</span>
<span class="normal"> 280</span>
<span class="normal"> 281</span>
<span class="normal"> 282</span>
<span class="normal"> 283</span>
<span class="normal"> 284</span>
<span class="normal"> 285</span>
<span class="normal"> 286</span>
<span class="normal"> 287</span>
<span class="normal"> 288</span>
<span class="normal"> 289</span>
<span class="normal"> 290</span>
<span class="normal"> 291</span>
<span class="normal"> 292</span>
<span class="normal"> 293</span>
<span class="normal"> 294</span>
<span class="normal"> 295</span>
<span class="normal"> 296</span>
<span class="normal"> 297</span>
<span class="normal"> 298</span>
<span class="normal"> 299</span>
<span class="normal"> 300</span>
<span class="normal"> 301</span>
<span class="normal"> 302</span>
<span class="normal"> 303</span>
<span class="normal"> 304</span>
<span class="normal"> 305</span>
<span class="normal"> 306</span>
<span class="normal"> 307</span>
<span class="normal"> 308</span>
<span class="normal"> 309</span>
<span class="normal"> 310</span>
<span class="normal"> 311</span>
<span class="normal"> 312</span>
<span class="normal"> 313</span>
<span class="normal"> 314</span>
<span class="normal"> 315</span>
<span class="normal"> 316</span>
<span class="normal"> 317</span>
<span class="normal"> 318</span>
<span class="normal"> 319</span>
<span class="normal"> 320</span>
<span class="normal"> 321</span>
<span class="normal"> 322</span>
<span class="normal"> 323</span>
<span class="normal"> 324</span>
<span class="normal"> 325</span>
<span class="normal"> 326</span>
<span class="normal"> 327</span>
<span class="normal"> 328</span>
<span class="normal"> 329</span>
<span class="normal"> 330</span>
<span class="normal"> 331</span>
<span class="normal"> 332</span>
<span class="normal"> 333</span>
<span class="normal"> 334</span>
<span class="normal"> 335</span>
<span class="normal"> 336</span>
<span class="normal"> 337</span>
<span class="normal"> 338</span>
<span class="normal"> 339</span>
<span class="normal"> 340</span>
<span class="normal"> 341</span>
<span class="normal"> 342</span>
<span class="normal"> 343</span>
<span class="normal"> 344</span>
<span class="normal"> 345</span>
<span class="normal"> 346</span>
<span class="normal"> 347</span>
<span class="normal"> 348</span>
<span class="normal"> 349</span>
<span class="normal"> 350</span>
<span class="normal"> 351</span>
<span class="normal"> 352</span>
<span class="normal"> 353</span>
<span class="normal"> 354</span>
<span class="normal"> 355</span>
<span class="normal"> 356</span>
<span class="normal"> 357</span>
<span class="normal"> 358</span>
<span class="normal"> 359</span>
<span class="normal"> 360</span>
<span class="normal"> 361</span>
<span class="normal"> 362</span>
<span class="normal"> 363</span>
<span class="normal"> 364</span>
<span class="normal"> 365</span>
<span class="normal"> 366</span>
<span class="normal"> 367</span>
<span class="normal"> 368</span>
<span class="normal"> 369</span>
<span class="normal"> 370</span>
<span class="normal"> 371</span>
<span class="normal"> 372</span>
<span class="normal"> 373</span>
<span class="normal"> 374</span>
<span class="normal"> 375</span>
<span class="normal"> 376</span>
<span class="normal"> 377</span>
<span class="normal"> 378</span>
<span class="normal"> 379</span>
<span class="normal"> 380</span>
<span class="normal"> 381</span>
<span class="normal"> 382</span>
<span class="normal"> 383</span>
<span class="normal"> 384</span>
<span class="normal"> 385</span>
<span class="normal"> 386</span>
<span class="normal"> 387</span>
<span class="normal"> 388</span>
<span class="normal"> 389</span>
<span class="normal"> 390</span>
<span class="normal"> 391</span>
<span class="normal"> 392</span>
<span class="normal"> 393</span>
<span class="normal"> 394</span>
<span class="normal"> 395</span>
<span class="normal"> 396</span>
<span class="normal"> 397</span>
<span class="normal"> 398</span>
<span class="normal"> 399</span>
<span class="normal"> 400</span>
<span class="normal"> 401</span>
<span class="normal"> 402</span>
<span class="normal"> 403</span>
<span class="normal"> 404</span>
<span class="normal"> 405</span>
<span class="normal"> 406</span>
<span class="normal"> 407</span>
<span class="normal"> 408</span>
<span class="normal"> 409</span>
<span class="normal"> 410</span>
<span class="normal"> 411</span>
<span class="normal"> 412</span>
<span class="normal"> 413</span>
<span class="normal"> 414</span>
<span class="normal"> 415</span>
<span class="normal"> 416</span>
<span class="normal"> 417</span>
<span class="normal"> 418</span>
<span class="normal"> 419</span>
<span class="normal"> 420</span>
<span class="normal"> 421</span>
<span class="normal"> 422</span>
<span class="normal"> 423</span>
<span class="normal"> 424</span>
<span class="normal"> 425</span>
<span class="normal"> 426</span>
<span class="normal"> 427</span>
<span class="normal"> 428</span>
<span class="normal"> 429</span>
<span class="normal"> 430</span>
<span class="normal"> 431</span>
<span class="normal"> 432</span>
<span class="normal"> 433</span>
<span class="normal"> 434</span>
<span class="normal"> 435</span>
<span class="normal"> 436</span>
<span class="normal"> 437</span>
<span class="normal"> 438</span>
<span class="normal"> 439</span>
<span class="normal"> 440</span>
<span class="normal"> 441</span>
<span class="normal"> 442</span>
<span class="normal"> 443</span>
<span class="normal"> 444</span>
<span class="normal"> 445</span>
<span class="normal"> 446</span>
<span class="normal"> 447</span>
<span class="normal"> 448</span>
<span class="normal"> 449</span>
<span class="normal"> 450</span>
<span class="normal"> 451</span>
<span class="normal"> 452</span>
<span class="normal"> 453</span>
<span class="normal"> 454</span>
<span class="normal"> 455</span>
<span class="normal"> 456</span>
<span class="normal"> 457</span>
<span class="normal"> 458</span>
<span class="normal"> 459</span>
<span class="normal"> 460</span>
<span class="normal"> 461</span>
<span class="normal"> 462</span>
<span class="normal"> 463</span>
<span class="normal"> 464</span>
<span class="normal"> 465</span>
<span class="normal"> 466</span>
<span class="normal"> 467</span>
<span class="normal"> 468</span>
<span class="normal"> 469</span>
<span class="normal"> 470</span>
<span class="normal"> 471</span>
<span class="normal"> 472</span>
<span class="normal"> 473</span>
<span class="normal"> 474</span>
<span class="normal"> 475</span>
<span class="normal"> 476</span>
<span class="normal"> 477</span>
<span class="normal"> 478</span>
<span class="normal"> 479</span>
<span class="normal"> 480</span>
<span class="normal"> 481</span>
<span class="normal"> 482</span>
<span class="normal"> 483</span>
<span class="normal"> 484</span>
<span class="normal"> 485</span>
<span class="normal"> 486</span>
<span class="normal"> 487</span>
<span class="normal"> 488</span>
<span class="normal"> 489</span>
<span class="normal"> 490</span>
<span class="normal"> 491</span>
<span class="normal"> 492</span>
<span class="normal"> 493</span>
<span class="normal"> 494</span>
<span class="normal"> 495</span>
<span class="normal"> 496</span>
<span class="normal"> 497</span>
<span class="normal"> 498</span>
<span class="normal"> 499</span>
<span class="normal"> 500</span>
<span class="normal"> 501</span>
<span class="normal"> 502</span>
<span class="normal"> 503</span>
<span class="normal"> 504</span>
<span class="normal"> 505</span>
<span class="normal"> 506</span>
<span class="normal"> 507</span>
<span class="normal"> 508</span>
<span class="normal"> 509</span>
<span class="normal"> 510</span>
<span class="normal"> 511</span>
<span class="normal"> 512</span>
<span class="normal"> 513</span>
<span class="normal"> 514</span>
<span class="normal"> 515</span>
<span class="normal"> 516</span>
<span class="normal"> 517</span>
<span class="normal"> 518</span>
<span class="normal"> 519</span>
<span class="normal"> 520</span>
<span class="normal"> 521</span>
<span class="normal"> 522</span>
<span class="normal"> 523</span>
<span class="normal"> 524</span>
<span class="normal"> 525</span>
<span class="normal"> 526</span>
<span class="normal"> 527</span>
<span class="normal"> 528</span>
<span class="normal"> 529</span>
<span class="normal"> 530</span>
<span class="normal"> 531</span>
<span class="normal"> 532</span>
<span class="normal"> 533</span>
<span class="normal"> 534</span>
<span class="normal"> 535</span>
<span class="normal"> 536</span>
<span class="normal"> 537</span>
<span class="normal"> 538</span>
<span class="normal"> 539</span>
<span class="normal"> 540</span>
<span class="normal"> 541</span>
<span class="normal"> 542</span>
<span class="normal"> 543</span>
<span class="normal"> 544</span>
<span class="normal"> 545</span>
<span class="normal"> 546</span>
<span class="normal"> 547</span>
<span class="normal"> 548</span>
<span class="normal"> 549</span>
<span class="normal"> 550</span>
<span class="normal"> 551</span>
<span class="normal"> 552</span>
<span class="normal"> 553</span>
<span class="normal"> 554</span>
<span class="normal"> 555</span>
<span class="normal"> 556</span>
<span class="normal"> 557</span>
<span class="normal"> 558</span>
<span class="normal"> 559</span>
<span class="normal"> 560</span>
<span class="normal"> 561</span>
<span class="normal"> 562</span>
<span class="normal"> 563</span>
<span class="normal"> 564</span>
<span class="normal"> 565</span>
<span class="normal"> 566</span>
<span class="normal"> 567</span>
<span class="normal"> 568</span>
<span class="normal"> 569</span>
<span class="normal"> 570</span>
<span class="normal"> 571</span>
<span class="normal"> 572</span>
<span class="normal"> 573</span>
<span class="normal"> 574</span>
<span class="normal"> 575</span>
<span class="normal"> 576</span>
<span class="normal"> 577</span>
<span class="normal"> 578</span>
<span class="normal"> 579</span>
<span class="normal"> 580</span>
<span class="normal"> 581</span>
<span class="normal"> 582</span>
<span class="normal"> 583</span>
<span class="normal"> 584</span>
<span class="normal"> 585</span>
<span class="normal"> 586</span>
<span class="normal"> 587</span>
<span class="normal"> 588</span>
<span class="normal"> 589</span>
<span class="normal"> 590</span>
<span class="normal"> 591</span>
<span class="normal"> 592</span>
<span class="normal"> 593</span>
<span class="normal"> 594</span>
<span class="normal"> 595</span>
<span class="normal"> 596</span>
<span class="normal"> 597</span>
<span class="normal"> 598</span>
<span class="normal"> 599</span>
<span class="normal"> 600</span>
<span class="normal"> 601</span>
<span class="normal"> 602</span>
<span class="normal"> 603</span>
<span class="normal"> 604</span>
<span class="normal"> 605</span>
<span class="normal"> 606</span>
<span class="normal"> 607</span>
<span class="normal"> 608</span>
<span class="normal"> 609</span>
<span class="normal"> 610</span>
<span class="normal"> 611</span>
<span class="normal"> 612</span>
<span class="normal"> 613</span>
<span class="normal"> 614</span>
<span class="normal"> 615</span>
<span class="normal"> 616</span>
<span class="normal"> 617</span>
<span class="normal"> 618</span>
<span class="normal"> 619</span>
<span class="normal"> 620</span>
<span class="normal"> 621</span>
<span class="normal"> 622</span>
<span class="normal"> 623</span>
<span class="normal"> 624</span>
<span class="normal"> 625</span>
<span class="normal"> 626</span>
<span class="normal"> 627</span>
<span class="normal"> 628</span>
<span class="normal"> 629</span>
<span class="normal"> 630</span>
<span class="normal"> 631</span>
<span class="normal"> 632</span>
<span class="normal"> 633</span>
<span class="normal"> 634</span>
<span class="normal"> 635</span>
<span class="normal"> 636</span>
<span class="normal"> 637</span>
<span class="normal"> 638</span>
<span class="normal"> 639</span>
<span class="normal"> 640</span>
<span class="normal"> 641</span>
<span class="normal"> 642</span>
<span class="normal"> 643</span>
<span class="normal"> 644</span>
<span class="normal"> 645</span>
<span class="normal"> 646</span>
<span class="normal"> 647</span>
<span class="normal"> 648</span>
<span class="normal"> 649</span>
<span class="normal"> 650</span>
<span class="normal"> 651</span>
<span class="normal"> 652</span>
<span class="normal"> 653</span>
<span class="normal"> 654</span>
<span class="normal"> 655</span>
<span class="normal"> 656</span>
<span class="normal"> 657</span>
<span class="normal"> 658</span>
<span class="normal"> 659</span>
<span class="normal"> 660</span>
<span class="normal"> 661</span>
<span class="normal"> 662</span>
<span class="normal"> 663</span>
<span class="normal"> 664</span>
<span class="normal"> 665</span>
<span class="normal"> 666</span>
<span class="normal"> 667</span>
<span class="normal"> 668</span>
<span class="normal"> 669</span>
<span class="normal"> 670</span>
<span class="normal"> 671</span>
<span class="normal"> 672</span>
<span class="normal"> 673</span>
<span class="normal"> 674</span>
<span class="normal"> 675</span>
<span class="normal"> 676</span>
<span class="normal"> 677</span>
<span class="normal"> 678</span>
<span class="normal"> 679</span>
<span class="normal"> 680</span>
<span class="normal"> 681</span>
<span class="normal"> 682</span>
<span class="normal"> 683</span>
<span class="normal"> 684</span>
<span class="normal"> 685</span>
<span class="normal"> 686</span>
<span class="normal"> 687</span>
<span class="normal"> 688</span>
<span class="normal"> 689</span>
<span class="normal"> 690</span>
<span class="normal"> 691</span>
<span class="normal"> 692</span>
<span class="normal"> 693</span>
<span class="normal"> 694</span>
<span class="normal"> 695</span>
<span class="normal"> 696</span>
<span class="normal"> 697</span>
<span class="normal"> 698</span>
<span class="normal"> 699</span>
<span class="normal"> 700</span>
<span class="normal"> 701</span>
<span class="normal"> 702</span>
<span class="normal"> 703</span>
<span class="normal"> 704</span>
<span class="normal"> 705</span>
<span class="normal"> 706</span>
<span class="normal"> 707</span>
<span class="normal"> 708</span>
<span class="normal"> 709</span>
<span class="normal"> 710</span>
<span class="normal"> 711</span>
<span class="normal"> 712</span>
<span class="normal"> 713</span>
<span class="normal"> 714</span>
<span class="normal"> 715</span>
<span class="normal"> 716</span>
<span class="normal"> 717</span>
<span class="normal"> 718</span>
<span class="normal"> 719</span>
<span class="normal"> 720</span>
<span class="normal"> 721</span>
<span class="normal"> 722</span>
<span class="normal"> 723</span>
<span class="normal"> 724</span>
<span class="normal"> 725</span>
<span class="normal"> 726</span>
<span class="normal"> 727</span>
<span class="normal"> 728</span>
<span class="normal"> 729</span>
<span class="normal"> 730</span>
<span class="normal"> 731</span>
<span class="normal"> 732</span>
<span class="normal"> 733</span>
<span class="normal"> 734</span>
<span class="normal"> 735</span>
<span class="normal"> 736</span>
<span class="normal"> 737</span>
<span class="normal"> 738</span>
<span class="normal"> 739</span>
<span class="normal"> 740</span>
<span class="normal"> 741</span>
<span class="normal"> 742</span>
<span class="normal"> 743</span>
<span class="normal"> 744</span>
<span class="normal"> 745</span>
<span class="normal"> 746</span>
<span class="normal"> 747</span>
<span class="normal"> 748</span>
<span class="normal"> 749</span>
<span class="normal"> 750</span>
<span class="normal"> 751</span>
<span class="normal"> 752</span>
<span class="normal"> 753</span>
<span class="normal"> 754</span>
<span class="normal"> 755</span>
<span class="normal"> 756</span>
<span class="normal"> 757</span>
<span class="normal"> 758</span>
<span class="normal"> 759</span>
<span class="normal"> 760</span>
<span class="normal"> 761</span>
<span class="normal"> 762</span>
<span class="normal"> 763</span>
<span class="normal"> 764</span>
<span class="normal"> 765</span>
<span class="normal"> 766</span>
<span class="normal"> 767</span>
<span class="normal"> 768</span>
<span class="normal"> 769</span>
<span class="normal"> 770</span>
<span class="normal"> 771</span>
<span class="normal"> 772</span>
<span class="normal"> 773</span>
<span class="normal"> 774</span>
<span class="normal"> 775</span>
<span class="normal"> 776</span>
<span class="normal"> 777</span>
<span class="normal"> 778</span>
<span class="normal"> 779</span>
<span class="normal"> 780</span>
<span class="normal"> 781</span>
<span class="normal"> 782</span>
<span class="normal"> 783</span>
<span class="normal"> 784</span>
<span class="normal"> 785</span>
<span class="normal"> 786</span>
<span class="normal"> 787</span>
<span class="normal"> 788</span>
<span class="normal"> 789</span>
<span class="normal"> 790</span>
<span class="normal"> 791</span>
<span class="normal"> 792</span>
<span class="normal"> 793</span>
<span class="normal"> 794</span>
<span class="normal"> 795</span>
<span class="normal"> 796</span>
<span class="normal"> 797</span>
<span class="normal"> 798</span>
<span class="normal"> 799</span>
<span class="normal"> 800</span>
<span class="normal"> 801</span>
<span class="normal"> 802</span>
<span class="normal"> 803</span>
<span class="normal"> 804</span>
<span class="normal"> 805</span>
<span class="normal"> 806</span>
<span class="normal"> 807</span>
<span class="normal"> 808</span>
<span class="normal"> 809</span>
<span class="normal"> 810</span>
<span class="normal"> 811</span>
<span class="normal"> 812</span>
<span class="normal"> 813</span>
<span class="normal"> 814</span>
<span class="normal"> 815</span>
<span class="normal"> 816</span>
<span class="normal"> 817</span>
<span class="normal"> 818</span>
<span class="normal"> 819</span>
<span class="normal"> 820</span>
<span class="normal"> 821</span>
<span class="normal"> 822</span>
<span class="normal"> 823</span>
<span class="normal"> 824</span>
<span class="normal"> 825</span>
<span class="normal"> 826</span>
<span class="normal"> 827</span>
<span class="normal"> 828</span>
<span class="normal"> 829</span>
<span class="normal"> 830</span>
<span class="normal"> 831</span>
<span class="normal"> 832</span>
<span class="normal"> 833</span>
<span class="normal"> 834</span>
<span class="normal"> 835</span>
<span class="normal"> 836</span>
<span class="normal"> 837</span>
<span class="normal"> 838</span>
<span class="normal"> 839</span>
<span class="normal"> 840</span>
<span class="normal"> 841</span>
<span class="normal"> 842</span>
<span class="normal"> 843</span>
<span class="normal"> 844</span>
<span class="normal"> 845</span>
<span class="normal"> 846</span>
<span class="normal"> 847</span>
<span class="normal"> 848</span>
<span class="normal"> 849</span>
<span class="normal"> 850</span>
<span class="normal"> 851</span>
<span class="normal"> 852</span>
<span class="normal"> 853</span>
<span class="normal"> 854</span>
<span class="normal"> 855</span>
<span class="normal"> 856</span>
<span class="normal"> 857</span>
<span class="normal"> 858</span>
<span class="normal"> 859</span>
<span class="normal"> 860</span>
<span class="normal"> 861</span>
<span class="normal"> 862</span>
<span class="normal"> 863</span>
<span class="normal"> 864</span>
<span class="normal"> 865</span>
<span class="normal"> 866</span>
<span class="normal"> 867</span>
<span class="normal"> 868</span>
<span class="normal"> 869</span>
<span class="normal"> 870</span>
<span class="normal"> 871</span>
<span class="normal"> 872</span>
<span class="normal"> 873</span>
<span class="normal"> 874</span>
<span class="normal"> 875</span>
<span class="normal"> 876</span>
<span class="normal"> 877</span>
<span class="normal"> 878</span>
<span class="normal"> 879</span>
<span class="normal"> 880</span>
<span class="normal"> 881</span>
<span class="normal"> 882</span>
<span class="normal"> 883</span>
<span class="normal"> 884</span>
<span class="normal"> 885</span>
<span class="normal"> 886</span>
<span class="normal"> 887</span>
<span class="normal"> 888</span>
<span class="normal"> 889</span>
<span class="normal"> 890</span>
<span class="normal"> 891</span>
<span class="normal"> 892</span>
<span class="normal"> 893</span>
<span class="normal"> 894</span>
<span class="normal"> 895</span>
<span class="normal"> 896</span>
<span class="normal"> 897</span>
<span class="normal"> 898</span>
<span class="normal"> 899</span>
<span class="normal"> 900</span>
<span class="normal"> 901</span>
<span class="normal"> 902</span>
<span class="normal"> 903</span>
<span class="normal"> 904</span>
<span class="normal"> 905</span>
<span class="normal"> 906</span>
<span class="normal"> 907</span>
<span class="normal"> 908</span>
<span class="normal"> 909</span>
<span class="normal"> 910</span>
<span class="normal"> 911</span>
<span class="normal"> 912</span>
<span class="normal"> 913</span>
<span class="normal"> 914</span>
<span class="normal"> 915</span>
<span class="normal"> 916</span>
<span class="normal"> 917</span>
<span class="normal"> 918</span>
<span class="normal"> 919</span>
<span class="normal"> 920</span>
<span class="normal"> 921</span>
<span class="normal"> 922</span>
<span class="normal"> 923</span>
<span class="normal"> 924</span>
<span class="normal"> 925</span>
<span class="normal"> 926</span>
<span class="normal"> 927</span>
<span class="normal"> 928</span>
<span class="normal"> 929</span>
<span class="normal"> 930</span>
<span class="normal"> 931</span>
<span class="normal"> 932</span>
<span class="normal"> 933</span>
<span class="normal"> 934</span>
<span class="normal"> 935</span>
<span class="normal"> 936</span>
<span class="normal"> 937</span>
<span class="normal"> 938</span>
<span class="normal"> 939</span>
<span class="normal"> 940</span>
<span class="normal"> 941</span>
<span class="normal"> 942</span>
<span class="normal"> 943</span>
<span class="normal"> 944</span>
<span class="normal"> 945</span>
<span class="normal"> 946</span>
<span class="normal"> 947</span>
<span class="normal"> 948</span>
<span class="normal"> 949</span>
<span class="normal"> 950</span>
<span class="normal"> 951</span>
<span class="normal"> 952</span>
<span class="normal"> 953</span>
<span class="normal"> 954</span>
<span class="normal"> 955</span>
<span class="normal"> 956</span>
<span class="normal"> 957</span>
<span class="normal"> 958</span>
<span class="normal"> 959</span>
<span class="normal"> 960</span>
<span class="normal"> 961</span>
<span class="normal"> 962</span>
<span class="normal"> 963</span>
<span class="normal"> 964</span>
<span class="normal"> 965</span>
<span class="normal"> 966</span>
<span class="normal"> 967</span>
<span class="normal"> 968</span>
<span class="normal"> 969</span>
<span class="normal"> 970</span>
<span class="normal"> 971</span>
<span class="normal"> 972</span>
<span class="normal"> 973</span>
<span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span>
<span class="normal">1124</span>
<span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span>
<span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span>
<span class="normal">1161</span>
<span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span>
<span class="normal">1174</span>
<span class="normal">1175</span>
<span class="normal">1176</span>
<span class="normal">1177</span>
<span class="normal">1178</span>
<span class="normal">1179</span>
<span class="normal">1180</span>
<span class="normal">1181</span>
<span class="normal">1182</span>
<span class="normal">1183</span>
<span class="normal">1184</span>
<span class="normal">1185</span>
<span class="normal">1186</span>
<span class="normal">1187</span>
<span class="normal">1188</span>
<span class="normal">1189</span>
<span class="normal">1190</span>
<span class="normal">1191</span>
<span class="normal">1192</span>
<span class="normal">1193</span>
<span class="normal">1194</span>
<span class="normal">1195</span>
<span class="normal">1196</span>
<span class="normal">1197</span>
<span class="normal">1198</span>
<span class="normal">1199</span>
<span class="normal">1200</span>
<span class="normal">1201</span>
<span class="normal">1202</span>
<span class="normal">1203</span>
<span class="normal">1204</span>
<span class="normal">1205</span>
<span class="normal">1206</span>
<span class="normal">1207</span>
<span class="normal">1208</span>
<span class="normal">1209</span>
<span class="normal">1210</span>
<span class="normal">1211</span>
<span class="normal">1212</span>
<span class="normal">1213</span>
<span class="normal">1214</span>
<span class="normal">1215</span>
<span class="normal">1216</span>
<span class="normal">1217</span>
<span class="normal">1218</span>
<span class="normal">1219</span>
<span class="normal">1220</span>
<span class="normal">1221</span>
<span class="normal">1222</span>
<span class="normal">1223</span>
<span class="normal">1224</span>
<span class="normal">1225</span>
<span class="normal">1226</span>
<span class="normal">1227</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">FinnGenUkbMvpMetaSummaryStatistics</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;FinnGen meta summary statistics ingestion and harmonisation.&quot;&quot;&quot;</span>

    <span class="n">N_THREAD_MAX</span> <span class="o">=</span> <span class="mi">32</span>
    <span class="n">N_THREAD_OPTIMAL</span> <span class="o">=</span> <span class="mi">10</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">extract_study_phenotype_from_path</span><span class="p">(</span><span class="n">file_path</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract the study phenotype from FinnGen file path.</span>

<span class="sd">        Note:</span>
<span class="sd">            Assumes the file name format is some_path/to/&lt;studyPhenotype&gt;_meta_out.tsv.gz</span>


<span class="sd">        Args:</span>
<span class="sd">            file_path (Column): Column containing the file path as a string.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Extracted study phenotype as a string.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; data = [(&quot;/path/to/AB1_meta_out.tsv.gz&quot;,), (&quot;/another/path/CD2_meta_out.tsv.gz&quot;,)]</span>
<span class="sd">            &gt;&gt;&gt; schema = &quot;filePath STRING&quot;</span>
<span class="sd">            &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">            &gt;&gt;&gt; df = df.withColumn(&quot;studyPhenotype&quot;, FinnGenUkbMvpMetaSummaryStatistics.extract_study_phenotype_from_path(f.col(&quot;filePath&quot;)))</span>
<span class="sd">            &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">            +---------------------------------+--------------+</span>
<span class="sd">            |filePath                         |studyPhenotype|</span>
<span class="sd">            +---------------------------------+--------------+</span>
<span class="sd">            |/path/to/AB1_meta_out.tsv.gz     |AB1           |</span>
<span class="sd">            |/another/path/CD2_meta_out.tsv.gz|CD2           |</span>
<span class="sd">            +---------------------------------+--------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">regexp_replace</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">element_at</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s2">&quot;/&quot;</span><span class="p">),</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="s2">&quot;_meta_out.tsv.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span>
        <span class="p">)</span>

    <span class="c1"># raw_schema represents the order of columns in the original summary statistics files</span>
    <span class="n">raw_schema</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">StructType</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;#CHR&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;POS&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">LongType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;REF&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;ALT&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;SNP&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="c1"># FinnGen</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;fg_beta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;fg_sebeta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;fg_pval&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;fg_af_alt&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;fg_af_alt_cases&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;fg_af_alt_controls&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="c1"># MVP_EUR</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_beta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_sebeta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_pval&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_af_alt&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_r2&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="c1"># MVP_AFR</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_beta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_sebeta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_pval&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_af_alt&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_r2&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="c1"># MVP_HIS</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_beta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_sebeta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_pval&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_af_alt&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_r2&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="c1"># UKBB</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;ukbb_beta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;ukbb_sebeta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;ukbb_pval&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;ukbb_af_alt&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="c1"># Meta</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;all_meta_N&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">IntegerType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_beta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_sebeta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_p&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_mlogp&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;all_inv_var_het_p&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="c1"># Leave-one-out: FinnGen</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_fg_N&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">IntegerType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_fg_inv_var_meta_beta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_fg_inv_var_meta_sebeta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_fg_inv_var_meta_p&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_fg_inv_var_meta_mlogp&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_fg_inv_var_meta_het_p&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="c1"># Leave-one-out: MVP_EUR</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_EUR_N&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">IntegerType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_EUR_inv_var_meta_beta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_EUR_inv_var_meta_sebeta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_EUR_inv_var_meta_p&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_EUR_inv_var_meta_mlogp&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_EUR_inv_var_meta_het_p&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="c1"># Leave-one-out: MVP_AFR</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_AFR_N&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">IntegerType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_AFR_inv_var_meta_beta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_AFR_inv_var_meta_sebeta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_AFR_inv_var_meta_p&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_AFR_inv_var_meta_mlogp&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_AFR_inv_var_meta_het_p&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="c1"># Leave-one-out: MVP_HIS</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_HIS_N&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">IntegerType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_HIS_inv_var_meta_beta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_HIS_inv_var_meta_sebeta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_HIS_inv_var_meta_p&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_HIS_inv_var_meta_mlogp&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_MVP_HIS_inv_var_meta_het_p&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="c1"># Leave-one-out: UKBB</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_ukbb_N&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">IntegerType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_ukbb_inv_var_meta_beta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_ukbb_inv_var_meta_sebeta&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_ukbb_inv_var_meta_p&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_ukbb_inv_var_meta_mlogp&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;leave_ukbb_inv_var_meta_het_p&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
            <span class="c1"># RSID</span>
            <span class="n">t</span><span class="o">.</span><span class="n">StructField</span><span class="p">(</span><span class="s2">&quot;rsid&quot;</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">(),</span> <span class="kc">True</span><span class="p">),</span>
        <span class="p">]</span>
    <span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">bgzip_to_parquet</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">,</span>
        <span class="n">session</span><span class="p">:</span> <span class="n">Session</span><span class="p">,</span>
        <span class="n">summary_statistics_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">datasource</span><span class="p">:</span> <span class="n">MetaAnalysisDataSource</span><span class="p">,</span>
        <span class="n">raw_summary_statistics_output_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">n_threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Convert gzipped summary statistics to Parquet format.</span>

<span class="sd">        This is a pre-step that needs to be performed once to convert the block gzipped to parquet format. This step</span>
<span class="sd">        should be run before the actual harmonisation step is performed and the output Parquet files can be used as input</span>
<span class="sd">        for the `from_source` method.</span>

<span class="sd">        Args:</span>
<span class="sd">            session (Session): Session object.</span>
<span class="sd">            summary_statistics_list (list[str]): List of paths where summary statistics files are located.</span>
<span class="sd">            datasource (MetaAnalysisDataSource): Data source, can be FinnGenMetaDataSource.FINNGEN_UKBB_MVP.</span>
<span class="sd">            raw_summary_statistics_output_path (str): Output path for the Parquet files.</span>
<span class="sd">            n_threads (int): Maximum number of threads to use for ThreadPoolExecutor (default is 10).</span>

<span class="sd">        The output requires a single path that will be populated with Parquet files partitioned by `studyId` extracted</span>
<span class="sd">        from the input file names.</span>

<span class="sd">        !!! note &quot;Block gzipped input files&quot;</span>

<span class="sd">            Since the individual summary statistics files are **block gzipped** we use the enhanced bgzip codec for efficient reading.</span>

<span class="sd">        !!! note &quot;Reading multiple files with divergent schemas&quot;</span>

<span class="sd">            Since the schema for individual summary statistics **is not strictly the same we have to enforce the schema**.</span>

<span class="sd">            **_enforcing schema_**</span>
<span class="sd">            using the `enforceSchema` option in `spark.read.csv` **does not map columns that exist in the file provided schema**,</span>
<span class="sd">            but rather aligns columns positionally, which breaks the column order per individual file.</span>

<span class="sd">            **_inferring schema_**</span>
<span class="sd">            Attempting to use the `inferSchema` option in `spark.read.csv` while reading multiple files in bulk drops columns, due</span>
<span class="sd">            to the random sampling of files to infer the schema. (Files chosen to infer the schema may not contain entire superset of column space.)</span>

<span class="sd">            **_manual schema enforcement_**</span>
<span class="sd">            The only way to keep the columns in order and use full column superset is to loop over the files with `inferSchema` and manually</span>
<span class="sd">            add missing columns with null values casted to expected type. The looping can be parallelized using a **thread pool** (ThreadPoolExecutor)</span>
<span class="sd">            with `n_threads` as the maximum load of jobs to spark cluster.</span>

<span class="sd">        ??? warning &quot;Performance considerations&quot;</span>
<span class="sd">            This function requires a _Session_ with `use_enhanced_bgzip_codec` to be set to True. This function is strongly encouraged to be used in</span>
<span class="sd">            a distributed environment.</span>

<span class="sd">        Raises:</span>
<span class="sd">            KeyError: If `use_enhanced_bgzip_codec` is set to False in the Session configuration.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">summary_statistics_list</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">session</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No summary statistics paths found to process.&quot;</span><span class="p">)</span>
            <span class="k">return</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">session</span><span class="o">.</span><span class="n">use_enhanced_bgzip_codec</span><span class="p">:</span>
            <span class="n">session</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="s2">&quot;The use_enhanced_bgzip_codec is set to False. This will lead to inefficient reading of block gzipped files.&quot;</span>
            <span class="p">)</span>
            <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span>
                <span class="s2">&quot;Please set `session.spark.use_enhanced_bgzip_codec` to True in the Session configuration.&quot;</span>
            <span class="p">)</span>

        <span class="c1"># Handle n_threads limits and warnings</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">n_threads</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">or</span> <span class="n">n_threads</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">session</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Invalid n_threads value: </span><span class="si">{</span><span class="n">n_threads</span><span class="si">}</span><span class="s2">. Falling back to 10 threads.&quot;</span>
            <span class="p">)</span>
            <span class="n">n_threads</span> <span class="o">=</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">N_THREAD_OPTIMAL</span>
        <span class="k">if</span> <span class="n">n_threads</span> <span class="o">&lt;</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">N_THREAD_OPTIMAL</span><span class="p">:</span>
            <span class="n">session</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Using low n_threads value: </span><span class="si">{</span><span class="n">n_threads</span><span class="si">}</span><span class="s2">. This may lead to sub-optimal performance.&quot;</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">n_threads</span> <span class="o">&gt;</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">N_THREAD_MAX</span><span class="p">:</span>
            <span class="n">session</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Using high n_threads value: </span><span class="si">{</span><span class="n">n_threads</span><span class="si">}</span><span class="s2">, this may lead to overloading spark driver. Limiting to 32.&quot;</span>
            <span class="p">)</span>
            <span class="n">n_threads</span> <span class="o">=</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">N_THREAD_MAX</span>

        <span class="k">def</span><span class="w"> </span><span class="nf">process_one</span><span class="p">(</span>
            <span class="n">input_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">session</span><span class="p">:</span> <span class="n">Session</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="nb">str</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;Function to process one finngen-ukbb-mvp summary statistics file to schema superset.</span>

<span class="sd">            Args:</span>
<span class="sd">                input_path (str): Input path to the gzipped summary statistics file.</span>
<span class="sd">                session (Session): Session object.</span>
<span class="sd">                output_path (str): Output path for the Parquet files.</span>

<span class="sd">            Returns:</span>
<span class="sd">                DataFrame: Processed dataframe.</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">session</span><span class="o">.</span><span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span>
                <span class="n">input_path</span><span class="p">,</span>
                <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">inferSchema</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">enforceSchema</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="c1"># Apply schema enforcement by selecting columns in the expected order with proper types</span>
            <span class="c1"># NOTE: Here we only add missing columns with null values casted to expected type</span>
            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">cls</span><span class="o">.</span><span class="n">raw_schema</span><span class="o">.</span><span class="n">names</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">c</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="bp">cls</span><span class="o">.</span><span class="n">raw_schema</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">dataType</span><span class="p">))</span>

            <span class="c1"># Replace all NA with nulls and cast to the expected type</span>
            <span class="c1"># NOTE: Here we apply the transformation on full schema set (including added missing columns)</span>
            <span class="c1"># NOTE: If we do not transform the `NA` to nulls, we will still have divergent schemas, as `NA` forces the column to StringType</span>
            <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="bp">cls</span><span class="o">.</span><span class="n">raw_schema</span><span class="o">.</span><span class="n">fields</span><span class="p">:</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="n">field</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">field</span><span class="o">.</span><span class="n">name</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">field</span><span class="o">.</span><span class="n">name</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">field</span><span class="o">.</span><span class="n">dataType</span><span class="p">),</span>
                <span class="p">)</span>
            <span class="c1"># Add studyId based on the input path</span>
            <span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">concat_ws</span><span class="p">(</span>
                        <span class="s2">&quot;_&quot;</span><span class="p">,</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="n">datasource</span><span class="o">.</span><span class="n">value</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span>
                            <span class="bp">cls</span><span class="o">.</span><span class="n">extract_study_phenotype_from_path</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">input_file_name</span><span class="p">())</span>
                        <span class="p">),</span>
                    <span class="p">),</span>
                    <span class="c1"># Optimal partition size is ~ 100MB, assuming the total size of the dataset is 2Tb</span>
                    <span class="c1"># we can have up to 60 partitions per study (330 studies)</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">orderBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;#CHR&quot;</span><span class="p">,</span> <span class="s2">&quot;POS&quot;</span><span class="p">)</span>
                <span class="o">.</span><span class="n">repartition</span><span class="p">(</span><span class="mi">60</span><span class="p">,</span> <span class="s2">&quot;#CHR&quot;</span><span class="p">,</span> <span class="s2">&quot;POS&quot;</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="c1"># Write out the processed dataframe to Parquet</span>
            <span class="c1"># NOTE: Write is done per studyId partition from the thread pool to</span>
            <span class="c1"># make sure we do not need to collect all data after the thread execution.</span>
            <span class="n">df</span><span class="o">.</span><span class="n">write</span><span class="o">.</span><span class="n">mode</span><span class="p">(</span><span class="s2">&quot;append&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">partitionBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">parquet</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">df</span>

        <span class="n">session</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Converting gzipped summary statistics from </span><span class="si">{</span><span class="n">summary_statistics_list</span><span class="si">}</span><span class="s2"> to Parquet at </span><span class="si">{</span><span class="n">raw_summary_statistics_output_path</span><span class="si">}</span><span class="s2">.&quot;</span>
        <span class="p">)</span>
        <span class="k">with</span> <span class="n">ThreadPoolExecutor</span><span class="p">(</span><span class="n">max_workers</span><span class="o">=</span><span class="n">n_threads</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
            <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">path</span><span class="p">:</span> <span class="n">process_one</span><span class="p">(</span>
                    <span class="n">path</span><span class="p">,</span>
                    <span class="n">session</span><span class="o">=</span><span class="n">session</span><span class="p">,</span>
                    <span class="n">output_path</span><span class="o">=</span><span class="n">raw_summary_statistics_output_path</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="n">summary_statistics_list</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">from_source</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">,</span>
        <span class="n">raw_summary_statistics</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span>
        <span class="n">finngen_manifest</span><span class="p">:</span> <span class="n">FinnGenMetaManifest</span><span class="p">,</span>
        <span class="n">variant_annotations</span><span class="p">:</span> <span class="n">VariantDirection</span><span class="p">,</span>
        <span class="n">perform_meta_analysis_filter</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">imputation_score_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.8</span><span class="p">,</span>
        <span class="n">perform_imputation_score_filter</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">min_allele_count_threshold</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
        <span class="n">perform_min_allele_count_filter</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">min_allele_frequency_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1e-4</span><span class="p">,</span>
        <span class="n">perform_min_allele_frequency_filter</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">filter_out_ambiguous_variants</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SummaryStatistics</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Build the summary statistics dataset from raw summary statistics.</span>

<span class="sd">        See original issue to find out more details on the harmonisation logic https://github.com/opentargets/issues/issues/3474</span>

<span class="sd">        ??? note &quot;The logic behind the harmonisation&quot;</span>
<span class="sd">            1. Build a slice of FinnGen Manifest to bring the information about nCases and nSamples per cohort (broadcast join).</span>
<span class="sd">            2. Build a variant direction (gnomAD) dataset partitioned by `chromosome` and `variantId` for joining using Sort-Merge strategy.</span>
<span class="sd">            3. Select required columns from raw summary statistics</span>
<span class="sd">            4. Remove all non-meta analyzed variants (nBiobanks &lt; 2) - by default</span>
<span class="sd">            5. Remove all variants with low imputation score - by default</span>
<span class="sd">            6. Join with Finngen Manifest and Variant Direction datasets</span>
<span class="sd">            7. Flip `beta` and `min allele frequency` based on the `direction` from Variant Direction dataset</span>
<span class="sd">            8. Use originalVariantId from Variant Direction dataset if available, otherwise fall back to variantId (for variants missing from Variant Direction)</span>
<span class="sd">            9. Calculate combined effect allele frequency from cohorts</span>
<span class="sd">            10. Remove all variants with low Min Allele Count - by default</span>
<span class="sd">            11. Remove all variants with low Min Allele Frequency - optional</span>
<span class="sd">            12. Remove strand ambiguous variants - optional</span>

<span class="sd">        ??? tip &quot;Variant Directionality&quot;</span>
<span class="sd">            **Variant Direction**</span>
<span class="sd">            By default we:</span>

<span class="sd">            1. keep all strand ambiguous variants as is</span>
<span class="sd">            2. keep all variants found in gnomAD aligned to gnomAD reference ( if alleles are flipped we flip the beta and allele frequency)</span>
<span class="sd">            3. keep all variants not found in gnomAD as is (cannot determine strand or alignment) - we assume these are correct.</span>

<span class="sd">        ??? note &quot;Important considerations&quot;</span>
<span class="sd">            * The input summary statistics are expected to be already parquet formatted and partitioned by `studyId`.</span>
<span class="sd">            * Both `perform_allele_count_filter` and `perform_allele_frequency_filter` are redundant, if both</span>
<span class="sd">            are set to True, only `perform_allele_count_filter` will be applied.</span>
<span class="sd">            * Threshold value for Min Allele Count &gt;= 20 means that with a MAF of 1e-4 we would expect to see</span>
<span class="sd">            at least 1_000 samples with minor allele in a cohort. This is quiet stringent for very rare variants and</span>
<span class="sd">            smaller cohorts, like MVP_AFR (nSamples ~120k) and MVP_HIS (~60k).</span>
<span class="sd">            * MVP_HIS cohort has been mapped to admixed American population - see https://www.science.org/doi/10.1126/science.adj1182 for more details.</span>

<span class="sd">        Args:</span>
<span class="sd">            raw_summary_statistics (DataFrame): Raw summary statistics dataframe.</span>
<span class="sd">            finngen_manifest (FinnGenMetaManifest): FinnGen meta analysis manifest.</span>
<span class="sd">            variant_annotations (VariantDirection): Variant direction dataset.</span>
<span class="sd">            perform_meta_analysis_filter (bool): Whether to remove variants found in just 1 biobank. Default is True.</span>
<span class="sd">            imputation_score_threshold (float): Imputation score threshold for MVP cohorts. Default is 0.8.</span>
<span class="sd">            perform_imputation_score_filter (bool): Whether to perform the imputation score filter. Default is True.</span>
<span class="sd">            min_allele_count_threshold (int): Minimum allele count threshold. Default is 20.</span>
<span class="sd">            perform_min_allele_count_filter (bool): Whether to perform the minimum allele count filter. Default is False.</span>
<span class="sd">            min_allele_frequency_threshold (float): Minimum allele frequency threshold. Default is 1e-4.</span>
<span class="sd">            perform_min_allele_frequency_filter (bool): Whether to perform the minimum allele frequency filter. Default is True.</span>
<span class="sd">            filter_out_ambiguous_variants (bool): Whether to filter out strand ambiguous variants. Default is False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            SummaryStatistics: Processed summary statistics dataset.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">perform_min_allele_count_filter</span><span class="p">:</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="n">min_allele_count_threshold</span> <span class="o">&gt;</span> <span class="mi">0</span>
            <span class="p">),</span> <span class="s2">&quot;Allele count threshold should be positive.&quot;</span>
        <span class="k">if</span> <span class="n">perform_min_allele_frequency_filter</span><span class="p">:</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="mf">0.0</span> <span class="o">&lt;=</span> <span class="n">min_allele_frequency_threshold</span> <span class="o">&lt;=</span> <span class="mf">0.5</span>
            <span class="p">),</span> <span class="s2">&quot;MAF needs to be between 0 and 0.5.&quot;</span>

        <span class="k">if</span> <span class="n">perform_min_allele_count_filter</span> <span class="ow">and</span> <span class="n">perform_min_allele_frequency_filter</span><span class="p">:</span>
            <span class="c1"># NOTE - MAC filter would be more stringent at low allele frequencies, so no</span>
            <span class="c1"># need to have both filters active at the same time</span>
            <span class="n">perform_min_allele_frequency_filter</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="n">si_slice</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">broadcast</span><span class="p">(</span>
            <span class="n">finngen_manifest</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nCasesPerCohort&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nSamples&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nSamplesPerCohort&quot;</span><span class="p">),</span>
            <span class="p">)</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">vd_slice</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">variant_annotations</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;originalVariantId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;direction&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;isStrandAmbiguous&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># NOTE: repartition(&quot;chromosome&quot;) produces very uneven partitions,</span>
            <span class="c1"># Spark attempts then to fall back to `dynamic partitioning` algorithm</span>
            <span class="c1"># which fails after N failures.</span>
            <span class="o">.</span><span class="n">repartitionByRange</span><span class="p">(</span><span class="mi">4_000</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;variantId&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="n">sumstats</span> <span class="o">=</span> <span class="p">(</span>
            <span class="c1"># Pre-select columns that are needed downstream</span>
            <span class="c1"># NOTE: full set of columns is not required.</span>
            <span class="n">raw_summary_statistics</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;#CHR&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;POS&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;REF&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ALT&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;fg_af_alt&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_r2&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_af_alt&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_r2&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_af_alt&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_r2&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_af_alt&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ukbb_af_alt&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_mlogp&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_beta&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_sebeta&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="c1"># NOTE: make sure the chromosome is coded to 1:22, X, Y</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
                <span class="n">normalize_chromosome</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;#CHR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">())),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;#CHR&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;POS&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">IntegerType</span><span class="p">()))</span>
            <span class="o">.</span><span class="n">withColumnRenamed</span><span class="p">(</span><span class="s2">&quot;REF&quot;</span><span class="p">,</span> <span class="s2">&quot;referenceAllele&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">withColumnRenamed</span><span class="p">(</span><span class="s2">&quot;ALT&quot;</span><span class="p">,</span> <span class="s2">&quot;alternateAllele&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;neglogpval&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_mlogp&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">())</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_beta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">()))</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;standardError&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_sebeta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">())</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;variantId&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">concat_ws</span><span class="p">(</span>
                    <span class="s2">&quot;_&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;referenceAllele&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;alternateAllele&quot;</span><span class="p">),</span>
                <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;referenceAllele&quot;</span><span class="p">,</span> <span class="s2">&quot;alternateAllele&quot;</span><span class="p">)</span>
            <span class="c1"># Initial filters based on statistics presence</span>
            <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;neglogpval&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">())</span>
            <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">())</span>
            <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">())</span>
        <span class="p">)</span>

        <span class="c1"># Filter out variants that are not meta analyzed (nBiobanks &lt; 1)</span>
        <span class="k">if</span> <span class="n">perform_meta_analysis_filter</span><span class="p">:</span>
            <span class="n">sumstats</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">sumstats</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;cohorts&quot;</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">cohorts</span><span class="p">())</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;isMetaAnalyzedVariant&quot;</span><span class="p">,</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">is_meta_analyzed_variant</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohorts&quot;</span><span class="p">)),</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;isMetaAnalyzedVariant&quot;</span><span class="p">))</span>
                <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;isMetaAnalyzedVariant&quot;</span><span class="p">,</span> <span class="s2">&quot;cohorts&quot;</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="c1"># Filter out variants with low INFO score</span>
        <span class="k">if</span> <span class="n">perform_imputation_score_filter</span><span class="p">:</span>
            <span class="k">assert</span> <span class="p">(</span>
                <span class="n">imputation_score_threshold</span> <span class="o">&gt;=</span> <span class="mf">0.0</span>
            <span class="p">),</span> <span class="s2">&quot;Imputation score threshold should be positive.&quot;</span>
            <span class="n">sumstats</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">sumstats</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;hasLowImputationScore&quot;</span><span class="p">,</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">has_low_imputation_score</span><span class="p">(</span><span class="n">imputation_score_threshold</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasLowImputationScore&quot;</span><span class="p">))</span>
                <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;hasLowImputationScore&quot;</span><span class="p">,</span> <span class="s2">&quot;imputationScore&quot;</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="n">sumstats</span> <span class="o">=</span> <span class="p">(</span>
            <span class="c1"># Annotate with StudyIndex nCases, nSamples to obtain</span>
            <span class="c1"># the cases Minor Allele Count and Samples for combined AF calculation</span>
            <span class="n">sumstats</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">si_slice</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
            <span class="c1"># Join with variant direction dataset</span>
            <span class="c1"># Keep variants if not found in gnomAD (left join)</span>
            <span class="o">.</span><span class="n">repartitionByRange</span><span class="p">(</span><span class="mi">4_000</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;variantId&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">vd_slice</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;variantId&quot;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
            <span class="c1"># Use originalVariantId (already flipped) or fall back to variantId if not found in gnomAD</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;originalVariantId&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">))</span>
            <span class="p">)</span>
            <span class="c1"># Compute allele frequency per cohort and align with direction</span>
            <span class="c1"># NOTE: `direction` column represents if variant aligned to gnomAD variant or it&#39;s flipped</span>
            <span class="c1"># version, the values are `1` - direct, `-1` - flipped</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;cohortAlleleFrequency&quot;</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">allele_frequencies</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;direction&quot;</span><span class="p">))</span>
            <span class="p">)</span>
            <span class="c1"># Make sure the beta is alined with the direction, if direction is null, keep beta as is</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;beta&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">)</span> <span class="o">*</span> <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;direction&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
            <span class="p">)</span>
            <span class="c1"># Calculate the combined effect allele frequency from cohorts</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;effectAlleleFrequencyFromSource&quot;</span><span class="p">,</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">combined_allele_frequency</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohortAlleleFrequency&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nSamplesPerCohort&quot;</span><span class="p">)</span>
                <span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="c1"># Remove strand ambiguous variants, if not found in gnomAD, we keep the variant</span>
        <span class="c1"># Not run by default</span>
        <span class="k">if</span> <span class="n">filter_out_ambiguous_variants</span><span class="p">:</span>
            <span class="n">sumstats</span> <span class="o">=</span> <span class="n">sumstats</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;isStrandAmbiguous&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">perform_min_allele_count_filter</span> <span class="ow">or</span> <span class="n">perform_min_allele_frequency_filter</span><span class="p">:</span>
            <span class="c1"># Calculate the MAF per cohort</span>
            <span class="n">sumstats</span> <span class="o">=</span> <span class="n">sumstats</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;cohortMinAlleleFrequency&quot;</span><span class="p">,</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">min_allele_frequency</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohortAlleleFrequency&quot;</span><span class="p">)),</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">perform_min_allele_count_filter</span><span class="p">:</span>
            <span class="n">sumstats</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">sumstats</span>
                <span class="c1"># Make sure to only keep cohorts that have nSamples &gt; 0</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;nSamplesPerCohort&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nSamplesPerCohort&quot;</span><span class="p">),</span>
                        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;nSamples&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
                        <span class="o">&amp;</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;nSamples&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">),</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;cohortMinAlleleCount&quot;</span><span class="p">,</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">min_allele_count</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohortMinAlleleFrequency&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nSamplesPerCohort&quot;</span><span class="p">),</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;hasLowMinAlleleCount&quot;</span><span class="p">,</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">has_low_min_allele_count</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohortMinAlleleCount&quot;</span><span class="p">),</span>
                        <span class="n">min_allele_count_threshold</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasLowMinAlleleCount&quot;</span><span class="p">))</span>
                <span class="o">.</span><span class="n">drop</span><span class="p">(</span>
                    <span class="s2">&quot;hasLowMinAlleleCount&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;cohortMinAlleleCount&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;nCasesPerCohort&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="c1"># Not run by default.</span>
        <span class="k">if</span> <span class="n">perform_min_allele_frequency_filter</span><span class="p">:</span>
            <span class="n">sumstats</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">sumstats</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                    <span class="s2">&quot;hasLowMinAlleleFrequency&quot;</span><span class="p">,</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">has_low_min_allele_frequency</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohortMinAlleleFrequency&quot;</span><span class="p">),</span>
                        <span class="n">min_allele_frequency_threshold</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasLowMinAlleleFrequency&quot;</span><span class="p">))</span>
                <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;hasLowMinAlleleFrequency&quot;</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="c1"># Convert to final summary statistics schema</span>
        <span class="n">sumstats</span> <span class="o">=</span> <span class="n">sumstats</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nSamples&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sampleSize&quot;</span><span class="p">),</span>
            <span class="o">*</span><span class="n">pvalue_from_neglogpval</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;neglogpval&quot;</span><span class="p">)),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;effectAlleleFrequencyFromSource&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sampleSize&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;effectAlleleFrequencyFromSource&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">),</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">SummaryStatistics</span><span class="p">(</span><span class="n">sumstats</span><span class="p">)</span><span class="o">.</span><span class="n">sanity_filter</span><span class="p">()</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">has_low_min_allele_frequency</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">,</span> <span class="n">maf</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1e-4</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find if variant has a low minor allele frequency in any cohort.</span>

<span class="sd">        Args:</span>
<span class="sd">            maf (Column): Column containing array of structs with `cohort` and `minAlleleFrequency` fields.</span>
<span class="sd">            threshold (float): Threshold below which the minor allele frequency is considered low. Default is 1e-4.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Boolean column indicating if any cohort has a low minor allele frequency.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; maf = {&quot;v1&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleFrequency&quot;: 0.0001}, {&quot;cohort&quot;: &quot;B&quot;, &quot;minAlleleFrequency&quot;: 0.0002}],</span>
<span class="sd">            ...         &quot;v2&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleFrequency&quot;: None}, {&quot;cohort&quot;: &quot;D&quot;, &quot;minAlleleFrequency&quot;: 0.15}],</span>
<span class="sd">            ...         &quot;v3&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleFrequency&quot;: 0.00001}, {&quot;cohort&quot;: &quot;B&quot;, &quot;minAlleleFrequency&quot;: 0.2}],}</span>
<span class="sd">            &gt;&gt;&gt; data = [(&quot;v1&quot;, maf[&quot;v1&quot;]),</span>
<span class="sd">            ...         (&quot;v2&quot;, maf[&quot;v2&quot;]),</span>
<span class="sd">            ...         (&quot;v3&quot;, maf[&quot;v3&quot;]),]</span>
<span class="sd">            &gt;&gt;&gt; schema = &quot;variantId STRING, cohortMinAlleleFrequency ARRAY&lt;STRUCT&lt;cohort: STRING, minAlleleFrequency: DOUBLE&gt;&gt;&quot;</span>
<span class="sd">            &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">            &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">            +---------+--------------------------+</span>
<span class="sd">            |variantId|cohortMinAlleleFrequency  |</span>
<span class="sd">            +---------+--------------------------+</span>
<span class="sd">            |v1       |[{A, 1.0E-4}, {B, 2.0E-4}]|</span>
<span class="sd">            |v2       |[{A, NULL}, {D, 0.15}]    |</span>
<span class="sd">            |v3       |[{A, 1.0E-5}, {B, 0.2}]   |</span>
<span class="sd">            +---------+--------------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>

<span class="sd">            &gt;&gt;&gt; df = df.withColumn(&quot;hasMinAlleleFrequency&quot;, FinnGenUkbMvpMetaSummaryStatistics.has_low_min_allele_frequency(f.col(&quot;cohortMinAlleleFrequency&quot;)))</span>
<span class="sd">            &gt;&gt;&gt; df.select(&quot;variantId&quot;, &quot;hasMinAlleleFrequency&quot;).show(truncate=False)</span>
<span class="sd">            +---------+---------------------+</span>
<span class="sd">            |variantId|hasMinAlleleFrequency|</span>
<span class="sd">            +---------+---------------------+</span>
<span class="sd">            |v1       |false                |</span>
<span class="sd">            |v2       |false                |</span>
<span class="sd">            |v3       |true                 |</span>
<span class="sd">            +---------+---------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">non_empty_maf</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">maf</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="n">n_cohorts_with_maf_below_threshold</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">non_empty_maf</span><span class="p">,</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">)</span>
                <span class="o">&lt;</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="n">threshold</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DecimalType</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span> <span class="mi">10</span><span class="p">)),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">n_cohorts_with_maf_below_threshold</span> <span class="o">&gt;</span> <span class="mi">0</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">normalize_af</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">af</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">flip</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Normalize allele frequency based on variant direction.</span>

<span class="sd">        Args:</span>
<span class="sd">            af (Column): Allele frequency column.</span>
<span class="sd">            flip (Column): Direction column indicating if the allele needs to be flipped.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Normalized allele frequency.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; data = [(&quot;v1&quot;, 0.1, 1),</span>
<span class="sd">            ...       (&quot;v2&quot;, 0.2, -1),</span>
<span class="sd">            ...       (&quot;v3&quot;, None, -1),</span>
<span class="sd">            ...       (&quot;V4&quot;, 0.1, None),]</span>
<span class="sd">            &gt;&gt;&gt; schema = &quot;variantId STRING, af DOUBLE, flip INT&quot;</span>
<span class="sd">            &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">            &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">            +---------+----+----+</span>
<span class="sd">            |variantId|af  |flip|</span>
<span class="sd">            +---------+----+----+</span>
<span class="sd">            |v1       |0.1 |1   |</span>
<span class="sd">            |v2       |0.2 |-1  |</span>
<span class="sd">            |v3       |NULL|-1  |</span>
<span class="sd">            |V4       |0.1 |NULL|</span>
<span class="sd">            +---------+----+----+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>

<span class="sd">            &gt;&gt;&gt; df = df.withColumn(&quot;normalizedAf&quot;, FinnGenUkbMvpMetaSummaryStatistics.normalize_af(f.col(&quot;af&quot;), f.col(&quot;flip&quot;)))</span>
<span class="sd">            &gt;&gt;&gt; df.select(&quot;variantId&quot;, &quot;normalizedAf&quot;).show(truncate=False)</span>
<span class="sd">            +---------+------------+</span>
<span class="sd">            |variantId|normalizedAf|</span>
<span class="sd">            +---------+------------+</span>
<span class="sd">            |v1       |0.1         |</span>
<span class="sd">            |v2       |0.8         |</span>
<span class="sd">            |v3       |NULL        |</span>
<span class="sd">            |V4       |0.1         |</span>
<span class="sd">            +---------+------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">((</span><span class="n">flip</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">af</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">1.0</span><span class="p">)</span> <span class="o">-</span> <span class="n">af</span><span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">af</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">allele_frequencies</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">flip</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">scale</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Extract the allele frequencies per cohort.</span>

<span class="sd">        Note:</span>
<span class="sd">            if the `flip` column is -1, then the allele frequency is flipped (1 - af).</span>

<span class="sd">        Args:</span>
<span class="sd">            flip (Column): Direction column indicating if the allele needs to be flipped. (-1 for flip, 1 for no flip, null for no information)</span>
<span class="sd">            scale (int): Scale for the decimal type conversion. Default is 10.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Column containing array of structs with `cohort` and `alleleFrequency` fields.</span>


<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; data = [(&quot;v1&quot;, 0.1, 0.2, None, 0.3, 0.4, -1),</span>
<span class="sd">            ...        (&quot;v2&quot;, 0.000000001, 0.999999999, None, None, None,1),</span>
<span class="sd">            ...        (&quot;v3&quot;, 0.1, 0.1, None, None, None, None),]</span>
<span class="sd">            &gt;&gt;&gt; schema = &quot;variantId STRING, MVP_EUR_af_alt DOUBLE, MVP_AFR_af_alt DOUBLE, MVP_HIS_af_alt DOUBLE, fg_af_alt DOUBLE, ukbb_af_alt DOUBLE, flip INT&quot;</span>
<span class="sd">            &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">            &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">            +---------+--------------+--------------+--------------+---------+-----------+----+</span>
<span class="sd">            |variantId|MVP_EUR_af_alt|MVP_AFR_af_alt|MVP_HIS_af_alt|fg_af_alt|ukbb_af_alt|flip|</span>
<span class="sd">            +---------+--------------+--------------+--------------+---------+-----------+----+</span>
<span class="sd">            |v1       |0.1           |0.2           |NULL          |0.3      |0.4        |-1  |</span>
<span class="sd">            |v2       |1.0E-9        |0.999999999   |NULL          |NULL     |NULL       |1   |</span>
<span class="sd">            |v3       |0.1           |0.1           |NULL          |NULL     |NULL       |NULL|</span>
<span class="sd">            +---------+--------------+--------------+--------------+---------+-----------+----+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>

<span class="sd">            &gt;&gt;&gt; df = df.withColumn(&quot;alleleFrequencies&quot;, FinnGenUkbMvpMetaSummaryStatistics.allele_frequencies(f.col(&quot;flip&quot;)))</span>
<span class="sd">            &gt;&gt;&gt; df.select(&quot;alleleFrequencies&quot;).show(truncate=False)</span>
<span class="sd">            +-------------------------------------------------------------------------------------------------+</span>
<span class="sd">            |alleleFrequencies                                                                                |</span>
<span class="sd">            +-------------------------------------------------------------------------------------------------+</span>
<span class="sd">            |[{MVP_EUR, 0.9000000000}, {MVP_AFR, 0.8000000000}, {FinnGen, 0.7000000000}, {UKBB, 0.6000000000}]|</span>
<span class="sd">            |[{MVP_EUR, 0.0000000010}, {MVP_AFR, 0.9999999990}]                                               |</span>
<span class="sd">            |[{MVP_EUR, 0.1000000000}, {MVP_AFR, 0.1000000000}]                                               |</span>
<span class="sd">            +-------------------------------------------------------------------------------------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">precision</span> <span class="o">=</span> <span class="n">scale</span> <span class="o">+</span> <span class="mi">1</span>  <span class="c1"># to ensure we can represent values like 1.0000</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_EUR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">normalize_af</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_af_alt&quot;</span><span class="p">),</span> <span class="n">flip</span><span class="p">)</span>
                    <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DecimalType</span><span class="p">(</span><span class="n">precision</span><span class="p">,</span> <span class="n">scale</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">),</span>
                <span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_AFR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">normalize_af</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_af_alt&quot;</span><span class="p">),</span> <span class="n">flip</span><span class="p">)</span>
                    <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DecimalType</span><span class="p">(</span><span class="n">precision</span><span class="p">,</span> <span class="n">scale</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">),</span>
                <span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_AMR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                    <span class="c1"># Note: HIS in sumstats is AMR in study index</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">normalize_af</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_af_alt&quot;</span><span class="p">),</span> <span class="n">flip</span><span class="p">)</span>
                    <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DecimalType</span><span class="p">(</span><span class="n">precision</span><span class="p">,</span> <span class="n">scale</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">),</span>
                <span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;FinnGen&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">normalize_af</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;fg_af_alt&quot;</span><span class="p">),</span> <span class="n">flip</span><span class="p">)</span>
                    <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DecimalType</span><span class="p">(</span><span class="n">precision</span><span class="p">,</span> <span class="n">scale</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">),</span>
                <span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;UKBB&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">normalize_af</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ukbb_af_alt&quot;</span><span class="p">),</span> <span class="n">flip</span><span class="p">)</span>
                    <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DecimalType</span><span class="p">(</span><span class="n">precision</span><span class="p">,</span> <span class="n">scale</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">),</span>
                <span class="p">),</span>
            <span class="p">),</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alleleFrequencies&quot;</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">combined_allele_frequency</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">,</span> <span class="n">allele_freq</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">n_samples_per_cohort</span><span class="p">:</span> <span class="n">Column</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Combined Allele Frequency across all cohorts.</span>

<span class="sd">        Args:</span>
<span class="sd">            allele_freq (Column): Column containing array of structs with `cohort` and `alleleFrequency` fields.</span>
<span class="sd">            n_samples_per_cohort (Column): Column containing array of structs with `cohort` and `nSamples` fields.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Combined allele frequency across all cohorts.</span>

<span class="sd">        Note:</span>
<span class="sd">            The combination is made by weighting the allele frequencies by the number of samples in each cohort.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; data = [</span>
<span class="sd">            ...    (&quot;v1&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;alleleFrequency&quot;: 0.6}, {&quot;cohort&quot;: &quot;B&quot;, &quot;alleleFrequency&quot;: 0.2}, {&quot;cohort&quot;: &quot;C&quot;, &quot;alleleFrequency&quot;: 0.3}],</span>
<span class="sd">            ...           [{&quot;cohort&quot;: &quot;A&quot;, &quot;nSamples&quot;: 100}, {&quot;cohort&quot;: &quot;B&quot;, &quot;nSamples&quot;: 200}, {&quot;cohort&quot;: &quot;D&quot;, &quot;nSamples&quot;: 20}]),</span>
<span class="sd">            ...    (&quot;v2&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;alleleFrequency&quot;: None},], [{&quot;cohort&quot;: &quot;A&quot;, &quot;nSamples&quot;: 50}]),</span>
<span class="sd">            ...    (&quot;v3&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;alleleFrequency&quot;: 0.05},], [{&quot;cohort&quot;: &quot;A&quot;, &quot;nSamples&quot;: None}]),]</span>
<span class="sd">            &gt;&gt;&gt; schema = &quot;variantId STRING, alleleFrequencies ARRAY&lt;STRUCT&lt;cohort: STRING, alleleFrequency: DOUBLE&gt;&gt;, nSamplesPerCohort ARRAY&lt;STRUCT&lt;cohort: STRING, nSamples: INT&gt;&gt;&quot;</span>
<span class="sd">            &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">            &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">            +---------+------------------------------+-----------------------------+</span>
<span class="sd">            |variantId|alleleFrequencies             |nSamplesPerCohort            |</span>
<span class="sd">            +---------+------------------------------+-----------------------------+</span>
<span class="sd">            |v1       |[{A, 0.6}, {B, 0.2}, {C, 0.3}]|[{A, 100}, {B, 200}, {D, 20}]|</span>
<span class="sd">            |v2       |[{A, NULL}]                   |[{A, 50}]                    |</span>
<span class="sd">            |v3       |[{A, 0.05}]                   |[{A, NULL}]                  |</span>
<span class="sd">            +---------+------------------------------+-----------------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>

<span class="sd">            &gt;&gt;&gt; df = df.withColumn(&quot;combinedAlleleFrequency&quot;, FinnGenUkbMvpMetaSummaryStatistics.combined_allele_frequency(f.col(&quot;alleleFrequencies&quot;), f.col(&quot;nSamplesPerCohort&quot;)))</span>
<span class="sd">            &gt;&gt;&gt; df.select(&quot;variantId&quot;, f.round(&quot;combinedAlleleFrequency&quot;, 2).alias(&quot;caf&quot;)).show(truncate=False)</span>
<span class="sd">            +---------+----+</span>
<span class="sd">            |variantId|caf |</span>
<span class="sd">            +---------+----+</span>
<span class="sd">            |v1       |0.33|</span>
<span class="sd">            |v2       |NULL|</span>
<span class="sd">            |v3       |NULL|</span>
<span class="sd">            +---------+----+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">af_filtered</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">allele_freq</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">samples_filtered</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">n_samples_per_cohort</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;nSamples&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="n">intersect</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">array_intersect</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">af_filtered</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">)),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">samples_filtered</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">)),</span>
        <span class="p">)</span>
        <span class="n">af_map</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">map_from_entries</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">af_filtered</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span><span class="n">intersect</span><span class="p">,</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">))</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">n_samples_map</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">map_from_entries</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">samples_filtered</span><span class="p">,</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span><span class="n">intersect</span><span class="p">,</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">)),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="c1"># Compute numerator: sum(AF * n)</span>
        <span class="n">af_times_n</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">map_entries</span><span class="p">(</span><span class="n">af_map</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">0.0</span><span class="p">),</span>
            <span class="k">lambda</span> <span class="n">acc</span><span class="p">,</span> <span class="n">kv</span><span class="p">:</span> <span class="n">acc</span>
            <span class="o">+</span> <span class="n">kv</span><span class="p">[</span><span class="s2">&quot;value&quot;</span><span class="p">]</span>
            <span class="o">*</span> <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">element_at</span><span class="p">(</span><span class="n">n_samples_map</span><span class="p">,</span> <span class="n">kv</span><span class="p">[</span><span class="s2">&quot;key&quot;</span><span class="p">]),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)),</span>
        <span class="p">)</span>

        <span class="c1"># Compute denominator: sum(n)</span>
        <span class="n">n_total</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">map_entries</span><span class="p">(</span><span class="n">n_samples_map</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
            <span class="k">lambda</span> <span class="n">acc</span><span class="p">,</span> <span class="n">kv</span><span class="p">:</span> <span class="n">acc</span> <span class="o">+</span> <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">kv</span><span class="p">[</span><span class="s2">&quot;value&quot;</span><span class="p">],</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mi">0</span><span class="p">)),</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">((</span><span class="n">n_total</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">af_times_n</span> <span class="o">==</span> <span class="mi">0</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">))</span>
            <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">af_times_n</span> <span class="o">/</span> <span class="n">n_total</span><span class="p">)</span>
            <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">FloatType</span><span class="p">())</span>
        <span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">min_allele_frequency</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">allele_freq</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Minor Allele Frequency (MAF) per cohort.</span>

<span class="sd">        Note:</span>
<span class="sd">            The resulting value is of DecimalType(11, 10) to ensure precision for low frequency variants.</span>

<span class="sd">        Args:</span>
<span class="sd">            allele_freq (Column): Column containing array of structs with `cohort` and `alleleFrequency` fields.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Column containing array of structs with `cohort` and `minAlleleFrequency` fields.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; data = [(&quot;v1&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;alleleFrequency&quot;: 0.1}, {&quot;cohort&quot;: &quot;B&quot;, &quot;alleleFrequency&quot;: 0.7}]),]</span>
<span class="sd">            &gt;&gt;&gt; schema = &quot;variantId STRING, alleleFrequencies ARRAY&lt;STRUCT&lt;cohort: STRING, alleleFrequency: DOUBLE&gt;&gt;&quot;</span>
<span class="sd">            &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">            &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">            +---------+--------------------+</span>
<span class="sd">            |variantId|alleleFrequencies   |</span>
<span class="sd">            +---------+--------------------+</span>
<span class="sd">            |v1       |[{A, 0.1}, {B, 0.7}]|</span>
<span class="sd">            +---------+--------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>

<span class="sd">            &gt;&gt;&gt; df = df.withColumn(&quot;cohortMinAlleleFrequency&quot;, FinnGenUkbMvpMetaSummaryStatistics.min_allele_frequency(f.col(&quot;alleleFrequencies&quot;)))</span>
<span class="sd">            &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">            +---------+--------------------+--------------------------------------+</span>
<span class="sd">            |variantId|alleleFrequencies   |cohortMinAlleleFrequency              |</span>
<span class="sd">            +---------+--------------------+--------------------------------------+</span>
<span class="sd">            |v1       |[{A, 0.1}, {B, 0.7}]|[{A, 0.1000000000}, {B, 0.3000000000}]|</span>
<span class="sd">            +---------+--------------------+--------------------------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
            <span class="n">allele_freq</span><span class="p">,</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                <span class="n">maf</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">min_allele_count</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">,</span> <span class="n">cohort_min_allele_frequency</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">n_samples_per_cohort</span><span class="p">:</span> <span class="n">Column</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Minor Allele Count (MAC) per cohort.</span>

<span class="sd">        Note:</span>
<span class="sd">            If a cohort either do not have the maf or nCases, it will be dropped from the resulting MAC array.</span>

<span class="sd">        Args:</span>
<span class="sd">            cohort_min_allele_frequency (Column): Column containing array of structs with `cohort` and `minAlleleFrequency` fields.</span>
<span class="sd">            n_samples_per_cohort (Column): Column containing array of structs with `cohort` and `nSamples` fields.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Column containing array of structs with `cohort` and `minAlleleCount` fields.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; maf = {&quot;v1&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleFrequency&quot;: 0.1}, {&quot;cohort&quot;: &quot;B&quot;, &quot;minAlleleFrequency&quot;: 0.2}],</span>
<span class="sd">            ...         &quot;v2&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleFrequency&quot;: 0.05}, {&quot;cohort&quot;: &quot;D&quot;, &quot;minAlleleFrequency&quot;: 0.15}],</span>
<span class="sd">            ...         &quot;v3&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleFrequency&quot;: 0.01}, {&quot;cohort&quot;: &quot;B&quot;, &quot;minAlleleFrequency&quot;: 0.02}],}</span>
<span class="sd">            &gt;&gt;&gt; n_samples = {&quot;v1&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;nSamples&quot;: 100}, {&quot;cohort&quot;: &quot;B&quot;, &quot;nSamples&quot;: 200}],</span>
<span class="sd">            ...            &quot;v2&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;nSamples&quot;: 150}, {&quot;cohort&quot;: &quot;C&quot;, &quot;nSamples&quot;: 250}],</span>
<span class="sd">            ...            &quot;v3&quot;: [{&quot;cohort&quot;: &quot;C&quot;, &quot;nSamples&quot;: 50}, {&quot;cohort&quot;: &quot;D&quot;, &quot;nSamples&quot;: 80}],}</span>
<span class="sd">            &gt;&gt;&gt; data = [(&quot;v1&quot;, maf[&quot;v1&quot;], n_samples[&quot;v1&quot;]),</span>
<span class="sd">            ...         (&quot;v2&quot;, maf[&quot;v2&quot;], n_samples[&quot;v2&quot;]),</span>
<span class="sd">            ...         (&quot;v3&quot;, maf[&quot;v3&quot;], n_samples[&quot;v3&quot;]),]</span>
<span class="sd">            &gt;&gt;&gt; schema = &quot;variantId STRING, cohortMinAlleleFrequency ARRAY&lt;STRUCT&lt;cohort: STRING, minAlleleFrequency: DOUBLE&gt;&gt;, nSamplesPerCohort ARRAY&lt;STRUCT&lt;cohort: STRING, nSamples: INT&gt;&gt;&quot;</span>
<span class="sd">            &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">            &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">            +---------+------------------------+--------------------+</span>
<span class="sd">            |variantId|cohortMinAlleleFrequency|nSamplesPerCohort   |</span>
<span class="sd">            +---------+------------------------+--------------------+</span>
<span class="sd">            |v1       |[{A, 0.1}, {B, 0.2}]    |[{A, 100}, {B, 200}]|</span>
<span class="sd">            |v2       |[{A, 0.05}, {D, 0.15}]  |[{A, 150}, {C, 250}]|</span>
<span class="sd">            |v3       |[{A, 0.01}, {B, 0.02}]  |[{C, 50}, {D, 80}]  |</span>
<span class="sd">            +---------+------------------------+--------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">            &gt;&gt;&gt; df = df.withColumn(&quot;cohortMinAlleleCount&quot;, FinnGenUkbMvpMetaSummaryStatistics.min_allele_count(f.col(&quot;cohortMinAlleleFrequency&quot;), f.col(&quot;nSamplesPerCohort&quot;)))</span>
<span class="sd">            &gt;&gt;&gt; df.select(&quot;variantId&quot;, &quot;cohortMinAlleleCount&quot;).show(truncate=False)</span>
<span class="sd">            +---------+--------------------+</span>
<span class="sd">            |variantId|cohortMinAlleleCount|</span>
<span class="sd">            +---------+--------------------+</span>
<span class="sd">            |v1       |[{A, 20}, {B, 80}]  |</span>
<span class="sd">            |v2       |[{A, 15}]           |</span>
<span class="sd">            |v3       |[]                  |</span>
<span class="sd">            +---------+--------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">cohort_min_allele_frequency</span><span class="p">,</span>
                <span class="k">lambda</span> <span class="n">left</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span>
                    <span class="n">n_samples_per_cohort</span><span class="p">,</span>
                    <span class="k">lambda</span> <span class="n">right</span><span class="p">:</span> <span class="n">right</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">left</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                <span class="p">),</span>
            <span class="p">),</span>
            <span class="k">lambda</span> <span class="n">left</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">left</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                <span class="n">mac</span><span class="p">(</span>
                    <span class="n">left</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                        <span class="n">n_samples_per_cohort</span><span class="p">,</span>
                        <span class="k">lambda</span> <span class="n">right</span><span class="p">:</span> <span class="n">right</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">)</span>
                        <span class="o">==</span> <span class="n">left</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                    <span class="p">)</span>
                    <span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
                    <span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;nSamples&quot;</span><span class="p">),</span>
                <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;minAlleleCount&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">has_low_min_allele_count</span><span class="p">(</span>
        <span class="bp">cls</span><span class="p">,</span> <span class="n">min_allele_count</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">min_allele_count_threshold</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find if variant has a low minor allele count in any of the cohorts.</span>

<span class="sd">        Note:</span>
<span class="sd">            If any cohort has a minor allele count below the threshold, the variant is considered to have a low minor allele count.</span>


<span class="sd">        Args:</span>
<span class="sd">            min_allele_count (Column): Column containing array of structs with `cohort` and `minAlleleCount` fields.</span>
<span class="sd">            min_allele_count_threshold (int): Threshold below which the minor allele count is considered low.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Boolean column indicating if any cohort has a low minor allele count.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; data = [(&quot;v1&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleCount&quot;: 30}, {&quot;cohort&quot;: &quot;B&quot;, &quot;minAlleleCount&quot;: 25}]),</span>
<span class="sd">            ...         (&quot;v2&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleCount&quot;: 15}, {&quot;cohort&quot;: &quot;B&quot;, &quot;minAlleleCount&quot;: 25}]),</span>
<span class="sd">            ...         (&quot;v3&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleCount&quot;: 30}, {&quot;cohort&quot;: &quot;B&quot;, &quot;minAlleleCount&quot;: 10}]),</span>
<span class="sd">            ...         (&quot;v4&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleCount&quot;: 5},],)]</span>
<span class="sd">            &gt;&gt;&gt; schema = &quot;variantId STRING, cohortMinAlleleCount ARRAY&lt;STRUCT&lt;cohort: STRING, minAlleleCount: INT&gt;&gt;&quot;</span>
<span class="sd">            &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">            &gt;&gt;&gt; df = df.withColumn(&quot;hasLowMinAlleleCount&quot;, FinnGenUkbMvpMetaSummaryStatistics.has_low_min_allele_count(f.col(&quot;cohortMinAlleleCount&quot;), 20))</span>
<span class="sd">            &gt;&gt;&gt; df.select(&quot;variantId&quot;, &quot;hasLowMinAlleleCount&quot;).show()</span>
<span class="sd">            +---------+--------------------+</span>
<span class="sd">            |variantId|hasLowMinAlleleCount|</span>
<span class="sd">            +---------+--------------------+</span>
<span class="sd">            |       v1|               false|</span>
<span class="sd">            |       v2|                true|</span>
<span class="sd">            |       v3|                true|</span>
<span class="sd">            |       v4|                true|</span>
<span class="sd">            +---------+--------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">min_allele_count</span><span class="p">,</span>
                    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;minAlleleCount&quot;</span><span class="p">)</span>
                    <span class="o">&lt;</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="n">min_allele_count_threshold</span><span class="p">),</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;hasLowMinAlleleCount&quot;</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">has_low_imputation_score</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">imputation_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find if variant has a low r2 imputation score in any of the MVP cohorts.</span>

<span class="sd">        Note:</span>
<span class="sd">            A missing imputation score is considered as passing the threshold, since it means that the variant was not</span>
<span class="sd">            present in that cohort.</span>

<span class="sd">        Note:</span>
<span class="sd">            If any r2 imputation score is below the threshold, the variant is considered to have a low imputation score.</span>

<span class="sd">        Args:</span>
<span class="sd">            imputation_threshold (float): Threshold below which the imputation score is considered low.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Boolean column indicating if any cohort has a low imputation score.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; data = [(&quot;v1&quot;, 0.9, 0.8, 1.0), (&quot;v2&quot;,0.7, 0.9, 0.9), (&quot;v3&quot;, None, None, 0.8), (&quot;v4&quot;, None, None, 0.7)]</span>
<span class="sd">            &gt;&gt;&gt; schema = &quot;variantId STRING, MVP_EUR_r2 DOUBLE, MVP_AFR_r2 DOUBLE, MVP_HIS_r2 DOUBLE&quot;</span>
<span class="sd">            &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">            &gt;&gt;&gt; df = df.withColumn(&quot;hasLowImputationScore&quot;, FinnGenUkbMvpMetaSummaryStatistics.has_low_imputation_score(0.8))</span>
<span class="sd">            &gt;&gt;&gt; df.select(&quot;variantId&quot;, &quot;hasLowImputationScore&quot;).show()</span>
<span class="sd">            +---------+---------------------+</span>
<span class="sd">            |variantId|hasLowImputationScore|</span>
<span class="sd">            +---------+---------------------+</span>
<span class="sd">            |       v1|                false|</span>
<span class="sd">            |       v2|                 true|</span>
<span class="sd">            |       v3|                false|</span>
<span class="sd">            |       v4|                 true|</span>
<span class="sd">            +---------+---------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_r2&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;r2&quot;</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_EUR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_r2&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">()</span>
                                <span class="o">|</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_r2&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">imputation_threshold</span><span class="p">),</span>
                                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">),</span>
                            <span class="p">)</span>
                            <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                            <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;filter&quot;</span><span class="p">),</span>
                        <span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_r2&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;r2&quot;</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_AFR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_r2&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">()</span>
                                <span class="o">|</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_r2&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">imputation_threshold</span><span class="p">),</span>
                                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">),</span>
                            <span class="p">)</span>
                            <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                            <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;filter&quot;</span><span class="p">),</span>
                        <span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_r2&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;r2&quot;</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_AMR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_r2&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">()</span>
                                <span class="o">|</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_r2&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">imputation_threshold</span><span class="p">),</span>
                                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">),</span>
                            <span class="p">)</span>
                            <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                            <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;filter&quot;</span><span class="p">),</span>
                        <span class="p">),</span>
                    <span class="p">),</span>
                    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="o">~</span><span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;filter&quot;</span><span class="p">),</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;hasLowImputationScore&quot;</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">cohorts</span><span class="p">(</span><span class="bp">cls</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Cohorts involved in the meta-analysis.</span>

<span class="sd">        This method creates an array of structs containing biobank and cohort information</span>
<span class="sd">        for variants that have allele frequency data available in each respective cohort.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Array of structs with fields &#39;biobank&#39; and &#39;cohort&#39;.</span>

<span class="sd">        Examples:</span>
<span class="sd">            # Test case 1: All cohorts have data</span>
<span class="sd">            &gt;&gt;&gt; data1 = [(0.3, 0.2, 0.4, 0.1, 0.25)]</span>
<span class="sd">            &gt;&gt;&gt; schema1 = &quot;MVP_EUR_af_alt DOUBLE, MVP_AFR_af_alt DOUBLE, MVP_HIS_af_alt DOUBLE, fg_af_alt DOUBLE, ukbb_af_alt DOUBLE&quot;</span>
<span class="sd">            &gt;&gt;&gt; df1 = spark.createDataFrame(data1, schema1)</span>
<span class="sd">            &gt;&gt;&gt; df1.withColumn(&quot;cohorts&quot;, FinnGenUkbMvpMetaSummaryStatistics.cohorts()).select(&quot;cohorts&quot;).show(truncate=False)</span>
<span class="sd">            +----------------------------------------------------------------------------------+</span>
<span class="sd">            |cohorts                                                                           |</span>
<span class="sd">            +----------------------------------------------------------------------------------+</span>
<span class="sd">            |[{MVP, MVP_EUR}, {MVP, MVP_AFR}, {MVP, MVP_AMR}, {FinnGen, FinnGen}, {UKBB, UKBB}]|</span>
<span class="sd">            +----------------------------------------------------------------------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>

<span class="sd">            # Test case 2: Only some cohorts have data</span>
<span class="sd">            &gt;&gt;&gt; data2 = [(0.3, None, None, 0.1, None)]</span>
<span class="sd">            &gt;&gt;&gt; df2 = spark.createDataFrame(data2, schema1)</span>
<span class="sd">            &gt;&gt;&gt; df2.withColumn(&quot;cohorts&quot;, FinnGenUkbMvpMetaSummaryStatistics.cohorts()).select(&quot;cohorts&quot;).show(truncate=False)</span>
<span class="sd">            +------------------------------------+</span>
<span class="sd">            |cohorts                             |</span>
<span class="sd">            +------------------------------------+</span>
<span class="sd">            |[{MVP, MVP_EUR}, {FinnGen, FinnGen}]|</span>
<span class="sd">            +------------------------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_af_alt&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inCohort&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_EUR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;biobank&quot;</span><span class="p">),</span>
                    <span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_af_alt&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inCohort&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_AFR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;biobank&quot;</span><span class="p">),</span>
                    <span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_af_alt&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inCohort&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_AMR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;biobank&quot;</span><span class="p">),</span>
                    <span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;fg_af_alt&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inCohort&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;FinnGen&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;FinnGen&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;biobank&quot;</span><span class="p">),</span>
                    <span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ukbb_af_alt&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inCohort&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;UKBB&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;UKBB&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;biobank&quot;</span><span class="p">),</span>
                    <span class="p">),</span>
                <span class="p">),</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;inCohort&quot;</span><span class="p">],</span>
            <span class="p">),</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;biobank&quot;</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;cohort&quot;</span><span class="p">]),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohorts&quot;</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">is_meta_analyzed_variant</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">cohorts</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check if the variant is meta-analyzed (present in more than one biobank).</span>

<span class="sd">        Note:</span>
<span class="sd">            if the same biobank has multiple cohorts, it still counts as one biobank.</span>

<span class="sd">        Args:</span>
<span class="sd">            cohorts (Column): Array of structs with fields &#39;biobank&#39;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Boolean column indicating if the variant is meta-analyzed.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; data = [([(&quot;FinnGen&quot;, &quot;FinnGen&quot;), (&quot;MVP&quot;, &quot;MVP_EUR&quot;), (&quot;MVP&quot;, &quot;MVP_AMR&quot;)],), ([(&quot;MVP&quot;, &quot;MVP_AMR&quot;), (&quot;MVP&quot;, &quot;MVP_EUR&quot;)],),([(&quot;UKBB&quot;, &quot;UKBB&quot;)],)]</span>
<span class="sd">            &gt;&gt;&gt; schema = &quot;cohorts ARRAY&lt;STRUCT&lt;biobank: STRING, cohort: STRING&gt;&gt;&quot;</span>
<span class="sd">            &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">            &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">            +----------------------------------------------------+</span>
<span class="sd">            |cohorts                                             |</span>
<span class="sd">            +----------------------------------------------------+</span>
<span class="sd">            |[{FinnGen, FinnGen}, {MVP, MVP_EUR}, {MVP, MVP_AMR}]|</span>
<span class="sd">            |[{MVP, MVP_AMR}, {MVP, MVP_EUR}]                    |</span>
<span class="sd">            |[{UKBB, UKBB}]                                      |</span>
<span class="sd">            +----------------------------------------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>

<span class="sd">            &gt;&gt;&gt; df.withColumn(&quot;isMetaAnalyzedVariant&quot;, FinnGenUkbMvpMetaSummaryStatistics.is_meta_analyzed_variant(f.col(&quot;cohorts&quot;))).show(truncate=False)</span>
<span class="sd">            +----------------------------------------------------+---------------------+</span>
<span class="sd">            |cohorts                                             |isMetaAnalyzedVariant|</span>
<span class="sd">            +----------------------------------------------------+---------------------+</span>
<span class="sd">            |[{FinnGen, FinnGen}, {MVP, MVP_EUR}, {MVP, MVP_AMR}]|true                 |</span>
<span class="sd">            |[{MVP, MVP_AMR}, {MVP, MVP_EUR}]                    |false                |</span>
<span class="sd">            |[{UKBB, UKBB}]                                      |false                |</span>
<span class="sd">            +----------------------------------------------------+---------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>


<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">n_biobanks</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">array_distinct</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">cohorts</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;biobank&quot;</span><span class="p">))),</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">n_biobanks</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;isMetaAnalyzedVariant&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.allele_frequencies" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">allele_frequencies</span><span class="p">(</span><span class="n">flip</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">scale</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.allele_frequencies" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Extract the allele frequencies per cohort.</p>


<details class="note" open>
  <summary>Note</summary>
  <p>if the <code>flip</code> column is -1, then the allele frequency is flipped (1 - af).</p>
</details>

<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>flip</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Direction column indicating if the allele needs to be flipped. (-1 for flip, 1 for no flip, null for no information)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>scale</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Scale for the decimal type conversion. Default is 10.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Column containing array of structs with <code>cohort</code> and <code>alleleFrequency</code> fields.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;v1&quot;</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>       <span class="p">(</span><span class="s2">&quot;v2&quot;</span><span class="p">,</span> <span class="mf">0.000000001</span><span class="p">,</span> <span class="mf">0.999999999</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>       <span class="p">(</span><span class="s2">&quot;v3&quot;</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">),]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">schema</span> <span class="o">=</span> <span class="s2">&quot;variantId STRING, MVP_EUR_af_alt DOUBLE, MVP_AFR_af_alt DOUBLE, MVP_HIS_af_alt DOUBLE, fg_af_alt DOUBLE, ukbb_af_alt DOUBLE, flip INT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">schema</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+---------+--------------+--------------+--------------+---------+-----------+----+</span>
<span class="go">|variantId|MVP_EUR_af_alt|MVP_AFR_af_alt|MVP_HIS_af_alt|fg_af_alt|ukbb_af_alt|flip|</span>
<span class="go">+---------+--------------+--------------+--------------+---------+-----------+----+</span>
<span class="go">|v1       |0.1           |0.2           |NULL          |0.3      |0.4        |-1  |</span>
<span class="go">|v2       |1.0E-9        |0.999999999   |NULL          |NULL     |NULL       |1   |</span>
<span class="go">|v3       |0.1           |0.1           |NULL          |NULL     |NULL       |NULL|</span>
<span class="go">+---------+--------------+--------------+--------------+---------+-----------+----+</span>
</code></pre></div>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;alleleFrequencies&quot;</span><span class="p">,</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">allele_frequencies</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;flip&quot;</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;alleleFrequencies&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+-------------------------------------------------------------------------------------------------+</span>
<span class="go">|alleleFrequencies                                                                                |</span>
<span class="go">+-------------------------------------------------------------------------------------------------+</span>
<span class="go">|[{MVP_EUR, 0.9000000000}, {MVP_AFR, 0.8000000000}, {FinnGen, 0.7000000000}, {UKBB, 0.6000000000}]|</span>
<span class="go">|[{MVP_EUR, 0.0000000010}, {MVP_AFR, 0.9999999990}]                                               |</span>
<span class="go">|[{MVP_EUR, 0.1000000000}, {MVP_AFR, 0.1000000000}]                                               |</span>
<span class="go">+-------------------------------------------------------------------------------------------------+</span>
</code></pre></div>


            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/finngen_meta/summary_statistics.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">allele_frequencies</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">flip</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">scale</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract the allele frequencies per cohort.</span>

<span class="sd">    Note:</span>
<span class="sd">        if the `flip` column is -1, then the allele frequency is flipped (1 - af).</span>

<span class="sd">    Args:</span>
<span class="sd">        flip (Column): Direction column indicating if the allele needs to be flipped. (-1 for flip, 1 for no flip, null for no information)</span>
<span class="sd">        scale (int): Scale for the decimal type conversion. Default is 10.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Column containing array of structs with `cohort` and `alleleFrequency` fields.</span>


<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; data = [(&quot;v1&quot;, 0.1, 0.2, None, 0.3, 0.4, -1),</span>
<span class="sd">        ...        (&quot;v2&quot;, 0.000000001, 0.999999999, None, None, None,1),</span>
<span class="sd">        ...        (&quot;v3&quot;, 0.1, 0.1, None, None, None, None),]</span>
<span class="sd">        &gt;&gt;&gt; schema = &quot;variantId STRING, MVP_EUR_af_alt DOUBLE, MVP_AFR_af_alt DOUBLE, MVP_HIS_af_alt DOUBLE, fg_af_alt DOUBLE, ukbb_af_alt DOUBLE, flip INT&quot;</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">        &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">        +---------+--------------+--------------+--------------+---------+-----------+----+</span>
<span class="sd">        |variantId|MVP_EUR_af_alt|MVP_AFR_af_alt|MVP_HIS_af_alt|fg_af_alt|ukbb_af_alt|flip|</span>
<span class="sd">        +---------+--------------+--------------+--------------+---------+-----------+----+</span>
<span class="sd">        |v1       |0.1           |0.2           |NULL          |0.3      |0.4        |-1  |</span>
<span class="sd">        |v2       |1.0E-9        |0.999999999   |NULL          |NULL     |NULL       |1   |</span>
<span class="sd">        |v3       |0.1           |0.1           |NULL          |NULL     |NULL       |NULL|</span>
<span class="sd">        +---------+--------------+--------------+--------------+---------+-----------+----+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        &gt;&gt;&gt; df = df.withColumn(&quot;alleleFrequencies&quot;, FinnGenUkbMvpMetaSummaryStatistics.allele_frequencies(f.col(&quot;flip&quot;)))</span>
<span class="sd">        &gt;&gt;&gt; df.select(&quot;alleleFrequencies&quot;).show(truncate=False)</span>
<span class="sd">        +-------------------------------------------------------------------------------------------------+</span>
<span class="sd">        |alleleFrequencies                                                                                |</span>
<span class="sd">        +-------------------------------------------------------------------------------------------------+</span>
<span class="sd">        |[{MVP_EUR, 0.9000000000}, {MVP_AFR, 0.8000000000}, {FinnGen, 0.7000000000}, {UKBB, 0.6000000000}]|</span>
<span class="sd">        |[{MVP_EUR, 0.0000000010}, {MVP_AFR, 0.9999999990}]                                               |</span>
<span class="sd">        |[{MVP_EUR, 0.1000000000}, {MVP_AFR, 0.1000000000}]                                               |</span>
<span class="sd">        +-------------------------------------------------------------------------------------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">precision</span> <span class="o">=</span> <span class="n">scale</span> <span class="o">+</span> <span class="mi">1</span>  <span class="c1"># to ensure we can represent values like 1.0000</span>
    <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_EUR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">normalize_af</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_af_alt&quot;</span><span class="p">),</span> <span class="n">flip</span><span class="p">)</span>
                <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DecimalType</span><span class="p">(</span><span class="n">precision</span><span class="p">,</span> <span class="n">scale</span><span class="p">))</span>
                <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">),</span>
            <span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_AFR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">normalize_af</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_af_alt&quot;</span><span class="p">),</span> <span class="n">flip</span><span class="p">)</span>
                <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DecimalType</span><span class="p">(</span><span class="n">precision</span><span class="p">,</span> <span class="n">scale</span><span class="p">))</span>
                <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">),</span>
            <span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_AMR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                <span class="c1"># Note: HIS in sumstats is AMR in study index</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">normalize_af</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_af_alt&quot;</span><span class="p">),</span> <span class="n">flip</span><span class="p">)</span>
                <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DecimalType</span><span class="p">(</span><span class="n">precision</span><span class="p">,</span> <span class="n">scale</span><span class="p">))</span>
                <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">),</span>
            <span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;FinnGen&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">normalize_af</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;fg_af_alt&quot;</span><span class="p">),</span> <span class="n">flip</span><span class="p">)</span>
                <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DecimalType</span><span class="p">(</span><span class="n">precision</span><span class="p">,</span> <span class="n">scale</span><span class="p">))</span>
                <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">),</span>
            <span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;UKBB&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">normalize_af</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ukbb_af_alt&quot;</span><span class="p">),</span> <span class="n">flip</span><span class="p">)</span>
                <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DecimalType</span><span class="p">(</span><span class="n">precision</span><span class="p">,</span> <span class="n">scale</span><span class="p">))</span>
                <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">),</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span>
    <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;alleleFrequencies&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.bgzip_to_parquet" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">bgzip_to_parquet</span><span class="p">(</span><span class="n">session</span><span class="p">:</span> <span class="n">Session</span><span class="p">,</span> <span class="n">summary_statistics_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">datasource</span><span class="p">:</span> <span class="n">MetaAnalysisDataSource</span><span class="p">,</span> <span class="n">raw_summary_statistics_output_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">n_threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.bgzip_to_parquet" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Convert gzipped summary statistics to Parquet format.</p>
<p>This is a pre-step that needs to be performed once to convert the block gzipped to parquet format. This step
should be run before the actual harmonisation step is performed and the output Parquet files can be used as input
for the <code>from_source</code> method.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>session</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.common.session.Session" href="../../../common/session/#gentropy.common.session.Session">Session</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Session object.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>summary_statistics_list</code>
            </td>
            <td>
                  <code><span title="list">list</span>[<span title="str">str</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of paths where summary statistics files are located.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>datasource</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.datasource.finngen_meta.MetaAnalysisDataSource" href="../finngen_manifest/#gentropy.datasource.finngen_meta.MetaAnalysisDataSource">MetaAnalysisDataSource</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Data source, can be FinnGenMetaDataSource.FINNGEN_UKBB_MVP.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>raw_summary_statistics_output_path</code>
            </td>
            <td>
                  <code><span title="str">str</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Output path for the Parquet files.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_threads</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum number of threads to use for ThreadPoolExecutor (default is 10).</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
      </tbody>
    </table>
        <p>The output requires a single path that will be populated with Parquet files partitioned by <code>studyId</code> extracted
from the input file names.</p>
<div class="admonition note">
<p class="admonition-title">Block gzipped input files</p>
<p>Since the individual summary statistics files are <strong>block gzipped</strong> we use the enhanced bgzip codec for efficient reading.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Reading multiple files with divergent schemas</p>
<p>Since the schema for individual summary statistics <strong>is not strictly the same we have to enforce the schema</strong>.</p>
<p><strong><em>enforcing schema</em></strong>
using the <code>enforceSchema</code> option in <code>spark.read.csv</code> <strong>does not map columns that exist in the file provided schema</strong>,
but rather aligns columns positionally, which breaks the column order per individual file.</p>
<p><strong><em>inferring schema</em></strong>
Attempting to use the <code>inferSchema</code> option in <code>spark.read.csv</code> while reading multiple files in bulk drops columns, due
to the random sampling of files to infer the schema. (Files chosen to infer the schema may not contain entire superset of column space.)</p>
<p><strong><em>manual schema enforcement</em></strong>
The only way to keep the columns in order and use full column superset is to loop over the files with <code>inferSchema</code> and manually
add missing columns with null values casted to expected type. The looping can be parallelized using a <strong>thread pool</strong> (ThreadPoolExecutor)
with <code>n_threads</code> as the maximum load of jobs to spark cluster.</p>
</div>
<details class="warning">
<summary>Performance considerations</summary>
<p>This function requires a <em>Session</em> with <code>use_enhanced_bgzip_codec</code> to be set to True. This function is strongly encouraged to be used in
a distributed environment.</p>
</details>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="KeyError">KeyError</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If <code>use_enhanced_bgzip_codec</code> is set to False in the Session configuration.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/finngen_meta/summary_statistics.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">bgzip_to_parquet</span><span class="p">(</span>
    <span class="bp">cls</span><span class="p">,</span>
    <span class="n">session</span><span class="p">:</span> <span class="n">Session</span><span class="p">,</span>
    <span class="n">summary_statistics_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">datasource</span><span class="p">:</span> <span class="n">MetaAnalysisDataSource</span><span class="p">,</span>
    <span class="n">raw_summary_statistics_output_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">n_threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert gzipped summary statistics to Parquet format.</span>

<span class="sd">    This is a pre-step that needs to be performed once to convert the block gzipped to parquet format. This step</span>
<span class="sd">    should be run before the actual harmonisation step is performed and the output Parquet files can be used as input</span>
<span class="sd">    for the `from_source` method.</span>

<span class="sd">    Args:</span>
<span class="sd">        session (Session): Session object.</span>
<span class="sd">        summary_statistics_list (list[str]): List of paths where summary statistics files are located.</span>
<span class="sd">        datasource (MetaAnalysisDataSource): Data source, can be FinnGenMetaDataSource.FINNGEN_UKBB_MVP.</span>
<span class="sd">        raw_summary_statistics_output_path (str): Output path for the Parquet files.</span>
<span class="sd">        n_threads (int): Maximum number of threads to use for ThreadPoolExecutor (default is 10).</span>

<span class="sd">    The output requires a single path that will be populated with Parquet files partitioned by `studyId` extracted</span>
<span class="sd">    from the input file names.</span>

<span class="sd">    !!! note &quot;Block gzipped input files&quot;</span>

<span class="sd">        Since the individual summary statistics files are **block gzipped** we use the enhanced bgzip codec for efficient reading.</span>

<span class="sd">    !!! note &quot;Reading multiple files with divergent schemas&quot;</span>

<span class="sd">        Since the schema for individual summary statistics **is not strictly the same we have to enforce the schema**.</span>

<span class="sd">        **_enforcing schema_**</span>
<span class="sd">        using the `enforceSchema` option in `spark.read.csv` **does not map columns that exist in the file provided schema**,</span>
<span class="sd">        but rather aligns columns positionally, which breaks the column order per individual file.</span>

<span class="sd">        **_inferring schema_**</span>
<span class="sd">        Attempting to use the `inferSchema` option in `spark.read.csv` while reading multiple files in bulk drops columns, due</span>
<span class="sd">        to the random sampling of files to infer the schema. (Files chosen to infer the schema may not contain entire superset of column space.)</span>

<span class="sd">        **_manual schema enforcement_**</span>
<span class="sd">        The only way to keep the columns in order and use full column superset is to loop over the files with `inferSchema` and manually</span>
<span class="sd">        add missing columns with null values casted to expected type. The looping can be parallelized using a **thread pool** (ThreadPoolExecutor)</span>
<span class="sd">        with `n_threads` as the maximum load of jobs to spark cluster.</span>

<span class="sd">    ??? warning &quot;Performance considerations&quot;</span>
<span class="sd">        This function requires a _Session_ with `use_enhanced_bgzip_codec` to be set to True. This function is strongly encouraged to be used in</span>
<span class="sd">        a distributed environment.</span>

<span class="sd">    Raises:</span>
<span class="sd">        KeyError: If `use_enhanced_bgzip_codec` is set to False in the Session configuration.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">summary_statistics_list</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">session</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No summary statistics paths found to process.&quot;</span><span class="p">)</span>
        <span class="k">return</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">session</span><span class="o">.</span><span class="n">use_enhanced_bgzip_codec</span><span class="p">:</span>
        <span class="n">session</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
            <span class="s2">&quot;The use_enhanced_bgzip_codec is set to False. This will lead to inefficient reading of block gzipped files.&quot;</span>
        <span class="p">)</span>
        <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span>
            <span class="s2">&quot;Please set `session.spark.use_enhanced_bgzip_codec` to True in the Session configuration.&quot;</span>
        <span class="p">)</span>

    <span class="c1"># Handle n_threads limits and warnings</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">n_threads</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">or</span> <span class="n">n_threads</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">session</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Invalid n_threads value: </span><span class="si">{</span><span class="n">n_threads</span><span class="si">}</span><span class="s2">. Falling back to 10 threads.&quot;</span>
        <span class="p">)</span>
        <span class="n">n_threads</span> <span class="o">=</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">N_THREAD_OPTIMAL</span>
    <span class="k">if</span> <span class="n">n_threads</span> <span class="o">&lt;</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">N_THREAD_OPTIMAL</span><span class="p">:</span>
        <span class="n">session</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Using low n_threads value: </span><span class="si">{</span><span class="n">n_threads</span><span class="si">}</span><span class="s2">. This may lead to sub-optimal performance.&quot;</span>
        <span class="p">)</span>
    <span class="k">if</span> <span class="n">n_threads</span> <span class="o">&gt;</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">N_THREAD_MAX</span><span class="p">:</span>
        <span class="n">session</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Using high n_threads value: </span><span class="si">{</span><span class="n">n_threads</span><span class="si">}</span><span class="s2">, this may lead to overloading spark driver. Limiting to 32.&quot;</span>
        <span class="p">)</span>
        <span class="n">n_threads</span> <span class="o">=</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">N_THREAD_MAX</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">process_one</span><span class="p">(</span>
        <span class="n">input_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">session</span><span class="p">:</span> <span class="n">Session</span><span class="p">,</span> <span class="n">output_path</span><span class="p">:</span> <span class="nb">str</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Function to process one finngen-ukbb-mvp summary statistics file to schema superset.</span>

<span class="sd">        Args:</span>
<span class="sd">            input_path (str): Input path to the gzipped summary statistics file.</span>
<span class="sd">            session (Session): Session object.</span>
<span class="sd">            output_path (str): Output path for the Parquet files.</span>

<span class="sd">        Returns:</span>
<span class="sd">            DataFrame: Processed dataframe.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">session</span><span class="o">.</span><span class="n">spark</span><span class="o">.</span><span class="n">read</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span>
            <span class="n">input_path</span><span class="p">,</span>
            <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">inferSchema</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">enforceSchema</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># Apply schema enforcement by selecting columns in the expected order with proper types</span>
        <span class="c1"># NOTE: Here we only add missing columns with null values casted to expected type</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="bp">cls</span><span class="o">.</span><span class="n">raw_schema</span><span class="o">.</span><span class="n">names</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">c</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="bp">cls</span><span class="o">.</span><span class="n">raw_schema</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">.</span><span class="n">dataType</span><span class="p">))</span>

        <span class="c1"># Replace all NA with nulls and cast to the expected type</span>
        <span class="c1"># NOTE: Here we apply the transformation on full schema set (including added missing columns)</span>
        <span class="c1"># NOTE: If we do not transform the `NA` to nulls, we will still have divergent schemas, as `NA` forces the column to StringType</span>
        <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="bp">cls</span><span class="o">.</span><span class="n">raw_schema</span><span class="o">.</span><span class="n">fields</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="n">field</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">field</span><span class="o">.</span><span class="n">name</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">))</span>
                <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">field</span><span class="o">.</span><span class="n">name</span><span class="p">))</span>
                <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">field</span><span class="o">.</span><span class="n">dataType</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="c1"># Add studyId based on the input path</span>
        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">concat_ws</span><span class="p">(</span>
                    <span class="s2">&quot;_&quot;</span><span class="p">,</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="n">datasource</span><span class="o">.</span><span class="n">value</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span>
                        <span class="bp">cls</span><span class="o">.</span><span class="n">extract_study_phenotype_from_path</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">input_file_name</span><span class="p">())</span>
                    <span class="p">),</span>
                <span class="p">),</span>
                <span class="c1"># Optimal partition size is ~ 100MB, assuming the total size of the dataset is 2Tb</span>
                <span class="c1"># we can have up to 60 partitions per study (330 studies)</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">orderBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;#CHR&quot;</span><span class="p">,</span> <span class="s2">&quot;POS&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">repartition</span><span class="p">(</span><span class="mi">60</span><span class="p">,</span> <span class="s2">&quot;#CHR&quot;</span><span class="p">,</span> <span class="s2">&quot;POS&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="c1"># Write out the processed dataframe to Parquet</span>
        <span class="c1"># NOTE: Write is done per studyId partition from the thread pool to</span>
        <span class="c1"># make sure we do not need to collect all data after the thread execution.</span>
        <span class="n">df</span><span class="o">.</span><span class="n">write</span><span class="o">.</span><span class="n">mode</span><span class="p">(</span><span class="s2">&quot;append&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">partitionBy</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">parquet</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df</span>

    <span class="n">session</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;Converting gzipped summary statistics from </span><span class="si">{</span><span class="n">summary_statistics_list</span><span class="si">}</span><span class="s2"> to Parquet at </span><span class="si">{</span><span class="n">raw_summary_statistics_output_path</span><span class="si">}</span><span class="s2">.&quot;</span>
    <span class="p">)</span>
    <span class="k">with</span> <span class="n">ThreadPoolExecutor</span><span class="p">(</span><span class="n">max_workers</span><span class="o">=</span><span class="n">n_threads</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">path</span><span class="p">:</span> <span class="n">process_one</span><span class="p">(</span>
                <span class="n">path</span><span class="p">,</span>
                <span class="n">session</span><span class="o">=</span><span class="n">session</span><span class="p">,</span>
                <span class="n">output_path</span><span class="o">=</span><span class="n">raw_summary_statistics_output_path</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">summary_statistics_list</span><span class="p">,</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.cohorts" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">cohorts</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.cohorts" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Cohorts involved in the meta-analysis.</p>
<p>This method creates an array of structs containing biobank and cohort information
for variants that have allele frequency data available in each respective cohort.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array of structs with fields 'biobank' and 'cohort'.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <h5 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.cohorts--test-case-1-all-cohorts-have-data">Test case 1: All cohorts have data<a class="headerlink" href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.cohorts--test-case-1-all-cohorts-have-data" title="Permanent link">&para;</a></h5>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">data1</span> <span class="o">=</span> <span class="p">[(</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">schema1</span> <span class="o">=</span> <span class="s2">&quot;MVP_EUR_af_alt DOUBLE, MVP_AFR_af_alt DOUBLE, MVP_HIS_af_alt DOUBLE, fg_af_alt DOUBLE, ukbb_af_alt DOUBLE&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df1</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">data1</span><span class="p">,</span> <span class="n">schema1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df1</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;cohorts&quot;</span><span class="p">,</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">cohorts</span><span class="p">())</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;cohorts&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+----------------------------------------------------------------------------------+</span>
<span class="go">|cohorts                                                                           |</span>
<span class="go">+----------------------------------------------------------------------------------+</span>
<span class="go">|[{MVP, MVP_EUR}, {MVP, MVP_AFR}, {MVP, MVP_AMR}, {FinnGen, FinnGen}, {UKBB, UKBB}]|</span>
<span class="go">+----------------------------------------------------------------------------------+</span>
</code></pre></div>
    <h5 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.cohorts--test-case-2-only-some-cohorts-have-data">Test case 2: Only some cohorts have data<a class="headerlink" href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.cohorts--test-case-2-only-some-cohorts-have-data" title="Permanent link">&para;</a></h5>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">data2</span> <span class="o">=</span> <span class="p">[(</span><span class="mf">0.3</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="kc">None</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df2</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">data2</span><span class="p">,</span> <span class="n">schema1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df2</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;cohorts&quot;</span><span class="p">,</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">cohorts</span><span class="p">())</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;cohorts&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+------------------------------------+</span>
<span class="go">|cohorts                             |</span>
<span class="go">+------------------------------------+</span>
<span class="go">|[{MVP, MVP_EUR}, {FinnGen, FinnGen}]|</span>
<span class="go">+------------------------------------+</span>
</code></pre></div>


            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/finngen_meta/summary_statistics.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span>
<span class="normal">1124</span>
<span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span>
<span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span>
<span class="normal">1161</span>
<span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span>
<span class="normal">1174</span>
<span class="normal">1175</span>
<span class="normal">1176</span>
<span class="normal">1177</span>
<span class="normal">1178</span>
<span class="normal">1179</span>
<span class="normal">1180</span>
<span class="normal">1181</span>
<span class="normal">1182</span>
<span class="normal">1183</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">cohorts</span><span class="p">(</span><span class="bp">cls</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Cohorts involved in the meta-analysis.</span>

<span class="sd">    This method creates an array of structs containing biobank and cohort information</span>
<span class="sd">    for variants that have allele frequency data available in each respective cohort.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Array of structs with fields &#39;biobank&#39; and &#39;cohort&#39;.</span>

<span class="sd">    Examples:</span>
<span class="sd">        # Test case 1: All cohorts have data</span>
<span class="sd">        &gt;&gt;&gt; data1 = [(0.3, 0.2, 0.4, 0.1, 0.25)]</span>
<span class="sd">        &gt;&gt;&gt; schema1 = &quot;MVP_EUR_af_alt DOUBLE, MVP_AFR_af_alt DOUBLE, MVP_HIS_af_alt DOUBLE, fg_af_alt DOUBLE, ukbb_af_alt DOUBLE&quot;</span>
<span class="sd">        &gt;&gt;&gt; df1 = spark.createDataFrame(data1, schema1)</span>
<span class="sd">        &gt;&gt;&gt; df1.withColumn(&quot;cohorts&quot;, FinnGenUkbMvpMetaSummaryStatistics.cohorts()).select(&quot;cohorts&quot;).show(truncate=False)</span>
<span class="sd">        +----------------------------------------------------------------------------------+</span>
<span class="sd">        |cohorts                                                                           |</span>
<span class="sd">        +----------------------------------------------------------------------------------+</span>
<span class="sd">        |[{MVP, MVP_EUR}, {MVP, MVP_AFR}, {MVP, MVP_AMR}, {FinnGen, FinnGen}, {UKBB, UKBB}]|</span>
<span class="sd">        +----------------------------------------------------------------------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        # Test case 2: Only some cohorts have data</span>
<span class="sd">        &gt;&gt;&gt; data2 = [(0.3, None, None, 0.1, None)]</span>
<span class="sd">        &gt;&gt;&gt; df2 = spark.createDataFrame(data2, schema1)</span>
<span class="sd">        &gt;&gt;&gt; df2.withColumn(&quot;cohorts&quot;, FinnGenUkbMvpMetaSummaryStatistics.cohorts()).select(&quot;cohorts&quot;).show(truncate=False)</span>
<span class="sd">        +------------------------------------+</span>
<span class="sd">        |cohorts                             |</span>
<span class="sd">        +------------------------------------+</span>
<span class="sd">        |[{MVP, MVP_EUR}, {FinnGen, FinnGen}]|</span>
<span class="sd">        +------------------------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_af_alt&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inCohort&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_EUR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;biobank&quot;</span><span class="p">),</span>
                <span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_af_alt&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inCohort&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_AFR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;biobank&quot;</span><span class="p">),</span>
                <span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_af_alt&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inCohort&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_AMR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;biobank&quot;</span><span class="p">),</span>
                <span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;fg_af_alt&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inCohort&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;FinnGen&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;FinnGen&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;biobank&quot;</span><span class="p">),</span>
                <span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ukbb_af_alt&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                    <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;inCohort&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;UKBB&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;UKBB&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;biobank&quot;</span><span class="p">),</span>
                <span class="p">),</span>
            <span class="p">),</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;inCohort&quot;</span><span class="p">],</span>
        <span class="p">),</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;biobank&quot;</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;cohort&quot;</span><span class="p">]),</span>
    <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohorts&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.combined_allele_frequency" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">combined_allele_frequency</span><span class="p">(</span><span class="n">allele_freq</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">n_samples_per_cohort</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.combined_allele_frequency" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Combined Allele Frequency across all cohorts.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>allele_freq</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Column containing array of structs with <code>cohort</code> and <code>alleleFrequency</code> fields.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_samples_per_cohort</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Column containing array of structs with <code>cohort</code> and <code>nSamples</code> fields.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Combined allele frequency across all cohorts.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="note" open>
  <summary>Note</summary>
  <p>The combination is made by weighting the allele frequencies by the number of samples in each cohort.</p>
</details>

<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">... </span>   <span class="p">(</span><span class="s2">&quot;v1&quot;</span><span class="p">,</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;alleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.6</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;alleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.2</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;alleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.3</span><span class="p">}],</span>
<span class="gp">... </span>          <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;nSamples&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;nSamples&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;D&quot;</span><span class="p">,</span> <span class="s2">&quot;nSamples&quot;</span><span class="p">:</span> <span class="mi">20</span><span class="p">}]),</span>
<span class="gp">... </span>   <span class="p">(</span><span class="s2">&quot;v2&quot;</span><span class="p">,</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;alleleFrequency&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">},],</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;nSamples&quot;</span><span class="p">:</span> <span class="mi">50</span><span class="p">}]),</span>
<span class="gp">... </span>   <span class="p">(</span><span class="s2">&quot;v3&quot;</span><span class="p">,</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;alleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.05</span><span class="p">},],</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;nSamples&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}]),]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">schema</span> <span class="o">=</span> <span class="s2">&quot;variantId STRING, alleleFrequencies ARRAY&lt;STRUCT&lt;cohort: STRING, alleleFrequency: DOUBLE&gt;&gt;, nSamplesPerCohort ARRAY&lt;STRUCT&lt;cohort: STRING, nSamples: INT&gt;&gt;&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">schema</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+---------+------------------------------+-----------------------------+</span>
<span class="go">|variantId|alleleFrequencies             |nSamplesPerCohort            |</span>
<span class="go">+---------+------------------------------+-----------------------------+</span>
<span class="go">|v1       |[{A, 0.6}, {B, 0.2}, {C, 0.3}]|[{A, 100}, {B, 200}, {D, 20}]|</span>
<span class="go">|v2       |[{A, NULL}]                   |[{A, 50}]                    |</span>
<span class="go">|v3       |[{A, 0.05}]                   |[{A, NULL}]                  |</span>
<span class="go">+---------+------------------------------+-----------------------------+</span>
</code></pre></div>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;combinedAlleleFrequency&quot;</span><span class="p">,</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">combined_allele_frequency</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;alleleFrequencies&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nSamplesPerCohort&quot;</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="s2">&quot;combinedAlleleFrequency&quot;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;caf&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+---------+----+</span>
<span class="go">|variantId|caf |</span>
<span class="go">+---------+----+</span>
<span class="go">|v1       |0.33|</span>
<span class="go">|v2       |NULL|</span>
<span class="go">|v3       |NULL|</span>
<span class="go">+---------+----+</span>
</code></pre></div>


            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/finngen_meta/summary_statistics.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">combined_allele_frequency</span><span class="p">(</span>
    <span class="bp">cls</span><span class="p">,</span> <span class="n">allele_freq</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">n_samples_per_cohort</span><span class="p">:</span> <span class="n">Column</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Combined Allele Frequency across all cohorts.</span>

<span class="sd">    Args:</span>
<span class="sd">        allele_freq (Column): Column containing array of structs with `cohort` and `alleleFrequency` fields.</span>
<span class="sd">        n_samples_per_cohort (Column): Column containing array of structs with `cohort` and `nSamples` fields.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Combined allele frequency across all cohorts.</span>

<span class="sd">    Note:</span>
<span class="sd">        The combination is made by weighting the allele frequencies by the number of samples in each cohort.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; data = [</span>
<span class="sd">        ...    (&quot;v1&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;alleleFrequency&quot;: 0.6}, {&quot;cohort&quot;: &quot;B&quot;, &quot;alleleFrequency&quot;: 0.2}, {&quot;cohort&quot;: &quot;C&quot;, &quot;alleleFrequency&quot;: 0.3}],</span>
<span class="sd">        ...           [{&quot;cohort&quot;: &quot;A&quot;, &quot;nSamples&quot;: 100}, {&quot;cohort&quot;: &quot;B&quot;, &quot;nSamples&quot;: 200}, {&quot;cohort&quot;: &quot;D&quot;, &quot;nSamples&quot;: 20}]),</span>
<span class="sd">        ...    (&quot;v2&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;alleleFrequency&quot;: None},], [{&quot;cohort&quot;: &quot;A&quot;, &quot;nSamples&quot;: 50}]),</span>
<span class="sd">        ...    (&quot;v3&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;alleleFrequency&quot;: 0.05},], [{&quot;cohort&quot;: &quot;A&quot;, &quot;nSamples&quot;: None}]),]</span>
<span class="sd">        &gt;&gt;&gt; schema = &quot;variantId STRING, alleleFrequencies ARRAY&lt;STRUCT&lt;cohort: STRING, alleleFrequency: DOUBLE&gt;&gt;, nSamplesPerCohort ARRAY&lt;STRUCT&lt;cohort: STRING, nSamples: INT&gt;&gt;&quot;</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">        &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">        +---------+------------------------------+-----------------------------+</span>
<span class="sd">        |variantId|alleleFrequencies             |nSamplesPerCohort            |</span>
<span class="sd">        +---------+------------------------------+-----------------------------+</span>
<span class="sd">        |v1       |[{A, 0.6}, {B, 0.2}, {C, 0.3}]|[{A, 100}, {B, 200}, {D, 20}]|</span>
<span class="sd">        |v2       |[{A, NULL}]                   |[{A, 50}]                    |</span>
<span class="sd">        |v3       |[{A, 0.05}]                   |[{A, NULL}]                  |</span>
<span class="sd">        +---------+------------------------------+-----------------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        &gt;&gt;&gt; df = df.withColumn(&quot;combinedAlleleFrequency&quot;, FinnGenUkbMvpMetaSummaryStatistics.combined_allele_frequency(f.col(&quot;alleleFrequencies&quot;), f.col(&quot;nSamplesPerCohort&quot;)))</span>
<span class="sd">        &gt;&gt;&gt; df.select(&quot;variantId&quot;, f.round(&quot;combinedAlleleFrequency&quot;, 2).alias(&quot;caf&quot;)).show(truncate=False)</span>
<span class="sd">        +---------+----+</span>
<span class="sd">        |variantId|caf |</span>
<span class="sd">        +---------+----+</span>
<span class="sd">        |v1       |0.33|</span>
<span class="sd">        |v2       |NULL|</span>
<span class="sd">        |v3       |NULL|</span>
<span class="sd">        +---------+----+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">af_filtered</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
        <span class="n">allele_freq</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">samples_filtered</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
        <span class="n">n_samples_per_cohort</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;nSamples&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="n">intersect</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">array_intersect</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">af_filtered</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">)),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">samples_filtered</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">)),</span>
    <span class="p">)</span>
    <span class="n">af_map</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">map_from_entries</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">af_filtered</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span><span class="n">intersect</span><span class="p">,</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">))</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">n_samples_map</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">map_from_entries</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">samples_filtered</span><span class="p">,</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">array_contains</span><span class="p">(</span><span class="n">intersect</span><span class="p">,</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">)),</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="c1"># Compute numerator: sum(AF * n)</span>
    <span class="n">af_times_n</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">map_entries</span><span class="p">(</span><span class="n">af_map</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">0.0</span><span class="p">),</span>
        <span class="k">lambda</span> <span class="n">acc</span><span class="p">,</span> <span class="n">kv</span><span class="p">:</span> <span class="n">acc</span>
        <span class="o">+</span> <span class="n">kv</span><span class="p">[</span><span class="s2">&quot;value&quot;</span><span class="p">]</span>
        <span class="o">*</span> <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">element_at</span><span class="p">(</span><span class="n">n_samples_map</span><span class="p">,</span> <span class="n">kv</span><span class="p">[</span><span class="s2">&quot;key&quot;</span><span class="p">]),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)),</span>
    <span class="p">)</span>

    <span class="c1"># Compute denominator: sum(n)</span>
    <span class="n">n_total</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">map_entries</span><span class="p">(</span><span class="n">n_samples_map</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span>
        <span class="k">lambda</span> <span class="n">acc</span><span class="p">,</span> <span class="n">kv</span><span class="p">:</span> <span class="n">acc</span> <span class="o">+</span> <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">kv</span><span class="p">[</span><span class="s2">&quot;value&quot;</span><span class="p">],</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mi">0</span><span class="p">)),</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">((</span><span class="n">n_total</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">af_times_n</span> <span class="o">==</span> <span class="mi">0</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">None</span><span class="p">))</span>
        <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">af_times_n</span> <span class="o">/</span> <span class="n">n_total</span><span class="p">)</span>
        <span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">FloatType</span><span class="p">())</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.extract_study_phenotype_from_path" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">extract_study_phenotype_from_path</span><span class="p">(</span><span class="n">file_path</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

<a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.extract_study_phenotype_from_path" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Extract the study phenotype from FinnGen file path.</p>


<details class="note" open>
  <summary>Note</summary>
  <p>Assumes the file name format is some_path/to/<studyPhenotype>_meta_out.tsv.gz</p>
</details>

<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>file_path</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Column containing the file path as a string.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Extracted study phenotype as a string.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;/path/to/AB1_meta_out.tsv.gz&quot;</span><span class="p">,),</span> <span class="p">(</span><span class="s2">&quot;/another/path/CD2_meta_out.tsv.gz&quot;</span><span class="p">,)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">schema</span> <span class="o">=</span> <span class="s2">&quot;filePath STRING&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">schema</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;studyPhenotype&quot;</span><span class="p">,</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">extract_study_phenotype_from_path</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;filePath&quot;</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+---------------------------------+--------------+</span>
<span class="go">|filePath                         |studyPhenotype|</span>
<span class="go">+---------------------------------+--------------+</span>
<span class="go">|/path/to/AB1_meta_out.tsv.gz     |AB1           |</span>
<span class="go">|/another/path/CD2_meta_out.tsv.gz|CD2           |</span>
<span class="go">+---------------------------------+--------------+</span>
</code></pre></div>


            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/finngen_meta/summary_statistics.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">extract_study_phenotype_from_path</span><span class="p">(</span><span class="n">file_path</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract the study phenotype from FinnGen file path.</span>

<span class="sd">    Note:</span>
<span class="sd">        Assumes the file name format is some_path/to/&lt;studyPhenotype&gt;_meta_out.tsv.gz</span>


<span class="sd">    Args:</span>
<span class="sd">        file_path (Column): Column containing the file path as a string.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Extracted study phenotype as a string.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; data = [(&quot;/path/to/AB1_meta_out.tsv.gz&quot;,), (&quot;/another/path/CD2_meta_out.tsv.gz&quot;,)]</span>
<span class="sd">        &gt;&gt;&gt; schema = &quot;filePath STRING&quot;</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">        &gt;&gt;&gt; df = df.withColumn(&quot;studyPhenotype&quot;, FinnGenUkbMvpMetaSummaryStatistics.extract_study_phenotype_from_path(f.col(&quot;filePath&quot;)))</span>
<span class="sd">        &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">        +---------------------------------+--------------+</span>
<span class="sd">        |filePath                         |studyPhenotype|</span>
<span class="sd">        +---------------------------------+--------------+</span>
<span class="sd">        |/path/to/AB1_meta_out.tsv.gz     |AB1           |</span>
<span class="sd">        |/another/path/CD2_meta_out.tsv.gz|CD2           |</span>
<span class="sd">        +---------------------------------+--------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">regexp_replace</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">element_at</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s2">&quot;/&quot;</span><span class="p">),</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="s2">&quot;_meta_out.tsv.gz&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.from_source" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">from_source</span><span class="p">(</span><span class="n">raw_summary_statistics</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">finngen_manifest</span><span class="p">:</span> <span class="n">FinnGenMetaManifest</span><span class="p">,</span> <span class="n">variant_annotations</span><span class="p">:</span> <span class="n">VariantDirection</span><span class="p">,</span> <span class="n">perform_meta_analysis_filter</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">imputation_score_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.8</span><span class="p">,</span> <span class="n">perform_imputation_score_filter</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">min_allele_count_threshold</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span> <span class="n">perform_min_allele_count_filter</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">min_allele_frequency_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0001</span><span class="p">,</span> <span class="n">perform_min_allele_frequency_filter</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">filter_out_ambiguous_variants</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SummaryStatistics</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.from_source" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Build the summary statistics dataset from raw summary statistics.</p>
<p>See original issue to find out more details on the harmonisation logic https://github.com/opentargets/issues/issues/3474</p>
<details class="note">
<summary>The logic behind the harmonisation</summary>
<ol>
<li>Build a slice of FinnGen Manifest to bring the information about nCases and nSamples per cohort (broadcast join).</li>
<li>Build a variant direction (gnomAD) dataset partitioned by <code>chromosome</code> and <code>variantId</code> for joining using Sort-Merge strategy.</li>
<li>Select required columns from raw summary statistics</li>
<li>Remove all non-meta analyzed variants (nBiobanks &lt; 2) - by default</li>
<li>Remove all variants with low imputation score - by default</li>
<li>Join with Finngen Manifest and Variant Direction datasets</li>
<li>Flip <code>beta</code> and <code>min allele frequency</code> based on the <code>direction</code> from Variant Direction dataset</li>
<li>Use originalVariantId from Variant Direction dataset if available, otherwise fall back to variantId (for variants missing from Variant Direction)</li>
<li>Calculate combined effect allele frequency from cohorts</li>
<li>Remove all variants with low Min Allele Count - by default</li>
<li>Remove all variants with low Min Allele Frequency - optional</li>
<li>Remove strand ambiguous variants - optional</li>
</ol>
</details>
<details class="tip">
<summary>Variant Directionality</summary>
<p><strong>Variant Direction</strong>
By default we:</p>
<ol>
<li>keep all strand ambiguous variants as is</li>
<li>keep all variants found in gnomAD aligned to gnomAD reference ( if alleles are flipped we flip the beta and allele frequency)</li>
<li>keep all variants not found in gnomAD as is (cannot determine strand or alignment) - we assume these are correct.</li>
</ol>
</details>
<details class="note">
<summary>Important considerations</summary>
<ul>
<li>The input summary statistics are expected to be already parquet formatted and partitioned by <code>studyId</code>.</li>
<li>Both <code>perform_allele_count_filter</code> and <code>perform_allele_frequency_filter</code> are redundant, if both
are set to True, only <code>perform_allele_count_filter</code> will be applied.</li>
<li>Threshold value for Min Allele Count &gt;= 20 means that with a MAF of 1e-4 we would expect to see
at least 1_000 samples with minor allele in a cohort. This is quiet stringent for very rare variants and
smaller cohorts, like MVP_AFR (nSamples ~120k) and MVP_HIS (~60k).</li>
<li>MVP_HIS cohort has been mapped to admixed American population - see https://www.science.org/doi/10.1126/science.adj1182 for more details.</li>
</ul>
</details>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>raw_summary_statistics</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Raw summary statistics dataframe.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>finngen_manifest</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.datasource.finngen_meta.FinnGenMetaManifest" href="../finngen_manifest/#gentropy.datasource.finngen_meta.FinnGenMetaManifest">FinnGenMetaManifest</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>FinnGen meta analysis manifest.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>variant_annotations</code>
            </td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.variant_direction.VariantDirection" href="../../../datasets/variant_direction/#gentropy.dataset.variant_direction.VariantDirection">VariantDirection</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Variant direction dataset.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>perform_meta_analysis_filter</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to remove variants found in just 1 biobank. Default is True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>imputation_score_threshold</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Imputation score threshold for MVP cohorts. Default is 0.8.</p>
              </div>
            </td>
            <td>
                  <code>0.8</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>perform_imputation_score_filter</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to perform the imputation score filter. Default is True.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_allele_count_threshold</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum allele count threshold. Default is 20.</p>
              </div>
            </td>
            <td>
                  <code>20</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>perform_min_allele_count_filter</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to perform the minimum allele count filter. Default is False.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_allele_frequency_threshold</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minimum allele frequency threshold. Default is 1e-4.</p>
              </div>
            </td>
            <td>
                  <code>0.0001</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>perform_min_allele_frequency_filter</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to perform the minimum allele frequency filter. Default is True.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>filter_out_ambiguous_variants</code>
            </td>
            <td>
                  <code><span title="bool">bool</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to filter out strand ambiguous variants. Default is False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>SummaryStatistics</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.summary_statistics.SummaryStatistics" href="../../../datasets/summary_statistics/#gentropy.dataset.summary_statistics.SummaryStatistics">SummaryStatistics</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Processed summary statistics dataset.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/finngen_meta/summary_statistics.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">from_source</span><span class="p">(</span>
    <span class="bp">cls</span><span class="p">,</span>
    <span class="n">raw_summary_statistics</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">finngen_manifest</span><span class="p">:</span> <span class="n">FinnGenMetaManifest</span><span class="p">,</span>
    <span class="n">variant_annotations</span><span class="p">:</span> <span class="n">VariantDirection</span><span class="p">,</span>
    <span class="n">perform_meta_analysis_filter</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">imputation_score_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.8</span><span class="p">,</span>
    <span class="n">perform_imputation_score_filter</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">min_allele_count_threshold</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
    <span class="n">perform_min_allele_count_filter</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">min_allele_frequency_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1e-4</span><span class="p">,</span>
    <span class="n">perform_min_allele_frequency_filter</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">filter_out_ambiguous_variants</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SummaryStatistics</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Build the summary statistics dataset from raw summary statistics.</span>

<span class="sd">    See original issue to find out more details on the harmonisation logic https://github.com/opentargets/issues/issues/3474</span>

<span class="sd">    ??? note &quot;The logic behind the harmonisation&quot;</span>
<span class="sd">        1. Build a slice of FinnGen Manifest to bring the information about nCases and nSamples per cohort (broadcast join).</span>
<span class="sd">        2. Build a variant direction (gnomAD) dataset partitioned by `chromosome` and `variantId` for joining using Sort-Merge strategy.</span>
<span class="sd">        3. Select required columns from raw summary statistics</span>
<span class="sd">        4. Remove all non-meta analyzed variants (nBiobanks &lt; 2) - by default</span>
<span class="sd">        5. Remove all variants with low imputation score - by default</span>
<span class="sd">        6. Join with Finngen Manifest and Variant Direction datasets</span>
<span class="sd">        7. Flip `beta` and `min allele frequency` based on the `direction` from Variant Direction dataset</span>
<span class="sd">        8. Use originalVariantId from Variant Direction dataset if available, otherwise fall back to variantId (for variants missing from Variant Direction)</span>
<span class="sd">        9. Calculate combined effect allele frequency from cohorts</span>
<span class="sd">        10. Remove all variants with low Min Allele Count - by default</span>
<span class="sd">        11. Remove all variants with low Min Allele Frequency - optional</span>
<span class="sd">        12. Remove strand ambiguous variants - optional</span>

<span class="sd">    ??? tip &quot;Variant Directionality&quot;</span>
<span class="sd">        **Variant Direction**</span>
<span class="sd">        By default we:</span>

<span class="sd">        1. keep all strand ambiguous variants as is</span>
<span class="sd">        2. keep all variants found in gnomAD aligned to gnomAD reference ( if alleles are flipped we flip the beta and allele frequency)</span>
<span class="sd">        3. keep all variants not found in gnomAD as is (cannot determine strand or alignment) - we assume these are correct.</span>

<span class="sd">    ??? note &quot;Important considerations&quot;</span>
<span class="sd">        * The input summary statistics are expected to be already parquet formatted and partitioned by `studyId`.</span>
<span class="sd">        * Both `perform_allele_count_filter` and `perform_allele_frequency_filter` are redundant, if both</span>
<span class="sd">        are set to True, only `perform_allele_count_filter` will be applied.</span>
<span class="sd">        * Threshold value for Min Allele Count &gt;= 20 means that with a MAF of 1e-4 we would expect to see</span>
<span class="sd">        at least 1_000 samples with minor allele in a cohort. This is quiet stringent for very rare variants and</span>
<span class="sd">        smaller cohorts, like MVP_AFR (nSamples ~120k) and MVP_HIS (~60k).</span>
<span class="sd">        * MVP_HIS cohort has been mapped to admixed American population - see https://www.science.org/doi/10.1126/science.adj1182 for more details.</span>

<span class="sd">    Args:</span>
<span class="sd">        raw_summary_statistics (DataFrame): Raw summary statistics dataframe.</span>
<span class="sd">        finngen_manifest (FinnGenMetaManifest): FinnGen meta analysis manifest.</span>
<span class="sd">        variant_annotations (VariantDirection): Variant direction dataset.</span>
<span class="sd">        perform_meta_analysis_filter (bool): Whether to remove variants found in just 1 biobank. Default is True.</span>
<span class="sd">        imputation_score_threshold (float): Imputation score threshold for MVP cohorts. Default is 0.8.</span>
<span class="sd">        perform_imputation_score_filter (bool): Whether to perform the imputation score filter. Default is True.</span>
<span class="sd">        min_allele_count_threshold (int): Minimum allele count threshold. Default is 20.</span>
<span class="sd">        perform_min_allele_count_filter (bool): Whether to perform the minimum allele count filter. Default is False.</span>
<span class="sd">        min_allele_frequency_threshold (float): Minimum allele frequency threshold. Default is 1e-4.</span>
<span class="sd">        perform_min_allele_frequency_filter (bool): Whether to perform the minimum allele frequency filter. Default is True.</span>
<span class="sd">        filter_out_ambiguous_variants (bool): Whether to filter out strand ambiguous variants. Default is False.</span>

<span class="sd">    Returns:</span>
<span class="sd">        SummaryStatistics: Processed summary statistics dataset.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">perform_min_allele_count_filter</span><span class="p">:</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">min_allele_count_threshold</span> <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="p">),</span> <span class="s2">&quot;Allele count threshold should be positive.&quot;</span>
    <span class="k">if</span> <span class="n">perform_min_allele_frequency_filter</span><span class="p">:</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="mf">0.0</span> <span class="o">&lt;=</span> <span class="n">min_allele_frequency_threshold</span> <span class="o">&lt;=</span> <span class="mf">0.5</span>
        <span class="p">),</span> <span class="s2">&quot;MAF needs to be between 0 and 0.5.&quot;</span>

    <span class="k">if</span> <span class="n">perform_min_allele_count_filter</span> <span class="ow">and</span> <span class="n">perform_min_allele_frequency_filter</span><span class="p">:</span>
        <span class="c1"># NOTE - MAC filter would be more stringent at low allele frequencies, so no</span>
        <span class="c1"># need to have both filters active at the same time</span>
        <span class="n">perform_min_allele_frequency_filter</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="n">si_slice</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">broadcast</span><span class="p">(</span>
        <span class="n">finngen_manifest</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nCasesPerCohort&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nSamples&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nSamplesPerCohort&quot;</span><span class="p">),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">vd_slice</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">variant_annotations</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;originalVariantId&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;direction&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;isStrandAmbiguous&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="c1"># NOTE: repartition(&quot;chromosome&quot;) produces very uneven partitions,</span>
        <span class="c1"># Spark attempts then to fall back to `dynamic partitioning` algorithm</span>
        <span class="c1"># which fails after N failures.</span>
        <span class="o">.</span><span class="n">repartitionByRange</span><span class="p">(</span><span class="mi">4_000</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;variantId&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">persist</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="n">sumstats</span> <span class="o">=</span> <span class="p">(</span>
        <span class="c1"># Pre-select columns that are needed downstream</span>
        <span class="c1"># NOTE: full set of columns is not required.</span>
        <span class="n">raw_summary_statistics</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;#CHR&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;POS&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;REF&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ALT&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;fg_af_alt&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_r2&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_af_alt&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_r2&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_af_alt&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_r2&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_af_alt&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ukbb_af_alt&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_mlogp&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_beta&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_sebeta&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="c1"># NOTE: make sure the chromosome is coded to 1:22, X, Y</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
            <span class="n">normalize_chromosome</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;#CHR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">StringType</span><span class="p">())),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;#CHR&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;POS&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">IntegerType</span><span class="p">()))</span>
        <span class="o">.</span><span class="n">withColumnRenamed</span><span class="p">(</span><span class="s2">&quot;REF&quot;</span><span class="p">,</span> <span class="s2">&quot;referenceAllele&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">withColumnRenamed</span><span class="p">(</span><span class="s2">&quot;ALT&quot;</span><span class="p">,</span> <span class="s2">&quot;alternateAllele&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;neglogpval&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_mlogp&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">())</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_beta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">()))</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;standardError&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;all_inv_var_meta_sebeta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DoubleType</span><span class="p">())</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;variantId&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">concat_ws</span><span class="p">(</span>
                <span class="s2">&quot;_&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;referenceAllele&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;alternateAllele&quot;</span><span class="p">),</span>
            <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;referenceAllele&quot;</span><span class="p">,</span> <span class="s2">&quot;alternateAllele&quot;</span><span class="p">)</span>
        <span class="c1"># Initial filters based on statistics presence</span>
        <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;neglogpval&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">())</span>
        <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">())</span>
        <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">())</span>
    <span class="p">)</span>

    <span class="c1"># Filter out variants that are not meta analyzed (nBiobanks &lt; 1)</span>
    <span class="k">if</span> <span class="n">perform_meta_analysis_filter</span><span class="p">:</span>
        <span class="n">sumstats</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sumstats</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;cohorts&quot;</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">cohorts</span><span class="p">())</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;isMetaAnalyzedVariant&quot;</span><span class="p">,</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">is_meta_analyzed_variant</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohorts&quot;</span><span class="p">)),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;isMetaAnalyzedVariant&quot;</span><span class="p">))</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;isMetaAnalyzedVariant&quot;</span><span class="p">,</span> <span class="s2">&quot;cohorts&quot;</span><span class="p">)</span>
        <span class="p">)</span>

    <span class="c1"># Filter out variants with low INFO score</span>
    <span class="k">if</span> <span class="n">perform_imputation_score_filter</span><span class="p">:</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">imputation_score_threshold</span> <span class="o">&gt;=</span> <span class="mf">0.0</span>
        <span class="p">),</span> <span class="s2">&quot;Imputation score threshold should be positive.&quot;</span>
        <span class="n">sumstats</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sumstats</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;hasLowImputationScore&quot;</span><span class="p">,</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">has_low_imputation_score</span><span class="p">(</span><span class="n">imputation_score_threshold</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasLowImputationScore&quot;</span><span class="p">))</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;hasLowImputationScore&quot;</span><span class="p">,</span> <span class="s2">&quot;imputationScore&quot;</span><span class="p">)</span>
        <span class="p">)</span>

    <span class="n">sumstats</span> <span class="o">=</span> <span class="p">(</span>
        <span class="c1"># Annotate with StudyIndex nCases, nSamples to obtain</span>
        <span class="c1"># the cases Minor Allele Count and Samples for combined AF calculation</span>
        <span class="n">sumstats</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">si_slice</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
        <span class="c1"># Join with variant direction dataset</span>
        <span class="c1"># Keep variants if not found in gnomAD (left join)</span>
        <span class="o">.</span><span class="n">repartitionByRange</span><span class="p">(</span><span class="mi">4_000</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;variantId&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">vd_slice</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;variantId&quot;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
        <span class="c1"># Use originalVariantId (already flipped) or fall back to variantId if not found in gnomAD</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;originalVariantId&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">))</span>
        <span class="p">)</span>
        <span class="c1"># Compute allele frequency per cohort and align with direction</span>
        <span class="c1"># NOTE: `direction` column represents if variant aligned to gnomAD variant or it&#39;s flipped</span>
        <span class="c1"># version, the values are `1` - direct, `-1` - flipped</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;cohortAlleleFrequency&quot;</span><span class="p">,</span> <span class="bp">cls</span><span class="o">.</span><span class="n">allele_frequencies</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;direction&quot;</span><span class="p">))</span>
        <span class="p">)</span>
        <span class="c1"># Make sure the beta is alined with the direction, if direction is null, keep beta as is</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;beta&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">)</span> <span class="o">*</span> <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;direction&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
        <span class="p">)</span>
        <span class="c1"># Calculate the combined effect allele frequency from cohorts</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;effectAlleleFrequencyFromSource&quot;</span><span class="p">,</span>
            <span class="bp">cls</span><span class="o">.</span><span class="n">combined_allele_frequency</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohortAlleleFrequency&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nSamplesPerCohort&quot;</span><span class="p">)</span>
            <span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="c1"># Remove strand ambiguous variants, if not found in gnomAD, we keep the variant</span>
    <span class="c1"># Not run by default</span>
    <span class="k">if</span> <span class="n">filter_out_ambiguous_variants</span><span class="p">:</span>
        <span class="n">sumstats</span> <span class="o">=</span> <span class="n">sumstats</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;isStrandAmbiguous&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
        <span class="p">)</span>
    <span class="k">if</span> <span class="n">perform_min_allele_count_filter</span> <span class="ow">or</span> <span class="n">perform_min_allele_frequency_filter</span><span class="p">:</span>
        <span class="c1"># Calculate the MAF per cohort</span>
        <span class="n">sumstats</span> <span class="o">=</span> <span class="n">sumstats</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;cohortMinAlleleFrequency&quot;</span><span class="p">,</span>
            <span class="bp">cls</span><span class="o">.</span><span class="n">min_allele_frequency</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohortAlleleFrequency&quot;</span><span class="p">)),</span>
        <span class="p">)</span>
    <span class="k">if</span> <span class="n">perform_min_allele_count_filter</span><span class="p">:</span>
        <span class="n">sumstats</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sumstats</span>
            <span class="c1"># Make sure to only keep cohorts that have nSamples &gt; 0</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;nSamplesPerCohort&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nSamplesPerCohort&quot;</span><span class="p">),</span>
                    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;nSamples&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;nSamples&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">),</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;cohortMinAlleleCount&quot;</span><span class="p">,</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">min_allele_count</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohortMinAlleleFrequency&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nSamplesPerCohort&quot;</span><span class="p">),</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;hasLowMinAlleleCount&quot;</span><span class="p">,</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">has_low_min_allele_count</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohortMinAlleleCount&quot;</span><span class="p">),</span>
                    <span class="n">min_allele_count_threshold</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasLowMinAlleleCount&quot;</span><span class="p">))</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span>
                <span class="s2">&quot;hasLowMinAlleleCount&quot;</span><span class="p">,</span>
                <span class="s2">&quot;cohortMinAlleleCount&quot;</span><span class="p">,</span>
                <span class="s2">&quot;nCasesPerCohort&quot;</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="c1"># Not run by default.</span>
    <span class="k">if</span> <span class="n">perform_min_allele_frequency_filter</span><span class="p">:</span>
        <span class="n">sumstats</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sumstats</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;hasLowMinAlleleFrequency&quot;</span><span class="p">,</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">has_low_min_allele_frequency</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohortMinAlleleFrequency&quot;</span><span class="p">),</span>
                    <span class="n">min_allele_frequency_threshold</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="o">~</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;hasLowMinAlleleFrequency&quot;</span><span class="p">))</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;hasLowMinAlleleFrequency&quot;</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="c1"># Convert to final summary statistics schema</span>
    <span class="n">sumstats</span> <span class="o">=</span> <span class="n">sumstats</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nSamples&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sampleSize&quot;</span><span class="p">),</span>
        <span class="o">*</span><span class="n">pvalue_from_neglogpval</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;neglogpval&quot;</span><span class="p">)),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;effectAlleleFrequencyFromSource&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">),</span>
    <span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;position&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sampleSize&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;effectAlleleFrequencyFromSource&quot;</span><span class="p">),</span>
        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">),</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">SummaryStatistics</span><span class="p">(</span><span class="n">sumstats</span><span class="p">)</span><span class="o">.</span><span class="n">sanity_filter</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.has_low_imputation_score" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">has_low_imputation_score</span><span class="p">(</span><span class="n">imputation_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.has_low_imputation_score" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Find if variant has a low r2 imputation score in any of the MVP cohorts.</p>


<details class="note" open>
  <summary>Note</summary>
  <p>A missing imputation score is considered as passing the threshold, since it means that the variant was not
present in that cohort.</p>
</details>

<details class="note" open>
  <summary>Note</summary>
  <p>If any r2 imputation score is below the threshold, the variant is considered to have a low imputation score.</p>
</details>

<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>imputation_threshold</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Threshold below which the imputation score is considered low.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Boolean column indicating if any cohort has a low imputation score.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;v1&quot;</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;v2&quot;</span><span class="p">,</span><span class="mf">0.7</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;v3&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="mf">0.8</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;v4&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="mf">0.7</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">schema</span> <span class="o">=</span> <span class="s2">&quot;variantId STRING, MVP_EUR_r2 DOUBLE, MVP_AFR_r2 DOUBLE, MVP_HIS_r2 DOUBLE&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">schema</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;hasLowImputationScore&quot;</span><span class="p">,</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">has_low_imputation_score</span><span class="p">(</span><span class="mf">0.8</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="s2">&quot;hasLowImputationScore&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+---------+---------------------+</span>
<span class="go">|variantId|hasLowImputationScore|</span>
<span class="go">+---------+---------------------+</span>
<span class="go">|       v1|                false|</span>
<span class="go">|       v2|                 true|</span>
<span class="go">|       v3|                false|</span>
<span class="go">|       v4|                 true|</span>
<span class="go">+---------+---------------------+</span>
</code></pre></div>


            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/finngen_meta/summary_statistics.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">has_low_imputation_score</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">imputation_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find if variant has a low r2 imputation score in any of the MVP cohorts.</span>

<span class="sd">    Note:</span>
<span class="sd">        A missing imputation score is considered as passing the threshold, since it means that the variant was not</span>
<span class="sd">        present in that cohort.</span>

<span class="sd">    Note:</span>
<span class="sd">        If any r2 imputation score is below the threshold, the variant is considered to have a low imputation score.</span>

<span class="sd">    Args:</span>
<span class="sd">        imputation_threshold (float): Threshold below which the imputation score is considered low.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Boolean column indicating if any cohort has a low imputation score.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; data = [(&quot;v1&quot;, 0.9, 0.8, 1.0), (&quot;v2&quot;,0.7, 0.9, 0.9), (&quot;v3&quot;, None, None, 0.8), (&quot;v4&quot;, None, None, 0.7)]</span>
<span class="sd">        &gt;&gt;&gt; schema = &quot;variantId STRING, MVP_EUR_r2 DOUBLE, MVP_AFR_r2 DOUBLE, MVP_HIS_r2 DOUBLE&quot;</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">        &gt;&gt;&gt; df = df.withColumn(&quot;hasLowImputationScore&quot;, FinnGenUkbMvpMetaSummaryStatistics.has_low_imputation_score(0.8))</span>
<span class="sd">        &gt;&gt;&gt; df.select(&quot;variantId&quot;, &quot;hasLowImputationScore&quot;).show()</span>
<span class="sd">        +---------+---------------------+</span>
<span class="sd">        |variantId|hasLowImputationScore|</span>
<span class="sd">        +---------+---------------------+</span>
<span class="sd">        |       v1|                false|</span>
<span class="sd">        |       v2|                 true|</span>
<span class="sd">        |       v3|                false|</span>
<span class="sd">        |       v4|                 true|</span>
<span class="sd">        +---------+---------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_r2&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;r2&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_EUR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_r2&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">()</span>
                            <span class="o">|</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_EUR_r2&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">imputation_threshold</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">),</span>
                        <span class="p">)</span>
                        <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;filter&quot;</span><span class="p">),</span>
                    <span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_r2&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;r2&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_AFR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_r2&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">()</span>
                            <span class="o">|</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_AFR_r2&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">imputation_threshold</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">),</span>
                        <span class="p">)</span>
                        <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;filter&quot;</span><span class="p">),</span>
                    <span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_r2&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;r2&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;MVP_AMR&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_r2&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">()</span>
                            <span class="o">|</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;MVP_HIS_r2&quot;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">imputation_threshold</span><span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">True</span><span class="p">),</span>
                        <span class="p">)</span>
                        <span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="kc">False</span><span class="p">))</span>
                        <span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;filter&quot;</span><span class="p">),</span>
                    <span class="p">),</span>
                <span class="p">),</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="o">~</span><span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;filter&quot;</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="o">&gt;</span> <span class="mi">0</span>
    <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;hasLowImputationScore&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.has_low_min_allele_count" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">has_low_min_allele_count</span><span class="p">(</span><span class="n">min_allele_count</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">min_allele_count_threshold</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.has_low_min_allele_count" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Find if variant has a low minor allele count in any of the cohorts.</p>


<details class="note" open>
  <summary>Note</summary>
  <p>If any cohort has a minor allele count below the threshold, the variant is considered to have a low minor allele count.</p>
</details>

<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>min_allele_count</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Column containing array of structs with <code>cohort</code> and <code>minAlleleCount</code> fields.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>min_allele_count_threshold</code>
            </td>
            <td>
                  <code><span title="int">int</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Threshold below which the minor allele count is considered low.</p>
              </div>
            </td>
            <td>
                  <code>20</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Boolean column indicating if any cohort has a low minor allele count.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;v1&quot;</span><span class="p">,</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleCount&quot;</span><span class="p">:</span> <span class="mi">30</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleCount&quot;</span><span class="p">:</span> <span class="mi">25</span><span class="p">}]),</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s2">&quot;v2&quot;</span><span class="p">,</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleCount&quot;</span><span class="p">:</span> <span class="mi">15</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleCount&quot;</span><span class="p">:</span> <span class="mi">25</span><span class="p">}]),</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s2">&quot;v3&quot;</span><span class="p">,</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleCount&quot;</span><span class="p">:</span> <span class="mi">30</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleCount&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">}]),</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s2">&quot;v4&quot;</span><span class="p">,</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleCount&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">},],)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">schema</span> <span class="o">=</span> <span class="s2">&quot;variantId STRING, cohortMinAlleleCount ARRAY&lt;STRUCT&lt;cohort: STRING, minAlleleCount: INT&gt;&gt;&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">schema</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;hasLowMinAlleleCount&quot;</span><span class="p">,</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">has_low_min_allele_count</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohortMinAlleleCount&quot;</span><span class="p">),</span> <span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="s2">&quot;hasLowMinAlleleCount&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+---------+--------------------+</span>
<span class="go">|variantId|hasLowMinAlleleCount|</span>
<span class="go">+---------+--------------------+</span>
<span class="go">|       v1|               false|</span>
<span class="go">|       v2|                true|</span>
<span class="go">|       v3|                true|</span>
<span class="go">|       v4|                true|</span>
<span class="go">+---------+--------------------+</span>
</code></pre></div>


            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/finngen_meta/summary_statistics.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">has_low_min_allele_count</span><span class="p">(</span>
    <span class="bp">cls</span><span class="p">,</span> <span class="n">min_allele_count</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">min_allele_count_threshold</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find if variant has a low minor allele count in any of the cohorts.</span>

<span class="sd">    Note:</span>
<span class="sd">        If any cohort has a minor allele count below the threshold, the variant is considered to have a low minor allele count.</span>


<span class="sd">    Args:</span>
<span class="sd">        min_allele_count (Column): Column containing array of structs with `cohort` and `minAlleleCount` fields.</span>
<span class="sd">        min_allele_count_threshold (int): Threshold below which the minor allele count is considered low.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Boolean column indicating if any cohort has a low minor allele count.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; data = [(&quot;v1&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleCount&quot;: 30}, {&quot;cohort&quot;: &quot;B&quot;, &quot;minAlleleCount&quot;: 25}]),</span>
<span class="sd">        ...         (&quot;v2&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleCount&quot;: 15}, {&quot;cohort&quot;: &quot;B&quot;, &quot;minAlleleCount&quot;: 25}]),</span>
<span class="sd">        ...         (&quot;v3&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleCount&quot;: 30}, {&quot;cohort&quot;: &quot;B&quot;, &quot;minAlleleCount&quot;: 10}]),</span>
<span class="sd">        ...         (&quot;v4&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleCount&quot;: 5},],)]</span>
<span class="sd">        &gt;&gt;&gt; schema = &quot;variantId STRING, cohortMinAlleleCount ARRAY&lt;STRUCT&lt;cohort: STRING, minAlleleCount: INT&gt;&gt;&quot;</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">        &gt;&gt;&gt; df = df.withColumn(&quot;hasLowMinAlleleCount&quot;, FinnGenUkbMvpMetaSummaryStatistics.has_low_min_allele_count(f.col(&quot;cohortMinAlleleCount&quot;), 20))</span>
<span class="sd">        &gt;&gt;&gt; df.select(&quot;variantId&quot;, &quot;hasLowMinAlleleCount&quot;).show()</span>
<span class="sd">        +---------+--------------------+</span>
<span class="sd">        |variantId|hasLowMinAlleleCount|</span>
<span class="sd">        +---------+--------------------+</span>
<span class="sd">        |       v1|               false|</span>
<span class="sd">        |       v2|                true|</span>
<span class="sd">        |       v3|                true|</span>
<span class="sd">        |       v4|                true|</span>
<span class="sd">        +---------+--------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">min_allele_count</span><span class="p">,</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;minAlleleCount&quot;</span><span class="p">)</span>
                <span class="o">&lt;</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="n">min_allele_count_threshold</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="o">&gt;</span> <span class="mi">0</span>
    <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;hasLowMinAlleleCount&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.has_low_min_allele_frequency" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">has_low_min_allele_frequency</span><span class="p">(</span><span class="n">maf</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0001</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.has_low_min_allele_frequency" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Find if variant has a low minor allele frequency in any cohort.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>maf</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Column containing array of structs with <code>cohort</code> and <code>minAlleleFrequency</code> fields.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>threshold</code>
            </td>
            <td>
                  <code><span title="float">float</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Threshold below which the minor allele frequency is considered low. Default is 1e-4.</p>
              </div>
            </td>
            <td>
                  <code>0.0001</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Boolean column indicating if any cohort has a low minor allele frequency.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">maf</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;v1&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.0001</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.0002</span><span class="p">}],</span>
<span class="gp">... </span>        <span class="s2">&quot;v2&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;D&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.15</span><span class="p">}],</span>
<span class="gp">... </span>        <span class="s2">&quot;v3&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.00001</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.2</span><span class="p">}],}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;v1&quot;</span><span class="p">,</span> <span class="n">maf</span><span class="p">[</span><span class="s2">&quot;v1&quot;</span><span class="p">]),</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s2">&quot;v2&quot;</span><span class="p">,</span> <span class="n">maf</span><span class="p">[</span><span class="s2">&quot;v2&quot;</span><span class="p">]),</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s2">&quot;v3&quot;</span><span class="p">,</span> <span class="n">maf</span><span class="p">[</span><span class="s2">&quot;v3&quot;</span><span class="p">]),]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">schema</span> <span class="o">=</span> <span class="s2">&quot;variantId STRING, cohortMinAlleleFrequency ARRAY&lt;STRUCT&lt;cohort: STRING, minAlleleFrequency: DOUBLE&gt;&gt;&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">schema</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+---------+--------------------------+</span>
<span class="go">|variantId|cohortMinAlleleFrequency  |</span>
<span class="go">+---------+--------------------------+</span>
<span class="go">|v1       |[{A, 1.0E-4}, {B, 2.0E-4}]|</span>
<span class="go">|v2       |[{A, NULL}, {D, 0.15}]    |</span>
<span class="go">|v3       |[{A, 1.0E-5}, {B, 0.2}]   |</span>
<span class="go">+---------+--------------------------+</span>
</code></pre></div>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;hasMinAlleleFrequency&quot;</span><span class="p">,</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">has_low_min_allele_frequency</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohortMinAlleleFrequency&quot;</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="s2">&quot;hasMinAlleleFrequency&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+---------+---------------------+</span>
<span class="go">|variantId|hasMinAlleleFrequency|</span>
<span class="go">+---------+---------------------+</span>
<span class="go">|v1       |false                |</span>
<span class="go">|v2       |false                |</span>
<span class="go">|v3       |true                 |</span>
<span class="go">+---------+---------------------+</span>
</code></pre></div>


            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/finngen_meta/summary_statistics.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">has_low_min_allele_frequency</span><span class="p">(</span>
    <span class="bp">cls</span><span class="p">,</span> <span class="n">maf</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1e-4</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find if variant has a low minor allele frequency in any cohort.</span>

<span class="sd">    Args:</span>
<span class="sd">        maf (Column): Column containing array of structs with `cohort` and `minAlleleFrequency` fields.</span>
<span class="sd">        threshold (float): Threshold below which the minor allele frequency is considered low. Default is 1e-4.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Boolean column indicating if any cohort has a low minor allele frequency.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; maf = {&quot;v1&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleFrequency&quot;: 0.0001}, {&quot;cohort&quot;: &quot;B&quot;, &quot;minAlleleFrequency&quot;: 0.0002}],</span>
<span class="sd">        ...         &quot;v2&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleFrequency&quot;: None}, {&quot;cohort&quot;: &quot;D&quot;, &quot;minAlleleFrequency&quot;: 0.15}],</span>
<span class="sd">        ...         &quot;v3&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleFrequency&quot;: 0.00001}, {&quot;cohort&quot;: &quot;B&quot;, &quot;minAlleleFrequency&quot;: 0.2}],}</span>
<span class="sd">        &gt;&gt;&gt; data = [(&quot;v1&quot;, maf[&quot;v1&quot;]),</span>
<span class="sd">        ...         (&quot;v2&quot;, maf[&quot;v2&quot;]),</span>
<span class="sd">        ...         (&quot;v3&quot;, maf[&quot;v3&quot;]),]</span>
<span class="sd">        &gt;&gt;&gt; schema = &quot;variantId STRING, cohortMinAlleleFrequency ARRAY&lt;STRUCT&lt;cohort: STRING, minAlleleFrequency: DOUBLE&gt;&gt;&quot;</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">        &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">        +---------+--------------------------+</span>
<span class="sd">        |variantId|cohortMinAlleleFrequency  |</span>
<span class="sd">        +---------+--------------------------+</span>
<span class="sd">        |v1       |[{A, 1.0E-4}, {B, 2.0E-4}]|</span>
<span class="sd">        |v2       |[{A, NULL}, {D, 0.15}]    |</span>
<span class="sd">        |v3       |[{A, 1.0E-5}, {B, 0.2}]   |</span>
<span class="sd">        +---------+--------------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        &gt;&gt;&gt; df = df.withColumn(&quot;hasMinAlleleFrequency&quot;, FinnGenUkbMvpMetaSummaryStatistics.has_low_min_allele_frequency(f.col(&quot;cohortMinAlleleFrequency&quot;)))</span>
<span class="sd">        &gt;&gt;&gt; df.select(&quot;variantId&quot;, &quot;hasMinAlleleFrequency&quot;).show(truncate=False)</span>
<span class="sd">        +---------+---------------------+</span>
<span class="sd">        |variantId|hasMinAlleleFrequency|</span>
<span class="sd">        +---------+---------------------+</span>
<span class="sd">        |v1       |false                |</span>
<span class="sd">        |v2       |false                |</span>
<span class="sd">        |v3       |true                 |</span>
<span class="sd">        +---------+---------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">non_empty_maf</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
        <span class="n">maf</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="n">n_cohorts_with_maf_below_threshold</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">non_empty_maf</span><span class="p">,</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">)</span>
            <span class="o">&lt;</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="n">threshold</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">DecimalType</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span> <span class="mi">10</span><span class="p">)),</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">n_cohorts_with_maf_below_threshold</span> <span class="o">&gt;</span> <span class="mi">0</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.is_meta_analyzed_variant" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">is_meta_analyzed_variant</span><span class="p">(</span><span class="n">cohorts</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.is_meta_analyzed_variant" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Check if the variant is meta-analyzed (present in more than one biobank).</p>


<details class="note" open>
  <summary>Note</summary>
  <p>if the same biobank has multiple cohorts, it still counts as one biobank.</p>
</details>

<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>cohorts</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array of structs with fields 'biobank'.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Boolean column indicating if the variant is meta-analyzed.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">[([(</span><span class="s2">&quot;FinnGen&quot;</span><span class="p">,</span> <span class="s2">&quot;FinnGen&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;MVP&quot;</span><span class="p">,</span> <span class="s2">&quot;MVP_EUR&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;MVP&quot;</span><span class="p">,</span> <span class="s2">&quot;MVP_AMR&quot;</span><span class="p">)],),</span> <span class="p">([(</span><span class="s2">&quot;MVP&quot;</span><span class="p">,</span> <span class="s2">&quot;MVP_AMR&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;MVP&quot;</span><span class="p">,</span> <span class="s2">&quot;MVP_EUR&quot;</span><span class="p">)],),([(</span><span class="s2">&quot;UKBB&quot;</span><span class="p">,</span> <span class="s2">&quot;UKBB&quot;</span><span class="p">)],)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">schema</span> <span class="o">=</span> <span class="s2">&quot;cohorts ARRAY&lt;STRUCT&lt;biobank: STRING, cohort: STRING&gt;&gt;&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">schema</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+----------------------------------------------------+</span>
<span class="go">|cohorts                                             |</span>
<span class="go">+----------------------------------------------------+</span>
<span class="go">|[{FinnGen, FinnGen}, {MVP, MVP_EUR}, {MVP, MVP_AMR}]|</span>
<span class="go">|[{MVP, MVP_AMR}, {MVP, MVP_EUR}]                    |</span>
<span class="go">|[{UKBB, UKBB}]                                      |</span>
<span class="go">+----------------------------------------------------+</span>
</code></pre></div>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;isMetaAnalyzedVariant&quot;</span><span class="p">,</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">is_meta_analyzed_variant</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohorts&quot;</span><span class="p">)))</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+----------------------------------------------------+---------------------+</span>
<span class="go">|cohorts                                             |isMetaAnalyzedVariant|</span>
<span class="go">+----------------------------------------------------+---------------------+</span>
<span class="go">|[{FinnGen, FinnGen}, {MVP, MVP_EUR}, {MVP, MVP_AMR}]|true                 |</span>
<span class="go">|[{MVP, MVP_AMR}, {MVP, MVP_EUR}]                    |false                |</span>
<span class="go">|[{UKBB, UKBB}]                                      |false                |</span>
<span class="go">+----------------------------------------------------+---------------------+</span>
</code></pre></div>


            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/finngen_meta/summary_statistics.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1185</span>
<span class="normal">1186</span>
<span class="normal">1187</span>
<span class="normal">1188</span>
<span class="normal">1189</span>
<span class="normal">1190</span>
<span class="normal">1191</span>
<span class="normal">1192</span>
<span class="normal">1193</span>
<span class="normal">1194</span>
<span class="normal">1195</span>
<span class="normal">1196</span>
<span class="normal">1197</span>
<span class="normal">1198</span>
<span class="normal">1199</span>
<span class="normal">1200</span>
<span class="normal">1201</span>
<span class="normal">1202</span>
<span class="normal">1203</span>
<span class="normal">1204</span>
<span class="normal">1205</span>
<span class="normal">1206</span>
<span class="normal">1207</span>
<span class="normal">1208</span>
<span class="normal">1209</span>
<span class="normal">1210</span>
<span class="normal">1211</span>
<span class="normal">1212</span>
<span class="normal">1213</span>
<span class="normal">1214</span>
<span class="normal">1215</span>
<span class="normal">1216</span>
<span class="normal">1217</span>
<span class="normal">1218</span>
<span class="normal">1219</span>
<span class="normal">1220</span>
<span class="normal">1221</span>
<span class="normal">1222</span>
<span class="normal">1223</span>
<span class="normal">1224</span>
<span class="normal">1225</span>
<span class="normal">1226</span>
<span class="normal">1227</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">is_meta_analyzed_variant</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">cohorts</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check if the variant is meta-analyzed (present in more than one biobank).</span>

<span class="sd">    Note:</span>
<span class="sd">        if the same biobank has multiple cohorts, it still counts as one biobank.</span>

<span class="sd">    Args:</span>
<span class="sd">        cohorts (Column): Array of structs with fields &#39;biobank&#39;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Boolean column indicating if the variant is meta-analyzed.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; data = [([(&quot;FinnGen&quot;, &quot;FinnGen&quot;), (&quot;MVP&quot;, &quot;MVP_EUR&quot;), (&quot;MVP&quot;, &quot;MVP_AMR&quot;)],), ([(&quot;MVP&quot;, &quot;MVP_AMR&quot;), (&quot;MVP&quot;, &quot;MVP_EUR&quot;)],),([(&quot;UKBB&quot;, &quot;UKBB&quot;)],)]</span>
<span class="sd">        &gt;&gt;&gt; schema = &quot;cohorts ARRAY&lt;STRUCT&lt;biobank: STRING, cohort: STRING&gt;&gt;&quot;</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">        &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">        +----------------------------------------------------+</span>
<span class="sd">        |cohorts                                             |</span>
<span class="sd">        +----------------------------------------------------+</span>
<span class="sd">        |[{FinnGen, FinnGen}, {MVP, MVP_EUR}, {MVP, MVP_AMR}]|</span>
<span class="sd">        |[{MVP, MVP_AMR}, {MVP, MVP_EUR}]                    |</span>
<span class="sd">        |[{UKBB, UKBB}]                                      |</span>
<span class="sd">        +----------------------------------------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        &gt;&gt;&gt; df.withColumn(&quot;isMetaAnalyzedVariant&quot;, FinnGenUkbMvpMetaSummaryStatistics.is_meta_analyzed_variant(f.col(&quot;cohorts&quot;))).show(truncate=False)</span>
<span class="sd">        +----------------------------------------------------+---------------------+</span>
<span class="sd">        |cohorts                                             |isMetaAnalyzedVariant|</span>
<span class="sd">        +----------------------------------------------------+---------------------+</span>
<span class="sd">        |[{FinnGen, FinnGen}, {MVP, MVP_EUR}, {MVP, MVP_AMR}]|true                 |</span>
<span class="sd">        |[{MVP, MVP_AMR}, {MVP, MVP_EUR}]                    |false                |</span>
<span class="sd">        |[{UKBB, UKBB}]                                      |false                |</span>
<span class="sd">        +----------------------------------------------------+---------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>


<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">n_biobanks</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">array_distinct</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">cohorts</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;biobank&quot;</span><span class="p">))),</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">n_biobanks</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;isMetaAnalyzedVariant&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.min_allele_count" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">min_allele_count</span><span class="p">(</span><span class="n">cohort_min_allele_frequency</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">n_samples_per_cohort</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.min_allele_count" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Minor Allele Count (MAC) per cohort.</p>


<details class="note" open>
  <summary>Note</summary>
  <p>If a cohort either do not have the maf or nCases, it will be dropped from the resulting MAC array.</p>
</details>

<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>cohort_min_allele_frequency</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Column containing array of structs with <code>cohort</code> and <code>minAlleleFrequency</code> fields.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_samples_per_cohort</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Column containing array of structs with <code>cohort</code> and <code>nSamples</code> fields.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Column containing array of structs with <code>cohort</code> and <code>minAlleleCount</code> fields.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">maf</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;v1&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.2</span><span class="p">}],</span>
<span class="gp">... </span>        <span class="s2">&quot;v2&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.05</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;D&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.15</span><span class="p">}],</span>
<span class="gp">... </span>        <span class="s2">&quot;v3&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.01</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.02</span><span class="p">}],}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n_samples</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;v1&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;nSamples&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;nSamples&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">}],</span>
<span class="gp">... </span>           <span class="s2">&quot;v2&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;nSamples&quot;</span><span class="p">:</span> <span class="mi">150</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;nSamples&quot;</span><span class="p">:</span> <span class="mi">250</span><span class="p">}],</span>
<span class="gp">... </span>           <span class="s2">&quot;v3&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;nSamples&quot;</span><span class="p">:</span> <span class="mi">50</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;D&quot;</span><span class="p">,</span> <span class="s2">&quot;nSamples&quot;</span><span class="p">:</span> <span class="mi">80</span><span class="p">}],}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;v1&quot;</span><span class="p">,</span> <span class="n">maf</span><span class="p">[</span><span class="s2">&quot;v1&quot;</span><span class="p">],</span> <span class="n">n_samples</span><span class="p">[</span><span class="s2">&quot;v1&quot;</span><span class="p">]),</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s2">&quot;v2&quot;</span><span class="p">,</span> <span class="n">maf</span><span class="p">[</span><span class="s2">&quot;v2&quot;</span><span class="p">],</span> <span class="n">n_samples</span><span class="p">[</span><span class="s2">&quot;v2&quot;</span><span class="p">]),</span>
<span class="gp">... </span>        <span class="p">(</span><span class="s2">&quot;v3&quot;</span><span class="p">,</span> <span class="n">maf</span><span class="p">[</span><span class="s2">&quot;v3&quot;</span><span class="p">],</span> <span class="n">n_samples</span><span class="p">[</span><span class="s2">&quot;v3&quot;</span><span class="p">]),]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">schema</span> <span class="o">=</span> <span class="s2">&quot;variantId STRING, cohortMinAlleleFrequency ARRAY&lt;STRUCT&lt;cohort: STRING, minAlleleFrequency: DOUBLE&gt;&gt;, nSamplesPerCohort ARRAY&lt;STRUCT&lt;cohort: STRING, nSamples: INT&gt;&gt;&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">schema</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+---------+------------------------+--------------------+</span>
<span class="go">|variantId|cohortMinAlleleFrequency|nSamplesPerCohort   |</span>
<span class="go">+---------+------------------------+--------------------+</span>
<span class="go">|v1       |[{A, 0.1}, {B, 0.2}]    |[{A, 100}, {B, 200}]|</span>
<span class="go">|v2       |[{A, 0.05}, {D, 0.15}]  |[{A, 150}, {C, 250}]|</span>
<span class="go">|v3       |[{A, 0.01}, {B, 0.02}]  |[{C, 50}, {D, 80}]  |</span>
<span class="go">+---------+------------------------+--------------------+</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;cohortMinAlleleCount&quot;</span><span class="p">,</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">min_allele_count</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;cohortMinAlleleFrequency&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;nSamplesPerCohort&quot;</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="s2">&quot;cohortMinAlleleCount&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+---------+--------------------+</span>
<span class="go">|variantId|cohortMinAlleleCount|</span>
<span class="go">+---------+--------------------+</span>
<span class="go">|v1       |[{A, 20}, {B, 80}]  |</span>
<span class="go">|v2       |[{A, 15}]           |</span>
<span class="go">|v3       |[]                  |</span>
<span class="go">+---------+--------------------+</span>
</code></pre></div>


            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/finngen_meta/summary_statistics.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span>
<span class="normal">917</span>
<span class="normal">918</span>
<span class="normal">919</span>
<span class="normal">920</span>
<span class="normal">921</span>
<span class="normal">922</span>
<span class="normal">923</span>
<span class="normal">924</span>
<span class="normal">925</span>
<span class="normal">926</span>
<span class="normal">927</span>
<span class="normal">928</span>
<span class="normal">929</span>
<span class="normal">930</span>
<span class="normal">931</span>
<span class="normal">932</span>
<span class="normal">933</span>
<span class="normal">934</span>
<span class="normal">935</span>
<span class="normal">936</span>
<span class="normal">937</span>
<span class="normal">938</span>
<span class="normal">939</span>
<span class="normal">940</span>
<span class="normal">941</span>
<span class="normal">942</span>
<span class="normal">943</span>
<span class="normal">944</span>
<span class="normal">945</span>
<span class="normal">946</span>
<span class="normal">947</span>
<span class="normal">948</span>
<span class="normal">949</span>
<span class="normal">950</span>
<span class="normal">951</span>
<span class="normal">952</span>
<span class="normal">953</span>
<span class="normal">954</span>
<span class="normal">955</span>
<span class="normal">956</span>
<span class="normal">957</span>
<span class="normal">958</span>
<span class="normal">959</span>
<span class="normal">960</span>
<span class="normal">961</span>
<span class="normal">962</span>
<span class="normal">963</span>
<span class="normal">964</span>
<span class="normal">965</span>
<span class="normal">966</span>
<span class="normal">967</span>
<span class="normal">968</span>
<span class="normal">969</span>
<span class="normal">970</span>
<span class="normal">971</span>
<span class="normal">972</span>
<span class="normal">973</span>
<span class="normal">974</span>
<span class="normal">975</span>
<span class="normal">976</span>
<span class="normal">977</span>
<span class="normal">978</span>
<span class="normal">979</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">min_allele_count</span><span class="p">(</span>
    <span class="bp">cls</span><span class="p">,</span> <span class="n">cohort_min_allele_frequency</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">n_samples_per_cohort</span><span class="p">:</span> <span class="n">Column</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Minor Allele Count (MAC) per cohort.</span>

<span class="sd">    Note:</span>
<span class="sd">        If a cohort either do not have the maf or nCases, it will be dropped from the resulting MAC array.</span>

<span class="sd">    Args:</span>
<span class="sd">        cohort_min_allele_frequency (Column): Column containing array of structs with `cohort` and `minAlleleFrequency` fields.</span>
<span class="sd">        n_samples_per_cohort (Column): Column containing array of structs with `cohort` and `nSamples` fields.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Column containing array of structs with `cohort` and `minAlleleCount` fields.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; maf = {&quot;v1&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleFrequency&quot;: 0.1}, {&quot;cohort&quot;: &quot;B&quot;, &quot;minAlleleFrequency&quot;: 0.2}],</span>
<span class="sd">        ...         &quot;v2&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleFrequency&quot;: 0.05}, {&quot;cohort&quot;: &quot;D&quot;, &quot;minAlleleFrequency&quot;: 0.15}],</span>
<span class="sd">        ...         &quot;v3&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;minAlleleFrequency&quot;: 0.01}, {&quot;cohort&quot;: &quot;B&quot;, &quot;minAlleleFrequency&quot;: 0.02}],}</span>
<span class="sd">        &gt;&gt;&gt; n_samples = {&quot;v1&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;nSamples&quot;: 100}, {&quot;cohort&quot;: &quot;B&quot;, &quot;nSamples&quot;: 200}],</span>
<span class="sd">        ...            &quot;v2&quot;: [{&quot;cohort&quot;: &quot;A&quot;, &quot;nSamples&quot;: 150}, {&quot;cohort&quot;: &quot;C&quot;, &quot;nSamples&quot;: 250}],</span>
<span class="sd">        ...            &quot;v3&quot;: [{&quot;cohort&quot;: &quot;C&quot;, &quot;nSamples&quot;: 50}, {&quot;cohort&quot;: &quot;D&quot;, &quot;nSamples&quot;: 80}],}</span>
<span class="sd">        &gt;&gt;&gt; data = [(&quot;v1&quot;, maf[&quot;v1&quot;], n_samples[&quot;v1&quot;]),</span>
<span class="sd">        ...         (&quot;v2&quot;, maf[&quot;v2&quot;], n_samples[&quot;v2&quot;]),</span>
<span class="sd">        ...         (&quot;v3&quot;, maf[&quot;v3&quot;], n_samples[&quot;v3&quot;]),]</span>
<span class="sd">        &gt;&gt;&gt; schema = &quot;variantId STRING, cohortMinAlleleFrequency ARRAY&lt;STRUCT&lt;cohort: STRING, minAlleleFrequency: DOUBLE&gt;&gt;, nSamplesPerCohort ARRAY&lt;STRUCT&lt;cohort: STRING, nSamples: INT&gt;&gt;&quot;</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">        &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">        +---------+------------------------+--------------------+</span>
<span class="sd">        |variantId|cohortMinAlleleFrequency|nSamplesPerCohort   |</span>
<span class="sd">        +---------+------------------------+--------------------+</span>
<span class="sd">        |v1       |[{A, 0.1}, {B, 0.2}]    |[{A, 100}, {B, 200}]|</span>
<span class="sd">        |v2       |[{A, 0.05}, {D, 0.15}]  |[{A, 150}, {C, 250}]|</span>
<span class="sd">        |v3       |[{A, 0.01}, {B, 0.02}]  |[{C, 50}, {D, 80}]  |</span>
<span class="sd">        +---------+------------------------+--------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">        &gt;&gt;&gt; df = df.withColumn(&quot;cohortMinAlleleCount&quot;, FinnGenUkbMvpMetaSummaryStatistics.min_allele_count(f.col(&quot;cohortMinAlleleFrequency&quot;), f.col(&quot;nSamplesPerCohort&quot;)))</span>
<span class="sd">        &gt;&gt;&gt; df.select(&quot;variantId&quot;, &quot;cohortMinAlleleCount&quot;).show(truncate=False)</span>
<span class="sd">        +---------+--------------------+</span>
<span class="sd">        |variantId|cohortMinAlleleCount|</span>
<span class="sd">        +---------+--------------------+</span>
<span class="sd">        |v1       |[{A, 20}, {B, 80}]  |</span>
<span class="sd">        |v2       |[{A, 15}]           |</span>
<span class="sd">        |v3       |[]                  |</span>
<span class="sd">        +---------+--------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
        <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">cohort_min_allele_frequency</span><span class="p">,</span>
            <span class="k">lambda</span> <span class="n">left</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span>
                <span class="n">n_samples_per_cohort</span><span class="p">,</span>
                <span class="k">lambda</span> <span class="n">right</span><span class="p">:</span> <span class="n">right</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">left</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">),</span>
        <span class="k">lambda</span> <span class="n">left</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">left</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
            <span class="n">mac</span><span class="p">(</span>
                <span class="n">left</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">n_samples_per_cohort</span><span class="p">,</span>
                    <span class="k">lambda</span> <span class="n">right</span><span class="p">:</span> <span class="n">right</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">)</span>
                    <span class="o">==</span> <span class="n">left</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="o">.</span><span class="n">getItem</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
                <span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;nSamples&quot;</span><span class="p">),</span>
            <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;minAlleleCount&quot;</span><span class="p">),</span>
        <span class="p">),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.min_allele_frequency" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">min_allele_frequency</span><span class="p">(</span><span class="n">allele_freq</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.min_allele_frequency" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Minor Allele Frequency (MAF) per cohort.</p>


<details class="note" open>
  <summary>Note</summary>
  <p>The resulting value is of DecimalType(11, 10) to ensure precision for low frequency variants.</p>
</details>

<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>allele_freq</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Column containing array of structs with <code>cohort</code> and <code>alleleFrequency</code> fields.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Column containing array of structs with <code>cohort</code> and <code>minAlleleFrequency</code> fields.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;v1&quot;</span><span class="p">,</span> <span class="p">[{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;alleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;cohort&quot;</span><span class="p">:</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;alleleFrequency&quot;</span><span class="p">:</span> <span class="mf">0.7</span><span class="p">}]),]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">schema</span> <span class="o">=</span> <span class="s2">&quot;variantId STRING, alleleFrequencies ARRAY&lt;STRUCT&lt;cohort: STRING, alleleFrequency: DOUBLE&gt;&gt;&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">schema</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+---------+--------------------+</span>
<span class="go">|variantId|alleleFrequencies   |</span>
<span class="go">+---------+--------------------+</span>
<span class="go">|v1       |[{A, 0.1}, {B, 0.7}]|</span>
<span class="go">+---------+--------------------+</span>
</code></pre></div>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;cohortMinAlleleFrequency&quot;</span><span class="p">,</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">min_allele_frequency</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;alleleFrequencies&quot;</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+---------+--------------------+--------------------------------------+</span>
<span class="go">|variantId|alleleFrequencies   |cohortMinAlleleFrequency              |</span>
<span class="go">+---------+--------------------+--------------------------------------+</span>
<span class="go">|v1       |[{A, 0.1}, {B, 0.7}]|[{A, 0.1000000000}, {B, 0.3000000000}]|</span>
<span class="go">+---------+--------------------+--------------------------------------+</span>
</code></pre></div>


            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/finngen_meta/summary_statistics.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">min_allele_frequency</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">allele_freq</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Minor Allele Frequency (MAF) per cohort.</span>

<span class="sd">    Note:</span>
<span class="sd">        The resulting value is of DecimalType(11, 10) to ensure precision for low frequency variants.</span>

<span class="sd">    Args:</span>
<span class="sd">        allele_freq (Column): Column containing array of structs with `cohort` and `alleleFrequency` fields.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Column containing array of structs with `cohort` and `minAlleleFrequency` fields.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; data = [(&quot;v1&quot;, [{&quot;cohort&quot;: &quot;A&quot;, &quot;alleleFrequency&quot;: 0.1}, {&quot;cohort&quot;: &quot;B&quot;, &quot;alleleFrequency&quot;: 0.7}]),]</span>
<span class="sd">        &gt;&gt;&gt; schema = &quot;variantId STRING, alleleFrequencies ARRAY&lt;STRUCT&lt;cohort: STRING, alleleFrequency: DOUBLE&gt;&gt;&quot;</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">        &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">        +---------+--------------------+</span>
<span class="sd">        |variantId|alleleFrequencies   |</span>
<span class="sd">        +---------+--------------------+</span>
<span class="sd">        |v1       |[{A, 0.1}, {B, 0.7}]|</span>
<span class="sd">        +---------+--------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        &gt;&gt;&gt; df = df.withColumn(&quot;cohortMinAlleleFrequency&quot;, FinnGenUkbMvpMetaSummaryStatistics.min_allele_frequency(f.col(&quot;alleleFrequencies&quot;)))</span>
<span class="sd">        &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">        +---------+--------------------+--------------------------------------+</span>
<span class="sd">        |variantId|alleleFrequencies   |cohortMinAlleleFrequency              |</span>
<span class="sd">        +---------+--------------------+--------------------------------------+</span>
<span class="sd">        |v1       |[{A, 0.1}, {B, 0.7}]|[{A, 0.1000000000}, {B, 0.3000000000}]|</span>
<span class="sd">        +---------+--------------------+--------------------------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
        <span class="n">allele_freq</span><span class="p">,</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;cohort&quot;</span><span class="p">),</span>
            <span class="n">maf</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">getField</span><span class="p">(</span><span class="s2">&quot;alleleFrequency&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;minAlleleFrequency&quot;</span><span class="p">),</span>
        <span class="p">),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.normalize_af" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">normalize_af</span><span class="p">(</span><span class="n">af</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">flip</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.datasource.finngen_meta.summary_statistics.FinnGenUkbMvpMetaSummaryStatistics.normalize_af" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Normalize allele frequency based on variant direction.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>af</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Allele frequency column.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>flip</code>
            </td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Direction column indicating if the allele needs to be flipped.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Normalized allele frequency.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;v1&quot;</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>      <span class="p">(</span><span class="s2">&quot;v2&quot;</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>      <span class="p">(</span><span class="s2">&quot;v3&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>      <span class="p">(</span><span class="s2">&quot;V4&quot;</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="kc">None</span><span class="p">),]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">schema</span> <span class="o">=</span> <span class="s2">&quot;variantId STRING, af DOUBLE, flip INT&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">schema</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+---------+----+----+</span>
<span class="go">|variantId|af  |flip|</span>
<span class="go">+---------+----+----+</span>
<span class="go">|v1       |0.1 |1   |</span>
<span class="go">|v2       |0.2 |-1  |</span>
<span class="go">|v3       |NULL|-1  |</span>
<span class="go">|V4       |0.1 |NULL|</span>
<span class="go">+---------+----+----+</span>
</code></pre></div>
    <div class="language-pycon highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;normalizedAf&quot;</span><span class="p">,</span> <span class="n">FinnGenUkbMvpMetaSummaryStatistics</span><span class="o">.</span><span class="n">normalize_af</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;af&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;flip&quot;</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="s2">&quot;normalizedAf&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">truncate</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">+---------+------------+</span>
<span class="go">|variantId|normalizedAf|</span>
<span class="go">+---------+------------+</span>
<span class="go">|v1       |0.1         |</span>
<span class="go">|v2       |0.8         |</span>
<span class="go">|v3       |NULL        |</span>
<span class="go">|V4       |0.1         |</span>
<span class="go">+---------+------------+</span>
</code></pre></div>


            <details class="quote">
              <summary>Source code in <code>src/gentropy/datasource/finngen_meta/summary_statistics.py</code></summary>
              <div class="language-python highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span><span class="w"> </span><span class="nf">normalize_af</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">af</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">flip</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Normalize allele frequency based on variant direction.</span>

<span class="sd">    Args:</span>
<span class="sd">        af (Column): Allele frequency column.</span>
<span class="sd">        flip (Column): Direction column indicating if the allele needs to be flipped.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Normalized allele frequency.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; data = [(&quot;v1&quot;, 0.1, 1),</span>
<span class="sd">        ...       (&quot;v2&quot;, 0.2, -1),</span>
<span class="sd">        ...       (&quot;v3&quot;, None, -1),</span>
<span class="sd">        ...       (&quot;V4&quot;, 0.1, None),]</span>
<span class="sd">        &gt;&gt;&gt; schema = &quot;variantId STRING, af DOUBLE, flip INT&quot;</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame(data, schema)</span>
<span class="sd">        &gt;&gt;&gt; df.show(truncate=False)</span>
<span class="sd">        +---------+----+----+</span>
<span class="sd">        |variantId|af  |flip|</span>
<span class="sd">        +---------+----+----+</span>
<span class="sd">        |v1       |0.1 |1   |</span>
<span class="sd">        |v2       |0.2 |-1  |</span>
<span class="sd">        |v3       |NULL|-1  |</span>
<span class="sd">        |V4       |0.1 |NULL|</span>
<span class="sd">        +---------+----+----+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>

<span class="sd">        &gt;&gt;&gt; df = df.withColumn(&quot;normalizedAf&quot;, FinnGenUkbMvpMetaSummaryStatistics.normalize_af(f.col(&quot;af&quot;), f.col(&quot;flip&quot;)))</span>
<span class="sd">        &gt;&gt;&gt; df.select(&quot;variantId&quot;, &quot;normalizedAf&quot;).show(truncate=False)</span>
<span class="sd">        +---------+------------+</span>
<span class="sd">        |variantId|normalizedAf|</span>
<span class="sd">        +---------+------------+</span>
<span class="sd">        |v1       |0.1         |</span>
<span class="sd">        |v2       |0.8         |</span>
<span class="sd">        |v3       |NULL        |</span>
<span class="sd">        |V4       |0.1         |</span>
<span class="sd">        +---------+------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">((</span><span class="n">flip</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">af</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">1.0</span><span class="p">)</span> <span class="o">-</span> <span class="n">af</span><span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">af</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>
<hr />
<div class="md-source-file">
  <small>
    <ul
      class="metadata page-metadata"
      data-bi-name="page info"
      lang="en-us"
      dir="ltr"
    >
      <!-- mkdocs-git-revision-date-plugin -->
      
      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3h-2Z"/></svg>
        </span>
        <span class="label"> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="October 27, 2025 18:26:57 UTC"><span class="timeago" datetime="2025-10-27T18:26:57+00:00" locale="en"></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="October 27, 2025 18:26:57 UTC">2025-10-27</span> </span>
      </li>
       
      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg>
        </span>
        <span class="label"> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago" title="October 27, 2025 18:26:57 UTC"><span class="timeago" datetime="2025-10-27T18:26:57+00:00" locale="en"></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date" title="October 27, 2025 18:26:57 UTC">2025-10-27</span> </span>
      </li>
      

      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2A10 10 0 0 0 2 12c0 4.42 2.87 8.17 6.84 9.5.5.08.66-.23.66-.5v-1.69c-2.77.6-3.36-1.34-3.36-1.34-.46-1.16-1.11-1.47-1.11-1.47-.91-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.87 1.52 2.34 1.07 2.91.83.09-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.92 0-1.11.38-2 1.03-2.71-.1-.25-.45-1.29.1-2.64 0 0 .84-.27 2.75 1.02.79-.22 1.65-.33 2.5-.33.85 0 1.71.11 2.5.33 1.91-1.29 2.75-1.02 2.75-1.02.55 1.35.2 2.39.1 2.64.65.71 1.03 1.6 1.03 2.71 0 3.82-2.34 4.66-4.57 4.91.36.31.69.92.69 1.85V21c0 .27.16.59.67.5C19.14 20.16 22 16.42 22 12A10 10 0 0 0 12 2Z"/></svg>
        </span>
        <span class="label"> Contributors </span>
      </li>
      <li class="contributors-holder">
        <ul class="contributors" data-bi-name="contributors"></ul>
      </li>
    </ul>
  </small>
</div>




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/gentropy" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../../..", "features": ["navigation.instant", "navigation.top", "navigation.tabs", "navigation.tabs.sticky", "content.code.annotate", "search.suggest", "search.highlight"], "search": "../../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../../assets/javascripts/bundle.3220b9d7.min.js"></script>
      
        <script src="../../../../js/timeago.min.js"></script>
      
        <script src="../../../../js/timeago_mkdocs_material.js"></script>
      
        <script src="../../../../assets/javascripts/extra.js"></script>
      
    
  </body>
</html>